{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Phenotype acronym extraction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import sys\n",
    "import cPickle\n",
    "import numpy as np\n",
    "\n",
    "# import snorkel and gwasdb\n",
    "sys.path.append('../snorkel')\n",
    "sys.path.append('../src')\n",
    "sys.path.append('../src/crawler')\n",
    "\n",
    "# set up paths\n",
    "abstract_dir = '../data/db/papers'\n",
    "\n",
    "# set up matplotlib\n",
    "import matplotlib\n",
    "%matplotlib inline\n",
    "matplotlib.rcParams['figure.figsize'] = (12,4)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load corpus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from snorkel.parser import XMLDocParser\n",
    "from extractor.parser import UnicodeXMLTableDocParser\n",
    "\n",
    "xml_parser = UnicodeXMLTableDocParser(\n",
    "    path=abstract_dir,\n",
    "    doc='./*',\n",
    "    text='.//table',\n",
    "    id='.//article-id[@pub-id-type=\"pmid\"]/text()',\n",
    "    keep_xml_tree=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 4min 31s, sys: 13.4 s, total: 4min 45s\n",
      "Wall time: 7min 1s\n"
     ]
    }
   ],
   "source": [
    "from snorkel.parser import CorpusParser, OmniParser\n",
    "import cPickle\n",
    "\n",
    "table_parser = OmniParser()\n",
    "\n",
    "corpus_name = 'gwas-table-corpus.pkl'\n",
    "try:\n",
    "    with open(corpus_name,\"r\") as pkl:\n",
    "        corpus = cPickle.load(pkl)\n",
    "except:\n",
    "    cp = CorpusParser(xml_parser, table_parser, max_docs=100)\n",
    "    %time corpus = cp.parse_corpus(name='GWAS Corpus')\n",
    "    # pickling currently doesn't work...\n",
    "#     with open(corpus_name,\"w\") as pkl:\n",
    "#         corpus = cPickle.dump(corpus, pkl)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Candidate extraction"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### From tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from snorkel.matchers import DictionaryMatch, Union, CellDictNameMatcher\n",
    "from snorkel.candidates import EntityExtractor\n",
    "from snorkel.candidates import TableNgrams, CellSpace\n",
    "\n",
    "# Define a candidate space\n",
    "cells = CellSpace()\n",
    "\n",
    "# Create a list of possible words that could denote phenotypes\n",
    "acro_words = ['abbreviation', 'acronym', 'phenotype']\n",
    "phen_words = ['trait', 'phenotype', 'description']\n",
    "\n",
    "# Define matchers\n",
    "phen_matcher = CellDictNameMatcher(axis='col', d=phen_words, n_max=3, ignore_case=True)\n",
    "acro_matcher = CellDictNameMatcher(axis='col', d=acro_words, n_max=3, ignore_case=True)\n",
    "\n",
    "phen_extractor = EntityExtractor(cells, phen_matcher)\n",
    "acro_extractor = EntityExtractor(cells, acro_matcher)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from snorkel.candidates import AlignedTableRelationExtractor\n",
    "relation_extractor = AlignedTableRelationExtractor(acro_extractor, phen_extractor, axis='row', induced=True)\n",
    "tables = corpus.get_tables()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 29min 20s, sys: 18.9 s, total: 29min 39s\n",
      "Wall time: 30min 8s\n",
      "267 relations extracted, e.g.\n",
      "SpanPair(Span(\"Candidate gene\", context=None, chars=[0,13], words=[0,1]), Span(\"PHENOTYPE\", context=None, chars=[0,8], words=[0,0]))\n",
      "SpanPair(Span(\"PHENOTYPE\", context=None, chars=[0,8], words=[0,0]), Span(\"Candidate gene\", context=None, chars=[0,13], words=[0,1]))\n",
      "SpanPair(Span(\"CST3\", context=None, chars=[0,3], words=[0,0]), Span(\"CysC\", context=None, chars=[0,3], words=[0,0]))\n",
      "SpanPair(Span(\"CysC\", context=None, chars=[0,3], words=[0,0]), Span(\"CST3\", context=None, chars=[0,3], words=[0,0]))\n",
      "SpanPair(Span(\"CST3\", context=None, chars=[0,3], words=[0,0]), Span(\"CysC\", context=None, chars=[0,3], words=[0,0]))\n"
     ]
    }
   ],
   "source": [
    "%time candidates = relation_extractor.extract(tables, name='all')\n",
    "table_c = candidates\n",
    "print \"%s relations extracted, e.g.\" % len(table_c)\n",
    "for cand in table_c[:5]: \n",
    "    print cand"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### From text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 839 ms, sys: 59 ms, total: 898 ms\n",
      "Wall time: 10.7 s\n"
     ]
    }
   ],
   "source": [
    "from snorkel.parser import SentenceParser, CorpusParser\n",
    "from extractor.parser import UnicodeXMLDocParser, GWASXMLDocParser\n",
    "\n",
    "xml_parser = GWASXMLDocParser(\n",
    "    path=abstract_dir,\n",
    "    doc='./*',\n",
    "    title='.//front//article-title//text()',\n",
    "    abstract='.//abstract//p//text()',\n",
    "    n_par=5,\n",
    "    id='.//article-id[@pub-id-type=\"pmid\"]/text()',\n",
    "    keep_xml_tree=True)\n",
    "\n",
    "sent_parser = SentenceParser()\n",
    "cp = CorpusParser(xml_parser, sent_parser, max_docs=15)\n",
    "%time text_corpus = cp.parse_corpus(name='GWAS Text Corpus')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from snorkel.candidates import Ngrams, CellSpace, TableNgrams\n",
    "from snorkel.matchers import RegexMatchSpan\n",
    "from snorkel.candidates import EntityExtractor, RelationExtractor, UnionExtractor\n",
    "\n",
    "# Define a candidate space\n",
    "ngrams8 = Ngrams(n_max=8)\n",
    "ngrams3 = Ngrams(n_max=3)\n",
    "cells = CellSpace()\n",
    "table_ngrams3 = TableNgrams(n_max=3)\n",
    "\n",
    "# Define matchers\n",
    "phen_matcher = RegexMatchSpan(rgx=r'.+ \\([a-zA-Z0-9_-]{1,10}[\\);]')\n",
    "acro_matcher = RegexMatchSpan(rgx=r'\\([a-zA-Z0-9_-]{1,10}[\\);]')\n",
    "\n",
    "# Extractors\n",
    "phen_txt_ngram_extractor = EntityExtractor(ngrams8, phen_matcher)\n",
    "phen_txt_cells_extractor = EntityExtractor(cells, phen_matcher)\n",
    "acro_txt_ngram_extractor = EntityExtractor(ngrams3, acro_matcher)\n",
    "acro_txt_cells_extractor = EntityExtractor(table_ngrams3, acro_matcher)\n",
    "\n",
    "# Filtering functions\n",
    "def overlap_filter_fn(span0, span1):\n",
    "    if hasattr(span0.context, 'cell') and hasattr(span1.context, 'cell'):\n",
    "        if span0.context.cell != span1.context.cell: return False\n",
    "    if len(span1.get_span().split()) >= 15: return False\n",
    "    start0, end0 = span0.char_start, span0.char_end\n",
    "    start1, end1 = span1.char_start, span1.char_end\n",
    "    return True if start1 <= start0 <= end0 <= end1 else False\n",
    "\n",
    "# Relation extractor\n",
    "txt_tab_ngram_extractor = RelationExtractor(acro_txt_ngram_extractor, phen_txt_ngram_extractor, filter_fn=overlap_filter_fn)\n",
    "txt_tab_cells_extractor = RelationExtractor(acro_txt_cells_extractor, phen_txt_cells_extractor, filter_fn=overlap_filter_fn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Extract acroynms from tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 3.4 s, sys: 262 ms, total: 3.67 s\n",
      "Wall time: 3.76 s\n",
      "1305 candidates extracted from text in tables\n"
     ]
    }
   ],
   "source": [
    "%time txt_tab_c = txt_tab_ngram_extractor.extract(corpus.get_phrases(), name='all')\n",
    "print len(txt_tab_c), 'candidates extracted from text in tables'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Extract acronyms from full table cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 5.37 s, sys: 119 ms, total: 5.49 s\n",
      "Wall time: 5.42 s\n",
      "761 candidates extracted from text in full table cells\n"
     ]
    }
   ],
   "source": [
    "%time txt_cel_c = txt_tab_cells_extractor.extract(corpus.get_tables(), name='all')\n",
    "print len(txt_cel_c), 'candidates extracted from text in full table cells'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Extract acroynms from abstracts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 2.05 s, sys: 31.2 ms, total: 2.09 s\n",
      "Wall time: 2.09 s\n",
      "884 candidates extracted from text in abstracts\n"
     ]
    }
   ],
   "source": [
    "%time txt_txt_c = txt_tab_ngram_extractor.extract(text_corpus.get_sentences(), name='all')\n",
    "print len(txt_txt_c), 'candidates extracted from text in abstracts'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Learning the correctness of relations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2950\n"
     ]
    }
   ],
   "source": [
    "from snorkel.candidates import UnionExtractor\n",
    "\n",
    "contexts = [corpus.get_tables(), corpus.get_phrases(), text_corpus.get_sentences()]\n",
    "extractors = [txt_tab_cells_extractor, txt_tab_ngram_extractor, txt_tab_ngram_extractor]\n",
    "joint_extractor = UnionExtractor(extractor_list=extractors, context_list=contexts)\n",
    "all_c = joint_extractor.union()\n",
    "print len(all_c)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Creating a gold set"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To create a gold set, we save all extracted relations into a csv file. We annotate it manually, and save the result to a second file. It contains pairs of phenotype and rsid strings; if that file exists, we take these as gold truth."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 456,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Physical Position (bp)†\n",
      "Physical Position (bp)\n",
      "('17903294', u'Physical Position (bp)', u'(bp)') 0\n",
      "\n",
      "Physical Position (bp)†\n",
      "Physical Position (bp)\n",
      "('17903294', u'Physical Position (bp)', u'(bp)') 0\n",
      "\n",
      "GEE p-val (Rank)††\n",
      "GEE p-val (Rank)\n",
      "('17903294', u'GEE p-val (Rank)', u'(Rank)') 0\n",
      "\n",
      "Physical Position (bp)†\n",
      "Physical Position (bp)\n",
      "('17903294', u'Physical Position (bp)', u'(bp)') 0\n",
      "\n",
      "Physical Position (bp)†\n",
      "Physical Position (bp)\n",
      "('17903294', u'Physical Position (bp)', u'(bp)') 0\n",
      "\n",
      "GEE Pval (Rank)††\n",
      "GEE Pval (Rank)\n",
      "('17903294', u'GEE Pval (Rank)', u'(Rank)') 0\n",
      "\n",
      "Physical Position (bp)†\n",
      "Physical Position (bp)\n",
      "('17903294', u'Physical Position (bp)', u'(bp)') 0\n",
      "\n",
      "Physical Position (bp)†\n",
      "Physical Position (bp)\n",
      "('17903294', u'Physical Position (bp)', u'(bp)') 0\n",
      "\n",
      "Total Cerebral Brain Volume (ATCBV)\n",
      "Total Cerebral Brain Volume (ATCBV)\n",
      "('17903297', u'Total Cerebral Brain Volume (ATCBV)', u'(ATCBV)') 1\n",
      "\n",
      "Factor 1:Verbal Memory (F1)\n",
      "Factor 1:Verbal Memory (F1)\n",
      "('17903297', u'Factor 1:Verbal Memory (F1)', u'(F1)') 1\n",
      "\n",
      "Factor 1:Verbal Memory (F1)\n",
      "Factor 1:Verbal Memory (F1)\n",
      "('17903297', u'Factor 1:Verbal Memory (F1)', u'(F1)') 1\n",
      "\n",
      "Factor 2:Visual Memory and Organization (F2)\n",
      "Factor 2:Visual Memory and Organization (F2)\n",
      "('17903297', u'Factor 2:Visual Memory and Organization (F2)', u'(F2)') 1\n",
      "\n",
      "Factor 2:Visual Memory and Organization (F2)\n",
      "Factor 2:Visual Memory and Organization (F2)\n",
      "('17903297', u'Factor 2:Visual Memory and Organization (F2)', u'(F2)') 1\n",
      "\n",
      "Factor 3: Measure of attention and executive function-Trails A and B (F3)\n",
      "Factor 3: Measure of attention and executive function-Trails A and B (F3)\n",
      "('17903297', u'Factor 3: Measure of attention and executive function-Trails A and B (F3)', u'(F3)') 1\n",
      "\n",
      "Factor 3: Measure of attention and executive function-Trails A and B (F3)\n",
      "Factor 3: Measure of attention and executive function-Trails A and B (F3)\n",
      "('17903297', u'Factor 3: Measure of attention and executive function-Trails A and B (F3)', u'(F3)') 1\n",
      "\n",
      "28 yr time averaged FPG (tFPG)\n",
      "28 yr time averaged FPG (tFPG)\n",
      "('17903298', u'28 yr time averaged FPG (tFPG)', u'(tFPG)') 1\n",
      "\n",
      "Stage 2 Exercise systolic blood pressure (SBP)\n",
      "Stage 2 Exercise systolic blood pressure (SBP)\n",
      "('17903301', u'Stage 2 Exercise systolic blood pressure (SBP)', u'(SBP)') 1\n",
      "\n",
      "Stage 2 Exercise systolic blood pressure (SBP)\n",
      "Stage 2 Exercise systolic blood pressure (SBP)\n",
      "('17903301', u'Stage 2 Exercise systolic blood pressure (SBP)', u'(SBP)') 1\n",
      "\n",
      "Stage 2 Exercise diastolic blood pressure (DBP)\n",
      "Stage 2 Exercise diastolic blood pressure (DBP)\n",
      "('17903301', u'Stage 2 Exercise diastolic blood pressure (DBP)', u'(DBP)') 1\n",
      "\n",
      "Stage 2 Exercise diastolic blood pressure (DBP)\n",
      "Stage 2 Exercise diastolic blood pressure (DBP)\n",
      "('17903301', u'Stage 2 Exercise diastolic blood pressure (DBP)', u'(DBP)') 1\n",
      "\n",
      "2a.Top 25 SNPs for association with GFR (examination 7), UAE (examination 6), cysC (examination 7), and mean TSH (examinations 3 and 4) based on the lowest p value of the GEE test. Corresponding phenotype names on the web are GFRMV7 (GFR), UAELNMV6 (UAE), CYSCMV7 (CysC), and TSHMEAN34MV (TSH)\n",
      "(GFR), UAELNMV6 (UAE)\n",
      "('17903292', u'(GFR), UAELNMV6 (UAE)', u'(GFR)') 0\n",
      "\n",
      "2a.Top 25 SNPs for association with GFR (examination 7), UAE (examination 6), cysC (examination 7), and mean TSH (examinations 3 and 4) based on the lowest p value of the GEE test. Corresponding phenotype names on the web are GFRMV7 (GFR), UAELNMV6 (UAE), CYSCMV7 (CysC), and TSHMEAN34MV (TSH)\n",
      "(GFR), UAELNMV6 (UAE)\n",
      "('17903292', u'(GFR), UAELNMV6 (UAE)', u'(UAE)') 0\n",
      "\n",
      "2a.Top 25 SNPs for association with GFR (examination 7), UAE (examination 6), cysC (examination 7), and mean TSH (examinations 3 and 4) based on the lowest p value of the GEE test. Corresponding phenotype names on the web are GFRMV7 (GFR), UAELNMV6 (UAE), CYSCMV7 (CysC), and TSHMEAN34MV (TSH)\n",
      "CysC), and TSHMEAN34MV (TSH)\n",
      "('17903292', u'CysC), and TSHMEAN34MV (TSH)', u'(TSH)') 0\n",
      "\n",
      "2a.Top 25 SNPs for association with GFR (examination 7), UAE (examination 6), cysC (examination 7), and mean TSH (examinations 3 and 4) based on the lowest p value of the GEE test. Corresponding phenotype names on the web are GFRMV7 (GFR), UAELNMV6 (UAE), CYSCMV7 (CysC), and TSHMEAN34MV (TSH)\n",
      ", and TSHMEAN34MV (TSH)\n",
      "('17903292', u', and TSHMEAN34MV (TSH)', u'(TSH)') 0\n",
      "\n",
      "2a.Top 25 SNPs for association with GFR (examination 7), UAE (examination 6), cysC (examination 7), and mean TSH (examinations 3 and 4) based on the lowest p value of the GEE test. Corresponding phenotype names on the web are GFRMV7 (GFR), UAELNMV6 (UAE), CYSCMV7 (CysC), and TSHMEAN34MV (TSH)\n",
      ", and TSHMEAN34MV (TSH)\n",
      "('17903292', u', and TSHMEAN34MV (TSH)', u'(TSH)') 0\n",
      "\n",
      "2a.Top 25 SNPs for association with GFR (examination 7), UAE (examination 6), cysC (examination 7), and mean TSH (examinations 3 and 4) based on the lowest p value of the GEE test. Corresponding phenotype names on the web are GFRMV7 (GFR), UAELNMV6 (UAE), CYSCMV7 (CysC), and TSHMEAN34MV (TSH)\n",
      "GFR), UAELNMV6 (UAE)\n",
      "('17903292', u'GFR), UAELNMV6 (UAE)', u'(UAE)') 0\n",
      "\n",
      "2a.Top 25 SNPs for association with GFR (examination 7), UAE (examination 6), cysC (examination 7), and mean TSH (examinations 3 and 4) based on the lowest p value of the GEE test. Corresponding phenotype names on the web are GFRMV7 (GFR), UAELNMV6 (UAE), CYSCMV7 (CysC), and TSHMEAN34MV (TSH)\n",
      "), and TSHMEAN34MV (TSH)\n",
      "('17903292', u'), and TSHMEAN34MV (TSH)', u'(TSH)') 0\n",
      "\n",
      "2a.Top 25 SNPs for association with GFR (examination 7), UAE (examination 6), cysC (examination 7), and mean TSH (examinations 3 and 4) based on the lowest p value of the GEE test. Corresponding phenotype names on the web are GFRMV7 (GFR), UAELNMV6 (UAE), CYSCMV7 (CysC), and TSHMEAN34MV (TSH)\n",
      "), UAELNMV6 (UAE)\n",
      "('17903292', u'), UAELNMV6 (UAE)', u'(UAE)') 0\n",
      "\n",
      "2a.Top 25 SNPs for association with GFR (examination 7), UAE (examination 6), cysC (examination 7), and mean TSH (examinations 3 and 4) based on the lowest p value of the GEE test. Corresponding phenotype names on the web are GFRMV7 (GFR), UAELNMV6 (UAE), CYSCMV7 (CysC), and TSHMEAN34MV (TSH)\n",
      ", and TSHMEAN34MV (TSH)\n",
      "('17903292', u', and TSHMEAN34MV (TSH)', u'(TSH)') 0\n",
      "\n",
      "2a.Top 25 SNPs for association with GFR (examination 7), UAE (examination 6), cysC (examination 7), and mean TSH (examinations 3 and 4) based on the lowest p value of the GEE test. Corresponding phenotype names on the web are GFRMV7 (GFR), UAELNMV6 (UAE), CYSCMV7 (CysC), and TSHMEAN34MV (TSH)\n",
      ", UAELNMV6 (UAE)\n",
      "('17903292', u', UAELNMV6 (UAE)', u'(UAE)') 0\n",
      "\n",
      "2b.Top 25 SNPs for association with GFR (examination 7), UAE (examination 6), cysC (examination 7), and mean TSH (examinations 3 and 4) based on the lowest p-value of the FBAT test. Corresponding phenotype names on the web are GFRMV7 (GFR), UAELNMV6 (UAE), CYSCMV7 (CysC), and TSHMEAN34MV (TSH)\n",
      "(GFR), UAELNMV6 (UAE)\n",
      "('17903292', u'(GFR), UAELNMV6 (UAE)', u'(GFR)') 0\n",
      "\n",
      "2b.Top 25 SNPs for association with GFR (examination 7), UAE (examination 6), cysC (examination 7), and mean TSH (examinations 3 and 4) based on the lowest p-value of the FBAT test. Corresponding phenotype names on the web are GFRMV7 (GFR), UAELNMV6 (UAE), CYSCMV7 (CysC), and TSHMEAN34MV (TSH)\n",
      "(GFR), UAELNMV6 (UAE)\n",
      "('17903292', u'(GFR), UAELNMV6 (UAE)', u'(UAE)') 0\n",
      "\n",
      "2b.Top 25 SNPs for association with GFR (examination 7), UAE (examination 6), cysC (examination 7), and mean TSH (examinations 3 and 4) based on the lowest p-value of the FBAT test. Corresponding phenotype names on the web are GFRMV7 (GFR), UAELNMV6 (UAE), CYSCMV7 (CysC), and TSHMEAN34MV (TSH)\n",
      "CysC), and TSHMEAN34MV (TSH)\n",
      "('17903292', u'CysC), and TSHMEAN34MV (TSH)', u'(TSH)') 0\n",
      "\n",
      "2b.Top 25 SNPs for association with GFR (examination 7), UAE (examination 6), cysC (examination 7), and mean TSH (examinations 3 and 4) based on the lowest p-value of the FBAT test. Corresponding phenotype names on the web are GFRMV7 (GFR), UAELNMV6 (UAE), CYSCMV7 (CysC), and TSHMEAN34MV (TSH)\n",
      ", and TSHMEAN34MV (TSH)\n",
      "('17903292', u', and TSHMEAN34MV (TSH)', u'(TSH)') 0\n",
      "\n",
      "2b.Top 25 SNPs for association with GFR (examination 7), UAE (examination 6), cysC (examination 7), and mean TSH (examinations 3 and 4) based on the lowest p-value of the FBAT test. Corresponding phenotype names on the web are GFRMV7 (GFR), UAELNMV6 (UAE), CYSCMV7 (CysC), and TSHMEAN34MV (TSH)\n",
      ", and TSHMEAN34MV (TSH)\n",
      "('17903292', u', and TSHMEAN34MV (TSH)', u'(TSH)') 0\n",
      "\n",
      "2b.Top 25 SNPs for association with GFR (examination 7), UAE (examination 6), cysC (examination 7), and mean TSH (examinations 3 and 4) based on the lowest p-value of the FBAT test. Corresponding phenotype names on the web are GFRMV7 (GFR), UAELNMV6 (UAE), CYSCMV7 (CysC), and TSHMEAN34MV (TSH)\n",
      "GFR), UAELNMV6 (UAE)\n",
      "('17903292', u'GFR), UAELNMV6 (UAE)', u'(UAE)') 0\n",
      "\n",
      "2b.Top 25 SNPs for association with GFR (examination 7), UAE (examination 6), cysC (examination 7), and mean TSH (examinations 3 and 4) based on the lowest p-value of the FBAT test. Corresponding phenotype names on the web are GFRMV7 (GFR), UAELNMV6 (UAE), CYSCMV7 (CysC), and TSHMEAN34MV (TSH)\n",
      "), and TSHMEAN34MV (TSH)\n",
      "('17903292', u'), and TSHMEAN34MV (TSH)', u'(TSH)') 0\n",
      "\n",
      "2b.Top 25 SNPs for association with GFR (examination 7), UAE (examination 6), cysC (examination 7), and mean TSH (examinations 3 and 4) based on the lowest p-value of the FBAT test. Corresponding phenotype names on the web are GFRMV7 (GFR), UAELNMV6 (UAE), CYSCMV7 (CysC), and TSHMEAN34MV (TSH)\n",
      "), UAELNMV6 (UAE)\n",
      "('17903292', u'), UAELNMV6 (UAE)', u'(UAE)') 0\n",
      "\n",
      "2b.Top 25 SNPs for association with GFR (examination 7), UAE (examination 6), cysC (examination 7), and mean TSH (examinations 3 and 4) based on the lowest p-value of the FBAT test. Corresponding phenotype names on the web are GFRMV7 (GFR), UAELNMV6 (UAE), CYSCMV7 (CysC), and TSHMEAN34MV (TSH)\n",
      ", and TSHMEAN34MV (TSH)\n",
      "('17903292', u', and TSHMEAN34MV (TSH)', u'(TSH)') 0\n",
      "\n",
      "2b.Top 25 SNPs for association with GFR (examination 7), UAE (examination 6), cysC (examination 7), and mean TSH (examinations 3 and 4) based on the lowest p-value of the FBAT test. Corresponding phenotype names on the web are GFRMV7 (GFR), UAELNMV6 (UAE), CYSCMV7 (CysC), and TSHMEAN34MV (TSH)\n",
      ", UAELNMV6 (UAE)\n",
      "('17903292', u', UAELNMV6 (UAE)', u'(UAE)') 0\n",
      "\n",
      "Physical Position (bp)†\n",
      "Physical Position (bp)\n",
      "('17903294', u'Physical Position (bp)', u'(bp)') 0\n",
      "\n",
      "Physical Position (bp)†\n",
      "Physical Position (bp)\n",
      "('17903294', u'Physical Position (bp)', u'(bp)') 0\n",
      "\n",
      "Physical Position (bp)†\n",
      "Position (bp)\n",
      "('17903294', u'Position (bp)', u'(bp)') 0\n",
      "\n",
      "Physical Position (bp)†\n",
      "Position (bp)\n",
      "('17903294', u'Position (bp)', u'(bp)') 0\n",
      "\n",
      "Physical Position (bp)†\n",
      "Physical Position (bp)\n",
      "('17903294', u'Physical Position (bp)', u'(bp)') 0\n",
      "\n",
      "Physical Position (bp)†\n",
      "Position (bp)\n",
      "('17903294', u'Position (bp)', u'(bp)') 0\n",
      "\n",
      "GEE p-val (Rank)††\n",
      "GEE p-val (Rank)\n",
      "('17903294', u'GEE p-val (Rank)', u'(Rank)') 0\n",
      "\n",
      "GEE p-val (Rank)††\n",
      "p-val (Rank)\n",
      "('17903294', u'p-val (Rank)', u'(Rank)') 0\n",
      "\n",
      "Physical Position (bp)†\n",
      "Physical Position (bp)\n",
      "('17903294', u'Physical Position (bp)', u'(bp)') 0\n",
      "\n",
      "Physical Position (bp)†\n",
      "Physical Position (bp)\n",
      "('17903294', u'Physical Position (bp)', u'(bp)') 0\n",
      "\n",
      "Physical Position (bp)†\n",
      "Position (bp)\n",
      "('17903294', u'Position (bp)', u'(bp)') 0\n",
      "\n",
      "Physical Position (bp)†\n",
      "Position (bp)\n",
      "('17903294', u'Position (bp)', u'(bp)') 0\n",
      "\n",
      "Physical Position (bp)†\n",
      "Physical Position (bp)\n",
      "('17903294', u'Physical Position (bp)', u'(bp)') 0\n",
      "\n",
      "Physical Position (bp)†\n",
      "Position (bp)\n",
      "('17903294', u'Position (bp)', u'(bp)') 0\n",
      "\n",
      "GEE Pval (Rank)††\n",
      "GEE Pval (Rank)\n",
      "('17903294', u'GEE Pval (Rank)', u'(Rank)') 0\n",
      "\n",
      "GEE Pval (Rank)††\n",
      "Pval (Rank)\n",
      "('17903294', u'Pval (Rank)', u'(Rank)') 0\n",
      "\n",
      "Physical Position (bp)†\n",
      "Physical Position (bp)\n",
      "('17903294', u'Physical Position (bp)', u'(bp)') 0\n",
      "\n",
      "Physical Position (bp)†\n",
      "Physical Position (bp)\n",
      "('17903294', u'Physical Position (bp)', u'(bp)') 0\n",
      "\n",
      "Physical Position (bp)†\n",
      "Position (bp)\n",
      "('17903294', u'Position (bp)', u'(bp)') 0\n",
      "\n",
      "Physical Position (bp)†\n",
      "Position (bp)\n",
      "('17903294', u'Position (bp)', u'(bp)') 0\n",
      "\n",
      "Physical Position (bp)†\n",
      "Physical Position (bp)\n",
      "('17903294', u'Physical Position (bp)', u'(bp)') 0\n",
      "\n",
      "Physical Position (bp)†\n",
      "Position (bp)\n",
      "('17903294', u'Position (bp)', u'(bp)') 0\n",
      "\n",
      "Factor 1:Verbal Memory (F1)\n",
      "Factor 1:Verbal Memory (F1)\n",
      "('17903297', u'Factor 1:Verbal Memory (F1)', u'(F1)') 1\n",
      "\n",
      "Factor 1:Verbal Memory (F1)\n",
      ":Verbal Memory (F1)\n",
      "('17903297', u':Verbal Memory (F1)', u'(F1)') 0\n",
      "\n",
      "Factor 1:Verbal Memory (F1)\n",
      ":Verbal Memory (F1)\n",
      "('17903297', u':Verbal Memory (F1)', u'(F1)') 0\n",
      "\n",
      "Factor 1:Verbal Memory (F1)\n",
      "1:Verbal Memory (F1)\n",
      "('17903297', u'1:Verbal Memory (F1)', u'(F1)') 0\n",
      "\n",
      "Factor 1:Verbal Memory (F1)\n",
      "Verbal Memory (F1)\n",
      "('17903297', u'Verbal Memory (F1)', u'(F1)') 0\n",
      "\n",
      "Factor 1:Verbal Memory (F1)\n",
      "Verbal Memory (F1)\n",
      "('17903297', u'Verbal Memory (F1)', u'(F1)') 0\n",
      "\n",
      "Factor 1:Verbal Memory (F1)\n",
      ":Verbal Memory (F1)\n",
      "('17903297', u':Verbal Memory (F1)', u'(F1)') 0\n",
      "\n",
      "Factor 1:Verbal Memory (F1)\n",
      "Memory (F1)\n",
      "('17903297', u'Memory (F1)', u'(F1)') 0\n",
      "\n",
      "Factor 1:Verbal Memory (F1)\n",
      "Memory (F1)\n",
      "('17903297', u'Memory (F1)', u'(F1)') 0\n",
      "\n",
      "Factor 1:Verbal Memory (F1)\n",
      "Verbal Memory (F1)\n",
      "('17903297', u'Verbal Memory (F1)', u'(F1)') 0\n",
      "\n",
      "Factor 1:Verbal Memory (F1)\n",
      "Memory (F1)\n",
      "('17903297', u'Memory (F1)', u'(F1)') 0\n",
      "\n",
      "Factor 2:Visual Memory and Organization (F2)\n",
      ":Visual Memory and Organization (F2)\n",
      "('17903297', u':Visual Memory and Organization (F2)', u'(F2)') 0\n",
      "\n",
      "Factor 2:Visual Memory and Organization (F2)\n",
      "Memory and Organization (F2)\n",
      "('17903297', u'Memory and Organization (F2)', u'(F2)') 0\n",
      "\n",
      "Factor 2:Visual Memory and Organization (F2)\n",
      "Memory and Organization (F2)\n",
      "('17903297', u'Memory and Organization (F2)', u'(F2)') 0\n",
      "\n",
      "Factor 2:Visual Memory and Organization (F2)\n",
      "Visual Memory and Organization (F2)\n",
      "('17903297', u'Visual Memory and Organization (F2)', u'(F2)') 0\n",
      "\n",
      "Factor 2:Visual Memory and Organization (F2)\n",
      "and Organization (F2)\n",
      "('17903297', u'and Organization (F2)', u'(F2)') 0\n",
      "\n",
      "Factor 2:Visual Memory and Organization (F2)\n",
      "and Organization (F2)\n",
      "('17903297', u'and Organization (F2)', u'(F2)') 0\n",
      "\n",
      "Factor 2:Visual Memory and Organization (F2)\n",
      "Memory and Organization (F2)\n",
      "('17903297', u'Memory and Organization (F2)', u'(F2)') 0\n",
      "\n",
      "Factor 2:Visual Memory and Organization (F2)\n",
      "Organization (F2)\n",
      "('17903297', u'Organization (F2)', u'(F2)') 0\n",
      "\n",
      "Factor 2:Visual Memory and Organization (F2)\n",
      "Organization (F2)\n",
      "('17903297', u'Organization (F2)', u'(F2)') 0\n",
      "\n",
      "Factor 2:Visual Memory and Organization (F2)\n",
      "and Organization (F2)\n",
      "('17903297', u'and Organization (F2)', u'(F2)') 0\n",
      "\n",
      "Factor 2:Visual Memory and Organization (F2)\n",
      "Organization (F2)\n",
      "('17903297', u'Organization (F2)', u'(F2)') 0\n",
      "\n",
      "Factor 3: Measure of attention and executive function-Trails A and B (F3)\n",
      "executive function-Trails A and B (F3)\n",
      "('17903297', u'executive function-Trails A and B (F3)', u'(F3)') 0\n",
      "\n",
      "Factor 3: Measure of attention and executive function-Trails A and B (F3)\n",
      "A and B (F3)\n",
      "('17903297', u'A and B (F3)', u'(F3)') 0\n",
      "\n",
      "Factor 3: Measure of attention and executive function-Trails A and B (F3)\n",
      "A and B (F3)\n",
      "('17903297', u'A and B (F3)', u'(F3)') 0\n",
      "\n",
      "Factor 3: Measure of attention and executive function-Trails A and B (F3)\n",
      "function-Trails A and B (F3)\n",
      "('17903297', u'function-Trails A and B (F3)', u'(F3)') 0\n",
      "\n",
      "Factor 3: Measure of attention and executive function-Trails A and B (F3)\n",
      "and B (F3)\n",
      "('17903297', u'and B (F3)', u'(F3)') 0\n",
      "\n",
      "Factor 3: Measure of attention and executive function-Trails A and B (F3)\n",
      "and B (F3)\n",
      "('17903297', u'and B (F3)', u'(F3)') 0\n",
      "\n",
      "Factor 3: Measure of attention and executive function-Trails A and B (F3)\n",
      "A and B (F3)\n",
      "('17903297', u'A and B (F3)', u'(F3)') 0\n",
      "\n",
      "Factor 3: Measure of attention and executive function-Trails A and B (F3)\n",
      "B (F3)\n",
      "('17903297', u'B (F3)', u'(F3)') 0\n",
      "\n",
      "Factor 3: Measure of attention and executive function-Trails A and B (F3)\n",
      "B (F3)\n",
      "('17903297', u'B (F3)', u'(F3)') 0\n",
      "\n",
      "Factor 3: Measure of attention and executive function-Trails A and B (F3)\n",
      "and B (F3)\n",
      "('17903297', u'and B (F3)', u'(F3)') 0\n",
      "\n",
      "Factor 3: Measure of attention and executive function-Trails A and B (F3)\n",
      "B (F3)\n",
      "('17903297', u'B (F3)', u'(F3)') 0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# store relations to annotate\n",
    "with open('acronyms.unannotated.tsv', 'w') as f:\n",
    "    for span_pair in all_c:\n",
    "        doc_id = span_pair.span0.context.document.name\n",
    "        str1 = span_pair.span0.get_span()\n",
    "        str2 = span_pair.span1.get_span()\n",
    "        if (doc_id, str1, str2) not in K: \n",
    "            if hasattr(span_pair.span1.context, 'cell'):\n",
    "                print span_pair.span0.context.cell.text\n",
    "                print span_pair.span0.get_span()\n",
    "                print (doc_id, str1, str2), LF_full_cell(span_pair)\n",
    "                print \n",
    "        try:\n",
    "            f.write('%s\\t%s\\t%s\\n' % (doc_id, str2, str1))\n",
    "        except:\n",
    "            continue"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "We now load the results of our annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "annotations = dict()\n",
    "with open('acronyms.anotated.txt') as f:\n",
    "    text = f.read()\n",
    "    for line in text.split('\\r'):\n",
    "        doc_id, str1, str2, res = line.strip().split('\\t')\n",
    "        res = 1 if int(res) == 1 else -1\n",
    "        annotations[(doc_id, str2, str1)] = res"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Classifying table acronyms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Building feature index...\n",
      "Extracting features...\n",
      "0/7631\n",
      "5000/7631\n"
     ]
    }
   ],
   "source": [
    "from snorkel.features import TableNgramPairFeaturizer\n",
    "\n",
    "pkl_f = 'acro_table_feats.pkl'\n",
    "try:\n",
    "    with open(pkl_f, 'rb') as f:\n",
    "        featurizer = cPickle.load(f)\n",
    "except:\n",
    "    featurizer = TableNgramPairFeaturizer()\n",
    "    featurizer.fit_transform(candidates)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Training set size: 210\n",
      "Gold set size: 57\n",
      "Positive labels in training set: 0\n",
      "Negative labels in training set: 0\n",
      "Positive labels in gold set: 36\n",
      "Negative labels in gold set: 21\n"
     ]
    }
   ],
   "source": [
    "def spair2uid(span_pair):\n",
    "    doc_id = span_pair.span0.context.document.name\n",
    "    str1 = span_pair.span0.get_span()\n",
    "    str2 = span_pair.span1.get_span()\n",
    "    return (doc_id, str1, str2)\n",
    "\n",
    "# Split into train and test set\n",
    "training_candidates = []\n",
    "gold_candidates     = []\n",
    "gold_labels         = []\n",
    "n_half = len(candidates)/2\n",
    "for c in candidates[:n_half]:\n",
    "    uid = spair2uid(c)\n",
    "    if uid in annotations:\n",
    "        gold_candidates.append(c)\n",
    "        gold_labels.append(annotations[uid])\n",
    "    else:\n",
    "        training_candidates.append(c)\n",
    "training_candidates.extend(candidates[n_half:])\n",
    "gold_labels = np.array(gold_labels)\n",
    "print \"Training set size: %s\" % len(training_candidates)\n",
    "print \"Gold set size: %s\" % len(gold_candidates)\n",
    "print \"Positive labels in training set: %s\" % len([c for c in training_candidates if annotations.get(spair2uid(c),0)==1])\n",
    "print \"Negative labels in training set: %s\" % len([c for c in training_candidates if annotations.get(spair2uid(c),0)==-1])\n",
    "print \"Positive labels in gold set: %s\" % len([c for c in gold_candidates if annotations[spair2uid(c)]==1])\n",
    "print \"Negative labels in gold set: %s\" % len([c for c in gold_candidates if annotations[spair2uid(c)]==-1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def LF1_digits(m):\n",
    "    txt = m.span1.get_span()\n",
    "    frac_num = len([ch for ch in txt if ch.isdigit()]) / float(len(txt))\n",
    "    return -1 if frac_num > 0.5 else +1\n",
    "def LF1_short(m):\n",
    "    txt = m.span1.get_span()\n",
    "    return -1 if len(txt) < 5 else 0\n",
    "\n",
    "LF_tables = [LF1_digits, LF1_short]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Library/Python/2.7/site-packages/matplotlib/__init__.py:1155: UserWarning:  This call to matplotlib.use() has no effect\n",
      "because the backend has already been chosen;\n",
      "matplotlib.use() must be called *before* pylab, matplotlib.pyplot,\n",
      "or matplotlib.backends is imported for the first time.\n",
      "\n",
      "  warnings.warn(_use_error_msg)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Applying LFs...\n",
      "Featurizing...\n",
      "Building feature index...\n",
      "Extracting features...\n",
      "0/6567\n",
      "5000/6567\n",
      "============================================================\n",
      "LF Summary Statistics: 2 LFs applied to 210 candidates\n",
      "------------------------------------------------------------\n",
      "Coverage (candidates w/ > 0 labels):\t\t64.29%\n",
      "Overlap (candidates w/ > 1 labels):\t\t48.10%\n",
      "Conflict (candidates w/ conflicting labels):\t35.71%\n",
      "============================================================\n"
     ]
    }
   ],
   "source": [
    "from snorkel.snorkel import TrainingSet\n",
    "from snorkel.features import NgramFeaturizer\n",
    "\n",
    "training_set = TrainingSet(training_candidates, LF_tables, featurizer=TableNgramPairFeaturizer())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[<matplotlib.axes.AxesSubplot object at 0x130178b10>]], dtype=object)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAsUAAAEKCAYAAAD6nSjkAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAGdhJREFUeJzt3X2QXXV9x/H3hoiPSKw4KhDYGYkWaGvAKaCtci10QFtE\nRyqGMWVrxqHTYqdWZ5AZa6DWdrDjw1CmRuv4kPEBWG0dFBSL5lQdRasYquFBcNxCjEFRg2ltBWT7\nx+8se7nZ5P6Sc3d/97u/92tmyZ57z958ySfZfHL2e++CJEmSJEmSJEmSJEmSJEmSJEmSJEmSJEmS\nJEmSJEmSJEmSJEmSJElSACtKDyBJy5GfXCVp/60G/gX4EXAv8I/ABPAmYAa4B/gQ8MT2/M8Cfz7w\nGDcDL23f/3Xg34CfALcBf9R33geBdwPXAf8N9IA/AL4F3AfcBWwceOw/Bv6rnW1uptPa+yaANwJ3\ntvdfBTwp+/9ckiRJAg4iFdq3A48FDgZ+B3g1cAcwCTwe+ASwuf2Y9cCX+x7jOOBnwKPac+8Gzidd\nqDgB+HF7DqRSvAt4bnv8aOBU4Pj2+DeBncDZfY+9G3he+/j/ANwP/F57/18CXwEOb+/fBHz0AH4d\nJEmSVLHnkq4QD36l7fPAn/YdP5NURlcAh5Cu8q5u73sr8L72/XOBLw481nuAN7fvf7B925d3Ae9o\n338z8JG++x4L/JL5UnxL3/sAT++bU5Kq5SdBSdo/q0mrCQ8N3P709vY5dwErgaeSrtxeC6xr7zuX\n+eJ6NHAy6crx3Nt57ccBzJKuJPc7GdhCKue7gAuAJ7f3HQ5s7zv3f0lrGXMmgX/t+7luAR7s+/kk\nqUqWYknaP3cDR5HWKPrtIBXOOUeRyuY97fHHSKX4uaSrt1va2+8C/p201zv3dgh77iD3+yjwSeBI\nYBVpBWKib44j+859LPOFee7nO3Pg53sc8MN9/HySJEnSI6wAtpJ2dR8HPIa0U7wB+C6pGD8B+Djz\nO8WQdo9/CnyOtI885wmkJ8K9irTj+yjgt0lPvoO0OvGWgRnuIT2ZDuCk9nju5zoe+DmpfB8MvI09\nd4q3kEo7wFOAl+T9r0uSJEnzVpNWEO4lPSnuXaQrtX9NuhL7I1JJPXTg494H/Ap4zsDtzwQ+zfyr\nWdwA/FZ73weAvxk4/+WkIv1z4FPA5TyygJ/PI199YjupuNPO+TrSq1z8nPQqFH+b978tSXV7P+kq\nxLf3cc7lpGdd30x65rQkaTw8AXiAtLssSerg+aSiu7dS/GLS62dCevLHjUsxlCRpr84irXY8nrRv\n/M2y40jS8jHJ3kvxJtIzqefchs9ilqSS/pn0yhK7SN8UZE3ZcSRp+Zhk76X4U6QXiZ9zA3vuy0mS\nJEljaxQvyTbB/EsBzRl8/U5JkiRpbK0cwWNsZ/67NEF6fcwdgycdfvjhszt27HGzJEmSNErfA47Z\n3w8axZXia5h/vcxTSDts9wyetGPHDmZnZ31bxLerr76aQw45h/QNsEb5tnERHrPL27tZv/6C4r/e\nUd42btxYfAbfzK/GN7OL/bYc80tK/x2+FG88gwOQc6X4Y8CpwGGk7+S0kfTi8gDvIb3yxItJr3X5\nP8CfHMggGmczpQdQBzMzM6VHUAfmF5fZxWZ+9ckpxesyzrmw6yCSJElSKaNYn9CyN1V6AHUwNTVV\negR1YH5xmV1s5lcfS7Ey9EoPoA56vV7pEdSB+cVldrGZX30sxcrQlB5AHTRNU3oEdWB+cZldbOZX\nH0uxJEmSqmcpVoZe6QHUgV8CjM384jK72MyvPpZiSZIkVc9SrAxN6QHUgXtxsZlfXGYXm/nVx1Is\nSZKk6lmKlaFXegB14F5cbOYXl9nFZn71sRRLkiSpepZiZWhKD6AO3IuLzfziMrvYzK8+lmJJkiRV\nz1KsDL3SA6gD9+JiM7+4zC4286uPpViSJEnVsxQrQ1N6AHXgXlxs5heX2cVmfvWxFEuSJKl6lmJl\n6JUeQB24Fxeb+cVldrGZX30sxZIkSaqepVgZmtIDqAP34mIzv7jMLjbzq4+lWJIkSdWzFCtDr/QA\n6sC9uNjMLy6zi8386mMpliRJUvUsxcrQlB5AHbgXF5v5xWV2sZlffSzFkiRJqp6lWBl6pQdQB+7F\nxWZ+cZldbOZXH0uxJEmSqmcpVoam9ADqwL242MwvLrOLzfzqYymWJElS9SzFytArPYA6cC8uNvOL\ny+xiM7/6WIolSZJUPUuxMjSlB1AH7sXFZn5xmV1s5lcfS7EkSZKqZylWhl7pAdSBe3GxmV9cZheb\n+dXHUixJkqTqWYqVoSk9gDpwLy4284vL7GIzv/pYiiVJklS9nFJ8JnAbcAdw0QL3HwVsAW4CbgZe\nNLLpNCZ6pQdQB+7FxWZ+cZldbOZXn2Gl+CDgClIxPg5YBxw7cM6bgCuBE4FXAv804hklSZKkRTWs\nFJ8E3AnMAA+Qyu/ZA+c8BBzavr8K+MEI59NYaEoPoA7ci4vN/OIyu9jMrz4rh9x/BHB33/F24OSB\ncy4BPge8Fng8cNqohpMkSZKWwrBSvJDZgePzgA8A7wROAT4MHL/QB05NTTE5OQnAqlWrWLt27cM7\nO3P/IvP4wI+3bdvGvKb9sTeC496IH6/78c6dO2iaZqx+/cf1uNfrjdU8Hpufxx57XO649N/fi3O8\nFdjVHs9woCaG3H8K6Urwme3xxaR1icv6zvkOcAbzaxPfI11NvnfgsWZnZwf7tEZpenqaDRuuZvfu\n6dKjLLJNrF+/lc2bN5UeRJKkMCYmJtjz2uZyNPHwf/bHiiH3fwNYA0wCBwPnAtcMnHMXcHr7/rHA\nY9izECu0pvQA6mDuKoFiMr+4zC4286vPsFL8IHAhcD1wC3AVcCtwKXBWe87rgdeQrl1/FDh/USaV\nJEmSFknOTvFn2rd+G/vevxX43ZFNpDHUKz2AOpjfI1NE5heX2cVmfvUZdqVYkiRJWvYsxcrQlB5A\nHbgXF5v5xWV2sZlffSzFkiRJqp6lWBl6pQdQB+7FxWZ+cZldbOZXH0uxJEmSqmcpVoam9ADqwL24\n2MwvLrOLzfzqYymWJElS9SzFytArPYA6cC8uNvOLy+xiM7/6WIolSZJUPUuxMjSlB1AH7sXFZn5x\nmV1s5lcfS7EkSZKqZylWhl7pAdSBe3GxmV9cZheb+dXHUixJkqTqWYqVoSk9gDpwLy4284vL7GIz\nv/pYiiVJklQ9S7Ey9EoPoA7ci4vN/OIyu9jMrz6WYkmSJFXPUqwMTekB1IF7cbGZX1xmF5v51cdS\nLEmSpOpZipWhV3oAdeBeXGzmF5fZxWZ+9bEUS5IkqXqWYmVoSg+gDtyLi8384jK72MyvPpZiSZIk\nVc9SrAy90gOoA/fiYjO/uMwuNvOrj6VYkiRJ1bMUK0NTegB14F5cbOYXl9nFZn71sRRLkiSpepZi\nZeiVHkAduBcXm/nFZXaxmV99LMWSJEmqnqVYGZrSA6gD9+JiM7+4zC4286uPpViSJEnVsxQrQ6/0\nAOrAvbjYzC8us4vN/OpjKZYkSVL1LMXK0JQeQB24Fxeb+cVldrGZX30sxZIkSapeTik+E7gNuAO4\naC/nvALYBnwH+MhoRtP46JUeQB24Fxeb+cVldrGZX31WDrn/IOAK4HTgB8B/ANcAt/adswZ4I/A8\n4D7gsNGPKUmSJC2eYVeKTwLuBGaAB4ArgbMHznkNqTjf1x7fO8L5NBaa0gOoA/fiYjO/uMwuNvOr\nz7BSfARwd9/x9va2fmuAZwFfBr4KnDGy6SRJkqQlMGx9YiGzCzzGMcCpwGrgS8BvMH/lWOH1Sg+g\nDtyLi8384jK72MyvPsNK8XZS0Z2zmrRbPHjOjcCvSGsWt5NK8jcHH2xqaorJyUkAVq1axdq1ax/+\nTTf3ZQqPD/x427ZtzGvaH3vL8njnzh00TTNWv/4ee+yxxx57PO7Hpf/+XpzjrcCu9niGAzUx5P6V\npJJ7GrAD+Dqwjkc+0e6M9rYp0pPsbgKeDfxs4LFmZ2cHLzJrlKanp9mw4Wp2754e8SM3zP/mGweb\nWL9+K5s3byo9SAhN0/R9MlQ05heX2cW2HPObmJhgzy/4L0cTD/9nf6wYcv+DwIXA9cAtwFWkQnwp\ncFZ7zvXAT0gvyfYF4A3sWYglSZKksbXfLboDrxQvssW7UjxuvFIsSdL+8krxvg27UixJkiQte5Zi\nZWhKD6AO5p5koZjMLy6zi8386mMpliRJUvUsxcrQKz2AOlhuz56ujfnFZXaxmV99LMWSJEmqnqVY\nGZrSA6gD9+JiM7+4zC4286uPpViSJEnVsxQrQ6/0AOrAvbjYzC8us4vN/OpjKZYkSVL1LMXK0JQe\nQB24Fxeb+cVldrGZX30sxZIkSaqepVgZeqUHUAfuxcVmfnGZXWzmVx9LsSRJkqpnKVaGpvQA6sC9\nuNjMLy6zi8386mMpliRJUvUsxcrQKz2AOnAvLjbzi8vsYjO/+liKJUmSVD1LsTI0pQdQB+7FxWZ+\ncZldbOZXH0uxJEmSqmcpVoZe6QHUgXtxsZlfXGYXm/nVx1IsSZKk6lmKlaEpPYA6cC8uNvOLy+xi\nM7/6WIolSZJUPUuxMvRKD6AO3IuLzfziMrvYzK8+lmJJkiRVz1KsDE3pAdSBe3GxmV9cZheb+dXH\nUixJkqTqWYqVoVd6AHXgXlxs5heX2cVmfvWxFEuSJKl6lmJlaEoPoA7ci4vN/OIyu9jMrz6WYkmS\nJFXPUqwMvdIDqAP34mIzv7jMLjbzq4+lWJIkSdWzFCtDU3oAdeBeXGzmF5fZxWZ+9bEUS5IkqXqW\nYmXolR5AHbgXF5v5xWV2sZlffXJK8ZnAbcAdwEX7OO8c4CHgxBHMJUmSJC2ZYaX4IOAKUjE+DlgH\nHLvAeYcAfwHcONLpNCaa0gOoA/fiYjO/uMwuNvOrz7BSfBJwJzADPABcCZy9wHlvAS4DfglMjHA+\nSZIkadENK8VHAHf3HW9vb+t3Qnvbte3x7GhG0/jolR5AHbgXF5v5xWV2sZlffVYewMf0l94VwDuB\n8/tu80qxJEmSQhlWircDq/uOVwM/6Ds+BDie+aXTpwHXAGcBNw0+2NTUFJOTkwCsWrWKtWvXPvwv\nsbndHY8P/Hjbtm3Ma9ofeyM4nnt/VI/X/Xjnzh00TTNWv/7jety/FzcO83hsfrUcz902LvN4vH/H\nc7eNyzyjOi799/fiHG8FdrXHMxyoYVd1VwK3A6cBO4Cvk55sd+tezt8CvJ4FCjEwOzvrZsVimp6e\nZsOGq9m9e3rEj9ww/5tvHGxi/fqtbN68qfQgITRN0/fJUNGYX1xmF9tyzG9iYoI6tlwnHv7P/lgx\n5P4HgQuB64FbgKtIhfhS0tVgVaFXegB1sNw+qdfG/OIyu9jMrz45O8Wfad/6bdzLuS/sNo4kSZK0\n9IZdKZaY399RRP37cYrH/OIyu9jMrz6WYkmSJFXPUqwMvdIDqAP34mIzv7jMLjbzq4+lWJIkSdWz\nFCtDU3oAdeBeXGzmF5fZxWZ+9bEUS5IkqXqWYmXolR5AHbgXF5v5xWV2sZlffSzFkiRJqp6lWBma\n0gOoA/fiYjO/uMwuNvOrj6VYkiRJ1bMUK0Ov9ADqwL242MwvLrOLzfzqYymWJElS9SzFytCUHkAd\nuBcXm/nFZXaxmV99LMWSJEmqnqVYGXqlB1AH7sXFZn5xmV1s5lcfS7EkSZKqZylWhqb0AOrAvbjY\nzC8us4vN/OpjKZYkSVL1LMXK0Cs9gDpwLy4284vL7GIzv/pYiiVJklQ9S7EyNKUHUAfuxcVmfnGZ\nXWzmVx9LsSRJkqpnKVaGXukB1IF7cbGZX1xmF5v51cdSLEmSpOpZipWhKT2AOnAvLjbzi8vsYjO/\n+liKJUmSVD1LsTL0Sg+gDtyLi8384jK72MyvPpZiSZIkVc9SrAxN6QHUgXtxsZlfXGYXm/nVx1Is\nSZKk6lmKlaFXegB14F5cbOYXl9nFZn71sRRLkiSpepZiZWhKD6AO3IuLzfziMrvYzK8+lmJJkiRV\nz1KsDL3SA6gD9+JiM7+4zC4286uPpViSJEnVyy3FZwK3AXcAFy1w/18B24CbgRuAo0YyncZEU3oA\ndeBeXGzmF5fZxWZ+9ckpxQcBV5CK8XHAOuDYgXNuAp4DPBv4OPC2Ec4oSZIkLaqcUnwScCcwAzwA\nXAmcPXBOA/xf+/7XgCNHM57GQ6/0AOrAvbjYzC8us4vN/OqTU4qPAO7uO97e3rY3G4DrugwlSZIk\nLaWVB/hxs3u5/VXAicDrFrpzamqKyclJAFatWsXatWsf/pfY3O6Oxwd+vG3bNuY17Y+9ERzPvT+q\nx+t+vHPnDpqmGatf/3E97t+LG4d5PDa/Wo7nbhuXeTzev+O528ZlnlEdl/77e3GOtwK72uMZDtRE\nxjmnAJeQdooBLgYeAi4bOO904HLgBcC9CzzO7Ozs3rq0RmF6epoNG65m9+7pET9yw/xvvnGwifXr\nt7J586bSg4TQNE3fJ0NFY35xmV1syzG/iYkJ9n5dczmZePg/+2NFxjnfANYAk8DBwLnANQPnnABs\nAs5i4UKs0HqlB1AHy+2Tem3MLy6zi8386pNTih8ELgSuB24BrgJuBS4F/rA9523A40mvPPEt4JMj\nn1SSJElaJDmlGOAzwLOAY4C/b2/bCHy6ff/3gaeTrhifALx0hDOquKb0AOqgfz9O8ZhfXGYXm/nV\nJ7cUS5IkScuWpVgZeqUHUAfuxcVmfnGZXWzmVx9LsSRJkqpnKVaGpvQA6sC9uNjMLy6zi8386mMp\nliRJUvUsxcrQKz2AOnAvLjbzi8vsYjO/+liKJUmSVD1LsTI0pQdQB+7FxWZ+cZldbOZXH0uxJEmS\nqmcpVoZe6QHUgXtxsZlfXGYXm/nVx1IsSZKk6lmKlaEpPYA6cC8uNvOLy+xiM7/6WIolSZJUPUux\nMvRKD6AO3IuLzfziMrvYzK8+lmJJkiRVz1KsDE3pAdSBe3GxmV9cZheb+dXHUixJkqTqWYqVoVd6\nAHXgXlxs5heX2cVmfvWxFEuSJKl6lmJlaEoPoA7ci4vN/OIyu9jMrz6WYkmSJFXPUqwMvdIDqAP3\n4mIzv7jMLjbzq4+lWJIkSdWzFCtDU3oAdeBeXGzmF5fZxWZ+9bEUS5IkqXqWYmXolR5AHbgXF5v5\nxWV2sZlffSzFkiRJqp6lWBma0gOoA/fiYjO/uMwuNvOrj6VYkiRJ1bMUK0Ov9ADqwL242MwvLrOL\nzfzqYymWJElS9SzFytCUHkAduBcXm/nFZXaxmV99LMWSJEmqnqVYGXqlB1AH7sXFZn5xmV1s5lcf\nS7EkSZKql1OKzwRuA+4ALlrg/kcDV7X33wgcPbLpNCaa0gOoA/fiYjO/uMwuNvOrz7BSfBBwBakY\nHwesA44dOGcD8BNgDfBO4LIRz6jitpYeQB1s3Wp+kZlfXGYXm/nVZ1gpPgm4E5gBHgCuBM4eOOcl\nwIfa9z8BnDbC+TQWdpUeQB3s2mV+kZlfXGYXm/nVZ1gpPgK4u+94e3vb3s55ELgP+LWRTCdJkiQt\ngZUH8DGzA8cTGedoCaxYsYIHHvgyT3ziWSN93F/84ls87nHfHOljdnH//TOsXPn80mOEMTMzU3oE\ndWB+cZldbOZXn4UKbb9TgEtIO8UAFwMP8ci94c+259xIKtk/BJ6ywGPdCTzjwEeVJEmShvoecMyo\nH3Rl+8CTwMGkZ1wNPtHuz4B3t++/krR3LEmSJC0rLwJuJ13pvbi97VJg7mv0jwauZv4l2SaXeD5J\nkiRJkiRJ0rgZ9s0+5pxD2k8+cSmGUpZh2U0BPwa+1b69eskmU46cP3uvALYB3wE+skRzabhh2b2D\n+T93twM/W7rRlGFYfkcBW4CbgJtJX4HV+BiW39HA50nZbWHPV+FSGe8H7gG+vY9zLiflejNwwlIM\n1e8g0prFJPAoFt5BBjgE+CLwFSzF4yInu/NJv8E0fnLyW0P6S/nQ9viwpRpO+5T7eXPOhcD7Fn8s\nZcrJ773ABe37xwLfX6rhNFROftPA+vb9FwKbl2o47dPzSUV3b6X4xcB17fsnk1Z89ynn2zzvj5xv\n9gHwFtIrWPyS4a+AoaWRk90E5jWucvJ7Dek7VN7XHt+7VMNpn3I/b845D/jY4o+lTDn5PcT8P0ZX\nAT9YquE0VE5+x5KuFAM0C9yvMr7Evr9q1v/N5b5G+rP31H094KhLcc43+zihve3a9tjXNB4POdnN\nAi8nfRliGjhyaUZThpz81gDPAr4MfBU4Y2lG0xA52c05mnRF6wuLPJPy5eR3CfCq9rxrgdcuyWTK\nkZPfzaSVT4CXkb7a/aTFH00dLZTtPnvLqEvxQvpL7wrgncAb+m7zyuP4GvwHy6dIfyk/G7iB+X+B\naTwN5reS9LqNpwLrSF+CP3TwgzQW9nax4JWkf5B6MWG8DeZzHvABYDXpS7ofXvKJtD8G83sD6fPm\nTcALSFf6H1zqobTfFvrq9kP7+oBRl+LtpD/0c1bzyC8THQIcT/ryw/dJ3xzkGtwrHgfDsgP4KenL\nS5AK1XOWYC7lyclvO+nP269IXyq8nUV4cXPtt5zs5pyLqxPjJie/V5NeuhTSXuNjcKd/XOTk90PS\nV0lPBN7U3rZ78UdTR4PZHgnsWMoBcr7ZR78tWIjHRU52T+t7/2WkJ0pqPOTkdwbwwfb9w4C78EuA\n4yD38+az8Ala4ygnv+tIT1Smvc+d4vGRk9+Tmb+I+FbSOozGwyR5T7Q7hYwn2i2GYd/so5+leLwM\ny+7vSC/ltZX0pINnLvWA2qecP3tvJ70k23+SXp5N4yEnu42kP4MaP8PyO5a0y7+V9LJ6py/1gNqn\nYfmdA3y3Pee9pFepUHkfI135vZ+0O/xq0qu8XNB3zhWkXG/GvilJkiRJkiRJkiRJkiRJkiRJkiRJ\nkiRJkiRJkiRJkiRJkiRJkiTt3f8DwIGbCrkGj40AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x12edb7510>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "lf_stats = training_set.lf_stats()\n",
    "lf_stats[:5]\n",
    "lf_stats.hist(\"coverage\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "============================================================\n",
      "Testing mu = 1.00e-05, lf_w0 = 1.00e+00\n",
      "============================================================\n",
      "Begin training for rate=0.01, mu=1e-05\n",
      "\tLearning epoch = 0\tGradient mag. = 0.015879\n",
      "\tLearning epoch = 250\tGradient mag. = 0.021617\n",
      "\tLearning epoch = 500\tGradient mag. = 0.028912\n",
      "\tLearning epoch = 750\tGradient mag. = 0.039393\n",
      "Final gradient magnitude for rate=0.01, mu=1e-05: 0.054\n",
      "Applying LFs...\n",
      "Featurizing...\n",
      "============================================================\n",
      "Testing mu = 1.00e-05, lf_w0 = 2.00e+00\n",
      "============================================================\n",
      "Begin training for rate=0.01, mu=1e-05\n",
      "\tLearning epoch = 0\tGradient mag. = 0.031249\n",
      "\tLearning epoch = 250\tGradient mag. = 0.041009\n",
      "\tLearning epoch = 500\tGradient mag. = 0.051939\n",
      "\tLearning epoch = 750\tGradient mag. = 0.066466\n",
      "Final gradient magnitude for rate=0.01, mu=1e-05: 0.085\n",
      "============================================================\n",
      "Testing mu = 1.00e-07, lf_w0 = 1.00e+00\n",
      "============================================================\n",
      "Begin training for rate=0.01, mu=1e-07\n",
      "\tLearning epoch = 0\tGradient mag. = 0.015879\n",
      "\tLearning epoch = 250\tGradient mag. = 0.021679\n",
      "\tLearning epoch = 500\tGradient mag. = 0.029072\n",
      "\tLearning epoch = 750\tGradient mag. = 0.039686\n",
      "Final gradient magnitude for rate=0.01, mu=1e-07: 0.054\n",
      "============================================================\n",
      "Testing mu = 1.00e-07, lf_w0 = 2.00e+00\n",
      "============================================================\n",
      "Begin training for rate=0.01, mu=1e-07\n",
      "\tLearning epoch = 0\tGradient mag. = 0.031249\n",
      "\tLearning epoch = 250\tGradient mag. = 0.041060\n",
      "\tLearning epoch = 500\tGradient mag. = 0.052060\n",
      "\tLearning epoch = 750\tGradient mag. = 0.066679\n",
      "Final gradient magnitude for rate=0.01, mu=1e-07: 0.086\n"
     ]
    }
   ],
   "source": [
    "from snorkel.snorkel import Learner\n",
    "import snorkel.learning\n",
    "from snorkel.learning import LogReg\n",
    "\n",
    "learner = Learner(training_set, model=LogReg())\n",
    "\n",
    "# Splitting into CV and test set\n",
    "n_half = len(gold_candidates)/2\n",
    "test_candidates = gold_candidates[:n_half]\n",
    "test_labels     = gold_labels[:n_half]\n",
    "cv_candidates   = gold_candidates[n_half:]\n",
    "cv_labels       = gold_labels[n_half:]\n",
    "\n",
    "from snorkel.learning_utils import GridSearch\n",
    "gs       = GridSearch(learner, ['mu', 'lf_w0'], [[1e-5, 1e-7],[1.0,2.0]])\n",
    "gs_stats = gs.fit(cv_candidates, cv_labels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mu</th>\n",
       "      <th>lf_w0</th>\n",
       "      <th>Prec.</th>\n",
       "      <th>Rec.</th>\n",
       "      <th>F1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td> 0.00001</td>\n",
       "      <td> 1</td>\n",
       "      <td> 1</td>\n",
       "      <td> 1</td>\n",
       "      <td> 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td> 0.00001</td>\n",
       "      <td> 2</td>\n",
       "      <td> 1</td>\n",
       "      <td> 1</td>\n",
       "      <td> 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 1</td>\n",
       "      <td> 1</td>\n",
       "      <td> 1</td>\n",
       "      <td> 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 2</td>\n",
       "      <td> 1</td>\n",
       "      <td> 1</td>\n",
       "      <td> 1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        mu  lf_w0  Prec.  Rec.  F1\n",
       "0  0.00001      1      1     1   1\n",
       "1  0.00001      2      1     1   1\n",
       "2  0.00000      1      1     1   1\n",
       "3  0.00000      2      1     1   1\n",
       "\n",
       "[4 rows x 5 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gs_stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Applying LFs...\n",
      "Featurizing...\n",
      "========================================\n",
      "Test set size:\t28\n",
      "----------------------------------------\n",
      "Precision:\t1.0\n",
      "Recall:\t\t1.0\n",
      "F1 Score:\t1.0\n",
      "----------------------------------------\n",
      "TP: 14 | FP: 0 | TN: 14 | FN: 0\n",
      "========================================\n"
     ]
    }
   ],
   "source": [
    "learner.test_wmv(test_candidates, test_labels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Applying LFs...\n",
      "Featurizing...\n",
      "[('17903292', u'Candidate gene', u'PHENOTYPE'), ('17903292', u'PHENOTYPE', u'Candidate gene'), ('17903292', u'CST3', u'CysC'), ('17903292', u'CysC', u'CST3'), ('17903292', u'CST3', u'CysC'), ('17903292', u'CysC', u'CST3'), ('17903292', u'CST3', u'CysC'), ('17903292', u'CysC', u'CST3'), ('17903292', u'CST3', u'CysC'), ('17903292', u'CysC', u'CST3')]\n"
     ]
    }
   ],
   "source": [
    "preds = learner.predict_wmv(candidates)\n",
    "resolved_acronyms = [spair2uid(c) for (c, p) in zip(candidates, preds) if p == 1]\n",
    "print resolved_acronyms[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# mislabeled_cand = learner.mislabeled_test_candidates(test_candidates, test_labels)\n",
    "# for (c,p,g) in mislabeled_cand[:10]:\n",
    "#     print c.span0.context.document.name\n",
    "#     print c.span0.context    \n",
    "#     print c.span1.context\n",
    "#     print"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Classifying sentence acronyms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2950\n"
     ]
    }
   ],
   "source": [
    "from snorkel.candidates import UnionExtractor\n",
    "contexts = [corpus.get_tables(), corpus.get_phrases(), text_corpus.get_sentences()]\n",
    "extractors = [txt_tab_cells_extractor, txt_tab_ngram_extractor, txt_tab_ngram_extractor]\n",
    "joint_extractor = UnionExtractor(extractor_list=extractors, context_list=contexts)\n",
    "\n",
    "text_c = joint_extractor.union()\n",
    "print len(text_c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SpanPair(Span(\"(min-max)\", context=None, chars=[12,20], words=[2,4]), Span(\"Sample size (min-max)\", context=None, chars=[0,20], words=[0,4]))\n",
      "SpanPair(Span(\"(bp)\", context=None, chars=[18,21], words=[2,4]), Span(\"Physical Position (bp)\", context=None, chars=[0,21], words=[0,4]))\n",
      "SpanPair(Span(\"(Rank)\", context=None, chars=[10,15], words=[2,4]), Span(\"GEE p-val (Rank)\", context=None, chars=[0,15], words=[0,4]))\n",
      "SpanPair(Span(\"(bp)\", context=None, chars=[18,21], words=[2,4]), Span(\"Physical Position (bp)\", context=None, chars=[0,21], words=[0,4]))\n",
      "SpanPair(Span(\"(Rank)\", context=None, chars=[9,14], words=[2,4]), Span(\"GEE Pval (Rank)\", context=None, chars=[0,14], words=[0,4]))\n",
      "SpanPair(Span(\"(bp)\", context=None, chars=[18,21], words=[2,4]), Span(\"Physical Position (bp)\", context=None, chars=[0,21], words=[0,4]))\n",
      "SpanPair(Span(\"(min-max)\", context=None, chars=[12,20], words=[2,4]), Span(\"Sample size (min-max)\", context=None, chars=[0,20], words=[0,4]))\n",
      "SpanPair(Span(\"(min-max)\", context=None, chars=[12,20], words=[2,4]), Span(\"size (min-max)\", context=None, chars=[7,20], words=[1,4]))\n",
      "SpanPair(Span(\"(bp)\", context=None, chars=[18,21], words=[2,4]), Span(\"Physical Position (bp)\", context=None, chars=[0,21], words=[0,4]))\n",
      "SpanPair(Span(\"(bp)\", context=None, chars=[18,21], words=[2,4]), Span(\"Position (bp)\", context=None, chars=[9,21], words=[1,4]))\n",
      "SpanPair(Span(\"(Rank)\", context=None, chars=[10,15], words=[2,4]), Span(\"GEE p-val (Rank)\", context=None, chars=[0,15], words=[0,4]))\n",
      "SpanPair(Span(\"(Rank)\", context=None, chars=[10,15], words=[2,4]), Span(\"p-val (Rank)\", context=None, chars=[4,15], words=[1,4]))\n",
      "SpanPair(Span(\"(bp)\", context=None, chars=[18,21], words=[2,4]), Span(\"Physical Position (bp)\", context=None, chars=[0,21], words=[0,4]))\n",
      "SpanPair(Span(\"(bp)\", context=None, chars=[18,21], words=[2,4]), Span(\"Position (bp)\", context=None, chars=[9,21], words=[1,4]))\n",
      "SpanPair(Span(\"(Rank)\", context=None, chars=[9,14], words=[2,4]), Span(\"GEE Pval (Rank)\", context=None, chars=[0,14], words=[0,4]))\n",
      "SpanPair(Span(\"(Rank)\", context=None, chars=[9,14], words=[2,4]), Span(\"Pval (Rank)\", context=None, chars=[4,14], words=[1,4]))\n",
      "SpanPair(Span(\"(bp)\", context=None, chars=[18,21], words=[2,4]), Span(\"Physical Position (bp)\", context=None, chars=[0,21], words=[0,4]))\n",
      "SpanPair(Span(\"(bp)\", context=None, chars=[18,21], words=[2,4]), Span(\"Position (bp)\", context=None, chars=[9,21], words=[1,4]))\n",
      "SpanPair(Span(\"(bp)\", context=None, chars=[18,21], words=[2,4]), Span(\"Physical Position (bp)\", context=None, chars=[0,21], words=[0,4]))\n",
      "SpanPair(Span(\"(bp)\", context=None, chars=[18,21], words=[2,4]), Span(\"Position (bp)\", context=None, chars=[9,21], words=[1,4]))\n",
      "SpanPair(Span(\"(bp)\", context=None, chars=[18,21], words=[2,4]), Span(\"Physical Position (bp)\", context=None, chars=[0,21], words=[0,4]))\n",
      "SpanPair(Span(\"(bp)\", context=None, chars=[18,21], words=[2,4]), Span(\"Position (bp)\", context=None, chars=[9,21], words=[1,4]))\n",
      "SpanPair(Span(\"(bp)\", context=None, chars=[18,21], words=[2,4]), Span(\"Physical Position (bp)\", context=None, chars=[0,21], words=[0,4]))\n",
      "SpanPair(Span(\"(bp)\", context=None, chars=[18,21], words=[2,4]), Span(\"Position (bp)\", context=None, chars=[9,21], words=[1,4]))\n",
      "SpanPair(Span(\"(bp)\", context=None, chars=[18,21], words=[2,4]), Span(\"Physical Position (bp)\", context=None, chars=[0,21], words=[0,4]))\n",
      "SpanPair(Span(\"(bp)\", context=None, chars=[18,21], words=[2,4]), Span(\"Position (bp)\", context=None, chars=[9,21], words=[1,4]))\n",
      "SpanPair(Span(\"(CVD)\", context=None, chars=[136,140], words=[19,21]), Span(\"increased risk of cardiovascular disease (CVD)\", context=None, chars=[95,140], words=[14,21]))\n",
      "SpanPair(Span(\"(CVD)\", context=None, chars=[136,140], words=[19,21]), Span(\"risk of cardiovascular disease (CVD)\", context=None, chars=[105,140], words=[15,21]))\n",
      "SpanPair(Span(\"(CVD)\", context=None, chars=[136,140], words=[19,21]), Span(\"of cardiovascular disease (CVD)\", context=None, chars=[110,140], words=[16,21]))\n",
      "SpanPair(Span(\"(CVD)\", context=None, chars=[136,140], words=[19,21]), Span(\"cardiovascular disease (CVD)\", context=None, chars=[113,140], words=[17,21]))\n",
      "SpanPair(Span(\"(CVD)\", context=None, chars=[136,140], words=[19,21]), Span(\"disease (CVD)\", context=None, chars=[128,140], words=[18,21]))\n",
      "SpanPair(Span(\"(SNP)\", context=None, chars=[211,215], words=[30,32]), Span(\"GeneChip 100K single nucleotide polymorphism (SNP)\", context=None, chars=[166,215], words=[25,32]))\n",
      "SpanPair(Span(\"(SNP)\", context=None, chars=[211,215], words=[30,32]), Span(\"100K single nucleotide polymorphism (SNP)\", context=None, chars=[175,215], words=[26,32]))\n",
      "SpanPair(Span(\"(SNP)\", context=None, chars=[211,215], words=[30,32]), Span(\"single nucleotide polymorphism (SNP)\", context=None, chars=[180,215], words=[27,32]))\n",
      "SpanPair(Span(\"(SNP)\", context=None, chars=[211,215], words=[30,32]), Span(\"nucleotide polymorphism (SNP)\", context=None, chars=[187,215], words=[28,32]))\n",
      "SpanPair(Span(\"(SNP)\", context=None, chars=[211,215], words=[30,32]), Span(\"polymorphism (SNP)\", context=None, chars=[198,215], words=[29,32]))\n",
      "SpanPair(Span(\"(FHS)\", context=None, chars=[394,398], words=[59,61]), Span(\"approximately 1000 Framingham Heart Study (FHS)\", context=None, chars=[352,398], words=[54,61]))\n",
      "SpanPair(Span(\"(FHS)\", context=None, chars=[394,398], words=[59,61]), Span(\"1000 Framingham Heart Study (FHS)\", context=None, chars=[366,398], words=[55,61]))\n",
      "SpanPair(Span(\"(FHS)\", context=None, chars=[394,398], words=[59,61]), Span(\"Framingham Heart Study (FHS)\", context=None, chars=[371,398], words=[56,61]))\n",
      "SpanPair(Span(\"(FHS)\", context=None, chars=[394,398], words=[59,61]), Span(\"Heart Study (FHS)\", context=None, chars=[382,398], words=[57,61]))\n",
      "SpanPair(Span(\"(FHS)\", context=None, chars=[394,398], words=[59,61]), Span(\"Study (FHS)\", context=None, chars=[388,398], words=[58,61]))\n",
      "SpanPair(Span(\"(GEE)\", context=None, chars=[81,85], words=[8,10]), Span(\"using the generalized estimating equations (GEE)\", context=None, chars=[38,85], words=[3,10]))\n",
      "SpanPair(Span(\"(GEE)\", context=None, chars=[81,85], words=[8,10]), Span(\"the generalized estimating equations (GEE)\", context=None, chars=[44,85], words=[4,10]))\n",
      "SpanPair(Span(\"(GEE)\", context=None, chars=[81,85], words=[8,10]), Span(\"generalized estimating equations (GEE)\", context=None, chars=[48,85], words=[5,10]))\n",
      "SpanPair(Span(\"(GEE)\", context=None, chars=[81,85], words=[8,10]), Span(\"estimating equations (GEE)\", context=None, chars=[60,85], words=[6,10]))\n",
      "SpanPair(Span(\"(GEE)\", context=None, chars=[81,85], words=[8,10]), Span(\"equations (GEE)\", context=None, chars=[71,85], words=[7,10]))\n",
      "SpanPair(Span(\"(FBAT)\", context=None, chars=[118,123], words=[15,17]), Span(\"), family-based association test (FBAT)\", context=None, chars=[85,123], words=[10,17]))\n",
      "SpanPair(Span(\"(FBAT)\", context=None, chars=[118,123], words=[15,17]), Span(\", family-based association test (FBAT)\", context=None, chars=[86,123], words=[11,17]))\n",
      "SpanPair(Span(\"(FBAT)\", context=None, chars=[118,123], words=[15,17]), Span(\"family-based association test (FBAT)\", context=None, chars=[88,123], words=[12,17]))\n",
      "SpanPair(Span(\"(FBAT)\", context=None, chars=[118,123], words=[15,17]), Span(\"association test (FBAT)\", context=None, chars=[101,123], words=[13,17]))\n",
      "SpanPair(Span(\"(FBAT)\", context=None, chars=[118,123], words=[15,17]), Span(\"test (FBAT)\", context=None, chars=[113,123], words=[14,17]))\n",
      "SpanPair(Span(\"(LD)\", context=None, chars=[199,202], words=[37,39]), Span(\"and in complete linkage disequilibrium (LD)\", context=None, chars=[160,202], words=[32,39]))\n",
      "SpanPair(Span(\"(LD)\", context=None, chars=[199,202], words=[37,39]), Span(\"in complete linkage disequilibrium (LD)\", context=None, chars=[164,202], words=[33,39]))\n",
      "SpanPair(Span(\"(LD)\", context=None, chars=[199,202], words=[37,39]), Span(\"complete linkage disequilibrium (LD)\", context=None, chars=[167,202], words=[34,39]))\n",
      "SpanPair(Span(\"(LD)\", context=None, chars=[199,202], words=[37,39]), Span(\"linkage disequilibrium (LD)\", context=None, chars=[176,202], words=[35,39]))\n",
      "SpanPair(Span(\"(LD)\", context=None, chars=[199,202], words=[37,39]), Span(\"disequilibrium (LD)\", context=None, chars=[184,202], words=[36,39]))\n",
      "SpanPair(Span(\"(PAI-1)\", context=None, chars=[109,115], words=[17,19]), Span(\"], plasminogen activator inhibitor (PAI-1)\", context=None, chars=[74,115], words=[12,19]))\n",
      "SpanPair(Span(\"(PAI-1)\", context=None, chars=[109,115], words=[17,19]), Span(\", plasminogen activator inhibitor (PAI-1)\", context=None, chars=[75,115], words=[13,19]))\n",
      "SpanPair(Span(\"(PAI-1)\", context=None, chars=[109,115], words=[17,19]), Span(\"plasminogen activator inhibitor (PAI-1)\", context=None, chars=[77,115], words=[14,19]))\n",
      "SpanPair(Span(\"(PAI-1)\", context=None, chars=[109,115], words=[17,19]), Span(\"activator inhibitor (PAI-1)\", context=None, chars=[89,115], words=[15,19]))\n",
      "SpanPair(Span(\"(PAI-1)\", context=None, chars=[109,115], words=[17,19]), Span(\"inhibitor (PAI-1)\", context=None, chars=[99,115], words=[16,19]))\n",
      "SpanPair(Span(\"(vWF)\", context=None, chars=[146,150], words=[27,29]), Span(\"], von Willebrand factor (vWF)\", context=None, chars=[121,150], words=[22,29]))\n",
      "SpanPair(Span(\"(vWF)\", context=None, chars=[146,150], words=[27,29]), Span(\", von Willebrand factor (vWF)\", context=None, chars=[122,150], words=[23,29]))\n",
      "SpanPair(Span(\"(vWF)\", context=None, chars=[146,150], words=[27,29]), Span(\"von Willebrand factor (vWF)\", context=None, chars=[124,150], words=[24,29]))\n",
      "SpanPair(Span(\"(vWF)\", context=None, chars=[146,150], words=[27,29]), Span(\"Willebrand factor (vWF)\", context=None, chars=[128,150], words=[25,29]))\n",
      "SpanPair(Span(\"(vWF)\", context=None, chars=[146,150], words=[27,29]), Span(\"factor (vWF)\", context=None, chars=[139,150], words=[26,29]))\n",
      "SpanPair(Span(\"(tPA)\", context=None, chars=[186,190], words=[37,39]), Span(\"], tissue plasminogen activator (tPA)\", context=None, chars=[154,190], words=[32,39]))\n",
      "SpanPair(Span(\"(tPA)\", context=None, chars=[186,190], words=[37,39]), Span(\", tissue plasminogen activator (tPA)\", context=None, chars=[155,190], words=[33,39]))\n",
      "SpanPair(Span(\"(tPA)\", context=None, chars=[186,190], words=[37,39]), Span(\"tissue plasminogen activator (tPA)\", context=None, chars=[157,190], words=[34,39]))\n",
      "SpanPair(Span(\"(tPA)\", context=None, chars=[186,190], words=[37,39]), Span(\"plasminogen activator (tPA)\", context=None, chars=[164,190], words=[35,39]))\n",
      "SpanPair(Span(\"(tPA)\", context=None, chars=[186,190], words=[37,39]), Span(\"activator (tPA)\", context=None, chars=[176,190], words=[36,39]))\n",
      "SpanPair(Span(\"(FVII)\", context=None, chars=[212,217], words=[46,48]), Span(\"4,5,7], factor VII (FVII)\", context=None, chars=[193,217], words=[41,48]))\n",
      "SpanPair(Span(\"(FVII)\", context=None, chars=[212,217], words=[46,48]), Span(\"], factor VII (FVII)\", context=None, chars=[198,217], words=[42,48]))\n",
      "SpanPair(Span(\"(FVII)\", context=None, chars=[212,217], words=[46,48]), Span(\", factor VII (FVII)\", context=None, chars=[199,217], words=[43,48]))\n",
      "SpanPair(Span(\"(FVII)\", context=None, chars=[212,217], words=[46,48]), Span(\"factor VII (FVII)\", context=None, chars=[201,217], words=[44,48]))\n",
      "SpanPair(Span(\"(FVII)\", context=None, chars=[212,217], words=[46,48]), Span(\"VII (FVII)\", context=None, chars=[208,217], words=[45,48]))\n",
      "SpanPair(Span(\"(CHD)\", context=None, chars=[341,345], words=[73,75]), Span(\"markers for coronary heart disease (CHD)\", context=None, chars=[306,345], words=[68,75]))\n",
      "SpanPair(Span(\"(CHD)\", context=None, chars=[341,345], words=[73,75]), Span(\"for coronary heart disease (CHD)\", context=None, chars=[314,345], words=[69,75]))\n",
      "SpanPair(Span(\"(CHD)\", context=None, chars=[341,345], words=[73,75]), Span(\"coronary heart disease (CHD)\", context=None, chars=[318,345], words=[70,75]))\n",
      "SpanPair(Span(\"(CHD)\", context=None, chars=[341,345], words=[73,75]), Span(\"heart disease (CHD)\", context=None, chars=[327,345], words=[71,75]))\n",
      "SpanPair(Span(\"(CHD)\", context=None, chars=[341,345], words=[73,75]), Span(\"disease (CHD)\", context=None, chars=[333,345], words=[72,75]))\n",
      "SpanPair(Span(\"(CVD)\", context=None, chars=[388,392], words=[82,84]), Span(\"stroke and other cardiovascular disease (CVD)\", context=None, chars=[348,392], words=[77,84]))\n",
      "SpanPair(Span(\"(CVD)\", context=None, chars=[388,392], words=[82,84]), Span(\"and other cardiovascular disease (CVD)\", context=None, chars=[355,392], words=[78,84]))\n",
      "SpanPair(Span(\"(CVD)\", context=None, chars=[388,392], words=[82,84]), Span(\"other cardiovascular disease (CVD)\", context=None, chars=[359,392], words=[79,84]))\n",
      "SpanPair(Span(\"(CVD)\", context=None, chars=[388,392], words=[82,84]), Span(\"cardiovascular disease (CVD)\", context=None, chars=[365,392], words=[80,84]))\n",
      "SpanPair(Span(\"(CVD)\", context=None, chars=[388,392], words=[82,84]), Span(\"disease (CVD)\", context=None, chars=[380,392], words=[81,84]))\n",
      "SpanPair(Span(\"(HCT)\", context=None, chars=[221,225], words=[32,34]), Span(\"hematological phenotypes such as hematocrit (HCT)\", context=None, chars=[177,225], words=[27,34]))\n",
      "SpanPair(Span(\"(HCT)\", context=None, chars=[221,225], words=[32,34]), Span(\"(HCT), hemoglobin (Hgb)\", context=None, chars=[221,243], words=[32,39]))\n",
      "SpanPair(Span(\"(HCT)\", context=None, chars=[221,225], words=[32,34]), Span(\"phenotypes such as hematocrit (HCT)\", context=None, chars=[191,225], words=[28,34]))\n",
      "SpanPair(Span(\"(HCT)\", context=None, chars=[221,225], words=[32,34]), Span(\"such as hematocrit (HCT)\", context=None, chars=[202,225], words=[29,34]))\n",
      "SpanPair(Span(\"(HCT)\", context=None, chars=[221,225], words=[32,34]), Span(\"as hematocrit (HCT)\", context=None, chars=[207,225], words=[30,34]))\n",
      "SpanPair(Span(\"(HCT)\", context=None, chars=[221,225], words=[32,34]), Span(\"hematocrit (HCT)\", context=None, chars=[210,225], words=[31,34]))\n",
      "SpanPair(Span(\"(Hgb)\", context=None, chars=[239,243], words=[37,39]), Span(\"(HCT), hemoglobin (Hgb)\", context=None, chars=[221,243], words=[32,39]))\n",
      "SpanPair(Span(\"(Hgb)\", context=None, chars=[239,243], words=[37,39]), Span(\"HCT), hemoglobin (Hgb)\", context=None, chars=[222,243], words=[33,39]))\n",
      "SpanPair(Span(\"(Hgb)\", context=None, chars=[239,243], words=[37,39]), Span(\"), hemoglobin (Hgb)\", context=None, chars=[225,243], words=[34,39]))\n",
      "SpanPair(Span(\"(Hgb)\", context=None, chars=[239,243], words=[37,39]), Span(\", hemoglobin (Hgb)\", context=None, chars=[226,243], words=[35,39]))\n",
      "SpanPair(Span(\"(Hgb)\", context=None, chars=[239,243], words=[37,39]), Span(\"hemoglobin (Hgb)\", context=None, chars=[228,243], words=[36,39]))\n",
      "SpanPair(Span(\"(RBCC)\", context=None, chars=[267,272], words=[45,47]), Span(\", red blood cell count (RBCC)\", context=None, chars=[244,272], words=[40,47]))\n",
      "SpanPair(Span(\"(RBCC)\", context=None, chars=[267,272], words=[45,47]), Span(\"red blood cell count (RBCC)\", context=None, chars=[246,272], words=[41,47]))\n",
      "SpanPair(Span(\"(RBCC)\", context=None, chars=[267,272], words=[45,47]), Span(\"blood cell count (RBCC)\", context=None, chars=[250,272], words=[42,47]))\n",
      "SpanPair(Span(\"(RBCC)\", context=None, chars=[267,272], words=[45,47]), Span(\"cell count (RBCC)\", context=None, chars=[256,272], words=[43,47]))\n",
      "SpanPair(Span(\"(RBCC)\", context=None, chars=[267,272], words=[45,47]), Span(\"count (RBCC)\", context=None, chars=[261,272], words=[44,47]))\n",
      "SpanPair(Span(\"(MCV)\", context=None, chars=[308,312], words=[54,56]), Span(\"size, mean corpuscular volume (MCV)\", context=None, chars=[278,312], words=[49,56]))\n",
      "SpanPair(Span(\"(MCV)\", context=None, chars=[308,312], words=[54,56]), Span(\", mean corpuscular volume (MCV)\", context=None, chars=[282,312], words=[50,56]))\n",
      "SpanPair(Span(\"(MCV)\", context=None, chars=[308,312], words=[54,56]), Span(\"mean corpuscular volume (MCV)\", context=None, chars=[284,312], words=[51,56]))\n",
      "SpanPair(Span(\"(MCV)\", context=None, chars=[308,312], words=[54,56]), Span(\"corpuscular volume (MCV)\", context=None, chars=[289,312], words=[52,56]))\n",
      "SpanPair(Span(\"(MCV)\", context=None, chars=[308,312], words=[54,56]), Span(\"volume (MCV)\", context=None, chars=[301,312], words=[53,56]))\n",
      "SpanPair(Span(\"(MCH)\", context=None, chars=[346,350], words=[61,63]), Span(\") and mean corpuscular hemoglobin (MCH)\", context=None, chars=[312,350], words=[56,63]))\n",
      "SpanPair(Span(\"(MCH)\", context=None, chars=[346,350], words=[61,63]), Span(\"and mean corpuscular hemoglobin (MCH)\", context=None, chars=[314,350], words=[57,63]))\n",
      "SpanPair(Span(\"(MCH)\", context=None, chars=[346,350], words=[61,63]), Span(\"mean corpuscular hemoglobin (MCH)\", context=None, chars=[318,350], words=[58,63]))\n",
      "SpanPair(Span(\"(MCH)\", context=None, chars=[346,350], words=[61,63]), Span(\"corpuscular hemoglobin (MCH)\", context=None, chars=[323,350], words=[59,63]))\n",
      "SpanPair(Span(\"(MCH)\", context=None, chars=[346,350], words=[61,63]), Span(\"hemoglobin (MCH)\", context=None, chars=[335,350], words=[60,63]))\n",
      "SpanPair(Span(\"(ADP)\", context=None, chars=[442,446], words=[83,85]), Span(\"adenosine 5'-diphosphate (ADP)\", context=None, chars=[417,446], words=[78,85]))\n",
      "SpanPair(Span(\"(ADP)\", context=None, chars=[442,446], words=[83,85]), Span(\"(ADP), epinephrine (Epi)\", context=None, chars=[442,465], words=[83,90]))\n",
      "SpanPair(Span(\"(ADP)\", context=None, chars=[442,446], words=[83,85]), Span(\"5'-diphosphate (ADP)\", context=None, chars=[427,446], words=[79,85]))\n",
      "SpanPair(Span(\"(ADP)\", context=None, chars=[442,446], words=[83,85]), Span(\"'-diphosphate (ADP)\", context=None, chars=[428,446], words=[80,85]))\n",
      "SpanPair(Span(\"(ADP)\", context=None, chars=[442,446], words=[83,85]), Span(\"-diphosphate (ADP)\", context=None, chars=[429,446], words=[81,85]))\n",
      "SpanPair(Span(\"(ADP)\", context=None, chars=[442,446], words=[83,85]), Span(\"diphosphate (ADP)\", context=None, chars=[430,446], words=[82,85]))\n",
      "SpanPair(Span(\"(Epi)\", context=None, chars=[461,465], words=[88,90]), Span(\"(ADP), epinephrine (Epi)\", context=None, chars=[442,465], words=[83,90]))\n",
      "SpanPair(Span(\"(Epi)\", context=None, chars=[461,465], words=[88,90]), Span(\"ADP), epinephrine (Epi)\", context=None, chars=[443,465], words=[84,90]))\n",
      "SpanPair(Span(\"(Epi)\", context=None, chars=[461,465], words=[88,90]), Span(\"), epinephrine (Epi)\", context=None, chars=[446,465], words=[85,90]))\n",
      "SpanPair(Span(\"(Epi)\", context=None, chars=[461,465], words=[88,90]), Span(\", epinephrine (Epi)\", context=None, chars=[447,465], words=[86,90]))\n",
      "SpanPair(Span(\"(Epi)\", context=None, chars=[461,465], words=[88,90]), Span(\"epinephrine (Epi)\", context=None, chars=[449,465], words=[87,90]))\n",
      "SpanPair(Span(\"(FGB)\", context=None, chars=[67,71], words=[9,11]), Span(\"the following genes – fibrinogen-β (FGB)\", context=None, chars=[32,71], words=[4,11]))\n",
      "SpanPair(Span(\"(FGB)\", context=None, chars=[67,71], words=[9,11]), Span(\"(FGB), fibrinogen-α (FGA)\", context=None, chars=[67,91], words=[9,16]))\n",
      "SpanPair(Span(\"(FGB)\", context=None, chars=[67,71], words=[9,11]), Span(\"following genes – fibrinogen-β (FGB)\", context=None, chars=[36,71], words=[5,11]))\n",
      "SpanPair(Span(\"(FGB)\", context=None, chars=[67,71], words=[9,11]), Span(\"genes – fibrinogen-β (FGB)\", context=None, chars=[46,71], words=[6,11]))\n",
      "SpanPair(Span(\"(FGB)\", context=None, chars=[67,71], words=[9,11]), Span(\"– fibrinogen-β (FGB)\", context=None, chars=[52,71], words=[7,11]))\n",
      "SpanPair(Span(\"(FGB)\", context=None, chars=[67,71], words=[9,11]), Span(\"fibrinogen-β (FGB)\", context=None, chars=[54,71], words=[8,11]))\n",
      "SpanPair(Span(\"(FGA)\", context=None, chars=[87,91], words=[14,16]), Span(\"(FGB), fibrinogen-α (FGA)\", context=None, chars=[67,91], words=[9,16]))\n",
      "SpanPair(Span(\"(FGA)\", context=None, chars=[87,91], words=[14,16]), Span(\"(FGA), fibrinogen-γ (FGG)\", context=None, chars=[87,111], words=[14,21]))\n",
      "SpanPair(Span(\"(FGA)\", context=None, chars=[87,91], words=[14,16]), Span(\"FGB), fibrinogen-α (FGA)\", context=None, chars=[68,91], words=[10,16]))\n",
      "SpanPair(Span(\"(FGA)\", context=None, chars=[87,91], words=[14,16]), Span(\"), fibrinogen-α (FGA)\", context=None, chars=[71,91], words=[11,16]))\n",
      "SpanPair(Span(\"(FGA)\", context=None, chars=[87,91], words=[14,16]), Span(\", fibrinogen-α (FGA)\", context=None, chars=[72,91], words=[12,16]))\n",
      "SpanPair(Span(\"(FGA)\", context=None, chars=[87,91], words=[14,16]), Span(\"fibrinogen-α (FGA)\", context=None, chars=[74,91], words=[13,16]))\n",
      "SpanPair(Span(\"(FGG)\", context=None, chars=[107,111], words=[19,21]), Span(\"(FGA), fibrinogen-γ (FGG)\", context=None, chars=[87,111], words=[14,21]))\n",
      "SpanPair(Span(\"(FGG)\", context=None, chars=[107,111], words=[19,21]), Span(\"(FGG), FVII (F7)\", context=None, chars=[107,122], words=[19,26]))\n",
      "SpanPair(Span(\"(FGG)\", context=None, chars=[107,111], words=[19,21]), Span(\"FGA), fibrinogen-γ (FGG)\", context=None, chars=[88,111], words=[15,21]))\n",
      "SpanPair(Span(\"(FGG)\", context=None, chars=[107,111], words=[19,21]), Span(\"), fibrinogen-γ (FGG)\", context=None, chars=[91,111], words=[16,21]))\n",
      "SpanPair(Span(\"(FGG)\", context=None, chars=[107,111], words=[19,21]), Span(\", fibrinogen-γ (FGG)\", context=None, chars=[92,111], words=[17,21]))\n",
      "SpanPair(Span(\"(FGG)\", context=None, chars=[107,111], words=[19,21]), Span(\"fibrinogen-γ (FGG)\", context=None, chars=[94,111], words=[18,21]))\n",
      "SpanPair(Span(\"(F7)\", context=None, chars=[119,122], words=[24,26]), Span(\"(FGG), FVII (F7)\", context=None, chars=[107,122], words=[19,26]))\n",
      "SpanPair(Span(\"(F7)\", context=None, chars=[119,122], words=[24,26]), Span(\"FGG), FVII (F7)\", context=None, chars=[108,122], words=[20,26]))\n",
      "SpanPair(Span(\"(F7)\", context=None, chars=[119,122], words=[24,26]), Span(\"), FVII (F7)\", context=None, chars=[111,122], words=[21,26]))\n",
      "SpanPair(Span(\"(F7)\", context=None, chars=[119,122], words=[24,26]), Span(\", FVII (F7)\", context=None, chars=[112,122], words=[22,26]))\n",
      "SpanPair(Span(\"(F7)\", context=None, chars=[119,122], words=[24,26]), Span(\"FVII (F7)\", context=None, chars=[114,122], words=[23,26]))\n",
      "SpanPair(Span(\"(SERPINE1)\", context=None, chars=[135,144], words=[30,32]), Span(\"F7), and PAI-1 (SERPINE1)\", context=None, chars=[120,144], words=[25,32]))\n",
      "SpanPair(Span(\"(SERPINE1)\", context=None, chars=[135,144], words=[30,32]), Span(\"), and PAI-1 (SERPINE1)\", context=None, chars=[122,144], words=[26,32]))\n",
      "SpanPair(Span(\"(SERPINE1)\", context=None, chars=[135,144], words=[30,32]), Span(\", and PAI-1 (SERPINE1)\", context=None, chars=[123,144], words=[27,32]))\n",
      "SpanPair(Span(\"(SERPINE1)\", context=None, chars=[135,144], words=[30,32]), Span(\"and PAI-1 (SERPINE1)\", context=None, chars=[125,144], words=[28,32]))\n",
      "SpanPair(Span(\"(SERPINE1)\", context=None, chars=[135,144], words=[30,32]), Span(\"PAI-1 (SERPINE1)\", context=None, chars=[129,144], words=[29,32]))\n",
      "SpanPair(Span(\"(SNP)\", context=None, chars=[124,128], words=[17,19]), Span(\"Mapping 100K single nucleotide polymorphism (SNP)\", context=None, chars=[80,128], words=[12,19]))\n",
      "SpanPair(Span(\"(SNP)\", context=None, chars=[124,128], words=[17,19]), Span(\"100K single nucleotide polymorphism (SNP)\", context=None, chars=[88,128], words=[13,19]))\n",
      "SpanPair(Span(\"(SNP)\", context=None, chars=[124,128], words=[17,19]), Span(\"single nucleotide polymorphism (SNP)\", context=None, chars=[93,128], words=[14,19]))\n",
      "SpanPair(Span(\"(SNP)\", context=None, chars=[124,128], words=[17,19]), Span(\"nucleotide polymorphism (SNP)\", context=None, chars=[100,128], words=[15,19]))\n",
      "SpanPair(Span(\"(SNP)\", context=None, chars=[124,128], words=[17,19]), Span(\"polymorphism (SNP)\", context=None, chars=[111,128], words=[16,19]))\n",
      "SpanPair(Span(\"(GWAS)\", context=None, chars=[247,252], words=[37,39]), Span(\"conduct a genome-wide association study (GWAS)\", context=None, chars=[207,252], words=[32,39]))\n",
      "SpanPair(Span(\"(GWAS)\", context=None, chars=[247,252], words=[37,39]), Span(\"a genome-wide association study (GWAS)\", context=None, chars=[215,252], words=[33,39]))\n",
      "SpanPair(Span(\"(GWAS)\", context=None, chars=[247,252], words=[37,39]), Span(\"genome-wide association study (GWAS)\", context=None, chars=[217,252], words=[34,39]))\n",
      "SpanPair(Span(\"(GWAS)\", context=None, chars=[247,252], words=[37,39]), Span(\"association study (GWAS)\", context=None, chars=[229,252], words=[35,39]))\n",
      "SpanPair(Span(\"(GWAS)\", context=None, chars=[247,252], words=[37,39]), Span(\"study (GWAS)\", context=None, chars=[241,252], words=[36,39]))\n"
     ]
    }
   ],
   "source": [
    "for c in text_c:\n",
    "    if c.span0.context.document.name == '17903294':\n",
    "        print unicode(c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# helpers\n",
    "def spairtxt2uid(span_pair):\n",
    "    doc_id = span_pair.span0.context.document.name\n",
    "    str1 = span_pair.span1.get_span()\n",
    "    str2 = span_pair.span0.get_span()\n",
    "    acro = str2[1:-1]\n",
    "    phen = str1.split(' (')[0]\n",
    "    return (doc_id, acro, phen)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We really trust the first stage, so if an acronym has been resolved there, remove it from this stage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2950 2948\n"
     ]
    }
   ],
   "source": [
    "new_text_c = list()\n",
    "resolved = {(doc_id, acro) for doc_id, _, acro in resolved_acronyms}\n",
    "for c in text_c:\n",
    "    doc_id, phen, acro = spairtxt2uid(c)\n",
    "#     print doc_id, phen, acro\n",
    "    if (doc_id, acro) not in resolved:\n",
    "        new_text_c.append(c)\n",
    "\n",
    "print len(text_c), len(new_text_c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Building feature index...\n",
      "Extracting features...\n",
      "0/27654\n",
      "5000/27654\n",
      "10000/27654\n",
      "15000/27654\n",
      "20000/27654\n",
      "25000/27654\n"
     ]
    }
   ],
   "source": [
    "from snorkel.features import NgramPairFeaturizer\n",
    "\n",
    "pkl_f = 'acro_text_feats.pkl'\n",
    "try:\n",
    "    with open(pkl_f, 'rb') as f:\n",
    "        featurizer = cPickle.load(f)\n",
    "except:\n",
    "    featurizer = NgramPairFeaturizer()\n",
    "    featurizer.fit_transform(text_c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Training set size: 2755\n",
      "Gold set size: 195\n",
      "Positive labels in training set: 39\n",
      "Negative labels in training set: 165\n",
      "Positive labels in gold set: 48\n",
      "Negative labels in gold set: 147\n"
     ]
    }
   ],
   "source": [
    "# Split into train and test set\n",
    "training_candidates = []\n",
    "gold_candidates     = []\n",
    "gold_labels         = []\n",
    "n_half = len(text_c)/2\n",
    "for c in text_c[:n_half]:\n",
    "    uid = spair2uid(c)\n",
    "    if uid in annotations:\n",
    "        gold_candidates.append(c)\n",
    "        gold_labels.append(annotations[uid])\n",
    "    else:\n",
    "        training_candidates.append(c)\n",
    "training_candidates.extend(text_c[n_half:])\n",
    "gold_labels = np.array(gold_labels)\n",
    "print \"Training set size: %s\" % len(training_candidates)\n",
    "print \"Gold set size: %s\" % len(gold_candidates)\n",
    "print \"Positive labels in training set: %s\" % len([c for c in training_candidates if annotations.get(spair2uid(c),0)==1])\n",
    "print \"Negative labels in training set: %s\" % len([c for c in training_candidates if annotations.get(spair2uid(c),0)==-1])\n",
    "print \"Positive labels in gold set: %s\" % len([c for c in gold_candidates if annotations[spair2uid(c)]==1])\n",
    "print \"Negative labels in gold set: %s\" % len([c for c in gold_candidates if annotations[spair2uid(c)]==-1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "# positive LFs\n",
    "def LF_acro_matches(m):\n",
    "    _, acro, phen = spairtxt2uid(m)\n",
    "    words = phen.strip().split()\n",
    "    if len(acro) == len(words):\n",
    "        w_acro = ''.join([w[0] for w in words])\n",
    "        if w_acro.lower() == acro.lower():\n",
    "            return +1\n",
    "    return 0\n",
    "\n",
    "def LF_acro_matches_with_dashes(m):\n",
    "    _, acro, phen = spairtxt2uid(m)\n",
    "    words = re.split(' |-', phen)\n",
    "    if len(acro) == len(words):\n",
    "        w_acro = ''.join([w[0] for w in words])\n",
    "        if w_acro.lower() == acro.lower():\n",
    "            return +1\n",
    "    return 0\n",
    "\n",
    "def LF_acro_first_letter(m):\n",
    "    _, acro, phen = spairtxt2uid(m)\n",
    "    if not any(l.islower() for l in phen): return 0\n",
    "    words = phen.strip().split()\n",
    "    if len(acro) <= len(words):\n",
    "        if words[0].lower() == acro[0].lower():\n",
    "            return +1\n",
    "    return 0\n",
    "\n",
    "def LF_acro_prefix(m):\n",
    "    _, acro, phen = spairtxt2uid(m)\n",
    "    phen = phen.replace('-', '')\n",
    "    if phen[:2].lower() == acro[:2].lower():\n",
    "        return +1\n",
    "    return 0\n",
    "\n",
    "def LF_acro_matches_last_letters(m):\n",
    "    _, acro, phen = spairtxt2uid(m)\n",
    "    words = phen.strip().split()\n",
    "    prev_words = m.span1.pre_window(d=1) + words\n",
    "    w_prev_acro = ''.join([w[0] for w in prev_words])\n",
    "    if w_prev_acro.lower() == acro.lower(): return 0\n",
    "    for r in (1,2):\n",
    "        new_acro = acro[r:]\n",
    "        if len(new_acro) < 3: continue\n",
    "        if len(new_acro) == len(words):\n",
    "            w_acro = ''.join([w[0] for w in words])\n",
    "            if w_acro.lower() == new_acro.lower():\n",
    "                return +1\n",
    "    return 0\n",
    "\n",
    "def LF_full_cell(m):\n",
    "    \"\"\"If only phrase in cell is A B C (XYZ), then it's correct\"\"\"\n",
    "    if not hasattr(m.span1.context, 'cell'): return 0\n",
    "    return 1 if m.span1.context.cell.text == m.span1.get_span() else 0\n",
    "\n",
    "def LF_start(m):\n",
    "    punc = ',.;!?()\\'\"'\n",
    "    if m.span1.get_word_start() == 0 or any(c in punc for c in m.span1.pre_window(d=1)):\n",
    "        _, acro, phen = spairtxt2uid(m)\n",
    "        if phen[0].lower() == acro[0].lower(): \n",
    "            return +1\n",
    "    return 0\n",
    "\n",
    "LF_txt_pos = [LF_acro_matches, LF_acro_matches_with_dashes, LF_acro_first_letter, LF_acro_prefix, LF_acro_matches_last_letters, LF_full_cell, LF_start]\n",
    "\n",
    "# negative LFs\n",
    "def LF_no_pos(m):\n",
    "    return -1 if not any(LF(m) for LF in LF_txt_pos) else 0\n",
    "\n",
    "def LF_short(m):\n",
    "    _, acro, phen = spairtxt2uid(m)\n",
    "    return -1 if len(acro) == 1 else 0\n",
    "\n",
    "def LF_lc(m):\n",
    "    _, acro, phen = spairtxt2uid(m)\n",
    "    return -1 if all(l.islower() for l in acro) else 0\n",
    "\n",
    "def LF_uc(m):\n",
    "    _, acro, phen = spairtxt2uid(m)\n",
    "    return -2 if not any(l.islower() for l in phen) else 0\n",
    "\n",
    "def LF_punc(m):\n",
    "    _, acro, phen = spairtxt2uid(m)\n",
    "    punc = ',.;!?()'\n",
    "    return -1 if any(c in punc for c in phen) else 0\n",
    "    \n",
    "\n",
    "LF_txt_neg = [LF_no_pos, LF_short, LF_lc, LF_uc, LF_punc]\n",
    "\n",
    "LF_txt = LF_txt_pos + LF_txt_neg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Applying LFs...\n",
      "Featurizing...\n",
      "Building feature index...\n",
      "Extracting features...\n",
      "0/26916\n",
      "5000/26916\n",
      "10000/26916\n",
      "15000/26916\n",
      "20000/26916\n",
      "25000/26916\n",
      "============================================================\n",
      "LF Summary Statistics: 12 LFs applied to 2755 candidates\n",
      "------------------------------------------------------------\n",
      "Coverage (candidates w/ > 0 labels):\t\t91.25%\n",
      "Overlap (candidates w/ > 1 labels):\t\t65.01%\n",
      "Conflict (candidates w/ conflicting labels):\t39.93%\n",
      "============================================================\n"
     ]
    }
   ],
   "source": [
    "from snorkel.snorkel import TrainingSet\n",
    "from snorkel.features import NgramFeaturizer\n",
    "\n",
    "training_set = TrainingSet(training_candidates, LF_txt, featurizer=NgramPairFeaturizer())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[<matplotlib.axes.AxesSubplot object at 0x131b0ad10>]], dtype=object)"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAArwAAAEKCAYAAADwwYa2AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAF9ZJREFUeJzt3X2wZHddJvBnkslIhkwYXCiBZWB8IYgUboACxZf1KilF\nBcRSF0KhzGJZAqvoWrvlSylh112q1AJSauFb5MUSEQW0CMQtUNIIhaKruTG8ruNyTbIrARYCySYr\ngcz+cU5nem56ps+993u6T5/7+VTdmv5N9+3+5cnpvt/p+3R3AgAAAAAAAAAAAAAAAAAAAAAAAAAA\nAAAAAADsyXmr3gDAGHlwBTjTsSRvTvLxJJ9M8itJDiT52SRbSW5J8tokF7eX/29J/t2267g+yTPa\n01+Z5B1J/k+SDyf5vpnLvSbJryW5JsntSTaSfGeS65J8JsmNSa7Ydt0/kOQf271N9/Tk9rwDSX4q\nycn2/DckuX/n/3IAAEbv/DTD6suSXJjkUJKvT/K8JH+f5HiS+yZ5U5Lfab/n+5O8Z+Y6virJp5Nc\n0F72piTPTfMEw2OTfKK9TNIMvLcmeVK7/qIk35Tk0e36MUk+luS7Zq77tiRf117/LyX5XJJvac//\n8STvTfKQ9vxfT/J7u8gBAICRelKaZ3a3//brz5I8f2Z9SZpB87wkR9I8O3usPe+/JrmqPf3MJH++\n7bp+I8mL29Ovab/O5cokL29PvzjJ62bOuzDJP+f0wPvBmdNJ8uCZfQLsWx4EAU47lqYucPe2v39w\n+/dTNyY5mORL0jzj+rYkl7fnPTOnh9KHJ/maNM/4Tr+e3X5fkpxK8wzwrK9Jcm2awfvWJD+c5F+0\n5z0kyc0zl70zTVVi6niSP5q5rQ8m+fzM7QHsSwZegNNuSvKwNNWGWf87zTA59bA0g+Qt7fr1aQbe\nJ6V51vXa9u9vTPKuND3a6deR3LvzO+v3kvxxkocmOZqmlnBgZh8PnbnshTk9DE9v7ynbbu9wkn86\nx+0BALCPnJdkM0039nCS+6Tp8P5gkv+RZui9KMkbc7rDmzRd308leXua/u/URWleVPacNJ3aC5I8\nIc0L2ZKmzvDz2/ZwS5oXpiXJE9v19LYeneSzaQbrQ0l+Mffu8F6bZiBPkgcmeXq3/3QAAPaLY2lq\nAZ9M8wKzK9M8w/pzaZ5B/XiaAfR+277vqiRfSPL4bX9/SZK35vS7Pvxpkq9uz3t1kv+87fLfk2ZI\n/mySq5P8cs4crp+bM9+l4eY0Q3naff77NO8G8dk079bwX7r9ZwPsb49M8xY506/PJHnRSncEQNI8\ng3xXmq4wAEXOS9MFO7boggD04mlp6hb3TdPv/ZvVbgdgfL41Z77fJADL9Vtp3oHh1jQfaPGI1W4H\nYHxeleSFq94EAAD04VCaF3A8cNUbAQCArg7u4LLfnqYr9oltf3+qbjsAAHBW/5DkK3b6TTv54InL\n07y5+hyn9vnXy/OCF/x4Tp06teevK664ouR6fMlTnsP+kqU8h/wlT1kO9SvJl+9gdr1H14H3cJLL\nkrx5NzdCd1tbW6vewqjIs5Y868iyljxrybOOLIeha6XhjiQP6HMjAADQh51UGliCEydOrHoLoyLP\nWvKsI8ta8qwlzzqyHIYDBddxyuvWXpEXvODGvPKVr1j1RgAARuvAgQPJLuZXz/AOzGQyWfUWRkWe\nteRZR5a15FlLnnVkOQwGXgAARk2loYRKAwBA31QaAABgDgPvwOj61JJnLXnWkWUtedaSZx1ZDoOB\nFwCAUdPhLaHDCwDQNx1eAACYw8A7MLo+teRZS551ZFlLnrXkWUeWw2DgBQBg1HR4S+jwAgD0TYcX\nAADmMPAOjK5PLXnWkmcdWdaSZy151pHlMBh4AQAYNR3eEjq8AAB90+EFAIA5DLwDo+tTS5615FlH\nlrXkWUuedWQ5DAZeAABGTYe3hA4vAEDfdHgBAGAOA+/A6PrUkmctedaRZS151pJnHVkOQ5eB92iS\nNyb5UJIPJvnaXncEAACFunQgXpvkXUleleRgkvsm+czM+Tq8OrwAAL3bbYf34ILzL07yjUme264/\nnzOHXQAAGLRFlYYvS/KJJK9O8rdJfivJ4b43tZ/p+tSSZy151pFlLXnWkmcdWQ7Domd4DyZ5XJIf\nSfLXSa5M8lNJXnzmxU4kOd6ePprk0iQb7XrS/jnm9ckkh5pVe2BvbGzsar25ubmn77c+cy3P2rU8\nra2trXe2nhrKftZtPT29tbWVvVjUgXhQkr9I8qXt+hvSDLxPnbmMDq8OLwBA7/p6H96PJbkpySXt\n+rIkH9jpjQAAwKosGniT5EeTvC7J9Um+OslLe93RPrf9VyDsjTxrybOOLGvJs5Y868hyGBZ1eJNm\n0H1C3xsBAIA+7LgDMYcOrw4vAEDv+urwAgDAWjPwDoyuTy151pJnHVnWkmctedaR5TAYeAEAGDUd\n3hI6vAAAfdPhBQCAOQy8A6PrU0ueteRZR5a15FlLnnVkOQwGXgAARk2Ht4QOLwBA33R4AQBgDgPv\nwOj61JJnLXnWkWUtedaSZx1ZDoOBFwCAUdPhLaHDCwDQNx1eAACYw8A7MLo+teRZS551ZFlLnrXk\nWUeWw2DgBQBg1HR4S+jwAgD0TYcXAADmMPAOjK5PLXnWkmcdWdaSZy151pHlMBh4AQAYNR3eEjq8\nAAB90+EFAIA5DLwDo+tTS5615FlHlrXkWUuedWQ5DAc7Xm4ryWeTfCHJXUme2NeGAACgUtcOxEeT\nPD7Jp+acp8OrwwsA0LtldHgrXuAGAABL1XXgPZXk7Un+e5If6m876PrUkmctedaRZS151pJnHVkO\nQ9cO79cl+ViSByZ5R5IPJ3n36bNPJDnenj6a5NIkG+160v455vXJJIeaVXtgb2xs7Gq9ubm5p++3\nPnMtz9q1PK2tra13tp4ayn7WbT09vbW1lb3YTU3hiiS3J3lZu9bh1eEFAOhdnx3ew0mOtKfvm+Rb\nk9yw0xsCAIBV6DLwfkma+sJmkvcleWuaPi892P4rEPZGnrXkWUeWteRZS551ZDkMXTq8H01TygUA\ngLVT8VZjOrw6vAAAvVvG+/ACAMDaMfAOjK5PLXnWkmcdWdaSZy151pHlMBh4AQAYNR3eEjq8AAB9\n0+EFAIA5DLwDo+tTS5615FlHlrXkWUuedWQ5DAZeAABGTYe3hA4vAEDfdHgBAGAOA+/A6PrUkmct\nedaRZS151pJnHVkOg4EXAIBR0+EtocMLANA3HV4AAJjDwDswuj615FlLnnVkWUueteRZR5bDYOAF\nAGDUdHhL6PACAPRNhxcAAOYw8A6Mrk8tedaSZx1Z1pJnLXnWkeUwGHgBABg1Hd4SOrwAAH3T4QUA\ngDkMvAOj61NLnrXkWUeWteRZS551ZDkMXQfe85Ncl+TqHvcCAADlunYgfiLJ45McSfL0befp8Orw\nAgD0rs8O70OTfEeSq3ZzAwAAsEpdBt5XJPmPSe7ueS9E16eaPGvJs44sa8mzljzryHIYDi44/6lJ\nPp6mv7tx9oudSHK8PX00yaUzF5+0f455fTLJoWbVHtgbGxu7Wm9ubu7p+63PXMuzdi1Pa2tr652t\np4ayn3VbT09vbW1lLxZVFF6a5PuTfD7JfZJcnORNSX5g5jI6vDq8AAC966vD+zNJjiX50iTPSvLO\nnDnsAgDAoC0aeLfb70/l9m77r0DYG3nWkmcdWdaSZy151pHlMCzq8M56V/sFAABro+JtxnR4dXgB\nAHrX5/vwAgDA2jLwDoyuTy151pJnHVnWkmctedaR5TAYeAEAGDUd3hI6vAAAfdPhBQCAOQy8A6Pr\nU0ueteRZR5a15FlLnnVkOQwGXgAARk2Ht4QOLwBA33R4AQBgDgPvwOj61JJnLXnWkWUtedaSZx1Z\nDoOBFwCAUdPhLaHDCwDQNx1eAACYw8A7MLo+teRZS551ZFlLnrXkWUeWw2DgBQBg1HR4S+jwAgD0\nTYcXAADmMPAOjK5PLXnWkmcdWdaSZy151pHlMBh4AQAYNR3eEjq8AAB90+EFAIA5DLwDo+tTS561\n5FlHlrXkWUuedWQ5DF0G3vskeV+SzSTvT/KSPjcEAACVunYgDie5I8nBJO9J8mNphuBEhzc6vAAA\n/eu7w3tH++ehJBckuXunNwQAAKvQdeA9L02l4ZYkb0/y173taJ/T9aklz1ryrCPLWvKsJc86shyG\ngx0vd3eSS5PcL8kfJXl0kg+cPvtEkuPt6aPtRTfa9aT9c8zrk2me/D59YG9sbOxqvbm5uafvtz5z\nLc/atTytra2td7aeGsp+1m09Pb21tZW92M378L44yf9N8rJ2rcOrwwsA0Ls+O7wPSPO0bZJcmOSy\nJB/a6Q0BAMAqdBl4H5zknUmuT/JXaTq81/S5qf1s+69A2Bt51pJnHVnWkmctedaR5TB06fDekORx\nfW8EAAD6sJsO73Y6vDq8AAC96/t9eAEAYC0ZeAdG16eWPGvJs44sa8mzljzryHIYDLwAAIyaDm8J\nHV4AgL7p8AIAwBwG3oHR9aklz1ryrCPLWvKsJc86shwGAy8AAKOmw1tChxcAoG86vAAAMIeBd2B0\nfWrJs5Y868iyljxrybOOLIfBwAsAwKjp8JbQ4QUA6JsOLwAAzGHgHRhdn1ryrCXPOrKsJc9a8qwj\ny2Ew8AIAMGo6vCV0eAEA+qbDCwAAcxh4B0bXp5Y8a8mzjixrybOWPOvIchgMvAAAjJoObwkdXgCA\nvunwAgDAHAbegdH1qSXPWvKsI8ta8qwlzzqyHIYuA++xJNcm+WCS9yd5Ua87AgCAQl06EA9qvzaT\nXJTkb5I8I8mH2vN1eHV4AQB612eH92Npht0kuT3NoPuQnd4QAACswk47vMeTPDbJ++q3QqLrU02e\nteRZR5a15FlLnnVkOQwHd3DZi5K8McmPpXmmd8aJNLNwkhxNcmmSjXY9af8c8/pkkkPNqj2wNzY2\ndrXe3Nzc0/even348JHceee2w2OfufDCi3LHHbclWf3/j+r1uh+f1tbW1steTw1lP+u2np7e2trK\nXnTtQFyQ5K1J/iTJldvO0+HV4b1H063Z78fDgZw6td8zAIB6fXZ4DyT57TTv0rB92AUAgEHrMvB+\nfZLnJPnmJNe1X0/pc1P72fZfgbBXk1VvYFQcn3VkWUueteRZR5bD0KXD+574gAoAANbUjjsQc+jw\n6vDeQ4c30eEFgH702eEFAIC1ZeAdGF2fapNVb2BUHJ91ZFlLnrXkWUeWw2DgBQBg1HR4S+jwTunw\nJjq8ANAPHV4AAJjDwDswuj7VJqvewKg4PuvIspY8a8mzjiyHwcALAMCo6fCW0OGd0uFNdHgBoB86\nvAAAMIeBd2B0fapNVr2BUXF81pFlLXnWkmcdWQ6DgRcAgFHT4S2hwzulw5vo8AJAP3R4AQBgDgPv\nwOj6VJusegOj4visI8ta8qwlzzqyHAYDLwAAo6bDW0KHd0qHN9HhBYB+6PACAMAcBt6B0fWpNln1\nBkbF8VlHlrXkWUuedWQ5DAZeAABGTYe3hA7vlA5vosMLAP3Q4QUAgDkMvAOj61NtsuoNjIrjs44s\na8mzljzryHIYugy8r0pyS5Ibet4LAACU69KB+MYktyf5nSSPmXO+Dq8O7z10eBMdXgDoR58d3ncn\n+fROrxgAAIZAh3dgdH2qTVa9gVFxfNaRZS151pJnHVkOw8GaqzmR5Hh7+miSS5NstOtJ++eY1yeT\nHGpW7YG9sbGxq/Xm5uaevn/V68Ykw/n/s7mi229XA/v/s9+Pz2WvDx8+kjvvvD1ckOSuVW9ipY4c\nuX/e8pY3JxnO8Wm9nPXU7PkXX/zFue22/f3L8wsvvCjXXHN1p/wmk0m2trb2dHtdOxDHk1wdHd6z\n0OGd0uFNdHiZcn+YkoPHBWZ5bEh2e5/wPrwAADBHl4H39Unem+SSJDcl+be97mif2/4rEPZqsuoN\njIrjs9Jk1RsYmcmqNzAq7ut1ZDkMXTq8l/e+CwAA6MmOOxBz6PDq8N5DLynR1WPK/WFKDh4XmOWx\nIdHhBQCAQgbegdH1qTZZ9QZGxfFZabLqDYzMZNUbGBX39TqyHAYDLwAAo6bDW0KHd0ovKdHVY8r9\nYUoOHheY5bEh0eEFAIBCBt6B0fWpNln1BkbF8VlpsuoNjMxk1RsYFff1OrIcBgMvAACjpsNbQod3\nSi8p0dVjyv1hSg4eF5jlsSHR4QUAgEIG3oHR9ak2WfUGRsXxWWmy6g2MzGTVGxgV9/U6shwGAy8A\nAKOmw1tCh3dKLynR1WPK/WFKDh4XmOWxIdHhBQCAQgbegdH1qTZZ9QZGxfFZabLqDYzMZNUbGBX3\n9TqyHAYDLwAAo6bDW0KHd0ovKdHVY8r9YUoOHheY5bEh0eEFAIBCBt6B0fWpNln1BkbF8VlpsuoN\njMxk1RsYFff1OrIcBgMvAACjpsNbQod3Si8p0dVjyv1hSg4eF5jlsSHR4QUAgEJdBt6nJPlwkr9P\n8pP9bgddn2qTVW9gVByflSar3sDITFa9gVFxX68jy2FYNPCen+RX0wy9X5Xk8iSP6ntT+9nm5uaq\ntzAy8qzk+Kwky1ryrOS+XkeWw7Bo4H1ikpNJtpLcleT3k3xXz3va12699dZVb2Fk5FnJ8VlJlrXk\nWcl9vY4sh2HRwPsvk9w0s765/TsAAFgLB3fxPfd6Sd3FFz+tYCvr63Of+585//ynlFzX1tZWyfUw\ntbXqDYyK47PS1qo3MDJbq97AqLiv15HlMCx6W4evTfKSNB3eJPnpJHcn+YWZy5xM8uXlOwMAgDP9\nQ5KvqL7Sg+0VH09yKM2rArxoDQCAUfn2JB9J80zuT694LwAAAAAA7MaiD6D4oiRvaM//yyQPX97W\n1tKiPP91kr9N83Zw37PEfa2rRXn+RJIPJLk+yZ8medjytrZ2FmX5/CR/l+S6JO+OmtMiXT+853vT\nvEbiccvY1BpblOeJJJ9Ic3xel+R5S9vZ+ulybP6bNI+d70/yuiXta10tyvPlOX1cfiTJp5e3tbW0\nKM+HJbk2zax0fZpGwp6dn6bScDzJBZnf5X1hkle2p5+Z5j17ma9Lng9P8pgkr42Bd5EueW4kuU97\n+vlxfJ5NlyyPzJx+WpI/WcrO1lOXPJMm0z9P8t4YeM+lS57PTfLLy93WWuqS5SPSDBP3a9cPWNbm\n1lDX+/rUjyS5qv9tra0uef5mkh9uTz8qyUfPdYVdPlo46fYBFE9PM5wlyZuSPLnjde9HXfL8xyQ3\npHnGh3Prkuckyf9rT78vyUOXtLd10yXL22ZOXxTH6Ll0/fCen0/z7jf/nMXvnrOfdcnzQGTYRZcs\nfyjNp61+pl1/clmbW0M7/aCuZyd5ff/bWltd8rw7p/8xdjTJ/zrXFXYdeLt8AMXsZT6f5g7yxR2v\nf7/xgR61dprnDya5ptcdra+uWb4wzYPRLyR50RL2ta665PnY9u/e1q7v9V7n3KNLnqfS/Fbs+iR/\nGP+4PZsuWT4iySOTvCfJXyT5tuVsbS3t5OfQw9M8c/nOnve0zrrk+ZIkz2kv97YkP3quK+w68M6z\n/UF53r+oPXB3J6taZ8vzOWl+ZfxLS9zLupuX5SvTvA/iTyb5ueVuZ+3N5nleklck+Q8zf+fZyZ3Z\nfnxenWag+Fdp+vqvvdd3cDbbszyY5n7+TUkuT/Mr+Ptt/ybO6mw/h56V5h9jfu7vzPa8np3k1UmO\nJfmOJL97rm/uOvDe3F7h1LHc+6njm3P6hUAH09wpFLLn65LnLHeKc+ua52VJfiZN/eauJexrHe30\n2HxDkmf0uqP1tijPI0kenaZy89E0H/bzlujxnk2X4/NTOX3/virJ45ewr3XU9ef6W5J8Ic2vlj+S\nHt7wfyR28tj5zKgzLNIlz+cl+YP29F+meZ3OnnvmXT6A4oVJfq09/ax4UdC57OQDPV4TL1pbpEue\nj41PBeyiS5azP/CeluSvlrKz9bTTD++5Nobdc+mS54NmTn93mhcCcm9dsvy2ND+DkmaQuDHJ/Zez\nvbXT9b7+yCx4cRVJuuV5TZoXqaY975wd3p2Y9wEU/ynND7ykeVuyP8jptyU7XnXDI7Uozyek6aXc\nnuaFAjcse4Nr5mx5PrU9/Y4k/5TTbwnzx8ve4BpZdGxemeYtiq5L8mfxtmSLLMpzloF3sUV5vjTN\n8bmZ5vi8ZNkbXCNdjs2XpXlbsr9L8xZlnF2XPK9Ic4yy2KI8H5WmX76Z5ufRZcveIAAAAAAAAAAA\nAAAAAAAAAAAAAAAAAAAAAEvw/wHzIPBNHlSrtAAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x13009e450>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "lf_stats = training_set.lf_stats()\n",
    "lf_stats[:5]\n",
    "lf_stats.hist(\"coverage\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "============================================================\n",
      "Testing mu = 1.00e-05, lf_w0 = 1.00e+00\n",
      "============================================================\n",
      "Begin training for rate=0.01, mu=1e-05\n",
      "\tLearning epoch = 0\tGradient mag. = 0.046642\n",
      "\tLearning epoch = 250\tGradient mag. = 0.064083\n",
      "\tLearning epoch = 500\tGradient mag. = 0.084458\n",
      "\tLearning epoch = 750\tGradient mag. = 0.110822\n",
      "Final gradient magnitude for rate=0.01, mu=1e-05: 0.144\n",
      "Applying LFs...\n",
      "Featurizing...\n",
      "============================================================\n",
      "Testing mu = 1.00e-05, lf_w0 = 2.00e+00\n",
      "============================================================\n",
      "Begin training for rate=0.01, mu=1e-05\n",
      "\tLearning epoch = 0\tGradient mag. = 0.070882\n",
      "\tLearning epoch = 250\tGradient mag. = 0.087098\n",
      "\tLearning epoch = 500\tGradient mag. = 0.101449\n",
      "\tLearning epoch = 750\tGradient mag. = 0.116702\n",
      "Final gradient magnitude for rate=0.01, mu=1e-05: 0.135\n",
      "============================================================\n",
      "Testing mu = 1.00e-07, lf_w0 = 1.00e+00\n",
      "============================================================\n",
      "Begin training for rate=0.01, mu=1e-07\n",
      "\tLearning epoch = 0\tGradient mag. = 0.046642\n",
      "\tLearning epoch = 250\tGradient mag. = 0.064100\n",
      "\tLearning epoch = 500\tGradient mag. = 0.084495\n",
      "\tLearning epoch = 750\tGradient mag. = 0.110882\n",
      "Final gradient magnitude for rate=0.01, mu=1e-07: 0.144\n",
      "============================================================\n",
      "Testing mu = 1.00e-07, lf_w0 = 2.00e+00\n",
      "============================================================\n",
      "Begin training for rate=0.01, mu=1e-07\n",
      "\tLearning epoch = 0\tGradient mag. = 0.070882\n",
      "\tLearning epoch = 250\tGradient mag. = 0.087108\n",
      "\tLearning epoch = 500\tGradient mag. = 0.101468\n",
      "\tLearning epoch = 750\tGradient mag. = 0.116727\n",
      "Final gradient magnitude for rate=0.01, mu=1e-07: 0.135\n"
     ]
    }
   ],
   "source": [
    "from snorkel.snorkel import Learner\n",
    "import snorkel.learning\n",
    "from snorkel.learning import LogReg\n",
    "\n",
    "text_learner = Learner(training_set, model=LogReg())\n",
    "\n",
    "# Splitting into CV and test set\n",
    "n_half = len(gold_candidates)/2\n",
    "test_candidates = gold_candidates[:n_half]\n",
    "test_labels     = gold_labels[:n_half]\n",
    "cv_candidates   = gold_candidates[n_half:]\n",
    "cv_labels       = gold_labels[n_half:]\n",
    "\n",
    "from snorkel.learning_utils import GridSearch\n",
    "gs       = GridSearch(text_learner, ['mu', 'lf_w0'], [[1e-5, 1e-7],[1.0,2.0]])\n",
    "gs_stats = gs.fit(cv_candidates, cv_labels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mu</th>\n",
       "      <th>lf_w0</th>\n",
       "      <th>Prec.</th>\n",
       "      <th>Rec.</th>\n",
       "      <th>F1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td> 0.00001</td>\n",
       "      <td> 1</td>\n",
       "      <td> 0.909091</td>\n",
       "      <td> 0.833333</td>\n",
       "      <td> 0.869565</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td> 0.00001</td>\n",
       "      <td> 2</td>\n",
       "      <td> 0.846154</td>\n",
       "      <td> 0.916667</td>\n",
       "      <td> 0.880000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 1</td>\n",
       "      <td> 0.909091</td>\n",
       "      <td> 0.833333</td>\n",
       "      <td> 0.869565</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 2</td>\n",
       "      <td> 0.846154</td>\n",
       "      <td> 0.916667</td>\n",
       "      <td> 0.880000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        mu  lf_w0     Prec.      Rec.        F1\n",
       "0  0.00001      1  0.909091  0.833333  0.869565\n",
       "1  0.00001      2  0.846154  0.916667  0.880000\n",
       "2  0.00000      1  0.909091  0.833333  0.869565\n",
       "3  0.00000      2  0.846154  0.916667  0.880000\n",
       "\n",
       "[4 rows x 5 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gs_stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Applying LFs...\n",
      "Featurizing...\n",
      "========================================\n",
      "Test set size:\t97\n",
      "----------------------------------------\n",
      "Precision:\t0.615384615385\n",
      "Recall:\t\t1.0\n",
      "F1 Score:\t0.761904761905\n",
      "----------------------------------------\n",
      "TP: 24 | FP: 15 | TN: 58 | FN: 0\n",
      "========================================\n"
     ]
    }
   ],
   "source": [
    "text_learner.test_wmv(test_candidates, test_labels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Applying LFs...\n",
      "Featurizing...\n",
      "[('17903292', u'GFR', u'Glomerular Filtration Rate'), ('17903292', u'TSH', u'Thyroid stimulation hormone'), ('17903292', u'DHEAS', u'Dehydroepiandrosterone sulfate'), ('17903292', u'Mb', u'Physical Location'), ('17903292', u'Mb', u'Physical Location'), ('17903292', u'Mb', u'Physical location'), ('17903292', u'GEE', u'Mean p-value'), ('17903292', u'FBAT', u'Mean p-value'), ('17903292', u'GEE', u'P_value'), ('17903294', u'min-max', u'Sample size')]\n",
      "17903292 1.0 -1\n",
      "Phrase('17903292', 1, 3, 0, u'Physical Location (Mb)')\n",
      "(Mb) Physical Location (Mb)\n",
      "[0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0]\n",
      "\n",
      "17903292 1.0 -1\n",
      "Phrase('17903292', 1, 189, 0, u'Physical Location (Mb)')\n",
      "(Mb) Physical Location (Mb)\n",
      "[0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0]\n",
      "\n",
      "17903292 1.0 -1\n",
      "Phrase('17903292', 1, 383, 0, u'Physical location (Mb)')\n",
      "(Mb) Physical location (Mb)\n",
      "[0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0]\n",
      "\n",
      "17903292 1.0 -1\n",
      "Phrase('17903292', 2, 5, 0, u'Mean p-value (GEE)')\n",
      "(GEE) Mean p-value (GEE)\n",
      "[0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0]\n",
      "\n",
      "17903292 1.0 -1\n",
      "Phrase('17903292', 2, 6, 0, u'Mean p-value (FBAT)')\n",
      "(FBAT) Mean p-value (FBAT)\n",
      "[0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0]\n",
      "\n",
      "17903292 1.0 -1\n",
      "Phrase('17903292', 3, 7, 0, u'P_value (GEE)')\n",
      "(GEE) P_value (GEE)\n",
      "[0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0]\n",
      "\n",
      "17903297 1.0 -1\n",
      "Phrase('17903297', 0, 93, 0, u'Similarities (Sim)')\n",
      "(Sim) Similarities (Sim)\n",
      "[0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0]\n",
      "\n",
      "17903300 1.0 -1\n",
      "Phrase('17903300', 1, 3, 0, u'Physical Position (Mb)')\n",
      "(Mb) Physical Position (Mb)\n",
      "[0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0]\n",
      "\n",
      "17903300 1.0 -1\n",
      "Phrase('17903300', 1, 190, 0, u'Physical Location (Mb)')\n",
      "(Mb) Physical Location (Mb)\n",
      "[0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0]\n",
      "\n",
      "17903292 1.0 -1\n",
      "Phrase('17903292', 1, 3, 0, u'Physical Location (Mb)')\n",
      "(Mb) Physical Location (Mb)\n",
      "[0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0]\n",
      "\n",
      "17903292 1.0 -1\n",
      "Phrase('17903292', 1, 189, 0, u'Physical Location (Mb)')\n",
      "(Mb) Physical Location (Mb)\n",
      "[0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0]\n",
      "\n",
      "17903292 1.0 -1\n",
      "Phrase('17903292', 1, 383, 0, u'Physical location (Mb)')\n",
      "(Mb) Physical location (Mb)\n",
      "[0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0]\n",
      "\n",
      "17903292 1.0 -1\n",
      "Phrase('17903292', 2, 5, 0, u'Mean p-value (GEE)')\n",
      "(GEE) Mean p-value (GEE)\n",
      "[0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0]\n",
      "\n",
      "17903292 1.0 -1\n",
      "Phrase('17903292', 2, 6, 0, u'Mean p-value (FBAT)')\n",
      "(FBAT) Mean p-value (FBAT)\n",
      "[0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0]\n",
      "\n",
      "17903292 1.0 -1\n",
      "Phrase('17903292', 3, 7, 0, u'P_value (GEE)')\n",
      "(GEE) P_value (GEE)\n",
      "[0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0]\n",
      "\n",
      "17903297 1.0 -1\n",
      "Phrase('17903297', 0, 87, 0, u'Boston Naming Test (Nam)')\n",
      "(Nam) Naming Test (Nam)\n",
      "[0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0]\n",
      "\n",
      "17903297 1.0 -1\n",
      "Phrase('17903297', 0, 93, 0, u'Similarities (Sim)')\n",
      "(Sim) Similarities (Sim)\n",
      "[0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0]\n",
      "\n",
      "17903298 -1.0 1\n",
      "Phrase('17903298', 0, 18, 0, u'28 yr time averaged FPG (tFPG)')\n",
      "(tFPG) time averaged FPG (tFPG)\n",
      "[0, 0, 0, 0, 0, 0, 0, -1, 0, 0, 0, 0]\n",
      "\n",
      "17903300 1.0 -1\n",
      "Phrase('17903300', 1, 3, 0, u'Physical Position (Mb)')\n",
      "(Mb) Physical Position (Mb)\n",
      "[0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0]\n",
      "\n",
      "17903300 1.0 -1\n",
      "Phrase('17903300', 1, 190, 0, u'Physical Location (Mb)')\n",
      "(Mb) Physical Location (Mb)\n",
      "[0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0]\n",
      "\n",
      "17903301 -1.0 1\n",
      "Phrase('17903301', 0, 50, 0, u'Stage 2 Exercise diastolic blood pressure (DBP)')\n",
      "(DBP) blood pressure (DBP)\n",
      "[0, 0, 0, 0, 0, 0, 0, -1, 0, 0, 0, 0]\n",
      "\n",
      "17903292 1.0 -1\n",
      "Sentence(Document('17903292', Corpus (GWAS Text Corpus)), 3, u'We tested for association between the Affymetrix GeneChip Human Mapping 100K single nucleotide polymorphism (SNP) set and measures of kidney function and endocrine traits.')\n",
      "(SNP) single nucleotide polymorphism (SNP)\n",
      "[1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]\n",
      "\n",
      "17903294 -1.0 1\n",
      "Sentence(Document('17903294', Corpus (GWAS Text Corpus)), 1, u'Increased circulating levels of hemostatic factors as well as anemia have been associated with increased risk of cardiovascular disease (CVD).')\n",
      "(CVD) cardiovascular disease (CVD)\n",
      "[0, 0, 0, 0, 0, 0, 0, -1, 0, 0, 0, 0]\n",
      "\n",
      "17903294 1.0 -1\n",
      "Sentence(Document('17903294', Corpus (GWAS Text Corpus)), 3, u'We sought to confirm known putative loci and identify novel loci that may influence either trait in genome-wide association and linkage analyses using the Affymetrix GeneChip 100K single nucleotide polymorphism (SNP) set.')\n",
      "(SNP) single nucleotide polymorphism (SNP)\n",
      "[1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]\n",
      "\n",
      "17903294 -1.0 1\n",
      "Sentence(Document('17903294', Corpus (GWAS Text Corpus)), 14, u'Elevated circulating levels of hemostatic factors, such as fibrinogen [1-3], plasminogen activator inhibitor (PAI-1) [4,5], von Willebrand factor (vWF) [6], tissue plasminogen activator (tPA) [4,5,7], factor VII (FVII) [8], and D-dimer [9,10] are linked to the development of atherothrombosis and are risk markers for coronary heart disease (CHD), stroke and other cardiovascular disease (CVD) events.')\n",
      "(CVD) cardiovascular disease (CVD)\n",
      "[0, 0, 0, 0, 0, 0, 0, -1, 0, 0, 0, 0]\n",
      "\n",
      "17903294 1.0 -1\n",
      "Sentence(Document('17903294', Corpus (GWAS Text Corpus)), 21, u'The recent completion of a genome-wide scan using the Affymetrix GeneChip Human Mapping 100K single nucleotide polymorphism (SNP) set on participants in the Framingham Heart Study offered the opportunity to conduct a genome-wide association study (GWAS) and linkage scan for variants that influence hemostatic factors and hematological phenotypes.')\n",
      "(SNP) single nucleotide polymorphism (SNP)\n",
      "[1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]\n",
      "\n",
      "17903300 -1.0 1\n",
      "Sentence(Document('17903300', Corpus (GWAS Text Corpus)), 1, u'Obesity is related to multiple cardiovascular disease (CVD) risk factors as well as CVD and has a strong familial component.')\n",
      "(CVD) cardiovascular disease (CVD)\n",
      "[0, 0, 0, 0, 0, 0, 0, -1, 0, 0, 0, 0]\n",
      "\n",
      "17903301 1.0 -1\n",
      "Sentence(Document('17903301', Corpus (GWAS Text Corpus)), 7, u'Generalized estimating equations (GEE), family-based association tests (FBAT) and variance-components linkage were used to relate multivariable-adjusted trait residuals to 70,987 SNPs (Human 100K GeneChip, Affymetrix) restricted to autosomal SNPs with minor allele frequency \\u22650.10, genotype call rate \\u22650.80, and Hardy-Weinberg equilibrium p \\u2265 0.001.')\n",
      "(GEE) Generalized estimating equations (GEE)\n",
      "[1, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0]\n",
      "\n",
      "17903304 1.0 -1\n",
      "Sentence(Document('17903304', Corpus (GWAS Text Corpus)), 2, u'In many industrialized countries, cardiovascular disease (CVD) claims more lives each year than any other disease.')\n",
      "(CVD) cardiovascular disease (CVD)\n",
      "[0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "text_preds = text_learner.predict_wmv(text_c)\n",
    "text_acronyms = [spairtxt2uid(c) for (c, p) in zip(text_c, text_preds) if p == 1]\n",
    "print text_acronyms[:10]\n",
    "mislabeled_cand = [(c,p, annotations.get(spair2uid(c), None)) for c, p in zip(text_c, text_preds) if p != annotations.get(spair2uid(c), p)]\n",
    "for (c,p,g) in mislabeled_cand[:50]:\n",
    "    _, phen, acro = spairtxt2uid(c)\n",
    "    print c.span0.context.document.name, p, g\n",
    "    print c.span0.context    \n",
    "    print c.span0.get_span(), c.span1.get_span()\n",
    "    print [LF(c) for LF in LF_txt]\n",
    "    print"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# punc=','\n",
    "# for (c) in (text_c):\n",
    "#     if c.span0.context.document.name != '17903294': continue\n",
    "#     _, phen, acro = spairtxt2uid(c)\n",
    "#     print c.span0.context.document.name\n",
    "#     print c.span0.context    \n",
    "#     print c.span0.get_span(), c.span1.get_span()\n",
    "#     print [LF(c) for LF in LF_txt]\n",
    "#     print"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### Collect all the candidates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Applying LFs...\n",
      "Featurizing...\n",
      "[('17903292', u'Candidate gene', u'PHENOTYPE'), ('17903292', u'PHENOTYPE', u'Candidate gene'), ('17903292', u'CST3', u'CysC'), ('17903292', u'CysC', u'CST3'), ('17903292', u'CST3', u'CysC'), ('17903292', u'CysC', u'CST3'), ('17903292', u'CST3', u'CysC'), ('17903292', u'CysC', u'CST3'), ('17903292', u'CST3', u'CysC'), ('17903292', u'CysC', u'CST3')]\n",
      "Applying LFs...\n",
      "Featurizing...\n",
      "[('17903292', u'GFR', u'Glomerular Filtration Rate'), ('17903292', u'TSH', u'Thyroid stimulation hormone'), ('17903292', u'DHEAS', u'Dehydroepiandrosterone sulfate'), ('17903292', u'Mb', u'Physical Location'), ('17903292', u'Mb', u'Physical Location'), ('17903292', u'Mb', u'Physical location'), ('17903292', u'GEE', u'Mean p-value'), ('17903292', u'FBAT', u'Mean p-value'), ('17903292', u'GEE', u'P_value'), ('17903294', u'min-max', u'Sample size')]\n"
     ]
    }
   ],
   "source": [
    "table_preds = learner.predict_wmv(candidates)\n",
    "table_acronyms = [spair2uid(c) for (c, p) in zip(candidates, preds) if p == 1]\n",
    "print table_acronyms[:10]\n",
    "\n",
    "text_preds = text_learner.predict_wmv(text_c)\n",
    "text_acronyms = [spairtxt2uid(c) for (c, p) in zip(text_c, text_preds) if p == 1]\n",
    "print text_acronyms[:10]\n",
    "\n",
    "acronyms = table_acronyms + text_acronyms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "693"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(acronyms)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# store relations to annotate\n",
    "with open('acronyms.extracted.all.tsv', 'w') as f:\n",
    "    for doc_id, str1, str2 in acronyms:\n",
    "        try:\n",
    "            out = u'{}\\t{}\\t{}\\n'.format(doc_id, unicode(str2), str1)\n",
    "            f.write(out.encode(\"UTF-8\"))\n",
    "        except:\n",
    "            print 'ERROR:', str1, str2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
