{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Phenotype acronym extraction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import sys\n",
    "import cPickle\n",
    "import numpy as np\n",
    "\n",
    "# import snorkel and gwasdb\n",
    "sys.path.append('../snorkel')\n",
    "sys.path.append('../src')\n",
    "sys.path.append('../src/crawler')\n",
    "\n",
    "# set up paths\n",
    "abstract_dir = '../data/db/papers'\n",
    "\n",
    "# set up matplotlib\n",
    "import matplotlib\n",
    "%matplotlib inline\n",
    "matplotlib.rcParams['figure.figsize'] = (12,4)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load corpus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from snorkel.parser import XMLDocParser\n",
    "from extractor.parser import UnicodeXMLTableDocParser\n",
    "\n",
    "xml_parser = UnicodeXMLTableDocParser(\n",
    "    path=abstract_dir,\n",
    "    doc='./*',\n",
    "    text='.//table',\n",
    "    id='.//article-id[@pub-id-type=\"pmid\"]/text()',\n",
    "    keep_xml_tree=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 54.7 s, sys: 2.99 s, total: 57.7 s\n",
      "Wall time: 1min 24s\n"
     ]
    }
   ],
   "source": [
    "from snorkel.parser import CorpusParser, OmniParser\n",
    "import cPickle\n",
    "\n",
    "table_parser = OmniParser()\n",
    "\n",
    "corpus_name = 'gwas-table-corpus.pkl'\n",
    "try:\n",
    "    with open(corpus_name,\"r\") as pkl:\n",
    "        corpus = cPickle.load(pkl)\n",
    "except:\n",
    "    cp = CorpusParser(xml_parser, table_parser, max_docs=15)\n",
    "    %time corpus = cp.parse_corpus(name='GWAS Corpus')\n",
    "    # pickling currently doesn't work...\n",
    "#     with open(corpus_name,\"w\") as pkl:\n",
    "#         corpus = cPickle.dump(corpus, pkl)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Candidate extraction"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### From tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from snorkel.matchers import DictionaryMatch, Union, CellNameMatcher, CellDictNameMatcher\n",
    "from snorkel.candidates import EntityExtractor\n",
    "from snorkel.candidates import TableNgrams, CellSpace\n",
    "\n",
    "# Define a candidate space\n",
    "cells = CellSpace()\n",
    "\n",
    "# Create a list of possible words that could denote phenotypes\n",
    "acro_words = ['abbreviation', 'acronym']\n",
    "phen_words = ['trait', 'phenotype']\n",
    "\n",
    "# Define matchers\n",
    "phen_matcher = CellDictNameMatcher(axis='col', d=phen_words, n_max=3, ignore_case=True)\n",
    "acro_matcher = CellDictNameMatcher(axis='col', d=acro_words, n_max=3, ignore_case=True)\n",
    "\n",
    "phen_extractor = EntityExtractor(cells, phen_matcher)\n",
    "acro_extractor = EntityExtractor(cells, acro_matcher)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from snorkel.candidates import AlignedTableRelationExtractor\n",
    "relation_extractor = AlignedTableRelationExtractor(phen_extractor, acro_extractor, axis='row', induced=False)\n",
    "tables = corpus.get_tables()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 23min 2s, sys: 11.2 s, total: 23min 13s\n",
      "Wall time: 23min 20s\n",
      "58 relations extracted, e.g.\n",
      "SpanPair(Span(\"CD40 Ligand, serum & plasma\", context=None, chars=[0,26], words=[0,5]), Span(\"CD40L\", context=None, chars=[0,4], words=[0,0]))\n",
      "SpanPair(Span(\"2\", context=None, chars=[0,0], words=[0,0]), Span(\"CD40L\", context=None, chars=[0,4], words=[0,0]))\n",
      "SpanPair(Span(\"C-reactive protein\", context=None, chars=[0,17], words=[0,1]), Span(\"CRP\", context=None, chars=[0,2], words=[0,0]))\n",
      "SpanPair(Span(\"5\", context=None, chars=[0,0], words=[0,0]), Span(\"CRP\", context=None, chars=[0,2], words=[0,0]))\n",
      "SpanPair(Span(\"Intercellular adhesion molecule-1\", context=None, chars=[0,32], words=[0,2]), Span(\"ICAM1\", context=None, chars=[0,4], words=[0,0]))\n"
     ]
    }
   ],
   "source": [
    "%time candidates = relation_extractor.extract(tables, name='all')\n",
    "print \"%s relations extracted, e.g.\" % len(candidates)\n",
    "for cand in candidates[:5]: \n",
    "    print cand"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table('17903293', 0) SpanPair(Span(\"CD40 Ligand, serum & plasma\", context=None, chars=[0,26], words=[0,5]), Span(\"CD40L\", context=None, chars=[0,4], words=[0,0]))\n",
      "Table('17903293', 0) SpanPair(Span(\"2\", context=None, chars=[0,0], words=[0,0]), Span(\"CD40L\", context=None, chars=[0,4], words=[0,0]))\n",
      "Table('17903293', 0) SpanPair(Span(\"C-reactive protein\", context=None, chars=[0,17], words=[0,1]), Span(\"CRP\", context=None, chars=[0,2], words=[0,0]))\n",
      "Table('17903293', 0) SpanPair(Span(\"5\", context=None, chars=[0,0], words=[0,0]), Span(\"CRP\", context=None, chars=[0,2], words=[0,0]))\n",
      "Table('17903293', 0) SpanPair(Span(\"Intercellular adhesion molecule-1\", context=None, chars=[0,32], words=[0,2]), Span(\"ICAM1\", context=None, chars=[0,4], words=[0,0]))\n"
     ]
    }
   ],
   "source": [
    "for cand in candidates[:5]: \n",
    "    print cand.span0.context.table, cand"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### From text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 524 ms, sys: 145 ms, total: 668 ms\n",
      "Wall time: 8.12 s\n"
     ]
    }
   ],
   "source": [
    "from snorkel.parser import SentenceParser, CorpusParser\n",
    "from extractor.parser import UnicodeXMLDocParser, GWASXMLDocParser\n",
    "\n",
    "xml_parser = GWASXMLDocParser(\n",
    "    path=abstract_dir,\n",
    "    doc='./*',\n",
    "    title='.//front//article-title//text()',\n",
    "    abstract='.//abstract//p//text()',\n",
    "    par1='.//body/p[1]//text()',\n",
    "    id='.//article-id[@pub-id-type=\"pmid\"]/text()',\n",
    "    keep_xml_tree=True)\n",
    "\n",
    "sent_parser = SentenceParser()\n",
    "cp = CorpusParser(xml_parser, sent_parser, max_docs=15)\n",
    "%time text_corpus = cp.parse_corpus(name='GWAS Text Corpus')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from snorkel.candidates import Ngrams\n",
    "from snorkel.matchers import RegexMatchSpan\n",
    "from snorkel.candidates import EntityExtractor, RelationExtractor\n",
    "\n",
    "# Define a candidate space\n",
    "ngrams6 = Ngrams(n_max=6)\n",
    "ngrams3 = Ngrams(n_max=3)\n",
    "\n",
    "# Define matchers\n",
    "phen_matcher = RegexMatchSpan(rgx=r'.+ \\([a-zA-Z]{0,5}[\\);]')\n",
    "acro_matcher = RegexMatchSpan(rgx=r'\\([a-zA-Z]{0,5}[\\);]')\n",
    "\n",
    "# Extractors\n",
    "phen_txt_extractor = EntityExtractor(ngrams6, phen_matcher)\n",
    "acro_txt_extractor = EntityExtractor(ngrams3, acro_matcher)\n",
    "\n",
    "# Filtering function\n",
    "def overlap_filter_fn(span0, span1):\n",
    "    start0, end0 = span0.char_start, span0.char_end\n",
    "    start1, end1 = span1.char_start, span1.char_end\n",
    "    return True if start0 <= start1 <= end1 <= end0 else False\n",
    "\n",
    "# Relation extractor\n",
    "txt_tab_extractor = RelationExtractor(phen_txt_extractor, acro_txt_extractor, filter_fn=overlap_filter_fn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Extract acroynms from tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15369\n"
     ]
    },
    {
     "ename": "IndexError",
     "evalue": "list index out of range",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-35-d22060ddb5bf>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;32mprint\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mS\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mlong_S\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0ms\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0ms\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mS\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0ms\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcell\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m \u001b[0;32mprint\u001b[0m \u001b[0mlong_S\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m: list index out of range"
     ]
    }
   ],
   "source": [
    "S=corpus.get_phrases()\n",
    "print len(S)\n",
    "long_S = [s for s in S if not s.cell]\n",
    "print long_S[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 1.1 s, sys: 260 ms, total: 1.36 s\n",
      "Wall time: 2 s\n",
      "347 candidates extracted from text in tables\n"
     ]
    }
   ],
   "source": [
    "# %time txt_tab_c = acro_txt_extractor.extract(corpus.get_sentences(), name='all')\n",
    "%time txt_tab_c = txt_tab_extractor.extract(corpus.get_phrases(), name='all')\n",
    "print len(txt_tab_c), 'candidates extracted from text in tables'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Extract acroynms from abstracts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 753 ms, sys: 16.2 ms, total: 770 ms\n",
      "Wall time: 781 ms\n",
      "169 candidates extracted from text in abstracts\n"
     ]
    }
   ],
   "source": [
    "%time txt_txt_c = txt_tab_extractor.extract(text_corpus.get_sentences(), name='all')\n",
    "print len(txt_txt_c), 'candidates extracted from text in abstracts'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "for cand in txt_tab_c[:100]:\n",
    "    print cand.span0.context.document.name, cand"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Learning the correctness of relations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from snorkel.candidates import UnionExtractor\n",
    "\n",
    "contexts = [corpus.get_tables(), corpus.get_sentences(), text_corpus.get_sentences()]\n",
    "extractors = [relation_extractor, txt_tab_extractor, txt_tab_extractor]\n",
    "joint_extractor = UnionExtractor(extractor_list=extractors, context_list=contexts)\n",
    "all_c = joint_extractor.union()\n",
    "print len(all_c)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Creating a gold set"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To create a gold set, we save all extracted relations into a csv file. We annotate it manually, and save the result to a second file. It contains pairs of phenotype and rsid strings; if that file exists, we take these as gold truth."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'all_c' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-21-d1ba58fb1031>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# store relations to annotate\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;32mwith\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'acronyms.unannotated.tsv'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'w'\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m     \u001b[0;32mfor\u001b[0m \u001b[0mspan_pair\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mall_c\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m         \u001b[0mdoc_id\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mspan_pair\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mspan0\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcontext\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdocument\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m         \u001b[0mstr1\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mspan_pair\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mspan0\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_span\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'all_c' is not defined"
     ]
    }
   ],
   "source": [
    "# store relations to annotate\n",
    "with open('acronyms.unannotated.tsv', 'w') as f:\n",
    "    for span_pair in all_c:\n",
    "        doc_id = span_pair.span0.context.document.name\n",
    "        str1 = span_pair.span0.get_span()\n",
    "        str2 = span_pair.span1.get_span()\n",
    "        try:\n",
    "            f.write('%s\\t%s\\t%s\\n' % (doc_id, str1, str2))\n",
    "        except:\n",
    "            continue"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "We now load the results of our annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "annotations = dict()\n",
    "with open('acronyms.anotated.txt') as f:\n",
    "    text = f.read()\n",
    "    for line in text.split('\\r'):\n",
    "        doc_id, str1, str2, res = line.strip().split('\\t')\n",
    "        res = 1 if int(res) == 1 else -1\n",
    "        annotations[(doc_id, str1, str2)] = res"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Classifying table acronyms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Building feature index...\n",
      "Extracting features...\n",
      "0/2579\n"
     ]
    }
   ],
   "source": [
    "from snorkel.features import TableNgramPairFeaturizer\n",
    "\n",
    "pkl_f = 'acro_table_feats.pkl'\n",
    "try:\n",
    "    with open(pkl_f, 'rb') as f:\n",
    "        featurizer = cPickle.load(f)\n",
    "except:\n",
    "    featurizer = TableNgramPairFeaturizer()\n",
    "    featurizer.fit_transform(candidates)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Training set size: 30\n",
      "Gold set size: 28\n",
      "Positive labels in training set: 22\n",
      "Negative labels in training set: 7\n",
      "Positive labels in gold set: 14\n",
      "Negative labels in gold set: 14\n"
     ]
    }
   ],
   "source": [
    "def spair2uid(span_pair):\n",
    "    doc_id = span_pair.span0.context.document.name\n",
    "    str1 = span_pair.span0.get_span()\n",
    "    str2 = span_pair.span1.get_span()\n",
    "    return (doc_id, str1, str2)\n",
    "\n",
    "# Split into train and test set\n",
    "training_candidates = []\n",
    "gold_candidates     = []\n",
    "gold_labels         = []\n",
    "n_half = len(candidates)/2\n",
    "for c in candidates[:n_half]:\n",
    "    uid = spair2uid(c)\n",
    "    if uid in annotations:\n",
    "        gold_candidates.append(c)\n",
    "        gold_labels.append(annotations[uid])\n",
    "    else:\n",
    "        training_candidates.append(c)\n",
    "training_candidates.extend(candidates[n_half:])\n",
    "gold_labels = np.array(gold_labels)\n",
    "print \"Training set size: %s\" % len(training_candidates)\n",
    "print \"Gold set size: %s\" % len(gold_candidates)\n",
    "print \"Positive labels in training set: %s\" % len([c for c in training_candidates if annotations.get(spair2uid(c),0)==1])\n",
    "print \"Negative labels in training set: %s\" % len([c for c in training_candidates if annotations.get(spair2uid(c),0)==-1])\n",
    "print \"Positive labels in gold set: %s\" % len([c for c in gold_candidates if annotations[spair2uid(c)]==1])\n",
    "print \"Negative labels in gold set: %s\" % len([c for c in gold_candidates if annotations[spair2uid(c)]==-1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def LF1_digits(m):\n",
    "    txt = m.span0.get_span()\n",
    "    frac_num = len([ch for ch in txt if ch.isdigit()]) / float(len(txt))\n",
    "    return -1 if frac_num > 0.5 else +1\n",
    "def LF1_short(m):\n",
    "    txt = m.span0.get_span()\n",
    "    return -1 if len(txt) < 5 else 0\n",
    "\n",
    "LF_tables = [LF1_digits, LF1_short]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Library/Python/2.7/site-packages/matplotlib/__init__.py:1155: UserWarning:  This call to matplotlib.use() has no effect\n",
      "because the backend has already been chosen;\n",
      "matplotlib.use() must be called *before* pylab, matplotlib.pyplot,\n",
      "or matplotlib.backends is imported for the first time.\n",
      "\n",
      "  warnings.warn(_use_error_msg)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Applying LFs...\n",
      "Featurizing...\n",
      "Building feature index...\n",
      "Extracting features...\n",
      "0/1926\n",
      "============================================================\n",
      "LF Summary Statistics: 2 LFs applied to 30 candidates\n",
      "------------------------------------------------------------\n",
      "Coverage (candidates w/ > 0 labels):\t\t100.00%\n",
      "Overlap (candidates w/ > 1 labels):\t\t23.33%\n",
      "Conflict (candidates w/ conflicting labels):\t0.00%\n",
      "============================================================\n"
     ]
    }
   ],
   "source": [
    "from snorkel.snorkel import TrainingSet\n",
    "from snorkel.features import NgramFeaturizer\n",
    "\n",
    "training_set = TrainingSet(training_candidates, LF_tables, featurizer=TableNgramPairFeaturizer())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[<matplotlib.axes.AxesSubplot object at 0x12885efd0>]], dtype=object)"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAsUAAAEKCAYAAAD6nSjkAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAGgBJREFUeJzt3XuQZGV5x/HvXkBcIKwRSwUWpkqQLCRxwQoXU8oYSIFG\nREsjLoVhAqG0DFrxkkJS6kKMSWGibBHKC6G8lQpETSwUFEX3RI0iibDrBRiBcgLLGhR10cTogpA/\n3tM7vb3T0z29z/Q75z3fT9VIn+4zs4+/mt19tufXp0GSJEmSJEmSJEmSJEmSJEmSJEmSJEmSJEmS\nJEmSJEmSJEmSJEmSJKkBluceQJJK5B+ukrRwa4B/AX4IPAj8I7AMeDMwAzwAfAj4jfr8zwF/3vM1\ntgAvqm//FvAF4MfAncAfd533QeA9wA3A/wCTwB8BtwEPAfcCG3q+9p8A/1XP1pnp5PqxZcCbgLvr\nx68FnjD0/3NJkiQJWEFaaN8JPB7YG/h94FzgLmAC2Bf4JPDh+nNeAXy162scBfwU2Ks+9z7gHNIT\nFccAP6rPgbQUbwdOrI8fB5wEHF0f/w7w38AZXV/758Cz6q//98AO4A/qx/8C+BpwUP34e4GPjZCD\nJEmSWuxE0jPEvT9p+yLwqq7jp5OW0eXA/qRnedfUj70duKq+fSbw5Z6v9T7grfXtD9Yf89kIvKu+\n/Vbgo12PPR74FbNL8e1dtwGe2jWnJLWWfwhK0sKsIVUTHu25/6n1/R33AiuBJ5Oeub0eWF8/diaz\ni+thwPGkZ447H2fVnwfwGOmZ5G7HA5tIy/l24JXAE+vHDgK2dp37f6RaRscE8K9dv9btwCNdv54k\ntZJLsSQtzH3AoaQaRbdtpIWz41DSsvlAfXw1aSk+kfTs7ab6/nuBfyP1ejsf+7N7B7nbx4BPAYcA\nq0kViGVdcxzSde7jmV2YO7/eaT2/3irgB/P8epIkSdIulgObSV3dVcA+pE7xecD3SIvxfsAnmO0U\nQ+oe/wT4PKmP3LEf6YVwZ5M6vnsBv0d68R2k6sTbemZ4gPRiOoDj6uPOr3U08DPS8r038A527xRv\nIi3tAE8CXjjc/3VJkiRp1hpSBeFB0oviNpKeqX0L6ZnYH5KW1AN6Pu8q4NfAM3vufzrwGWavZnET\n8Lv1Yx8A/rrn/JeQFumfAZ8GLmfXBfwcdr36xFbS4k495+tIV7n4GekqFH8z3P9tSWq395Oehfj2\nPOdcTnrV9RbSK6clSUvDfsDDpO6yJGkPPJu06PZbip9Pun4mpBd/3DyOoSRJfZ1OqnbsS+obfzPv\nOJJUjgn6L8XvJb2SuuNOfBWzJOX0T6QrS2wnvSnIEXnHkaRyTNB/Kf406SLxHTexe19OkiRJWrIi\nLsm2jNlLAXX0Xr9TkiRJWrJWBnyNrcy+SxOk62Nu6z3poIMOemzbtt3uliRJkiLdAxy+0E+KeKb4\nOmavl3kCqcP2QO9J27Zt47HHHhvLx9q1JwL/TnojqFI/NvR9bNWqP+PKK68cW95N/9iwYUP2GUr6\nME/zXKofZmmeS/kjMs8k956S84OnMYJhnim+GjgJOJD0Tk4bSBeXB3gf6coTzydd6/J/gT8dZRAt\n1EzuAYoxMzOTe4SimGcs84xjlrHMM5Z55jfMUrx+iHMu2NNBJEmSpFwi6hPKYir3AMWYmprKPUJR\nzDOWecYxy1jmGcs88+u9asRiemy257K4jjrqWdxxxz+w65Xi2mPVqvPZuPE4zj///NyjSJKkMVu2\nbBl1t7allu38n4XwmeLGqnIPUIyqqnKPUBTzjGWeccwylnnGMs/8XIolSZLUetYnCmR9QpKk9rI+\nYX1CkiRJGolLcWNVuQcohj2uWOYZyzzjmGUs84xlnvm5FEuSJKn17BQXyE6xJEntZafYTrEkSZI0\nEpfixqpyD1AMe1yxzDOWecYxy1jmGcs883MpliRJUuvZKS6QnWJJktrLTrGdYkmSJGkkLsWNVeUe\noBj2uGKZZyzzjGOWscwzlnnm51IsSZKk1rNTXCA7xZIktZedYjvFkiRJ0khcihuryj1AMexxxTLP\nWOYZxyxjmWcs88zPpViSJEmtZ6e4QHaKJUlqLzvFdoolSZKkkbgUN1aVe4Bi2OOKZZ6xzDOOWcYy\nz1jmmZ9LsSRJklrPTnGB7BRLktRedortFEuSJEkjcSlurCr3AMWwxxXLPGOZZxyzjGWescwzP5di\nSZIktZ6d4gLZKZYkqb3sFNspliRJkkbiUtxYVe4BimGPK5Z5xjLPOGYZyzxjmWd+LsWSJElqPTvF\nBbJTLElSe9kptlMsSZIkjcSluLGq3AMUwx5XLPOMZZ5xzDKWecYyz/xciiVJktR6wyzFpwF3AncB\nF87x+KHAJuBWYAvwvLDpNI/J3AMUY3JyMvcIRTHPWOYZxyxjmWcs88xv0FK8AriCtBgfBawH1vac\n82bgGuBY4OXAu4NnlCRJkhbVoKX4OOBuYAZ4mLT8ntFzzqPAAfXt1cD9gfOpryr3AMWwxxXLPGOZ\nZxyzjGWescwzv5UDHj8YuK/reCtwfM85FwOfB14D7AucHDWcJEmSNA6DluK59F747izgA8BlwAnA\nR4Cj5/rEqakpJiYmAFi9ejXr1q3b2aHp/Asp6jhVnHcw272t6v+Wcty5b+7Hp6enqapq0fIt6Xhy\ncnJJzdP0Y/M0T4899jj/8azO8WTBx5uB7fXxDKMadGHjE0jPBJ9WH19Eqktc2nXOd4BTma1N3EN6\nNvnBnq/lm3eMiW/eIUlSe/nmHYvz5h3/CRwBTAB7A2cC1/Wccy9wSn17LbAPuy/EClflHqAYu/+r\nWnvCPGOZZxyzjGWescwzv0FL8SPABcCNwO3AtcAdwCXA6fU5bwDOJz13/THgnEWZVJIkSVokC35q\neQ9YnxgT6xOSJLWX9YnFqU9IkiRJxXMpbqwq9wDFsMcVyzxjmWccs4xlnrHMMz+XYkmSJLWeneIC\n2SmWJKm97BTbKZYkSZJG4lLcWFXuAYphjyuWecYyzzhmGcs8Y5lnfi7FkiRJaj07xQWyUyxJUnvZ\nKbZTLEmSJI3EpbixqtwDFMMeVyzzjGWeccwylnnGMs/8XIolSZLUenaKC2SnWJKk9rJTbKdYkiRJ\nGolLcWNVuQcohj2uWOYZyzzjmGUs84xlnvm5FEuSJKn17BQXyE6xJEntZafYTrEkSZI0Epfixqpy\nD1AMe1yxzDOWecYxy1jmGcs883MpliRJUuvZKS6QnWJJktrLTrGdYkmSJGkkLsWNVeUeoBj2uGKZ\nZyzzjGOWscwzlnnm51IsSZKk1rNTXCA7xZIktZedYjvFkiRJ0khcihuryj1AMexxxTLPWOYZxyxj\nmWcs88zPpViSJEmtZ6e4QHaKJUlqLzvFdoolSZKkkbgUN1aVe4Bi2OOKZZ6xzDOOWcYyz1jmmZ9L\nsSRJklrPTnGB7BRLktRedortFEuSJEkjcSlurCr3AMWwxxXLPGOZZxyzjGWescwzP5diSZIktd4w\nS/FpwJ3AXcCFfc55GfBd4DvAR2NG0/wmcw9QjMnJydwjFMU8Y5lnHLOMZZ6xzDO/lQMeXwFcAZwC\n3A/8B3AdcEfXOUcAbyK9qu0h4MD4MSVJkqTFM+iZ4uOAu4EZ4GHgGuCMnnPOJy3OD9XHDwbOp76q\n3AMUwx5XLPOMZZ5xzDKWecYyz/wGLcUHA/d1HW+t7+t2BHAk8FXg68CpYdNJkiRJYzCoPjGX3gvf\nrQQOB04C1gBfAX6b2WeOtSgmcw9QDHtcscwzlnnGMctY5hnLPPMbtBRvJS26HWtI3eLec24Gfk2q\nWUyTluRv9n6xqakpJiYmAFi9ejXr1q3b+U3Q+bFB1DHcCuxgdnms6v+243h6epqqqhYtX4899thj\njz32eOkez+ocTxZ8vBnYXh/PMKpB7/axkrTkngxsA24B1rPrC+1Ore+bIr3I7lbgGcBPe76W72gX\nqmL2G2JXvqPdwlRVtfMPE+0584xlnnHMMpZ5xorM03e0W5x3tHsEuAC4EbgduJa0EF8CnF6fcyPw\nY9Il2b4EvJHdF2JJkiRpyVrwFr0HfKZ4THymWJKk9vKZ4sV5pliSJEkqnktxY1W5ByjG7i9K0J4w\nz1jmGccsY5lnLPPMz6VYkiRJrWenuEB2iiVJai87xXaKJUmSpJG4FDdWlXuAYtjjimWescwzjlnG\nMs9Y5pmfS7EkSZJaz05xgewUS5LUXnaK7RRLkiRJI3Epbqwq9wDFsMcVyzxjmWccs4xlnrHMMz+X\nYkmSJLWeneIC2SmWJKm97BTbKZYkSZJG4lLcWFXuAYphjyuWecYyzzhmGcs8Y5lnfi7FkiRJaj07\nxQWyUyxJUnvZKbZTLEmSJI3EpbixqtwDFMMeVyzzjGWeccwylnnGMs/8XIolSZLUenaKC2SnWJKk\n9rJTbKdYkiRJGolLcWNVuQcohj2uWOYZyzzjmGUs84xlnvm5FEuSJKn17BQXyE6xJEntZafYTrEk\nSZI0EpfixqpyD1AMe1yxzDOWecYxy1jmGcs883MpliRJUuvZKS6QnWJJktrLTrGdYkmSJGkkLsWN\nVeUeoBj2uGKZZyzzjGOWscwzlnnm51IsSZKk1rNTXCA7xZIktZedYjvFkiRJ0khcihuryj1AMexx\nxTLPWOYZxyxjmWcs88zPpViSJEmtZ6e4QHaKJUlqLzvFi9cpPg24E7gLuHCe814KPAocu9AhJEmS\npJwGLcUrgCtIi/FRwHpg7Rzn7Q+8Frg5dDrNo8o9QDHsccUyz1jmGccsY5lnLPPMb9BSfBxwNzAD\nPAxcA5wxx3lvAy4FfsV4KxmSJEnSHhu0FB8M3Nd1vLW+r9sx9X3X18dtLrGM0WTuAYoxOTmZe4Si\nmGcs84xjlrHMM5Z55rdyhM/pXnqXA5cB53Td5zPFkiRJapRBS/FWYE3X8Rrg/q7j/YGjmS24PgW4\nDjgduLX3i01NTTExMQHA6tWrWbdu3c5/GXW6NFHH6Zffwewzqp0RSzneCKzr+/j09DRVVS1aviUd\nd/e4lsI8TT82T/Ncqsed+5bKPE0/7ty3VOZp+nHnvsivl3SOJws+3gxsr49nGNWgZ3VXAtPAycA2\n4BbSi+3u6HP+JuANzLEQ4yXZglXMfkPsykuyLUxVVTv/MNGeM89Y5hnHLGOZZ6zIPL0k2+Jcku0R\n4ALgRuB24FrSQnwJ6dlgZTOZe4Bi+Id6LPOMZZ5xzDKWecYyz/yG6RR/tv7otqHPuc/ds3EkSZKk\n8Rv0TLGWrCr3AMXYvX+lPWGescwzjlnGMs9Y5pmfS7EkSZJab5yXT/OFdmPiC+0kSWovX2i3OC+0\nkyRJkornUtxYVe4BimGPK5Z5xjLPOGYZyzxjmWd+LsWSJElqPTvFBbJTLElSe9kptlMsSZIkjcSl\nuLGq3AMUwx5XLPOMZZ5xzDKWecYyz/xciiVJktR6dooLZKdYkqT2slNsp1iSJEkaiUtxY1W5ByiG\nPa5Y5hnLPOOYZSzzjGWe+bkUS5IkqfXsFBfITrEkSe1lp9hOsSRJkjQSl+LGqnIPUAx7XLHMM5Z5\nxjHLWOYZyzzzcymWJElS69kpLpCdYkmS2stOsZ1iSZIkaSQuxY1V5R6gGPa4YplnLPOMY5axzDOW\neebnUixJkqTWs1NcIDvFkiS1l51iO8WSJEnSSFyKG6vKPUAx7HHFMs9Y5hnHLGOZZyzzzM+lWJIk\nSa1np7hAdoolSWovO8V2iiVJkqSRuBQ3VpV7gGLY44plnrHMM45ZxjLPWOaZn0uxJEmSWs9OcYHs\nFEuS1F52iu0US5IkSSNxKW6sKvcAxbDHFcs8Y5lnHLOMZZ6xzDM/l2JJkiS1np3iAtkpliSpvewU\n2ymWJEmSRjLsUnwacCdwF3DhHI+/HvgusAW4CTg0ZDrNo8o9QDHsccUyz1jmGccsY5lnLPPMb5il\neAVwBWkxPgpYD6ztOedW4JnAM4BPAO8InFGSJElaVMMsxccBdwMzwMPANcAZPedUwC/r298ADokZ\nT/1N5h6gGJOTk7lHKIp5xjLPOGYZyzxjmWd+wyzFBwP3dR1vre/r5zzghj0ZSpIkSRqnlSN+Xr+X\nNJ4NHAu8bq4Hp6ammJiYAGD16tWsW7du57+MOl2aqOPU6NjB7DOqVf3fUo43Auv6Pj49PU1VVYuW\nb0nH3T2upTBP04/N0zyX6nHnvqUyT9OPO/ctlXmafty5L/LrJZ3jyYKPNwPb6+MZRjXM5SpOAC4m\ndYoBLgIeBS7tOe8U4HLgOcCDc3wdL8kWqmL2G2JXXpJtYaqq2vmHifacecYyzzhmGcs8Y0Xm6SXZ\nRrsk2zCfsBKYBk4GtgG3kF5sd0fXOccAHwdOBe7p83VcisfEpViSpPZyKV686xQ/AlwA3AjcDlxL\nWogvAV5Qn/MOYF/SlSduAz610EEkSZKkXIZZigE+CxwJHA78XX3fBuAz9e0/BJ5Kesb4GOBFgTNq\nTlXuAYqxe/9Ke8I8Y5lnHLOMZZ6xzDO/YZdiSZIkqVgL7lvsATvFY2KnWJKk9rJTvHidYkmSJKlo\nLsWNVeUeoBj2uGKZZyzzjGOWscwzlnnm51IsSZKk1rNTXCA7xZIktZedYjvFkiRJ0khcihuryj1A\nMexxxTLPWOYZxyxjmWcs88zPpViSJEmtZ6e4QHaKJUlqLzvFdoolSZKkkbgUN1aVe4Bi2OOKZZ6x\nzDOOWcYyz1jmmZ9LsSRJklrPTnGB7BRLktRedortFEuSJEkjcSlurCr3AMWwxxXLPGOZZxyzjGWe\nscwzP5diSZIktZ6d4gLZKZYkqb3sFNspliRJkkbiUtxYVe4BimGPK5Z5xjLPOGYZyzxjmWd+LsWS\nJElqPTvFBbJTLElSe9kptlMsSZIkjcSluLGq3AMUwx5XLPOMZZ5xzDKWecYyz/xciiVJktR6dooL\nZKdYkqT2slNsp1iSJEkaiUtxY1W5ByiGPa5Y5hnLPOOYZSzzjGWe+bkUS5IkqfXsFBfITrEkSe1l\np9hOsSRJkjQSl+LGqnIPUAx7XLHMM5Z5xjHLWOYZyzzzcymWJElS69kpLpCdYkmS2stOsZ1iSZIk\naSTDLMWnAXcCdwEXzvH444Br68dvBg4Lm07zqHIPUAx7XLHMM5Z5xjHLWOYZyzzzG7QUrwCuIC3G\nRwHrgbU955wH/Bg4ArgMuDR4Rs1pc+4BirF5s1lGMs9Y5hnHLGOZZyzzzG/QUnwccDcwAzwMXAOc\n0XPOC4EP1bc/CZwcOJ/62p57gGJs326WkcwzlnnGMctY5hnLPPMbtBQfDNzXdby1vq/fOY8ADwG/\nGTKdJEmSNAYrR/ic3pczzvXqvqwvedxrr+Xsu+9fsmJFubv5L35xG6tWfXPOx3bs2MLy5SeMeaLm\nmpmZyT1CUcwzlnnGMctY5hnLPPMbdLmKE4CLSZ1igIuAR9m1N/y5+pybSUv2D4AnzfG17gaeNvqo\nkiRJ0kD3AIdHf9GV9ReeAPYmvbqr94V2rwbeU99+Oal3LEmSJBXlecA06Znei+r7LgFOr28/Dvhn\nZi/JNjHm+SRJkiRJkiRJS82gN/t4PfBdYAtwE3Do+EZrnEFZvgr4FnAb8BV2r7ZoV4Py7HgpqTt/\n7DiGarBBeU4BPyJ9f94GnDu2yZpnmO/Nl5H+7PwO8NExzdVUg/J8F7Pfl9PAT8c3WiMNyvNQYBNw\nK+nv9ueNb7TGGZTlYcAXSTluYvcrfmnW+4EHgG/Pc87lpKy3AMeMY6huK0g1iwlgL+buIE8C+9S3\nX4Ud5H6GyXL/rtunA58dy2TNNEyekDL9MvA1XIrnM0ye55D+QNL8hsnyCNLCcUB9fOC4hmugYX+v\nd1wAXLX4YzXWMHleCbyyvr0W+P64hmuYYbL8OPCK+vZzgQ+Pa7gGejZp0e23FD8fuKG+fTyp4juv\nYd7meSGGebOPCvhlffsbwCHBM5RimCx/3nV7P9Kzm5rbMHkCvI10dZVfMfjqLG02TJ7LMMNhDJPl\n+aR3F32oPn5wXMM10LC/1zvOAq5e/LEaa5g8H2X2H2yrgfvHNVzDDJPlWtIzxZD2pfm+d9vuK8z/\nU57uN5f7Bul788nzfcHopXiYN/vodh6zW7x2NWyWryb9JrsUeO0Y5mqqYfI8pr7v+vo46/W2l7hh\n8nwMeAnpx1Yfx38A9zNMlkcARwJfBb4OnDqe0RppIX8PHUZ61u5LizxTkw2T58XA2fV51wOvGctk\nzTNMlltIFT6AF5N+evmExR+tSHPlPe/fQ9FL8Vz6LRZnk348/fdjmKEUc2X5btK1+C4E3jLecRqv\nO8/lwGXAG7vu81nOhen9/vw0ael4Bun1Ax/a7TPUT2+WK0m/z08C1pN+3H9A7yepr35/D72c9A82\n/wG8ML15nQV8AFhD+pH1R8Y+UXP1ZvlG0u/zW4HnkJ51f2TcQxVirp9WzvsT9eileCvpN0XHGub+\nMcopwF+Rntp+OHiGUgybZce1wIsWdaJmG5Tn/sDRpB9XfZ/0xjXXYa+4n2G+P3/C7O/vq4BnjmGu\nJhomy62k78dfk370Os0iXJi+EAv5s/NMrE4MMkye55IuzQqpt7kP9t7nMkyWPyD9hO1Y4M31fT9H\no+jN+xBg2zgHGObNPo7Bd7cbxjBZdv+leDpwy1gma6Zh8uy2CRfi+QyT51O6br+Y9OJF7W6YLE8F\nPljfPhC4F3+k2s+wv9ePxBeEDWOYPG8gvbCW+jE7xXMbJssnMvuE5dtJ1RT1N8FwL7Q7gSFeaLcY\n+r3Zxwvq218g/UuoczmcT417wAYZ9MYpG0mXZ7qNVMz3kmzzG5RnN5fiwQbl+bek78/NpO/Pp497\nwAYZ5nvznaRLsn2LdHk29TdMnhtI36MabFCea0l9982kv49OGfeADTIoy5cC36vPuZJ0lQrN7WrS\nM787SN3hc0lXQXll1zlXkLLegn+nS5IkSZIkSZIkSZIkSZIkSZIkSZIkSZIkSZIkSZIkSZIkSZIk\nSf39P/4wCr1zq96eAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x1273c9f50>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "lf_stats = training_set.lf_stats()\n",
    "lf_stats[:5]\n",
    "lf_stats.hist(\"coverage\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "============================================================\n",
      "Testing mu = 1.00e-05, lf_w0 = 1.00e+00\n",
      "============================================================\n",
      "Begin training for rate=0.01, mu=1e-05\n",
      "\tLearning epoch = 0\tGradient mag. = 0.030888\n",
      "\tLearning epoch = 250\tGradient mag. = 0.084110\n",
      "\tLearning epoch = 500\tGradient mag. = 0.211097\n",
      "\tLearning epoch = 750\tGradient mag. = 0.275662\n",
      "Final gradient magnitude for rate=0.01, mu=1e-05: 0.282\n",
      "Applying LFs...\n",
      "Featurizing...\n",
      "============================================================\n",
      "Testing mu = 1.00e-05, lf_w0 = 2.00e+00\n",
      "============================================================\n",
      "Begin training for rate=0.01, mu=1e-05\n",
      "\tLearning epoch = 0\tGradient mag. = 0.061684\n",
      "\tLearning epoch = 250\tGradient mag. = 0.158546\n",
      "\tLearning epoch = 500\tGradient mag. = 0.277430\n",
      "\tLearning epoch = 750\tGradient mag. = 0.284230\n",
      "Final gradient magnitude for rate=0.01, mu=1e-05: 0.294\n",
      "============================================================\n",
      "Testing mu = 1.00e-07, lf_w0 = 1.00e+00\n",
      "============================================================\n",
      "Begin training for rate=0.01, mu=1e-07\n",
      "\tLearning epoch = 0\tGradient mag. = 0.030888\n",
      "\tLearning epoch = 250\tGradient mag. = 0.084282\n",
      "\tLearning epoch = 500\tGradient mag. = 0.211653\n",
      "\tLearning epoch = 750\tGradient mag. = 0.275691\n",
      "Final gradient magnitude for rate=0.01, mu=1e-07: 0.282\n",
      "============================================================\n",
      "Testing mu = 1.00e-07, lf_w0 = 2.00e+00\n",
      "============================================================\n",
      "Begin training for rate=0.01, mu=1e-07\n",
      "\tLearning epoch = 0\tGradient mag. = 0.061684\n",
      "\tLearning epoch = 250\tGradient mag. = 0.158704\n",
      "\tLearning epoch = 500\tGradient mag. = 0.277446\n",
      "\tLearning epoch = 750\tGradient mag. = 0.284270\n",
      "Final gradient magnitude for rate=0.01, mu=1e-07: 0.294\n"
     ]
    }
   ],
   "source": [
    "from snorkel.snorkel import Learner\n",
    "import snorkel.learning\n",
    "from snorkel.learning import LogReg\n",
    "\n",
    "learner = Learner(training_set, model=LogReg())\n",
    "\n",
    "# Splitting into CV and test set\n",
    "n_half = len(gold_candidates)/2\n",
    "test_candidates = gold_candidates[:n_half]\n",
    "test_labels     = gold_labels[:n_half]\n",
    "cv_candidates   = gold_candidates[n_half:]\n",
    "cv_labels       = gold_labels[n_half:]\n",
    "\n",
    "from snorkel.learning_utils import GridSearch\n",
    "gs       = GridSearch(learner, ['mu', 'lf_w0'], [[1e-5, 1e-7],[1.0,2.0]])\n",
    "gs_stats = gs.fit(cv_candidates, cv_labels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mu</th>\n",
       "      <th>lf_w0</th>\n",
       "      <th>Prec.</th>\n",
       "      <th>Rec.</th>\n",
       "      <th>F1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td> 0.00001</td>\n",
       "      <td> 1</td>\n",
       "      <td> 1</td>\n",
       "      <td> 1</td>\n",
       "      <td> 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td> 0.00001</td>\n",
       "      <td> 2</td>\n",
       "      <td> 1</td>\n",
       "      <td> 1</td>\n",
       "      <td> 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 1</td>\n",
       "      <td> 1</td>\n",
       "      <td> 1</td>\n",
       "      <td> 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 2</td>\n",
       "      <td> 1</td>\n",
       "      <td> 1</td>\n",
       "      <td> 1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        mu  lf_w0  Prec.  Rec.  F1\n",
       "0  0.00001      1      1     1   1\n",
       "1  0.00001      2      1     1   1\n",
       "2  0.00000      1      1     1   1\n",
       "3  0.00000      2      1     1   1\n",
       "\n",
       "[4 rows x 5 columns]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gs_stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Applying LFs...\n",
      "Featurizing...\n",
      "========================================\n",
      "Test set size:\t14\n",
      "----------------------------------------\n",
      "Precision:\t1.0\n",
      "Recall:\t\t1.0\n",
      "F1 Score:\t1.0\n",
      "----------------------------------------\n",
      "TP: 7 | FP: 0 | TN: 7 | FN: 0\n",
      "========================================\n"
     ]
    }
   ],
   "source": [
    "learner.test_wmv(test_candidates, test_labels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Applying LFs...\n",
      "Featurizing...\n",
      "[('17903293', u'CD40 Ligand, serum & plasma', u'CD40L'), ('17903293', u'C-reactive protein', u'CRP'), ('17903293', u'Intercellular adhesion molecule-1', u'ICAM1'), ('17903293', u'Interleukin-6', u'IL6'), ('17903293', u'Urinary isoprostanes/creatinine', u'IsoCrUrine'), ('17903293', u'Monocyte chemoattractant protein-1', u'MCP1'), ('17903293', u'Myeloperoxidase', u'MPO'), ('17903293', u'Osteoprotegerin', u'OPG'), ('17903293', u'P-selectin', u'Pselectin'), ('17903293', u'Tumor necrosis factor alpha', u'TNFA')]\n"
     ]
    }
   ],
   "source": [
    "preds = learner.predict_wmv(candidates)\n",
    "resolved_acronyms = [spair2uid(c) for (c, p) in zip(candidates, preds) if p == 1]\n",
    "print resolved_acronyms[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# mislabeled_cand = learner.mislabeled_test_candidates(test_candidates, test_labels)\n",
    "# for (c,p,g) in mislabeled_cand[:10]:\n",
    "#     print c.span0.context.document.name\n",
    "#     print c.span0.context    \n",
    "#     print c.span1.context\n",
    "#     print"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Classifying sentence acronyms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "516\n"
     ]
    }
   ],
   "source": [
    "from snorkel.candidates import UnionExtractor\n",
    "\n",
    "contexts = [corpus.get_phrases(), text_corpus.get_sentences()]\n",
    "extractors = [txt_tab_extractor, txt_tab_extractor]\n",
    "joint_extractor = UnionExtractor(extractor_list=extractors, context_list=contexts)\n",
    "text_c = joint_extractor.union()\n",
    "print len(text_c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# helpers\n",
    "def spairtxt2uid(span_pair):\n",
    "    doc_id = span_pair.span0.context.document.name\n",
    "    str1 = span_pair.span0.get_span()\n",
    "    str2 = span_pair.span1.get_span()\n",
    "    acro = str2[1:-1]\n",
    "    phen = str1.split(' (')[0]\n",
    "    return (doc_id, phen, acro)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We really trust the first stage, so if an acronym has been resolved there, remove it from this stage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "17903292 Glomerular Filtration Rate GFR\n",
      "17903292 Glomerular Filtration Rate GFR\n",
      "17903292 Filtration Rate GFR\n",
      "17903292 Filtration Rate GFR\n",
      "17903292 Glomerular Filtration Rate GFR\n",
      "17903292 Rate GFR\n",
      "17903292 Rate GFR\n",
      "17903292 Filtration Rate GFR\n",
      "17903292 Rate GFR\n",
      "17903292 Thyroid stimulation hormone TSH\n",
      "17903292 Thyroid stimulation hormone TSH\n",
      "17903292 stimulation hormone TSH\n",
      "17903292 stimulation hormone TSH\n",
      "17903292 Thyroid stimulation hormone TSH\n",
      "17903292 hormone TSH\n",
      "17903292 hormone TSH\n",
      "17903292 stimulation hormone TSH\n",
      "17903292 hormone TSH\n",
      "17903292 Follicle stimulating hormone FSH\n",
      "17903292 stimulating hormone FSH\n",
      "17903292 hormone FSH\n",
      "17903292 Dehydroepiandrosterone sulfate DHEAS\n",
      "17903292 sulfate DHEAS\n",
      "17903292 on the web are GFRMV7 GFR\n",
      "17903292 (GFR), UAELNMV6 GFR\n",
      "17903292 (GFR), UAELNMV6 UAE\n",
      "17903292 CysC), and TSHMEAN34MV TSH\n",
      "17903292 , and TSHMEAN34MV TSH\n",
      "17903292 , and TSHMEAN34MV TSH\n",
      "17903292 the web are GFRMV7 GFR\n",
      "17903292 GFR), UAELNMV6 UAE\n",
      "17903292 ), and TSHMEAN34MV TSH\n",
      "17903292 and TSHMEAN34MV TSH\n",
      "17903292 and TSHMEAN34MV TSH\n",
      "17903292 web are GFRMV7 GFR\n",
      "17903292 ), UAELNMV6 UAE\n",
      "17903292 , and TSHMEAN34MV TSH\n",
      "17903292 TSHMEAN34MV TSH\n",
      "17903292 TSHMEAN34MV TSH\n",
      "17903292 are GFRMV7 GFR\n",
      "17903292 , UAELNMV6 UAE\n",
      "17903292 and TSHMEAN34MV TSH\n",
      "17903292 GFRMV7 GFR\n",
      "17903292 UAELNMV6 UAE\n",
      "17903292 TSHMEAN34MV TSH\n",
      "17903292 Physical Location Mb\n",
      "17903292 Physical Location Mb\n",
      "17903292 Location Mb\n",
      "17903292 Location Mb\n",
      "17903292 Physical Location Mb\n",
      "17903292 Location Mb\n",
      "17903292 on the web are GFRMV7 GFR\n",
      "17903292 (GFR), UAELNMV6 GFR\n",
      "17903292 (GFR), UAELNMV6 UAE\n",
      "17903292 CysC), and TSHMEAN34MV TSH\n",
      "17903292 , and TSHMEAN34MV TSH\n",
      "17903292 , and TSHMEAN34MV TSH\n",
      "17903292 the web are GFRMV7 GFR\n",
      "17903292 GFR), UAELNMV6 UAE\n",
      "17903292 ), and TSHMEAN34MV TSH\n",
      "17903292 and TSHMEAN34MV TSH\n",
      "17903292 and TSHMEAN34MV TSH\n",
      "17903292 web are GFRMV7 GFR\n",
      "17903292 ), UAELNMV6 UAE\n",
      "17903292 , and TSHMEAN34MV TSH\n",
      "17903292 TSHMEAN34MV TSH\n",
      "17903292 TSHMEAN34MV TSH\n",
      "17903292 are GFRMV7 GFR\n",
      "17903292 , UAELNMV6 UAE\n",
      "17903292 and TSHMEAN34MV TSH\n",
      "17903292 GFRMV7 GFR\n",
      "17903292 UAELNMV6 UAE\n",
      "17903292 TSHMEAN34MV TSH\n",
      "17903292 Physical Location Mb\n",
      "17903292 Physical Location Mb\n",
      "17903292 Location Mb\n",
      "17903292 Location Mb\n",
      "17903292 Physical Location Mb\n",
      "17903292 Location Mb\n",
      "17903292 Physical location Mb\n",
      "17903292 Physical location Mb\n",
      "17903292 location Mb\n",
      "17903292 location Mb\n",
      "17903292 Physical location Mb\n",
      "17903292 location Mb\n",
      "17903292 LOD 1.5 Lower\n",
      "17903292 1.5 Lower\n",
      "17903292 LOD 1.5 Upper\n",
      "17903292 1.5 Upper\n",
      "17903292 SNP rsID\n",
      "17903292 Mean p-value GEE\n",
      "17903292 Mean p-value GEE\n",
      "17903292 p-value GEE\n",
      "17903292 p-value GEE\n",
      "17903292 Mean p-value GEE\n",
      "17903292 p-value GEE\n",
      "17903292 Mean p-value FBAT\n",
      "17903292 p-value FBAT\n",
      "17903292 (CysC) and CKDMV7 CKD\n",
      "17903292 CysC) and CKDMV7 CKD\n",
      "17903292 ) and CKDMV7 CKD\n",
      "17903292 and CKDMV7 CKD\n",
      "17903292 CKDMV7 CKD\n",
      "17903292 P_value GEE\n",
      "17903292 P_value GEE\n",
      "17903292 P_value GEE\n",
      "17903293 Physical location bp\n",
      "17903293 Physical location bp\n",
      "17903293 location bp\n",
      "17903293 location bp\n",
      "17903293 Physical location bp\n",
      "17903293 location bp\n",
      "17903293 Physical location bp\n",
      "17903293 Physical location bp\n",
      "17903293 location bp\n",
      "17903293 location bp\n",
      "17903293 Physical location bp\n",
      "17903293 location bp\n",
      "17903293 Physical location bp\n",
      "17903293 Physical location bp\n",
      "17903293 location bp\n",
      "17903293 location bp\n",
      "17903293 Physical location bp\n",
      "17903293 location bp\n",
      "17903293 Physical location bp\n",
      "17903293 Physical location bp\n",
      "17903293 location bp\n",
      "17903293 location bp\n",
      "17903293 Physical location bp\n",
      "17903293 location bp\n",
      "17903294 Physical Position bp\n",
      "17903294 Physical Position bp\n",
      "17903294 Position bp\n",
      "17903294 Position bp\n",
      "17903294 Physical Position bp\n",
      "17903294 Position bp\n",
      "17903294 GEE p-val Rank\n",
      "17903294 p-val Rank\n",
      "17903294 Physical Position bp\n",
      "17903294 Physical Position bp\n",
      "17903294 Position bp\n",
      "17903294 Position bp\n",
      "17903294 Physical Position bp\n",
      "17903294 Position bp\n",
      "17903294 GEE Pval Rank\n",
      "17903294 Pval Rank\n",
      "17903294 Physical Position bp\n",
      "17903294 Physical Position bp\n",
      "17903294 Position bp\n",
      "17903294 Position bp\n",
      "17903294 Physical Position bp\n",
      "17903294 Position bp\n",
      "17903296 Gender N\n",
      "17903296 Gender N\n",
      "17903296 Gender N\n",
      "17903296 Bone mineral density BMD\n",
      "17903296 Bone mineral density BMD\n",
      "17903296 mineral density BMD\n",
      "17903296 mineral density BMD\n",
      "17903296 Bone mineral density BMD\n",
      "17903296 density BMD\n",
      "17903296 density BMD\n",
      "17903296 mineral density BMD\n",
      "17903296 density BMD\n",
      "17903296 Quantitative ultrasound QUS\n",
      "17903296 Quantitative ultrasound QUS\n",
      "17903296 ultrasound QUS\n",
      "17903296 ultrasound QUS\n",
      "17903296 Quantitative ultrasound QUS\n",
      "17903296 ultrasound QUS\n",
      "17903297 Cerebral Brain Volume ATCBV\n",
      "17903297 Brain Volume ATCBV\n",
      "17903297 Volume ATCBV\n",
      "17903297 Frontal Brain volume AFBV\n",
      "17903297 Brain volume AFBV\n",
      "17903297 volume AFBV\n",
      "17903297 Parietal Brain Volume APBV\n",
      "17903297 Brain Volume APBV\n",
      "17903297 Volume APBV\n",
      "17903297 Occipital Brain Volume AOBV\n",
      "17903297 Brain Volume AOBV\n",
      "17903297 Volume AOBV\n",
      "17903297 Temporal Brain volume ATBV\n",
      "17903297 Brain volume ATBV\n",
      "17903297 volume ATBV\n",
      "17903297 Hippocampal Volume AHPV\n",
      "17903297 Volume AHPV\n",
      "17903297 Boston Naming Test Nam\n",
      "17903297 Boston Naming Test Nam\n",
      "17903297 Naming Test Nam\n",
      "17903297 Naming Test Nam\n",
      "17903297 Boston Naming Test Nam\n",
      "17903297 Test Nam\n",
      "17903297 Test Nam\n",
      "17903297 Naming Test Nam\n",
      "17903297 Test Nam\n",
      "17903297 Similarities Sim\n",
      "17903297 Similarities Sim\n",
      "17903297 Similarities Sim\n",
      "17903297 Wide-Range Achievement Test WRAT\n",
      "17903297 Achievement Test WRAT\n",
      "17903297 Test WRAT\n",
      "17903298 Fasting plasma glucose FPG\n",
      "17903298 Fasting plasma glucose FPG\n",
      "17903298 plasma glucose FPG\n",
      "17903298 plasma glucose FPG\n",
      "17903298 Fasting plasma glucose FPG\n",
      "17903298 glucose FPG\n",
      "17903298 glucose FPG\n",
      "17903298 plasma glucose FPG\n",
      "17903298 glucose FPG\n",
      "17903298 time averaged FPG tFPG\n",
      "17903298 averaged FPG tFPG\n",
      "17903298 FPG tFPG\n",
      "17903300 Physical Position Mb\n",
      "17903300 Physical Position Mb\n",
      "17903300 Position Mb\n",
      "17903300 Position Mb\n",
      "17903300 Physical Position Mb\n",
      "17903300 Position Mb\n",
      "17903300 Physical Location Mb\n",
      "17903300 Physical Location Mb\n",
      "17903300 Location Mb\n",
      "17903300 Location Mb\n",
      "17903300 Physical Location Mb\n",
      "17903300 Location Mb\n",
      "17903300 Physical Position bp\n",
      "17903300 Physical Position bp\n",
      "17903300 Position bp\n",
      "17903300 Position bp\n",
      "17903300 Physical Position bp\n",
      "17903300 Position bp\n",
      "17903300 .5 Lower Bound bp\n",
      "17903300 .5 Lower Bound bp\n",
      "17903300 LOD1.5 Lower Bound bp\n",
      "17903300 Lower Bound bp\n",
      "17903300 Lower Bound bp\n",
      "17903300 .5 Lower Bound bp\n",
      "17903300 Bound bp\n",
      "17903300 Bound bp\n",
      "17903300 Lower Bound bp\n",
      "17903300 Bound bp\n",
      "17903300 .5 Upper Bound bp\n",
      "17903300 .5 Upper Bound bp\n",
      "17903300 LOD1.5 Upper Bound bp\n",
      "17903300 Upper Bound bp\n",
      "17903300 Upper Bound bp\n",
      "17903300 .5 Upper Bound bp\n",
      "17903300 Bound bp\n",
      "17903300 Bound bp\n",
      "17903300 Upper Bound bp\n",
      "17903300 Bound bp\n",
      "17903300 Physical Position bp\n",
      "17903300 Physical Position bp\n",
      "17903300 Position bp\n",
      "17903300 Position bp\n",
      "17903300 Physical Position bp\n",
      "17903300 Position bp\n",
      "17903300 Location Candidate SNP bp\n",
      "17903300 Location Candidate SNP bp\n",
      "17903300 Candidate SNP bp\n",
      "17903300 Candidate SNP bp\n",
      "17903300 Location Candidate SNP bp\n",
      "17903300 SNP bp\n",
      "17903300 SNP bp\n",
      "17903300 Candidate SNP bp\n",
      "17903300 SNP bp\n",
      "17903300 Location 100K SNP bp\n",
      "17903300 Location 100K SNP bp\n",
      "17903300 100K SNP bp\n",
      "17903300 100K SNP bp\n",
      "17903300 Location 100K SNP bp\n",
      "17903300 SNP bp\n",
      "17903300 SNP bp\n",
      "17903300 100K SNP bp\n",
      "17903300 SNP bp\n",
      "17903300 Physical Position bp\n",
      "17903300 Physical Position bp\n",
      "17903300 Position bp\n",
      "17903300 Position bp\n",
      "17903300 Physical Position bp\n",
      "17903300 Position bp\n",
      "17903301 LV mass LVM\n",
      "17903301 LV mass LVM\n",
      "17903301 mass LVM\n",
      "17903301 mass LVM\n",
      "17903301 LV mass LVM\n",
      "17903301 mass LVM\n",
      "17903301 LV diastolic dimension LVDD\n",
      "17903301 diastolic dimension LVDD\n",
      "17903301 dimension LVDD\n",
      "17903301 LV systolic dimension LVDS\n",
      "17903301 systolic dimension LVDS\n",
      "17903301 dimension LVDS\n",
      "17903301 LV wall thickness LVWT\n",
      "17903301 wall thickness LVWT\n",
      "17903301 thickness LVWT\n",
      "17903301 LV fractional shortening LVFS\n",
      "17903301 fractional shortening LVFS\n",
      "17903301 shortening LVFS\n",
      "17903301 Left atrial diameter LAD\n",
      "17903301 Left atrial diameter LAD\n",
      "17903301 atrial diameter LAD\n",
      "17903301 atrial diameter LAD\n",
      "17903301 Left atrial diameter LAD\n",
      "17903301 diameter LAD\n",
      "17903301 diameter LAD\n",
      "17903301 atrial diameter LAD\n",
      "17903301 diameter LAD\n",
      "17903301 Aortic root diameter AOR\n",
      "17903301 Aortic root diameter AOR\n",
      "17903301 root diameter AOR\n",
      "17903301 root diameter AOR\n",
      "17903301 Aortic root diameter AOR\n",
      "17903301 diameter AOR\n",
      "17903301 diameter AOR\n",
      "17903301 root diameter AOR\n",
      "17903301 diameter AOR\n",
      "17903301 B. Exercise Treadmill Test ETT\n",
      "17903301 Exercise Treadmill Test ETT\n",
      "17903301 Treadmill Test ETT\n",
      "17903301 Test ETT\n",
      "17903301 2 Exercise systolic blood pressure SBP\n",
      "17903301 systolic blood pressure SBP\n",
      "17903301 systolic blood pressure SBP\n",
      "17903301 Exercise systolic blood pressure SBP\n",
      "17903301 blood pressure SBP\n",
      "17903301 blood pressure SBP\n",
      "17903301 systolic blood pressure SBP\n",
      "17903301 pressure SBP\n",
      "17903301 pressure SBP\n",
      "17903301 blood pressure SBP\n",
      "17903301 pressure SBP\n",
      "17903301 2 Exercise diastolic blood pressure DBP\n",
      "17903301 diastolic blood pressure DBP\n",
      "17903301 diastolic blood pressure DBP\n",
      "17903301 Exercise diastolic blood pressure DBP\n",
      "17903301 blood pressure DBP\n",
      "17903301 blood pressure DBP\n",
      "17903301 diastolic blood pressure DBP\n",
      "17903301 pressure DBP\n",
      "17903301 pressure DBP\n",
      "17903301 blood pressure DBP\n",
      "17903301 pressure DBP\n",
      "17903301 BA flow-mediated dilation FMD\n",
      "17903301 flow-mediated dilation FMD\n",
      "17903301 dilation FMD\n",
      "17684544 sub-entity of inflammatory bowel disease IBD\n",
      "17684544 of inflammatory bowel disease IBD\n",
      "17684544 inflammatory bowel disease IBD\n",
      "17684544 bowel disease IBD\n",
      "17684544 disease IBD\n",
      "17903292 ) and urinary albumin excretion UAE\n",
      "17903292 and urinary albumin excretion UAE\n",
      "17903292 Glomerular filtration rate GFR\n",
      "17903292 urinary albumin excretion UAE\n",
      "17903292 filtration rate GFR\n",
      "17903292 albumin excretion UAE\n",
      "17903292 rate GFR\n",
      "17903292 excretion UAE\n",
      "17903292 Mapping 100K single nucleotide polymorphism SNP\n",
      "17903292 100K single nucleotide polymorphism SNP\n",
      "17903292 single nucleotide polymorphism SNP\n",
      "17903292 nucleotide polymorphism SNP\n",
      "17903292 polymorphism SNP\n",
      "17903292 creatinine and cystatin-C cysC\n",
      "17903292 and cystatin-C cysC\n",
      "17903292 cystatin-C cysC\n",
      "17903292 Thyroid stimulating hormone TSH\n",
      "17903292 stimulating hormone TSH\n",
      "17903292 hormone TSH\n",
      "17903292 data using generalized estimating equations GEE\n",
      "17903292 using generalized estimating equations GEE\n",
      "17903292 generalized estimating equations GEE\n",
      "17903292 estimating equations GEE\n",
      "17903292 equations GEE\n",
      "17903293 We used Generalized Estimating Equations GEE\n",
      "17903293 used Generalized Estimating Equations GEE\n",
      "17903293 Generalized Estimating Equations GEE\n",
      "17903293 Estimating Equations GEE\n",
      "17903293 Equations GEE\n",
      "17903293 -8 ) for C-reactive protein CRP\n",
      "17903293 ) for C-reactive protein CRP\n",
      "17903293 for C-reactive protein CRP\n",
      "17903293 C-reactive protein CRP\n",
      "17903293 protein CRP\n",
      "17903294 increased risk of cardiovascular disease CVD\n",
      "17903294 risk of cardiovascular disease CVD\n",
      "17903294 of cardiovascular disease CVD\n",
      "17903294 cardiovascular disease CVD\n",
      "17903294 disease CVD\n",
      "17903294 GeneChip 100K single nucleotide polymorphism SNP\n",
      "17903294 100K single nucleotide polymorphism SNP\n",
      "17903294 single nucleotide polymorphism SNP\n",
      "17903294 nucleotide polymorphism SNP\n",
      "17903294 polymorphism SNP\n",
      "17903294 approximately 1000 Framingham Heart Study FHS\n",
      "17903294 1000 Framingham Heart Study FHS\n",
      "17903294 Framingham Heart Study FHS\n",
      "17903294 Heart Study FHS\n",
      "17903294 Study FHS\n",
      "17903294 using the generalized estimating equations GEE\n",
      "17903294 the generalized estimating equations GEE\n",
      "17903294 generalized estimating equations GEE\n",
      "17903294 estimating equations GEE\n",
      "17903294 equations GEE\n",
      "17903295 SNPs for generalized estimating equation GEE\n",
      "17903295 for generalized estimating equation GEE\n",
      "17903295 generalized estimating equation GEE\n",
      "17903295 estimating equation GEE\n",
      "17903295 equation GEE\n",
      "17903296 in the Framingham Heart Study FHS\n",
      "17903296 traits: bone mineral density BMD\n",
      "17903296 the Framingham Heart Study FHS\n",
      "17903296 : bone mineral density BMD\n",
      "17903296 Framingham Heart Study FHS\n",
      "17903296 bone mineral density BMD\n",
      "17903296 Heart Study FHS\n",
      "17903296 mineral density BMD\n",
      "17903296 Study FHS\n",
      "17903296 density BMD\n",
      "17903296 used additive generalized estimating equation GEE\n",
      "17903296 additive generalized estimating equation GEE\n",
      "17903296 generalized estimating equation GEE\n",
      "17903296 estimating equation GEE\n",
      "17903296 equation GEE\n",
      "17903297 Brain magnetic resonance imaging MRI\n",
      "17903297 magnetic resonance imaging MRI\n",
      "17903297 resonance imaging MRI\n",
      "17903297 imaging MRI\n",
      "17903297 We conducted a genome-wide association GWA\n",
      "17903297 conducted a genome-wide association GWA\n",
      "17903297 a genome-wide association GWA\n",
      "17903297 genome-wide association GWA\n",
      "17903297 association GWA\n",
      "17903297 relationships via generalized estimating equations GEE\n",
      "17903297 ventricular and white matter hyperintensity WMH\n",
      "17903297 via generalized estimating equations GEE\n",
      "17903297 and white matter hyperintensity WMH\n",
      "17903297 generalized estimating equations GEE\n",
      "17903297 white matter hyperintensity WMH\n",
      "17903297 estimating equations GEE\n",
      "17903297 matter hyperintensity WMH\n",
      "17903297 equations GEE\n",
      "17903297 hyperintensity WMH\n",
      "17903298 traits FPG\n",
      "17903298 (fasting plasma glucose FPG\n",
      "17903298 fasting plasma glucose FPG\n",
      "17903298 plasma glucose FPG\n",
      "17903298 glucose FPG\n",
      "17903298 used additive generalized estimating equations GEE\n",
      "17903298 additive generalized estimating equations GEE\n",
      "17903298 generalized estimating equations GEE\n",
      "17903298 estimating equations GEE\n",
      "17903298 equations GEE\n",
      "17903300 related to multiple cardiovascular disease CVD\n",
      "17903300 to multiple cardiovascular disease CVD\n",
      "17903300 multiple cardiovascular disease CVD\n",
      "17903300 cardiovascular disease CVD\n",
      "17903300 disease CVD\n",
      "17903300 Body mass index BMI\n",
      "17903300 mass index BMI\n",
      "17903300 index BMI\n",
      "17903300 using additive Generalized Estimating Equations GEE\n",
      "17903300 additive Generalized Estimating Equations GEE\n",
      "17903300 Generalized Estimating Equations GEE\n",
      "17903300 Estimating Equations GEE\n",
      "17903300 Equations GEE\n",
      "17903301 responses to standardized treadmill test ETT\n",
      "17903301 to standardized treadmill test ETT\n",
      "17903301 standardized treadmill test ETT\n",
      "17903301 treadmill test ETT\n",
      "17903301 test ETT\n",
      "17903301 vessel diameter, flow-mediated dilation FMD\n",
      "17903301 diameter, flow-mediated dilation FMD\n",
      "17903301 , flow-mediated dilation FMD\n",
      "17903301 flow-mediated dilation FMD\n",
      "17903301 dilation FMD\n",
      "17903301 Generalized estimating equations GEE\n",
      "17903301 estimating equations GEE\n",
      "17903301 equations GEE\n",
      "17903302 GeneChip) associations with systolic SBP\n",
      "17903302 (SBP) and diastolic SBP\n",
      "17903302 (SBP) and diastolic DBP\n",
      "17903302 ) associations with systolic SBP\n",
      "17903302 SBP) and diastolic DBP\n",
      "17903302 associations with systolic SBP\n",
      "17903302 ) and diastolic DBP\n",
      "17903302 with systolic SBP\n",
      "17903302 and diastolic DBP\n",
      "17903302 systolic SBP\n",
      "17903302 diastolic DBP\n",
      "17903303 Subclinical atherosclerosis SCA\n",
      "17903303 atherosclerosis SCA\n",
      "17903303 and carotid intimal medial thickness IMT\n",
      "17903303 carotid intimal medial thickness IMT\n",
      "17903303 intimal medial thickness IMT\n",
      "17903303 medial thickness IMT\n",
      "17903303 thickness IMT\n",
      "17903303 regression with generalized estimating equations GEE\n",
      "17903303 with generalized estimating equations GEE\n",
      "17903303 generalized estimating equations GEE\n",
      "17903303 estimating equations GEE\n",
      "17903303 equations GEE\n",
      "17903304 – including coronary heart disease CHD\n",
      "17903304 including coronary heart disease CHD\n",
      "17903304 coronary heart disease CHD\n",
      "17903304 Cardiovascular disease CVD\n",
      "17903304 heart disease CHD\n",
      "17903304 disease CVD\n",
      "17903304 disease CHD\n",
      "17903304 industrialized countries, cardiovascular disease CVD\n",
      "17903304 countries, cardiovascular disease CVD\n",
      "17903304 , cardiovascular disease CVD\n",
      "17903304 cardiovascular disease CVD\n",
      "17903304 disease CVD\n",
      "511\n"
     ]
    }
   ],
   "source": [
    "new_text_c = list()\n",
    "resolved = {(doc_id, acro) for doc_id, _, acro in resolved_acronyms}\n",
    "for c in text_c:\n",
    "    doc_id, phen, acro = spairtxt2uid(c)\n",
    "    print doc_id, phen, acro\n",
    "    if (doc_id, acro) not in resolved:\n",
    "        new_text_c.append(c)\n",
    "\n",
    "print len(new_text_c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Building feature index...\n",
      "Extracting features...\n",
      "0/5970\n",
      "5000/5970\n"
     ]
    }
   ],
   "source": [
    "from snorkel.features import NgramPairFeaturizer\n",
    "\n",
    "pkl_f = 'acro_text_feats.pkl'\n",
    "try:\n",
    "    with open(pkl_f, 'rb') as f:\n",
    "        featurizer = cPickle.load(f)\n",
    "except:\n",
    "    featurizer = NgramPairFeaturizer()\n",
    "    featurizer.fit_transform(text_c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Training set size: 300\n",
      "Gold set size: 216\n",
      "Positive labels in training set: 54\n",
      "Negative labels in training set: 197\n",
      "Positive labels in gold set: 28\n",
      "Negative labels in gold set: 188\n"
     ]
    }
   ],
   "source": [
    "# Split into train and test set\n",
    "training_candidates = []\n",
    "gold_candidates     = []\n",
    "gold_labels         = []\n",
    "n_half = len(text_c)/2\n",
    "for c in text_c[:n_half]:\n",
    "    uid = spair2uid(c)\n",
    "    if uid in annotations:\n",
    "        gold_candidates.append(c)\n",
    "        gold_labels.append(annotations[uid])\n",
    "    else:\n",
    "        training_candidates.append(c)\n",
    "training_candidates.extend(text_c[n_half:])\n",
    "gold_labels = np.array(gold_labels)\n",
    "print \"Training set size: %s\" % len(training_candidates)\n",
    "print \"Gold set size: %s\" % len(gold_candidates)\n",
    "print \"Positive labels in training set: %s\" % len([c for c in training_candidates if annotations.get(spair2uid(c),0)==1])\n",
    "print \"Negative labels in training set: %s\" % len([c for c in training_candidates if annotations.get(spair2uid(c),0)==-1])\n",
    "print \"Positive labels in gold set: %s\" % len([c for c in gold_candidates if annotations[spair2uid(c)]==1])\n",
    "print \"Negative labels in gold set: %s\" % len([c for c in gold_candidates if annotations[spair2uid(c)]==-1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "# positive LFs\n",
    "def LF_acro_matches(m):\n",
    "    _, phen, acro = spairtxt2uid(m)\n",
    "    words = phen.strip().split()\n",
    "    if len(acro) == len(words):\n",
    "        w_acro = ''.join([w[0] for w in words])\n",
    "        if w_acro.lower() == acro.lower():\n",
    "            return +1\n",
    "    return 0\n",
    "\n",
    "def LF_acro_matches_with_dashes(m):\n",
    "    _, phen, acro = spairtxt2uid(m)\n",
    "    words = re.split(' |-', phen.split())\n",
    "    if len(acro) == len(words):\n",
    "        w_acro = ''.join([w[0] for w in words])\n",
    "        if w_acro.lower() == acro.lower():\n",
    "            return +1\n",
    "    return 0\n",
    "\n",
    "def LF_acro_first_letter(m):\n",
    "    _, phen, acro = spairtxt2uid(m)\n",
    "    if not any(l.islower() for l in phen): return 0\n",
    "    words = phen.strip().split()\n",
    "    if len(acro) <= len(words):\n",
    "        if words[0].lower() == acro[0].lower():\n",
    "            return +1\n",
    "    return 0\n",
    "\n",
    "def LF_acro_second_letter(m):\n",
    "    _, phen, acro = spairtxt2uid(m)\n",
    "    if not any(l.islower() for l in phen): return 0\n",
    "    if len(acro) < 4: return 0\n",
    "    acro = acro[1:]\n",
    "    words = phen.strip().split()\n",
    "    if len(acro) <= len(words):\n",
    "        if words[0].lower() == acro[0].lower():\n",
    "            return +1\n",
    "    return 0\n",
    "\n",
    "# NOTE: not used\n",
    "def LF_acro_second_letter(m):\n",
    "    _, phen, acro = spairtxt2uid(m)\n",
    "    if not any(l.islower() for l in phen): return 0\n",
    "    if len(acro) < 4: return 0\n",
    "    acro = acro[1:]\n",
    "    words = phen.strip().split()\n",
    "    if len(acro) <= len(words):\n",
    "        if words[0].lower() == acro[0].lower():\n",
    "            return +1\n",
    "    return 0\n",
    "\n",
    "def LF_acro_prefix(m):\n",
    "    _, phen, acro = spairtxt2uid(m)\n",
    "    phen = phen.replace('-', '')\n",
    "    if phen[:2].lower() == acro[:2].lower():\n",
    "        return +1\n",
    "    return 0\n",
    "\n",
    "def LF_acro_matches_last_letters(m):\n",
    "    _, phen, acro = spairtxt2uid(m)\n",
    "    words = phen.strip().split()\n",
    "    for r in (1,2):\n",
    "        new_acro = acro[r:]\n",
    "        if len(new_acro) < 3: continue\n",
    "        if len(new_acro) == len(words):\n",
    "            w_acro = ''.join([w[0] for w in words])\n",
    "            if w_acro.lower() == new_acro.lower():\n",
    "                return +1\n",
    "    return 0\n",
    "\n",
    "LF_txt_pos = [LF_acro_matches, LF_acro_first_letter, LF_acro_prefix, LF_acro_matches_last_letters]\n",
    "\n",
    "# negative LFs\n",
    "def LF_no_pos(m):\n",
    "    return -1 if not any(LF(m) for LF in LF_txt_pos) else 0\n",
    "\n",
    "def LF_short(m):\n",
    "    _, phen, acro = spairtxt2uid(m)\n",
    "    return -1 if len(acro) == 1 else 0\n",
    "\n",
    "def LF_lc(m):\n",
    "    _, phen, acro = spairtxt2uid(m)\n",
    "    return -1 if all(l.islower() for l in acro) else 0\n",
    "\n",
    "def LF_uc(m):\n",
    "    _, phen, acro = spairtxt2uid(m)\n",
    "    return -1 if not any(l.islower() for l in phen) else 0\n",
    "\n",
    "LF_txt_neg = [LF_no_pos, LF_short, LF_lc, LF_uc]\n",
    "\n",
    "LF_txt = LF_txt_pos + LF_txt_neg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Applying LFs...\n",
      "Featurizing...\n",
      "Building feature index...\n",
      "Extracting features...\n",
      "0/5260\n",
      "5000/5260\n",
      "============================================================\n",
      "LF Summary Statistics: 8 LFs applied to 300 candidates\n",
      "------------------------------------------------------------\n",
      "Coverage (candidates w/ > 0 labels):\t\t100.00%\n",
      "Overlap (candidates w/ > 1 labels):\t\t20.67%\n",
      "Conflict (candidates w/ conflicting labels):\t0.00%\n",
      "============================================================\n"
     ]
    }
   ],
   "source": [
    "from snorkel.snorkel import TrainingSet\n",
    "from snorkel.features import NgramFeaturizer\n",
    "\n",
    "training_set = TrainingSet(training_candidates, LF_txt, featurizer=NgramPairFeaturizer())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[<matplotlib.axes.AxesSubplot object at 0x128f4f190>]], dtype=object)"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAArwAAAEKCAYAAADwwYa2AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAFy9JREFUeJzt3XuMrHddBvDntKdAoYVyMVzk0GMUsBD0FAKCNxYxiki5\nBJRLuFSIgYBAMRrEAEW5RCTAiRpQ0wglXARBCEhBEDogQfBCT0GQqyxQhHItBbkI7fGP9112up2e\n/c5033feeefzSTZn3pnZ2d88Z3b2u+8+804CAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAFwtxy17\nAQBj5MkV4IoOJPn7JF9K8pUkf55kX5KnJtlMcnGSc5Nct73+W5M8bsdtXJjkvu3pn0zy9iRfTfLR\nJL8xdb2XJnlxkvOSfCvJRpJfT3JBkm8k+WySs3fc9sOTfKZd29aa7t5eti/JHyT5ZHv5q5Ncv3zP\nAQAYvePTDKvPT3Jikmsk+bkkj0zyiSQHk1wnyeuSvKz9nIclec/UbdwmydeTnNBe93NJHpFmB8Pp\nSb7cXidpBt5Lktyl3b5mkrsmuW27fbskX0xyn6nb/maSn21v/3lJ/i/JL7WXn5XkvUlu1l7+l0le\nuUAOAACM1F3S7Nnd+devdyR5zNT2rdIMmsclOTnN3tkD7WXPTnJOe/qBSd6947b+KsnT29MvbT+O\n5XCSF7Snn57kFVOXnZjke9keeD8ydTpJbjq1ToC15UkQYNuBNHWBy3ecf9P2/C2fTbI/yY3T7HF9\nc5IHt5c9MNtD6alJfibNHt+tj4e0n5ckR9PsAZ72M0nOTzN4X5Lk0Ulu2F52syQXTV33O2mqElsO\nJnn91Nf6SJIfTH09gLVk4AXY9rkkt0hTbZj2P2mGyS23SDNIXtxuvyrNwHuXNHtdz2/P/2ySd6Xp\n0W59nJwrd36nvTLJG5LcPMkpaWoJ+6bWcfOp656Y7WF46+vdY8fXu3aSLxzj6wEAsEaOS3IkTTf2\n2kmulabD+6gkH08z9J6U5LXZ7vAmTdf3a0nelqb/u+WkNC8qe2iaTu0JSe6Y5oVsSVNneOaONVyc\n5oVpSXKndnvra902yaVpButrJPnTXLnDe36agTxJfiTJvWt3HQCAdXEgTS3gK2leYHY4zR7Wp6XZ\ng/qlNAPo9XZ83jlJLktyhx3n3yrJP2T7qA//lOSn2stekuSPd1z//mmG5EuTvCnJn+WKw/UjcsWj\nNFyUZihPu84npTkaxKVpjtbwrNrdBhivfbtfJcn2k+9lSb6fZq8DAMt1Upqu7k/kih1jABbw6SQ3\nWPYiAMgZaeoW10nT7/2P5S4HYPjmedFadW8wAN25d5LPtx8/nuRBy10OwHj8d5q9CP+e5LeXvBYA\nANhzN2n//ZE0r2D+hSWuBQAAyvYXr/fF9t8vp3n18p2S/HN73tG9XhQAAMzwqTQv1J1LpcN77TQH\nSk+aF0n8SpIPXfEqR9fw4x05/fS75ejRo+WPs88+e67rr+uHnGQlJzkN/UNWcpLVcj7SvHZhbpU9\nvDdOs1d36/qvSHNwdea0ubm57CWsBDnVyapGTjVyqpNVjZzqZNWtysD76SSHul4IAAB0YZ7DknE1\nnXnmmctewkqQU52sauRUI6c6WdXIqU5W3dqLY+seXc/Xrb0zp5/+rHzgA+9c9kIAANbCvn37kgXm\nV3t4ezSZTJa9hJUgpzpZ1cipRk51sqqRU52sumXgBQBg1FQaFqbSAADQJ5UGAACYwcDbI/2cGjnV\nyapGTjVyqpNVjZzqZNUtAy8AAKOmw7swHV4AgD7p8AIAwAwG3h7p59TIqU5WNXKqkVOdrGrkVCer\nbhl4AQAYNR3ehenwAgD0SYcXAABmMPD2SD+nRk51sqqRU42c6mRVI6c6WXXLwAsAwKjp8C5MhxcA\noE86vAAAMIOBt0f6OTVyqpNVjZxq5FQnqxo51cmqWwZeAABGTYd3YTq8AAB90uEFAIAZDLw90s+p\nkVOdrGrkVCOnOlnVyKlOVt0y8AIAMGo6vAvT4QUA6JMOLwAAzGDg7ZF+To2c6mRVI6caOdXJqkZO\ndbLqloEXAIBR0+FdmA4vAECfdHgBAGAGA2+P9HNq5FQnqxo51cipTlY1cqqTVbcMvAAAjJoO78J0\neAEA+qTDCwAAMxh4e6SfUyOnOlnVyKlGTnWyqpFTnay6ZeAFAGDUqh2I45P8e5KLkpyx4zIdXgAA\nOtd1h/eJST6S9ZxsAQBYYZWB9+ZJ7pnknOzNUR3Wln5OjZzqZFUjpxo51cmqRk51supWZeB9YZLf\nT3J5x2sBAIA9t3+Xy++V5EtJLkiycdVXOzPJwfb0KUkOTV190v47tu12q/2NbGNjY9ftjY2Nua6/\nzttbhrKeoW5vnTeU9dhe7e2t84ayniFvb3g+L29vGcp6hrq9dd5Q1jOU7a3Tm5ubuTp2qyg8J8nD\nkvwgybWSXDfJ65I8fOo6XrQGAEDnunrR2h8mOZDkx5I8KMk7c8Vhlzns/G2X2eRUJ6saOdXIqU5W\nNXKqk1W3dht4d1rHXbkAAKywvTjqgkoDAACd6/o4vAAAsJIMvD3Sz6mRU52sauRUI6c6WdXIqU5W\n3TLwAgAwajq8C9PhBQDokw4vAADMYODtkX5OjZzqZFUjpxo51cmqRk51suqWgRcAgFHT4V2YDi8A\nQJ90eAEAYAYDb4/0c2rkVCerGjnVyKlOVjVyqpNVtwy8AACMmg7vwnR4AQD6pMMLAAAzGHh7pJ9T\nI6c6WdXIqUZOdbKqkVOdrLpl4AUAYNR0eBemwwsA0CcdXgAAmMHA2yP9nBo51cmqRk41cqqTVY2c\n6mTVLQMvAACjpsO7MB1eAIA+6fACAMAMBt4e6efUyKlOVjVyqpFTnaxq5FQnq24ZeAEAGDUd3oXp\n8AIA9EmHFwAAZjDw9kg/p0ZOdbKqkVONnOpkVSOnOll1y8ALAMCo6fAuTIcXAKBPOrwAADCDgbdH\n+jk1cqqTVY2cauRUJ6saOdXJqlsGXgAARk2Hd2E6vAAAfdLhBQCAGQy8PdLPqZFTnaxq5FQjpzpZ\n1cipTlbdMvACADBqlQ7EtZK8K8k1k+xP8tokz5i6XIcXAIDOLdrh3V+4zneT3C3Jt9vrvyfJW5K8\nf94vBgAAfatWGr7d/nuNJCckubyb5Yybfk6NnOpkVSOnGjnVyapGTnWy6lZ14D0uyZEkFyd5W5J/\n62xFAACwh+btQFwvyeuTPD7Jh9vzdHgBAOhclx3ead9IMklyj2wPvEnOTHKwPX1KkkNJNtrtSfvv\n2LbbrfZPEBsbG7Zt27Zt27Zt27b3cHvr9ObmZq6OyoR8oyQ/SHJJkhOT/GOSP0lyXnu5PbxFk8nk\nh/+RXDU51cmqRk41cqqTVY2c6mRV0+Ue3psmOTfJ8Wm6vK/O9rALAACDNveEPIM9vAAAdG7RPbzH\n7f1SAABgOAy8PZouYHPV5FQnqxo51cipTlY1cqqTVbcMvAAAjJoO78J0eAEA+qTDCwAAMxh4e6Sf\nUyOnOlnVyKlGTnWyqpFTnay6ZeAFAGDUdHgXpsMLANAnHV4AAJjBwNsj/ZwaOdXJqkZONXKqk1WN\nnOpk1S0DLwAAo6bDuzAdXgCAPunwAgDADAbeHunn1MipTlY1cqqRU52sauRUJ6tuGXgBABg1Hd6F\n6fACAPRJhxcAAGYw8PZIP6dGTnWyqpFTjZzqZFUjpzpZdcvACwDAqOnwLkyHFwCgTzq8AAAwg4G3\nR/o5NXKqk1WNnGrkVCerGjnVyapbBl4AAEZNh3dhOrwAAH3S4QUAgBkMvD3Sz6mRU52sauRUI6c6\nWdXIqU5W3TLwAgAwajq8C9PhBQDokw4vAADMYODtkX5OjZzqZFUjpxo51cmqRk51suqWgRcAgFHT\n4V2YDi8AQJ90eAEAYAYDb4/0c2rkVCerGjnVyKlOVjVyqpNVtwy8AACMmg7vwnR4AQD61GWH90CS\n85N8JMl/JnnCvF8EAACWpTLwfj/Jk5LcJsmdkzwuyWldLmqs9HNq5FQnqxo51cipTlY1cqqTVbcq\nA+8XkxxpT38ryX8luVlnKwIAgD00bwfiYJJ3JbltmuE30eFd9kIAANZCH8fhPSnJa5M8MdvDLgAA\nDNr+4vVOSPK6JC9P8oYrX3xmmp2/SXJKkkNJNtrtSfvv2LaTCy9879ZvGmvn5JOvn0sv/doPO0cb\nGxtJsifbR44cyVlnnbVntzfm7cOHD+fQoUODWc9Qt7fOG8p6hrrt8VTf3vnYWvZ6hrrt+dz339Xd\n3jq9ubmZq6Myre1Lcm6Sr6Z58dpOa1tpSO6e+e77JMlGF4tZgn05erSb//fJZPLDBzzHJqsaOdXI\nqU5WNXKqk1XNopWGyif8fJJ3J/lgtqe7pyR5a3vawLuWuht4AQBmWXTgrVQa3hPvyAYAwIoyyPZq\nsuwFrITp3g7HJqsaOdXIqU5WNXKqk1W3DLwAAIzaXhxiQId3LenwAgD96uM4vAAAsHIMvL2aLHsB\nK0GPqU5WNXKqkVOdrGrkVCerbhl4AQAYNR3ehenw6vACAH3S4QUAgBkMvL2aLHsBK0GPqU5WNXKq\nkVOdrGrkVCerbhl4AQAYNR3ehenw6vACAH3S4QUAgBkMvL2aLHsBK0GPqU5WNXKqkVOdrGrkVCer\nbhl4AQAYNR3ehenw6vACAH3S4QUAgBkMvL2aLHsBK0GPqU5WNXKqkVOdrGrkVCerbhl4AQAYNR3e\nhenw6vACAH3S4QUAgBkMvL2aLHsBK0GPqU5WNXKqkVOdrGrkVCerbhl4AQAYNR3ehenw6vACAH3S\n4QUAgBkMvL2aLHsBK0GPqU5WNXKqkVOdrGrkVCerbhl4AQAYNR3ehenw6vACAH3S4QUAgBkMvL2a\nLHsBK0GPqU5WNXKqkVOdrGrkVCerbhl4AQAYNR3ehenw6vACAH3S4QUAgBkMvL2aLHsBK0GPqU5W\nNXKqkVOdrGrkVCerbhl4AQAYNR3ehenw6vACAH3S4QUAgBkqA+/fJLk4yYc6XssamCx7AStBj6lO\nVjVyqpFTnaxq5FQnq25VBt6XJLlH1wsBAIAuVDsQB5O8KcntZlymw7uWdHgBgH7p8AIAwAz79+Zm\nzkyzEzhJTklyKMlGuz1p/x3bdna5fNb29Od2vb6ut9uttnO0sbGxZ9tHjhzJWWed1dntj2n78OHD\nOXTo0GDWM9TtrfOGsp6hbns81bd3PraWvZ6hbq/a8/k973lGvvOdb2UdnXjiSTnvvDcN6v9jy2Qy\nyebm5kL3a4tKw8IWqTRMsj04rrruKg2TyeSHD3iOTVY1cqqRU52salYtp+bP5cuaaSZZ7oywGlXF\nRSsNBt6F6fCuwjcGAFQtd+BdttX4ud5lh/dVSd6b5FZJPpfkt+b9IgAAsCyVgffBSW6W5JpJDqQ5\nTBkLmSx7ASthurfDscmqRk41cqqTVY2c5jFZ9gJGrTLwAgDAypq7AzGDDu9aWo2uDwBU6fAO/747\nDi8AAMxg4O3VZNkLWAk6X3WyqpFTjZzqZFUjp3lMlr2AUTPwAgAwajq8C9PhXYWuDwBU6fAO/77r\n8AIAwAwG3l5Nlr2AlaDzVSerGjnVyKlOVjVymsdk2QsYNQMvAACjpsO7MB3eVej6AECVDu/w77sO\nLwAAzGDg7dVk2QtYCTpfdbKqkVONnOpkVSOneUyWvYBRM/ACADBqOrwL0+Fdha4PAFTp8A7/vuvw\nAgDADAbeXk2WvYCVoPNVJ6saOdXIqU5WNXKax2TZCxg1Ay8AAKOmw7swHd5V6PoAQJUO7/Dvuw4v\nAADMYODt1WTZC1gJOl91sqqRU42c6mRVI6d5TJa9gFEz8AIAMGo6vAvT4V2Frg8AVOnwDv++6/AC\nAMAMBt5eTZa9gJWg81Unqxo51cipTlY1cprHZNkLGDUDLwAAo6bDuzAd3lXo+gBAlQ7v8O+7Di8A\nAMxg4O3VZNkLWAk6X3WyqpFTjZzqZFUjp3lMlr2AUTPwAgAwajq8C9PhXYWuDwBU6fAO/77r8AIA\nwAwG3l5Nlr2AlaDzVSerGjnVyKlOVjVymsdk2QsYNQMvAACjpsO7MB3eVej6AECVDu/w77sOLwAA\nzFAZeO+R5KNJPpHkyd0uZ+wmy17AStD5qpNVjZxq5FQnqxo5zWOy7AWM2m4D7/FJ/iLN0HubJA9O\nclrXixqvI8tewEo4ckROVbKqkVONnOpkVSOneciqS7sNvHdK8skkm0m+n+Rvk9yn4zWN2CXLXsBK\nuOQSOVXJqkZONXKqk1WNnOYhqy7tNvD+aJLPTW1f1J4HAAArYf8Cn3Oll/Bd97pn7MFSVstll30l\n//u/837WZgcrGZ/Nzc1lL2FlyKpGTjVyqpNVjZzmsbnsBYzabod1uHOSZ6Tp8CbJU5JcnuS5U9f5\nZJIf3/OVAQDAFX0qyU/s9Y3ub2/4YJJrpGlUe9EaAACj8mtJPpZmT+5TlrwWAAAAAAAWsdsbUFwz\nyavby9+X5NT+ljY4u2X1i0k+kOZQb/fvcV1Ds1tOv5vkw0kuTPJPSW7R39IGZ7esHpPkg0kuSPLP\nWd/qUfWNch6Q5vUIt+9jUQO0W05nJvlymsfTBUke2dvKhqfymPrNNM9V/5nkFT2ta2h2y+kF2X48\nfSzJ1/tb2qDsltMtkpyfZka4MM1f2dfVblmdmuQdaXI6P3t0FLHj01QaDiY5IbO7vI9N8qL29APT\nHLN3HVWyOjXJ7ZKcm/UdeCs5bSS5Vnv6MfGYOpirzurkqdNnJHlLLysblkpOSZPVu5O8N+s58FZy\nekSSP+t3WYNUyeqWaYaT67XbN+prcQNS/d7b8jtJzul+WYNTyemvkzy6PX1akk/3tbiBqWT1d0ke\n1p6+W5KXHesGK28tnNTegOLeaQa4JHldkrsXb3tsKll9JsmH0uxhWleVnCZJvtuefn+Sm/e0tqGp\nZPXNqdMnZT0fW9U3ynlmmiPNfC+7H6lmjCo57ct6ZrNTJavfTvOOpN9ot7/S1+IGZN43qXpIkld1\nv6zBqeR0ebZ/eTolyef7WtzAVLI6Lc0e3qSZF475xmjVgbfyBhTT1/lBmm/+GxRvf0y8WUfNvDk9\nKsl5na5ouKpZPTbNE8Rzkzyhh3UNTSWn09vz3txuX+m44mugktPRNH99ujDNXpR1/WWzktUtk9w6\nyXuS/EuSX+1naYMyz/P5qWn22r2z4zUNUSWnZyR5aHu9Nyd5fC8rG55KVhemqaclyf3S/PXu+ld1\ng9WBd5adPyhm7Q1Yxx8ms8ih5qpyemiaPz0/r8e1DN2srF6U5tiET07ytH6XM1jTOR2X5IVJfm/q\nPHsxGzsfT29KM5j8dJr+/LlX+oz1tTOr/Wm+7+6a5MFp/lR/vZ2ftIau6vn8QWl+ifJzsbEzh4ck\neUmSA0numeTlva9ouHZm9Xtpvu8+kOa1UZ9Ps8N1purAe1Ga8LccyJV3s1+U7RcV7U/zDb+OpfRK\nVtPW9Zu+mtMvJ/nDNJWZ7/ewriGa9zH16iT37XRFw7RbTicnuW2aP319Os0b67wx69fjrTyevpbt\n77dzktyhh3UNUfVn3xuTXJbmz68fSwcHxR+4eZ6jHpj1rDMktZwemeQ17en3pXkdyzr2witZfSHN\nX6Jun+Sp7XnfzNVUeQOKxyZ5cXv6QVnfFxjN82YdL836vmitktPp8U5+SS2r6R+wZyT5115WNizz\nvlHO+Vm/YTep5XSTqdP3S/MCv3VUyepX0zyXJ81g8tkc48+qI1X93rt11vdFWEktp/PSvGg07WXr\n2uGtZHXDbO+4fXaaOsiemPUGFH+U5odr0hyW7DXZPizZwb36witot6zumKab8q00L3D4UN8LHIir\nyule7em3p/kNbutQNm/oe4EDsttj6nCaQyJdkKbEv66HJdstp2nrOvAmu+f0nDSPpyNpHk+36nuB\nA1J5TD0/zWHJPpjmEGXrqJLT2WkeW+tst5xOS9MHP5Lm+fyX+17ggOyW1QOSfLy9zl+nOZoDAAAA\nAAAAAAAAAAAAAAAAAAAAAAAAAAAAY/P/faQu4d+xjEoAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x128ebdfd0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "lf_stats = training_set.lf_stats()\n",
    "lf_stats[:5]\n",
    "lf_stats.hist(\"coverage\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "============================================================\n",
      "Testing mu = 1.00e-05, lf_w0 = 1.00e+00\n",
      "============================================================\n",
      "Begin training for rate=0.01, mu=1e-05\n",
      "\tLearning epoch = 0\tGradient mag. = 0.035677\n",
      "\tLearning epoch = 250\tGradient mag. = 0.044917\n",
      "\tLearning epoch = 500\tGradient mag. = 0.054043\n",
      "\tLearning epoch = 750\tGradient mag. = 0.065129\n",
      "Final gradient magnitude for rate=0.01, mu=1e-05: 0.078\n",
      "Applying LFs...\n",
      "Featurizing...\n",
      "============================================================\n",
      "Testing mu = 1.00e-05, lf_w0 = 2.00e+00\n",
      "============================================================\n",
      "Begin training for rate=0.01, mu=1e-05\n",
      "\tLearning epoch = 0\tGradient mag. = 0.061755\n",
      "\tLearning epoch = 250\tGradient mag. = 0.072695\n",
      "\tLearning epoch = 500\tGradient mag. = 0.081157\n",
      "\tLearning epoch = 750\tGradient mag. = 0.090401\n",
      "Final gradient magnitude for rate=0.01, mu=1e-05: 0.101\n",
      "============================================================\n",
      "Testing mu = 1.00e-07, lf_w0 = 1.00e+00\n",
      "============================================================\n",
      "Begin training for rate=0.01, mu=1e-07\n",
      "\tLearning epoch = 0\tGradient mag. = 0.035677\n",
      "\tLearning epoch = 250\tGradient mag. = 0.044941\n",
      "\tLearning epoch = 500\tGradient mag. = 0.054095\n",
      "\tLearning epoch = 750\tGradient mag. = 0.065213\n",
      "Final gradient magnitude for rate=0.01, mu=1e-07: 0.078\n",
      "============================================================\n",
      "Testing mu = 1.00e-07, lf_w0 = 2.00e+00\n",
      "============================================================\n",
      "Begin training for rate=0.01, mu=1e-07\n",
      "\tLearning epoch = 0\tGradient mag. = 0.061755\n",
      "\tLearning epoch = 250\tGradient mag. = 0.072710\n",
      "\tLearning epoch = 500\tGradient mag. = 0.081187\n",
      "\tLearning epoch = 750\tGradient mag. = 0.090445\n",
      "Final gradient magnitude for rate=0.01, mu=1e-07: 0.101\n"
     ]
    }
   ],
   "source": [
    "from snorkel.snorkel import Learner\n",
    "import snorkel.learning\n",
    "from snorkel.learning import LogReg\n",
    "\n",
    "text_learner = Learner(training_set, model=LogReg())\n",
    "\n",
    "# Splitting into CV and test set\n",
    "n_half = len(gold_candidates)/2\n",
    "test_candidates = gold_candidates[:n_half]\n",
    "test_labels     = gold_labels[:n_half]\n",
    "cv_candidates   = gold_candidates[n_half:]\n",
    "cv_labels       = gold_labels[n_half:]\n",
    "\n",
    "from snorkel.learning_utils import GridSearch\n",
    "gs       = GridSearch(text_learner, ['mu', 'lf_w0'], [[1e-5, 1e-7],[1.0,2.0]])\n",
    "gs_stats = gs.fit(cv_candidates, cv_labels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mu</th>\n",
       "      <th>lf_w0</th>\n",
       "      <th>Prec.</th>\n",
       "      <th>Rec.</th>\n",
       "      <th>F1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td> 0.00001</td>\n",
       "      <td> 1</td>\n",
       "      <td> 0.647059</td>\n",
       "      <td> 0.55</td>\n",
       "      <td> 0.594595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td> 0.00001</td>\n",
       "      <td> 2</td>\n",
       "      <td> 0.700000</td>\n",
       "      <td> 0.70</td>\n",
       "      <td> 0.700000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 1</td>\n",
       "      <td> 0.647059</td>\n",
       "      <td> 0.55</td>\n",
       "      <td> 0.594595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 2</td>\n",
       "      <td> 0.700000</td>\n",
       "      <td> 0.70</td>\n",
       "      <td> 0.700000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        mu  lf_w0     Prec.  Rec.        F1\n",
       "0  0.00001      1  0.647059  0.55  0.594595\n",
       "1  0.00001      2  0.700000  0.70  0.700000\n",
       "2  0.00000      1  0.647059  0.55  0.594595\n",
       "3  0.00000      2  0.700000  0.70  0.700000\n",
       "\n",
       "[4 rows x 5 columns]"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gs_stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Applying LFs...\n",
      "Featurizing...\n",
      "========================================\n",
      "Test set size:\t108\n",
      "----------------------------------------\n",
      "Precision:\t1.0\n",
      "Recall:\t\t0.875\n",
      "F1 Score:\t0.933333333333\n",
      "----------------------------------------\n",
      "TP: 7 | FP: 0 | TN: 100 | FN: 1\n",
      "========================================\n"
     ]
    }
   ],
   "source": [
    "text_learner.test_wmv(test_candidates, test_labels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Applying LFs...\n",
      "Featurizing...\n",
      "[('17903292', u'Glomerular Filtration Rate', u'GFR'), ('17903292', u'Glomerular Filtration Rate', u'GFR'), ('17903292', u'Glomerular Filtration Rate', u'GFR'), ('17903292', u'Thyroid stimulation hormone', u'TSH'), ('17903292', u'Thyroid stimulation hormone', u'TSH'), ('17903292', u'Thyroid stimulation hormone', u'TSH'), ('17903292', u'Follicle stimulating hormone', u'FSH'), ('17903296', u'Bone mineral density', u'BMD'), ('17903296', u'Bone mineral density', u'BMD'), ('17903296', u'Bone mineral density', u'BMD')]\n",
      "17903292 -1.0 1\n",
      "Phrase('17903292', 0, 80, 0, u'Dehydroepiandrosterone sulfate (DHEAS)')\n",
      "Dehydroepiandrosterone sulfate (DHEAS) (DHEAS)\n",
      "[0, 0, 0, 0, -1, 0, 0, 0]\n",
      "\n",
      "17903297 -1.0 1\n",
      "Phrase('17903297', 0, 43, 0, u'Hippocampal Volume (AHPV)')\n",
      "Hippocampal Volume (AHPV) (AHPV)\n",
      "[0, 0, 0, 0, -1, 0, 0, 0]\n",
      "\n",
      "17903297 -1.0 1\n",
      "Phrase('17903297', 0, 87, 0, u'Boston Naming Test (Nam)')\n",
      "Boston Naming Test (Nam) (Nam)\n",
      "[0, 0, 0, 0, -1, 0, 0, 0]\n",
      "\n",
      "17903297 -1.0 1\n",
      "Phrase('17903297', 0, 87, 0, u'Boston Naming Test (Nam)')\n",
      "Boston Naming Test (Nam) (Nam)\n",
      "[0, 0, 0, 0, -1, 0, 0, 0]\n",
      "\n",
      "17903297 1.0 -1\n",
      "Phrase('17903297', 0, 87, 0, u'Boston Naming Test (Nam)')\n",
      "Naming Test (Nam) (Nam)\n",
      "[0, 0, 1, 0, 0, 0, 0, 0]\n",
      "\n",
      "17903297 1.0 -1\n",
      "Phrase('17903297', 0, 87, 0, u'Boston Naming Test (Nam)')\n",
      "Naming Test (Nam) (Nam)\n",
      "[0, 0, 1, 0, 0, 0, 0, 0]\n",
      "\n",
      "17903297 -1.0 1\n",
      "Phrase('17903297', 0, 87, 0, u'Boston Naming Test (Nam)')\n",
      "Boston Naming Test (Nam) (Nam)\n",
      "[0, 0, 0, 0, -1, 0, 0, 0]\n",
      "\n",
      "17903297 1.0 -1\n",
      "Phrase('17903297', 0, 87, 0, u'Boston Naming Test (Nam)')\n",
      "Naming Test (Nam) (Nam)\n",
      "[0, 0, 1, 0, 0, 0, 0, 0]\n",
      "\n",
      "17903297 1.0 -1\n",
      "Phrase('17903297', 0, 93, 0, u'Similarities (Sim)')\n",
      "Similarities (Sim) (Sim)\n",
      "[0, 0, 1, 0, 0, 0, 0, 0]\n",
      "\n",
      "17903297 1.0 -1\n",
      "Phrase('17903297', 0, 93, 0, u'Similarities (Sim)')\n",
      "Similarities (Sim) (Sim)\n",
      "[0, 0, 1, 0, 0, 0, 0, 0]\n",
      "\n",
      "17903297 1.0 -1\n",
      "Phrase('17903297', 0, 93, 0, u'Similarities (Sim)')\n",
      "Similarities (Sim) (Sim)\n",
      "[0, 0, 1, 0, 0, 0, 0, 0]\n",
      "\n",
      "17903297 -1.0 1\n",
      "Phrase('17903297', 0, 99, 0, u'Wide-Range Achievement Test (WRAT)')\n",
      "Wide-Range Achievement Test (WRAT) (WRAT)\n",
      "[0, 0, 0, 0, -1, 0, 0, 0]\n",
      "\n",
      "17903298 -1.0 1\n",
      "Phrase('17903298', 0, 18, 0, u'28 yr time averaged FPG (tFPG)')\n",
      "time averaged FPG (tFPG) (tFPG)\n",
      "[0, 0, 0, 0, -1, 0, 0, 0]\n",
      "\n",
      "17903301 -1.0 1\n",
      "Phrase('17903301', 0, 50, 0, u'Stage 2 Exercise diastolic blood pressure (DBP)')\n",
      "blood pressure (DBP) (DBP)\n",
      "[0, 0, 0, 0, -1, 0, 0, 0]\n",
      "\n",
      "17903301 -1.0 1\n",
      "Phrase('17903301', 0, 50, 0, u'Stage 2 Exercise diastolic blood pressure (DBP)')\n",
      "blood pressure (DBP) (DBP)\n",
      "[0, 0, 0, 0, -1, 0, 0, 0]\n",
      "\n",
      "17903301 -1.0 1\n",
      "Phrase('17903301', 0, 50, 0, u'Stage 2 Exercise diastolic blood pressure (DBP)')\n",
      "blood pressure (DBP) (DBP)\n",
      "[0, 0, 0, 0, -1, 0, 0, 0]\n",
      "\n",
      "17903301 -1.0 1\n",
      "Phrase('17903301', 0, 88, 0, u'BA flow-mediated dilation (FMD) percent')\n",
      "flow-mediated dilation (FMD) (FMD)\n",
      "[0, 0, 0, 0, -1, 0, 0, 0]\n",
      "\n",
      "17903292 1.0 -1\n",
      "Sentence(Document('17903292', Corpus (GWAS Text Corpus)), 3, u'We tested for association between the Affymetrix GeneChip Human Mapping 100K single nucleotide polymorphism (SNP) set and measures of kidney function and endocrine traits.')\n",
      "single nucleotide polymorphism (SNP) (SNP)\n",
      "[1, 0, 0, 0, 0, 0, 0, 0]\n",
      "\n",
      "17903294 -1.0 1\n",
      "Sentence(Document('17903294', Corpus (GWAS Text Corpus)), 1, u'Increased circulating levels of hemostatic factors as well as anemia have been associated with increased risk of cardiovascular disease (CVD).')\n",
      "cardiovascular disease (CVD) (CVD)\n",
      "[0, 0, 0, 0, -1, 0, 0, 0]\n",
      "\n",
      "17903294 1.0 -1\n",
      "Sentence(Document('17903294', Corpus (GWAS Text Corpus)), 3, u'We sought to confirm known putative loci and identify novel loci that may influence either trait in genome-wide association and linkage analyses using the Affymetrix GeneChip 100K single nucleotide polymorphism (SNP) set.')\n",
      "single nucleotide polymorphism (SNP) (SNP)\n",
      "[1, 0, 0, 0, 0, 0, 0, 0]\n",
      "\n",
      "17903297 -1.0 1\n",
      "Sentence(Document('17903297', Corpus (GWAS Text Corpus)), 2, u'We conducted a genome-wide association (GWA) and linkage analysis exploring the genetic basis of these endophenotypes in a community-based sample.')\n",
      "genome-wide association (GWA) (GWA)\n",
      "[0, 0, 0, 0, -1, 0, 0, 0]\n",
      "\n",
      "17903300 -1.0 1\n",
      "Sentence(Document('17903300', Corpus (GWAS Text Corpus)), 1, u'Obesity is related to multiple cardiovascular disease (CVD) risk factors as well as CVD and has a strong familial component.')\n",
      "cardiovascular disease (CVD) (CVD)\n",
      "[0, 0, 0, 0, -1, 0, 0, 0]\n",
      "\n",
      "17903301 -1.0 1\n",
      "Sentence(Document('17903301', Corpus (GWAS Text Corpus)), 6, u'BA measures (single exam) included vessel diameter, flow-mediated dilation (FMD), and baseline and hyperemic flow responses.')\n",
      "flow-mediated dilation (FMD) (FMD)\n",
      "[0, 0, 0, 0, -1, 0, 0, 0]\n",
      "\n",
      "17903301 1.0 -1\n",
      "Sentence(Document('17903301', Corpus (GWAS Text Corpus)), 7, u'Generalized estimating equations (GEE), family-based association tests (FBAT) and variance-components linkage were used to relate multivariable-adjusted trait residuals to 70,987 SNPs (Human 100K GeneChip, Affymetrix) restricted to autosomal SNPs with minor allele frequency \\u22650.10, genotype call rate \\u22650.80, and Hardy-Weinberg equilibrium p \\u2265 0.001.')\n",
      "Generalized estimating equations (GEE) (GEE)\n",
      "[1, 0, 1, 0, 0, 0, 0, 0]\n",
      "\n",
      "17903303 -1.0 1\n",
      "Sentence(Document('17903303', Corpus (GWAS Text Corpus)), 1, u'Subclinical atherosclerosis (SCA) measures in multiple arterial beds are heritable phenotypes that are associated with increased incidence of cardiovascular disease.')\n",
      "Subclinical atherosclerosis (SCA) (SCA)\n",
      "[0, 0, 0, 0, -1, 0, 0, 0]\n",
      "\n",
      "17903304 -1.0 1\n",
      "Sentence(Document('17903304', Corpus (GWAS Text Corpus)), 1, u'Cardiovascular disease (CVD) and its most common manifestations \\u2013 including coronary heart disease (CHD), stroke, heart failure (HF), and atrial fibrillation (AF) \\u2013 are major causes of morbidity and mortality.')\n",
      "Cardiovascular disease (CVD) (CVD)\n",
      "[0, 0, 0, 0, -1, 0, 0, 0]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "text_preds = text_learner.predict_wmv(text_c)\n",
    "text_acronyms = [spairtxt2uid(c) for (c, p) in zip(text_c, text_preds) if p == 1]\n",
    "print text_acronyms[:10]\n",
    "mislabeled_cand = [(c,p, annotations.get(spair2uid(c), None)) for c, p in zip(text_c, text_preds) if p != annotations.get(spair2uid(c), p)]\n",
    "for (c,p,g) in mislabeled_cand[:50]:\n",
    "    print c.span0.context.document.name, p, g\n",
    "    print c.span0.context    \n",
    "    print c.span0.get_span(), c.span1.get_span()\n",
    "    print [LF(c) for LF in LF_txt]\n",
    "    print"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### Collect all the candidates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Applying LFs...\n",
      "Featurizing...\n",
      "[('17903293', u'CD40 Ligand, serum & plasma', u'CD40L'), ('17903293', u'C-reactive protein', u'CRP'), ('17903293', u'Intercellular adhesion molecule-1', u'ICAM1'), ('17903293', u'Interleukin-6', u'IL6'), ('17903293', u'Urinary isoprostanes/creatinine', u'IsoCrUrine'), ('17903293', u'Monocyte chemoattractant protein-1', u'MCP1'), ('17903293', u'Myeloperoxidase', u'MPO'), ('17903293', u'Osteoprotegerin', u'OPG'), ('17903293', u'P-selectin', u'Pselectin'), ('17903293', u'Tumor necrosis factor alpha', u'TNFA')]\n",
      "Applying LFs...\n",
      "Featurizing...\n",
      "[('17903292', u'Glomerular Filtration Rate', u'GFR'), ('17903292', u'Glomerular Filtration Rate', u'GFR'), ('17903292', u'Glomerular Filtration Rate', u'GFR'), ('17903292', u'Thyroid stimulation hormone', u'TSH'), ('17903292', u'Thyroid stimulation hormone', u'TSH'), ('17903292', u'Thyroid stimulation hormone', u'TSH'), ('17903292', u'Follicle stimulating hormone', u'FSH'), ('17903296', u'Bone mineral density', u'BMD'), ('17903296', u'Bone mineral density', u'BMD'), ('17903296', u'Bone mineral density', u'BMD')]\n"
     ]
    }
   ],
   "source": [
    "table_preds = learner.predict_wmv(candidates)\n",
    "table_acronyms = [spair2uid(c) for (c, p) in zip(candidates, preds) if p == 1]\n",
    "print table_acronyms[:10]\n",
    "\n",
    "text_preds = text_learner.predict_wmv(text_c)\n",
    "text_acronyms = [spairtxt2uid(c) for (c, p) in zip(text_c, text_preds) if p == 1]\n",
    "print text_acronyms[:10]\n",
    "\n",
    "acronyms = table_acronyms + text_acronyms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# store relations to annotate\n",
    "with open('acronyms.extracted.tsv', 'w') as f:\n",
    "    for doc_id, str1, str2 in acronyms:\n",
    "        f.write('%s\\t%s\\t%s\\n' % (doc_id, str1, str2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
