{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Phenotype extraction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import sys\n",
    "import cPickle\n",
    "\n",
    "# import snorkel and gwasdb\n",
    "sys.path.append('../snorkel')\n",
    "sys.path.append('../src')\n",
    "sys.path.append('../src/crawler')\n",
    "\n",
    "# set up paths\n",
    "abstract_dir = '../data/db/papers'\n",
    "\n",
    "# set up matplotlib\n",
    "import matplotlib\n",
    "%matplotlib inline\n",
    "matplotlib.rcParams['figure.figsize'] = (12,4)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extract phenotype candidates from papers"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load corpus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from snorkel.parser import XMLDocParser\n",
    "from extractor.parser import UnicodeXMLDocParser, GWASXMLAbstractParser\n",
    "\n",
    "xml_parser = GWASXMLAbstractParser(\n",
    "    path=abstract_dir,\n",
    "    doc='./*',\n",
    "    title='.//front//article-title//text()',\n",
    "    abstract='.//abstract//p//text()',\n",
    "    par1='.//body/p[1]//text()',\n",
    "    id='.//article-id[@pub-id-type=\"pmid\"]/text()',\n",
    "    keep_xml_tree=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 2.29 s, sys: 282 ms, total: 2.58 s\n",
      "Wall time: 14.2 s\n"
     ]
    }
   ],
   "source": [
    "from snorkel.parser import HTMLParser\n",
    "from snorkel.parser import SentenceParser\n",
    "from snorkel.parser import CorpusParser\n",
    "\n",
    "sent_parser = SentenceParser()\n",
    "html_parser = HTMLParser(path='../data/db/papers/')\n",
    "\n",
    "corpus_name = 'gwas-text-corpus.pkl'\n",
    "\n",
    "try:\n",
    "    with open(corpus_name,\"r\") as pkl:\n",
    "        corpus = cPickle.load(pkl)\n",
    "except:\n",
    "    cp = CorpusParser(xml_parser, sent_parser, max_docs=100)\n",
    "    %time corpus = cp.parse_corpus(name='GWAS Corpus')\n",
    "    # pickling currently doesn't work...\n",
    "#     with open(corpus_name,\"w\") as pkl:\n",
    "#         corpus = cPickle.dump(corpus, pkl)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extract candidates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 657,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 54.3 s, sys: 360 ms, total: 54.7 s\n",
      "Wall time: 54.6 s\n",
      "5650 candidates extracted\n"
     ]
    }
   ],
   "source": [
    "from snorkel.candidates import Ngrams\n",
    "from snorkel.matchers import DictionaryMatch, Union, RegexMatchSpan\n",
    "from snorkel.candidates import EntityExtractor\n",
    "from snorkel.utils import slice_into_ngrams\n",
    "\n",
    "from extractor.util import change_name\n",
    "from extractor.matcher import PhenotypeMatcher\n",
    "from db.kb import KnowledgeBase\n",
    "\n",
    "def make_ngrams(L, n_max=10, n_min=3, delim=' '):\n",
    "    for l in L:\n",
    "        yield l\n",
    "        tokens = l.strip().split(delim)\n",
    "        for ngram in slice_into_ngrams(tokens, n_max=n_max, n_min=n_min, delim=delim):\n",
    "            yield ngram\n",
    "\n",
    "# Define a candidate space\n",
    "ngrams = Ngrams(n_max=7)\n",
    "\n",
    "# collect phenotype list\n",
    "kb = KnowledgeBase()\n",
    "# efo phenotypes\n",
    "efo_phenotype_list0 = kb.get_phenotype_candidates(source='efo', peek=True) # TODO: remove peaking\n",
    "efo_phenotype_list = list(make_ngrams(efo_phenotype_list0))\n",
    "# mesh diseases\n",
    "mesh_phenotype_list0 = kb.get_phenotype_candidates(source='mesh')\n",
    "mesh_phenotype_list = list(make_ngrams(mesh_phenotype_list0))\n",
    "# mesh chemicals\n",
    "chem_phenotype_list = kb.get_phenotype_candidates(source='chemical')\n",
    "# regex matches\n",
    "rgx = u'[A-Za-z\\u2013-]+ (disease|trait|phenotype|outcome|response|quantitative trait|measurement|response)s?'\n",
    "\n",
    "# Define matchers\n",
    "efo_phen_matcher = PhenotypeMatcher(d=efo_phenotype_list, ignore_case=True, mod_fn=change_name)\n",
    "mesh_phen_matcher = PhenotypeMatcher(d=mesh_phenotype_list, ignore_case=True, mod_fn=change_name)\n",
    "chem_phen_matcher = DictionaryMatch(d=chem_phenotype_list, longest_match_only=True, ignore_case=True)\n",
    "regex_phen_matcher = RegexMatchSpan(rgx=rgx)\n",
    "phen_matcher = Union(efo_phen_matcher, mesh_phen_matcher, chem_phen_matcher, regex_phen_matcher)\n",
    "# phen_matcher = PhenotypeMatcher(d=phenotype_list, ignore_case=True, mod_fn=change_name)\n",
    "\n",
    "# Extractor\n",
    "phen_extractor = EntityExtractor(ngrams, phen_matcher)\n",
    "\n",
    "# collect candidates\n",
    "%time phen_c = phen_extractor.extract(corpus.get_sentences(), name='all')\n",
    "print len(phen_c), 'candidates extracted'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 658,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[u'cholesterol ester storage disease', u'familial hypercholesterolemia', u'low density lipoprotein cholesterol measurement', u'nhdl cholesterol', u'cholesterol ester', u'total cholesterol measurement', u'lecithin-cholesterol acyltransferase deficiency', u'hypercholesterolemia', u'cholesterol esters', u'xy disorder of sex development due to cholesterol synthesis defect', u'epicholesterol', u'7-dehydrocholesterol reductase deficiency', u'hdl cholesterol', u'cholesterol', u'cholesterol homeostasis', u'reduce cholesterol levels', u'alpha lipoprotein cholesterol', u'high density lipoprotein cholesterol measurement', u'homozygous familial hypercholesterolemia', u'cholesterol embolism', u'non-high density lipoprotein cholesterol measurement', u'nhdl cholesterol measurement', u'hypocholesterolemia', u'hypercholesterolemia due to cholesterol 7alpha-hydroxylase deficiency', u'cholesterol-ester transfer protein deficiency', u'non-high density lipoprotein cholesterol', u'hdl(2) cholesterol', u'ldl cholesterol', u'familial hypercholesterolaemia with hyperlipaemia', u'cholesterol inhibitor', u'alpha-lipoprotein cholesterol']\n"
     ]
    }
   ],
   "source": [
    "# phen_c0 = phen_c\n",
    "# phenotype_list0 = phenotype_list\n",
    "print [p for p in phenotype_list if 'cholesterol' in p]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 659,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n"
     ]
    }
   ],
   "source": [
    "print len(tmp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 660,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[u'', u'ccl5 measurement', u'cholera infantum', u'chemokine ccl2 level', u'chemokine ccl2 level', u'cigarettes per day measurement', u'cigarettes per day', u'cigarettes per day measurement', u'per day measurement', u'inflammatory marker measurement']\n",
      "[u'', u'ccl5 measurement', u'cholera infantum', u'chemokine ccl2 level', u'cigarettes per day measurement', u'inflammatory marker measurement', u'kidney diseases', u'white spot', u'response to sertraline', u'acylcarnitine measurement']\n",
      "parkinson'\n"
     ]
    }
   ],
   "source": [
    "len(phenotype_list0)\n",
    "phenotype_list = list(make_ngrams(phenotype_list0))\n",
    "len(phenotype_list)\n",
    "print phenotype_list[:10]\n",
    "print phenotype_list0[:10]\n",
    "print stemmer.stem(\"parkinson's\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 661,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "type i diabetes mellitus\n",
      "[u'type i diabetes mellitus', u'type i diabetes', u'type i diabetes mellitus', u'i diabetes mellitus', u'type i diabetes', u'type i diabetes', u'type i diabetes mellitus', u'type i diabetes', u'type i diabetes mellitus', u'i diabetes mellitus', u'i diabetes mellitus', u'i diabetes mellitus', u'diabetes mellitus type 02', u'diabetes mellitus type', u'diabetes mellitus type 02', u'mellitus type 02', u'diabetes mellitus type', u'diabetes mellitus type', u'diabetes mellitus type 02', u'diabetes mellitus type', u'diabetes mellitus type 02', u'mellitus type 02', u'dm - diabetes mellitus', u'dm - diabetes', u'dm - diabetes mellitus', u'- diabetes mellitus', u'dm - diabetes', u'dm - diabetes', u'dm - diabetes mellitus', u'dm - diabetes', u'dm - diabetes mellitus', u'- diabetes mellitus', u'- diabetes mellitus', u'- diabetes mellitus', u'type ii diabetes mellitus', u'type ii diabetes', u'type ii diabetes mellitus', u'ii diabetes mellitus', u'type ii diabetes', u'type ii diabetes', u'type ii diabetes mellitus', u'type ii diabetes', u'type ii diabetes mellitus', u'ii diabetes mellitus', u'ii diabetes mellitus', u'ii diabetes mellitus', u'type 1 diabetes nephropathy', u'type 1 diabetes', u'type 1 diabetes nephropathy', u'1 diabetes nephropathy', u'type 1 diabetes', u'type 1 diabetes', u'type 1 diabetes nephropathy', u'type 1 diabetes', u'type 1 diabetes nephropathy', u'1 diabetes nephropathy', u'1 diabetes nephropathy', u'1 diabetes nephropathy', u'gestational diabetes', u'type 2 diabetes nephropathy', u'type 2 diabetes', u'type 2 diabetes nephropathy', u'2 diabetes nephropathy', u'type 2 diabetes', u'type 2 diabetes', u'type 2 diabetes nephropathy', u'type 2 diabetes', u'type 2 diabetes nephropathy', u'2 diabetes nephropathy', u'2 diabetes nephropathy', u'2 diabetes nephropathy', u'diabetes mellitus type 2 associated cataract', u'diabetes mellitus type', u'diabetes mellitus type 2', u'diabetes mellitus type 2 associated', u'diabetes mellitus type 2 associated cataract', u'mellitus type 2', u'mellitus type 2 associated', u'mellitus type 2 associated cataract', u'type 2 associated', u'type 2 associated cataract', u'2 associated cataract', u'diabetes mellitus type', u'diabetes mellitus type', u'diabetes mellitus type 2', u'diabetes mellitus type', u'diabetes mellitus type 2', u'mellitus type 2', u'diabetes mellitus type 2 associated', u'diabetes mellitus type', u'diabetes mellitus type 2', u'diabetes mellitus type 2 associated', u'mellitus type 2', u'mellitus type 2 associated', u'type 2 associated', u'diabetes mellitus type 2 associated cataract', u'diabetes mellitus type', u'diabetes mellitus type 2', u'diabetes mellitus type 2 associated', u'diabetes mellitus type 2 associated cataract', u'mellitus type 2', u'mellitus type 2 associated', u'mellitus type 2 associated cataract', u'type 2 associated', u'type 2 associated cataract', u'2 associated cataract', u'diabetes mellitus', u'gestational diabetes mellitus', u'gestational diabetes mellitus', u'gestational diabetes mellitus', u'gestational diabetes mellitus', u'non-insulin dependent diabetes mellitus', u'non-insulin dependent diabetes', u'non-insulin dependent diabetes mellitus', u'dependent diabetes mellitus', u'non-insulin dependent diabetes', u'non-insulin dependent diabetes', u'non-insulin dependent diabetes mellitus', u'non-insulin dependent diabetes', u'non-insulin dependent diabetes mellitus', u'dependent diabetes mellitus', u'dependent diabetes mellitus', u'dependent diabetes mellitus', u'stable diabetes mellitus', u'stable diabetes mellitus', u'stable diabetes mellitus', u'stable diabetes mellitus', u'diabetes-nephrosis syndrome']\n"
     ]
    }
   ],
   "source": [
    "def make_ngrams(L, n_max=10, n_min=3, delim=' '):\n",
    "    for l in L:\n",
    "        yield l\n",
    "        tokens = l.strip().split(delim)\n",
    "        for ngram in slice_into_ngrams(tokens, n_max=n_max, n_min=n_min, delim=delim):\n",
    "            yield ngram\n",
    "            \n",
    "L=[p for p in phenotype_list if 'diabetes' in p]\n",
    "print L[0]\n",
    "print [n for n in make_ngrams(L)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We would like to remove nested candidates (manually, for now)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 662,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2562 candidates dropped, now we have 3088\n"
     ]
    }
   ],
   "source": [
    "# load existing candidates into a dict\n",
    "span_dict = { str(span.context) : list() for span in phen_c }\n",
    "for span in phen_c:\n",
    "    span_dict[str(span.context)].append( (span.char_start, span.char_end) )\n",
    "\n",
    "def nested(ivl1, ivl2):\n",
    "    if ivl1 != ivl2 and ivl2[0] <= ivl1[0] <= ivl1[1] <= ivl2[1]:\n",
    "        return True\n",
    "    else:\n",
    "        return False\n",
    "\n",
    "new_phen_c = list()\n",
    "for span in phen_c:\n",
    "    span_ivl = span.char_start, span.char_end\n",
    "    span_name = str(span.context)\n",
    "    if all([not nested(span_ivl, other_ivl) for other_ivl in span_dict[span_name]]):\n",
    "        new_phen_c.append(span)\n",
    "        \n",
    "print len(phen_c) - len(new_phen_c), 'candidates dropped, now we have', len(new_phen_c)\n",
    "phen_c = new_phen_c"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create gold-truth set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 663,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 322 gold mentions, e.g.:\n",
      "[('17658951', u'http://www.ebi.ac.uk/efo/EFO_0004764'), ('17903294', u'http://purl.obolibrary.org/obo/GO_0070527'), ('20038947', u'http://www.ebi.ac.uk/efo/EFO_0003761'), ('19197348', u'http://www.ebi.ac.uk/efo/EFO_0004748'), ('17658951', u'http://www.ebi.ac.uk/efo/EFO_0004626')]\n",
      "441 216\n"
     ]
    }
   ],
   "source": [
    "from db.kb import KnowledgeBase\n",
    "from nltk.stem import PorterStemmer\n",
    "from extractor.util import change_name\n",
    "\n",
    "kb = KnowledgeBase() # reload\n",
    "gold_set_phens = frozenset \\\n",
    "([ \n",
    "    (doc.name, phen.ontology_ref) for doc in corpus.documents \n",
    "                                  for phen in kb.phen_by_pmid(doc.name, source='efo')\n",
    "])\n",
    "\n",
    "# map phenotype names to their id\n",
    "phen2id = \\\n",
    "{\n",
    "    change_name(syn) : phen.ontology_ref for doc in corpus.documents\n",
    "                                                  for phen in kb.phen_by_pmid(doc.name, source='efo')\n",
    "                                                  for syn in [phen.name] + phen.synonyms.split('|')\n",
    "}\n",
    "\n",
    "# this is the more correct version (code below should be changed to use it)\n",
    "# we collect a set of EFOs matching a string; we use this only at the very end for now\n",
    "phen2idset = dict()\n",
    "for doc in corpus.documents:\n",
    "    for phen in kb.phen_by_pmid(doc.name, source='efo'):\n",
    "        synonyms = [phen.name] + phen.synonyms.split('|')\n",
    "        ngrams = make_ngrams(synonyms)\n",
    "        for ngram in ngrams:\n",
    "            syn_name = change_name(ngram)\n",
    "            if syn_name not in phen2idset: phen2idset[syn_name] = set()\n",
    "            phen2idset[syn_name].add(phen.id)\n",
    "\n",
    "id2phen = \\\n",
    "{\n",
    "    phen.ontology_ref : phen for doc in corpus.documents\n",
    "                             for phen in kb.phen_by_pmid(doc.name, source='efo')\n",
    "}\n",
    "\n",
    "print 'Found %d gold mentions, e.g.:' % len(gold_set_phens)\n",
    "print list(gold_set_phens)[:5]\n",
    "print len(phen2id), len(id2phen)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Statistics"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, we need to understand whether the extraction or the classification approach will be better."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Number of phenotypes per paper"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Number of EFO phenotypes per paper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 664,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "EFO phenotype numbers per paper: [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 3, 3, 3, 4, 4, 4, 4, 4, 5, 5, 5, 6, 6, 7, 7, 7, 10, 11, 13, 17, 37, 66]\n"
     ]
    }
   ],
   "source": [
    "docid2efo = \\\n",
    "{\n",
    "    doc.name : [ phen.name for phen in kb.phen_by_pmid(doc.name, source='efo') ] for doc in corpus.documents\n",
    "}\n",
    "print 'EFO phenotype numbers per paper:', sorted([len(v) for v in docid2efo.values()])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Number of GWAS catalog (i.e. aggregate) phenotypes per paper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 665,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GWAS catalog phenotype numbers per paper: [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 4, 4, 4, 4, 4, 4, 5, 5, 5, 5, 5, 5, 5, 5, 6, 6, 7, 7, 8, 8, 9, 9, 9, 9, 10, 11, 11, 11, 12, 12, 13, 13, 14, 14, 15, 16, 19, 20, 21, 21, 21, 22, 22, 24, 29, 33, 34, 37, 38, 40, 42, 49, 73]\n"
     ]
    }
   ],
   "source": [
    "docid2efo = \\\n",
    "{\n",
    "    doc.name : [ phen.name for phen in kb.phen_by_pmid(doc.name, source='gwas_catalog') ] for doc in corpus.documents\n",
    "}\n",
    "print 'GWAS catalog phenotype numbers per paper:', sorted([len(v) for d, v in docid2efo.items()])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Candidate recall statistics"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Over EFO phenotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 666,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Statistics over EFO phenotypes:\n",
      "# of gold annotations\t= 322\n",
      "# of candidates\t\t= 3088\n",
      "Candidate recall\t= 0.370\n",
      "Candidate precision\t= 0.039\n"
     ]
    }
   ],
   "source": [
    "from extractor.util import gold_phen_stats\n",
    "gold_phen_stats(phen_c, gold_set_phens, phen2id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Why is the recall low?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 667,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "203\n",
      "19798445\n",
      "blood metabolite measurement http://www.ebi.ac.uk/efo/EFO_0005664\n",
      "sphingolipid measurement http://www.ebi.ac.uk/efo/EFO_0004622\n",
      "blood metabolite measurement http://www.ebi.ac.uk/efo/EFO_0005664\n",
      "\n",
      "Genetic determinants of circulating sphingolipid concentrations in European populations.\n",
      "Genetic Determinants of Circulating Sphingolipid Concentrations in European Populations. Sphingolipids have essential roles as structural components of cell membranes and in cell signalling, and disruption of their metabolism causes several diseases, with diverse neurological, psychiatric, and metabolic consequences. Increasingly, variants within a few of the genes that encode enzymes involved in sphingolipid metabolism are being associated with complex disease phenotypes. Direct experimental evidence supports a role of specific sphingolipid species in several common complex chronic disease processes including atherosclerotic plaque formation, myocardial infarction (MI), cardiomyopathy, pancreatic β-cell failure, insulin resistance, and type 2 diabetes mellitus. Therefore, sphingolipids represent novel and important intermediate phenotypes for genetic analysis, yet little is known about the major genetic variants that influence their circulating levels in the general population. We performed a genome-wide association study (GWAS) between 318,237 single-nucleotide polymorphisms (SNPs) and levels of circulating sphingomyelin (SM), dihydrosphingomyelin (Dih-SM), ceramide (Cer), and glucosylceramide (GluCer) single lipid species (33 traits); and 43 matched metabolite ratios measured in 4,400 subjects from five diverse European populations. Associated variants (32) in five genomic regions were identified with genome-wide significant corrected  p -values ranging down to 9.08×10 −66 . The strongest associations were observed in or near 7 genes functionally involved in ceramide biosynthesis and trafficking:  SPTLC3 ,  LASS4 ,  SGPP1 ,  ATP10D , and  FADS1–3 . Variants in 3 loci ( ATP10D ,  FADS3 , and  SPTLC3 ) associate with MI in a series of three German MI studies. An additional 70 variants across 23 candidate genes involved in sphingolipid-metabolizing pathways also demonstrate association ( p  = 10 −4  or less). Circulating concentrations of several key components in sphingolipid metabolism are thus under strong genetic control, and variants in these loci can be tested for a role in the development of common cardiovascular, metabolic, neurological, and psychiatric diseases. Although several rare monogenic diseases are caused by defects in enzymes involved in sphingolipid biosynthesis and metabolism, little is known about the major variants that control the circulating levels of these important bioactive molecules. As well as being essential components of plasma membranes and endosomes, sphingolipids play critical roles in cell surface protection, protein and lipid transport and sorting, and cellular signalling cascades. Experimental evidence supports a role for sphingolipids in several common complex chronic metabolic, cardiovascular, or neurological disease processes. Therefore, sphingolipids represent novel and important intermediate phenotypes for genetic analysis, and discovering the genetic variants that influence their circulating concentrations is an important step towards understanding how the genetic control of sphingolipids might contribute to common human disease. We have identified 32 variants in 7 genes that have a strong effect on the circulating plasma levels of 33 distinct sphingolipids, and 43 matched metabolite ratios. In a series of 3 German MI studies, we see association with MI for variants in 3 of the genes tested. Further cardiovascular, metabolic, neurological, and psychiatric disease associations can be tested with the variants described here, which may identify additional disease risk and potentially useful therapeutic targets.\n",
      "\n",
      "17903300\n",
      "waist circumference http://www.ebi.ac.uk/efo/EFO_0004342\n",
      "body mass index http://www.ebi.ac.uk/efo/EFO_0004340\n",
      "longitudinal bmi measurement http://www.ebi.ac.uk/efo/EFO_0005937\n",
      "longitudinal bmi measurement http://www.ebi.ac.uk/efo/EFO_0005937\n",
      "body mass index change measurement\n",
      "Genome-wide association to body mass index and waist circumference: the Framingham Heart Study 100K project.\n",
      "Genome-wide association to body mass index and waist circumference: the Framingham Heart Study 100K project. Obesity is related to multiple cardiovascular disease (CVD) risk factors as well as CVD and has a strong familial component. We tested for association between SNPs on the Affymetrix 100K SNP GeneChip and measures of adiposity in the Framingham Heart Study. A total of 1341 Framingham Heart Study participants in 310 families genotyped with the Affymetrix 100K SNP GeneChip had adiposity traits measured over 30 years of follow up. Body mass index (BMI), waist circumference (WC), weight change, height, and radiographic measures of adiposity (subcutaneous adipose tissue, visceral adipose tissue, waist circumference, sagittal height) were measured at multiple examination cycles. Multivariable-adjusted residuals, adjusting for age, age-squared, sex, smoking, and menopausal status, were evaluated in association with the genotype data using additive Generalized Estimating Equations (GEE) and Family Based Association Test (FBAT) models. We prioritized mean BMI over offspring examinations (1–7) and cohort examinations (10, 16, 18, 20, 22, 24, 26) and mean WC over offspring examinations (4–7) for presentation. We evaluated associations with 70,987 SNPs on autosomes with minor allele frequencies of at least 0.10, Hardy-Weinberg equilibrium p ≥ 0.001, and call rates of at least 80%. The top SNPs to be associated with mean BMI and mean WC by GEE were rs110683 (p-value 1.22*10 -7 ) and rs4471028 (p-values 1.96*10 -7 ). Please see   for the complete set of results. We were able to validate SNPs in known genes that have been related to BMI or other adiposity traits, including the  ESR1  Xba1 SNP,  PPARG , and  ADIPOQ . Adiposity traits are associated with SNPs on the Affymetrix 100K SNP GeneChip. Replication of these initial findings is necessary. These data will serve as a resource for replication as more genes become identified with BMI and WC.\n",
      "\n",
      "19056611\n",
      "type ii diabetes mellitus http://www.ebi.ac.uk/efo/EFO_0001360\n",
      "type ii diabetes mellitus http://www.ebi.ac.uk/efo/EFO_0001360\n",
      "diabetes mellitus type 02|non-insulin dependent diabetes mellitus|stable diabetes mellitus|diabetes mellitus\n",
      "Adiposity-related heterogeneity in patterns of type 2 diabetes susceptibility observed in genome-wide association data.\n",
      "Adiposity-Related Heterogeneity in Patterns of Type 2 Diabetes Susceptibility Observed in Genome-Wide Association Data. OBJECTIVE— This study examined how differences in the BMI distribution of type 2 diabetic case subjects affected genome-wide patterns of type 2 diabetes association and considered the implications for the etiological heterogeneity of type 2 diabetes. RESEARCH DESIGN AND METHODS— We reanalyzed data from the Wellcome Trust Case Control Consortium genome-wide association scan (1,924 case subjects, 2,938 control subjects: 393,453 single-nucleotide polymorphisms [SNPs]) after stratifying case subjects (into “obese” and “nonobese”) according to median BMI (30.2 kg/m 2 ). Replication of signals in which alternative case-ascertainment strategies generated marked effect size heterogeneity in type 2 diabetes association signal was sought in additional samples. RESULTS— In the “obese-type 2 diabetes” scan,  FTO  variants had the strongest type 2 diabetes effect (rs8050136: relative risk [RR] 1.49 [95% CI 1.34–1.66],  P  = 1.3 × 10 −13 ), with only weak evidence for  TCF7L2  (rs7901695 RR 1.21 [1.09–1.35],  P  = 0.001). This situation was reversed in the “nonobese” scan, with  FTO  association undetectable (RR 1.07 [0.97–1.19],  P  = 0.19) and  TCF7L2  predominant (RR 1.53 [1.37–1.71],  P  = 1.3 × 10 −14 ). These patterns, confirmed by replication, generated strong combined evidence for between-stratum effect size heterogeneity ( FTO :  P DIFF  = 1.4 × 10 −7 ;  TCF7L2 :  P DIFF  = 4.0 × 10 −6 ). Other signals displaying evidence of effect size heterogeneity in the genome-wide analyses (on chromosomes 3, 12, 15, and 18) did not replicate. Analysis of the current list of type 2 diabetes susceptibility variants revealed nominal evidence for effect size heterogeneity for the  SLC30A8  locus alone (RR obese  1.08 [1.01–1.15]; RR nonobese  1.18 [1.10–1.27]:  P DIFF  = 0.04). CONCLUSIONS— This study demonstrates the impact of differences in case ascertainment on the power to detect and replicate genetic associations in genome-wide association studies. These data reinforce the notion that there is substantial etiological heterogeneity within type 2 diabetes.\n",
      "\n",
      "17903296\n",
      "bone density http://www.ebi.ac.uk/efo/EFO_0003923\n",
      "hip geometry http://www.ebi.ac.uk/efo/EFO_0004685\n",
      "bone density http://www.ebi.ac.uk/efo/EFO_0003923\n",
      "density\n",
      "Genome-wide association with bone mass and geometry in the Framingham Heart Study.\n",
      "Genome-wide association with bone mass and geometry in the Framingham Heart Study. Osteoporosis is characterized by low bone mass and compromised bone structure, heritable traits that contribute to fracture risk. There have been no genome-wide association and linkage studies for these traits using high-density genotyping platforms. We used the Affymetrix 100K SNP GeneChip marker set in the Framingham Heart Study (FHS) to examine genetic associations with ten primary quantitative traits: bone mineral density (BMD), calcaneal ultrasound, and geometric indices of the hip. To test associations with multivariable-adjusted residual trait values, we used additive generalized estimating equation (GEE) and family-based association tests (FBAT) models within each sex as well as sexes combined. We evaluated 70,987 autosomal SNPs with genotypic call rates ≥80%, HWE p ≥ 0.001, and MAF ≥10% in up to 1141 phenotyped individuals (495 men and 646 women, mean age 62.5 yrs). Variance component linkage analysis was performed using 11,200 markers. Heritability estimates for all bone phenotypes were 30–66%. LOD scores ≥3.0 were found on chromosomes 15 (1.5 LOD confidence interval: 51,336,679–58,934,236 bp) and 22 (35,890,398–48,603,847 bp) for femoral shaft section modulus. The ten primary phenotypes had 12 associations with 100K SNPs in GEE models at p < 0.000001 and 2 associations in FBAT models at p < 0.000001. The 25 most significant p-values for GEE and FBAT were all less than 3.5 × 10 -6  and 2.5 × 10 -5 , respectively. Of the 40 top SNPs with the greatest numbers of significantly associated BMD traits (including femoral neck, trochanter, and lumbar spine), one half to two-thirds were in or near genes that have  not  previously been studied for osteoporosis. Notably, pleiotropic associations between BMD and bone geometric traits were uncommon. Evidence for association (FBAT or GEE p < 0.05) was observed for several SNPs in candidate genes for osteoporosis, such as rs1801133 in  MTHFR ; rs1884052 and rs3778099 in  ESR1 ; rs4988300 in  LRP5 ; rs2189480 in  VDR ; rs2075555 in  COLIA1 ; rs10519297 and rs2008691 in  CYP19 , as well as SNPs in  PPARG  (rs10510418 and rs2938392) and  ANKH  (rs2454873 and rs379016). All GEE, FBAT and linkage results are provided as an open-access results resource at  . The FHS 100K SNP project offers an unbiased genome-wide strategy to identify new candidate loci and to replicate previously suggested candidate genes for osteoporosis.\n",
      "\n",
      "17903305\n",
      "prostate carcinoma http://www.ebi.ac.uk/efo/EFO_0001663\n",
      "breast carcinoma http://www.ebi.ac.uk/efo/EFO_0000305\n",
      "breast carcinoma http://www.ebi.ac.uk/efo/EFO_0000305\n",
      "breast neoplasms|neoplasm\n",
      "A genome-wide association study of breast and prostate cancer in the NHLBI's Framingham Heart Study.\n",
      "A genome-wide association study of breast and prostate cancer in the NHLBI's Framingham Heart Study. Breast and prostate cancer are two commonly diagnosed cancers in the United States. Prior work suggests that cancer causing genes and cancer susceptibility genes can be identified. We conducted a genome-wide association study (Affymetrix 100K SNP GeneChip) of cancer in the community-based Framingham Heart Study. We report on 2 cancer traits – prostate cancer and breast cancer – in up to 1335 participants from 330 families (54% women, mean entry age 33 years). Multivariable-adjusted residuals, computed using Cox proportional hazards models, were tested for association with qualifying SNPs (70, 987 autosomal SNPs with genotypic call rate ≥80%, minor allele frequency ≥10%, Hardy-Weinberg test p ≥ 0.001) using generalized estimating equations (GEE) models and family based association tests (FBAT). There were 58 women with breast cancer and 59 men with prostate cancer. No SNP associations attained genome-wide significance. The top SNP associations in GEE models for each trait were as follows: breast cancer, rs2075555, p = 8.0 × 10 -8  in  COL1A1 ; and prostate cancer, rs9311171, p = 1.75 × 10 -6  in  CTDSPL . In analysis of selected candidate cancer susceptibility genes, two  MSR1  SNPs (rs9325782, GEE p = 0.008 and rs2410373, FBAT p = 0.021) were associated with prostate cancer and three  ERBB4  SNPs (rs905883 GEE p = 0.0002, rs7564590 GEE p = 0.003, rs7558615 GEE p = 0.0078) were associated with breast cancer. The previously reported risk SNP for prostate cancer, rs1447295, was not included on the 100K chip. Results of cancer phenotype-genotype associations for all autosomal SNPs are web posted at  . Although no association attained genome-wide significance, several interesting associations emerged for breast and prostate cancer. These findings can serve as a resource for replication in other populations to identify novel biologic pathways contributing to cancer susceptibility.\n",
      "\n",
      "19570815\n",
      "body height http://www.ebi.ac.uk/efo/EFO_0004339\n",
      "anthropometry http://www.ebi.ac.uk/efo/EFO_0004302\n",
      "anthropometry http://www.ebi.ac.uk/efo/EFO_0004302\n",
      "anthropometric traits\n",
      "A genome-wide association study of northwestern Europeans involves the C-type natriuretic peptide signaling pathway in the etiology of human height variation.\n",
      "A genome-wide association study of northwestern Europeans involves the C-type natriuretic peptide signaling pathway in the etiology of human height variation. Northwestern Europeans are among the tallest of human populations. The increase in body height in these people appears to have reached a plateau, suggesting the ubiquitous presence of an optimal environment in which genetic factors may have exerted a particularly strong influence on human growth. Therefore, we performed a genome-wide association study (GWAS) of body height using 2.2 million markers in 10 074 individuals from three Dutch and one German population-based cohorts. Upon genotyping, the 12 most significantly height-associated single nucleotide polymorphisms (SNPs) from this GWAS in 6912 additional individuals of Dutch and Swedish origin, a genetic variant (rs6717918) on chromosome 2q37.1 was found to be associated with height at a genome-wide significance level ( P combined  = 3.4 × 10 −9 ). Notably, a second SNP (rs6718438) located ∼450 bp away and in strong LD ( r 2  = 0.77) with rs6717918 was previously found to be suggestive of a height association in 29 820 individuals of mainly northwestern European ancestry, and the over-expression of a nearby natriuretic peptide precursor type C ( NPPC ) gene, has been associated with overgrowth and skeletal anomalies. We also found a SNP (rs10472828) located on 5p14 near the natriuretic peptide receptor 3 ( NPR3 ) gene, encoding a receptor of the NPPC ligand, to be associated with body height ( P combined  = 2.1 × 10 −7 ). Taken together, these results suggest that variation in the C-type natriuretic peptide signaling pathway, involving the  NPPC  and  NPR3  genes, plays an important role in determining human body height.\n",
      "\n",
      "19966805\n",
      "type i diabetes mellitus http://www.ebi.ac.uk/efo/EFO_0001359\n",
      "type i diabetes mellitus http://www.ebi.ac.uk/efo/EFO_0001359\n",
      "iddm|diabetes mellitus\n",
      "The imprinted DLK1-MEG3 gene region on chromosome 14q32.2 alters susceptibility to type 1 diabetes.\n",
      "The imprinted DLK1-MEG3 gene region on chromosome 14q32.2 alters susceptibility to type 1 diabetes. Genomewide association studies to map common disease susceptibility loci have been hugely successful with over 300 reproducibly associated loci reported to date, 1  but, perhaps surprisingly, have not yet provided convincing evidence for any susceptibility locus subject to parent of origin effects. We used imputation to extend existing genomewide association datasets 2 ,  3 ,  4  and here report robust evidence, at rs941576, for paternally inherited risk of type 1 diabetes (T1D, ratio of allelic effects for paternal vs maternal transmissions = 0.75, 95%CI=0.71–0.79), in the imprinted region of chromosome 14q32.2, which contains a functional candidate gene,  DLK1 . Our meta-analysis also provided support at genomewide significance for a T1D locus at chromosome 19p13.2, with the highest association at marker rs2304256 (OR=0.86, 95%CI=0.82–0.90) in the  TYK2  gene, which has previously associated with systemic lupus erythematosus. 5\n",
      "\n",
      "19081515\n",
      "creutzfeldt jacob disease http://www.ebi.ac.uk/efo/EFO_0004226\n",
      "creutzfeldt jacob disease http://www.ebi.ac.uk/efo/EFO_0004226\n",
      "spongiform encephalopathy\n",
      "Genetic risk factors for variant Creutzfeldt-Jakob disease: a genome-wide association study.\n",
      "Genetic risk factors for variant Creutzfeldt–Jakob disease: a genome-wide association study. Human and animal prion diseases are under genetic control, but apart from  PRNP  (the gene that encodes the prion protein), we understand little about human susceptibility to bovine spongiform encephalopathy (BSE) prions, the causal agent of variant Creutzfeldt–Jakob disease (vCJD). We did a genome-wide association study of the risk of vCJD and tested for replication of our findings in samples from many categories of human prion disease (929 samples) and control samples from the UK and Papua New Guinea (4254 samples), including controls in the UK who were genotyped by the Wellcome Trust Case Control Consortium. We also did follow-up analyses of the genetic control of the clinical phenotype of prion disease and analysed candidate gene expression in a mouse cellular model of prion infection. The  PRNP  locus was strongly associated with risk across several markers and all categories of prion disease (best single SNP [single nucleotide polymorphism] association in vCJD p=2·5×10 −17 ; best haplotypic association in vCJD p=1×10 −24 ). Although the main contribution to disease risk was conferred by  PRNP  polymorphic codon 129, another nearby SNP conferred increased risk of vCJD. In addition to  PRNP , one technically validated SNP association upstream of  RARB  (the gene that encodes retinoic acid receptor beta) had nominal genome-wide significance (p=1·9×10 −7 ). A similar association was found in a small sample of patients with iatrogenic CJD (p=0·030) but not in patients with sporadic CJD (sCJD) or kuru. In cultured cells, retinoic acid regulates the expression of the prion protein. We found an association with acquired prion disease, including vCJD (p=5·6×10 −5 ), kuru incubation time (p=0·017), and resistance to kuru (p=2·5×10 −4 ), in a region upstream of  STMN2  (the gene that encodes SCG10). The risk genotype was not associated with sCJD but conferred an earlier age of onset. Furthermore, expression of  Stmn2  was reduced 30-fold post-infection in a mouse cellular model of prion disease. The polymorphic codon 129 of  PRNP  was the main genetic risk factor for vCJD; however, additional candidate loci have been identified, which justifies functional analyses of these biological pathways in prion disease. The UK Medical Research Council.\n",
      "\n",
      "19343178\n",
      "body height http://www.ebi.ac.uk/efo/EFO_0004339\n",
      "anthropometry http://www.ebi.ac.uk/efo/EFO_0004302\n",
      "anthropometry http://www.ebi.ac.uk/efo/EFO_0004302\n",
      "anthropometric traits\n",
      "Meta-analysis of genome-wide scans for human adult stature identifies novel Loci and associations with measures of skeletal frame size.\n",
      "Meta-Analysis of Genome-Wide Scans for Human Adult Stature Identifies Novel Loci and Associations with Measures of Skeletal Frame Size. Recent genome-wide (GW) scans have identified several independent loci affecting human stature, but their contribution through the different skeletal components of height is still poorly understood. We carried out a genome-wide scan in 12,611 participants, followed by replication in an additional 7,187 individuals, and identified 17 genomic regions with GW-significant association with height. Of these, two are entirely novel (rs11809207 in  CATSPER4 , combined  P -value = 6.1×10 −8  and rs910316 in  TMED10 ,  P -value = 1.4×10 −7 ) and two had previously been described with weak statistical support (rs10472828 in  NPR3 ,  P -value = 3×10 −7  and rs849141 in  JAZF1 ,  P -value = 3.2×10 −11 ). One locus (rs1182188 at  GNA12 ) identifies the first height eQTL. We also assessed the contribution of height loci to the upper- (trunk) and lower-body (hip axis and femur) skeletal components of height. We find evidence for several loci associated with trunk length (including rs6570507 in  GPR126 ,  P -value = 4×10 −5  and rs6817306 in  LCORL ,  P -value = 4×10 −4 ), hip axis length (including rs6830062 at  LCORL ,  P -value = 4.8×10 −4  and rs4911494 at  UQCC ,  P -value = 1.9×10 −4 ), and femur length (including rs710841 at  PRKG2 ,  P -value = 2.4×10 −5  and rs10946808 at  HIST1H1D ,  P -value = 6.4×10 −6 ). Finally, we used conditional analyses to explore a possible differential contribution of the height loci to these different skeletal size measurements. In addition to validating four novel loci controlling adult stature, our study represents the first effort to assess the contribution of genetic loci to three skeletal components of height. Further statistical tests in larger numbers of individuals will be required to verify if the height loci affect height preferentially through these subcomponents of height. The first genetic association studies of adult height have confirmed a role of many common variants in influencing human height, but to date, the genetic basis of differences between different skeletal components of height have not been addressed. Here, we take advantage of recent technical and methodological advances to examine the role of common genetic variants on both height and skeletal components of height. By examining nearly 20,000 individuals from the UK and the Netherlands, we provide statistically significant evidence that 17 genomic regions are associated with height, including four novel regions. We also examine, for the first time, the association of these 17 regions with skeletal size measurements of spine, femur, and hip axis length, a measurement of hip geometry known to influence the risk of osteoporotic fractures. We find that some height loci are also associated with these skeletal components, although further statistical tests will be required to verify if these genetic variants act differentially on the individual skeletal measurements. The knowledge generated by this and other studies will not only inform the genetics of human quantitative variation, but will also lead to the potential discovery of many medically important polymorphisms.\n",
      "\n",
      "19300499\n",
      "response to anticoagulant http://purl.obolibrary.org/obo/GO_0061476\n",
      "response to anticoagulant http://purl.obolibrary.org/obo/GO_0061476\n",
      "\n",
      "A genome-wide association study confirms VKORC1, CYP2C9, and CYP4F2 as principal genetic determinants of warfarin dose.\n",
      "A Genome-Wide Association Study Confirms  VKORC1 ,  CYP2C9 , and  CYP4F2  as Principal Genetic Determinants of Warfarin Dose. We report the first genome-wide association study (GWAS) whose sample size (1,053 Swedish subjects) is sufficiently powered to detect genome-wide significance ( p <1.5×10 −7 ) for polymorphisms that modestly alter therapeutic warfarin dose. The anticoagulant drug warfarin is widely prescribed for reducing the risk of stroke, thrombosis, pulmonary embolism, and coronary malfunction. However, Caucasians vary widely (20-fold) in the dose needed for therapeutic anticoagulation, and hence prescribed doses may be too low (risking serious illness) or too high (risking severe bleeding). Prior work established that ∼30% of the dose variance is explained by single nucleotide polymorphisms (SNPs) in the warfarin drug target  VKORC1  and another ∼12% by two non-synonymous SNPs ( *2 ,  *3 ) in the cytochrome P450 warfarin-metabolizing gene  CYP2C9 . We initially tested each of 325,997 GWAS SNPs for association with warfarin dose by univariate regression and found the strongest statistical signals ( p <10 −78 ) at SNPs clustering near  VKORC1  and the second lowest p-values ( p <10 −31 ) emanating from  CYP2C9 . No other SNPs approached genome-wide significance. To enhance detection of weaker effects, we conducted multiple regression adjusting for known influences on warfarin dose ( VKORC1 ,  CYP2C9 , age, gender) and identified a single SNP (rs2108622) with genome-wide significance ( p  = 8.3×10 −10 ) that alters protein coding of the  CYP4F2  gene. We confirmed this result in 588 additional Swedish patients ( p <0.0029) and, during our investigation, a second group provided independent confirmation from a scan of warfarin-metabolizing genes. We also thoroughly investigated copy number variations, haplotypes, and imputed SNPs, but found no additional highly significant warfarin associations. We present power analysis of our GWAS that is generalizable to other studies, and conclude we had 80% power to detect genome-wide significance for common causative variants or markers explaining at least 1.5% of dose variance. These GWAS results provide further impetus for conducting large-scale trials assessing patient benefit from genotype-based forecasting of warfarin dose. Recently, geneticists have begun assaying hundreds of thousands of genetic markers covering the entire human genome to systematically search for and identify genes that cause disease. We have extended this “genome-wide association study” (GWAS) method by assaying ∼326,000 markers in 1,053 Swedish patients in order to identify genes that alter response to the anticoagulant drug warfarin. Warfarin is widely prescribed to reduce blood clotting in order to protect high-risk patients from stroke, thrombosis, and heart attack. But patients vary widely (20-fold) in the warfarin dose needed for proper blood thinning, which means that initial doses in some patients are too high (risking severe bleeding) or too low (risking serious illness). Our GWAS detected two genes ( VKORC1 ,  CYP2C9 ) already known to cause ∼40% of the variability in warfarin dose and discovered a new gene ( CYP4F2 ) contributing 1%–2% of the variability. Since our GWAS searched the entire genome, additional genes having a major influence on warfarin dose might not exist or be found in the near-term. Hence, clinical trials assessing patient benefit from individualized dose forecasting based on a patient's genetic makeup at  VKORC1 ,  CYP2C9  and possibly  CYP4F2  could provide state-of-the-art clinical benchmarks for warfarin use during the foreseeable future.\n",
      "\n",
      "19801982\n",
      "bone density http://www.ebi.ac.uk/efo/EFO_0003923\n",
      "bone density http://www.ebi.ac.uk/efo/EFO_0003923\n",
      "density\n",
      "Twenty bone-mineral-density loci identified by large-scale meta-analysis of genome-wide association studies.\n",
      "Twenty bone mineral density loci identified by large-scale meta-analysis of genome-wide association studies. Bone mineral density (BMD) is a heritable complex trait used in the clinical diagnosis of osteoporosis and the assessment of fracture risk. We performed meta-analysis of five genome-wide association studies of femoral neck and lumbar spine BMD in 19,195 subjects of Northern European descent. We identified 20 loci reaching genome-wide significance (GWS; P<5×10 −8 ), of which 13 map to new regions including 1p31.3 ( GPR177 ), 2p21 ( SPTBN1 ), 3p22 ( CTNNB1 ), 4q21.1 ( MEPE ), 5q14 ( MEF2C ), 7p14 ( STARD3NL ), 7q21.3 ( FLJ42280 ), 11p11.2 ( LRP4 ;  ARHGAP1 ;  F2 ), 11p14.1 ( DCDC5 ), 11p15 ( SOX6 ), 16q24 ( FOXL1 ), 17q21 ( HDAC5 ) and 17q12 ( CRHR1 ). The metaanalysis also confirmed at GWS level, seven known BMD loci on 1p36 ( ZBTB40 ), 6q25 ( ESR1 ), 8q24 ( TNFRSF11B ), 11q13.4 ( LRP5 ), 12q13 ( SP7 ), 13q14 ( TNFSF11 ), and 18q21 ( TNFRSF11A ). The numerous SNPs associated with BMD map to genes in signaling pathways with relevance to bone metabolism, and highlight the complex genetic architecture underlying osteoporosis and BMD variation.\n",
      "\n",
      "17903297\n",
      "neuropsychological test http://www.ebi.ac.uk/efo/EFO_0003926\n",
      "volumetric brain mri http://www.ebi.ac.uk/efo/EFO_0004868\n",
      "neuropsychological test http://www.ebi.ac.uk/efo/EFO_0003926\n",
      "test\n",
      "Genetic correlates of brain aging on MRI and cognitive test measures: a genome-wide association and linkage analysis in the Framingham Study.\n",
      "Genetic correlates of brain aging on MRI and cognitive test measures: a genome-wide association and linkage analysis in the Framingham study. Brain magnetic resonance imaging (MRI) and cognitive tests can identify heritable endophenotypes associated with an increased risk of developing stroke, dementia and Alzheimer's disease (AD). We conducted a genome-wide association (GWA) and linkage analysis exploring the genetic basis of these endophenotypes in a community-based sample. A total of 705 stroke- and dementia-free Framingham participants (age 62 +9 yrs, 50% male) who underwent volumetric brain MRI and cognitive testing (1999–2002) were genotyped. We used linear models adjusting for first degree relationships via generalized estimating equations (GEE) and family based association tests (FBAT) in additive models to relate qualifying single nucleotide polymorphisms (SNPs, 70,987 autosomal on Affymetrix 100K Human Gene Chip with minor allele frequency ≥ 0.10, genotypic call rate ≥ 0.80, and Hardy-Weinberg equilibrium p-value ≥ 0.001) to multivariable-adjusted residuals of 9 MRI measures including total cerebral brain (TCBV), lobar, ventricular and white matter hyperintensity (WMH) volumes, and 6 cognitive factors/tests assessing verbal and visuospatial memory, visual scanning and motor speed, reading, abstract reasoning and naming. We determined multipoint identity-by-descent utilizing 10,592 informative SNPs and 613 short tandem repeats and used variance component analyses to compute LOD scores. The strongest gene-phenotype association in FBAT analyses was between  SORL1  (rs1131497; p = 3.2 × 10 -6 ) and abstract reasoning, and in GEE analyses between  CDH4  (rs1970546; p = 3.7 × 10 -8 ) and TCBV. SORL1  plays a role in amyloid precursor protein processing and has been associated with the risk of AD. Among the 50 strongest associations (25 each by GEE and FBAT) were other biologically interesting genes. Polymorphisms within 28 of 163 candidate genes for stroke, AD and memory impairment were associated with the endophenotypes studied at p < 0.001. We confirmed our previously reported linkage of WMH on chromosome 4 and describe linkage of reading performance to a marker on chromosome 18 (GATA11A06), previously linked to dyslexia (LOD scores = 2.2 and 5.1). Our results suggest that genes associated with clinical neurological disease also have detectable effects on subclinical phenotypes. These hypothesis generating data illustrate the use of an unbiased approach to discover novel pathways that may be involved in brain aging, and could be used to replicate observations made in other studies.\n",
      "\n",
      "17903296\n",
      "bone density http://www.ebi.ac.uk/efo/EFO_0003923\n",
      "hip geometry http://www.ebi.ac.uk/efo/EFO_0004685\n",
      "hip geometry http://www.ebi.ac.uk/efo/EFO_0004685\n",
      "\n",
      "Genome-wide association with bone mass and geometry in the Framingham Heart Study.\n",
      "Genome-wide association with bone mass and geometry in the Framingham Heart Study. Osteoporosis is characterized by low bone mass and compromised bone structure, heritable traits that contribute to fracture risk. There have been no genome-wide association and linkage studies for these traits using high-density genotyping platforms. We used the Affymetrix 100K SNP GeneChip marker set in the Framingham Heart Study (FHS) to examine genetic associations with ten primary quantitative traits: bone mineral density (BMD), calcaneal ultrasound, and geometric indices of the hip. To test associations with multivariable-adjusted residual trait values, we used additive generalized estimating equation (GEE) and family-based association tests (FBAT) models within each sex as well as sexes combined. We evaluated 70,987 autosomal SNPs with genotypic call rates ≥80%, HWE p ≥ 0.001, and MAF ≥10% in up to 1141 phenotyped individuals (495 men and 646 women, mean age 62.5 yrs). Variance component linkage analysis was performed using 11,200 markers. Heritability estimates for all bone phenotypes were 30–66%. LOD scores ≥3.0 were found on chromosomes 15 (1.5 LOD confidence interval: 51,336,679–58,934,236 bp) and 22 (35,890,398–48,603,847 bp) for femoral shaft section modulus. The ten primary phenotypes had 12 associations with 100K SNPs in GEE models at p < 0.000001 and 2 associations in FBAT models at p < 0.000001. The 25 most significant p-values for GEE and FBAT were all less than 3.5 × 10 -6  and 2.5 × 10 -5 , respectively. Of the 40 top SNPs with the greatest numbers of significantly associated BMD traits (including femoral neck, trochanter, and lumbar spine), one half to two-thirds were in or near genes that have  not  previously been studied for osteoporosis. Notably, pleiotropic associations between BMD and bone geometric traits were uncommon. Evidence for association (FBAT or GEE p < 0.05) was observed for several SNPs in candidate genes for osteoporosis, such as rs1801133 in  MTHFR ; rs1884052 and rs3778099 in  ESR1 ; rs4988300 in  LRP5 ; rs2189480 in  VDR ; rs2075555 in  COLIA1 ; rs10519297 and rs2008691 in  CYP19 , as well as SNPs in  PPARG  (rs10510418 and rs2938392) and  ANKH  (rs2454873 and rs379016). All GEE, FBAT and linkage results are provided as an open-access results resource at  . The FHS 100K SNP project offers an unbiased genome-wide strategy to identify new candidate loci and to replicate previously suggested candidate genes for osteoporosis.\n",
      "\n",
      "20038947\n",
      "unipolar depression http://www.ebi.ac.uk/efo/EFO_0003761\n",
      "unipolar depression http://www.ebi.ac.uk/efo/EFO_0003761\n",
      "melancholia\n",
      "Novel loci for major depression identified by genome-wide association study of Sequenced Treatment Alternatives to Relieve Depression and meta-analysis of three studies.\n",
      "Novel loci for major depression identified by genome-wide association study of STAR*D and meta-analysis of three studies. We report a genome-wide association study (GWAS) of major depressive disorder (MDD) in 1,221 cases from the Sequenced Treatment Alternatives to Relieve Depression (STAR*D) study and 1,636 screened controls. No genome-wide evidence for association was detected. We also carried out a meta-analysis of three European-ancestry MDD GWAS datasets: STAR*D, Genetics of Recurrent Early-Onset Depression (GenRED) and the publicly-available Genetic Association Information Network MDD dataset (GAIN-MDD). These datasets, totaling 3,957 cases and 3,428 controls, were genotyped using four different platforms (Affymetrix 6.0, 5.0 and 500K, and Perlegen). For each of 2.4 million HapMap II SNPs, using genotyped data where available and imputed data otherwise, single-SNP association tests were carried out in each sample with correction for ancestry-informative principal components. The strongest evidence for association in the meta-analysis was observed for intronic SNPs in  ATP6V1B2  (P = 6.78 × 10 −7 ),  SP4  ( P  = 7.68 × 10 −7 ) and  GRM7  ( P  = 1.11 × 10 −6 ). Additional exploratory analyses were carried out for a narrower phenotype (recurrent MDD with onset before age 31, N = 2,191 cases), and separately for males and females. Several of the best findings were supported primarily by evidence from narrow cases or from either males or females. Based on previous biological evidence, we consider  GRM7  a strong MDD candidate gene. Larger samples will be required to determine whether any common SNPs are significantly associated with MDD.\n",
      "\n",
      "18262040\n",
      "total cholesterol measurement http://www.ebi.ac.uk/efo/EFO_0004574\n",
      "lipid measurement http://www.ebi.ac.uk/efo/EFO_0004529\n",
      "low density lipoprotein cholesterol measurement http://www.ebi.ac.uk/efo/EFO_0004611\n",
      "lipid measurement http://www.ebi.ac.uk/efo/EFO_0004529\n",
      "\n",
      "LDL-cholesterol concentrations: a genome-wide association study.\n",
      "LDL-cholesterol concentrations: a genome-wide association study. LDL cholesterol has a causal role in the development of cardiovascular disease. Improved understanding of the biological mechanisms that underlie the metabolism and regulation of LDL cholesterol might help to identify novel therapeutic targets. We therefore did a genome-wide association study of LDL-cholesterol concentrations. We used genome-wide association data from up to 11 685 participants with measures of circulating LDL-cholesterol concentrations across five studies, including data for 293 461 autosomal single nucleotide polymorphisms (SNPs) with a minor allele frequency of 5% or more that passed our quality control criteria. We also used data from a second genome-wide array in up to 4337 participants from three of these five studies, with data for 290 140 SNPs. We did replication studies in two independent populations consisting of up to 4979 participants. Statistical approaches, including meta-analysis and linkage disequilibrium plots, were used to refine association signals; we analysed pooled data from all seven populations to determine the effect of each SNP on variations in circulating LDL-cholesterol concentrations. In our initial scan, we found two SNPs (rs599839 [p=1·7×10 −15 ] and rs4970834 [p=3·0×10 −11 ]) that showed genome-wide statistical association with LDL cholesterol at chromosomal locus 1p13.3. The second genome screen found a third statistically associated SNP at the same locus (rs646776 [p=4·3×10 −9 ]). Meta-analysis of data from all studies showed an association of SNPs rs599839 (combined p=1·2×10 −33 ) and rs646776 (p=4·8×10 −20 ) with LDL-cholesterol concentrations. SNPs rs599839 and rs646776 both explained around 1% of the variation in circulating LDL-cholesterol concentrations and were associated with about 15% of an SD change in LDL cholesterol per allele, assuming an SD of 1 mmol/L. We found evidence for a novel locus for LDL cholesterol on chromosome 1p13.3. These results potentially provide insight into the biological mechanisms that underlie the regulation of LDL cholesterol and might help in the discovery of novel therapeutic targets for cardiovascular disease.\n",
      "\n",
      "19430480\n",
      "type i diabetes mellitus http://www.ebi.ac.uk/efo/EFO_0001359\n",
      "type i diabetes mellitus http://www.ebi.ac.uk/efo/EFO_0001359\n",
      "iddm|diabetes mellitus\n",
      "Genome-wide association study and meta-analysis find that over 40 loci affect risk of type 1 diabetes.\n",
      "Genome-wide association study and meta-analysis finds over 40 loci affect risk of type 1 diabetes. Type 1 diabetes (T1D) is a common autoimmune disorder that arises from the action of multiple genetic and environmental risk factors. We report the findings of a new genome-wide association study of T1D, combined in a meta-analysis with two previously published studies. The total sample set included 7,514 cases and 9,045 reference samples. Forty-one distinct genomic locations provided evidence for association to T1D in the meta-analysis ( P  < 10 -6 ). After excluding previously reported associations, 27 regions were further tested in an independent set of 4,267 cases, 4,463 controls and 2,319 affected sib-pair (ASP) families. Of these, 18 regions were replicated ( P  < 0.01; overall  P  < 5 × 10 -8 ) and four additional regions provided nominal evidence of replication ( P  < 0.05). The many new candidate genes suggested by these results include  IL10, IL19, IL20, GLIS3, CD69  and  IL27 .\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from extractor.util import gold_phen_recall\n",
    "kb = KnowledgeBase()\n",
    "id2doc = {doc.name : doc for doc in corpus.documents}\n",
    "gold_dict_phen = { doc_id : set() for doc_id, phen_id in gold_set_phens }\n",
    "for doc_id, phen_id in gold_set_phens:\n",
    "    gold_dict_phen[doc_id].add(phen_id)\n",
    "\n",
    "phen_not_found = list(gold_phen_recall(phen_c, gold_set_phens, phen2id))\n",
    "print len(phen_not_found)\n",
    "for doc_id, phen_id in phen_not_found[:100]:\n",
    "    if len(gold_dict_phen[doc_id]) > 3: continue\n",
    "    print doc_id\n",
    "    for phen_id2 in gold_dict_phen[doc_id]:\n",
    "        print id2phen[phen_id2].name, phen_id2\n",
    "#     print gold_dict_phen[doc_id]\n",
    "    print id2phen[phen_id].name, id2phen[phen_id].ontology_ref\n",
    "    print id2phen[phen_id].synonyms    \n",
    "    print kb.title_by_pmid(doc_id)\n",
    "    print ' '.join(s.text for s in id2doc[doc_id].sentences)\n",
    "    print"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Over aggregate phenotypes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We say that a mention for an aggregate phenotype is correct, if it corresponds to the name of the GWC phenotype or to the phenotype of any equivalent EFO phenotype."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 668,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 322 gold mentions, e.g.:\n",
      "[('17658951', u'http://www.ebi.ac.uk/efo/EFO_0004764'), ('17903294', u'http://purl.obolibrary.org/obo/GO_0070527'), ('20038947', u'http://www.ebi.ac.uk/efo/EFO_0003761'), ('19197348', u'http://www.ebi.ac.uk/efo/EFO_0004748'), ('17658951', u'http://www.ebi.ac.uk/efo/EFO_0004626')]\n",
      "441 216\n",
      "Statistics over EFO phenotypes:\n",
      "# of gold annotations\t= 925\n",
      "# of candidates\t\t= 3088\n",
      "# of correct candidates\t= 802\n",
      "Candidate recall\t= 0.867\n",
      "Candidate precision\t= 0.260\n"
     ]
    }
   ],
   "source": [
    "from db.kb import KnowledgeBase\n",
    "from nltk.stem import PorterStemmer\n",
    "from extractor.util import change_name\n",
    "\n",
    "kb = KnowledgeBase() # reload\n",
    "gold_set_agg_phens = frozenset \\\n",
    "([ \n",
    "    (doc.name, phen.id) for doc in corpus.documents \n",
    "                        for phen in kb.phen_by_pmid(doc.name, source='gwas_catalog')\n",
    "])\n",
    "\n",
    "# map phenotype names to their id (EFO syn -> GWC id)\n",
    "agg_phen2id = dict()\n",
    "for doc in corpus.documents:\n",
    "    for phen in kb.phen_by_pmid(doc.name, source='gwas_catalog'):\n",
    "        for eq_phen in phen.equivalents:\n",
    "            for syn in [phen.name] + [eq_phen.name] + eq_phen.synonyms.split('|'):\n",
    "                syn_name = change_name(syn)\n",
    "                if eq_phen.name == 'personality': print syn_name\n",
    "                if syn_name not in agg_phen2id: agg_phen2id[syn_name] = set()\n",
    "                agg_phen2id[syn_name].add(phen.id)\n",
    "\n",
    "# map ids to phenotypes (GWC id -> GWC phen obj)                \n",
    "agg_id2phen = \\\n",
    "{\n",
    "    phen.id : phen for doc in corpus.documents\n",
    "                   for phen in kb.phen_by_pmid(doc.name, source='gwas_catalog')\n",
    "}\n",
    "\n",
    "print 'Found %d gold mentions, e.g.:' % len(gold_set_phens)\n",
    "print list(gold_set_phens)[:5]\n",
    "print len(phen2id), len(id2phen)\n",
    "\n",
    "from extractor.util import gold_phen_stats, gold_agg_phen_stats\n",
    "gold_agg_phen_stats(phen_c, gold_set_agg_phens, agg_phen2id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Why is the recall low?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 669,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "123\n",
      "19578366\n",
      "glioma (high-grade) 20856\n",
      "\tcentral nervous system cancer http://www.ebi.ac.uk/efo/EFO_0000326 cns cancer|malignant central nervous system neoplasm|malignant tumor of cns|cancer of cns|malignant tumor of central nervous system|malignant cns neoplasms|malignant neoplasm of cns|malignant tumor of the central nervous system|malignant cns tumor|malignant central nervous system tumor|malignant neoplasm of the cns|malignant neoplasm of the central nervous system|malignant cns neoplasm|central nervous system neoplasms\n",
      "glioma (high-grade) 20857\n",
      "\tcentral nervous system cancer http://www.ebi.ac.uk/efo/EFO_0000326 cns cancer|malignant central nervous system neoplasm|malignant tumor of cns|cancer of cns|malignant tumor of central nervous system|malignant cns neoplasms|malignant neoplasm of cns|malignant tumor of the central nervous system|malignant cns tumor|malignant central nervous system tumor|malignant neoplasm of the cns|malignant neoplasm of the central nervous system|malignant cns neoplasm|central nervous system neoplasms\n",
      "glioma (high-grade) 20855\n",
      "\tcentral nervous system cancer http://www.ebi.ac.uk/efo/EFO_0000326 cns cancer|malignant central nervous system neoplasm|malignant tumor of cns|cancer of cns|malignant tumor of central nervous system|malignant cns neoplasms|malignant neoplasm of cns|malignant tumor of the central nervous system|malignant cns tumor|malignant central nervous system tumor|malignant neoplasm of the cns|malignant neoplasm of the central nervous system|malignant cns neoplasm|central nervous system neoplasms\n",
      "glioma (high-grade)\n",
      "Variants in the  CDKN2B  and  RTEL1  regions are associated with high grade glioma susceptibility. The causes of glioblastoma and other devastating gliomas remain obscure  1 , 2 . To discover new candidate genes that might influence glioma susceptibility, we conducted a principal component  3  adjusted genome wide association study (GWAS) of 275,895 autosomal variants among 692 adult high grade glioma patients (622 from the San Francisco Adult Glioma Study (AGS) and 70 from the Cancer Genome Atlas (TCGA)  4  versus 3992 controls (602 from AGS and 3390 from Illumina iControlDB (iControls)). For replication, we then analyzed the 13 SNPs with p<10 −6  using completely independent data from 176 high grade glioma patients versus 174 controls from the Mayo Clinic. Rs1412829 in chromosome 9p21 (near  CDKN2B ) had discovery p-value 3.4*10 −8 , replication p-value 0.0038, and combined p-value 1.85*10 −10 . Rs6010620 intronic to  RTEL1  had discovery p-value 1.5*10 −7 , replication p-value 0.00035, and combined p-value 3.40*10 −9 . For both SNPs, the direction of association was the same in discovery and replication phases.\n",
      "\n",
      "19081515\n",
      "creutzfeldt-jakob disease 20257\n",
      "\tcreutzfeldt jacob disease http://www.ebi.ac.uk/efo/EFO_0004226 spongiform encephalopathy\n",
      "creutzfeldt-jakob disease 20258\n",
      "\tcreutzfeldt jacob disease http://www.ebi.ac.uk/efo/EFO_0004226 spongiform encephalopathy\n",
      "creutzfeldt-jakob disease\n",
      "Genetic risk factors for variant Creutzfeldt–Jakob disease: a genome-wide association study. Human and animal prion diseases are under genetic control, but apart from  PRNP  (the gene that encodes the prion protein), we understand little about human susceptibility to bovine spongiform encephalopathy (BSE) prions, the causal agent of variant Creutzfeldt–Jakob disease (vCJD). We did a genome-wide association study of the risk of vCJD and tested for replication of our findings in samples from many categories of human prion disease (929 samples) and control samples from the UK and Papua New Guinea (4254 samples), including controls in the UK who were genotyped by the Wellcome Trust Case Control Consortium. We also did follow-up analyses of the genetic control of the clinical phenotype of prion disease and analysed candidate gene expression in a mouse cellular model of prion infection. The  PRNP  locus was strongly associated with risk across several markers and all categories of prion disease (best single SNP [single nucleotide polymorphism] association in vCJD p=2·5×10 −17 ; best haplotypic association in vCJD p=1×10 −24 ). Although the main contribution to disease risk was conferred by  PRNP  polymorphic codon 129, another nearby SNP conferred increased risk of vCJD. In addition to  PRNP , one technically validated SNP association upstream of  RARB  (the gene that encodes retinoic acid receptor beta) had nominal genome-wide significance (p=1·9×10 −7 ). A similar association was found in a small sample of patients with iatrogenic CJD (p=0·030) but not in patients with sporadic CJD (sCJD) or kuru. In cultured cells, retinoic acid regulates the expression of the prion protein. We found an association with acquired prion disease, including vCJD (p=5·6×10 −5 ), kuru incubation time (p=0·017), and resistance to kuru (p=2·5×10 −4 ), in a region upstream of  STMN2  (the gene that encodes SCG10). The risk genotype was not associated with sCJD but conferred an earlier age of onset. Furthermore, expression of  Stmn2  was reduced 30-fold post-infection in a mouse cellular model of prion disease. The polymorphic codon 129 of  PRNP  was the main genetic risk factor for vCJD; however, additional candidate loci have been identified, which justifies functional analyses of these biological pathways in prion disease. The UK Medical Research Council.\n",
      "\n",
      "19421330\n",
      "hypertension (young onset) 20627\n",
      "\tearly onset hypertension http://www.ebi.ac.uk/efo/EFO_0004772 \n",
      "hypertension (young onset)\n",
      "Genome-Wide Association Study of Young-Onset Hypertension in the Han Chinese Population of Taiwan. Young-onset hypertension has a stronger genetic component than late-onset counterpart; thus, the identification of genes related to its susceptibility is a critical issue for the prevention and management of this disease. We carried out a two-stage association scan to map young-onset hypertension susceptibility genes. The first-stage analysis, a genome-wide association study, analyzed 175 matched case-control pairs; the second-stage analysis, a confirmatory association study, verified the results at the first stage based on a total of 1,008 patients and 1,008 controls. Single-locus association tests, multilocus association tests and pair-wise gene-gene interaction tests were performed to identify young-onset hypertension susceptibility genes. After considering stringent adjustments of multiple testing, gene annotation and single-nucleotide polymorphism (SNP) quality, four SNPs from two SNP triplets with strong association signals (−log 10 (p)>7) and 13 SNPs from 8 interactive SNP pairs with strong interactive signals (−log 10 (p)>8) were carefully re-examined. The confirmatory study verified the association for a SNP quartet 219 kb and 495 kb downstream of  LOC344371  (a hypothetical gene) and  RASGRP3  on chromosome 2p22.3, respectively. The latter has been implicated in the abnormal vascular responsiveness to endothelin-1 and angiotensin II in diabetic-hypertensive rats. Intrinsic synergy involving  IMPG1  on chromosome 6q14.2-q15 was also verified. IMPG1  encodes interphotoreceptor matrix proteoglycan 1 which has cation binding capacity. The genes are novel hypertension targets identified in this first genome-wide hypertension association study of the Han Chinese population.\n",
      "\n",
      "19578366\n",
      "glioma (high-grade) 20856\n",
      "\tcentral nervous system cancer http://www.ebi.ac.uk/efo/EFO_0000326 cns cancer|malignant central nervous system neoplasm|malignant tumor of cns|cancer of cns|malignant tumor of central nervous system|malignant cns neoplasms|malignant neoplasm of cns|malignant tumor of the central nervous system|malignant cns tumor|malignant central nervous system tumor|malignant neoplasm of the cns|malignant neoplasm of the central nervous system|malignant cns neoplasm|central nervous system neoplasms\n",
      "glioma (high-grade) 20857\n",
      "\tcentral nervous system cancer http://www.ebi.ac.uk/efo/EFO_0000326 cns cancer|malignant central nervous system neoplasm|malignant tumor of cns|cancer of cns|malignant tumor of central nervous system|malignant cns neoplasms|malignant neoplasm of cns|malignant tumor of the central nervous system|malignant cns tumor|malignant central nervous system tumor|malignant neoplasm of the cns|malignant neoplasm of the central nervous system|malignant cns neoplasm|central nervous system neoplasms\n",
      "glioma (high-grade) 20855\n",
      "\tcentral nervous system cancer http://www.ebi.ac.uk/efo/EFO_0000326 cns cancer|malignant central nervous system neoplasm|malignant tumor of cns|cancer of cns|malignant tumor of central nervous system|malignant cns neoplasms|malignant neoplasm of cns|malignant tumor of the central nervous system|malignant cns tumor|malignant central nervous system tumor|malignant neoplasm of the cns|malignant neoplasm of the central nervous system|malignant cns neoplasm|central nervous system neoplasms\n",
      "glioma (high-grade)\n",
      "Variants in the  CDKN2B  and  RTEL1  regions are associated with high grade glioma susceptibility. The causes of glioblastoma and other devastating gliomas remain obscure  1 , 2 . To discover new candidate genes that might influence glioma susceptibility, we conducted a principal component  3  adjusted genome wide association study (GWAS) of 275,895 autosomal variants among 692 adult high grade glioma patients (622 from the San Francisco Adult Glioma Study (AGS) and 70 from the Cancer Genome Atlas (TCGA)  4  versus 3992 controls (602 from AGS and 3390 from Illumina iControlDB (iControls)). For replication, we then analyzed the 13 SNPs with p<10 −6  using completely independent data from 176 high grade glioma patients versus 174 controls from the Mayo Clinic. Rs1412829 in chromosome 9p21 (near  CDKN2B ) had discovery p-value 3.4*10 −8 , replication p-value 0.0038, and combined p-value 1.85*10 −10 . Rs6010620 intronic to  RTEL1  had discovery p-value 1.5*10 −7 , replication p-value 0.00035, and combined p-value 3.40*10 −9 . For both SNPs, the direction of association was the same in discovery and replication phases.\n",
      "\n",
      "19081515\n",
      "creutzfeldt-jakob disease 20257\n",
      "\tcreutzfeldt jacob disease http://www.ebi.ac.uk/efo/EFO_0004226 spongiform encephalopathy\n",
      "creutzfeldt-jakob disease 20258\n",
      "\tcreutzfeldt jacob disease http://www.ebi.ac.uk/efo/EFO_0004226 spongiform encephalopathy\n",
      "creutzfeldt-jakob disease\n",
      "Genetic risk factors for variant Creutzfeldt–Jakob disease: a genome-wide association study. Human and animal prion diseases are under genetic control, but apart from  PRNP  (the gene that encodes the prion protein), we understand little about human susceptibility to bovine spongiform encephalopathy (BSE) prions, the causal agent of variant Creutzfeldt–Jakob disease (vCJD). We did a genome-wide association study of the risk of vCJD and tested for replication of our findings in samples from many categories of human prion disease (929 samples) and control samples from the UK and Papua New Guinea (4254 samples), including controls in the UK who were genotyped by the Wellcome Trust Case Control Consortium. We also did follow-up analyses of the genetic control of the clinical phenotype of prion disease and analysed candidate gene expression in a mouse cellular model of prion infection. The  PRNP  locus was strongly associated with risk across several markers and all categories of prion disease (best single SNP [single nucleotide polymorphism] association in vCJD p=2·5×10 −17 ; best haplotypic association in vCJD p=1×10 −24 ). Although the main contribution to disease risk was conferred by  PRNP  polymorphic codon 129, another nearby SNP conferred increased risk of vCJD. In addition to  PRNP , one technically validated SNP association upstream of  RARB  (the gene that encodes retinoic acid receptor beta) had nominal genome-wide significance (p=1·9×10 −7 ). A similar association was found in a small sample of patients with iatrogenic CJD (p=0·030) but not in patients with sporadic CJD (sCJD) or kuru. In cultured cells, retinoic acid regulates the expression of the prion protein. We found an association with acquired prion disease, including vCJD (p=5·6×10 −5 ), kuru incubation time (p=0·017), and resistance to kuru (p=2·5×10 −4 ), in a region upstream of  STMN2  (the gene that encodes SCG10). The risk genotype was not associated with sCJD but conferred an earlier age of onset. Furthermore, expression of  Stmn2  was reduced 30-fold post-infection in a mouse cellular model of prion disease. The polymorphic codon 129 of  PRNP  was the main genetic risk factor for vCJD; however, additional candidate loci have been identified, which justifies functional analyses of these biological pathways in prion disease. The UK Medical Research Council.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from extractor.util import gold_agg_phen_recall\n",
    "kb = KnowledgeBase()\n",
    "id2doc = {doc.name : doc for doc in corpus.documents}\n",
    "gold_agg_dict_phen = { doc_id : set() for doc_id, phen_id in gold_set_agg_phens }\n",
    "for doc_id, phen_id in gold_set_agg_phens:\n",
    "    gold_agg_dict_phen[doc_id].add(phen_id)\n",
    "\n",
    "# this contains (doc_id, gwc_id) pairs from gold_set_agg_phens that haven't been found\n",
    "phen_not_found = list(gold_agg_phen_recall(phen_c, gold_set_agg_phens, agg_phen2id))\n",
    "print len(phen_not_found)\n",
    "for doc_id, phen_id in phen_not_found[:100]:\n",
    "    if len(gold_agg_dict_phen[doc_id]) > 3: continue # skip if >3 gwc_phen_ids in doc\n",
    "    print doc_id\n",
    "    for phen_id2 in gold_agg_dict_phen[doc_id]: # iterate over gwc_id's in doc\n",
    "        print agg_id2phen[phen_id2].name, phen_id2 # print its name and id\n",
    "        for eq_phen in agg_id2phen[phen_id2].equivalents: # loop over equivalent phens\n",
    "            print '\\t', eq_phen.name, eq_phen.ontology_ref, eq_phen.synonyms # print name, ref, synonyms\n",
    "    print agg_id2phen[phen_id].name # print the one that we haven't found\n",
    "    print ' '.join(s.text for s in id2doc[doc_id].sentences)\n",
    "    print"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 670,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# FOR DEBUGGING WHY SPANS ARENT MATCHED\n",
    "# from extractor.util import change_name\n",
    "\n",
    "# doc_id, phen_id = phen_not_found[3]\n",
    "# print agg_id2phen[phen_id].name \n",
    "# print id2doc[doc_id].sentences[0]\n",
    "# for span in ngrams.apply(id2doc[doc_id].sentences[0]):\n",
    "#     print span.get_span()\n",
    "#     if phen_matcher._f(span):    \n",
    "#         phen_name = span.get_span()\n",
    "#         print phen_name\n",
    "#         print '...', change_name(phen_name) in phenotype_list, phen_name in phen_matcher.d, change_name(phen_name) in phen_matcher.d\n",
    "#         phen_id = phen2id.get(change_name(phen_name), None)\n",
    "#         print phen_id\n",
    "#         if not phen_id or phen_id not in gold_dict_phen[span.context.document.name]:\n",
    "#             print span.context.document.name, phen_id\n",
    "#             print gold_dict_phen[span.context.document.name]\n",
    "        \n",
    "#         print"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 671,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# query_word = 'personality'\n",
    "\n",
    "# # phenotypes = db_session.query(Phenotype).filter(Phenotype.source=='efo').all()\n",
    "# phenotypes == kb.get_phenotype_candidates_cheating()\n",
    "# phenotype_names = set()\n",
    "# for phenotype in phenotypes:\n",
    "#     if phenotype.name:\n",
    "#         phenotype_names.add((phenotype.name))\n",
    "#         synonyms = [(syn) for syn in phenotype.synonyms.split('|')]\n",
    "#         if query_word in synonyms or query_word == phenotype.name:\n",
    "#             print phenotype.name, phenotype.ontology_ref\n",
    "#         phenotype_names.update(synonyms)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 672,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# print len(phenotype_names)\n",
    "# [(word, change_name(word)) for word in phenotype_list if change_name(word) == change_name('personalized')]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extraction performance"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, extract some features for each candidate mention."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 673,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Building feature index...\n",
      "Extracting features...\n",
      "0/53842\n",
      "5000/53842\n",
      "10000/53842\n",
      "15000/53842\n",
      "20000/53842\n",
      "25000/53842\n",
      "30000/53842\n",
      "35000/53842\n",
      "40000/53842\n",
      "45000/53842\n",
      "50000/53842\n",
      "Example features: [u'DDLIB_WORD_SEQ_[Crohn Disease]', u'DDLIB_LEMMA_SEQ_[crohn disease]', u'DDLIB_POS_SEQ_[NN NN]', u'DDLIB_DEP_SEQ_[compound compound]', u'DDLIB_W_LEFT_1_[novel]', u'DDLIB_W_LEFT_POS_1_[JJ]', u'DDLIB_W_LEFT_2_[. novel]', u'DDLIB_W_LEFT_POS_2_[. JJ]', u'DDLIB_W_LEFT_3_[ptger4 . novel]', u'DDLIB_W_LEFT_POS_3_[NN . JJ]']\n"
     ]
    }
   ],
   "source": [
    "import cPickle        \n",
    "from snorkel.features import NgramFeaturizer\n",
    "\n",
    "pkl_f = 'phenotype_feats.pkl'\n",
    "try:\n",
    "    with open(pkl_f, 'rb') as f:\n",
    "        featurizer = cPickle.load(f)\n",
    "except:\n",
    "    featurizer = NgramFeaturizer()\n",
    "    featurizer.fit_transform(phen_c)\n",
    "\n",
    "print 'Example features:', [f for f in featurizer.get_features_by_candidate(phen_c[0])[:10]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 674,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Defined 547 positive and 2541 negative gold mentions\n"
     ]
    }
   ],
   "source": [
    "# classify candidates as correct or not\n",
    "\n",
    "# doc_id -> set of correct gwc_ids\n",
    "gold_agg_dict_phen = { doc_id : set() for doc_id, phen_id in gold_set_agg_phens }\n",
    "for doc_id, phen_id in gold_set_agg_phens:\n",
    "    gold_agg_dict_phen[doc_id].add(phen_id)\n",
    "\n",
    "def c2uid(candidate):\n",
    "    return candidate.context.document.name, candidate.context.position, candidate.char_start, candidate.char_end\n",
    "    \n",
    "gt_dict_pos = dict()\n",
    "gt_dict_neg = dict()\n",
    "for candidate in phen_c:\n",
    "    doc_id = candidate.context.document.name\n",
    "    agg_ids = agg_phen2id.get(change_name(candidate.get_span()), set())\n",
    "    uid = c2uid(candidate)\n",
    "    if agg_ids & gold_agg_dict_phen[doc_id]:\n",
    "        gt_dict_pos[uid] = +1\n",
    "    else:\n",
    "        gt_dict_neg[uid] = -1\n",
    "\n",
    "gt_dict = dict(gt_dict_pos.items() + gt_dict_neg.items())\n",
    "print 'Defined %d positive and %d negative gold mentions' % (len(gt_dict_pos), len(gt_dict_neg))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 675,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "17447842\n",
      "crohn's disease 19827\n",
      "\tcrohn's disease http://www.ebi.ac.uk/efo/EFO_0000384 gastritis associated with crohn's disease|ileitis\n",
      "crohn's disease 19333\n",
      "\tcrohn's disease http://www.ebi.ac.uk/efo/EFO_0000384 gastritis associated with crohn's disease|ileitis\n",
      "crohn's disease 19486\n",
      "\tcrohn's disease http://www.ebi.ac.uk/efo/EFO_0000384 gastritis associated with crohn's disease|ileitis\n",
      "Novel Crohn Disease Locus Identified by Genome-Wide Association Maps to a Gene Desert on 5p13.1 and Modulates Expression of  PTGER4. To identify novel susceptibility loci for Crohn disease (CD), we undertook a genome-wide association study with more than 300,000 SNPs characterized in 547 patients and 928 controls. We found three chromosome regions that provided evidence of disease association with  p -values between 10 −6  and 10 −9 . Two of these ( IL23R  on Chromosome 1 and  CARD15  on Chromosome 16) correspond to genes previously reported to be associated with CD. In addition, a 250-kb region of Chromosome 5p13.1 was found to contain multiple markers with strongly suggestive evidence of disease association (including four markers with  p  < 10 −7 ). We replicated the results for 5p13.1 by studying 1,266 additional CD patients, 559 additional controls, and 428 trios. Significant evidence of association ( p  < 4 × 10 −4 ) was found in case/control comparisons with the replication data, while associated alleles were over-transmitted to affected offspring ( p  < 0.05), thus confirming that the 5p13.1 locus contributes to CD susceptibility. The CD-associated 250-kb region was saturated with 111 SNP markers. Haplotype analysis supports a complex locus architecture with multiple variants contributing to disease susceptibility. The novel 5p13.1 CD locus is contained within a 1.25-Mb gene desert. We present evidence that disease-associated alleles correlate with quantitative expression levels of the prostaglandin receptor EP4,  PTGER4,  the gene that resides closest to the associated region. Our results identify a major new susceptibility locus for CD, and suggest that genetic variants associated with disease risk at this locus could modulate  cis -acting regulatory elements of  PTGER4 . Individual susceptibility to many common diseases is determined by a combination of environmental and genetic factors. Identifying these genetic risk factors is one of the most important objectives of modern medical genetics, as it paves the way towards personalized medicine and drug target identification. Recent advances in SNP genotyping technology allows systematic association scanning of the entire genome for the detection of novel susceptibility loci. We herein apply this approach to Crohn disease, which afflicts an estimated 0.15% of the people in the developed world and identify a novel susceptibility locus on Chromosome 5. A unique feature of the novel 5p13.1 locus is that it coincides with a 1.25-Mb gene desert. We present evidence that genetic variants at this locus influence the expression levels of the closest gene,  PTGER4,  located 270 kb away, in the direction of the centromere. PTGER4  encodes the prostaglandin receptor EP4 and is a strong candidate susceptibility gene for Crohn disease as PTGER4 knock-out mice have increased susceptibility to colitis.\n",
      "1 Crohn Disease | crohn's disease http://www.ebi.ac.uk/efo/EFO_0000384\n",
      "-1 Maps | mean arterial pressure http://www.ebi.ac.uk/efo/EFO_0006340\n",
      "-1 .1\n",
      "-1 (CD),\n",
      "1 Crohn disease ( | crohn's disease http://www.ebi.ac.uk/efo/EFO_0000384\n",
      "-1 disease association with\n",
      "-1 of disease\n",
      "-1 on Chromosome 16)\n",
      "-1 Chromosome 1 and\n",
      "-1 CD.\n",
      "-1 of disease\n",
      "-1 (including\n",
      "-1 .1\n",
      "-1 was\n",
      "-1 multiple\n",
      "-1 .1\n",
      "-1 CD\n",
      "-1 ) was\n",
      "-1 with the\n",
      "-1 .1\n",
      "-1 CD\n",
      "-1 was\n",
      "-1 disease susceptibility.\n",
      "-1 to disease\n",
      "-1 multiple\n",
      "-1 .1\n",
      "-1 CD\n",
      "-1 is\n",
      "-1 CD,\n",
      "-1 with disease\n",
      "-1 elements\n",
      "-1 susceptibility to\n",
      "-1 common diseases\n",
      "-1 is\n",
      "-1 , as\n",
      "-1 is\n",
      "-1 personalized\n",
      "-1 locus on Chromosome 5.\n",
      "1 Crohn disease, | crohn's disease http://www.ebi.ac.uk/efo/EFO_0000384\n",
      "-1 an\n",
      "-1 .1\n",
      "-1 is\n",
      "-1 direction\n",
      "1 Crohn disease | crohn's disease http://www.ebi.ac.uk/efo/EFO_0000384\n",
      "-1 susceptibility to\n",
      "-1 colitis.\n",
      "-1 is\n",
      "-1 as\n",
      "\n",
      "17658951\n",
      "obesity-related traits 19256\n",
      "\tadiponectin measurement http://www.ebi.ac.uk/efo/EFO_0004502 adiponectin levels\n",
      "\thoma-ir http://www.ebi.ac.uk/efo/EFO_0004501 homeostasis model assessment insulin resistance\n",
      "\tccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749 monocyte chemotactic protein-1 measurement|chemokine ccl2 level|monocyte chemoattractant protein-1 measurement|chemokine ccl2 levels|monocyte chemotactic protein-1 levels|monocyte chemotactic protein-1 level\n",
      "\tthyroid stimulating hormone measurement http://www.ebi.ac.uk/efo/EFO_0004748 thyrotropin measurement|thyroid stimulating hormone level|thyroid stimulating hormone levels\n",
      "\tprotein measurement http://www.ebi.ac.uk/efo/EFO_0004747 protein levels\n",
      "\tblood urea nitrogen measurement http://www.ebi.ac.uk/efo/EFO_0004741 \n",
      "\tlean body mass http://www.ebi.ac.uk/efo/EFO_0004995 \n",
      "\tserum alanine aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004735 alanine transaminase measurement|alanine aminotransferase measurement\n",
      "\taspartate aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004736 aspartate transaminase measurement\n",
      "\thormone measurement http://www.ebi.ac.uk/efo/EFO_0004730 \n",
      "\tfat body mass http://www.ebi.ac.uk/efo/EFO_0005409 total body fat measurement\n",
      "\tadipose tissue measurement http://www.ebi.ac.uk/efo/EFO_0004764 \n",
      "\ticam-1 measurement http://www.ebi.ac.uk/efo/EFO_0004520 soluble icam1 measurement|soluble intercellular adhesion molecule-1|sicam-1\n",
      "\ttriglyceride measurement http://www.ebi.ac.uk/efo/EFO_0004530 triglyceride levels\n",
      "\twaist circumference http://www.ebi.ac.uk/efo/EFO_0004342 \n",
      "\tbody mass index http://www.ebi.ac.uk/efo/EFO_0004340 quetelet's index|bmi\n",
      "\tobesity http://www.ebi.ac.uk/efo/EFO_0001073 overweight and obesity|obesity (disorder)|obese|obesity nos|obese (finding)|adiposity|adiposis|obesity [ambiguous]|obesity\n",
      "\tbirth weight http://www.ebi.ac.uk/efo/EFO_0004344 birthweight\n",
      "\twaist-hip ratio http://www.ebi.ac.uk/efo/EFO_0004343 \n",
      "\theart rate http://www.ebi.ac.uk/efo/EFO_0004326 pulse rate|cardiac chronotropy\n",
      "\tbody height http://www.ebi.ac.uk/efo/EFO_0004339 height\n",
      "\tbody weight http://www.ebi.ac.uk/efo/EFO_0004338 weight\n",
      "\thomocysteine measurement http://www.ebi.ac.uk/efo/EFO_0004578 homocysteine levels\n",
      "\ttotal cholesterol measurement http://www.ebi.ac.uk/efo/EFO_0004574 cholesterol\n",
      "\tccl5 measurement http://www.ebi.ac.uk/efo/EFO_0005117 regulated on activation\n",
      "\tigfbp-1 measurement http://www.ebi.ac.uk/efo/EFO_0005118 \n",
      "\tantioxidant measurement http://www.ebi.ac.uk/efo/EFO_0005119 \n",
      "\thead circumference http://www.ebi.ac.uk/efo/EFO_0005114 \n",
      "\tmetabolic rate measurement http://www.ebi.ac.uk/efo/EFO_0005115 \n",
      "\turinary metabolite measurement http://www.ebi.ac.uk/efo/EFO_0005116 \n",
      "\tfatty acid measurement http://www.ebi.ac.uk/efo/EFO_0005110 \n",
      "\tfolic acid measurement http://www.ebi.ac.uk/efo/EFO_0005111 folate measurement|vitamin b9 measurement\n",
      "\tgestational age http://www.ebi.ac.uk/efo/EFO_0005112 \n",
      "\tbody composition measurement http://www.ebi.ac.uk/efo/EFO_0005106 \n",
      "\tarm span http://www.ebi.ac.uk/efo/EFO_0005108 \n",
      "\tenergy expenditure http://www.ebi.ac.uk/efo/EFO_0005109 \n",
      "\tamino acid measurement http://www.ebi.ac.uk/efo/EFO_0005134 \n",
      "\tc-peptide measurement http://www.ebi.ac.uk/efo/EFO_0005187 \n",
      "\tccl11 measurement http://www.ebi.ac.uk/efo/EFO_0005188 \n",
      "\trespiratory quotient http://www.ebi.ac.uk/efo/EFO_0005189 \n",
      "\turinary nitrogen measurement http://www.ebi.ac.uk/efo/EFO_0005190 \n",
      "\twaist height ratio http://www.ebi.ac.uk/efo/EFO_0005191 \n",
      "\ttestosterone measurement http://www.ebi.ac.uk/efo/EFO_0004908 testosterone levels|testosterone level\n",
      "\tinterleukin-6 measurement http://www.ebi.ac.uk/efo/EFO_0004810 il-6 levels|il-6 level|il-6 measurement\n",
      "\ttransforming growth factor beta measurement http://www.ebi.ac.uk/efo/EFO_0004818 \n",
      "\tigf-1 measurement http://www.ebi.ac.uk/efo/EFO_0004627 insulin-like growth factor 1 levels\n",
      "\tigfbp-3 measurement http://www.ebi.ac.uk/efo/EFO_0004626 insulin-like growth factor binding protein 3 levels\n",
      "\tvitamin b12 measurement http://www.ebi.ac.uk/efo/EFO_0004620 vitamin b12 levels\n",
      "\tsleep measurement http://www.ebi.ac.uk/efo/EFO_0004870 \n",
      "\thigh density lipoprotein cholesterol measurement http://www.ebi.ac.uk/efo/EFO_0004612 hdl measurement\n",
      "\tlow density lipoprotein cholesterol measurement http://www.ebi.ac.uk/efo/EFO_0004611 ldl measurement\n",
      "\tmaximal oxygen uptake measurement http://www.ebi.ac.uk/efo/EFO_0004887 maximal oxygen consumption|peak oxygen uptake|maximal oxygen uptake|maximal aerobic capacity|vo2 max\n",
      "\tcortisol secretion measurement http://www.ebi.ac.uk/efo/EFO_0004642 \n",
      "\tfasting blood insulin measurement http://www.ebi.ac.uk/efo/EFO_0004466 fasting blood insulin|fasting blood insulin level|fasting insulin-related traits\n",
      "\tfasting blood glucose measurement http://www.ebi.ac.uk/efo/EFO_0004465 fasting glucose-related traits|fasting blood glucose level|fasting plasma glucose\n",
      "\tinsulin sensitivity measurement http://www.ebi.ac.uk/efo/EFO_0004471 insulin sensitivity test\n",
      "\ttumor necrosis factor-alpha measurement http://www.ebi.ac.uk/efo/EFO_0004684 \n",
      "\tc-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458 c-reactive protein level\n",
      "\testradiol measurement http://www.ebi.ac.uk/efo/EFO_0004697 oestrogen level|oestrogen levels|estradiol levels\n",
      "\tsystolic blood pressure http://www.ebi.ac.uk/efo/EFO_0006335 sysbp|systolic pressure\n",
      "\tdiastolic blood pressure http://www.ebi.ac.uk/efo/EFO_0006336 diastolic pressure|diabp\n",
      "\tleptin measurement http://www.ebi.ac.uk/efo/EFO_0005000 \n",
      "\thip circumference http://www.ebi.ac.uk/efo/EFO_0005093 \n",
      "\tenergy intake http://www.ebi.ac.uk/efo/EFO_0003939 caloric intake\n",
      "\tphysical activity http://www.ebi.ac.uk/efo/EFO_0003940 motor activity|activities\n",
      "\tbone density http://www.ebi.ac.uk/efo/EFO_0003923 density\n",
      "obesity-related traits 19254\n",
      "\tadiponectin measurement http://www.ebi.ac.uk/efo/EFO_0004502 adiponectin levels\n",
      "\thoma-ir http://www.ebi.ac.uk/efo/EFO_0004501 homeostasis model assessment insulin resistance\n",
      "\tccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749 monocyte chemotactic protein-1 measurement|chemokine ccl2 level|monocyte chemoattractant protein-1 measurement|chemokine ccl2 levels|monocyte chemotactic protein-1 levels|monocyte chemotactic protein-1 level\n",
      "\tthyroid stimulating hormone measurement http://www.ebi.ac.uk/efo/EFO_0004748 thyrotropin measurement|thyroid stimulating hormone level|thyroid stimulating hormone levels\n",
      "\tprotein measurement http://www.ebi.ac.uk/efo/EFO_0004747 protein levels\n",
      "\tblood urea nitrogen measurement http://www.ebi.ac.uk/efo/EFO_0004741 \n",
      "\tlean body mass http://www.ebi.ac.uk/efo/EFO_0004995 \n",
      "\tserum alanine aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004735 alanine transaminase measurement|alanine aminotransferase measurement\n",
      "\taspartate aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004736 aspartate transaminase measurement\n",
      "\thormone measurement http://www.ebi.ac.uk/efo/EFO_0004730 \n",
      "\tfat body mass http://www.ebi.ac.uk/efo/EFO_0005409 total body fat measurement\n",
      "\tadipose tissue measurement http://www.ebi.ac.uk/efo/EFO_0004764 \n",
      "\ticam-1 measurement http://www.ebi.ac.uk/efo/EFO_0004520 soluble icam1 measurement|soluble intercellular adhesion molecule-1|sicam-1\n",
      "\ttriglyceride measurement http://www.ebi.ac.uk/efo/EFO_0004530 triglyceride levels\n",
      "\twaist circumference http://www.ebi.ac.uk/efo/EFO_0004342 \n",
      "\tbody mass index http://www.ebi.ac.uk/efo/EFO_0004340 quetelet's index|bmi\n",
      "\tobesity http://www.ebi.ac.uk/efo/EFO_0001073 overweight and obesity|obesity (disorder)|obese|obesity nos|obese (finding)|adiposity|adiposis|obesity [ambiguous]|obesity\n",
      "\tbirth weight http://www.ebi.ac.uk/efo/EFO_0004344 birthweight\n",
      "\twaist-hip ratio http://www.ebi.ac.uk/efo/EFO_0004343 \n",
      "\theart rate http://www.ebi.ac.uk/efo/EFO_0004326 pulse rate|cardiac chronotropy\n",
      "\tbody height http://www.ebi.ac.uk/efo/EFO_0004339 height\n",
      "\tbody weight http://www.ebi.ac.uk/efo/EFO_0004338 weight\n",
      "\thomocysteine measurement http://www.ebi.ac.uk/efo/EFO_0004578 homocysteine levels\n",
      "\ttotal cholesterol measurement http://www.ebi.ac.uk/efo/EFO_0004574 cholesterol\n",
      "\tccl5 measurement http://www.ebi.ac.uk/efo/EFO_0005117 regulated on activation\n",
      "\tigfbp-1 measurement http://www.ebi.ac.uk/efo/EFO_0005118 \n",
      "\tantioxidant measurement http://www.ebi.ac.uk/efo/EFO_0005119 \n",
      "\thead circumference http://www.ebi.ac.uk/efo/EFO_0005114 \n",
      "\tmetabolic rate measurement http://www.ebi.ac.uk/efo/EFO_0005115 \n",
      "\turinary metabolite measurement http://www.ebi.ac.uk/efo/EFO_0005116 \n",
      "\tfatty acid measurement http://www.ebi.ac.uk/efo/EFO_0005110 \n",
      "\tfolic acid measurement http://www.ebi.ac.uk/efo/EFO_0005111 folate measurement|vitamin b9 measurement\n",
      "\tgestational age http://www.ebi.ac.uk/efo/EFO_0005112 \n",
      "\tbody composition measurement http://www.ebi.ac.uk/efo/EFO_0005106 \n",
      "\tarm span http://www.ebi.ac.uk/efo/EFO_0005108 \n",
      "\tenergy expenditure http://www.ebi.ac.uk/efo/EFO_0005109 \n",
      "\tamino acid measurement http://www.ebi.ac.uk/efo/EFO_0005134 \n",
      "\tc-peptide measurement http://www.ebi.ac.uk/efo/EFO_0005187 \n",
      "\tccl11 measurement http://www.ebi.ac.uk/efo/EFO_0005188 \n",
      "\trespiratory quotient http://www.ebi.ac.uk/efo/EFO_0005189 \n",
      "\turinary nitrogen measurement http://www.ebi.ac.uk/efo/EFO_0005190 \n",
      "\twaist height ratio http://www.ebi.ac.uk/efo/EFO_0005191 \n",
      "\ttestosterone measurement http://www.ebi.ac.uk/efo/EFO_0004908 testosterone levels|testosterone level\n",
      "\tinterleukin-6 measurement http://www.ebi.ac.uk/efo/EFO_0004810 il-6 levels|il-6 level|il-6 measurement\n",
      "\ttransforming growth factor beta measurement http://www.ebi.ac.uk/efo/EFO_0004818 \n",
      "\tigf-1 measurement http://www.ebi.ac.uk/efo/EFO_0004627 insulin-like growth factor 1 levels\n",
      "\tigfbp-3 measurement http://www.ebi.ac.uk/efo/EFO_0004626 insulin-like growth factor binding protein 3 levels\n",
      "\tvitamin b12 measurement http://www.ebi.ac.uk/efo/EFO_0004620 vitamin b12 levels\n",
      "\tsleep measurement http://www.ebi.ac.uk/efo/EFO_0004870 \n",
      "\thigh density lipoprotein cholesterol measurement http://www.ebi.ac.uk/efo/EFO_0004612 hdl measurement\n",
      "\tlow density lipoprotein cholesterol measurement http://www.ebi.ac.uk/efo/EFO_0004611 ldl measurement\n",
      "\tmaximal oxygen uptake measurement http://www.ebi.ac.uk/efo/EFO_0004887 maximal oxygen consumption|peak oxygen uptake|maximal oxygen uptake|maximal aerobic capacity|vo2 max\n",
      "\tcortisol secretion measurement http://www.ebi.ac.uk/efo/EFO_0004642 \n",
      "\tfasting blood insulin measurement http://www.ebi.ac.uk/efo/EFO_0004466 fasting blood insulin|fasting blood insulin level|fasting insulin-related traits\n",
      "\tfasting blood glucose measurement http://www.ebi.ac.uk/efo/EFO_0004465 fasting glucose-related traits|fasting blood glucose level|fasting plasma glucose\n",
      "\tinsulin sensitivity measurement http://www.ebi.ac.uk/efo/EFO_0004471 insulin sensitivity test\n",
      "\ttumor necrosis factor-alpha measurement http://www.ebi.ac.uk/efo/EFO_0004684 \n",
      "\tc-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458 c-reactive protein level\n",
      "\testradiol measurement http://www.ebi.ac.uk/efo/EFO_0004697 oestrogen level|oestrogen levels|estradiol levels\n",
      "\tsystolic blood pressure http://www.ebi.ac.uk/efo/EFO_0006335 sysbp|systolic pressure\n",
      "\tdiastolic blood pressure http://www.ebi.ac.uk/efo/EFO_0006336 diastolic pressure|diabp\n",
      "\tleptin measurement http://www.ebi.ac.uk/efo/EFO_0005000 \n",
      "\thip circumference http://www.ebi.ac.uk/efo/EFO_0005093 \n",
      "\tenergy intake http://www.ebi.ac.uk/efo/EFO_0003939 caloric intake\n",
      "\tphysical activity http://www.ebi.ac.uk/efo/EFO_0003940 motor activity|activities\n",
      "\tbone density http://www.ebi.ac.uk/efo/EFO_0003923 density\n",
      "obesity-related traits 19255\n",
      "\tadiponectin measurement http://www.ebi.ac.uk/efo/EFO_0004502 adiponectin levels\n",
      "\thoma-ir http://www.ebi.ac.uk/efo/EFO_0004501 homeostasis model assessment insulin resistance\n",
      "\tccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749 monocyte chemotactic protein-1 measurement|chemokine ccl2 level|monocyte chemoattractant protein-1 measurement|chemokine ccl2 levels|monocyte chemotactic protein-1 levels|monocyte chemotactic protein-1 level\n",
      "\tthyroid stimulating hormone measurement http://www.ebi.ac.uk/efo/EFO_0004748 thyrotropin measurement|thyroid stimulating hormone level|thyroid stimulating hormone levels\n",
      "\tprotein measurement http://www.ebi.ac.uk/efo/EFO_0004747 protein levels\n",
      "\tblood urea nitrogen measurement http://www.ebi.ac.uk/efo/EFO_0004741 \n",
      "\tlean body mass http://www.ebi.ac.uk/efo/EFO_0004995 \n",
      "\tserum alanine aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004735 alanine transaminase measurement|alanine aminotransferase measurement\n",
      "\taspartate aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004736 aspartate transaminase measurement\n",
      "\thormone measurement http://www.ebi.ac.uk/efo/EFO_0004730 \n",
      "\tfat body mass http://www.ebi.ac.uk/efo/EFO_0005409 total body fat measurement\n",
      "\tadipose tissue measurement http://www.ebi.ac.uk/efo/EFO_0004764 \n",
      "\ticam-1 measurement http://www.ebi.ac.uk/efo/EFO_0004520 soluble icam1 measurement|soluble intercellular adhesion molecule-1|sicam-1\n",
      "\ttriglyceride measurement http://www.ebi.ac.uk/efo/EFO_0004530 triglyceride levels\n",
      "\twaist circumference http://www.ebi.ac.uk/efo/EFO_0004342 \n",
      "\tbody mass index http://www.ebi.ac.uk/efo/EFO_0004340 quetelet's index|bmi\n",
      "\tobesity http://www.ebi.ac.uk/efo/EFO_0001073 overweight and obesity|obesity (disorder)|obese|obesity nos|obese (finding)|adiposity|adiposis|obesity [ambiguous]|obesity\n",
      "\tbirth weight http://www.ebi.ac.uk/efo/EFO_0004344 birthweight\n",
      "\twaist-hip ratio http://www.ebi.ac.uk/efo/EFO_0004343 \n",
      "\theart rate http://www.ebi.ac.uk/efo/EFO_0004326 pulse rate|cardiac chronotropy\n",
      "\tbody height http://www.ebi.ac.uk/efo/EFO_0004339 height\n",
      "\tbody weight http://www.ebi.ac.uk/efo/EFO_0004338 weight\n",
      "\thomocysteine measurement http://www.ebi.ac.uk/efo/EFO_0004578 homocysteine levels\n",
      "\ttotal cholesterol measurement http://www.ebi.ac.uk/efo/EFO_0004574 cholesterol\n",
      "\tccl5 measurement http://www.ebi.ac.uk/efo/EFO_0005117 regulated on activation\n",
      "\tigfbp-1 measurement http://www.ebi.ac.uk/efo/EFO_0005118 \n",
      "\tantioxidant measurement http://www.ebi.ac.uk/efo/EFO_0005119 \n",
      "\thead circumference http://www.ebi.ac.uk/efo/EFO_0005114 \n",
      "\tmetabolic rate measurement http://www.ebi.ac.uk/efo/EFO_0005115 \n",
      "\turinary metabolite measurement http://www.ebi.ac.uk/efo/EFO_0005116 \n",
      "\tfatty acid measurement http://www.ebi.ac.uk/efo/EFO_0005110 \n",
      "\tfolic acid measurement http://www.ebi.ac.uk/efo/EFO_0005111 folate measurement|vitamin b9 measurement\n",
      "\tgestational age http://www.ebi.ac.uk/efo/EFO_0005112 \n",
      "\tbody composition measurement http://www.ebi.ac.uk/efo/EFO_0005106 \n",
      "\tarm span http://www.ebi.ac.uk/efo/EFO_0005108 \n",
      "\tenergy expenditure http://www.ebi.ac.uk/efo/EFO_0005109 \n",
      "\tamino acid measurement http://www.ebi.ac.uk/efo/EFO_0005134 \n",
      "\tc-peptide measurement http://www.ebi.ac.uk/efo/EFO_0005187 \n",
      "\tccl11 measurement http://www.ebi.ac.uk/efo/EFO_0005188 \n",
      "\trespiratory quotient http://www.ebi.ac.uk/efo/EFO_0005189 \n",
      "\turinary nitrogen measurement http://www.ebi.ac.uk/efo/EFO_0005190 \n",
      "\twaist height ratio http://www.ebi.ac.uk/efo/EFO_0005191 \n",
      "\ttestosterone measurement http://www.ebi.ac.uk/efo/EFO_0004908 testosterone levels|testosterone level\n",
      "\tinterleukin-6 measurement http://www.ebi.ac.uk/efo/EFO_0004810 il-6 levels|il-6 level|il-6 measurement\n",
      "\ttransforming growth factor beta measurement http://www.ebi.ac.uk/efo/EFO_0004818 \n",
      "\tigf-1 measurement http://www.ebi.ac.uk/efo/EFO_0004627 insulin-like growth factor 1 levels\n",
      "\tigfbp-3 measurement http://www.ebi.ac.uk/efo/EFO_0004626 insulin-like growth factor binding protein 3 levels\n",
      "\tvitamin b12 measurement http://www.ebi.ac.uk/efo/EFO_0004620 vitamin b12 levels\n",
      "\tsleep measurement http://www.ebi.ac.uk/efo/EFO_0004870 \n",
      "\thigh density lipoprotein cholesterol measurement http://www.ebi.ac.uk/efo/EFO_0004612 hdl measurement\n",
      "\tlow density lipoprotein cholesterol measurement http://www.ebi.ac.uk/efo/EFO_0004611 ldl measurement\n",
      "\tmaximal oxygen uptake measurement http://www.ebi.ac.uk/efo/EFO_0004887 maximal oxygen consumption|peak oxygen uptake|maximal oxygen uptake|maximal aerobic capacity|vo2 max\n",
      "\tcortisol secretion measurement http://www.ebi.ac.uk/efo/EFO_0004642 \n",
      "\tfasting blood insulin measurement http://www.ebi.ac.uk/efo/EFO_0004466 fasting blood insulin|fasting blood insulin level|fasting insulin-related traits\n",
      "\tfasting blood glucose measurement http://www.ebi.ac.uk/efo/EFO_0004465 fasting glucose-related traits|fasting blood glucose level|fasting plasma glucose\n",
      "\tinsulin sensitivity measurement http://www.ebi.ac.uk/efo/EFO_0004471 insulin sensitivity test\n",
      "\ttumor necrosis factor-alpha measurement http://www.ebi.ac.uk/efo/EFO_0004684 \n",
      "\tc-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458 c-reactive protein level\n",
      "\testradiol measurement http://www.ebi.ac.uk/efo/EFO_0004697 oestrogen level|oestrogen levels|estradiol levels\n",
      "\tsystolic blood pressure http://www.ebi.ac.uk/efo/EFO_0006335 sysbp|systolic pressure\n",
      "\tdiastolic blood pressure http://www.ebi.ac.uk/efo/EFO_0006336 diastolic pressure|diabp\n",
      "\tleptin measurement http://www.ebi.ac.uk/efo/EFO_0005000 \n",
      "\thip circumference http://www.ebi.ac.uk/efo/EFO_0005093 \n",
      "\tenergy intake http://www.ebi.ac.uk/efo/EFO_0003939 caloric intake\n",
      "\tphysical activity http://www.ebi.ac.uk/efo/EFO_0003940 motor activity|activities\n",
      "\tbone density http://www.ebi.ac.uk/efo/EFO_0003923 density\n",
      "Genome-Wide Association Scan Shows Genetic Variants in the  FTO  Gene Are Associated with Obesity-Related Traits. The obesity epidemic is responsible for a substantial economic burden in developed countries and is a major risk factor for type 2 diabetes and cardiovascular disease. The disease is the result not only of several environmental risk factors, but also of genetic predisposition. To take advantage of recent advances in gene-mapping technology, we executed a genome-wide association scan to identify genetic variants associated with obesity-related quantitative traits in the genetically isolated population of Sardinia. Initial analysis suggested that several SNPs in the  FTO  and  PFKP  genes were associated with increased BMI, hip circumference, and weight. Within the  FTO  gene, rs9930506 showed the strongest association with BMI ( p =  8.6 ×10 − 7 ), hip circumference ( p =  3.4 × 10 − 8 ), and weight ( p =  9.1 × 10 − 7 ). In Sardinia, homozygotes for the rare “G” allele of this SNP (minor allele frequency = 0.46) were 1.3 BMI units heavier than homozygotes for the common “A” allele. Within the PFKP gene, rs6602024 showed very strong association with BMI ( p =  4.9 × 10 − 6 ). Homozygotes for the rare “A” allele of this SNP (minor allele frequency = 0.12) were 1.8 BMI units heavier than homozygotes for the common “G” allele. To replicate our findings, we genotyped these two SNPs in the GenNet study. In European Americans ( N  = 1,496) and in Hispanic Americans ( N  = 839), we replicated significant association between rs9930506 in the FTO gene and BMI ( p -value for meta-analysis of European American and Hispanic American follow-up samples,  p =  0.001), weight ( p =  0.001), and hip circumference ( p =  0.0005). We did not replicate association between rs6602024 and obesity-related traits in the GenNet sample, although we found that in European Americans, Hispanic Americans, and African Americans, homozygotes for the rare “A” allele were, on average, 1.0–3.0 BMI units heavier than homozygotes for the more common “G” allele. In summary, we have completed a whole genome–association scan for three obesity-related quantitative traits and report that common genetic variants in the  FTO  gene are associated with substantial changes in BMI, hip circumference, and body weight. These changes could have a significant impact on the risk of obesity-related morbidity in the general population. Although twin and family studies have clearly shown that genes play a role in obesity, it has proven quite difficult to identify the specific genetic variants involved. Here, we take advantage of recent technical and methodological advances to examine the role of common genetic variants on several obesity-related traits. By examining >4,000 Sardinians, we show that a specific genetic variant, rs9930506, and other nearby variants on human Chromosome 16 are associated with body mass index, hip circumference, and total body weight. The variants overlap  FTO,  a gene with poorly understood function. Further studies of the region may implicate new biological pathways affecting susceptibility to obesity. We also show that the association is not restricted to Sardinia but is also seen in independent samples of European Americans and Hispanic Americans. This finding is particularly important because obesity is associated with increased risk of cardiovascular disease and diabetes.\n",
      "1 Obesity-Related Traits\n",
      "-1 type 2 diabetes\n",
      "-1 cardiovascular disease. | cardiovascular disease http://www.ebi.ac.uk/efo/EFO_0000319\n",
      "1 obesity | obesity http://www.ebi.ac.uk/efo/EFO_0001073\n",
      "-1 is\n",
      "-1 is\n",
      "-1 genetic predisposition.\n",
      "-1 The disease\n",
      "-1 is\n",
      "-1 obesity-related quantitative traits\n",
      "-1 advantage\n",
      "1 , hip circumference, | hip circumference http://www.ebi.ac.uk/efo/EFO_0005093\n",
      "-1 associated with increased\n",
      "1 BMI, | body mass index http://www.ebi.ac.uk/efo/EFO_0004340\n",
      "1 weight. | body weight http://www.ebi.ac.uk/efo/EFO_0004338\n",
      "1 ), hip circumference ( | hip circumference http://www.ebi.ac.uk/efo/EFO_0005093\n",
      "1 BMI ( | body mass index http://www.ebi.ac.uk/efo/EFO_0004340\n",
      "1 weight ( | body weight http://www.ebi.ac.uk/efo/EFO_0004338\n",
      "1 BMI | body mass index http://www.ebi.ac.uk/efo/EFO_0004340\n",
      "1 BMI ( | body mass index http://www.ebi.ac.uk/efo/EFO_0004340\n",
      "1 BMI | body mass index http://www.ebi.ac.uk/efo/EFO_0004340\n",
      "1 ), weight ( | body weight http://www.ebi.ac.uk/efo/EFO_0004338\n",
      "1 hip circumference ( | hip circumference http://www.ebi.ac.uk/efo/EFO_0005093\n",
      "1 BMI ( | body mass index http://www.ebi.ac.uk/efo/EFO_0004340\n",
      "-1 , and African Americans,\n",
      "-1 Americans, and African\n",
      "1 obesity-related traits\n",
      "1 BMI | body mass index http://www.ebi.ac.uk/efo/EFO_0004340\n",
      "1 , hip circumference, | hip circumference http://www.ebi.ac.uk/efo/EFO_0005093\n",
      "-1 obesity-related quantitative traits\n",
      "1 body weight. | body weight http://www.ebi.ac.uk/efo/EFO_0004338\n",
      "1 BMI, | body mass index http://www.ebi.ac.uk/efo/EFO_0004340\n",
      "1 obesity, | obesity http://www.ebi.ac.uk/efo/EFO_0001073\n",
      "1 obesity-related traits\n",
      "-1 advantage\n",
      "1 body mass index, | body mass index http://www.ebi.ac.uk/efo/EFO_0004340\n",
      "1 , hip circumference, | hip circumference http://www.ebi.ac.uk/efo/EFO_0005093\n",
      "1 body weight. | body weight http://www.ebi.ac.uk/efo/EFO_0004338\n",
      "-1 susceptibility to obesity.\n",
      "-1 is\n",
      "-1 is\n",
      "-1 disease and diabetes.\n",
      "-1 associated with increased\n",
      "-1 cardiovascular disease | cardiovascular disease http://www.ebi.ac.uk/efo/EFO_0000319\n",
      "-1 is\n",
      "1 obesity | obesity http://www.ebi.ac.uk/efo/EFO_0001073\n",
      "-1 is\n",
      "\n",
      "17684544\n",
      "crohn's disease 19532\n",
      "\tcrohn's disease http://www.ebi.ac.uk/efo/EFO_0000384 gastritis associated with crohn's disease|ileitis\n",
      "crohn's disease 19845\n",
      "\tcrohn's disease http://www.ebi.ac.uk/efo/EFO_0000384 gastritis associated with crohn's disease|ileitis\n",
      "crohn's disease 19846\n",
      "\tcrohn's disease http://www.ebi.ac.uk/efo/EFO_0000384 gastritis associated with crohn's disease|ileitis\n",
      "Systematic Association Mapping Identifies  NELL1  as a Novel IBD Disease Gene. Crohn disease (CD), a sub-entity of inflammatory bowel disease (IBD), is a complex polygenic disorder. Although recent studies have successfully identified CD-associated genetic variants, these susceptibility loci explain only a fraction of the heritability of the disease. Here, we report on a multi-stage genome-wide scan of 393 German CD cases and 399 controls. Among the 116,161 single-nucleotide polymorphisms tested, an association with the known CD susceptibility gene  NOD2 , the  5q31  haplotype, and the recently reported CD locus at  5p13.1  was confirmed. In addition, SNP rs1793004 in the gene encoding nel-like 1 precursor ( NELL1,  chromosome  11p15.1 ) showed a consistent disease-association in independent German population- and family-based samples (942 cases, 1082 controls, 375 trios). Subsequent fine mapping and replication in an independent sample of 454 French/Canadian CD trios supported the authenticity of the  NELL1  association. Further confirmation in a large German ulcerative colitis (UC) sample indicated that  NELL1  is a ubiquitous IBD susceptibility locus (combined p<10 −6 ; OR = 1.66, 95% CI: 1.30–2.11). The novel  5p13.1  locus was also replicated in the French/Canadian sample and in an independent UK CD patient panel (453 cases, 521 controls, combined p<10 −6  for SNP rs1992660). Several associations were replicated in at least one independent sample, point to an involvement of  ITGB6  (upstream),  GRM8  (downstream),  OR5V1  (downstream),  PPP3R2  (downstream),  NM_152575  (upstream) and  HNF4G  (intron).\n",
      "-1 IBD Disease\n",
      "-1 Mapping | mean arterial pressure http://www.ebi.ac.uk/efo/EFO_0006340\n",
      "-1 as\n",
      "-1 (CD),\n",
      "-1 inflammatory bowel disease (\n",
      "1 Crohn disease ( | crohn's disease http://www.ebi.ac.uk/efo/EFO_0000384\n",
      "-1 ), is\n",
      "-1 disorder. | autism http://www.ebi.ac.uk/efo/EFO_0003758\n",
      "-1 the disease\n",
      "-1 disease. | systemic lupus erythematosus http://www.ebi.ac.uk/efo/EFO_0002690\n",
      "-1 CD\n",
      "-1 tested, | pulmonary function measurement http://www.ebi.ac.uk/efo/EFO_0003892\n",
      "-1 , an\n",
      "-1 with the\n",
      "-1 CD\n",
      "-1 CD\n",
      "-1 .1\n",
      "-1 was\n",
      "-1 .1 )\n",
      "-1 1\n",
      "-1 mapping | mean arterial pressure http://www.ebi.ac.uk/efo/EFO_0006340\n",
      "-1 an\n",
      "-1 CD\n",
      "-1 ulcerative colitis ( | ulcerative colitis http://www.ebi.ac.uk/efo/EFO_0000729\n",
      "-1 indicated | ankle brachial index http://www.ebi.ac.uk/efo/EFO_0003912\n",
      "-1 is\n",
      "-1 .1\n",
      "-1 was\n",
      "-1 an\n",
      "-1 CD\n",
      "-1 involvement of\n",
      "-1 an\n",
      "\n",
      "17903292\n",
      "cystatin c 19169\n",
      "\tcystatin c measurement http://www.ebi.ac.uk/efo/EFO_0004617 \n",
      "thyroid stimulating hormone 19170\n",
      "\tthyroid stimulating hormone measurement http://www.ebi.ac.uk/efo/EFO_0004748 thyrotropin measurement|thyroid stimulating hormone level|thyroid stimulating hormone levels\n",
      "\tthyroid function http://www.ebi.ac.uk/efo/EFO_0004296 \n",
      "thyroid stimulating hormone 19171\n",
      "\tthyroid stimulating hormone measurement http://www.ebi.ac.uk/efo/EFO_0004748 thyrotropin measurement|thyroid stimulating hormone level|thyroid stimulating hormone levels\n",
      "\tthyroid function http://www.ebi.ac.uk/efo/EFO_0004296 \n",
      "thyroid stimulating hormone 19172\n",
      "\tthyroid stimulating hormone measurement http://www.ebi.ac.uk/efo/EFO_0004748 thyrotropin measurement|thyroid stimulating hormone level|thyroid stimulating hormone levels\n",
      "\tthyroid function http://www.ebi.ac.uk/efo/EFO_0004296 \n",
      "urinary albumin excretion 19173\n",
      "\talbuminuria http://www.ebi.ac.uk/efo/EFO_0004285 urinary albumin excretion\n",
      "urinary albumin excretion 19626\n",
      "\talbuminuria http://www.ebi.ac.uk/efo/EFO_0004285 urinary albumin excretion\n",
      "A genome-wide association for kidney function and endocrine-related traits in the NHLBI's Framingham Heart Study. Glomerular filtration rate (GFR) and urinary albumin excretion (UAE) are markers of kidney function that are known to be heritable. Many endocrine conditions have strong familial components. We tested for association between the Affymetrix GeneChip Human Mapping 100K single nucleotide polymorphism (SNP) set and measures of kidney function and endocrine traits. Genotype information on the Affymetrix GeneChip Human Mapping 100K SNP set was available on 1345 participants. Serum creatinine and cystatin-C (cysC; n = 981) were measured at the seventh examination cycle (1998–2001); GFR (n = 1010) was estimated via the Modification of Diet in Renal Disease (MDRD) equation; UAE was measured on spot urine samples during the sixth examination cycle (1995–1998) and was indexed to urinary creatinine (n = 822). Thyroid stimulating hormone (TSH) was measured at the third and fourth examination cycles (1981–1984; 1984–1987) and mean value of the measurements were used (n = 810). Age-sex-adjusted and multivariable-adjusted residuals for these measurements were used in association with genotype data using generalized estimating equations (GEE) and family-based association tests (FBAT) models. We presented the results for association tests using additive allele model. We evaluated associations with 70,987 SNPs on autosomes with minor allele frequencies of at least 0.10, Hardy-Weinberg Equilibrium p-value ≥ 0.001, and call rates of at least 80%. The top SNPs associated with these traits using the GEE method were rs2839235 with GFR (p-value 1.6*10 -05 ), rs1158167 with cysC (p-value 8.5*10 -09 ), rs1712790 with UAE (p-value 1.9*10 -06 ), and rs6977660 with TSH (p-value 3.7*10 -06 ), respectively. The top SNPs associated with these traits using the FBAT method were rs6434804 with GFR(p-value 2.4*10 -5 ), rs563754 with cysC (p-value 4.7*10 -5 ), rs1243400 with UAE (p-value 4.8*10 -6 ), and rs4128956 with TSH (p-value 3.6*10 -5 ), respectively. Detailed association test results can be found at  . Four SNPs in or near the  CST 3 gene were highly associated with cysC levels (p-value 8.5*10 -09  to 0.007). Kidney function traits and TSH are associated with SNPs on the Affymetrix GeneChip Human Mapping 100K SNP set. These data will serve as a valuable resource for replication as more SNPs associated with kidney function and endocrine traits are identified.\n",
      "-1 endocrine-related traits\n",
      "-1 Glomerular filtration rate (\n",
      "1 urinary albumin excretion ( | albuminuria http://www.ebi.ac.uk/efo/EFO_0004285\n",
      "-1 (GFR)\n",
      "-1 of kidney\n",
      "-1 of kidney\n",
      "-1 endocrine traits\n",
      "-1 tested | pulmonary function measurement http://www.ebi.ac.uk/efo/EFO_0003892\n",
      "-1 Mapping | mean arterial pressure http://www.ebi.ac.uk/efo/EFO_0006340\n",
      "-1 single\n",
      "-1 Mapping | mean arterial pressure http://www.ebi.ac.uk/efo/EFO_0006340\n",
      "-1 was\n",
      "-1 ); GFR (\n",
      "-1 Renal Disease (\n",
      "-1 Serum creatinine\n",
      "1 cystatin-C ( | cystatin c measurement http://www.ebi.ac.uk/efo/EFO_0004617\n",
      "-1 ) was\n",
      "-1 creatinine (\n",
      "-1 was\n",
      "-1 spot\n",
      "-1 was\n",
      "1 Thyroid stimulating hormone ( | thyroid stimulating hormone measurement http://www.ebi.ac.uk/efo/EFO_0004748\n",
      "-1 third and fourth\n",
      "-1 ) was\n",
      "-1 the measurements\n",
      "-1 these measurements\n",
      "-1 tests ( | pulmonary function measurement http://www.ebi.ac.uk/efo/EFO_0003892\n",
      "-1 tests | pulmonary function measurement http://www.ebi.ac.uk/efo/EFO_0003892\n",
      "-1 these traits\n",
      "-1 GFR (\n",
      "-1 these traits\n",
      "-1 GFR(\n",
      "-1 can\n",
      "-1 function traits\n",
      "-1 Mapping | mean arterial pressure http://www.ebi.ac.uk/efo/EFO_0006340\n",
      "-1 endocrine traits\n",
      "-1 as\n",
      "-1 as\n",
      "\n",
      "17903293\n",
      "select biomarker traits 19638\n",
      "\tccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749 monocyte chemotactic protein-1 measurement|chemokine ccl2 level|monocyte chemoattractant protein-1 measurement|chemokine ccl2 levels|monocyte chemotactic protein-1 levels|monocyte chemotactic protein-1 level\n",
      "\tserum alanine aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004735 alanine transaminase measurement|alanine aminotransferase measurement\n",
      "\tatrial natriuretic factor measurement http://www.ebi.ac.uk/efo/EFO_0004789 \n",
      "\tcd40 ligand measurement http://www.ebi.ac.uk/efo/EFO_0004790 cd154 ligand measurement\n",
      "\talkaline phosphatase measurement http://www.ebi.ac.uk/efo/EFO_0004533 ap\n",
      "\tbilirubin measurement http://www.ebi.ac.uk/efo/EFO_0004570 bilirubin levels\n",
      "\tvitamin d measurement http://www.ebi.ac.uk/efo/EFO_0004631 \n",
      "\tvitamin k measurement http://www.ebi.ac.uk/efo/EFO_0004618 phylloquinone measurement\n",
      "\ttumor necrosis factor-alpha measurement http://www.ebi.ac.uk/efo/EFO_0004684 \n",
      "\tc-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458 c-reactive protein level\n",
      "select biomarker traits 19627\n",
      "\tccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749 monocyte chemotactic protein-1 measurement|chemokine ccl2 level|monocyte chemoattractant protein-1 measurement|chemokine ccl2 levels|monocyte chemotactic protein-1 levels|monocyte chemotactic protein-1 level\n",
      "\tserum alanine aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004735 alanine transaminase measurement|alanine aminotransferase measurement\n",
      "\tatrial natriuretic factor measurement http://www.ebi.ac.uk/efo/EFO_0004789 \n",
      "\tcd40 ligand measurement http://www.ebi.ac.uk/efo/EFO_0004790 cd154 ligand measurement\n",
      "\talkaline phosphatase measurement http://www.ebi.ac.uk/efo/EFO_0004533 ap\n",
      "\tbilirubin measurement http://www.ebi.ac.uk/efo/EFO_0004570 bilirubin levels\n",
      "\tvitamin d measurement http://www.ebi.ac.uk/efo/EFO_0004631 \n",
      "\tvitamin k measurement http://www.ebi.ac.uk/efo/EFO_0004618 phylloquinone measurement\n",
      "\ttumor necrosis factor-alpha measurement http://www.ebi.ac.uk/efo/EFO_0004684 \n",
      "\tc-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458 c-reactive protein level\n",
      "select biomarker traits 19628\n",
      "\tccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749 monocyte chemotactic protein-1 measurement|chemokine ccl2 level|monocyte chemoattractant protein-1 measurement|chemokine ccl2 levels|monocyte chemotactic protein-1 levels|monocyte chemotactic protein-1 level\n",
      "\tserum alanine aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004735 alanine transaminase measurement|alanine aminotransferase measurement\n",
      "\tatrial natriuretic factor measurement http://www.ebi.ac.uk/efo/EFO_0004789 \n",
      "\tcd40 ligand measurement http://www.ebi.ac.uk/efo/EFO_0004790 cd154 ligand measurement\n",
      "\talkaline phosphatase measurement http://www.ebi.ac.uk/efo/EFO_0004533 ap\n",
      "\tbilirubin measurement http://www.ebi.ac.uk/efo/EFO_0004570 bilirubin levels\n",
      "\tvitamin d measurement http://www.ebi.ac.uk/efo/EFO_0004631 \n",
      "\tvitamin k measurement http://www.ebi.ac.uk/efo/EFO_0004618 phylloquinone measurement\n",
      "\ttumor necrosis factor-alpha measurement http://www.ebi.ac.uk/efo/EFO_0004684 \n",
      "\tc-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458 c-reactive protein level\n",
      "select biomarker traits 19629\n",
      "\tccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749 monocyte chemotactic protein-1 measurement|chemokine ccl2 level|monocyte chemoattractant protein-1 measurement|chemokine ccl2 levels|monocyte chemotactic protein-1 levels|monocyte chemotactic protein-1 level\n",
      "\tserum alanine aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004735 alanine transaminase measurement|alanine aminotransferase measurement\n",
      "\tatrial natriuretic factor measurement http://www.ebi.ac.uk/efo/EFO_0004789 \n",
      "\tcd40 ligand measurement http://www.ebi.ac.uk/efo/EFO_0004790 cd154 ligand measurement\n",
      "\talkaline phosphatase measurement http://www.ebi.ac.uk/efo/EFO_0004533 ap\n",
      "\tbilirubin measurement http://www.ebi.ac.uk/efo/EFO_0004570 bilirubin levels\n",
      "\tvitamin d measurement http://www.ebi.ac.uk/efo/EFO_0004631 \n",
      "\tvitamin k measurement http://www.ebi.ac.uk/efo/EFO_0004618 phylloquinone measurement\n",
      "\ttumor necrosis factor-alpha measurement http://www.ebi.ac.uk/efo/EFO_0004684 \n",
      "\tc-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458 c-reactive protein level\n",
      "select biomarker traits 19630\n",
      "\tccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749 monocyte chemotactic protein-1 measurement|chemokine ccl2 level|monocyte chemoattractant protein-1 measurement|chemokine ccl2 levels|monocyte chemotactic protein-1 levels|monocyte chemotactic protein-1 level\n",
      "\tserum alanine aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004735 alanine transaminase measurement|alanine aminotransferase measurement\n",
      "\tatrial natriuretic factor measurement http://www.ebi.ac.uk/efo/EFO_0004789 \n",
      "\tcd40 ligand measurement http://www.ebi.ac.uk/efo/EFO_0004790 cd154 ligand measurement\n",
      "\talkaline phosphatase measurement http://www.ebi.ac.uk/efo/EFO_0004533 ap\n",
      "\tbilirubin measurement http://www.ebi.ac.uk/efo/EFO_0004570 bilirubin levels\n",
      "\tvitamin d measurement http://www.ebi.ac.uk/efo/EFO_0004631 \n",
      "\tvitamin k measurement http://www.ebi.ac.uk/efo/EFO_0004618 phylloquinone measurement\n",
      "\ttumor necrosis factor-alpha measurement http://www.ebi.ac.uk/efo/EFO_0004684 \n",
      "\tc-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458 c-reactive protein level\n",
      "select biomarker traits 19631\n",
      "\tccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749 monocyte chemotactic protein-1 measurement|chemokine ccl2 level|monocyte chemoattractant protein-1 measurement|chemokine ccl2 levels|monocyte chemotactic protein-1 levels|monocyte chemotactic protein-1 level\n",
      "\tserum alanine aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004735 alanine transaminase measurement|alanine aminotransferase measurement\n",
      "\tatrial natriuretic factor measurement http://www.ebi.ac.uk/efo/EFO_0004789 \n",
      "\tcd40 ligand measurement http://www.ebi.ac.uk/efo/EFO_0004790 cd154 ligand measurement\n",
      "\talkaline phosphatase measurement http://www.ebi.ac.uk/efo/EFO_0004533 ap\n",
      "\tbilirubin measurement http://www.ebi.ac.uk/efo/EFO_0004570 bilirubin levels\n",
      "\tvitamin d measurement http://www.ebi.ac.uk/efo/EFO_0004631 \n",
      "\tvitamin k measurement http://www.ebi.ac.uk/efo/EFO_0004618 phylloquinone measurement\n",
      "\ttumor necrosis factor-alpha measurement http://www.ebi.ac.uk/efo/EFO_0004684 \n",
      "\tc-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458 c-reactive protein level\n",
      "select biomarker traits 19632\n",
      "\tccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749 monocyte chemotactic protein-1 measurement|chemokine ccl2 level|monocyte chemoattractant protein-1 measurement|chemokine ccl2 levels|monocyte chemotactic protein-1 levels|monocyte chemotactic protein-1 level\n",
      "\tserum alanine aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004735 alanine transaminase measurement|alanine aminotransferase measurement\n",
      "\tatrial natriuretic factor measurement http://www.ebi.ac.uk/efo/EFO_0004789 \n",
      "\tcd40 ligand measurement http://www.ebi.ac.uk/efo/EFO_0004790 cd154 ligand measurement\n",
      "\talkaline phosphatase measurement http://www.ebi.ac.uk/efo/EFO_0004533 ap\n",
      "\tbilirubin measurement http://www.ebi.ac.uk/efo/EFO_0004570 bilirubin levels\n",
      "\tvitamin d measurement http://www.ebi.ac.uk/efo/EFO_0004631 \n",
      "\tvitamin k measurement http://www.ebi.ac.uk/efo/EFO_0004618 phylloquinone measurement\n",
      "\ttumor necrosis factor-alpha measurement http://www.ebi.ac.uk/efo/EFO_0004684 \n",
      "\tc-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458 c-reactive protein level\n",
      "select biomarker traits 19633\n",
      "\tccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749 monocyte chemotactic protein-1 measurement|chemokine ccl2 level|monocyte chemoattractant protein-1 measurement|chemokine ccl2 levels|monocyte chemotactic protein-1 levels|monocyte chemotactic protein-1 level\n",
      "\tserum alanine aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004735 alanine transaminase measurement|alanine aminotransferase measurement\n",
      "\tatrial natriuretic factor measurement http://www.ebi.ac.uk/efo/EFO_0004789 \n",
      "\tcd40 ligand measurement http://www.ebi.ac.uk/efo/EFO_0004790 cd154 ligand measurement\n",
      "\talkaline phosphatase measurement http://www.ebi.ac.uk/efo/EFO_0004533 ap\n",
      "\tbilirubin measurement http://www.ebi.ac.uk/efo/EFO_0004570 bilirubin levels\n",
      "\tvitamin d measurement http://www.ebi.ac.uk/efo/EFO_0004631 \n",
      "\tvitamin k measurement http://www.ebi.ac.uk/efo/EFO_0004618 phylloquinone measurement\n",
      "\ttumor necrosis factor-alpha measurement http://www.ebi.ac.uk/efo/EFO_0004684 \n",
      "\tc-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458 c-reactive protein level\n",
      "select biomarker traits 19634\n",
      "\tccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749 monocyte chemotactic protein-1 measurement|chemokine ccl2 level|monocyte chemoattractant protein-1 measurement|chemokine ccl2 levels|monocyte chemotactic protein-1 levels|monocyte chemotactic protein-1 level\n",
      "\tserum alanine aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004735 alanine transaminase measurement|alanine aminotransferase measurement\n",
      "\tatrial natriuretic factor measurement http://www.ebi.ac.uk/efo/EFO_0004789 \n",
      "\tcd40 ligand measurement http://www.ebi.ac.uk/efo/EFO_0004790 cd154 ligand measurement\n",
      "\talkaline phosphatase measurement http://www.ebi.ac.uk/efo/EFO_0004533 ap\n",
      "\tbilirubin measurement http://www.ebi.ac.uk/efo/EFO_0004570 bilirubin levels\n",
      "\tvitamin d measurement http://www.ebi.ac.uk/efo/EFO_0004631 \n",
      "\tvitamin k measurement http://www.ebi.ac.uk/efo/EFO_0004618 phylloquinone measurement\n",
      "\ttumor necrosis factor-alpha measurement http://www.ebi.ac.uk/efo/EFO_0004684 \n",
      "\tc-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458 c-reactive protein level\n",
      "select biomarker traits 19635\n",
      "\tccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749 monocyte chemotactic protein-1 measurement|chemokine ccl2 level|monocyte chemoattractant protein-1 measurement|chemokine ccl2 levels|monocyte chemotactic protein-1 levels|monocyte chemotactic protein-1 level\n",
      "\tserum alanine aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004735 alanine transaminase measurement|alanine aminotransferase measurement\n",
      "\tatrial natriuretic factor measurement http://www.ebi.ac.uk/efo/EFO_0004789 \n",
      "\tcd40 ligand measurement http://www.ebi.ac.uk/efo/EFO_0004790 cd154 ligand measurement\n",
      "\talkaline phosphatase measurement http://www.ebi.ac.uk/efo/EFO_0004533 ap\n",
      "\tbilirubin measurement http://www.ebi.ac.uk/efo/EFO_0004570 bilirubin levels\n",
      "\tvitamin d measurement http://www.ebi.ac.uk/efo/EFO_0004631 \n",
      "\tvitamin k measurement http://www.ebi.ac.uk/efo/EFO_0004618 phylloquinone measurement\n",
      "\ttumor necrosis factor-alpha measurement http://www.ebi.ac.uk/efo/EFO_0004684 \n",
      "\tc-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458 c-reactive protein level\n",
      "select biomarker traits 19636\n",
      "\tccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749 monocyte chemotactic protein-1 measurement|chemokine ccl2 level|monocyte chemoattractant protein-1 measurement|chemokine ccl2 levels|monocyte chemotactic protein-1 levels|monocyte chemotactic protein-1 level\n",
      "\tserum alanine aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004735 alanine transaminase measurement|alanine aminotransferase measurement\n",
      "\tatrial natriuretic factor measurement http://www.ebi.ac.uk/efo/EFO_0004789 \n",
      "\tcd40 ligand measurement http://www.ebi.ac.uk/efo/EFO_0004790 cd154 ligand measurement\n",
      "\talkaline phosphatase measurement http://www.ebi.ac.uk/efo/EFO_0004533 ap\n",
      "\tbilirubin measurement http://www.ebi.ac.uk/efo/EFO_0004570 bilirubin levels\n",
      "\tvitamin d measurement http://www.ebi.ac.uk/efo/EFO_0004631 \n",
      "\tvitamin k measurement http://www.ebi.ac.uk/efo/EFO_0004618 phylloquinone measurement\n",
      "\ttumor necrosis factor-alpha measurement http://www.ebi.ac.uk/efo/EFO_0004684 \n",
      "\tc-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458 c-reactive protein level\n",
      "select biomarker traits 19637\n",
      "\tccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749 monocyte chemotactic protein-1 measurement|chemokine ccl2 level|monocyte chemoattractant protein-1 measurement|chemokine ccl2 levels|monocyte chemotactic protein-1 levels|monocyte chemotactic protein-1 level\n",
      "\tserum alanine aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004735 alanine transaminase measurement|alanine aminotransferase measurement\n",
      "\tatrial natriuretic factor measurement http://www.ebi.ac.uk/efo/EFO_0004789 \n",
      "\tcd40 ligand measurement http://www.ebi.ac.uk/efo/EFO_0004790 cd154 ligand measurement\n",
      "\talkaline phosphatase measurement http://www.ebi.ac.uk/efo/EFO_0004533 ap\n",
      "\tbilirubin measurement http://www.ebi.ac.uk/efo/EFO_0004570 bilirubin levels\n",
      "\tvitamin d measurement http://www.ebi.ac.uk/efo/EFO_0004631 \n",
      "\tvitamin k measurement http://www.ebi.ac.uk/efo/EFO_0004618 phylloquinone measurement\n",
      "\ttumor necrosis factor-alpha measurement http://www.ebi.ac.uk/efo/EFO_0004684 \n",
      "\tc-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458 c-reactive protein level\n",
      "select biomarker traits 19158\n",
      "\tccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749 monocyte chemotactic protein-1 measurement|chemokine ccl2 level|monocyte chemoattractant protein-1 measurement|chemokine ccl2 levels|monocyte chemotactic protein-1 levels|monocyte chemotactic protein-1 level\n",
      "\tserum alanine aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004735 alanine transaminase measurement|alanine aminotransferase measurement\n",
      "\tatrial natriuretic factor measurement http://www.ebi.ac.uk/efo/EFO_0004789 \n",
      "\tcd40 ligand measurement http://www.ebi.ac.uk/efo/EFO_0004790 cd154 ligand measurement\n",
      "\talkaline phosphatase measurement http://www.ebi.ac.uk/efo/EFO_0004533 ap\n",
      "\tbilirubin measurement http://www.ebi.ac.uk/efo/EFO_0004570 bilirubin levels\n",
      "\tvitamin d measurement http://www.ebi.ac.uk/efo/EFO_0004631 \n",
      "\tvitamin k measurement http://www.ebi.ac.uk/efo/EFO_0004618 phylloquinone measurement\n",
      "\ttumor necrosis factor-alpha measurement http://www.ebi.ac.uk/efo/EFO_0004684 \n",
      "\tc-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458 c-reactive protein level\n",
      "select biomarker traits 19639\n",
      "\tccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749 monocyte chemotactic protein-1 measurement|chemokine ccl2 level|monocyte chemoattractant protein-1 measurement|chemokine ccl2 levels|monocyte chemotactic protein-1 levels|monocyte chemotactic protein-1 level\n",
      "\tserum alanine aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004735 alanine transaminase measurement|alanine aminotransferase measurement\n",
      "\tatrial natriuretic factor measurement http://www.ebi.ac.uk/efo/EFO_0004789 \n",
      "\tcd40 ligand measurement http://www.ebi.ac.uk/efo/EFO_0004790 cd154 ligand measurement\n",
      "\talkaline phosphatase measurement http://www.ebi.ac.uk/efo/EFO_0004533 ap\n",
      "\tbilirubin measurement http://www.ebi.ac.uk/efo/EFO_0004570 bilirubin levels\n",
      "\tvitamin d measurement http://www.ebi.ac.uk/efo/EFO_0004631 \n",
      "\tvitamin k measurement http://www.ebi.ac.uk/efo/EFO_0004618 phylloquinone measurement\n",
      "\ttumor necrosis factor-alpha measurement http://www.ebi.ac.uk/efo/EFO_0004684 \n",
      "\tc-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458 c-reactive protein level\n",
      "select biomarker traits 19640\n",
      "\tccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749 monocyte chemotactic protein-1 measurement|chemokine ccl2 level|monocyte chemoattractant protein-1 measurement|chemokine ccl2 levels|monocyte chemotactic protein-1 levels|monocyte chemotactic protein-1 level\n",
      "\tserum alanine aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004735 alanine transaminase measurement|alanine aminotransferase measurement\n",
      "\tatrial natriuretic factor measurement http://www.ebi.ac.uk/efo/EFO_0004789 \n",
      "\tcd40 ligand measurement http://www.ebi.ac.uk/efo/EFO_0004790 cd154 ligand measurement\n",
      "\talkaline phosphatase measurement http://www.ebi.ac.uk/efo/EFO_0004533 ap\n",
      "\tbilirubin measurement http://www.ebi.ac.uk/efo/EFO_0004570 bilirubin levels\n",
      "\tvitamin d measurement http://www.ebi.ac.uk/efo/EFO_0004631 \n",
      "\tvitamin k measurement http://www.ebi.ac.uk/efo/EFO_0004618 phylloquinone measurement\n",
      "\ttumor necrosis factor-alpha measurement http://www.ebi.ac.uk/efo/EFO_0004684 \n",
      "\tc-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458 c-reactive protein level\n",
      "select biomarker traits 19161\n",
      "\tccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749 monocyte chemotactic protein-1 measurement|chemokine ccl2 level|monocyte chemoattractant protein-1 measurement|chemokine ccl2 levels|monocyte chemotactic protein-1 levels|monocyte chemotactic protein-1 level\n",
      "\tserum alanine aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004735 alanine transaminase measurement|alanine aminotransferase measurement\n",
      "\tatrial natriuretic factor measurement http://www.ebi.ac.uk/efo/EFO_0004789 \n",
      "\tcd40 ligand measurement http://www.ebi.ac.uk/efo/EFO_0004790 cd154 ligand measurement\n",
      "\talkaline phosphatase measurement http://www.ebi.ac.uk/efo/EFO_0004533 ap\n",
      "\tbilirubin measurement http://www.ebi.ac.uk/efo/EFO_0004570 bilirubin levels\n",
      "\tvitamin d measurement http://www.ebi.ac.uk/efo/EFO_0004631 \n",
      "\tvitamin k measurement http://www.ebi.ac.uk/efo/EFO_0004618 phylloquinone measurement\n",
      "\ttumor necrosis factor-alpha measurement http://www.ebi.ac.uk/efo/EFO_0004684 \n",
      "\tc-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458 c-reactive protein level\n",
      "select biomarker traits 19642\n",
      "\tccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749 monocyte chemotactic protein-1 measurement|chemokine ccl2 level|monocyte chemoattractant protein-1 measurement|chemokine ccl2 levels|monocyte chemotactic protein-1 levels|monocyte chemotactic protein-1 level\n",
      "\tserum alanine aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004735 alanine transaminase measurement|alanine aminotransferase measurement\n",
      "\tatrial natriuretic factor measurement http://www.ebi.ac.uk/efo/EFO_0004789 \n",
      "\tcd40 ligand measurement http://www.ebi.ac.uk/efo/EFO_0004790 cd154 ligand measurement\n",
      "\talkaline phosphatase measurement http://www.ebi.ac.uk/efo/EFO_0004533 ap\n",
      "\tbilirubin measurement http://www.ebi.ac.uk/efo/EFO_0004570 bilirubin levels\n",
      "\tvitamin d measurement http://www.ebi.ac.uk/efo/EFO_0004631 \n",
      "\tvitamin k measurement http://www.ebi.ac.uk/efo/EFO_0004618 phylloquinone measurement\n",
      "\ttumor necrosis factor-alpha measurement http://www.ebi.ac.uk/efo/EFO_0004684 \n",
      "\tc-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458 c-reactive protein level\n",
      "select biomarker traits 19641\n",
      "\tccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749 monocyte chemotactic protein-1 measurement|chemokine ccl2 level|monocyte chemoattractant protein-1 measurement|chemokine ccl2 levels|monocyte chemotactic protein-1 levels|monocyte chemotactic protein-1 level\n",
      "\tserum alanine aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004735 alanine transaminase measurement|alanine aminotransferase measurement\n",
      "\tatrial natriuretic factor measurement http://www.ebi.ac.uk/efo/EFO_0004789 \n",
      "\tcd40 ligand measurement http://www.ebi.ac.uk/efo/EFO_0004790 cd154 ligand measurement\n",
      "\talkaline phosphatase measurement http://www.ebi.ac.uk/efo/EFO_0004533 ap\n",
      "\tbilirubin measurement http://www.ebi.ac.uk/efo/EFO_0004570 bilirubin levels\n",
      "\tvitamin d measurement http://www.ebi.ac.uk/efo/EFO_0004631 \n",
      "\tvitamin k measurement http://www.ebi.ac.uk/efo/EFO_0004618 phylloquinone measurement\n",
      "\ttumor necrosis factor-alpha measurement http://www.ebi.ac.uk/efo/EFO_0004684 \n",
      "\tc-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458 c-reactive protein level\n",
      "select biomarker traits 19160\n",
      "\tccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749 monocyte chemotactic protein-1 measurement|chemokine ccl2 level|monocyte chemoattractant protein-1 measurement|chemokine ccl2 levels|monocyte chemotactic protein-1 levels|monocyte chemotactic protein-1 level\n",
      "\tserum alanine aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004735 alanine transaminase measurement|alanine aminotransferase measurement\n",
      "\tatrial natriuretic factor measurement http://www.ebi.ac.uk/efo/EFO_0004789 \n",
      "\tcd40 ligand measurement http://www.ebi.ac.uk/efo/EFO_0004790 cd154 ligand measurement\n",
      "\talkaline phosphatase measurement http://www.ebi.ac.uk/efo/EFO_0004533 ap\n",
      "\tbilirubin measurement http://www.ebi.ac.uk/efo/EFO_0004570 bilirubin levels\n",
      "\tvitamin d measurement http://www.ebi.ac.uk/efo/EFO_0004631 \n",
      "\tvitamin k measurement http://www.ebi.ac.uk/efo/EFO_0004618 phylloquinone measurement\n",
      "\ttumor necrosis factor-alpha measurement http://www.ebi.ac.uk/efo/EFO_0004684 \n",
      "\tc-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458 c-reactive protein level\n",
      "select biomarker traits 19159\n",
      "\tccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749 monocyte chemotactic protein-1 measurement|chemokine ccl2 level|monocyte chemoattractant protein-1 measurement|chemokine ccl2 levels|monocyte chemotactic protein-1 levels|monocyte chemotactic protein-1 level\n",
      "\tserum alanine aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004735 alanine transaminase measurement|alanine aminotransferase measurement\n",
      "\tatrial natriuretic factor measurement http://www.ebi.ac.uk/efo/EFO_0004789 \n",
      "\tcd40 ligand measurement http://www.ebi.ac.uk/efo/EFO_0004790 cd154 ligand measurement\n",
      "\talkaline phosphatase measurement http://www.ebi.ac.uk/efo/EFO_0004533 ap\n",
      "\tbilirubin measurement http://www.ebi.ac.uk/efo/EFO_0004570 bilirubin levels\n",
      "\tvitamin d measurement http://www.ebi.ac.uk/efo/EFO_0004631 \n",
      "\tvitamin k measurement http://www.ebi.ac.uk/efo/EFO_0004618 phylloquinone measurement\n",
      "\ttumor necrosis factor-alpha measurement http://www.ebi.ac.uk/efo/EFO_0004684 \n",
      "\tc-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458 c-reactive protein level\n",
      "select biomarker traits 19157\n",
      "\tccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749 monocyte chemotactic protein-1 measurement|chemokine ccl2 level|monocyte chemoattractant protein-1 measurement|chemokine ccl2 levels|monocyte chemotactic protein-1 levels|monocyte chemotactic protein-1 level\n",
      "\tserum alanine aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004735 alanine transaminase measurement|alanine aminotransferase measurement\n",
      "\tatrial natriuretic factor measurement http://www.ebi.ac.uk/efo/EFO_0004789 \n",
      "\tcd40 ligand measurement http://www.ebi.ac.uk/efo/EFO_0004790 cd154 ligand measurement\n",
      "\talkaline phosphatase measurement http://www.ebi.ac.uk/efo/EFO_0004533 ap\n",
      "\tbilirubin measurement http://www.ebi.ac.uk/efo/EFO_0004570 bilirubin levels\n",
      "\tvitamin d measurement http://www.ebi.ac.uk/efo/EFO_0004631 \n",
      "\tvitamin k measurement http://www.ebi.ac.uk/efo/EFO_0004618 phylloquinone measurement\n",
      "\ttumor necrosis factor-alpha measurement http://www.ebi.ac.uk/efo/EFO_0004684 \n",
      "\tc-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458 c-reactive protein level\n",
      "Genome-wide association with select biomarker traits in the Framingham Heart Study. Systemic biomarkers provide insights into disease pathogenesis, diagnosis, and risk stratification. Many systemic biomarker concentrations are heritable phenotypes. Genome-wide association studies (GWAS) provide mechanisms to investigate the genetic contributions to biomarker variability unconstrained by current knowledge of physiological relations. We examined the association of Affymetrix 100K GeneChip single nucleotide polymorphisms (SNPs) to 22 systemic biomarker concentrations in 4 biological domains: inflammation/oxidative stress; natriuretic peptides; liver function; and vitamins. Related members of the Framingham Offspring cohort (n = 1012; mean age 59 ± 10 years, 51% women) had both phenotype and genotype data (minimum-maximum per phenotype n = 507–1008). We used Generalized Estimating Equations (GEE), Family Based Association Tests (FBAT) and variance components linkage to relate SNPs to multivariable-adjusted biomarker residuals. Autosomal SNPs (n = 70,987) meeting the following criteria were studied: minor allele frequency ≥ 10%, call rate ≥ 80% and Hardy-Weinberg equilibrium p ≥ 0.001. With  GEE , 58 SNPs had p < 10 -6 : the top SNPs were rs2494250 (p = 1.00*10 -14 ) and rs4128725 (p = 3.68*10 -12 ) for monocyte chemoattractant protein-1 (MCP1), and rs2794520 (p = 2.83*10 -8 ) and rs2808629 (p = 3.19*10 -8 ) for C-reactive protein (CRP) averaged from 3 examinations (over about 20 years). With  FBAT , 11 SNPs had p < 10 -6 : the top SNPs were the same for MCP1 (rs4128725, p = 3.28*10 -8 , and rs2494250, p = 3.55*10 -8 ), and also included B-type natriuretic peptide (rs437021, p = 1.01*10 -6 ) and Vitamin K percent undercarboxylated osteocalcin (rs2052028, p = 1.07*10 -6 ). The peak  LOD  (logarithm of the odds) scores were for MCP1 (4.38, chromosome 1) and CRP (3.28, chromosome 1; previously described) concentrations; of note the 1.5 support interval included the MCP1 and CRP SNPs reported above (GEE model). Previous candidate SNP associations with circulating CRP concentrations were replicated at p < 0.05; the SNPs rs2794520 and rs2808629 are in linkage disequilibrium with previously reported SNPs. GEE, FBAT and linkage results are posted at  . The Framingham GWAS represents a resource to describe potentially novel genetic influences on systemic biomarker variability. The newly described associations will need to be replicated in other studies.\n",
      "-1 biomarker traits\n",
      "-1 into disease\n",
      "-1 biomarkers\n",
      "-1 heritable phenotypes\n",
      "-1 variability\n",
      "-1 ; natriuretic peptides;\n",
      "-1 stress; | exercise test http://www.ebi.ac.uk/efo/EFO_0004328\n",
      "-1 vitamins.\n",
      "-1 single\n",
      "-1 both phenotype\n",
      "-1 per phenotype\n",
      "-1 Tests ( | pulmonary function measurement http://www.ebi.ac.uk/efo/EFO_0003892\n",
      "-1 relate\n",
      "1 monocyte chemoattractant protein-1 ( | ccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749\n",
      "1 C-reactive protein ( | c-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458\n",
      "-1 (over\n",
      "-1 B-type natriuretic peptide (\n",
      "1 Vitamin K | vitamin k measurement http://www.ebi.ac.uk/efo/EFO_0004618\n",
      "-1 included\n",
      "-1 , chromosome 1) and\n",
      "-1 1;\n",
      "-1 included\n",
      "-1 variability.\n",
      "\n",
      "17903294\n",
      "hemostatic factors and hematological phenotypes 19584\n",
      "\thematological measurement http://www.ebi.ac.uk/efo/EFO_0004503 \n",
      "\themoglobin measurement http://www.ebi.ac.uk/efo/EFO_0004509 hb|haemoglobin measurement|hemoglobin count\n",
      "\tplatelet aggregation http://purl.obolibrary.org/obo/GO_0070527 \n",
      "\terythrocyte count http://www.ebi.ac.uk/efo/EFO_0004305 erythrocyte number|red blood cell count|rbc\n",
      "\tblood viscosity http://www.ebi.ac.uk/efo/EFO_0004301 \n",
      "\tplasminogen activator inhibitor 1 measurement http://www.ebi.ac.uk/efo/EFO_0004792 \n",
      "\ttissue plasminogen activator measurement http://www.ebi.ac.uk/efo/EFO_0004791 \n",
      "\tmean corpuscular hemoglobin http://www.ebi.ac.uk/efo/EFO_0004527 mch|mean corpuscular haemoglobin\n",
      "\tprotein c measurement http://www.ebi.ac.uk/efo/EFO_0004633 coagulation factor xiv\n",
      "\tprotein s measurement http://www.ebi.ac.uk/efo/EFO_0004637 \n",
      "hemostatic factors and hematological phenotypes 19585\n",
      "\thematological measurement http://www.ebi.ac.uk/efo/EFO_0004503 \n",
      "\themoglobin measurement http://www.ebi.ac.uk/efo/EFO_0004509 hb|haemoglobin measurement|hemoglobin count\n",
      "\tplatelet aggregation http://purl.obolibrary.org/obo/GO_0070527 \n",
      "\terythrocyte count http://www.ebi.ac.uk/efo/EFO_0004305 erythrocyte number|red blood cell count|rbc\n",
      "\tblood viscosity http://www.ebi.ac.uk/efo/EFO_0004301 \n",
      "\tplasminogen activator inhibitor 1 measurement http://www.ebi.ac.uk/efo/EFO_0004792 \n",
      "\ttissue plasminogen activator measurement http://www.ebi.ac.uk/efo/EFO_0004791 \n",
      "\tmean corpuscular hemoglobin http://www.ebi.ac.uk/efo/EFO_0004527 mch|mean corpuscular haemoglobin\n",
      "\tprotein c measurement http://www.ebi.ac.uk/efo/EFO_0004633 coagulation factor xiv\n",
      "\tprotein s measurement http://www.ebi.ac.uk/efo/EFO_0004637 \n",
      "hemostatic factors and hematological phenotypes 19586\n",
      "\thematological measurement http://www.ebi.ac.uk/efo/EFO_0004503 \n",
      "\themoglobin measurement http://www.ebi.ac.uk/efo/EFO_0004509 hb|haemoglobin measurement|hemoglobin count\n",
      "\tplatelet aggregation http://purl.obolibrary.org/obo/GO_0070527 \n",
      "\terythrocyte count http://www.ebi.ac.uk/efo/EFO_0004305 erythrocyte number|red blood cell count|rbc\n",
      "\tblood viscosity http://www.ebi.ac.uk/efo/EFO_0004301 \n",
      "\tplasminogen activator inhibitor 1 measurement http://www.ebi.ac.uk/efo/EFO_0004792 \n",
      "\ttissue plasminogen activator measurement http://www.ebi.ac.uk/efo/EFO_0004791 \n",
      "\tmean corpuscular hemoglobin http://www.ebi.ac.uk/efo/EFO_0004527 mch|mean corpuscular haemoglobin\n",
      "\tprotein c measurement http://www.ebi.ac.uk/efo/EFO_0004633 coagulation factor xiv\n",
      "\tprotein s measurement http://www.ebi.ac.uk/efo/EFO_0004637 \n",
      "hemostatic factors and hematological phenotypes 19587\n",
      "\thematological measurement http://www.ebi.ac.uk/efo/EFO_0004503 \n",
      "\themoglobin measurement http://www.ebi.ac.uk/efo/EFO_0004509 hb|haemoglobin measurement|hemoglobin count\n",
      "\tplatelet aggregation http://purl.obolibrary.org/obo/GO_0070527 \n",
      "\terythrocyte count http://www.ebi.ac.uk/efo/EFO_0004305 erythrocyte number|red blood cell count|rbc\n",
      "\tblood viscosity http://www.ebi.ac.uk/efo/EFO_0004301 \n",
      "\tplasminogen activator inhibitor 1 measurement http://www.ebi.ac.uk/efo/EFO_0004792 \n",
      "\ttissue plasminogen activator measurement http://www.ebi.ac.uk/efo/EFO_0004791 \n",
      "\tmean corpuscular hemoglobin http://www.ebi.ac.uk/efo/EFO_0004527 mch|mean corpuscular haemoglobin\n",
      "\tprotein c measurement http://www.ebi.ac.uk/efo/EFO_0004633 coagulation factor xiv\n",
      "\tprotein s measurement http://www.ebi.ac.uk/efo/EFO_0004637 \n",
      "hemostatic factors and hematological phenotypes 19588\n",
      "\thematological measurement http://www.ebi.ac.uk/efo/EFO_0004503 \n",
      "\themoglobin measurement http://www.ebi.ac.uk/efo/EFO_0004509 hb|haemoglobin measurement|hemoglobin count\n",
      "\tplatelet aggregation http://purl.obolibrary.org/obo/GO_0070527 \n",
      "\terythrocyte count http://www.ebi.ac.uk/efo/EFO_0004305 erythrocyte number|red blood cell count|rbc\n",
      "\tblood viscosity http://www.ebi.ac.uk/efo/EFO_0004301 \n",
      "\tplasminogen activator inhibitor 1 measurement http://www.ebi.ac.uk/efo/EFO_0004792 \n",
      "\ttissue plasminogen activator measurement http://www.ebi.ac.uk/efo/EFO_0004791 \n",
      "\tmean corpuscular hemoglobin http://www.ebi.ac.uk/efo/EFO_0004527 mch|mean corpuscular haemoglobin\n",
      "\tprotein c measurement http://www.ebi.ac.uk/efo/EFO_0004633 coagulation factor xiv\n",
      "\tprotein s measurement http://www.ebi.ac.uk/efo/EFO_0004637 \n",
      "hemostatic factors and hematological phenotypes 19589\n",
      "\thematological measurement http://www.ebi.ac.uk/efo/EFO_0004503 \n",
      "\themoglobin measurement http://www.ebi.ac.uk/efo/EFO_0004509 hb|haemoglobin measurement|hemoglobin count\n",
      "\tplatelet aggregation http://purl.obolibrary.org/obo/GO_0070527 \n",
      "\terythrocyte count http://www.ebi.ac.uk/efo/EFO_0004305 erythrocyte number|red blood cell count|rbc\n",
      "\tblood viscosity http://www.ebi.ac.uk/efo/EFO_0004301 \n",
      "\tplasminogen activator inhibitor 1 measurement http://www.ebi.ac.uk/efo/EFO_0004792 \n",
      "\ttissue plasminogen activator measurement http://www.ebi.ac.uk/efo/EFO_0004791 \n",
      "\tmean corpuscular hemoglobin http://www.ebi.ac.uk/efo/EFO_0004527 mch|mean corpuscular haemoglobin\n",
      "\tprotein c measurement http://www.ebi.ac.uk/efo/EFO_0004633 coagulation factor xiv\n",
      "\tprotein s measurement http://www.ebi.ac.uk/efo/EFO_0004637 \n",
      "hemostatic factors and hematological phenotypes 19590\n",
      "\thematological measurement http://www.ebi.ac.uk/efo/EFO_0004503 \n",
      "\themoglobin measurement http://www.ebi.ac.uk/efo/EFO_0004509 hb|haemoglobin measurement|hemoglobin count\n",
      "\tplatelet aggregation http://purl.obolibrary.org/obo/GO_0070527 \n",
      "\terythrocyte count http://www.ebi.ac.uk/efo/EFO_0004305 erythrocyte number|red blood cell count|rbc\n",
      "\tblood viscosity http://www.ebi.ac.uk/efo/EFO_0004301 \n",
      "\tplasminogen activator inhibitor 1 measurement http://www.ebi.ac.uk/efo/EFO_0004792 \n",
      "\ttissue plasminogen activator measurement http://www.ebi.ac.uk/efo/EFO_0004791 \n",
      "\tmean corpuscular hemoglobin http://www.ebi.ac.uk/efo/EFO_0004527 mch|mean corpuscular haemoglobin\n",
      "\tprotein c measurement http://www.ebi.ac.uk/efo/EFO_0004633 coagulation factor xiv\n",
      "\tprotein s measurement http://www.ebi.ac.uk/efo/EFO_0004637 \n",
      "hemostatic factors and hematological phenotypes 19591\n",
      "\thematological measurement http://www.ebi.ac.uk/efo/EFO_0004503 \n",
      "\themoglobin measurement http://www.ebi.ac.uk/efo/EFO_0004509 hb|haemoglobin measurement|hemoglobin count\n",
      "\tplatelet aggregation http://purl.obolibrary.org/obo/GO_0070527 \n",
      "\terythrocyte count http://www.ebi.ac.uk/efo/EFO_0004305 erythrocyte number|red blood cell count|rbc\n",
      "\tblood viscosity http://www.ebi.ac.uk/efo/EFO_0004301 \n",
      "\tplasminogen activator inhibitor 1 measurement http://www.ebi.ac.uk/efo/EFO_0004792 \n",
      "\ttissue plasminogen activator measurement http://www.ebi.ac.uk/efo/EFO_0004791 \n",
      "\tmean corpuscular hemoglobin http://www.ebi.ac.uk/efo/EFO_0004527 mch|mean corpuscular haemoglobin\n",
      "\tprotein c measurement http://www.ebi.ac.uk/efo/EFO_0004633 coagulation factor xiv\n",
      "\tprotein s measurement http://www.ebi.ac.uk/efo/EFO_0004637 \n",
      "factor vii 19233\n",
      "\tfactor vii measurement http://www.ebi.ac.uk/efo/EFO_0004619 \n",
      "factor vii 19234\n",
      "\tfactor vii measurement http://www.ebi.ac.uk/efo/EFO_0004619 \n",
      "hemostatic factors and hematological phenotypes 19235\n",
      "\thematological measurement http://www.ebi.ac.uk/efo/EFO_0004503 \n",
      "\themoglobin measurement http://www.ebi.ac.uk/efo/EFO_0004509 hb|haemoglobin measurement|hemoglobin count\n",
      "\tplatelet aggregation http://purl.obolibrary.org/obo/GO_0070527 \n",
      "\terythrocyte count http://www.ebi.ac.uk/efo/EFO_0004305 erythrocyte number|red blood cell count|rbc\n",
      "\tblood viscosity http://www.ebi.ac.uk/efo/EFO_0004301 \n",
      "\tplasminogen activator inhibitor 1 measurement http://www.ebi.ac.uk/efo/EFO_0004792 \n",
      "\ttissue plasminogen activator measurement http://www.ebi.ac.uk/efo/EFO_0004791 \n",
      "\tmean corpuscular hemoglobin http://www.ebi.ac.uk/efo/EFO_0004527 mch|mean corpuscular haemoglobin\n",
      "\tprotein c measurement http://www.ebi.ac.uk/efo/EFO_0004633 coagulation factor xiv\n",
      "\tprotein s measurement http://www.ebi.ac.uk/efo/EFO_0004637 \n",
      "hemostatic factors and hematological phenotypes 19236\n",
      "\thematological measurement http://www.ebi.ac.uk/efo/EFO_0004503 \n",
      "\themoglobin measurement http://www.ebi.ac.uk/efo/EFO_0004509 hb|haemoglobin measurement|hemoglobin count\n",
      "\tplatelet aggregation http://purl.obolibrary.org/obo/GO_0070527 \n",
      "\terythrocyte count http://www.ebi.ac.uk/efo/EFO_0004305 erythrocyte number|red blood cell count|rbc\n",
      "\tblood viscosity http://www.ebi.ac.uk/efo/EFO_0004301 \n",
      "\tplasminogen activator inhibitor 1 measurement http://www.ebi.ac.uk/efo/EFO_0004792 \n",
      "\ttissue plasminogen activator measurement http://www.ebi.ac.uk/efo/EFO_0004791 \n",
      "\tmean corpuscular hemoglobin http://www.ebi.ac.uk/efo/EFO_0004527 mch|mean corpuscular haemoglobin\n",
      "\tprotein c measurement http://www.ebi.ac.uk/efo/EFO_0004633 coagulation factor xiv\n",
      "\tprotein s measurement http://www.ebi.ac.uk/efo/EFO_0004637 \n",
      "factor vii 19431\n",
      "\tfactor vii measurement http://www.ebi.ac.uk/efo/EFO_0004619 \n",
      "factor vii 19575\n",
      "\tfactor vii measurement http://www.ebi.ac.uk/efo/EFO_0004619 \n",
      "hemostatic factors and hematological phenotypes 19576\n",
      "\thematological measurement http://www.ebi.ac.uk/efo/EFO_0004503 \n",
      "\themoglobin measurement http://www.ebi.ac.uk/efo/EFO_0004509 hb|haemoglobin measurement|hemoglobin count\n",
      "\tplatelet aggregation http://purl.obolibrary.org/obo/GO_0070527 \n",
      "\terythrocyte count http://www.ebi.ac.uk/efo/EFO_0004305 erythrocyte number|red blood cell count|rbc\n",
      "\tblood viscosity http://www.ebi.ac.uk/efo/EFO_0004301 \n",
      "\tplasminogen activator inhibitor 1 measurement http://www.ebi.ac.uk/efo/EFO_0004792 \n",
      "\ttissue plasminogen activator measurement http://www.ebi.ac.uk/efo/EFO_0004791 \n",
      "\tmean corpuscular hemoglobin http://www.ebi.ac.uk/efo/EFO_0004527 mch|mean corpuscular haemoglobin\n",
      "\tprotein c measurement http://www.ebi.ac.uk/efo/EFO_0004633 coagulation factor xiv\n",
      "\tprotein s measurement http://www.ebi.ac.uk/efo/EFO_0004637 \n",
      "hemostatic factors and hematological phenotypes 19577\n",
      "\thematological measurement http://www.ebi.ac.uk/efo/EFO_0004503 \n",
      "\themoglobin measurement http://www.ebi.ac.uk/efo/EFO_0004509 hb|haemoglobin measurement|hemoglobin count\n",
      "\tplatelet aggregation http://purl.obolibrary.org/obo/GO_0070527 \n",
      "\terythrocyte count http://www.ebi.ac.uk/efo/EFO_0004305 erythrocyte number|red blood cell count|rbc\n",
      "\tblood viscosity http://www.ebi.ac.uk/efo/EFO_0004301 \n",
      "\tplasminogen activator inhibitor 1 measurement http://www.ebi.ac.uk/efo/EFO_0004792 \n",
      "\ttissue plasminogen activator measurement http://www.ebi.ac.uk/efo/EFO_0004791 \n",
      "\tmean corpuscular hemoglobin http://www.ebi.ac.uk/efo/EFO_0004527 mch|mean corpuscular haemoglobin\n",
      "\tprotein c measurement http://www.ebi.ac.uk/efo/EFO_0004633 coagulation factor xiv\n",
      "\tprotein s measurement http://www.ebi.ac.uk/efo/EFO_0004637 \n",
      "hemostatic factors and hematological phenotypes 19578\n",
      "\thematological measurement http://www.ebi.ac.uk/efo/EFO_0004503 \n",
      "\themoglobin measurement http://www.ebi.ac.uk/efo/EFO_0004509 hb|haemoglobin measurement|hemoglobin count\n",
      "\tplatelet aggregation http://purl.obolibrary.org/obo/GO_0070527 \n",
      "\terythrocyte count http://www.ebi.ac.uk/efo/EFO_0004305 erythrocyte number|red blood cell count|rbc\n",
      "\tblood viscosity http://www.ebi.ac.uk/efo/EFO_0004301 \n",
      "\tplasminogen activator inhibitor 1 measurement http://www.ebi.ac.uk/efo/EFO_0004792 \n",
      "\ttissue plasminogen activator measurement http://www.ebi.ac.uk/efo/EFO_0004791 \n",
      "\tmean corpuscular hemoglobin http://www.ebi.ac.uk/efo/EFO_0004527 mch|mean corpuscular haemoglobin\n",
      "\tprotein c measurement http://www.ebi.ac.uk/efo/EFO_0004633 coagulation factor xiv\n",
      "\tprotein s measurement http://www.ebi.ac.uk/efo/EFO_0004637 \n",
      "hemostatic factors and hematological phenotypes 19579\n",
      "\thematological measurement http://www.ebi.ac.uk/efo/EFO_0004503 \n",
      "\themoglobin measurement http://www.ebi.ac.uk/efo/EFO_0004509 hb|haemoglobin measurement|hemoglobin count\n",
      "\tplatelet aggregation http://purl.obolibrary.org/obo/GO_0070527 \n",
      "\terythrocyte count http://www.ebi.ac.uk/efo/EFO_0004305 erythrocyte number|red blood cell count|rbc\n",
      "\tblood viscosity http://www.ebi.ac.uk/efo/EFO_0004301 \n",
      "\tplasminogen activator inhibitor 1 measurement http://www.ebi.ac.uk/efo/EFO_0004792 \n",
      "\ttissue plasminogen activator measurement http://www.ebi.ac.uk/efo/EFO_0004791 \n",
      "\tmean corpuscular hemoglobin http://www.ebi.ac.uk/efo/EFO_0004527 mch|mean corpuscular haemoglobin\n",
      "\tprotein c measurement http://www.ebi.ac.uk/efo/EFO_0004633 coagulation factor xiv\n",
      "\tprotein s measurement http://www.ebi.ac.uk/efo/EFO_0004637 \n",
      "hemostatic factors and hematological phenotypes 19580\n",
      "\thematological measurement http://www.ebi.ac.uk/efo/EFO_0004503 \n",
      "\themoglobin measurement http://www.ebi.ac.uk/efo/EFO_0004509 hb|haemoglobin measurement|hemoglobin count\n",
      "\tplatelet aggregation http://purl.obolibrary.org/obo/GO_0070527 \n",
      "\terythrocyte count http://www.ebi.ac.uk/efo/EFO_0004305 erythrocyte number|red blood cell count|rbc\n",
      "\tblood viscosity http://www.ebi.ac.uk/efo/EFO_0004301 \n",
      "\tplasminogen activator inhibitor 1 measurement http://www.ebi.ac.uk/efo/EFO_0004792 \n",
      "\ttissue plasminogen activator measurement http://www.ebi.ac.uk/efo/EFO_0004791 \n",
      "\tmean corpuscular hemoglobin http://www.ebi.ac.uk/efo/EFO_0004527 mch|mean corpuscular haemoglobin\n",
      "\tprotein c measurement http://www.ebi.ac.uk/efo/EFO_0004633 coagulation factor xiv\n",
      "\tprotein s measurement http://www.ebi.ac.uk/efo/EFO_0004637 \n",
      "hemostatic factors and hematological phenotypes 19581\n",
      "\thematological measurement http://www.ebi.ac.uk/efo/EFO_0004503 \n",
      "\themoglobin measurement http://www.ebi.ac.uk/efo/EFO_0004509 hb|haemoglobin measurement|hemoglobin count\n",
      "\tplatelet aggregation http://purl.obolibrary.org/obo/GO_0070527 \n",
      "\terythrocyte count http://www.ebi.ac.uk/efo/EFO_0004305 erythrocyte number|red blood cell count|rbc\n",
      "\tblood viscosity http://www.ebi.ac.uk/efo/EFO_0004301 \n",
      "\tplasminogen activator inhibitor 1 measurement http://www.ebi.ac.uk/efo/EFO_0004792 \n",
      "\ttissue plasminogen activator measurement http://www.ebi.ac.uk/efo/EFO_0004791 \n",
      "\tmean corpuscular hemoglobin http://www.ebi.ac.uk/efo/EFO_0004527 mch|mean corpuscular haemoglobin\n",
      "\tprotein c measurement http://www.ebi.ac.uk/efo/EFO_0004633 coagulation factor xiv\n",
      "\tprotein s measurement http://www.ebi.ac.uk/efo/EFO_0004637 \n",
      "hemostatic factors and hematological phenotypes 19582\n",
      "\thematological measurement http://www.ebi.ac.uk/efo/EFO_0004503 \n",
      "\themoglobin measurement http://www.ebi.ac.uk/efo/EFO_0004509 hb|haemoglobin measurement|hemoglobin count\n",
      "\tplatelet aggregation http://purl.obolibrary.org/obo/GO_0070527 \n",
      "\terythrocyte count http://www.ebi.ac.uk/efo/EFO_0004305 erythrocyte number|red blood cell count|rbc\n",
      "\tblood viscosity http://www.ebi.ac.uk/efo/EFO_0004301 \n",
      "\tplasminogen activator inhibitor 1 measurement http://www.ebi.ac.uk/efo/EFO_0004792 \n",
      "\ttissue plasminogen activator measurement http://www.ebi.ac.uk/efo/EFO_0004791 \n",
      "\tmean corpuscular hemoglobin http://www.ebi.ac.uk/efo/EFO_0004527 mch|mean corpuscular haemoglobin\n",
      "\tprotein c measurement http://www.ebi.ac.uk/efo/EFO_0004633 coagulation factor xiv\n",
      "\tprotein s measurement http://www.ebi.ac.uk/efo/EFO_0004637 \n",
      "hemostatic factors and hematological phenotypes 19583\n",
      "\thematological measurement http://www.ebi.ac.uk/efo/EFO_0004503 \n",
      "\themoglobin measurement http://www.ebi.ac.uk/efo/EFO_0004509 hb|haemoglobin measurement|hemoglobin count\n",
      "\tplatelet aggregation http://purl.obolibrary.org/obo/GO_0070527 \n",
      "\terythrocyte count http://www.ebi.ac.uk/efo/EFO_0004305 erythrocyte number|red blood cell count|rbc\n",
      "\tblood viscosity http://www.ebi.ac.uk/efo/EFO_0004301 \n",
      "\tplasminogen activator inhibitor 1 measurement http://www.ebi.ac.uk/efo/EFO_0004792 \n",
      "\ttissue plasminogen activator measurement http://www.ebi.ac.uk/efo/EFO_0004791 \n",
      "\tmean corpuscular hemoglobin http://www.ebi.ac.uk/efo/EFO_0004527 mch|mean corpuscular haemoglobin\n",
      "\tprotein c measurement http://www.ebi.ac.uk/efo/EFO_0004633 coagulation factor xiv\n",
      "\tprotein s measurement http://www.ebi.ac.uk/efo/EFO_0004637 \n",
      "Genome-wide association and linkage analyses of hemostatic factors and hematological phenotypes in the Framingham Heart Study. Increased circulating levels of hemostatic factors as well as anemia have been associated with increased risk of cardiovascular disease (CVD). Known associations between hemostatic factors and sequence variants at genes encoding these factors explain only a small proportion of total phenotypic variation. We sought to confirm known putative loci and identify novel loci that may influence either trait in genome-wide association and linkage analyses using the Affymetrix GeneChip 100K single nucleotide polymorphism (SNP) set. Plasma levels of circulating hemostatic factors (fibrinogen, factor VII, plasminogen activator inhibitor-1, von Willebrand factor, tissue plasminogen activator, D-dimer) and hematological phenotypes (platelet aggregation, viscosity, hemoglobin, red blood cell count, mean corpuscular volume, mean corpuscular hemoglobin concentration) were obtained in approximately 1000 Framingham Heart Study (FHS) participants from 310 families. Population-based association analyses using the generalized estimating equations (GEE), family-based association test (FBAT), and multipoint variance components linkage analyses were performed on the multivariable adjusted residuals of hemostatic and hematological phenotypes. In association analysis, the lowest GEE p-value for hemostatic factors was p = 4.5*10 -16  for factor VII at SNP rs561241, a variant located near the  F7  gene and in complete linkage disequilibrium (LD) (r 2  = 1) with the Arg353Gln  F7  SNP previously shown to account for 9% of total phenotypic variance. The lowest GEE p-value for hematological phenotypes was 7*10 -8  at SNP rs2412522 on chromosome 4 for mean corpuscular hemoglobin concentration. We presented top 25 most significant GEE results with p-values in the range of 10 -6  to 10 -5  for hemostatic or hematological phenotypes. In relating 100K SNPs to known candidate genes, we identified two SNPs (rs1582055, rs4897475) in erythrocyte membrane protein band 4.1-like 2 ( EPB41L2 ) associated with hematological phenotypes (GEE p < 10 -3 ). In linkage analyses, the highest linkage LOD score for hemostatic factors was 3.3 for factor VII on chromosome 10 around 15 Mb, and for hematological phenotypes, LOD 3.4 for hemoglobin on chromosome 4 around 55 Mb. All GEE and FBAT association and variance components linkage results can be found at  Using genome-wide association methodology, we have successfully identified a SNP in complete LD with a sequence variant previously shown to be strongly associated with factor VII, providing proof of principle for this approach. Further study of additional strongly associated SNPs and linked regions may identify novel variants that influence the inter-individual variability in hemostatic factors and hematological phenotypes.\n",
      "-1 hematological phenotypes\n",
      "-1 associated with increased\n",
      "-1 cardiovascular disease ( | cardiovascular disease http://www.ebi.ac.uk/efo/EFO_0000319\n",
      "-1 as\n",
      "-1 as\n",
      "-1 anemia | anemia http://www.ebi.ac.uk/efo/EFO_0004272\n",
      "-1 either trait\n",
      "-1 single\n",
      "1 , red blood cell count, | erythrocyte count http://www.ebi.ac.uk/efo/EFO_0004305\n",
      "-1 , mean corpuscular hemoglobin concentration)\n",
      "1 , plasminogen activator inhibitor-1, | plasminogen activator inhibitor 1 measurement http://www.ebi.ac.uk/efo/EFO_0004792\n",
      "-1 , von Willebrand factor,\n",
      "1 , tissue plasminogen activator, | tissue plasminogen activator measurement http://www.ebi.ac.uk/efo/EFO_0004791\n",
      "-1 , mean corpuscular volume,\n",
      "1 , factor VII, | factor vii measurement http://www.ebi.ac.uk/efo/EFO_0004619\n",
      "1 (platelet aggregation, | platelet aggregation http://purl.obolibrary.org/obo/GO_0070527\n",
      "-1 corpuscular volume, mean\n",
      "-1 volume, mean corpuscular\n",
      "-1 (fibrinogen, | fibrinogen measurement http://www.ebi.ac.uk/efo/EFO_0004623\n",
      "-1 , D-dimer)\n",
      "1 , hemoglobin, | hemoglobin measurement http://www.ebi.ac.uk/efo/EFO_0004509\n",
      "-1 (FHS)\n",
      "-1 hematological phenotypes\n",
      "-1 hematological phenotypes\n",
      "1 factor VII at | factor vii measurement http://www.ebi.ac.uk/efo/EFO_0004619\n",
      "-1 = 1)\n",
      "-1 ) with the\n",
      "-1 was\n",
      "-1 mean corpuscular hemoglobin concentration.\n",
      "-1 hematological phenotypes\n",
      "-1 was\n",
      "-1 -6  to\n",
      "-1 hematological phenotypes\n",
      "-1 hematological phenotypes\n",
      "-1 erythrocyte\n",
      "-1 protein | protein measurement http://www.ebi.ac.uk/efo/EFO_0004747\n",
      "-1 on chromosome 10\n",
      "1 factor VII | factor vii measurement http://www.ebi.ac.uk/efo/EFO_0004619\n",
      "-1 Mb,\n",
      "-1 hematological phenotypes\n",
      "-1 Mb.\n",
      "-1 was\n",
      "1 hemoglobin | hemoglobin measurement http://www.ebi.ac.uk/efo/EFO_0004509\n",
      "1 factor VII, | factor vii measurement http://www.ebi.ac.uk/efo/EFO_0004619\n",
      "-1 All\n",
      "-1 can\n",
      "-1 hematological phenotypes\n",
      "-1 variability\n",
      "\n",
      "17903295\n",
      "aging traits 19811\n",
      "\tage at menopause http://www.ebi.ac.uk/efo/EFO_0004704 \n",
      "\texercise test http://www.ebi.ac.uk/efo/EFO_0004328 stress test\n",
      "\taging http://purl.obolibrary.org/obo/GO_0007568 senescence|aging\n",
      "\tage at death http://www.ebi.ac.uk/efo/EFO_0005056 \n",
      "aging traits 19812\n",
      "\tage at menopause http://www.ebi.ac.uk/efo/EFO_0004704 \n",
      "\texercise test http://www.ebi.ac.uk/efo/EFO_0004328 stress test\n",
      "\taging http://purl.obolibrary.org/obo/GO_0007568 senescence|aging\n",
      "\tage at death http://www.ebi.ac.uk/efo/EFO_0005056 \n",
      "aging traits 19813\n",
      "\tage at menopause http://www.ebi.ac.uk/efo/EFO_0004704 \n",
      "\texercise test http://www.ebi.ac.uk/efo/EFO_0004328 stress test\n",
      "\taging http://purl.obolibrary.org/obo/GO_0007568 senescence|aging\n",
      "\tage at death http://www.ebi.ac.uk/efo/EFO_0005056 \n",
      "aging traits 19814\n",
      "\tage at menopause http://www.ebi.ac.uk/efo/EFO_0004704 \n",
      "\texercise test http://www.ebi.ac.uk/efo/EFO_0004328 stress test\n",
      "\taging http://purl.obolibrary.org/obo/GO_0007568 senescence|aging\n",
      "\tage at death http://www.ebi.ac.uk/efo/EFO_0005056 \n",
      "aging traits 19815\n",
      "\tage at menopause http://www.ebi.ac.uk/efo/EFO_0004704 \n",
      "\texercise test http://www.ebi.ac.uk/efo/EFO_0004328 stress test\n",
      "\taging http://purl.obolibrary.org/obo/GO_0007568 senescence|aging\n",
      "\tage at death http://www.ebi.ac.uk/efo/EFO_0005056 \n",
      "aging traits 19816\n",
      "\tage at menopause http://www.ebi.ac.uk/efo/EFO_0004704 \n",
      "\texercise test http://www.ebi.ac.uk/efo/EFO_0004328 stress test\n",
      "\taging http://purl.obolibrary.org/obo/GO_0007568 senescence|aging\n",
      "\tage at death http://www.ebi.ac.uk/efo/EFO_0005056 \n",
      "aging traits 19817\n",
      "\tage at menopause http://www.ebi.ac.uk/efo/EFO_0004704 \n",
      "\texercise test http://www.ebi.ac.uk/efo/EFO_0004328 stress test\n",
      "\taging http://purl.obolibrary.org/obo/GO_0007568 senescence|aging\n",
      "\tage at death http://www.ebi.ac.uk/efo/EFO_0005056 \n",
      "morbidity-free survival 19818\n",
      "\tdisease free survival http://www.ebi.ac.uk/efo/EFO_0000409  disease-free survival|morbidity-free survival\n",
      "morbidity-free survival 19819\n",
      "\tdisease free survival http://www.ebi.ac.uk/efo/EFO_0000409  disease-free survival|morbidity-free survival\n",
      "morbidity-free survival 19820\n",
      "\tdisease free survival http://www.ebi.ac.uk/efo/EFO_0000409  disease-free survival|morbidity-free survival\n",
      "morbidity-free survival 19189\n",
      "\tdisease free survival http://www.ebi.ac.uk/efo/EFO_0000409  disease-free survival|morbidity-free survival\n",
      "morbidity-free survival 19190\n",
      "\tdisease free survival http://www.ebi.ac.uk/efo/EFO_0000409  disease-free survival|morbidity-free survival\n",
      "aging traits 19191\n",
      "\tage at menopause http://www.ebi.ac.uk/efo/EFO_0004704 \n",
      "\texercise test http://www.ebi.ac.uk/efo/EFO_0004328 stress test\n",
      "\taging http://purl.obolibrary.org/obo/GO_0007568 senescence|aging\n",
      "\tage at death http://www.ebi.ac.uk/efo/EFO_0005056 \n",
      "aging traits 19192\n",
      "\tage at menopause http://www.ebi.ac.uk/efo/EFO_0004704 \n",
      "\texercise test http://www.ebi.ac.uk/efo/EFO_0004328 stress test\n",
      "\taging http://purl.obolibrary.org/obo/GO_0007568 senescence|aging\n",
      "\tage at death http://www.ebi.ac.uk/efo/EFO_0005056 \n",
      "aging traits 19193\n",
      "\tage at menopause http://www.ebi.ac.uk/efo/EFO_0004704 \n",
      "\texercise test http://www.ebi.ac.uk/efo/EFO_0004328 stress test\n",
      "\taging http://purl.obolibrary.org/obo/GO_0007568 senescence|aging\n",
      "\tage at death http://www.ebi.ac.uk/efo/EFO_0005056 \n",
      "Genetic correlates of longevity and selected age-related phenotypes: a genome-wide association study in the Framingham Study. Family studies and heritability estimates provide evidence for a genetic contribution to variation in the human life span. We conducted a genome wide association study (Affymetrix 100K SNP GeneChip) for longevity-related traits in a community-based sample. We report on 5 longevity and aging traits in up to 1345 Framingham Study participants from 330 families. Multivariable-adjusted residuals were computed using appropriate models (Cox proportional hazards, logistic, or linear regression) and the residuals from these models were used to test for association with qualifying SNPs (70, 987 autosomal SNPs with genotypic call rate ≥80%, minor allele frequency ≥10%, Hardy-Weinberg test p ≥ 0.001). In family-based association test (FBAT) models, 8 SNPs in two regions approximately 500 kb apart on chromosome 1 (physical positions 73,091,610 and 73, 527,652) were associated with age at death (p-value < 10 -5 ). The two sets of SNPs were in high linkage disequilibrium (minimum r 2  = 0.58). The top 30 SNPs for generalized estimating equation (GEE) tests of association with age at death included rs10507486 (p = 0.0001) and rs4943794 (p = 0.0002), SNPs intronic to  FOXO1A , a gene implicated in lifespan extension in animal models. FBAT models identified 7 SNPs and GEE models identified 9 SNPs associated with both age at death and morbidity-free survival at age 65 including rs2374983 near  PON1 . In the analysis of selected candidate genes, SNP associations (FBAT or GEE p-value < 0.01) were identified for age at death in or near the following genes:  FOXO1A, GAPDH, KL, LEPR, PON1, PSEN1, SOD2, and WRN . Top ranked SNP associations in the GEE model for age at natural menopause included rs6910534 (p = 0.00003) near  FOXO3a  and rs3751591 (p = 0.00006) in  CYP19A1 . Results of all longevity phenotype-genotype associations for all autosomal SNPs are web posted at  . Longevity and aging traits are associated with SNPs on the Affymetrix 100K GeneChip. None of the associations achieved genome-wide significance. These data generate hypotheses and serve as a resource for replication as more genes and biologic pathways are proposed as contributing to longevity and healthy aging.\n",
      "-1 age-related phenotypes\n",
      "-1 longevity\n",
      "-1 longevity-related traits\n",
      "1 aging traits\n",
      "-1 longevity\n",
      "1 age at death ( | age at death http://www.ebi.ac.uk/efo/EFO_0005056\n",
      "-1 1 (\n",
      "1 age at death | age at death http://www.ebi.ac.uk/efo/EFO_0005056\n",
      "-1 ) tests | pulmonary function measurement http://www.ebi.ac.uk/efo/EFO_0003892\n",
      "-1 included\n",
      "1 morbidity-free survival at age | disease free survival http://www.ebi.ac.uk/efo/EFO_0000409\n",
      "1 age at death | age at death http://www.ebi.ac.uk/efo/EFO_0005056\n",
      "-1 including\n",
      "1 age at death | age at death http://www.ebi.ac.uk/efo/EFO_0005056\n",
      "-1 WRN .\n",
      "1 menopause | age at menopause http://www.ebi.ac.uk/efo/EFO_0004704\n",
      "-1 included\n",
      "-1 all\n",
      "-1 longevity\n",
      "-1 all\n",
      "1 aging traits\n",
      "-1 Longevity\n",
      "1 aging. | aging http://purl.obolibrary.org/obo/GO_0007568\n",
      "-1 as\n",
      "-1 as\n",
      "-1 as\n",
      "-1 longevity\n",
      "\n",
      "17903296\n",
      "bone mineral density 19843\n",
      "\tbone density http://www.ebi.ac.uk/efo/EFO_0003923 density\n",
      "bone mineral density 19844\n",
      "\tbone density http://www.ebi.ac.uk/efo/EFO_0003923 density\n",
      "hip geometry 19598\n",
      "\thip geometry http://www.ebi.ac.uk/efo/EFO_0004685 \n",
      "hip geometry 19599\n",
      "\thip geometry http://www.ebi.ac.uk/efo/EFO_0004685 \n",
      "hip geometry 19600\n",
      "\thip geometry http://www.ebi.ac.uk/efo/EFO_0004685 \n",
      "hip geometry 19601\n",
      "\thip geometry http://www.ebi.ac.uk/efo/EFO_0004685 \n",
      "hip geometry 19602\n",
      "\thip geometry http://www.ebi.ac.uk/efo/EFO_0004685 \n",
      "hip geometry 19603\n",
      "\thip geometry http://www.ebi.ac.uk/efo/EFO_0004685 \n",
      "hip geometry 19604\n",
      "\thip geometry http://www.ebi.ac.uk/efo/EFO_0004685 \n",
      "hip geometry 19605\n",
      "\thip geometry http://www.ebi.ac.uk/efo/EFO_0004685 \n",
      "hip geometry 19606\n",
      "\thip geometry http://www.ebi.ac.uk/efo/EFO_0004685 \n",
      "hip geometry 19607\n",
      "\thip geometry http://www.ebi.ac.uk/efo/EFO_0004685 \n",
      "hip geometry 19608\n",
      "\thip geometry http://www.ebi.ac.uk/efo/EFO_0004685 \n",
      "hip geometry 19609\n",
      "\thip geometry http://www.ebi.ac.uk/efo/EFO_0004685 \n",
      "bone mineral density 19610\n",
      "\tbone density http://www.ebi.ac.uk/efo/EFO_0003923 density\n",
      "bone mineral density 19611\n",
      "\tbone density http://www.ebi.ac.uk/efo/EFO_0003923 density\n",
      "bone mineral density 19612\n",
      "\tbone density http://www.ebi.ac.uk/efo/EFO_0003923 density\n",
      "bone mineral density 19613\n",
      "\tbone density http://www.ebi.ac.uk/efo/EFO_0003923 density\n",
      "bone mineral density 19614\n",
      "\tbone density http://www.ebi.ac.uk/efo/EFO_0003923 density\n",
      "bone mineral density 19174\n",
      "\tbone density http://www.ebi.ac.uk/efo/EFO_0003923 density\n",
      "hip geometry 19175\n",
      "\thip geometry http://www.ebi.ac.uk/efo/EFO_0004685 \n",
      "hip geometry 19176\n",
      "\thip geometry http://www.ebi.ac.uk/efo/EFO_0004685 \n",
      "hip geometry 19177\n",
      "\thip geometry http://www.ebi.ac.uk/efo/EFO_0004685 \n",
      "hip geometry 19178\n",
      "\thip geometry http://www.ebi.ac.uk/efo/EFO_0004685 \n",
      "Genome-wide association with bone mass and geometry in the Framingham Heart Study. Osteoporosis is characterized by low bone mass and compromised bone structure, heritable traits that contribute to fracture risk. There have been no genome-wide association and linkage studies for these traits using high-density genotyping platforms. We used the Affymetrix 100K SNP GeneChip marker set in the Framingham Heart Study (FHS) to examine genetic associations with ten primary quantitative traits: bone mineral density (BMD), calcaneal ultrasound, and geometric indices of the hip. To test associations with multivariable-adjusted residual trait values, we used additive generalized estimating equation (GEE) and family-based association tests (FBAT) models within each sex as well as sexes combined. We evaluated 70,987 autosomal SNPs with genotypic call rates ≥80%, HWE p ≥ 0.001, and MAF ≥10% in up to 1141 phenotyped individuals (495 men and 646 women, mean age 62.5 yrs). Variance component linkage analysis was performed using 11,200 markers. Heritability estimates for all bone phenotypes were 30–66%. LOD scores ≥3.0 were found on chromosomes 15 (1.5 LOD confidence interval: 51,336,679–58,934,236 bp) and 22 (35,890,398–48,603,847 bp) for femoral shaft section modulus. The ten primary phenotypes had 12 associations with 100K SNPs in GEE models at p < 0.000001 and 2 associations in FBAT models at p < 0.000001. The 25 most significant p-values for GEE and FBAT were all less than 3.5 × 10 -6  and 2.5 × 10 -5 , respectively. Of the 40 top SNPs with the greatest numbers of significantly associated BMD traits (including femoral neck, trochanter, and lumbar spine), one half to two-thirds were in or near genes that have  not  previously been studied for osteoporosis. Notably, pleiotropic associations between BMD and bone geometric traits were uncommon. Evidence for association (FBAT or GEE p < 0.05) was observed for several SNPs in candidate genes for osteoporosis, such as rs1801133 in  MTHFR ; rs1884052 and rs3778099 in  ESR1 ; rs4988300 in  LRP5 ; rs2189480 in  VDR ; rs2075555 in  COLIA1 ; rs10519297 and rs2008691 in  CYP19 , as well as SNPs in  PPARG  (rs10510418 and rs2938392) and  ANKH  (rs2454873 and rs379016). All GEE, FBAT and linkage results are provided as an open-access results resource at  . The FHS 100K SNP project offers an unbiased genome-wide strategy to identify new candidate loci and to replicate previously suggested candidate genes for osteoporosis.\n",
      "-1 bone\n",
      "-1 heritable traits\n",
      "-1 Osteoporosis | osteoporosis http://www.ebi.ac.uk/efo/EFO_0003882\n",
      "-1 is\n",
      "-1 bone\n",
      "-1 bone\n",
      "-1 these traits\n",
      "-1 high-density\n",
      "1 : bone mineral density (\n",
      "-1 (BMD),\n",
      "-1 of the hip.\n",
      "-1 (FHS)\n",
      "-1 primary quantitative traits\n",
      "-1 indices | ankle brachial index http://www.ebi.ac.uk/efo/EFO_0003912\n",
      "-1 residual trait\n",
      "-1 tests ( | pulmonary function measurement http://www.ebi.ac.uk/efo/EFO_0003892\n",
      "-1 as\n",
      "-1 as\n",
      "-1 was\n",
      "-1 bone phenotypes\n",
      "-1 all\n",
      "-1 on chromosomes 15 (\n",
      "-1 primary phenotypes\n",
      "-1 all\n",
      "-1 with the\n",
      "-1 BMD traits\n",
      "-1 (including\n",
      "-1 osteoporosis. | osteoporosis http://www.ebi.ac.uk/efo/EFO_0003882\n",
      "-1 and bone\n",
      "-1 geometric traits\n",
      "-1 BMD\n",
      "-1 ) was\n",
      "-1 osteoporosis, | osteoporosis http://www.ebi.ac.uk/efo/EFO_0003882\n",
      "-1 , as\n",
      "-1 as\n",
      "-1 as\n",
      "-1 All\n",
      "-1 as\n",
      "-1 an\n",
      "-1 osteoporosis. | osteoporosis http://www.ebi.ac.uk/efo/EFO_0003882\n",
      "-1 FHS\n",
      "-1 an\n",
      "\n",
      "17903297\n",
      "volumetric brain mri 19552\n",
      "\tvolumetric brain mri http://www.ebi.ac.uk/efo/EFO_0004868 \n",
      "volumetric brain mri 19553\n",
      "\tvolumetric brain mri http://www.ebi.ac.uk/efo/EFO_0004868 \n",
      "volumetric brain mri 19554\n",
      "\tvolumetric brain mri http://www.ebi.ac.uk/efo/EFO_0004868 \n",
      "cognitive test performance 19555\n",
      "\tneuropsychological test http://www.ebi.ac.uk/efo/EFO_0003926 test\n",
      "cognitive test performance 19556\n",
      "\tneuropsychological test http://www.ebi.ac.uk/efo/EFO_0003926 test\n",
      "cognitive test performance 19557\n",
      "\tneuropsychological test http://www.ebi.ac.uk/efo/EFO_0003926 test\n",
      "cognitive test performance 19558\n",
      "\tneuropsychological test http://www.ebi.ac.uk/efo/EFO_0003926 test\n",
      "cognitive test performance 19559\n",
      "\tneuropsychological test http://www.ebi.ac.uk/efo/EFO_0003926 test\n",
      "cognitive test performance 19560\n",
      "\tneuropsychological test http://www.ebi.ac.uk/efo/EFO_0003926 test\n",
      "cognitive test performance 19561\n",
      "\tneuropsychological test http://www.ebi.ac.uk/efo/EFO_0003926 test\n",
      "cognitive test performance 19562\n",
      "\tneuropsychological test http://www.ebi.ac.uk/efo/EFO_0003926 test\n",
      "cognitive test performance 19563\n",
      "\tneuropsychological test http://www.ebi.ac.uk/efo/EFO_0003926 test\n",
      "cognitive test performance 19564\n",
      "\tneuropsychological test http://www.ebi.ac.uk/efo/EFO_0003926 test\n",
      "cognitive test performance 19218\n",
      "\tneuropsychological test http://www.ebi.ac.uk/efo/EFO_0003926 test\n",
      "cognitive test performance 19219\n",
      "\tneuropsychological test http://www.ebi.ac.uk/efo/EFO_0003926 test\n",
      "cognitive test performance 19220\n",
      "\tneuropsychological test http://www.ebi.ac.uk/efo/EFO_0003926 test\n",
      "cognitive test performance 19221\n",
      "\tneuropsychological test http://www.ebi.ac.uk/efo/EFO_0003926 test\n",
      "volumetric brain mri 19222\n",
      "\tvolumetric brain mri http://www.ebi.ac.uk/efo/EFO_0004868 \n",
      "volumetric brain mri 19551\n",
      "\tvolumetric brain mri http://www.ebi.ac.uk/efo/EFO_0004868 \n",
      "Genetic correlates of brain aging on MRI and cognitive test measures: a genome-wide association and linkage analysis in the Framingham study. Brain magnetic resonance imaging (MRI) and cognitive tests can identify heritable endophenotypes associated with an increased risk of developing stroke, dementia and Alzheimer's disease (AD). We conducted a genome-wide association (GWA) and linkage analysis exploring the genetic basis of these endophenotypes in a community-based sample. A total of 705 stroke- and dementia-free Framingham participants (age 62 +9 yrs, 50% male) who underwent volumetric brain MRI and cognitive testing (1999–2002) were genotyped. We used linear models adjusting for first degree relationships via generalized estimating equations (GEE) and family based association tests (FBAT) in additive models to relate qualifying single nucleotide polymorphisms (SNPs, 70,987 autosomal on Affymetrix 100K Human Gene Chip with minor allele frequency ≥ 0.10, genotypic call rate ≥ 0.80, and Hardy-Weinberg equilibrium p-value ≥ 0.001) to multivariable-adjusted residuals of 9 MRI measures including total cerebral brain (TCBV), lobar, ventricular and white matter hyperintensity (WMH) volumes, and 6 cognitive factors/tests assessing verbal and visuospatial memory, visual scanning and motor speed, reading, abstract reasoning and naming. We determined multipoint identity-by-descent utilizing 10,592 informative SNPs and 613 short tandem repeats and used variance component analyses to compute LOD scores. The strongest gene-phenotype association in FBAT analyses was between  SORL1  (rs1131497; p = 3.2 × 10 -6 ) and abstract reasoning, and in GEE analyses between  CDH4  (rs1970546; p = 3.7 × 10 -8 ) and TCBV. SORL1  plays a role in amyloid precursor protein processing and has been associated with the risk of AD. Among the 50 strongest associations (25 each by GEE and FBAT) were other biologically interesting genes. Polymorphisms within 28 of 163 candidate genes for stroke, AD and memory impairment were associated with the endophenotypes studied at p < 0.001. We confirmed our previously reported linkage of WMH on chromosome 4 and describe linkage of reading performance to a marker on chromosome 18 (GATA11A06), previously linked to dyslexia (LOD scores = 2.2 and 5.1). Our results suggest that genes associated with clinical neurological disease also have detectable effects on subclinical phenotypes. These hypothesis generating data illustrate the use of an unbiased approach to discover novel pathways that may be involved in brain aging, and could be used to replicate observations made in other studies.\n",
      "1 cognitive test\n",
      "-1 brain\n",
      "-1 aging | aging http://purl.obolibrary.org/obo/GO_0007568\n",
      "-1 Alzheimer's disease ( | alzheimers disease http://www.ebi.ac.uk/efo/EFO_0000249\n",
      "-1 (AD).\n",
      "-1 stroke,\n",
      "-1 , dementia | alzheimers disease http://www.ebi.ac.uk/efo/EFO_0000249\n",
      "-1 Brain\n",
      "1 cognitive\n",
      "-1 tests | pulmonary function measurement http://www.ebi.ac.uk/efo/EFO_0003892\n",
      "-1 can\n",
      "-1 an\n",
      "1 volumetric brain MRI | volumetric brain mri http://www.ebi.ac.uk/efo/EFO_0004868\n",
      "-1 stroke-\n",
      "-1 testing ( | pulmonary function measurement http://www.ebi.ac.uk/efo/EFO_0003892\n",
      "1 cognitive\n",
      "-1 white matter hyperintensity (\n",
      "-1 , reading,\n",
      "-1 tests ( | pulmonary function measurement http://www.ebi.ac.uk/efo/EFO_0003892\n",
      "-1 brain (\n",
      "-1 memory,\n",
      "-1 relate\n",
      "-1 single\n",
      "-1 including\n",
      "1 cognitive\n",
      "-1 reasoning\n",
      "-1 reasoning,\n",
      "-1 was\n",
      "-1 with the\n",
      "-1 AD.\n",
      "-1 protein | protein measurement http://www.ebi.ac.uk/efo/EFO_0004747\n",
      "-1 processing\n",
      "-1 stroke,\n",
      "-1 , AD\n",
      "-1 with the\n",
      "-1 memory\n",
      "-1 on chromosome 18 (\n",
      "-1 reading performance\n",
      "-1 dyslexia (\n",
      "-1 neurological disease\n",
      "-1 subclinical phenotypes\n",
      "-1 use of\n",
      "-1 aging, | aging http://purl.obolibrary.org/obo/GO_0007568\n",
      "-1 an\n",
      "-1 brain\n",
      "\n",
      "17903298\n",
      "diabetes related insulin traits 19592\n",
      "\ta1c measurement http://www.ebi.ac.uk/efo/EFO_0004541 hba1c levels|glycosylated haemoglobin a level|glycosylated hemoglobin a level|glycated hemoglobin levels\n",
      "\tfasting blood insulin measurement http://www.ebi.ac.uk/efo/EFO_0004466 fasting blood insulin|fasting blood insulin level|fasting insulin-related traits\n",
      "\tinsulin sensitivity measurement http://www.ebi.ac.uk/efo/EFO_0004471 insulin sensitivity test\n",
      "\tinsulin resistance http://www.ebi.ac.uk/efo/EFO_0002614 drug resistance to insulin (disorder)\n",
      "diabetes related insulin traits 19593\n",
      "\ta1c measurement http://www.ebi.ac.uk/efo/EFO_0004541 hba1c levels|glycosylated haemoglobin a level|glycosylated hemoglobin a level|glycated hemoglobin levels\n",
      "\tfasting blood insulin measurement http://www.ebi.ac.uk/efo/EFO_0004466 fasting blood insulin|fasting blood insulin level|fasting insulin-related traits\n",
      "\tinsulin sensitivity measurement http://www.ebi.ac.uk/efo/EFO_0004471 insulin sensitivity test\n",
      "\tinsulin resistance http://www.ebi.ac.uk/efo/EFO_0002614 drug resistance to insulin (disorder)\n",
      "fasting plasma glucose 19594\n",
      "\tfasting blood glucose measurement http://www.ebi.ac.uk/efo/EFO_0004465 fasting glucose-related traits|fasting blood glucose level|fasting plasma glucose\n",
      "fasting plasma glucose 19595\n",
      "\tfasting blood glucose measurement http://www.ebi.ac.uk/efo/EFO_0004465 fasting glucose-related traits|fasting blood glucose level|fasting plasma glucose\n",
      "fasting plasma glucose 19596\n",
      "\tfasting blood glucose measurement http://www.ebi.ac.uk/efo/EFO_0004465 fasting glucose-related traits|fasting blood glucose level|fasting plasma glucose\n",
      "fasting plasma glucose 19597\n",
      "\tfasting blood glucose measurement http://www.ebi.ac.uk/efo/EFO_0004465 fasting glucose-related traits|fasting blood glucose level|fasting plasma glucose\n",
      "diabetes related insulin traits 19194\n",
      "\ta1c measurement http://www.ebi.ac.uk/efo/EFO_0004541 hba1c levels|glycosylated haemoglobin a level|glycosylated hemoglobin a level|glycated hemoglobin levels\n",
      "\tfasting blood insulin measurement http://www.ebi.ac.uk/efo/EFO_0004466 fasting blood insulin|fasting blood insulin level|fasting insulin-related traits\n",
      "\tinsulin sensitivity measurement http://www.ebi.ac.uk/efo/EFO_0004471 insulin sensitivity test\n",
      "\tinsulin resistance http://www.ebi.ac.uk/efo/EFO_0002614 drug resistance to insulin (disorder)\n",
      "diabetes related insulin traits 19195\n",
      "\ta1c measurement http://www.ebi.ac.uk/efo/EFO_0004541 hba1c levels|glycosylated haemoglobin a level|glycosylated hemoglobin a level|glycated hemoglobin levels\n",
      "\tfasting blood insulin measurement http://www.ebi.ac.uk/efo/EFO_0004466 fasting blood insulin|fasting blood insulin level|fasting insulin-related traits\n",
      "\tinsulin sensitivity measurement http://www.ebi.ac.uk/efo/EFO_0004471 insulin sensitivity test\n",
      "\tinsulin resistance http://www.ebi.ac.uk/efo/EFO_0002614 drug resistance to insulin (disorder)\n",
      "fasting plasma glucose 19196\n",
      "\tfasting blood glucose measurement http://www.ebi.ac.uk/efo/EFO_0004465 fasting glucose-related traits|fasting blood glucose level|fasting plasma glucose\n",
      "fasting plasma glucose 19197\n",
      "\tfasting blood glucose measurement http://www.ebi.ac.uk/efo/EFO_0004465 fasting glucose-related traits|fasting blood glucose level|fasting plasma glucose\n",
      "diabetes (incident) 19198\n",
      "\tdiabetes mellitus http://www.ebi.ac.uk/efo/EFO_0000400 dm - diabetes mellitus|diabetes mellitus\n",
      "Genome-wide association with diabetes-related traits in the Framingham Heart Study. Susceptibility to type 2 diabetes may be conferred by genetic variants having modest effects on risk. Genome-wide fixed marker arrays offer a novel approach to detect these variants. We used the Affymetrix 100K SNP array in 1,087 Framingham Offspring Study family members to examine genetic associations with three diabetes-related quantitative glucose traits (fasting plasma glucose (FPG), hemoglobin A1c, 28-yr time-averaged FPG (tFPG)), three insulin traits (fasting insulin, HOMA-insulin resistance, and 0–120 min insulin sensitivity index); and with risk for diabetes. We used additive generalized estimating equations (GEE) and family-based association test (FBAT) models to test associations of SNP genotypes with sex-age-age 2 -adjusted residual trait values, and Cox survival models to test incident diabetes. We found 415 SNPs associated (at p < 0.001) with at least one of the six quantitative traits in GEE, 242 in FBAT (18 overlapped with GEE for 639 non-overlapping SNPs), and 128 associated with incident diabetes (31 overlapped with the 639) giving 736 non-overlapping SNPs. Of these 736 SNPs, 439 were within 60 kb of a known gene. Additionally, 53 SNPs (of which 42 had r 2  < 0.80 with each other) had p < 0.01 for incident diabetes AND (all 3 glucose traits OR all 3 insulin traits, OR 2 glucose traits and 2 insulin traits); of these, 36 overlapped with the 736 other SNPs. Of 100K SNPs, one (rs7100927) was in moderate LD (r 2  = 0.50) with  TCF7L2  (rs7903146), and was associated with risk of diabetes (Cox p-value 0.007, additive hazard ratio for diabetes = 1.56) and with tFPG (GEE p-value 0.03). There were no common (MAF > 1%) 100K SNPs in LD (r 2  > 0.05) with  ABCC8  A1369S (rs757110),  KCNJ11  E23K (rs5219), or SNPs in  CAPN10  or  HNFa . PPARG  P12A (rs1801282) was not significantly associated with diabetes or related traits. Framingham 100K SNP data is a resource for association tests of known and novel genes with diabetes and related traits posted at  . Framingham 100K data replicate the  TCF7L2  association with diabetes.\n",
      "-1 diabetes-related traits\n",
      "-1 type 2 diabetes\n",
      "-1 Susceptibility to\n",
      "1 (fasting plasma glucose ( | fasting blood glucose measurement http://www.ebi.ac.uk/efo/EFO_0004465\n",
      "-1 ), hemoglobin | hemoglobin measurement http://www.ebi.ac.uk/efo/EFO_0004509\n",
      "-1 glucose traits\n",
      "1 A1c, | a1c measurement http://www.ebi.ac.uk/efo/EFO_0004541\n",
      "-1 insulin traits\n",
      "-1 insulin, | insulin measurement http://www.ebi.ac.uk/efo/EFO_0004467\n",
      "1 insulin sensitivity | insulin sensitivity measurement http://www.ebi.ac.uk/efo/EFO_0004471\n",
      "-1 residual trait\n",
      "-1 survival\n",
      "-1 six quantitative traits\n",
      "-1 with the\n",
      "-1 (all\n",
      "-1 glucose traits\n",
      "-1 insulin traits\n",
      "-1 glucose traits\n",
      "-1 insulin traits\n",
      "-1 with the\n",
      "-1 all\n",
      "-1 ) was\n",
      "-1 was\n",
      "-1 > 1%)\n",
      "-1 ) was\n",
      "-1 related traits\n",
      "-1 related traits\n",
      "-1 is\n",
      "-1 tests | pulmonary function measurement http://www.ebi.ac.uk/efo/EFO_0003892\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# let's look at a few examples\n",
    "for doc in corpus.documents[:10]:\n",
    "    doc_id = doc.name\n",
    "    print doc.name\n",
    "    for phen_id2 in gold_agg_dict_phen[doc_id]: # iterate over gwc_id's in doc\n",
    "        print agg_id2phen[phen_id2].name, phen_id2 # print its name and id\n",
    "        for eq_phen in agg_id2phen[phen_id2].equivalents: # loop over equivalent phens\n",
    "            print '\\t', eq_phen.name, eq_phen.ontology_ref, eq_phen.synonyms # print name, ref, synonyms\n",
    "    print ' '.join(s.text for s in id2doc[doc_id].sentences)\n",
    "    for candidate in phen_c:\n",
    "        if candidate.context.document != doc: continue\n",
    "#         print gt_dict[candidate.uid], candidate.get_span(), [agg_id2phen[phen_id].name for phen_id in phen2id.get(change_name(candidate.get_span()),set())]\n",
    "        print gt_dict[c2uid(candidate)], candidate.get_span(), \n",
    "        phen_id = phen2id.get(change_name(candidate.get_span()), None)\n",
    "        if phen_id: print '|', id2phen[phen_id].name, id2phen[phen_id].ontology_ref,\n",
    "        print\n",
    "    print"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create training set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 676,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Training set size: 1544\n",
      "Gold set size: 1544\n",
      "Positive labels in training set: 272\n",
      "Negative labels in training set: 1272\n",
      "Positive labels in gold set: 275\n",
      "Negative labels in gold set: 1269\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "\n",
    "# Split into train and test set\n",
    "candidates = phen_c\n",
    "training_candidates = []\n",
    "gold_candidates     = []\n",
    "gold_labels         = []\n",
    "n_half = len(candidates)/2\n",
    "for c in candidates[:n_half]:\n",
    "    uid = c2uid(c)\n",
    "    if uid in gt_dict:\n",
    "        gold_candidates.append(c)\n",
    "        gold_labels.append(gt_dict[uid])\n",
    "    else:\n",
    "        training_candidates.append(c)\n",
    "training_candidates.extend(candidates[n_half:])\n",
    "gold_labels = np.array(gold_labels)\n",
    "print \"Training set size: %s\" % len(training_candidates)\n",
    "print \"Gold set size: %s\" % len(gold_candidates)\n",
    "print \"Positive labels in training set: %s\" % len([c for c in training_candidates if gt_dict.get(c2uid(c),0)==1])\n",
    "print \"Negative labels in training set: %s\" % len([c for c in training_candidates if gt_dict.get(c2uid(c),0)==-1])\n",
    "print \"Positive labels in gold set: %s\" % len([c for c in gold_candidates if gt_dict[c2uid(c)]==1])\n",
    "print \"Negative labels in gold set: %s\" % len([c for c in gold_candidates if gt_dict[c2uid(c)]==-1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 677,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Span(\"Crohn Disease\", context=None, chars=[6,18], words=[1,2])\n",
      "[u'Disease', u'Locus', u'Identified', u'by']\n",
      "Crohn Disease\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "print phen_c[0]\n",
    "print phen_c[0].post_window(d=4)\n",
    "print phen_c[0].get_attrib_span('words')\n",
    "print re.search(r'\\([A-Z]{2,4}\\)', 'Test [AVC)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 723,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from nltk.stem import PorterStemmer\n",
    "import re\n",
    "stemmer = PorterStemmer()\n",
    "\n",
    "# load set of dictionary phenotypes\n",
    "kb = KnowledgeBase()\n",
    "phenotype_list = kb.get_phenotype_candidates() # TODO: load disease names from NCBI\n",
    "phenotype_list = [phenotype for phenotype in phenotype_list]\n",
    "phenotype_set = set(phenotype_list)\n",
    "\n",
    "# small helpers\n",
    "def get_phenotype(entity, stem=False):\n",
    "    phenotype = entity.get_span()\n",
    "    if stem: phenotype = stemmer.stem(phenotype)\n",
    "    return phenotype.lower()\n",
    "\n",
    "def stem_list(L):\n",
    "    return [stemmer.stem(l.lower()) for l in L]\n",
    "\n",
    "def LF_gt(m):\n",
    "    return gt_dict.get(m.uid, 0) if m in training_candidates else 0\n",
    "\n",
    "# positive LFs\n",
    "def LF_first_sentence(m):\n",
    "    return +10 if m.context.position == 0 else 0\n",
    "def LF_from_regex(m):\n",
    "    if m.context.position == 0 and not regex_phen_matcher._f(m) and not LF_bad_words(m): return +5\n",
    "    else: return 0\n",
    "def LF_with_acronym(m):\n",
    "    post_txt = ''.join(m.post_window('words',d=5))\n",
    "    return +1 if re.search(r'\\([A-Z]{2,4}\\)', post_txt) else 0\n",
    "def LF_associated(m):\n",
    "    return +1 if 'associ' in stem_list(m.pre_window('lemmas', d=5)) \\\n",
    "              or 'associ' in stem_list(m.post_window('lemmas', d=5)) \\\n",
    "              else 0\n",
    "def LF_influencing(m):\n",
    "    return +1 if 'influenc' in stem_list(m.pre_window('lemmas', d=5)) \\\n",
    "              or 'influenc' in stem_list(m.post_window('lemmas', d=5)) \\\n",
    "              else 0\n",
    "def LF_identifi(m):\n",
    "    return +1 if 'identifi' in stem_list(m.pre_window('lemmas', d=5)) \\\n",
    "              or 'identifi' in stem_list(m.post_window('lemmas', d=5)) \\\n",
    "              else 0\n",
    "\n",
    "LFs_pos = [LF_first_sentence, LF_with_acronym, LF_from_regex]\n",
    "\n",
    "# negative LFs\n",
    "def LF_bad_words(m):\n",
    "    bad_words = ['disease', 'single', 'map', 'genetic variation']\n",
    "    return -100 if any(m.get_span().lower().startswith(b) for b in bad_words) else 0\n",
    "def LF_short(m):\n",
    "    txt = m.get_attrib_span('words', 3)\n",
    "    return -50 if len(txt) < 5 else 0\n",
    "def LF_no_nouns(m):\n",
    "    return -5 if not any(t.startswith('NN') for t in m.get_attrib_tokens('poses')) else 0\n",
    "\n",
    "def LF_previously(m):\n",
    "    return -1 if 'previously' in m.pre_window('lemmas', 8) else 0\n",
    "def LF_further(m):\n",
    "    lemmas = m.get_attrib_span('lemmas').split(' ')    \n",
    "    return -1 if 'further' in lemmas or 'furthermore' in lemmas else 0\n",
    "def LF_also(m):\n",
    "    words = m.get_span().split(' ')        \n",
    "    return -1 if 'also' in words else 0\n",
    "def LF_recently(m):\n",
    "    words = m.get_span().split(' ')    \n",
    "    return -1 if 'recently' in words else 0\n",
    "def LF_addit(m):\n",
    "    lemmas = m.get_attrib_span('lemmas').split(' ')\n",
    "    return -1 if 'addit' in lemmas else 0\n",
    "def LF_may(m):\n",
    "    words = m.get_span().split(' ')\n",
    "    return -1 if 'may' in words else 0\n",
    "def LF_risk(m):\n",
    "    return +1 if 'risk' in stem_list(m.pre_window('lemmas', d=5)) else 0\n",
    "\n",
    "LFs_neg = [LF_bad_words, LF_short, LF_no_nouns]\n",
    "\n",
    "LFs = LFs_pos + LFs_neg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 724,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Applying LFs...\n",
      "Featurizing...\n",
      "Building feature index...\n",
      "Extracting features...\n",
      "0/28746\n",
      "5000/28746\n",
      "10000/28746\n",
      "15000/28746\n",
      "20000/28746\n",
      "25000/28746\n",
      "============================================================\n",
      "LF Summary Statistics: 6 LFs applied to 1544 candidates\n",
      "------------------------------------------------------------\n",
      "Coverage (candidates w/ > 0 labels):\t\t51.62%\n",
      "Overlap (candidates w/ > 1 labels):\t\t49.03%\n",
      "Conflict (candidates w/ conflicting labels):\t2.01%\n",
      "============================================================\n"
     ]
    }
   ],
   "source": [
    "from snorkel.snorkel import TrainingSet\n",
    "from snorkel.features import NgramFeaturizer\n",
    "\n",
    "training_set = TrainingSet(training_candidates, LFs, featurizer=NgramFeaturizer())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 725,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>conflicts</th>\n",
       "      <th>coverage</th>\n",
       "      <th>j</th>\n",
       "      <th>overlaps</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>LF_first_sentence</th>\n",
       "      <td> 0.168394</td>\n",
       "      <td>  0.673575</td>\n",
       "      <td> 0</td>\n",
       "      <td>  0.673575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LF_with_acronym</th>\n",
       "      <td> 0.003238</td>\n",
       "      <td>  0.029145</td>\n",
       "      <td> 1</td>\n",
       "      <td>  0.003238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LF_from_regex</th>\n",
       "      <td> 0.074482</td>\n",
       "      <td>  0.317358</td>\n",
       "      <td> 2</td>\n",
       "      <td>  0.317358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LF_bad_words</th>\n",
       "      <td> 0.388601</td>\n",
       "      <td>  3.950777</td>\n",
       "      <td> 3</td>\n",
       "      <td>  3.950777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LF_short</th>\n",
       "      <td> 0.615285</td>\n",
       "      <td> 15.738342</td>\n",
       "      <td> 4</td>\n",
       "      <td> 15.738342</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   conflicts   coverage  j   overlaps\n",
       "LF_first_sentence   0.168394   0.673575  0   0.673575\n",
       "LF_with_acronym     0.003238   0.029145  1   0.003238\n",
       "LF_from_regex       0.074482   0.317358  2   0.317358\n",
       "LF_bad_words        0.388601   3.950777  3   3.950777\n",
       "LF_short            0.615285  15.738342  4  15.738342\n",
       "\n",
       "[5 rows x 4 columns]"
      ]
     },
     "execution_count": 725,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lf_stats = training_set.lf_stats()\n",
    "lf_stats[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 726,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from snorkel.snorkel import Learner, PipelinedLearner\n",
    "import snorkel.learning\n",
    "from snorkel.learning import LogReg\n",
    "\n",
    "learner = Learner(training_set, model=snorkel.learning.LogReg())\n",
    "\n",
    "# Splitting into CV and test set\n",
    "n_half = len(gold_candidates)/2\n",
    "test_candidates = gold_candidates[:n_half]\n",
    "test_labels     = gold_labels[:n_half]\n",
    "cv_candidates   = gold_candidates[n_half:]\n",
    "cv_labels       = gold_labels[n_half:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 727,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "============================================================\n",
      "Testing mu = 1.00e-05, lf_w0 = 1.00e+00\n",
      "============================================================\n",
      "Begin training for rate=0.01, mu=1e-05\n",
      "\tLearning epoch = 0\tGradient mag. = 2.550242\n",
      "\tLearning epoch = 250\tGradient mag. = 1.378368\n",
      "\tLearning epoch = 500\tGradient mag. = 0.768510\n",
      "\tLearning epoch = 750\tGradient mag. = 0.490302\n",
      "Final gradient magnitude for rate=0.01, mu=1e-05: 0.359\n",
      "Applying LFs...\n",
      "Featurizing...\n",
      "============================================================\n",
      "Testing mu = 1.00e-05, lf_w0 = 2.00e+00\n",
      "============================================================\n",
      "Begin training for rate=0.01, mu=1e-05\n",
      "\tLearning epoch = 0\tGradient mag. = 2.266565\n",
      "\tLearning epoch = 250\tGradient mag. = 1.224469\n",
      "\tLearning epoch = 500\tGradient mag. = 0.681877\n",
      "\tLearning epoch = 750\tGradient mag. = 0.434128\n",
      "Final gradient magnitude for rate=0.01, mu=1e-05: 0.317\n",
      "============================================================\n",
      "Testing mu = 1.00e-07, lf_w0 = 1.00e+00\n",
      "============================================================\n",
      "Begin training for rate=0.01, mu=1e-07\n",
      "\tLearning epoch = 0\tGradient mag. = 2.550242\n",
      "\tLearning epoch = 250\tGradient mag. = 1.378356\n",
      "\tLearning epoch = 500\tGradient mag. = 0.768486\n",
      "\tLearning epoch = 750\tGradient mag. = 0.490273\n",
      "Final gradient magnitude for rate=0.01, mu=1e-07: 0.359\n",
      "============================================================\n",
      "Testing mu = 1.00e-07, lf_w0 = 2.00e+00\n",
      "============================================================\n",
      "Begin training for rate=0.01, mu=1e-07\n",
      "\tLearning epoch = 0\tGradient mag. = 2.266565\n",
      "\tLearning epoch = 250\tGradient mag. = 1.224455\n",
      "\tLearning epoch = 500\tGradient mag. = 0.681850\n",
      "\tLearning epoch = 750\tGradient mag. = 0.434096\n",
      "Final gradient magnitude for rate=0.01, mu=1e-07: 0.317\n"
     ]
    }
   ],
   "source": [
    "from snorkel.learning_utils import GridSearch\n",
    "\n",
    "gs       = GridSearch(learner, ['mu', 'lf_w0'], [[1e-5, 1e-7],[1.0,2.0]])\n",
    "gs_stats = gs.fit(cv_candidates, cv_labels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 728,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mu</th>\n",
       "      <th>lf_w0</th>\n",
       "      <th>Prec.</th>\n",
       "      <th>Rec.</th>\n",
       "      <th>F1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td> 0.00001</td>\n",
       "      <td> 1</td>\n",
       "      <td> 0.600000</td>\n",
       "      <td> 0.162162</td>\n",
       "      <td> 0.255319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td> 0.00001</td>\n",
       "      <td> 2</td>\n",
       "      <td> 0.609756</td>\n",
       "      <td> 0.168919</td>\n",
       "      <td> 0.264550</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 1</td>\n",
       "      <td> 0.600000</td>\n",
       "      <td> 0.162162</td>\n",
       "      <td> 0.255319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 2</td>\n",
       "      <td> 0.609756</td>\n",
       "      <td> 0.168919</td>\n",
       "      <td> 0.264550</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        mu  lf_w0     Prec.      Rec.        F1\n",
       "0  0.00001      1  0.600000  0.162162  0.255319\n",
       "1  0.00001      2  0.609756  0.168919  0.264550\n",
       "2  0.00000      1  0.600000  0.162162  0.255319\n",
       "3  0.00000      2  0.609756  0.168919  0.264550\n",
       "\n",
       "[4 rows x 5 columns]"
      ]
     },
     "execution_count": 728,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gs_stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 729,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Applying LFs...\n",
      "Featurizing...\n",
      "========================================\n",
      "Test set size:\t772\n",
      "----------------------------------------\n",
      "Precision:\t0.396825396825\n",
      "Recall:\t\t0.568181818182\n",
      "F1 Score:\t0.467289719626\n",
      "----------------------------------------\n",
      "TP: 25 | FP: 38 | TN: 277 | FN: 19\n",
      "========================================\n"
     ]
    }
   ],
   "source": [
    "learner.test_wmv(test_candidates, test_labels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 730,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Applying LFs...\n",
      "Featurizing...\n",
      "17447842 1.0 1\n",
      "Sentence(Document('17447842', Corpus (GWAS Corpus)), 0, u'Novel Crohn Disease Locus Identified by Genome-Wide Association Maps to a Gene Desert on 5p13.1 and Modulates Expression of  PTGER4.')\n",
      "Crohn Disease\n",
      "[10, 0, 0, 0, 0, 0]\n",
      "\n",
      "17658951 1.0 1\n",
      "Sentence(Document('17658951', Corpus (GWAS Corpus)), 0, u'Genome-Wide Association Scan Shows Genetic Variants in the  FTO  Gene Are Associated with Obesity-Related Traits.')\n",
      "Obesity-Related Traits\n",
      "[10, 0, 0, 0, 0, 0]\n",
      "\n",
      "17684544 1.0 -1\n",
      "Sentence(Document('17684544', Corpus (GWAS Corpus)), 0, u'Systematic Association Mapping Identifies  NELL1  as a Novel IBD Disease Gene.')\n",
      "IBD Disease\n",
      "[10, 0, 0, 0, 0, 0]\n",
      "\n",
      "17903292 1.0 -1\n",
      "Sentence(Document('17903292', Corpus (GWAS Corpus)), 0, u\"A genome-wide association for kidney function and endocrine-related traits in the NHLBI's Framingham Heart Study.\")\n",
      "endocrine-related traits\n",
      "[10, 0, 0, 0, 0, 0]\n",
      "\n",
      "17903293 1.0 -1\n",
      "Sentence(Document('17903293', Corpus (GWAS Corpus)), 0, u'Genome-wide association with select biomarker traits in the Framingham Heart Study.')\n",
      "biomarker traits\n",
      "[10, 0, 0, 0, 0, 0]\n",
      "\n",
      "17903294 1.0 -1\n",
      "Sentence(Document('17903294', Corpus (GWAS Corpus)), 0, u'Genome-wide association and linkage analyses of hemostatic factors and hematological phenotypes in the Framingham Heart Study.')\n",
      "hematological phenotypes\n",
      "[10, 0, 0, 0, 0, 0]\n",
      "\n",
      "17903295 1.0 -1\n",
      "Sentence(Document('17903295', Corpus (GWAS Corpus)), 0, u'Genetic correlates of longevity and selected age-related phenotypes: a genome-wide association study in the Framingham Study.')\n",
      "age-related phenotypes\n",
      "[10, 0, 0, 0, 0, 0]\n",
      "\n",
      "17903295 1.0 -1\n",
      "Sentence(Document('17903295', Corpus (GWAS Corpus)), 0, u'Genetic correlates of longevity and selected age-related phenotypes: a genome-wide association study in the Framingham Study.')\n",
      "longevity\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "17903297 1.0 1\n",
      "Sentence(Document('17903297', Corpus (GWAS Corpus)), 0, u'Genetic correlates of brain aging on MRI and cognitive test measures: a genome-wide association and linkage analysis in the Framingham study.')\n",
      "cognitive test\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "17903297 1.0 -1\n",
      "Sentence(Document('17903297', Corpus (GWAS Corpus)), 0, u'Genetic correlates of brain aging on MRI and cognitive test measures: a genome-wide association and linkage analysis in the Framingham study.')\n",
      "brain\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "17903297 1.0 -1\n",
      "Sentence(Document('17903297', Corpus (GWAS Corpus)), 0, u'Genetic correlates of brain aging on MRI and cognitive test measures: a genome-wide association and linkage analysis in the Framingham study.')\n",
      "aging\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "17903298 1.0 -1\n",
      "Sentence(Document('17903298', Corpus (GWAS Corpus)), 0, u'Genome-wide association with diabetes-related traits in the Framingham Heart Study.')\n",
      "diabetes-related traits\n",
      "[10, 0, 0, 0, 0, 0]\n",
      "\n",
      "17903300 1.0 1\n",
      "Sentence(Document('17903300', Corpus (GWAS Corpus)), 0, u'Genome-wide association to body mass index and waist circumference: the Framingham Heart Study 100K project.')\n",
      "body mass index\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "17903300 1.0 1\n",
      "Sentence(Document('17903300', Corpus (GWAS Corpus)), 0, u'Genome-wide association to body mass index and waist circumference: the Framingham Heart Study 100K project.')\n",
      "waist circumference:\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "17903301 1.0 -1\n",
      "Sentence(Document('17903301', Corpus (GWAS Corpus)), 0, u'Genome-wide association of echocardiographic dimensions, brachial artery endothelial function and treadmill exercise responses in the Framingham Heart Study.')\n",
      "exercise responses\n",
      "[10, 0, 0, 0, 0, 0]\n",
      "\n",
      "17903302 1.0 -1\n",
      "Sentence(Document('17903302', Corpus (GWAS Corpus)), 0, u'Framingham Heart Study 100K Project: genome-wide associations for blood pressure and arterial stiffness.')\n",
      "arterial stiffness.\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "17903302 1.0 1\n",
      "Sentence(Document('17903302', Corpus (GWAS Corpus)), 0, u'Framingham Heart Study 100K Project: genome-wide associations for blood pressure and arterial stiffness.')\n",
      "blood pressure\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "17903303 1.0 -1\n",
      "Sentence(Document('17903303', Corpus (GWAS Corpus)), 0, u\"Genome-wide association study for subclinical atherosclerosis in major arterial territories in the NHLBI's Framingham Heart Study.\")\n",
      "subclinical\n",
      "[10, 0, 5, 0, 0, -5]\n",
      "\n",
      "17903303 1.0 -1\n",
      "Sentence(Document('17903303', Corpus (GWAS Corpus)), 0, u\"Genome-wide association study for subclinical atherosclerosis in major arterial territories in the NHLBI's Framingham Heart Study.\")\n",
      "atherosclerosis\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "17903304 1.0 1\n",
      "Sentence(Document('17903304', Corpus (GWAS Corpus)), 0, u'Framingham Heart Study 100K project: genome-wide associations for cardiovascular disease outcomes.')\n",
      "cardiovascular disease\n",
      "[10, 0, 0, 0, 0, 0]\n",
      "\n",
      "17903305 1.0 1\n",
      "Sentence(Document('17903305', Corpus (GWAS Corpus)), 0, u\"A genome-wide association study of breast and prostate cancer in the NHLBI's Framingham Heart Study.\")\n",
      "prostate cancer\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "17903306 1.0 1\n",
      "Sentence(Document('17903306', Corpus (GWAS Corpus)), 0, u'Genome-wide association study of electrocardiographic and heart rate variability traits: the Framingham Heart Study.')\n",
      "heart rate\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "17903306 1.0 -1\n",
      "Sentence(Document('17903306', Corpus (GWAS Corpus)), 0, u'Genome-wide association study of electrocardiographic and heart rate variability traits: the Framingham Heart Study.')\n",
      "variability traits\n",
      "[10, 0, 0, 0, 0, 0]\n",
      "\n",
      "17903307 1.0 1\n",
      "Sentence(Document('17903307', Corpus (GWAS Corpus)), 0, u'Framingham Heart Study genome-wide association: results for pulmonary function measures.')\n",
      "pulmonary function\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "17903308 1.0 -1\n",
      "Sentence(Document('17903308', Corpus (GWAS Corpus)), 0, u'Genome-wide association of sleep and circadian phenotypes.')\n",
      "circadian phenotypes\n",
      "[10, 0, 0, 0, 0, 0]\n",
      "\n",
      "17903308 1.0 1\n",
      "Sentence(Document('17903308', Corpus (GWAS Corpus)), 0, u'Genome-wide association of sleep and circadian phenotypes.')\n",
      "sleep\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "17997608 1.0 -1\n",
      "Sentence(Document('17997608', Corpus (GWAS Corpus)), 0, u'The  GLUT9  Gene Is Associated with Serum Uric Acid Levels in Sardinia and Chianti Cohorts.')\n",
      "Uric Acid Levels\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "18159244 1.0 -1\n",
      "Sentence(Document('18159244', Corpus (GWAS Corpus)), 0, u'Genome Wide Association (GWA) Study for Early Onset Extreme Obesity Supports the Role of Fat Mass and Obesity Associated Gene  (FTO)  Variants.')\n",
      "Onset\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "18159244 1.0 1\n",
      "Sentence(Document('18159244', Corpus (GWAS Corpus)), 0, u'Genome Wide Association (GWA) Study for Early Onset Extreme Obesity Supports the Role of Fat Mass and Obesity Associated Gene  (FTO)  Variants.')\n",
      "Obesity\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "18159244 1.0 1\n",
      "Sentence(Document('18159244', Corpus (GWAS Corpus)), 0, u'Genome Wide Association (GWA) Study for Early Onset Extreme Obesity Supports the Role of Fat Mass and Obesity Associated Gene  (FTO)  Variants.')\n",
      "Obesity\n",
      "[10, 1, 5, 0, 0, 0]\n",
      "\n",
      "18282107 1.0 1\n",
      "Sentence(Document('18282107', Corpus (GWAS Corpus)), 0, u'Genome-Wide Association Identifies a Common Variant in the Reelin Gene That Increases the Risk of Schizophrenia Only in Women.')\n",
      "Schizophrenia\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "18369459 1.0 -1\n",
      "Sentence(Document('18369459', Corpus (GWAS Corpus)), 0, u'A Genome-Wide Association Study of Psoriasis and Psoriatic Arthritis Identifies New Disease Loci.')\n",
      "Psoriatic Arthritis\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "18369459 1.0 -1\n",
      "Sentence(Document('18369459', Corpus (GWAS Corpus)), 0, u'A Genome-Wide Association Study of Psoriasis and Psoriatic Arthritis Identifies New Disease Loci.')\n",
      "New Disease\n",
      "[10, 0, 0, 0, 0, 0]\n",
      "\n",
      "18369459 1.0 1\n",
      "Sentence(Document('18369459', Corpus (GWAS Corpus)), 0, u'A Genome-Wide Association Study of Psoriasis and Psoriatic Arthritis Identifies New Disease Loci.')\n",
      "Psoriasis\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "18455228 1.0 1\n",
      "Sentence(Document('18455228', Corpus (GWAS Corpus)), 0, u'Bone mineral density, osteoporosis, and osteoporotic fractures: a genome-wide association study.')\n",
      "Bone mineral density,\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "18455228 1.0 -1\n",
      "Sentence(Document('18455228', Corpus (GWAS Corpus)), 0, u'Bone mineral density, osteoporosis, and osteoporotic fractures: a genome-wide association study.')\n",
      ", osteoporosis,\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "18455228 1.0 -1\n",
      "Sentence(Document('18455228', Corpus (GWAS Corpus)), 0, u'Bone mineral density, osteoporosis, and osteoporotic fractures: a genome-wide association study.')\n",
      "osteoporotic fractures:\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "18464913 1.0 -1\n",
      "Sentence(Document('18464913', Corpus (GWAS Corpus)), 0, u'A Genome-Wide Association Study Identifies Protein Quantitative Trait Loci (pQTLs).')\n",
      "Protein Quantitative Trait\n",
      "[10, 0, 0, 0, 0, 0]\n",
      "\n",
      "18483556 1.0 -1\n",
      "Sentence(Document('18483556', Corpus (GWAS Corpus)), 0, u'A Genome-Wide Association Study Identifies Novel Alleles Associated with Hair Color and Skin Pigmentation.')\n",
      "Skin Pigmentation.\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "18483556 1.0 1\n",
      "Sentence(Document('18483556', Corpus (GWAS Corpus)), 0, u'A Genome-Wide Association Study Identifies Novel Alleles Associated with Hair Color and Skin Pigmentation.')\n",
      "Hair Color\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "18604267 1.0 1\n",
      "Sentence(Document('18604267', Corpus (GWAS Corpus)), 0, u'Novel Association of ABO Histo-Blood Group Antigen with Soluble ICAM-1: Results of a Genome-Wide Association Study of 6,578 Women.')\n",
      "ICAM-1:\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "18604267 1.0 -1\n",
      "Sentence(Document('18604267', Corpus (GWAS Corpus)), 0, u'Novel Association of ABO Histo-Blood Group Antigen with Soluble ICAM-1: Results of a Genome-Wide Association Study of 6,578 Women.')\n",
      "Antigen\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "18776929 1.0 1\n",
      "Sentence(Document('18776929', Corpus (GWAS Corpus)), 0, u'Identification of  PLCL1  Gene for Hip Bone Size Variation in Females in a Genome-Wide Association Study.')\n",
      "Hip Bone Size\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "18823527 1.0 1\n",
      "Sentence(Document('18823527', Corpus (GWAS Corpus)), 0, u\"A genome-wide association study for late-onset Alzheimer's disease using DNA pooling.\")\n",
      "late-onset Alzheimer's disease\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "18840781 1.0 1\n",
      "Sentence(Document('18840781', Corpus (GWAS Corpus)), 0, u'Follow-Up Analysis of Genome-Wide Association Data Identifies Novel Loci for Type 1 Diabetes.')\n",
      "Type 1 Diabetes.\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "18846228 1.0 1\n",
      "Sentence(Document('18846228', Corpus (GWAS Corpus)), 0, u'Genome-Wide Scan on Total Serum IgE Levels Identifies  FCER1A  as Novel Susceptibility Locus.')\n",
      "Serum IgE Levels\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "18941528 1.0 1\n",
      "Sentence(Document('18941528', Corpus (GWAS Corpus)), 0, u'Identification of a Novel Risk Locus for Multiple Sclerosis at 13q31.3 by a Pooled Genome-Wide Scan of 500,000 Single Nucleotide Polymorphisms.')\n",
      "Multiple Sclerosis at\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19043545 1.0 1\n",
      "Sentence(Document('19043545', Corpus (GWAS Corpus)), 0, u'Genetics Meets Metabolomics: A Genome-Wide Association Study of Metabolite Profiles in Human Serum.')\n",
      "Metabolite\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19056611 1.0 1\n",
      "Sentence(Document('19056611', Corpus (GWAS Corpus)), 0, u'Adiposity-Related Heterogeneity in Patterns of Type 2 Diabetes Susceptibility Observed in Genome-Wide Association Data.')\n",
      "Type 2 Diabetes\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19081515 1.0 -1\n",
      "Sentence(Document('19081515', Corpus (GWAS Corpus)), 0, u'Genetic risk factors for variant Creutzfeldt\\u2013Jakob disease: a genome-wide association study.')\n",
      "Jakob disease\n",
      "[10, 0, 0, 0, 0, 0]\n",
      "\n",
      "19096518 1.0 1\n",
      "Sentence(Document('19096518', Corpus (GWAS Corpus)), 0, u\"Novel Association of HK1 with Glycated Hemoglobin in a Non-Diabetic Population: A Genome-Wide Evaluation of 14,618 Participants in the Women's Genome Health Study.\")\n",
      "Glycated Hemoglobin\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19116933 1.0 1\n",
      "Sentence(Document('19116933', Corpus (GWAS Corpus)), 0, u'Identification of a novel susceptibility locus for juvenile idiopathic arthritis by genome-wide association analysis.')\n",
      "juvenile idiopathic arthritis\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19122664 1.0 1\n",
      "Sentence(Document('19122664', Corpus (GWAS Corpus)), 0, u'Ulcerative colitis loci on chromosomes 1p36 and 12q15 identified by genome-wide association study.')\n",
      "Ulcerative colitis\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19132087 1.0 1\n",
      "Sentence(Document('19132087', Corpus (GWAS Corpus)), 0, u'A Genome-Wide Association Study Identifies Novel and Functionally Related Susceptibility Loci for Kawasaki Disease.')\n",
      "Kawasaki Disease.\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19169254 1.0 1\n",
      "Sentence(Document('19169254', Corpus (GWAS Corpus)), 0, u'Genomewide Scan Reveals Association of Psoriasis with IL-23 and NF-\\u03baB Pathways.')\n",
      "Psoriasis\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19219042 1.0 1\n",
      "Sentence(Document('19219042', Corpus (GWAS Corpus)), 0, u'Genome-wide association study identifies a novel breast cancer susceptibility locus at 6q25.1.')\n",
      "breast cancer\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19247474 1.0 1\n",
      "Sentence(Document('19247474', Corpus (GWAS Corpus)), 0, u'Genome-Wide and Candidate Gene Association Study of Cigarette Smoking Behaviors.')\n",
      "Smoking Behaviors.\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19300482 1.0 1\n",
      "Sentence(Document('19300482', Corpus (GWAS Corpus)), 0, u'A Genome-Wide Association Study in Chronic Obstructive Pulmonary Disease (COPD): Identification of Two Major Susceptibility Loci A Genome-Wide Association Study of Pulmonary Function Measures in the Framingham Heart Study.')\n",
      "Chronic Obstructive Pulmonary Disease (\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19300482 1.0 -1\n",
      "Sentence(Document('19300482', Corpus (GWAS Corpus)), 0, u'A Genome-Wide Association Study in Chronic Obstructive Pulmonary Disease (COPD): Identification of Two Major Susceptibility Loci A Genome-Wide Association Study of Pulmonary Function Measures in the Framingham Heart Study.')\n",
      "(COPD):\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19300482 1.0 -1\n",
      "Sentence(Document('19300482', Corpus (GWAS Corpus)), 0, u'A Genome-Wide Association Study in Chronic Obstructive Pulmonary Disease (COPD): Identification of Two Major Susceptibility Loci A Genome-Wide Association Study of Pulmonary Function Measures in the Framingham Heart Study.')\n",
      "Pulmonary Function\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19300499 1.0 -1\n",
      "Sentence(Document('19300499', Corpus (GWAS Corpus)), 0, u'A Genome-Wide Association Study Confirms  VKORC1 ,  CYP2C9 , and  CYP4F2  as Principal Genetic Determinants of Warfarin Dose.')\n",
      "Warfarin\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19300500 1.0 -1\n",
      "Sentence(Document('19300500', Corpus (GWAS Corpus)), 0, u'A Genome-Wide Association Study of Pulmonary Function Measures in the Framingham Heart Study A Genome-Wide Association Study in Chronic Obstructive Pulmonary Disease (COPD): Identification of Two Major Susceptibility Loci.')\n",
      "Chronic Obstructive Pulmonary Disease (\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19300500 1.0 -1\n",
      "Sentence(Document('19300500', Corpus (GWAS Corpus)), 0, u'A Genome-Wide Association Study of Pulmonary Function Measures in the Framingham Heart Study A Genome-Wide Association Study in Chronic Obstructive Pulmonary Disease (COPD): Identification of Two Major Susceptibility Loci.')\n",
      "(COPD):\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19300500 1.0 1\n",
      "Sentence(Document('19300500', Corpus (GWAS Corpus)), 0, u'A Genome-Wide Association Study of Pulmonary Function Measures in the Framingham Heart Study A Genome-Wide Association Study in Chronic Obstructive Pulmonary Disease (COPD): Identification of Two Major Susceptibility Loci.')\n",
      "Pulmonary Function\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19304780 1.0 1\n",
      "Sentence(Document('19304780', Corpus (GWAS Corpus)), 0, u'Genome-wide scan identifies  CDH13  as a novel susceptibility locus contributing to blood pressure determination in two European populations.')\n",
      "blood pressure\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19359265 1.0 -1\n",
      "Sentence(Document('19359265', Corpus (GWAS Corpus)), 0, u'A genome-wide association study identifies a novel locus on chromosome 18q12.2 influencing white cell telomere length.')\n",
      "locus on chromosome\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19359265 1.0 1\n",
      "Sentence(Document('19359265', Corpus (GWAS Corpus)), 0, u'A genome-wide association study identifies a novel locus on chromosome 18q12.2 influencing white cell telomere length.')\n",
      "telomere length.\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19401414 1.0 1\n",
      "Sentence(Document('19401414', Corpus (GWAS Corpus)), 0, u'Confirmation of Multiple Risk Loci and Genetic Impacts by a Genome-Wide Association Study of Type 2 Diabetes in the Japanese Population.')\n",
      "Type 2 Diabetes\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19401414 1.0 -1\n",
      "Sentence(Document('19401414', Corpus (GWAS Corpus)), 0, u'Confirmation of Multiple Risk Loci and Genetic Impacts by a Genome-Wide Association Study of Type 2 Diabetes in the Japanese Population.')\n",
      "Multiple\n",
      "[10, 0, 5, 0, 0, -5]\n",
      "\n",
      "19412175 1.0 -1\n",
      "Sentence(Document('19412175', Corpus (GWAS Corpus)), 0, u'Common variations in  BARD1  influence susceptibility to high-risk neuroblastoma.')\n",
      "susceptibility to\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19412175 1.0 1\n",
      "Sentence(Document('19412175', Corpus (GWAS Corpus)), 0, u'Common variations in  BARD1  influence susceptibility to high-risk neuroblastoma.')\n",
      "neuroblastoma.\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19412176 1.0 -1\n",
      "Sentence(Document('19412176', Corpus (GWAS Corpus)), 0, u'Narcolepsy is strongly associated with the TCR alpha locus.')\n",
      "with the\n",
      "[10, 0, 5, 0, 0, -5]\n",
      "\n",
      "19412176 1.0 1\n",
      "Sentence(Document('19412176', Corpus (GWAS Corpus)), 0, u'Narcolepsy is strongly associated with the TCR alpha locus.')\n",
      "Narcolepsy\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19421330 1.0 -1\n",
      "Sentence(Document('19421330', Corpus (GWAS Corpus)), 0, u'Genome-Wide Association Study of Young-Onset Hypertension in the Han Chinese Population of Taiwan.')\n",
      "Hypertension\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19430480 1.0 1\n",
      "Sentence(Document('19430480', Corpus (GWAS Corpus)), 0, u'Genome-wide association study and meta-analysis finds over 40 loci affect risk of type 1 diabetes.')\n",
      "type 1 diabetes.\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19430483 1.0 -1\n",
      "Sentence(Document('19430483', Corpus (GWAS Corpus)), 0, u'Eight blood pressure loci identified by genome-wide association study of 34,433 people of European ancestry.')\n",
      "blood pressure\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19448621 1.0 1\n",
      "Sentence(Document('19448621', Corpus (GWAS Corpus)), 0, u'Genome-wide association studies identify novel loci associated with age at menarche and age at natural menopause.')\n",
      "age at menarche\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19448621 1.0 1\n",
      "Sentence(Document('19448621', Corpus (GWAS Corpus)), 0, u'Genome-wide association studies identify novel loci associated with age at menarche and age at natural menopause.')\n",
      "menopause.\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19454037 1.0 -1\n",
      "Sentence(Document('19454037', Corpus (GWAS Corpus)), 0, u'Genome-wide association study identifies single-nucleotide polymorphism in  KCNB1  associated with left ventricular mass in humans: The HyperGEN Study.')\n",
      "with left ventricular\n",
      "[10, 0, 5, 0, 0, -5]\n",
      "\n",
      "19503088 1.0 1\n",
      "Sentence(Document('19503088', Corpus (GWAS Corpus)), 0, u'REL, a member of the NF-\\u03baB family of transcription factors, is a newly defined risk locus for rheumatoid arthritis.')\n",
      "rheumatoid arthritis.\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19503597 1.0 -1\n",
      "Sentence(Document('19503597', Corpus (GWAS Corpus)), 0, u'Meta-Analysis of 28,141 Individuals Identifies Common Variants within Five New Loci That Influence Uric Acid Concentrations.')\n",
      "Uric Acid Concentrations.\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19557161 1.0 -1\n",
      "Sentence(Document('19557161', Corpus (GWAS Corpus)), 0, u'Genome-Wide Association Scan Meta-Analysis Identifies Three Loci Influencing Adiposity and Fat Distribution NRXN3  is a Novel Locus for Waist Circumference: A Genome-Wide Association Study from the CHARGE Consortium.')\n",
      "Waist Circumference:\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19557161 1.0 1\n",
      "Sentence(Document('19557161', Corpus (GWAS Corpus)), 0, u'Genome-Wide Association Scan Meta-Analysis Identifies Three Loci Influencing Adiposity and Fat Distribution NRXN3  is a Novel Locus for Waist Circumference: A Genome-Wide Association Study from the CHARGE Consortium.')\n",
      "Adiposity\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19557197 1.0 1\n",
      "Sentence(Document('19557197', Corpus (GWAS Corpus)), 0, u'NRXN3  Is a Novel Locus for Waist Circumference: A Genome-Wide Association Study from the CHARGE Consortium Genome-Wide Association Scan Meta-Analysis Identifies Three Loci Influencing Adiposity and Fat Distribution.')\n",
      "Waist Circumference:\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19557197 1.0 -1\n",
      "Sentence(Document('19557197', Corpus (GWAS Corpus)), 0, u'NRXN3  Is a Novel Locus for Waist Circumference: A Genome-Wide Association Study from the CHARGE Consortium Genome-Wide Association Scan Meta-Analysis Identifies Three Loci Influencing Adiposity and Fat Distribution.')\n",
      "Adiposity\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19570815 1.0 -1\n",
      "Sentence(Document('19570815', Corpus (GWAS Corpus)), 0, u'A genome-wide association study of northwestern Europeans involves the C-type natriuretic peptide signaling pathway in the etiology of human height variation.')\n",
      "natriuretic peptide\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19570815 1.0 1\n",
      "Sentence(Document('19570815', Corpus (GWAS Corpus)), 0, u'A genome-wide association study of northwestern Europeans involves the C-type natriuretic peptide signaling pathway in the etiology of human height variation.')\n",
      "height\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19571809 1.0 1\n",
      "Sentence(Document('19571809', Corpus (GWAS Corpus)), 0, u'Common variants on chromosome 6p22.1 are associated with schizophrenia.')\n",
      "schizophrenia.\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19578366 1.0 -1\n",
      "Sentence(Document('19578366', Corpus (GWAS Corpus)), 0, u'Variants in the  CDKN2B  and  RTEL1  regions are associated with high grade glioma susceptibility.')\n",
      "glioma\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19597492 1.0 1\n",
      "Sentence(Document('19597492', Corpus (GWAS Corpus)), 0, u'Variants in  ZFHX3  are associated with atrial fibrillation in individuals of European ancestry.')\n",
      "atrial fibrillation\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19609347 1.0 1\n",
      "Sentence(Document('19609347', Corpus (GWAS Corpus)), 0, u'A Genome-Wide Association Study of Hypertension and Blood Pressure in African Americans.')\n",
      "Blood Pressure\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19609347 1.0 1\n",
      "Sentence(Document('19609347', Corpus (GWAS Corpus)), 0, u'A Genome-Wide Association Study of Hypertension and Blood Pressure in African Americans.')\n",
      "Hypertension\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19651812 1.0 -1\n",
      "Sentence(Document('19651812', Corpus (GWAS Corpus)), 0, u'Common Genetic Variation Near Melatonin Receptor  MTNR1B  Contributes to Raised Plasma Glucose and Increased Risk of Type 2 Diabetes Among Indian Asians and European Caucasians.')\n",
      "Type 2 Diabetes\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19651812 1.0 -1\n",
      "Sentence(Document('19651812', Corpus (GWAS Corpus)), 0, u'Common Genetic Variation Near Melatonin Receptor  MTNR1B  Contributes to Raised Plasma Glucose and Increased Risk of Type 2 Diabetes Among Indian Asians and European Caucasians.')\n",
      "Melatonin\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19651812 1.0 -1\n",
      "Sentence(Document('19651812', Corpus (GWAS Corpus)), 0, u'Common Genetic Variation Near Melatonin Receptor  MTNR1B  Contributes to Raised Plasma Glucose and Increased Risk of Type 2 Diabetes Among Indian Asians and European Caucasians.')\n",
      "Glucose\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19668339 1.0 -1\n",
      "Sentence(Document('19668339', Corpus (GWAS Corpus)), 0, u\"Hippocampal Atrophy as a Quantitative Trait in a Genome-Wide Association Study Identifying Novel Susceptibility Genes for Alzheimer's Disease.\")\n",
      "as a Quantitative Trait\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19668339 1.0 -1\n",
      "Sentence(Document('19668339', Corpus (GWAS Corpus)), 0, u\"Hippocampal Atrophy as a Quantitative Trait in a Genome-Wide Association Study Identifying Novel Susceptibility Genes for Alzheimer's Disease.\")\n",
      "Alzheimer's Disease.\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19668339 1.0 1\n",
      "Sentence(Document('19668339', Corpus (GWAS Corpus)), 0, u\"Hippocampal Atrophy as a Quantitative Trait in a Genome-Wide Association Study Identifying Novel Susceptibility Genes for Alzheimer's Disease.\")\n",
      "Hippocampal Atrophy\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19714205 1.0 -1\n",
      "Sentence(Document('19714205', Corpus (GWAS Corpus)), 0, u'Genome-Wide Association Study Implicates Chromosome 9q21.31 as a Susceptibility Locus for Asthma in Mexican Children.')\n",
      "Asthma\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19714249 1.0 -1\n",
      "Sentence(Document('19714249', Corpus (GWAS Corpus)), 0, u'Powerful Bivariate Genome-Wide Association Analyses Suggest the  SOX6  Gene Influencing Both Obesity and Osteoporosis Phenotypes in Males.')\n",
      "Osteoporosis Phenotypes\n",
      "[10, 0, 0, 0, 0, 0]\n",
      "\n",
      "19714249 1.0 1\n",
      "Sentence(Document('19714249', Corpus (GWAS Corpus)), 0, u'Powerful Bivariate Genome-Wide Association Analyses Suggest the  SOX6  Gene Influencing Both Obesity and Osteoporosis Phenotypes in Males.')\n",
      "Obesity\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19721433 1.0 -1\n",
      "Sentence(Document('19721433', Corpus (GWAS Corpus)), 0, u'Genomewide Pharmacogenomic Analysis of Response to Treatment with Antipsychotics.')\n",
      "of Response\n",
      "[10, 0, 0, 0, 0, 0]\n",
      "\n",
      "19721433 1.0 -1\n",
      "Sentence(Document('19721433', Corpus (GWAS Corpus)), 0, u'Genomewide Pharmacogenomic Analysis of Response to Treatment with Antipsychotics.')\n",
      "Response to\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19721433 1.0 -1\n",
      "Sentence(Document('19721433', Corpus (GWAS Corpus)), 0, u'Genomewide Pharmacogenomic Analysis of Response to Treatment with Antipsychotics.')\n",
      "Antipsychotics\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19734902 1.0 1\n",
      "Sentence(Document('19734902', Corpus (GWAS Corpus)), 0, u\"Genome-wide association study identifies variants at  CLU  and  PICALM  associated with Alzheimer's disease, and shows evidence for additional susceptibility genes.\")\n",
      "Alzheimer's disease,\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19772629 1.0 1\n",
      "Sentence(Document('19772629', Corpus (GWAS Corpus)), 0, u'Genomewide association study for onset age in Parkinson disease.')\n",
      "Parkinson disease.\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19772629 1.0 1\n",
      "Sentence(Document('19772629', Corpus (GWAS Corpus)), 0, u'Genomewide association study for onset age in Parkinson disease.')\n",
      "onset age\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19779542 1.0 1\n",
      "Sentence(Document('19779542', Corpus (GWAS Corpus)), 0, u'A Genome-Wide Association Analysis Identified a Novel Susceptible Locus for Pathological Myopia at 11q24.1.')\n",
      "Pathological Myopia at\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19798445 1.0 1\n",
      "Sentence(Document('19798445', Corpus (GWAS Corpus)), 0, u'Genetic Determinants of Circulating Sphingolipid Concentrations in European Populations.')\n",
      "Sphingolipid\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19801982 1.0 -1\n",
      "Sentence(Document('19801982', Corpus (GWAS Corpus)), 0, u'Twenty bone mineral density loci identified by large-scale meta-analysis of genome-wide association studies.')\n",
      "bone mineral density\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19812673 1.0 1\n",
      "Sentence(Document('19812673', Corpus (GWAS Corpus)), 0, u'A GENOME-WIDE LINKAGE AND ASSOCIATION SCAN REVEALS NOVEL LOCI FOR AUTISM.')\n",
      "AUTISM.\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19820699 1.0 -1\n",
      "Sentence(Document('19820699', Corpus (GWAS Corpus)), 0, u'Common variants in  TMPRSS6  are associated with iron status and erythrocyte volume.')\n",
      "erythrocyte\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19874204 1.0 -1\n",
      "Sentence(Document('19874204', Corpus (GWAS Corpus)), 0, u'IL21R  and  PTH  May Underlie Variation of Femoral Neck Bone Mineral Density as Revealed by a Genome-wide Association Study.')\n",
      "Femoral Neck Bone\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19874204 1.0 1\n",
      "Sentence(Document('19874204', Corpus (GWAS Corpus)), 0, u'IL21R  and  PTH  May Underlie Variation of Femoral Neck Bone Mineral Density as Revealed by a Genome-wide Association Study.')\n",
      "Bone Mineral Density\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19915575 1.0 1\n",
      "Sentence(Document('19915575', Corpus (GWAS Corpus)), 0, u'Genome-Wide Association Study reveals genetic risk underlying Parkinson\\u2019s disease.')\n",
      "Parkinson’s disease.\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19936222 1.0 -1\n",
      "Sentence(Document('19936222', Corpus (GWAS Corpus)), 0, u'Forty-Three Loci Associated with Plasma Lipoprotein Size, Concentration, and Cholesterol Content in Genome-Wide Analysis.')\n",
      "Lipoprotein\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19936222 1.0 -1\n",
      "Sentence(Document('19936222', Corpus (GWAS Corpus)), 0, u'Forty-Three Loci Associated with Plasma Lipoprotein Size, Concentration, and Cholesterol Content in Genome-Wide Analysis.')\n",
      "Cholesterol\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19961619 1.0 1\n",
      "Sentence(Document('19961619', Corpus (GWAS Corpus)), 0, u'A pooling-based genome-wide analysis identifies new potential candidate genes for atopy in the European Community Respiratory Health Survey (ECRHS).')\n",
      "atopy\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19966805 1.0 -1\n",
      "Sentence(Document('19966805', Corpus (GWAS Corpus)), 0, u'The imprinted DLK1-MEG3 gene region on chromosome 14q32.2 alters susceptibility to type 1 diabetes.')\n",
      "susceptibility to type 1\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "19966805 1.0 1\n",
      "Sentence(Document('19966805', Corpus (GWAS Corpus)), 0, u'The imprinted DLK1-MEG3 gene region on chromosome 14q32.2 alters susceptibility to type 1 diabetes.')\n",
      "type 1 diabetes.\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "20011104 1.0 1\n",
      "Sentence(Document('20011104', Corpus (GWAS Corpus)), 0, u'A Genome-Wide Association Study Reveals Variants in  ARL15  that Influence Adiponectin Levels.')\n",
      "Adiponectin Levels.\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "20038947 1.0 -1\n",
      "Sentence(Document('20038947', Corpus (GWAS Corpus)), 0, u'Novel loci for major depression identified by genome-wide association study of STAR*D and meta-analysis of three studies.')\n",
      "major depression\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "20038947 1.0 -1\n",
      "Sentence(Document('20038947', Corpus (GWAS Corpus)), 0, u'Novel loci for major depression identified by genome-wide association study of STAR*D and meta-analysis of three studies.')\n",
      "STAR*\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "20041166 1.0 1\n",
      "Sentence(Document('20041166', Corpus (GWAS Corpus)), 0, u'Common Genetic Variation and the Control of HIV-1 in Humans.')\n",
      "HIV-1\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "20062060 1.0 1\n",
      "Sentence(Document('20062060', Corpus (GWAS Corpus)), 0, u'Genome-wide association study of PR interval.')\n",
      "PR interval.\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "20066028 1.0 1\n",
      "Sentence(Document('20066028', Corpus (GWAS Corpus)), 0, u'Modeling of Environmental Effects in Genome-Wide Association Studies Identifies  SLC2A2  and  HP  as Novel Loci Influencing Serum Cholesterol Levels.')\n",
      "Cholesterol Levels.\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "20072603 1.0 1\n",
      "Sentence(Document('20072603', Corpus (GWAS Corpus)), 0, u'Genome-Wide Association Study Identifies  ALDH7A1  as a Novel Susceptibility Gene for Osteoporosis.')\n",
      "Osteoporosis.\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "20081856 1.0 1\n",
      "Sentence(Document('20081856', Corpus (GWAS Corpus)), 0, u'Meta-analysis of genome-wide association data detects a risk locus for major mood disorders on chromosome 3p21.1.')\n",
      "major mood disorders\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "20081858 1.0 -1\n",
      "Sentence(Document('20081858', Corpus (GWAS Corpus)), 0, u'New genetic loci implicated in fasting glucose homeostasis and their impact on type 2 diabetes risk.')\n",
      "type 2 diabetes\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "20081858 1.0 -1\n",
      "Sentence(Document('20081858', Corpus (GWAS Corpus)), 0, u'New genetic loci implicated in fasting glucose homeostasis and their impact on type 2 diabetes risk.')\n",
      "glucose homeostasis\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "20169177 1.0 1\n",
      "Sentence(Document('20169177', Corpus (GWAS Corpus)), 0, u'Genome-Wide Association Study in Asian Populations Identifies Variants in  ETS1  and  WDFY4  Associated with Systemic Lupus Erythematosus.')\n",
      "Systemic Lupus Erythematosus.\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "20173747 1.0 1\n",
      "Sentence(Document('20173747', Corpus (GWAS Corpus)), 0, u'Common Variants in  KCNN3  are Associated with Lone Atrial Fibrillation.')\n",
      "Atrial Fibrillation.\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "20173748 1.0 1\n",
      "Sentence(Document('20173748', Corpus (GWAS Corpus)), 0, u'Variants in  FAM13A  are associated with chronic obstructive pulmonary disease.')\n",
      "chronic obstructive pulmonary disease.\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "20173748 1.0 -1\n",
      "Sentence(Document('20173748', Corpus (GWAS Corpus)), 0, u'Variants in  FAM13A  are associated with chronic obstructive pulmonary disease.')\n",
      "associated with chronic\n",
      "[10, 0, 5, 0, 0, -5]\n",
      "\n",
      "20174558 1.0 1\n",
      "Sentence(Document('20174558', Corpus (GWAS Corpus)), 0, u'A Genome-Wide Association Study Identifies Susceptibility Variants for Type 2 Diabetes in Han Chinese.')\n",
      "Type 2 Diabetes\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "20190752 1.0 1\n",
      "Sentence(Document('20190752', Corpus (GWAS Corpus)), 0, u'Multiple common variants for celiac disease influencing immune gene expression.')\n",
      "celiac disease\n",
      "[10, 0, 0, 0, 0, 0]\n",
      "\n",
      "20190752 1.0 -1\n",
      "Sentence(Document('20190752', Corpus (GWAS Corpus)), 0, u'Multiple common variants for celiac disease influencing immune gene expression.')\n",
      "Multiple\n",
      "[10, 0, 5, 0, 0, -5]\n",
      "\n",
      "20195266 1.0 -1\n",
      "Sentence(Document('20195266', Corpus (GWAS Corpus)), 0, u'Genomewide pharmacogenomic study of metabolic side effects to antipsychotic drugs.')\n",
      "to antipsychotic drugs.\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n",
      "20195266 1.0 -1\n",
      "Sentence(Document('20195266', Corpus (GWAS Corpus)), 0, u'Genomewide pharmacogenomic study of metabolic side effects to antipsychotic drugs.')\n",
      "metabolic side effects\n",
      "[10, 0, 5, 0, 0, 0]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "preds = learner.predict_wmv(phen_c)\n",
    "mislabeled_cand = [(c,p, gt_dict.get(c2uid(c), None)) for c, p in zip(phen_c, preds) if p == 1 or p != gt_dict.get(c2uid(c), p)]\n",
    "for (c,p,g) in mislabeled_cand:\n",
    "    if c.context.position != 0: continue\n",
    "    print c.context.document.name, p, g\n",
    "    print c.context    \n",
    "    print c.get_span()\n",
    "    print [LF(c) for LF in LFs]\n",
    "    print"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save and analyze the results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Analyze / Visualize"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "If a mention occurs in the title, its probably correct, we can take it.\n",
    "\n",
    "Question: what papers did not have any disease mentions in the title?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 731,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Applying LFs...\n",
      "Featurizing...\n",
      "6\n",
      "18262040 LDL-cholesterol concentrations: a genome-wide association study.\n",
      "19305408 Common variants at ten loci influence QT interval duration in the QTGEN Study.\n",
      "19587794 Common genetic variation near the phospholamban gene is associated with cardiac repolarisation: meta-analysis of three genome-wide association studies.\n",
      "19197348 Genome-wide association studies in an isolated founder population from the Pacific Island of Kosrae.\n",
      "17903296 Genome-wide association with bone mass and geometry in the Framingham Heart Study.\n",
      "19343178 Meta-analysis of genome-wide scans for human adult stature identifies novel Loci and associations with measures of skeletal frame size.\n"
     ]
    }
   ],
   "source": [
    "preds = learner.predict_wmv(phen_c)\n",
    "results = [c for p, c in zip(preds, phen_c) if p > 0 and c.context.position == 0]\n",
    "doc_set = {c.context.document.name for c in results}\n",
    "missing_docs = {doc.name for doc in corpus.documents} - doc_set\n",
    "docs = sorted(list(missing_docs))\n",
    "print len(docs)\n",
    "for d in missing_docs:\n",
    "    print d, kb.paper_by_pmid(d).title"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's not visualize what we found."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 734,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Applying LFs...\n",
      "Featurizing...\n"
     ]
    }
   ],
   "source": [
    "scores = learner.score_wmv(phen_c)\n",
    "score_dict = { doc.name : list() for doc in corpus.documents }\n",
    "for s, c in zip(scores, phen_c):\n",
    "    score_dict[c.context.document.name].append((s,c))\n",
    "\n",
    "results = dict()\n",
    "for pmid, preds in score_dict.items():\n",
    "    if preds: \n",
    "        best_c = sorted(preds, reverse=True)[0][1]\n",
    "        results[best_c.context.document.name] = best_c\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 736,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Document('17447842', Corpus (GWAS Corpus)) Novel Crohn disease locus identified by genome-wide association maps to a gene desert on 5p13.1 and modulates expression of PTGER4.\n",
      "Span(\"Crohn Disease\", context=None, chars=[6,18], words=[1,2])\n",
      "[(19.270490808768059, Span(\"Crohn Disease\", context=None, chars=[6,18], words=[1,2])), (1.9959427016534608, Span(\"Crohn disease (\", context=None, chars=[42,56], words=[6,8])), (0.0, Span(\"locus on Chromosome 5.\", context=None, chars=[155,176], words=[27,31])), (0.0, Span(\"Crohn disease,\", context=None, chars=[33,46], words=[6,8])), (0.0, Span(\"susceptibility to\", context=None, chars=[151,167], words=[22,23]))]\n",
      "\n",
      "Document('17658951', Corpus (GWAS Corpus)) Genome-wide association scan shows genetic variants in the FTO gene are associated with obesity-related traits.\n",
      "Span(\"Obesity-Related Traits\", context=None, chars=[90,111], words=[13,14])\n",
      "[(19.270490808768059, Span(\"Obesity-Related Traits\", context=None, chars=[90,111], words=[13,14])), (0.0, Span(\"BMI (\", context=None, chars=[68,72], words=[11,12])), (0.0, Span(\"weight (\", context=None, chars=[142,149], words=[35,36])), (0.0, Span(\"BMI (\", context=None, chars=[71,75], words=[11,12])), (0.0, Span(\"body mass index,\", context=None, chars=[153,168], words=[27,30]))]\n",
      "\n",
      "Document('17684544', Corpus (GWAS Corpus)) Systematic association mapping identifies NELL1 as a novel IBD disease gene.\n",
      "Span(\"IBD Disease\", context=None, chars=[61,71], words=[8,9])\n",
      "[(19.270490808768059, Span(\"IBD Disease\", context=None, chars=[61,71], words=[8,9])), (1.9959427016534608, Span(\"ulcerative colitis (\", context=None, chars=[39,58], words=[6,8])), (1.9959427016534608, Span(\"Crohn disease (\", context=None, chars=[0,14], words=[0,2])), (1.9959427016534608, Span(\"inflammatory bowel disease (\", context=None, chars=[36,63], words=[9,12])), (0.0, Span(\"involvement of\", context=None, chars=[85,98], words=[14,15]))]\n",
      "\n",
      "Document('17903292', Corpus (GWAS Corpus)) A genome-wide association for kidney function and endocrine-related traits in the NHLBI's Framingham Heart Study.\n",
      "Span(\"endocrine-related traits\", context=None, chars=[50,73], words=[7,8])\n",
      "[(19.270490808768059, Span(\"endocrine-related traits\", context=None, chars=[50,73], words=[7,8])), (1.9959427016534608, Span(\"urinary albumin excretion (\", context=None, chars=[37,63], words=[7,10])), (1.9959427016534608, Span(\"Glomerular filtration rate (\", context=None, chars=[0,27], words=[0,3])), (1.9959427016534608, Span(\"Thyroid stimulating hormone (\", context=None, chars=[0,28], words=[0,3])), (1.9959427016534608, Span(\"tests (\", context=None, chars=[195,201], words=[24,25]))]\n",
      "\n",
      "Document('17903293', Corpus (GWAS Corpus)) Genome-wide association with select biomarker traits in the Framingham Heart Study.\n",
      "Span(\"biomarker traits\", context=None, chars=[36,51], words=[4,5])\n",
      "[(19.270490808768059, Span(\"biomarker traits\", context=None, chars=[36,51], words=[4,5])), (1.9959427016534608, Span(\"C-reactive protein (\", context=None, chars=[231,250], words=[63,65])), (1.9959427016534608, Span(\"Tests (\", context=None, chars=[73,79], words=[12,13])), (0.0, Span(\"both phenotype\", context=None, chars=[101,114], words=[24,25])), (0.0, Span(\"Vitamin K\", context=None, chars=[212,220], words=[57,58]))]\n",
      "\n",
      "Document('17903294', Corpus (GWAS Corpus)) Genome-wide association and linkage analyses of hemostatic factors and hematological phenotypes in the Framingham Heart Study.\n",
      "Span(\"hematological phenotypes\", context=None, chars=[71,94], words=[9,10])\n",
      "[(19.270490808768059, Span(\"hematological phenotypes\", context=None, chars=[71,94], words=[9,10])), (1.9959427016534608, Span(\"cardiovascular disease (\", context=None, chars=[113,136], words=[17,19])), (0.0, Span(\"mean corpuscular hemoglobin concentration.\", context=None, chars=[102,143], words=[19,23])), (0.0, Span(\"hematological phenotypes\", context=None, chars=[174,197], words=[23,24])), (0.0, Span(\"factor VII,\", context=None, chars=[254,264], words=[39,41]))]\n",
      "\n",
      "Document('17903295', Corpus (GWAS Corpus)) Genetic correlates of longevity and selected age-related phenotypes: a genome-wide association study in the Framingham Study.\n",
      "Span(\"longevity\", context=None, chars=[22,30], words=[3,3])\n",
      "[(29.125847746080705, Span(\"longevity\", context=None, chars=[22,30], words=[3,3])), (19.270490808768059, Span(\"age-related phenotypes\", context=None, chars=[45,66], words=[6,7])), (0.0, Span(\"aging.\", context=None, chars=[161,166], words=[25,26])), (0.0, Span(\"longevity\", context=None, chars=[139,147], words=[22,22])), (0.0, Span(\"Longevity\", context=None, chars=[0,8], words=[0,0]))]\n",
      "\n",
      "Document('17903296', Corpus (GWAS Corpus)) Genome-wide association with bone mass and geometry in the Framingham Heart Study.\n",
      "Span(\"tests (\", context=None, chars=[156,162], words=[21,22])\n",
      "[(1.9959427016534608, Span(\"tests (\", context=None, chars=[156,162], words=[21,22])), (0.0, Span(\"osteoporosis.\", context=None, chars=[154,166], words=[23,24])), (0.0, Span(\"osteoporosis.\", context=None, chars=[229,241], words=[42,43])), (0.0, Span(\"and bone\", context=None, chars=[46,53], words=[6,7])), (0.0, Span(\"osteoporosis,\", context=None, chars=[101,113], words=[20,21]))]\n",
      "\n",
      "Document('17903297', Corpus (GWAS Corpus)) Genetic correlates of brain aging on MRI and cognitive test measures: a genome-wide association and linkage analysis in the Framingham Study.\n",
      "Span(\"brain\", context=None, chars=[22,26], words=[3,3])\n",
      "[(29.125847746080705, Span(\"brain\", context=None, chars=[22,26], words=[3,3])), (29.125847746080705, Span(\"aging\", context=None, chars=[28,32], words=[4,4])), (29.125847746080705, Span(\"cognitive test\", context=None, chars=[45,58], words=[8,9])), (1.9959427016534608, Span(\"Alzheimer's disease (\", context=None, chars=[166,186], words=[25,28])), (1.9959427016534608, Span(\"tests (\", context=None, chars=[135,141], words=[20,21]))]\n",
      "\n",
      "Document('17903298', Corpus (GWAS Corpus)) Genome-wide association with diabetes-related traits in the Framingham Heart Study.\n",
      "Span(\"diabetes-related traits\", context=None, chars=[29,51], words=[3,4])\n",
      "[(19.270490808768059, Span(\"diabetes-related traits\", context=None, chars=[29,51], words=[3,4])), (0.0, Span(\"insulin sensitivity\", context=None, chars=[335,353], words=[58,59])), (0.0, Span(\"insulin,\", context=None, chars=[287,294], words=[48,49])), (0.0, Span(\"insulin traits\", context=None, chars=[263,276], words=[44,45])), (0.0, Span(\"Susceptibility to\", context=None, chars=[0,16], words=[0,1]))]\n",
      "\n",
      "Document('17903300', Corpus (GWAS Corpus)) Genome-wide association to body mass index and waist circumference: the Framingham Heart Study 100K project.\n",
      "Span(\"waist circumference:\", context=None, chars=[47,66], words=[7,9])\n",
      "[(29.125847746080705, Span(\"waist circumference:\", context=None, chars=[47,66], words=[7,9])), (29.125847746080705, Span(\"body mass index\", context=None, chars=[27,41], words=[3,5])), (1.9959427016534608, Span(\"Body mass index (\", context=None, chars=[0,16], words=[0,3])), (1.9959427016534608, Span(\"cardiovascular disease (\", context=None, chars=[31,54], words=[5,7])), (0.0, Span(\"adiposity (\", context=None, chars=[101,111], words=[22,23]))]\n",
      "\n",
      "Document('17903301', Corpus (GWAS Corpus)) Genome-wide association of echocardiographic dimensions, brachial artery endothelial function and treadmill exercise responses in the Framingham Heart Study.\n",
      "Span(\"exercise responses\", context=None, chars=[108,125], words=[12,13])\n",
      "[(19.270490808768059, Span(\"exercise responses\", context=None, chars=[108,125], words=[12,13])), (1.9959427016534608, Span(\"tests (\", context=None, chars=[65,71], words=[9,10])), (0.0, Span(\"aortic root size.\", context=None, chars=[175,191], words=[26,29])), (0.0, Span(\"heart rate\", context=None, chars=[64,73], words=[11,12])), (0.0, Span(\"systolic blood pressure\", context=None, chars=[36,58], words=[7,9]))]\n",
      "\n",
      "Document('17903302', Corpus (GWAS Corpus)) Framingham Heart Study 100K Project: genome-wide associations for blood pressure and arterial stiffness.\n",
      "Span(\"arterial stiffness.\", context=None, chars=[85,103], words=[12,14])\n",
      "[(29.125847746080705, Span(\"arterial stiffness.\", context=None, chars=[85,103], words=[12,14])), (29.125847746080705, Span(\"blood pressure\", context=None, chars=[66,79], words=[9,10])), (1.9959427016534608, Span(\"blood-pressure (\", context=None, chars=[28,43], words=[3,4])), (0.0, Span(\"stiffness phenotypes\", context=None, chars=[81,100], words=[11,12])), (0.0, Span(\"blood pressure\", context=None, chars=[53,66], words=[7,8]))]\n",
      "\n",
      "Document('17903303', Corpus (GWAS Corpus)) Genome-wide association study for subclinical atherosclerosis in major arterial territories in the NHLBI's Framingham Heart Study.\n",
      "Span(\"atherosclerosis\", context=None, chars=[46,60], words=[5,5])\n",
      "[(29.125847746080705, Span(\"atherosclerosis\", context=None, chars=[46,60], words=[5,5])), (15.888639912619039, Span(\"subclinical\", context=None, chars=[34,44], words=[4,4])), (1.9959427016534608, Span(\"testing (\", context=None, chars=[315,323], words=[51,52])), (1.9959427016534608, Span(\"atherosclerosis (\", context=None, chars=[12,28], words=[1,2])), (0.0, Span(\"on chromosome 12.\", context=None, chars=[142,158], words=[21,24]))]\n",
      "\n",
      "Document('17903304', Corpus (GWAS Corpus)) Framingham Heart Study 100K project: genome-wide associations for cardiovascular disease outcomes.\n",
      "Span(\"cardiovascular disease\", context=None, chars=[66,87], words=[9,10])\n",
      "[(19.270490808768059, Span(\"cardiovascular disease\", context=None, chars=[66,87], words=[9,10])), (1.9959427016534608, Span(\", cardiovascular disease (\", context=None, chars=[32,57], words=[4,7])), (1.9959427016534608, Span(\"coronary heart disease (\", context=None, chars=[76,99], words=[12,15])), (1.9959427016534608, Span(\", heart failure (\", context=None, chars=[112,128], words=[20,23])), (1.9959427016534608, Span(\"atrial fibrillation (\", context=None, chars=[138,158], words=[28,30]))]\n",
      "\n",
      "Document('17903305', Corpus (GWAS Corpus)) A genome-wide association study of breast and prostate cancer in the NHLBI's Framingham Heart Study.\n",
      "Span(\"prostate cancer\", context=None, chars=[46,60], words=[7,8])\n",
      "[(29.125847746080705, Span(\"prostate cancer\", context=None, chars=[46,60], words=[7,8])), (1.9959427016534608, Span(\"tests (\", context=None, chars=[327,333], words=[56,57])), (0.0, Span(\"prostate cancer.\", context=None, chars=[115,130], words=[14,16])), (0.0, Span(\"cancer\", context=None, chars=[11,16], words=[2,2])), (0.0, Span(\"cancer\", context=None, chars=[128,133], words=[19,19]))]\n",
      "\n",
      "Document('17903306', Corpus (GWAS Corpus)) Genome-wide association study of electrocardiographic and heart rate variability traits: the Framingham Heart Study.\n",
      "Span(\"heart rate\", context=None, chars=[58,67], words=[6,7])\n",
      "[(29.125847746080705, Span(\"heart rate\", context=None, chars=[58,67], words=[6,7])), (19.270490808768059, Span(\"variability traits\", context=None, chars=[69,86], words=[8,9])), (1.9959427016534608, Span(\"variability (\", context=None, chars=[52,64], words=[8,9])), (1.9959427016534608, Span(\"RR intervals (\", context=None, chars=[259,272], words=[47,49])), (1.9959427016534608, Span(\"heart rate\", context=None, chars=[41,50], words=[6,7]))]\n",
      "\n",
      "Document('17903307', Corpus (GWAS Corpus)) Framingham Heart Study genome-wide association: results for pulmonary function measures.\n",
      "Span(\"pulmonary function\", context=None, chars=[60,77], words=[8,9])\n",
      "[(29.125847746080705, Span(\"pulmonary function\", context=None, chars=[60,77], words=[8,9])), (1.9959427016534608, Span(\"body mass index (\", context=None, chars=[212,228], words=[32,35])), (0.0, Span(\"spirometry phenotypes\", context=None, chars=[4,24], words=[1,2])), (0.0, Span(\"forced expiratory volume\", context=None, chars=[144,167], words=[22,24])), (0.0, Span(\"pack-years.\", context=None, chars=[121,131], words=[21,22]))]\n",
      "\n",
      "Document('17903308', Corpus (GWAS Corpus)) Genome-wide association of sleep and circadian phenotypes.\n",
      "Span(\"sleep\", context=None, chars=[27,31], words=[3,3])\n",
      "[(29.125847746080705, Span(\"sleep\", context=None, chars=[27,31], words=[3,3])), (19.270490808768059, Span(\"circadian phenotypes\", context=None, chars=[37,56], words=[5,6])), (0.0, Span(\"neuropeptide\", context=None, chars=[170,181], words=[28,28])), (0.0, Span(\"these traits\", context=None, chars=[48,59], words=[7,8])), (0.0, Span(\"tests\", context=None, chars=[12,16], words=[1,1]))]\n",
      "\n",
      "Document('17997608', Corpus (GWAS Corpus)) The GLUT9 gene is associated with serum uric acid levels in Sardinia and Chianti cohorts.\n",
      "Span(\"Uric Acid Levels\", context=None, chars=[42,57], words=[7,9])\n",
      "[(29.125847746080705, Span(\"Uric Acid Levels\", context=None, chars=[42,57], words=[7,9])), (0.0, Span(\"glucose\", context=None, chars=[53,59], words=[10,10])), (0.0, Span(\"uric acid,\", context=None, chars=[129,138], words=[19,21])), (0.0, Span(\"uric acid levels\", context=None, chars=[109,124], words=[13,15])), (0.0, Span(\"uric acid levels\", context=None, chars=[11,26], words=[2,4]))]\n",
      "\n",
      "Document('18159244', Corpus (GWAS Corpus)) Genome wide association (GWA) study for early onset extreme obesity supports the role of fat mass and obesity associated gene (FTO) variants.\n",
      "Span(\"Obesity\", context=None, chars=[102,108], words=[19,19])\n",
      "[(31.121790447734167, Span(\"Obesity\", context=None, chars=[102,108], words=[19,19])), (29.125847746080705, Span(\"Onset\", context=None, chars=[46,50], words=[9,9])), (29.125847746080705, Span(\"Obesity\", context=None, chars=[60,66], words=[11,11])), (0.0, Span(\"obesity\", context=None, chars=[116,122], words=[19,19])), (0.0, Span(\"onset\", context=None, chars=[102,106], words=[17,17]))]\n",
      "\n",
      "Document('18262040', Corpus (GWAS Corpus)) LDL-cholesterol concentrations: a genome-wide association study.\n",
      "Span(\"cardiovascular disease.\", context=None, chars=[186,208], words=[27,29])\n",
      "[(0.0, Span(\"cardiovascular disease.\", context=None, chars=[186,208], words=[27,29])), (0.0, Span(\"cholesterol\", context=None, chars=[44,54], words=[9,9])), (0.0, Span(\"cholesterol\", context=None, chars=[173,183], words=[29,29])), (0.0, Span(\"cholesterol at\", context=None, chars=[153,166], words=[43,44])), (0.0, Span(\"cholesterol\", context=None, chars=[103,113], words=[14,14]))]\n",
      "\n",
      "Document('18282107', Corpus (GWAS Corpus)) Genome-wide association identifies a common variant in the reelin gene that increases the risk of schizophrenia only in women.\n",
      "Span(\"Schizophrenia\", context=None, chars=[98,110], words=[15,15])\n",
      "[(29.125847746080705, Span(\"Schizophrenia\", context=None, chars=[98,110], words=[15,15])), (0.0, Span(\"schizophrenia\", context=None, chars=[19,31], words=[3,3])), (0.0, Span(\"schizophrenia\", context=None, chars=[13,25], words=[3,3])), (0.0, Span(\"brain\", context=None, chars=[81,85], words=[18,18])), (0.0, Span(\"schizophrenia\", context=None, chars=[69,81], words=[12,12]))]\n",
      "\n",
      "Document('18369459', Corpus (GWAS Corpus)) A genome-wide association study of psoriasis and psoriatic arthritis identifies new disease loci.\n",
      "Span(\"Psoriatic Arthritis\", context=None, chars=[49,67], words=[7,8])\n",
      "[(29.125847746080705, Span(\"Psoriatic Arthritis\", context=None, chars=[49,67], words=[7,8])), (29.125847746080705, Span(\"Psoriasis\", context=None, chars=[35,43], words=[5,5])), (19.270490808768059, Span(\"New Disease\", context=None, chars=[80,90], words=[10,11])), (1.9959427016534608, Span(\"psoriasis (\", context=None, chars=[104,114], words=[14,15])), (1.9959427016534608, Span(\"psoriatic arthritis (\", context=None, chars=[19,39], words=[5,7]))]\n",
      "\n",
      "Document('18455228', Corpus (GWAS Corpus)) Bone mineral density, osteoporosis, and osteoporotic fractures: a genome-wide association study.\n",
      "Span(\", osteoporosis,\", context=None, chars=[20,34], words=[3,5])\n",
      "[(29.125847746080705, Span(\", osteoporosis,\", context=None, chars=[20,34], words=[3,5])), (29.125847746080705, Span(\"osteoporotic fractures:\", context=None, chars=[40,62], words=[7,9])), (29.125847746080705, Span(\"Bone mineral density,\", context=None, chars=[0,20], words=[0,3])), (0.0, Span(\"osteoporotic fractures\", context=None, chars=[58,79], words=[9,10])), (0.0, Span(\"bone mineral density\", context=None, chars=[62,81], words=[8,10]))]\n",
      "\n",
      "Document('18464913', Corpus (GWAS Corpus)) A genome-wide association study identifies protein quantitative trait loci (pQTLs).\n",
      "Span(\"Protein Quantitative Trait\", context=None, chars=[43,68], words=[5,7])\n",
      "[(19.270490808768059, Span(\"Protein Quantitative Trait\", context=None, chars=[43,68], words=[5,7])), (1.9959427016534608, Span(\"proteins  (\", context=None, chars=[146,156], words=[23,24])), (1.9959427016534608, Span(\"proteins  (\", context=None, chars=[179,189], words=[29,30])), (0.0, Span(\"protein levels\", context=None, chars=[45,58], words=[7,8])), (0.0, Span(\"),  SHBG  (\", context=None, chars=[196,206], words=[65,68]))]\n",
      "\n",
      "Document('18483556', Corpus (GWAS Corpus)) A genome-wide association study identifies novel alleles associated with hair color and skin pigmentation.\n",
      "Span(\"Hair Color\", context=None, chars=[73,82], words=[9,10])\n",
      "[(29.125847746080705, Span(\"Hair Color\", context=None, chars=[73,82], words=[9,10])), (29.125847746080705, Span(\"Skin Pigmentation.\", context=None, chars=[88,105], words=[12,14])), (0.0, Span(\"hair color\", context=None, chars=[150,159], words=[21,22])), (0.0, Span(\"hair color.\", context=None, chars=[51,61], words=[9,11])), (0.0, Span(\"hair color\", context=None, chars=[71,80], words=[10,11]))]\n",
      "\n",
      "Document('18604267', Corpus (GWAS Corpus)) Novel association of ABO histo-blood group antigen with soluble ICAM-1: results of a genome-wide association study of 6,578 women.\n",
      "Span(\"ICAM-1:\", context=None, chars=[64,70], words=[9,10])\n",
      "[(29.125847746080705, Span(\"ICAM-1:\", context=None, chars=[64,70], words=[9,10])), (29.125847746080705, Span(\"Antigen\", context=None, chars=[43,49], words=[6,6])), (0.0, Span(\"inflammatory marker\", context=None, chars=[58,76], words=[10,11])), (0.0, Span(\", stroke,\", context=None, chars=[163,171], words=[26,28])), (0.0, Span(\"common diseases\", context=None, chars=[116,130], words=[18,19]))]\n",
      "\n",
      "Document('18776929', Corpus (GWAS Corpus)) Identification of PLCL1 gene for hip bone size variation in females in a genome-wide association study.\n",
      "Span(\"Hip Bone Size\", context=None, chars=[35,47], words=[5,7])\n",
      "[(29.125847746080705, Span(\"Hip Bone Size\", context=None, chars=[35,47], words=[5,7])), (1.9959427016534608, Span(\"Hip bone size (\", context=None, chars=[0,14], words=[0,3])), (1.9959427016534608, Span(\"trauma\", context=None, chars=[99,104], words=[16,16])), (1.9959427016534608, Span(\"hip fractures (\", context=None, chars=[106,120], words=[17,19])), (0.0, Span(\"the trait\", context=None, chars=[77,85], words=[11,12]))]\n",
      "\n",
      "Document('18823527', Corpus (GWAS Corpus)) A genome-wide association study for late-onset Alzheimer's disease using DNA pooling.\n",
      "Span(\"late-onset Alzheimer's disease\", context=None, chars=[36,65], words=[5,8])\n",
      "[(29.125847746080705, Span(\"late-onset Alzheimer's disease\", context=None, chars=[36,65], words=[5,8])), (0.0, Span(\"use of\", context=None, chars=[172,177], words=[31,32])), (0.0, Span(\"neurodegenerative disease\", context=None, chars=[56,80], words=[11,12])), (0.0, Span(\"an age\", context=None, chars=[41,46], words=[8,9])), (0.0, Span(\"Age at onset\", context=None, chars=[0,11], words=[0,2]))]\n",
      "\n",
      "Document('18840781', Corpus (GWAS Corpus)) Follow-up analysis of genome-wide association data identifies novel loci for type 1 diabetes.\n",
      "Span(\"Type 1 Diabetes.\", context=None, chars=[77,92], words=[10,13])\n",
      "[(29.125847746080705, Span(\"Type 1 Diabetes.\", context=None, chars=[77,92], words=[10,13])), (0.0, Span(\"tests\", context=None, chars=[102,106], words=[17,17])), (0.0, Span(\"with type 1\", context=None, chars=[152,162], words=[22,24])), (0.0, Span(\"type 1 diabetes\", context=None, chars=[175,189], words=[29,31])), (0.0, Span(\"type 1 diabetes.\", context=None, chars=[157,172], words=[23,26]))]\n",
      "\n",
      "Document('18846228', Corpus (GWAS Corpus)) Genome-wide scan on total serum IgE levels identifies FCER1A as novel susceptibility locus.\n",
      "Span(\"Serum IgE Levels\", context=None, chars=[26,41], words=[4,6])\n",
      "[(29.125847746080705, Span(\"Serum IgE Levels\", context=None, chars=[26,41], words=[4,6])), (0.0, Span(\"disorders,\", context=None, chars=[47,56], words=[8,9])), (0.0, Span(\"autoimmune diseases.\", context=None, chars=[128,147], words=[23,25])), (0.0, Span(\"serum IgE levels.\", context=None, chars=[67,83], words=[9,12])), (0.0, Span(\"serum IgE\", context=None, chars=[15,23], words=[3,4]))]\n",
      "\n",
      "Document('18941528', Corpus (GWAS Corpus)) Identification of a novel risk locus for multiple sclerosis at 13q31.3 by a pooled genome-wide scan of 500,000 single nucleotide polymorphisms.\n",
      "Span(\"Multiple Sclerosis at\", context=None, chars=[41,61], words=[7,9])\n",
      "[(29.125847746080705, Span(\"Multiple Sclerosis at\", context=None, chars=[41,61], words=[7,9])), (0.0, Span(\"elements\", context=None, chars=[181,188], words=[28,28])), (0.0, Span(\"the disease\", context=None, chars=[131,141], words=[22,23])), (0.0, Span(\"of chromosome 13\", context=None, chars=[12,27], words=[2,4])), (0.0, Span(\"the disease\", context=None, chars=[117,127], words=[18,19]))]\n",
      "\n",
      "Document('19043545', Corpus (GWAS Corpus)) Genetics meets metabolomics: a genome-wide association study of metabolite profiles in human serum.\n",
      "Span(\"Metabolite\", context=None, chars=[64,73], words=[9,9])\n",
      "[(29.125847746080705, Span(\"Metabolite\", context=None, chars=[64,73], words=[9,9])), (0.0, Span(\"polyunsaturated fatty acids,\", context=None, chars=[153,180], words=[22,25])), (0.0, Span(\"simultaneous measurements\", context=None, chars=[3,27], words=[1,2])), (0.0, Span(\"triglycerides.\", context=None, chars=[262,275], words=[40,41])), (0.0, Span(\"metabolic traits\", context=None, chars=[109,124], words=[22,23]))]\n",
      "\n",
      "Document('19056611', Corpus (GWAS Corpus)) Adiposity-related heterogeneity in patterns of type 2 diabetes susceptibility observed in genome-wide association data.\n",
      "Span(\"Type 2 Diabetes\", context=None, chars=[47,61], words=[5,7])\n",
      "[(29.125847746080705, Span(\"Type 2 Diabetes\", context=None, chars=[47,61], words=[5,7])), (0.0, Span(\"type 2 diabetes.\", context=None, chars=[91,106], words=[12,15])), (0.0, Span(\"type 2 diabetes\", context=None, chars=[137,151], words=[21,23])), (0.0, Span(\"type 2 diabetic\", context=None, chars=[74,88], words=[12,14])), (0.0, Span(\"BMI (\", context=None, chars=[301,305], words=[51,52]))]\n",
      "\n",
      "Document('19081515', Corpus (GWAS Corpus)) Genetic risk factors for variant Creutzfeldt-Jakob disease: a genome-wide association study.\n",
      "Span(\"Jakob disease\", context=None, chars=[45,57], words=[7,8])\n",
      "[(19.270490808768059, Span(\"Jakob disease\", context=None, chars=[45,57], words=[7,8])), (1.9959427016534608, Span(\"bovine spongiform encephalopathy (BSE\", context=None, chars=[175,211], words=[31,35])), (0.0, Span(\"sporadic CJD (\", context=None, chars=[117,130], words=[26,28])), (0.0, Span(\"kuru.\", context=None, chars=[140,144], words=[32,33])), (0.0, Span(\"retinoic acid receptor\", context=None, chars=[107,128], words=[18,20]))]\n",
      "\n",
      "Document('19096518', Corpus (GWAS Corpus)) Novel association of HK1 with glycated hemoglobin in a non-diabetic population: a genome-wide evaluation of 14,618 participants in the Women's Genome Health Study.\n",
      "Span(\"Glycated Hemoglobin\", context=None, chars=[30,48], words=[5,6])\n",
      "[(29.125847746080705, Span(\"Glycated Hemoglobin\", context=None, chars=[30,48], words=[5,6])), (0.0, Span(\"hemoglobin\", context=None, chars=[147,156], words=[23,23])), (0.0, Span(\"Type 2 diabetes\", context=None, chars=[0,14], words=[0,2])), (0.0, Span(\", glucose\", context=None, chars=[167,175], words=[25,26])), (0.0, Span(\"blood glucose\", context=None, chars=[67,79], words=[9,10]))]\n",
      "\n",
      "Document('19116933', Corpus (GWAS Corpus)) Identification of a novel susceptibility locus for juvenile idiopathic arthritis by genome-wide association analysis.\n",
      "Span(\"juvenile idiopathic arthritis\", context=None, chars=[51,79], words=[7,9])\n",
      "[(29.125847746080705, Span(\"juvenile idiopathic arthritis\", context=None, chars=[51,79], words=[7,9])), (1.9959427016534608, Span(\"Juvenile idiopathic arthritis (\", context=None, chars=[0,30], words=[0,3])), (0.0, Span(\"rheumatic disease\", context=None, chars=[49,65], words=[9,10])), (0.0, Span(\"susceptibility to disease;\", context=None, chars=[142,167], words=[23,26])), (0.0, Span(\"birth\", context=None, chars=[227,231], words=[39,39]))]\n",
      "\n",
      "Document('19122664', Corpus (GWAS Corpus)) Ulcerative colitis-risk loci on chromosomes 1p36 and 12q15 found by genome-wide association study.\n",
      "Span(\"Ulcerative colitis\", context=None, chars=[0,17], words=[0,1])\n",
      "[(29.125847746080705, Span(\"Ulcerative colitis\", context=None, chars=[0,17], words=[0,1])), (0.0, Span(\"locus on chromosome\", context=None, chars=[220,238], words=[42,44])), (0.0, Span(\", ulcerative colitis\", context=None, chars=[66,85], words=[10,12])), (0.0, Span(\"ulcerative colitis\", context=None, chars=[91,108], words=[13,14])), (0.0, Span(\"Ulcerative colitis\", context=None, chars=[0,17], words=[0,1]))]\n",
      "\n",
      "Document('19132087', Corpus (GWAS Corpus)) A genome-wide association study identifies novel and functionally related susceptibility Loci for Kawasaki disease.\n",
      "Span(\"Kawasaki Disease.\", context=None, chars=[98,114], words=[12,14])\n",
      "[(29.125847746080705, Span(\"Kawasaki Disease.\", context=None, chars=[98,114], words=[12,14])), (1.9959427016534608, Span(\"Kawasaki disease (\", context=None, chars=[0,17], words=[0,2])), (0.0, Span(\"infectious diseases\", context=None, chars=[35,53], words=[6,7])), (0.0, Span(\"Kawasaki disease\", context=None, chars=[123,138], words=[22,23])), (0.0, Span(\"heart disease\", context=None, chars=[175,187], words=[26,27]))]\n",
      "\n",
      "Document('19169254', Corpus (GWAS Corpus)) Genome-wide scan reveals association of psoriasis with IL-23 and NF-kappaB pathways.\n",
      "Span(\"Psoriasis\", context=None, chars=[39,47], words=[5,5])\n",
      "[(29.125847746080705, Span(\"Psoriasis\", context=None, chars=[39,47], words=[5,5])), (0.0, Span(\"proteins\", context=None, chars=[13,20], words=[2,2])), (0.0, Span(\"disorders.\", context=None, chars=[146,155], words=[21,22])), (0.0, Span(\"psoriasis\", context=None, chars=[61,69], words=[10,10])), (0.0, Span(\"genes involved in\", context=None, chars=[215,231], words=[40,42]))]\n",
      "\n",
      "Document('19197348', Corpus (GWAS Corpus)) Genome-wide association studies in an isolated founder population from the Pacific Island of Kosrae.\n",
      "Span(\"or trait\", context=None, chars=[131,138], words=[19,20])\n",
      "[(0.0, Span(\"or trait\", context=None, chars=[131,138], words=[19,20])), (0.0, Span(\"complex traits\", context=None, chars=[123,136], words=[18,19])), (0.0, Span(\"use of\", context=None, chars=[25,30], words=[4,5])), (0.0, Span(\"influencing trait\", context=None, chars=[60,76], words=[11,12])), (0.0, Span(\"cholesterol,\", context=None, chars=[70,81], words=[12,13]))]\n",
      "\n",
      "Document('19219042', Corpus (GWAS Corpus)) Genome-wide association study identifies a new breast cancer susceptibility locus at 6q25.1.\n",
      "Span(\"breast cancer\", context=None, chars=[49,61], words=[6,7])\n",
      "[(29.125847746080705, Span(\"breast cancer\", context=None, chars=[49,61], words=[6,7])), (0.0, Span(\"breast cancer.\", context=None, chars=[87,100], words=[13,15])), (0.0, Span(\"estrogen\", context=None, chars=[49,56], words=[10,10])), (0.0, Span(\"breast cancer\", context=None, chars=[133,145], words=[24,25])), (0.0, Span(\"breast cancer.\", context=None, chars=[96,109], words=[14,16]))]\n",
      "\n",
      "Document('19247474', Corpus (GWAS Corpus)) Genome-wide and candidate gene association study of cigarette smoking behaviors.\n",
      "Span(\"Smoking Behaviors.\", context=None, chars=[62,79], words=[8,10])\n",
      "[(29.125847746080705, Span(\"Smoking Behaviors.\", context=None, chars=[62,79], words=[8,10])), (0.0, Span(\"smoking,\", context=None, chars=[133,140], words=[26,27])), (0.0, Span(\"smoking behavior,\", context=None, chars=[30,46], words=[5,7])), (0.0, Span(\"smoking).\", context=None, chars=[273,281], words=[58,60])), (0.0, Span(\"cigarettes per day [\", context=None, chars=[219,238], words=[47,50]))]\n",
      "\n",
      "Document('19300482', Corpus (GWAS Corpus)) A genome-wide association study in chronic obstructive pulmonary disease (COPD): identification of two major susceptibility loci.\n",
      "Span(\"(COPD):\", context=None, chars=[73,79], words=[9,12])\n",
      "[(29.125847746080705, Span(\"(COPD):\", context=None, chars=[73,79], words=[9,12])), (29.125847746080705, Span(\"Chronic Obstructive Pulmonary Disease (\", context=None, chars=[35,73], words=[5,9])), (29.125847746080705, Span(\"Pulmonary Function\", context=None, chars=[164,181], words=[24,25])), (1.9959427016534608, Span(\"protein (\", context=None, chars=[25,33], words=[3,4])), (1.9959427016534608, Span(\"protein (\", context=None, chars=[40,48], words=[5,6]))]\n",
      "\n",
      "Document('19300499', Corpus (GWAS Corpus)) A genome-wide association study confirms VKORC1, CYP2C9, and CYP4F2 as principal genetic determinants of warfarin dose.\n",
      "Span(\"Warfarin\", context=None, chars=[111,118], words=[16,16])\n",
      "[(29.125847746080705, Span(\"Warfarin\", context=None, chars=[111,118], words=[16,16])), (0.0, Span(\"warfarin\", context=None, chars=[88,95], words=[15,15])), (0.0, Span(\"bleeding)\", context=None, chars=[167,175], words=[31,32])), (0.0, Span(\"warfarin\", context=None, chars=[42,49], words=[9,9])), (0.0, Span(\"variability.\", context=None, chars=[176,187], words=[39,40]))]\n",
      "\n",
      "Document('19300500', Corpus (GWAS Corpus)) A genome-wide association study of pulmonary function measures in the Framingham Heart Study.\n",
      "Span(\"Pulmonary Function\", context=None, chars=[35,52], words=[5,6])\n",
      "[(29.125847746080705, Span(\"Pulmonary Function\", context=None, chars=[35,52], words=[5,6])), (29.125847746080705, Span(\"(COPD):\", context=None, chars=[166,172], words=[22,25])), (29.125847746080705, Span(\"Chronic Obstructive Pulmonary Disease (\", context=None, chars=[128,166], words=[18,22])), (1.9959427016534608, Span(\"protein (\", context=None, chars=[94,102], words=[15,16])), (0.0, Span(\"obstruction phenotypes\", context=None, chars=[113,134], words=[17,18]))]\n",
      "\n",
      "Document('19304780', Corpus (GWAS Corpus)) Genome-wide scan identifies CDH13 as a novel susceptibility locus contributing to blood pressure determination in two European populations.\n",
      "Span(\"blood pressure\", context=None, chars=[84,97], words=[11,12])\n",
      "[(29.125847746080705, Span(\"blood pressure\", context=None, chars=[84,97], words=[11,12])), (1.9959427016534608, Span(\"); hypertension (\", context=None, chars=[89,105], words=[17,20])), (1.9959427016534608, Span(\"blood pressure (\", context=None, chars=[64,79], words=[8,10])), (0.0, Span(\"; HYP,\", context=None, chars=[190,195], words=[42,44])), (0.0, Span(\"BP traits\", context=None, chars=[28,36], words=[5,6]))]\n",
      "\n",
      "Document('19305408', Corpus (GWAS Corpus)) Common variants at ten loci influence QT interval duration in the QTGEN Study.\n",
      "Span(\"Cardiovascular\", context=None, chars=[170,183], words=[26,26])\n",
      "[(0.0, Span(\"Cardiovascular\", context=None, chars=[170,183], words=[26,26])), (0.0, Span(\"sudden cardiac death\", context=None, chars=[130,149], words=[18,20])), (0.0, Span(\"variation in QT interval.\", context=None, chars=[79,103], words=[16,20])), (0.0, Span(\"Long QT Syndromes.\", context=None, chars=[169,186], words=[33,36])), (0.0, Span(\"arrhythmias.\", context=None, chars=[168,179], words=[23,24]))]\n",
      "\n",
      "Document('19343178', Corpus (GWAS Corpus)) Meta-analysis of genome-wide scans for human adult stature identifies novel Loci and associations with measures of skeletal frame size.\n",
      "Span(\"height.\", context=None, chars=[182,188], words=[29,30])\n",
      "[(0.0, Span(\"height.\", context=None, chars=[182,188], words=[29,30])), (0.0, Span(\"height.\", context=None, chars=[130,136], words=[26,27])), (0.0, Span(\"height\", context=None, chars=[93,98], words=[14,14])), (0.0, Span(\"height\", context=None, chars=[93,98], words=[15,15])), (0.0, Span(\"height\", context=None, chars=[37,42], words=[6,6]))]\n",
      "\n",
      "Document('19359265', Corpus (GWAS Corpus)) A genome-wide association study identifies a novel locus on chromosome 18q12.2 influencing white cell telomere length.\n",
      "Span(\"telomere length.\", context=None, chars=[102,117], words=[15,17])\n",
      "[(29.125847746080705, Span(\"telomere length.\", context=None, chars=[102,117], words=[15,17])), (29.125847746080705, Span(\"locus on chromosome\", context=None, chars=[51,69], words=[7,9])), (0.0, Span(\"related diseases\", context=None, chars=[58,73], words=[11,12])), (0.0, Span(\"telomere length\", context=None, chars=[52,66], words=[8,9])), (0.0, Span(\"Telomere length\", context=None, chars=[0,14], words=[0,1]))]\n",
      "\n",
      "Document('19401414', Corpus (GWAS Corpus)) Confirmation of multiple risk Loci and genetic impacts by a genome-wide association study of type 2 diabetes in the Japanese population.\n",
      "Span(\"Type 2 Diabetes\", context=None, chars=[93,107], words=[14,16])\n",
      "[(29.125847746080705, Span(\"Type 2 Diabetes\", context=None, chars=[93,107], words=[14,16])), (15.888639912619039, Span(\"Multiple\", context=None, chars=[16,23], words=[2,2])), (0.0, Span(\"( PEPD\", context=None, chars=[43,48], words=[8,9])), (0.0, Span(\"the disease\", context=None, chars=[58,68], words=[9,10])), (0.0, Span(\"type 2 diabetes\", context=None, chars=[128,142], words=[21,23]))]\n",
      "\n",
      "Document('19412175', Corpus (GWAS Corpus)) Common variations in BARD1 influence susceptibility to high-risk neuroblastoma.\n",
      "Span(\"susceptibility to\", context=None, chars=[39,55], words=[5,6])\n",
      "[(29.125847746080705, Span(\"susceptibility to\", context=None, chars=[39,55], words=[5,6])), (29.125847746080705, Span(\"neuroblastoma.\", context=None, chars=[67,80], words=[8,9])), (0.0, Span(\"neuroblastoma.\", context=None, chars=[140,153], words=[22,23])), (0.0, Span(\"neuroblastoma\", context=None, chars=[110,122], words=[18,18])), (0.0, Span(\"neuroblastoma\", context=None, chars=[99,111], words=[15,15]))]\n",
      "\n",
      "Document('19412176', Corpus (GWAS Corpus)) Narcolepsy is strongly associated with the T-cell receptor alpha locus.\n",
      "Span(\"Narcolepsy\", context=None, chars=[0,9], words=[0,0])\n",
      "[(29.125847746080705, Span(\"Narcolepsy\", context=None, chars=[0,9], words=[0,0])), (15.888639912619039, Span(\"with the\", context=None, chars=[34,41], words=[4,5])), (0.0, Span(\"onset\", context=None, chars=[60,64], words=[7,7])), (0.0, Span(\"involvement of the\", context=None, chars=[37,54], words=[6,8])), (0.0, Span(\"sleep,\", context=None, chars=[75,80], words=[10,11]))]\n",
      "\n",
      "Document('19421330', Corpus (GWAS Corpus)) Genome-wide association study of young-onset hypertension in the Han Chinese population of Taiwan.\n",
      "Span(\"Hypertension\", context=None, chars=[45,56], words=[5,5])\n",
      "[(29.125847746080705, Span(\"Hypertension\", context=None, chars=[45,56], words=[5,5])), (0.0, Span(\"tests\", context=None, chars=[97,101], words=[11,11])), (0.0, Span(\"tests,\", context=None, chars=[25,30], words=[2,3])), (0.0, Span(\"hypertension\", context=None, chars=[142,153], words=[17,17])), (0.0, Span(\"hypertension\", context=None, chars=[12,23], words=[1,1]))]\n",
      "\n",
      "Document('19430480', Corpus (GWAS Corpus)) Genome-wide association study and meta-analysis find that over 40 loci affect risk of type 1 diabetes.\n",
      "Span(\"type 1 diabetes.\", context=None, chars=[82,97], words=[12,15])\n",
      "[(29.125847746080705, Span(\"type 1 diabetes.\", context=None, chars=[82,97], words=[12,15])), (0.0, Span(\"disorder\", context=None, chars=[45,52], words=[10,10])), (0.0, Span(\"(ASP)\", context=None, chars=[162,166], words=[25,27])), (0.0, Span(\"Type 1 diabetes (\", context=None, chars=[0,16], words=[0,3])), (-13.237207833461666, Span(\"included\", context=None, chars=[21,28], words=[4,4]))]\n",
      "\n",
      "Document('19430483', Corpus (GWAS Corpus)) Genome-wide association study identifies eight loci associated with blood pressure.\n",
      "Span(\"blood pressure\", context=None, chars=[6,19], words=[1,2])\n",
      "[(29.125847746080705, Span(\"blood pressure\", context=None, chars=[6,19], words=[1,2])), (0.0, Span(\"diastolic blood pressure\", context=None, chars=[46,69], words=[6,8])), (0.0, Span(\"blood pressure\", context=None, chars=[47,60], words=[6,7])), (0.0, Span(\"blood pressure\", context=None, chars=[129,142], words=[17,18])), (0.0, Span(\"hypertension.\", context=None, chars=[88,100], words=[11,12]))]\n",
      "\n",
      "Document('19448621', Corpus (GWAS Corpus)) Genome-wide association studies identify loci associated with age at menarche and age at natural menopause.\n",
      "Span(\"menopause.\", context=None, chars=[103,112], words=[15,16])\n",
      "[(29.125847746080705, Span(\"menopause.\", context=None, chars=[103,112], words=[15,16])), (29.125847746080705, Span(\"age at menarche\", context=None, chars=[68,82], words=[8,10])), (0.0, Span(\"mortality\", context=None, chars=[101,109], words=[16,16])), (0.0, Span(\"menopause\", context=None, chars=[35,43], words=[7,7])), (0.0, Span(\"Age at menarche\", context=None, chars=[0,14], words=[0,2]))]\n",
      "\n",
      "Document('19454037', Corpus (GWAS Corpus)) Genome-wide association study identifies single-nucleotide polymorphism in KCNB1 associated with left ventricular mass in humans: the HyperGEN Study.\n",
      "Span(\"with left ventricular\", context=None, chars=[94,114], words=[9,11])\n",
      "[(15.888639912619039, Span(\"with left ventricular\", context=None, chars=[94,114], words=[9,11])), (0.0, Span(\"hypertrophy\", context=None, chars=[72,82], words=[12,12])), (0.0, Span(\"on chromosomes 5,\", context=None, chars=[117,133], words=[19,22])), (0.0, Span(\"ages.\", context=None, chars=[102,106], words=[18,19])), (0.0, Span(\"hypertrophy\", context=None, chars=[135,145], words=[23,23]))]\n",
      "\n",
      "Document('19503088', Corpus (GWAS Corpus)) REL, encoding a member of the NF-kappaB family of transcription factors, is a newly defined risk locus for rheumatoid arthritis.\n",
      "Span(\"rheumatoid arthritis.\", context=None, chars=[94,114], words=[19,21])\n",
      "[(29.125847746080705, Span(\"rheumatoid arthritis.\", context=None, chars=[94,114], words=[19,21])), (0.0, Span(\"rheumatoid arthritis\", context=None, chars=[35,54], words=[5,6])), (0.0, Span(\"suggests disease\", context=None, chars=[124,139], words=[19,20])), (-13.237207833461666, Span(\"including\", context=None, chars=[213,221], words=[30,30])), (-461.98751011670691, Span(\", is\", context=None, chars=[58,61], words=[11,12]))]\n",
      "\n",
      "Document('19503597', Corpus (GWAS Corpus)) Meta-analysis of 28,141 individuals identifies common variants within five new loci that influence uric acid concentrations.\n",
      "Span(\"Uric Acid Concentrations.\", context=None, chars=[99,123], words=[13,16])\n",
      "[(29.125847746080705, Span(\"Uric Acid Concentrations.\", context=None, chars=[99,123], words=[13,16])), (1.9959427016534608, Span(\"protein (\", context=None, chars=[167,175], words=[27,28])), (0.0, Span(\"glucose\", context=None, chars=[95,101], words=[13,13])), (0.0, Span(\", metabolites\", context=None, chars=[126,138], words=[18,19])), (0.0, Span(\"hyperuricemia.\", context=None, chars=[211,224], words=[31,32]))]\n",
      "\n",
      "Document('19557161', Corpus (GWAS Corpus)) Genome-wide association scan meta-analysis identifies three Loci influencing adiposity and fat distribution.\n",
      "Span(\"Waist Circumference:\", context=None, chars=[136,155], words=[18,20])\n",
      "[(29.125847746080705, Span(\"Waist Circumference:\", context=None, chars=[136,155], words=[18,20])), (29.125847746080705, Span(\"Adiposity\", context=None, chars=[77,85], words=[8,8])), (1.9959427016534608, Span(\"waist circumference (\", context=None, chars=[270,290], words=[40,42])), (1.9959427016534608, Span(\"waist circumference (\", context=None, chars=[183,203], words=[30,32])), (0.0, Span(\"adiposity;\", context=None, chars=[118,127], words=[17,18]))]\n",
      "\n",
      "Document('19557197', Corpus (GWAS Corpus)) NRXN3 is a novel locus for waist circumference: a genome-wide association study from the CHARGE Consortium.\n",
      "Span(\"Adiposity\", context=None, chars=[185,193], words=[25,25])\n",
      "[(29.125847746080705, Span(\"Adiposity\", context=None, chars=[185,193], words=[25,25])), (29.125847746080705, Span(\"Waist Circumference:\", context=None, chars=[28,47], words=[6,8])), (1.9959427016534608, Span(\"body mass index (\", context=None, chars=[34,50], words=[6,9])), (1.9959427016534608, Span(\"waist circumference (\", context=None, chars=[125,145], words=[17,19])), (0.0, Span(\"obesity (\", context=None, chars=[150,158], words=[40,41]))]\n",
      "\n",
      "Document('19570815', Corpus (GWAS Corpus)) A genome-wide association study of northwestern Europeans involves the C-type natriuretic peptide signaling pathway in the etiology of human height variation.\n",
      "Span(\"height\", context=None, chars=[141,146], words=[19,19])\n",
      "[(29.125847746080705, Span(\"height\", context=None, chars=[141,146], words=[19,19])), (29.125847746080705, Span(\"natriuretic peptide\", context=None, chars=[78,96], words=[10,11])), (0.0, Span(\") of body\", context=None, chars=[61,69], words=[10,12])), (0.0, Span(\"body height\", context=None, chars=[66,76], words=[12,13])), (0.0, Span(\"height at\", context=None, chars=[258,266], words=[43,44]))]\n",
      "\n",
      "Document('19571809', Corpus (GWAS Corpus)) Common variants on chromosome 6p22.1 are associated with schizophrenia.\n",
      "Span(\"schizophrenia.\", context=None, chars=[57,70], words=[9,10])\n",
      "[(29.125847746080705, Span(\"schizophrenia.\", context=None, chars=[57,70], words=[9,10])), (1.9959427016534608, Span(\"Schizophrenia (\", context=None, chars=[98,112], words=[18,19])), (1.9959427016534608, Span(\"Schizophrenia\", context=None, chars=[200,212], words=[36,36])), (0.0, Span(\"schizophrenia\", context=None, chars=[38,50], words=[5,5])), (0.0, Span(\"Schizophrenia,\", context=None, chars=[0,13], words=[0,1]))]\n",
      "\n",
      "Document('19578366', Corpus (GWAS Corpus)) Variants in the CDKN2B and RTEL1 regions are associated with high-grade glioma susceptibility.\n",
      "Span(\"glioma\", context=None, chars=[76,81], words=[12,12])\n",
      "[(29.125847746080705, Span(\"glioma\", context=None, chars=[76,81], words=[12,12])), (1.9959427016534608, Span(\"Glioma\", context=None, chars=[268,273], words=[43,43])), (1.9959427016534608, Span(\"Cancer\", context=None, chars=[303,308], words=[52,52])), (0.0, Span(\"glioblastoma\", context=None, chars=[14,25], words=[3,3])), (0.0, Span(\"(AGS)\", context=None, chars=[281,285], words=[45,47]))]\n",
      "\n",
      "Document('19587794', Corpus (GWAS Corpus)) Common genetic variation near the phospholamban gene is associated with cardiac repolarisation: meta-analysis of three genome-wide association studies.\n",
      "Span(\"QT interval\", context=None, chars=[76,86], words=[11,12])\n",
      "[(0.0, Span(\"QT interval\", context=None, chars=[76,86], words=[11,12])), (0.0, Span(\"arrhythmias\", context=None, chars=[129,139], words=[19,19])), (0.0, Span(\"sudden cardiac death.\", context=None, chars=[145,165], words=[21,24])), (0.0, Span(\"QT interval at\", context=None, chars=[45,58], words=[6,8])), (0.0, Span(\"arrhythmias\", context=None, chars=[137,147], words=[19,19]))]\n",
      "\n",
      "Document('19597492', Corpus (GWAS Corpus)) Variants in ZFHX3 are associated with atrial fibrillation in individuals of European ancestry.\n",
      "Span(\"atrial fibrillation\", context=None, chars=[40,58], words=[6,7])\n",
      "[(29.125847746080705, Span(\"atrial fibrillation\", context=None, chars=[40,58], words=[6,7])), (1.9959427016534608, Span(\"atrial fibrillation (\", context=None, chars=[73,93], words=[11,13])), (0.0, Span(\"[RR]=\", context=None, chars=[161,165], words=[30,33])), (-13.237207833461666, Span(\"), an\", context=None, chars=[186,190], words=[43,45])), (-477.87615002932597, Span(\"(RR=\", context=None, chars=[96,99], words=[13,15]))]\n",
      "\n",
      "Document('19609347', Corpus (GWAS Corpus)) A genome-wide association study of hypertension and blood pressure in African Americans.\n",
      "Span(\"Blood Pressure\", context=None, chars=[52,65], words=[7,8])\n",
      "[(29.125847746080705, Span(\"Blood Pressure\", context=None, chars=[52,65], words=[7,8])), (29.125847746080705, Span(\"Hypertension\", context=None, chars=[35,46], words=[5,5])), (1.9959427016534608, Span(\"blood pressure (\", context=None, chars=[82,97], words=[12,14])), (1.9959427016534608, Span(\"blood pressure (\", context=None, chars=[38,53], words=[5,7])), (0.0, Span(\"hypertension\", context=None, chars=[227,238], words=[36,36]))]\n",
      "\n",
      "Document('19651812', Corpus (GWAS Corpus)) Common genetic variation near melatonin receptor MTNR1B contributes to raised plasma glucose and increased risk of type 2 diabetes among Indian Asians and European Caucasians.\n",
      "Span(\"Glucose\", context=None, chars=[87,93], words=[11,11])\n",
      "[(29.125847746080705, Span(\"Glucose\", context=None, chars=[87,93], words=[11,11])), (29.125847746080705, Span(\"Melatonin\", context=None, chars=[30,38], words=[4,4])), (29.125847746080705, Span(\"Type 2 Diabetes\", context=None, chars=[117,131], words=[16,18])), (0.0, Span(\"glucose\", context=None, chars=[94,100], words=[17,17])), (0.0, Span(\"blood glucose\", context=None, chars=[50,62], words=[6,7]))]\n",
      "\n",
      "Document('19668339', Corpus (GWAS Corpus)) Hippocampal atrophy as a quantitative trait in a genome-wide association study identifying novel susceptibility genes for Alzheimer's disease.\n",
      "Span(\"Hippocampal Atrophy\", context=None, chars=[0,18], words=[0,1])\n",
      "[(29.125847746080705, Span(\"Hippocampal Atrophy\", context=None, chars=[0,18], words=[0,1])), (29.125847746080705, Span(\"Alzheimer's Disease.\", context=None, chars=[122,141], words=[16,19])), (29.125847746080705, Span(\"as a Quantitative Trait\", context=None, chars=[20,42], words=[2,5])), (1.9959427016534608, Span(\"Alzheimer's Disease (\", context=None, chars=[83,103], words=[15,18])), (0.0, Span(\"protein\", context=None, chars=[101,107], words=[20,20]))]\n",
      "\n",
      "Document('19714205', Corpus (GWAS Corpus)) Genome-wide association study implicates chromosome 9q21.31 as a susceptibility locus for asthma in mexican children.\n",
      "Span(\"Asthma\", context=None, chars=[90,95], words=[12,12])\n",
      "[(29.125847746080705, Span(\"Asthma\", context=None, chars=[90,95], words=[12,12])), (0.0, Span(\"asthma.\", context=None, chars=[323,329], words=[48,49])), (0.0, Span(\"in the lung\", context=None, chars=[177,187], words=[24,26])), (0.0, Span(\"in the lung\", context=None, chars=[112,122], words=[14,16])), (0.0, Span(\"asthma\", context=None, chars=[83,88], words=[12,12]))]\n",
      "\n",
      "Document('19714249', Corpus (GWAS Corpus)) Powerful bivariate genome-wide association analyses suggest the SOX6 gene influencing both obesity and osteoporosis phenotypes in males.\n",
      "Span(\"Obesity\", context=None, chars=[93,99], words=[11,11])\n",
      "[(29.125847746080705, Span(\"Obesity\", context=None, chars=[93,99], words=[11,11])), (19.270490808768059, Span(\"Osteoporosis Phenotypes\", context=None, chars=[105,127], words=[13,14])), (1.9959427016534608, Span(\"bone mineral density (\", context=None, chars=[137,158], words=[25,28])), (0.0, Span(\"obesity phenotype\", context=None, chars=[53,69], words=[8,9])), (0.0, Span(\"(BMD),\", context=None, chars=[158,163], words=[28,31]))]\n",
      "\n",
      "Document('19721433', Corpus (GWAS Corpus)) Genome-wide pharmacogenomic analysis of response to treatment with antipsychotics.\n",
      "Span(\"Antipsychotics\", context=None, chars=[66,79], words=[8,8])\n",
      "[(29.125847746080705, Span(\"Antipsychotics\", context=None, chars=[66,79], words=[8,8])), (29.125847746080705, Span(\"Response to\", context=None, chars=[39,49], words=[4,5])), (19.270490808768059, Span(\"of Response\", context=None, chars=[36,46], words=[3,4])), (0.0, Span(\"antipsychotics\", context=None, chars=[94,107], words=[15,15])), (0.0, Span(\"risperidone\", context=None, chars=[88,98], words=[15,15]))]\n",
      "\n",
      "Document('19734902', Corpus (GWAS Corpus)) Genome-wide association study identifies variants at CLU and PICALM associated with Alzheimer's disease.\n",
      "Span(\"Alzheimer's disease,\", context=None, chars=[88,107], words=[11,14])\n",
      "[(29.125847746080705, Span(\"Alzheimer's disease,\", context=None, chars=[88,107], words=[11,14])), (0.0, Span(\"of Alzheimer's disease\", context=None, chars=[55,76], words=[7,10])), (-13.237207833461666, Span(\"), including\", context=None, chars=[94,105], words=[26,28])), (-13.237207833461666, Span(\"with the\", context=None, chars=[87,94], words=[16,17])), (-477.87615002932597, Span(\"AD\", context=None, chars=[135,136], words=[22,22]))]\n",
      "\n",
      "Document('19772629', Corpus (GWAS Corpus)) Genomewide association study for onset age in Parkinson disease.\n",
      "Span(\"onset age\", context=None, chars=[33,41], words=[4,5])\n",
      "[(29.125847746080705, Span(\"onset age\", context=None, chars=[33,41], words=[4,5])), (29.125847746080705, Span(\"Parkinson disease.\", context=None, chars=[46,63], words=[7,9])), (1.9959427016534608, Span(\"Parkinson disease (\", context=None, chars=[16,34], words=[4,6])), (0.0, Span(\"the disease\", context=None, chars=[94,104], words=[13,14])), (0.0, Span(\"onset age\", context=None, chars=[30,38], words=[4,5]))]\n",
      "\n",
      "Document('19779542', Corpus (GWAS Corpus)) A genome-wide association analysis identified a novel susceptible locus for pathological myopia at 11q24.1.\n",
      "Span(\"Pathological Myopia at\", context=None, chars=[76,97], words=[10,12])\n",
      "[(29.125847746080705, Span(\"Pathological Myopia at\", context=None, chars=[76,97], words=[10,12])), (0.0, Span(\"the disease\", context=None, chars=[134,144], words=[23,24])), (0.0, Span(\"pathological myopia\", context=None, chars=[49,67], words=[6,7])), (0.0, Span(\"legal blindness\", context=None, chars=[127,141], words=[26,27])), (0.0, Span(\"myopia,\", context=None, chars=[38,44], words=[6,7]))]\n",
      "\n",
      "Document('19798445', Corpus (GWAS Corpus)) Genetic determinants of circulating sphingolipid concentrations in European populations.\n",
      "Span(\"Sphingolipid\", context=None, chars=[36,47], words=[4,4])\n",
      "[(29.125847746080705, Span(\"Sphingolipid\", context=None, chars=[36,47], words=[4,4])), (1.9959427016534608, Span(\", myocardial infarction (\", context=None, chars=[172,196], words=[21,24])), (1.9959427016534608, Span(\"infarction (MI),\", context=None, chars=[185,200], words=[23,27])), (0.0, Span(\"sphingolipids\", context=None, chars=[256,268], words=[34,34])), (0.0, Span(\"human disease\", context=None, chars=[297,309], words=[39,40]))]\n",
      "\n",
      "Document('19801982', Corpus (GWAS Corpus)) Twenty bone-mineral-density loci identified by large-scale meta-analysis of genome-wide association studies.\n",
      "Span(\"bone mineral density\", context=None, chars=[7,26], words=[1,3])\n",
      "[(29.125847746080705, Span(\"bone mineral density\", context=None, chars=[7,26], words=[1,3])), (1.9959427016534608, Span(\"Bone mineral density (\", context=None, chars=[0,21], words=[0,3])), (0.0, Span(\"osteoporosis\", context=None, chars=[90,101], words=[17,17])), (0.0, Span(\"(BMD)\", context=None, chars=[21,25], words=[3,5])), (0.0, Span(\"complex trait\", context=None, chars=[42,54], words=[9,10]))]\n",
      "\n",
      "Document('19812673', Corpus (GWAS Corpus)) A genome-wide linkage and association scan reveals novel loci for autism.\n",
      "Span(\"AUTISM.\", context=None, chars=[66,72], words=[10,11])\n",
      "[(29.125847746080705, Span(\"AUTISM.\", context=None, chars=[66,72], words=[10,11])), (0.0, Span(\"autism\", context=None, chars=[129,134], words=[20,20])), (0.0, Span(\"brains\", context=None, chars=[63,68], words=[10,10])), (0.0, Span(\"autism (\", context=None, chars=[229,236], words=[35,36])), (0.0, Span(\"autism\", context=None, chars=[140,145], words=[23,23]))]\n",
      "\n",
      "Document('19820699', Corpus (GWAS Corpus)) Common variants in TMPRSS6 are associated with iron status and erythrocyte volume.\n",
      "Span(\"erythrocyte\", context=None, chars=[65,75], words=[10,10])\n",
      "[(29.125847746080705, Span(\"erythrocyte\", context=None, chars=[65,75], words=[10,10])), (0.0, Span(\"involvement of\", context=None, chars=[31,44], words=[4,5])), (0.0, Span(\"haemoglobin levels (\", context=None, chars=[59,78], words=[9,11])), (0.0, Span(\"erythrocyte\", context=None, chars=[159,169], words=[38,38])), (0.0, Span(\"(MCV,\", context=None, chars=[188,192], words=[42,44]))]\n",
      "\n",
      "Document('19874204', Corpus (GWAS Corpus)) IL21R and PTH may underlie variation of femoral neck bone mineral density as revealed by a genome-wide association study.\n",
      "Span(\"Femoral Neck Bone\", context=None, chars=[43,59], words=[7,9])\n",
      "[(29.125847746080705, Span(\"Femoral Neck Bone\", context=None, chars=[43,59], words=[7,9])), (29.125847746080705, Span(\"Bone Mineral Density\", context=None, chars=[56,75], words=[9,11])), (1.9959427016534608, Span(\", parathyroid hormone (\", context=None, chars=[63,85], words=[12,15])), (1.9959427016534608, Span(\"Bone mineral density (\", context=None, chars=[0,21], words=[0,3])), (0.0, Span(\"( PTH )\", context=None, chars=[85,91], words=[15,17]))]\n",
      "\n",
      "Document('19915575', Corpus (GWAS Corpus)) Genome-wide association study reveals genetic risk underlying Parkinson's disease.\n"
     ]
    },
    {
     "ename": "UnicodeEncodeError",
     "evalue": "'ascii' codec can't encode character u'\\u2019' in position 15: ordinal not in range(128)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mUnicodeEncodeError\u001b[0m                        Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-736-e7d210a915d8>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0md\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mcorpus\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdocuments\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m     \u001b[0;32mprint\u001b[0m \u001b[0md\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkb\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpaper_by_pmid\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0md\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtitle\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 7\u001b[0;31m     \u001b[0;32mprint\u001b[0m \u001b[0mresults\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0md\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      8\u001b[0m     \u001b[0;32mprint\u001b[0m \u001b[0msorted\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mscore_dict\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0md\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreverse\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;36m5\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      9\u001b[0m     \u001b[0;32mprint\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mUnicodeEncodeError\u001b[0m: 'ascii' codec can't encode character u'\\u2019' in position 15: ordinal not in range(128)"
     ]
    }
   ],
   "source": [
    "# doc_set = {c.context.document.name for c in results}\n",
    "# missing_docs = {doc.name for doc in corpus.documents} - doc_set\n",
    "# docs = sorted(list(missing_docs))\n",
    "# print len(docs)\n",
    "for d in corpus.documents:\n",
    "    print d, kb.paper_by_pmid(d.name).title\n",
    "    print results.get(d.name, None)\n",
    "    print sorted(score_dict[d.name], reverse=True)[:5]\n",
    "    print"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "\n",
    "### Save results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 742,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "with open('phenotypes.extracted.tsv', 'w') as f:\n",
    "    for d in corpus.documents:\n",
    "        # pick the top two results:\n",
    "        (_, r1), (_, r2) = sorted(score_dict[d.name], reverse=True)[:2]\n",
    "        # if both are in title, report both, otherwise report only the best one\n",
    "        if r2.context.position == 0:\n",
    "            phen = r1.get_span() + '|' + r2.get_span()\n",
    "        else:\n",
    "            phen = r1.get_span()\n",
    "        out_str = u'%s\\t%s\\t\\n' % (d.name, phen)        \n",
    "        f.write(out_str.encode(\"UTF-8\"))\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
