{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Phenotype extraction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import sys\n",
    "import cPickle\n",
    "\n",
    "# import snorkel and gwasdb\n",
    "sys.path.append('../snorkel')\n",
    "sys.path.append('../src')\n",
    "sys.path.append('../src/crawler')\n",
    "\n",
    "# set up paths\n",
    "abstract_dir = '../data/db/papers'\n",
    "\n",
    "# set up matplotlib\n",
    "import matplotlib\n",
    "%matplotlib inline\n",
    "matplotlib.rcParams['figure.figsize'] = (12,4)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extract phenotype candidates from papers"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load corpus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from snorkel.parser import XMLDocParser\n",
    "from extractor.parser import UnicodeXMLDocParser, GWASXMLAbstractParser\n",
    "\n",
    "xml_parser = GWASXMLAbstractParser(\n",
    "    path=abstract_dir,\n",
    "    doc='./*',\n",
    "    title='.//front//article-title//text()',\n",
    "    abstract='.//abstract//p//text()',\n",
    "    par1='.//body/p[1]//text()',\n",
    "    id='.//article-id[@pub-id-type=\"pmid\"]/text()',\n",
    "    keep_xml_tree=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 13.5 s, sys: 994 ms, total: 14.5 s\n",
      "Wall time: 1min 28s\n"
     ]
    }
   ],
   "source": [
    "from snorkel.parser import HTMLParser\n",
    "from snorkel.parser import SentenceParser\n",
    "from snorkel.parser import CorpusParser\n",
    "\n",
    "sent_parser = SentenceParser()\n",
    "html_parser = HTMLParser(path='../data/db/papers/')\n",
    "\n",
    "corpus_name = 'gwas-text-corpus.pkl'\n",
    "\n",
    "try:\n",
    "    with open(corpus_name,\"r\") as pkl:\n",
    "        corpus = cPickle.load(pkl)\n",
    "except:\n",
    "    cp = CorpusParser(xml_parser, sent_parser)\n",
    "    %time corpus = cp.parse_corpus(name='GWAS Corpus')\n",
    "    # pickling currently doesn't work...\n",
    "#     with open(corpus_name,\"w\") as pkl:\n",
    "#         corpus = cPickle.dump(corpus, pkl)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extract candidates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 7min 50s, sys: 4.49 s, total: 7min 55s\n",
      "Wall time: 7min 56s\n",
      "71170 candidates extracted\n"
     ]
    }
   ],
   "source": [
    "from snorkel.candidates import Ngrams\n",
    "from snorkel.matchers import DictionaryMatch, Union, RegexMatchSpan\n",
    "from snorkel.candidates import EntityExtractor\n",
    "from snorkel.utils import slice_into_ngrams\n",
    "\n",
    "from extractor.util import change_name\n",
    "from extractor.matcher import PhenotypeMatcher\n",
    "from db.kb import KnowledgeBase\n",
    "\n",
    "def make_ngrams(L, n_max=10, n_min=3, delim=' '):\n",
    "    for l in L:\n",
    "        yield l\n",
    "        tokens = l.strip().split(delim)\n",
    "        for ngram in slice_into_ngrams(tokens, n_max=n_max, n_min=n_min, delim=delim):\n",
    "            yield ngram\n",
    "\n",
    "# Define a candidate space\n",
    "ngrams = Ngrams(n_max=7)\n",
    "\n",
    "# collect phenotype list\n",
    "kb = KnowledgeBase()\n",
    "# efo phenotypes\n",
    "efo_phenotype_list0 = kb.get_phenotype_candidates(source='efo-matching', peek=False) # TODO: remove peaking\n",
    "efo_phenotype_list = list(make_ngrams(efo_phenotype_list0))\n",
    "# snomed keywords\n",
    "snomed_phenotype_list = kb.get_phenotype_candidates(source='snomed')\n",
    "# mesh diseases\n",
    "mesh_phenotype_list0 = kb.get_phenotype_candidates(source='mesh')\n",
    "mesh_phenotype_list = list(make_ngrams(mesh_phenotype_list0))\n",
    "# mesh chemicals\n",
    "chem_phenotype_list = kb.get_phenotype_candidates(source='chemical')\n",
    "# regex matches\n",
    "rgx = u'[A-Za-z\\u2013-]+ (disease|trait|phenotype|outcome|response|quantitative trait|measurement|response|side effects)s?'\n",
    "\n",
    "# Define matchers\n",
    "efo_phen_matcher = PhenotypeMatcher(d=efo_phenotype_list, ignore_case=True, mod_fn=change_name)\n",
    "snom_phen_matcher = PhenotypeMatcher(d=snomed_phenotype_list, ignore_case=True, mod_fn=change_name)\n",
    "mesh_phen_matcher = PhenotypeMatcher(d=mesh_phenotype_list, ignore_case=True, mod_fn=change_name)\n",
    "chem_phen_matcher = DictionaryMatch(d=chem_phenotype_list, longest_match_only=True, ignore_case=True)\n",
    "regex_phen_matcher = RegexMatchSpan(rgx=rgx)\n",
    "# phen_matcher = Union(efo_phen_matcher, mesh_phen_matcher, chem_phen_matcher, regex_phen_matcher)\n",
    "phen_matcher = Union(efo_phen_matcher, snom_phen_matcher, mesh_phen_matcher, chem_phen_matcher, regex_phen_matcher)\n",
    "\n",
    "# Extractor\n",
    "phen_extractor = EntityExtractor(ngrams, phen_matcher)\n",
    "\n",
    "# collect candidates\n",
    "%time phen_c = phen_extractor.extract(corpus.get_sentences(), name='all')\n",
    "print len(phen_c), 'candidates extracted'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def make_ngrams(L, n_max=10, n_min=3, delim=' '):\n",
    "    for l in L:\n",
    "        yield l\n",
    "        tokens = l.strip().split(delim)\n",
    "        for ngram in slice_into_ngrams(tokens, n_max=n_max, n_min=n_min, delim=delim):\n",
    "            yield ngram"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We would like to remove nested candidates (manually, for now)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20137 candidates dropped, now we have 51033\n"
     ]
    }
   ],
   "source": [
    "# load existing candidates into a dict\n",
    "span_dict = { str(span.context) : list() for span in phen_c }\n",
    "for span in phen_c:\n",
    "    span_dict[str(span.context)].append( (span.char_start, span.char_end) )\n",
    "\n",
    "def nested(ivl1, ivl2):\n",
    "    if ivl1 != ivl2 and ivl2[0] <= ivl1[0] <= ivl1[1] <= ivl2[1]:\n",
    "        return True\n",
    "    else:\n",
    "        return False\n",
    "\n",
    "new_phen_c = list()\n",
    "for span in phen_c:\n",
    "    span_ivl = span.char_start, span.char_end\n",
    "    span_name = str(span.context)\n",
    "    if all([not nested(span_ivl, other_ivl) for other_ivl in span_dict[span_name]]):\n",
    "        new_phen_c.append(span)\n",
    "        \n",
    "print len(phen_c) - len(new_phen_c), 'candidates dropped, now we have', len(new_phen_c)\n",
    "phen_c = new_phen_c"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create gold-truth set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 1282 gold mentions, e.g.:\n",
      "[('23251661', u'http://www.ebi.ac.uk/efo/EFO_0004749'), ('19587794', u'http://www.ebi.ac.uk/efo/EFO_0004682'), ('24376456', u'http://www.ebi.ac.uk/efo/EFO_0006340'), ('20081858', u'http://www.ebi.ac.uk/efo/EFO_0004501'), ('23776548', u'http://www.ebi.ac.uk/efo/EFO_0004731')]\n",
      "1063 535\n"
     ]
    }
   ],
   "source": [
    "from db.kb import KnowledgeBase\n",
    "from nltk.stem import PorterStemmer\n",
    "from extractor.util import change_name\n",
    "\n",
    "kb = KnowledgeBase() # reload\n",
    "gold_set_phens = frozenset \\\n",
    "([ \n",
    "    (doc.name, phen.ontology_ref) for doc in corpus.documents \n",
    "                                  for phen in kb.phen_by_pmid(doc.name, source='efo')\n",
    "])\n",
    "\n",
    "# map phenotype names to their id\n",
    "phen2id = \\\n",
    "{\n",
    "    change_name(syn) : phen.ontology_ref for doc in corpus.documents\n",
    "    \n",
    "                                                  for phen in kb.phen_by_pmid(doc.name, source='efo')\n",
    "                                                  for syn in [phen.name] + phen.synonyms.split('|')\n",
    "}\n",
    "\n",
    "# this is the more correct version (code below should be changed to use it)\n",
    "# we collect a set of EFOs matching a string; we use this only at the very end for now\n",
    "phen2idset = dict()\n",
    "for doc in corpus.documents:\n",
    "    for phen in kb.phen_by_pmid(doc.name, source='efo'):\n",
    "        synonyms = [phen.name] + phen.synonyms.split('|')\n",
    "        ngrams = make_ngrams(synonyms)\n",
    "        for ngram in ngrams:\n",
    "            syn_name = change_name(ngram)\n",
    "            if syn_name not in phen2idset: phen2idset[syn_name] = set()\n",
    "            phen2idset[syn_name].add(phen.id)\n",
    "\n",
    "id2phen = \\\n",
    "{\n",
    "    phen.ontology_ref : phen for doc in corpus.documents\n",
    "                             for phen in kb.phen_by_pmid(doc.name, source='efo')\n",
    "}\n",
    "\n",
    "print 'Found %d gold mentions, e.g.:' % len(gold_set_phens)\n",
    "print list(gold_set_phens)[:5]\n",
    "print len(phen2id), len(id2phen)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7\n"
     ]
    }
   ],
   "source": [
    "from db import db_session\n",
    "from db.schema import *\n",
    "\n",
    "# paper = db_session.query(Paper).filter(Paper.pubmed_id=='20195266').first()\n",
    "# print [ (assoc.phenotype.name, [(p.name, p.source) for p in assoc.phenotype.equivalents]) for assoc in paper.associations ]\n",
    "# print first_degree_phens\n",
    "\n",
    "from db.kb import KnowledgeBase\n",
    "kb = KnowledgeBase() # reload\n",
    "print len(kb.phen_by_pmid('20195266', source='efo'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Statistics"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, we need to understand whether the extraction or the classification approach will be better."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Number of phenotypes per paper"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Number of EFO phenotypes per paper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "docid2efo = \\\n",
    "{\n",
    "    doc.name : [ phen.name for phen in kb.phen_by_pmid(doc.name, source='efo') ] for doc in corpus.documents\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Number of GWAS catalog (i.e. aggregate) phenotypes per paper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GWAS catalog phenotype numbers per paper: [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 12, 12, 12, 12, 12, 12, 12, 12, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 14, 14, 14, 14, 14, 14, 14, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 16, 16, 16, 16, 16, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 18, 18, 18, 18, 18, 18, 18, 19, 19, 19, 19, 19, 19, 19, 19, 19, 20, 20, 20, 20, 21, 21, 21, 21, 21, 21, 22, 22, 22, 23, 23, 23, 24, 24, 24, 25, 25, 26, 26, 27, 28, 28, 28, 29, 29, 30, 30, 30, 31, 31, 31, 31, 32, 33, 33, 34, 34, 34, 34, 34, 35, 35, 35, 35, 36, 36, 37, 37, 38, 38, 38, 39, 39, 39, 40, 40, 41, 41, 42, 42, 47, 48, 48, 49, 50, 50, 50, 53, 56, 59, 60, 62, 69, 73, 74, 87, 89, 98, 151, 163, 166, 166, 168, 181, 199, 219, 241, 245, 299, 699, 949]\n"
     ]
    }
   ],
   "source": [
    "docid2efo = \\\n",
    "{\n",
    "    doc.name : [ phen.name for phen in kb.phen_by_pmid(doc.name, source='gwas_catalog') ] for doc in corpus.documents\n",
    "}\n",
    "print 'GWAS catalog phenotype numbers per paper:', sorted([len(v) for d, v in docid2efo.items()])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Candidate recall statistics"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Over EFO phenotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Statistics over EFO phenotypes:\n",
      "# of gold annotations\t= 1282\n",
      "# of candidates\t\t= 27861\n",
      "Candidate recall\t= 0.534\n",
      "Candidate precision\t= 0.025\n"
     ]
    }
   ],
   "source": [
    "from extractor.util import gold_phen_stats\n",
    "gold_phen_stats(phen_c, gold_set_phens, phen2id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Why is the recall low?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "597\n",
      "24204828\n",
      "homa-ir http://www.ebi.ac.uk/efo/EFO_0004501\n",
      "homa-b http://www.ebi.ac.uk/efo/EFO_0004469\n",
      "fasting blood insulin measurement http://www.ebi.ac.uk/efo/EFO_0004466\n",
      "fasting blood insulin measurement http://www.ebi.ac.uk/efo/EFO_0004466\n",
      "fasting blood insulin|fasting blood insulin level|fasting insulin-related traits\n",
      "Genome-wide contribution of genotype by environment interaction to variation of diabetes-related traits.\n",
      "Genome-Wide Contribution of Genotype by Environment Interaction to Variation of Diabetes-Related Traits. While genome-wide association studies (GWAS) and candidate gene approaches have identified many genetic variants that contribute to disease risk as main effects, the impact of genotype by environment (GxE) interactions remains rather under-surveyed. To explore the importance of GxE interactions for diabetes-related traits, a tool for Genome-wide Complex Trait Analysis (GCTA) was used to examine GxE variance contribution of 15 macronutrients and lifestyle to the total phenotypic variance of diabetes-related traits at the genome-wide level in a European American population. GCTA identified two key environmental factors making significant contributions to the GxE variance for diabetes-related traits: carbohydrate for fasting insulin (25.1% of total variance,  P -nominal = 0.032) and homeostasis model assessment of insulin resistance (HOMA-IR) (24.2% of total variance,  P -nominal = 0.035), n-6 polyunsaturated fatty acid (PUFA) for HOMA-β-cell-function (39.0% of total variance,  P -nominal = 0.005). To demonstrate and support the results from GCTA, a GxE GWAS was conducted with each of the significant dietary factors and a control E factor (dietary protein), which contributed a non-significant GxE variance. We observed that GxE GWAS for the environmental factor contributing a significant GxE variance yielded more significant SNPs than the control factor. For each trait, we selected all significant SNPs produced from GxE GWAS, and conducted anew the GCTA to estimate the variance they contributed. We noted the variance contributed by these SNPs is higher than that of the control. In conclusion, we utilized a novel method that demonstrates the importance of genome-wide GxE interactions in explaining the variance of diabetes-related traits.\n",
      "\n",
      "23209189\n",
      "type ii diabetes mellitus http://www.ebi.ac.uk/efo/EFO_0001360\n",
      "type ii diabetes mellitus http://www.ebi.ac.uk/efo/EFO_0001360\n",
      "diabetes mellitus type 02|non-insulin dependent diabetes mellitus|stable diabetes mellitus|diabetes mellitus\n",
      "Genome-wide association study for type 2 diabetes in Indians identifies a new susceptibility locus at 2q21.\n",
      "Genome-Wide Association Study for Type 2 Diabetes in Indians Identifies a New Susceptibility Locus at 2q21. Indians undergoing socioeconomic and lifestyle transitions will be maximally affected by epidemic of type 2 diabetes (T2D). We conducted a two-stage genome-wide association study of T2D in 12,535 Indians, a less explored but high-risk group. We identified a new type 2 diabetes–associated locus at 2q21, with the lead signal being rs6723108 (odds ratio 1.31;  P  = 3.32 × 10 −9 ). Imputation analysis refined the signal to rs998451 (odds ratio 1.56;  P  = 6.3 × 10 −12 ) within  TMEM163  that encodes a probable vesicular transporter in nerve terminals. TMEM163  variants also showed association with decreased fasting plasma insulin and homeostatic model assessment of insulin resistance, indicating a plausible effect through impaired insulin secretion. The 2q21 region also harbors  RAB3GAP1  and  ACMSD ; those are involved in neurologic disorders. Forty-nine of 56 previously reported signals showed consistency in direction with similar effect sizes in Indians and previous studies, and 25 of them were also associated ( P  < 0.05). Known loci and the newly identified 2q21 locus altogether explained 7.65% variance in the risk of T2D in Indians. Our study suggests that common susceptibility variants for T2D are largely the same across populations, but also reveals a population-specific locus and provides further insights into genetic architecture and etiology of T2D.\n",
      "\n",
      "23535730\n",
      "ovarian carcinoma http://www.ebi.ac.uk/efo/EFO_0001075\n",
      "ovarian carcinoma http://www.ebi.ac.uk/efo/EFO_0001075\n",
      "neoplasms\n",
      "GWAS meta-analysis and replication identifies three new susceptibility loci for ovarian cancer.\n",
      "GWAS meta-analysis and replication identifies three new susceptibility loci for ovarian cancer. Genome wide association studies (GWAS) have identified four susceptibility loci for epithelial ovarian cancer (EOC) with another two loci being close to genome-wide significance. We pooled data from a GWAS conducted in North America with another GWAS from the United Kingdom. We selected the top 24,551 SNPs for inclusion on the iCOGS custom genotyping array. Follow-up genotyping was carried out in 18,174 cases and 26,134 controls from 43 studies from the Ovarian Cancer Association Consortium. We validated the two loci at 3q25 and 17q21 previously near genome-wide significance and identified three novel loci associated with risk; two loci associated with all EOC subtypes, at 8q21 (rs11782652, P=5.5×10 -9 ) and 10p12 (rs1243180; P=1.8×10 -8 ), and another locus specific to the serous subtype at 17q12 (rs757210; P=8.1×10 -10 ). An integrated molecular analysis of genes and regulatory regions at these loci provided evidence for functional mechanisms underlying susceptibility that implicates  CHMP4C  in the pathogenesis of ovarian cancer.\n",
      "\n",
      "22044751\n",
      "renal sinus adipose tissue measurement http://www.ebi.ac.uk/efo/EFO_0004864\n",
      "renal sinus adipose tissue measurement http://www.ebi.ac.uk/efo/EFO_0004864\n",
      "\n",
      "Heritability and genome-wide association analysis of renal sinus fat accumulation in the Framingham Heart Study.\n",
      "Heritability and genome-wide association analysis of renal sinus fat accumulation in the Framingham Heart Study. Ectopic fat accumulation in the renal sinus is associated with chronic kidney disease and hypertension. The genetic contributions to renal sinus fat accumulation in humans have not been well characterized. The present analysis consists of participants from the Framingham Offspring and Third Generation who underwent computed tomography; renal sinus fat and visceral adipose tissue (VAT) were quantified. Renal sinus fat was natural log transformed and sex- and cohort-specific residuals were created, adjusted for (1) age, (2) age and body mass index (BMI), and (3) age and VAT. Residuals were pooled and used to calculate heritability using variance-components analysis in SOLAR. A genome-wide association study (GWAS) for renal sinus fat was performed using an additive model with approximately 2.5 million imputed single nucleotide polymorphisms (SNPs). Finally, we identified the associations of renal sinus fat in our GWAS results with validated SNPs for renal function (n = 16), BMI (n = 32), and waist-to-hip ratio (WHR, n = 14), and applied a multi-SNP genetic risk score method to determine if the SNPs for each renal and obesity trait were in aggregate associated with renal sinus fat. The heritability of renal sinus fat was 39% (p < 0.0001); results were not materially different after adjustment for BMI (39%) or VAT (40%). No SNPs reached genome-wide significance in our GWAS. In our candidate gene analysis, we observed nominal, direction consistent associations with renal sinus fat for one SNP associated with renal function (p = 0.01), two associated with BMI (p < 0.03), and two associated with WHR (p < 0.03); however, none remained significant after accounting for multiple testing. Finally, we observed that in aggregate, the 32 SNPs associated with BMI were nominally associated with renal sinus fat (multi-SNP genetic risk score p = 0.03). Renal sinus fat is a heritable trait, even after accounting for generalized and abdominal adiposity. This provides support for further research into the genetic determinants of renal sinus fat. While our study was underpowered to detect genome-wide significant loci, our candidate gene BMI risk score results suggest that variability in renal sinus fat may be associated with SNPs previously known to be associated with generalized adiposity.\n",
      "\n",
      "20932310\n",
      "beta-amyloid 1-42 measurement http://www.ebi.ac.uk/efo/EFO_0004670\n",
      "t-tau measurement http://www.ebi.ac.uk/efo/EFO_0004760\n",
      "p-tau measurement http://www.ebi.ac.uk/efo/EFO_0004763\n",
      "beta-amyloid 1-42 measurement http://www.ebi.ac.uk/efo/EFO_0004670\n",
      "csf a1-42 levels|csf ab1-42 levels|amyloid-beta 1-42 measurement\n",
      "Genome-wide association reveals genetic effects on human A&#x003b2;42 and &#x003c4; protein levels in cerebrospinal fluids: a case control study.\n",
      "Genome-wide association reveals genetic effects on human Aβ 42  and τ protein levels in cerebrospinal fluids: a case control study. Alzheimer's disease (AD) is common and highly heritable with many genes and gene variants associated with AD in one or more studies, including APOE ε2/ε3/ε4. However, the genetic backgrounds for normal cognition, mild cognitive impairment (MCI) and AD in terms of changes in cerebrospinal fluid (CSF) levels of Aβ 1-42 , T-tau, and P-tau 181P , have not been clearly delineated. We carried out a genome-wide association study (GWAS) in order to better define the genetic backgrounds to these three states in relation to CSF levels. Subjects were participants in the Alzheimer's Disease Neuroimaging Initiative (ADNI). The GWAS dataset consisted of 818 participants (mainly Caucasian) genotyped using the Illumina Human Genome 610 Quad BeadChips. This sample included 410 subjects (119 Normal, 115 MCI and 176 AD) with measurements of CSF Aβ 1-42 , T-tau, and P-tau 181P  Levels. We used PLINK to find genetic associations with the three CSF biomarker levels. Association of each of the 498,205 SNPs was tested using additive, dominant, and general association models while considering APOE genotype and age. Finally, an effort was made to better identify relevant biochemical pathways for associated genes using the ALIGATOR software. We found that there were some associations with APOE genotype although CSF levels were about the same for each subject group; CSF Aβ 1-42  levels decreased with APOE gene dose for each subject group. T-tau levels tended to be higher among AD cases than among normal subjects. From adjusted result using APOE genotype and age as covariates, no SNP was associated with CSF levels among AD subjects. CYP19A1  'aromatase' (rs2899472),  NCAM2 , and multiple SNPs located on chromosome 10 near the  ARL5B  gene demonstrated the strongest associations with Aβ 1-42  in normal subjects. Two genes found to be near the top SNPs,  CYP19A1  (rs2899472, p = 1.90 × 10 -7 ) and  NCAM2  (rs1022442, p = 2.75 × 10 -7 ) have been reported as genetic factors related to the progression of AD from previous studies. In AD subjects, APOE ε2/ε3 and ε2/ε4 genotypes were associated with elevated T-tau levels and ε4/ε4 genotype was associated with elevated T-tau and P-tau 181P  levels. Pathway analysis detected several biological pathways implicated in Normal with CSF β-amyloid peptide (Aβ 1-42 ). Our genome-wide association analysis identified several SNPs as important factors for CSF biomarker. We also provide new evidence for additional candidate genetic risk factors from pathway analysis that can be tested in further studies.\n",
      "\n",
      "22792070\n",
      "bone density http://www.ebi.ac.uk/efo/EFO_0003923\n",
      "bone density http://www.ebi.ac.uk/efo/EFO_0003923\n",
      "density\n",
      "Meta-analysis of genome-wide scans for total body BMD in children and adults reveals allelic heterogeneity and age-specific effects at the WNT16 locus.\n",
      "Meta-Analysis of Genome-Wide Scans for Total Body BMD in Children and Adults Reveals Allelic Heterogeneity and Age-Specific Effects at the  WNT16  Locus WNT16  Influences Bone Mineral Density, Cortical Bone Thickness, Bone Strength, and Osteoporotic Fracture Risk. To identify genetic loci influencing bone accrual, we performed a genome-wide association scan for total-body bone mineral density (TB-BMD) variation in 2,660 children of different ethnicities. We discovered variants in 7q31.31 associated with BMD measurements, with the lowest P = 4.1×10 −11  observed for rs917727 with minor allele frequency of 0.37. We sought replication for all SNPs located ±500 kb from rs917727 in 11,052 additional individuals from five independent studies including children and adults, together with  de novo  genotyping of rs3801387 (in perfect linkage disequilibrium (LD) with rs917727) in 1,014 mothers of children from the discovery cohort. The top signal mapping in the surroundings of  WNT16  was replicated across studies with a meta-analysis P = 2.6×10 −31  and an effect size explaining between 0.6%–1.8% of TB-BMD variance. Conditional analyses on this signal revealed a secondary signal for total body BMD (P = 1.42×10 −10 ) for rs4609139 and mapping to  C7orf58 . We also examined the genomic region for association with skull BMD to test if the associations were independent of skeletal loading. We identified two signals influencing skull BMD variation, including rs917727 (P = 1.9×10 −16 ) and rs7801723 (P = 8.9×10 −28 ), also mapping to  C7orf58 ( r 2  = 0.50 with rs4609139). Wnt16  knockout (KO) mice with reduced total body BMD and gene expression profiles in human bone biopsies support a role of  C7orf58  and  WNT16  on the BMD phenotypes observed at the human population level. In summary, we detected two independent signals influencing total body and skull BMD variation in children and adults, thus demonstrating the presence of allelic heterogeneity at the  WNT16  locus. One of the skull BMD signals mapping to  C7orf58  is mostly driven by children, suggesting temporal determination on peak bone mass acquisition. Our life-course approach postulates that these genetic effects influencing peak bone mass accrual may impact the risk of osteoporosis later in life. Genetic investigations on bone mineral density (BMD) variation in children allow the identification of factors determining peak bone mass and their influence on developing osteoporosis later in life. We ran a genome-wide association study (GWAS) for total body BMD based on 2,660 children of different ethnic backgrounds, followed by replication in an additional 12,066 individuals comprising children, young adults, and elderly populations. Our GWAS meta-analysis identified two independent signals in the 7q31.31 locus, arising from SNPs in the vicinity of  WNT16 ,  FAM3C , and  C7orf58 . These variants were also associated with skull BMD, a skeletal trait with much less environmental influence for which one of the signals displayed age-specific effects. Integration of functional studies in a  Wnt16  knockout mouse model and gene expression profiles in human bone tissue provided additional evidence that  WNT16  and  C7orf58  underlie the described associations. All together our findings demonstrate the relevance of these factors for bone biology, the attainment of peak bone mass, and their likely impact on bone fragility later in life.\n",
      "\n",
      "23834954\n",
      "post-operative sensory disturbance http://www.ebi.ac.uk/efo/EFO_0005324\n",
      "post-operative sensory disturbance http://www.ebi.ac.uk/efo/EFO_0005324\n",
      "\n",
      "Genome-wide association study of sensory disturbances in the inferior alveolar nerve after bilateral sagittal split ramus osteotomy.\n",
      "Genome-wide association study of sensory disturbances in the inferior alveolar nerve after bilateral sagittal split ramus osteotomy. Bilateral sagittal split ramus osteotomy (BSSRO) is a common orthognatic surgical procedure. Sensory disturbances in the inferior alveolar nerve, including hypoesthesia and dysesthesia, are frequently observed after BSSRO, even without distinct nerve injury. The mechanisms that underlie individual differences in the vulnerability to sensory disturbances have not yet been elucidated. The present study investigated the relationships between genetic polymorphisms and the vulnerability to sensory disturbances after BSSRO in a genome-wide association study (GWAS). A total of 304 and 303 patients who underwent BSSRO were included in the analyses of hypoesthesia and dysesthesia, respectively. Hypoesthesia was evaluated using the tactile test 1 week after surgery. Dysesthesia was evaluated by interview 4 weeks after surgery. Whole-genome genotyping was conducted using Illumina BeadChips including approximately 300,000 polymorphism markers. Hypoesthesia and dysesthesia occurred in 51 (16.8%) and 149 (49.2%) subjects, respectively. Significant associations were not observed between the clinical data (i.e., age, sex, body weight, body height, loss of blood volume, migration length of bone fragments, nerve exposure, duration of anesthesia, and duration of surgery) and the frequencies of hypoesthesia and dysesthesia. Significant associations were found between hypoesthesia and the rs502281 polymorphism (recessive model: combined  χ 2  = 24.72, nominal  P  = 6.633 × 10 -7 ), between hypoesthesia and the rs2063640 polymorphism (recessive model: combined  χ 2  = 23.07, nominal  P  = 1.563 × 10 -6 ), and between dysesthesia and the nonsynonymous rs2677879 polymorphism (trend model: combined  χ 2  = 16.56, nominal  P  = 4.722 × 10 -5 ; dominant model: combined  χ 2  = 16.31, nominal  P  = 5.369 × 10 -5 ). The rs502281 and rs2063640 polymorphisms were located in the flanking region of the  ARID1B  and  ZPLD1  genes on chromosomes 6 and 3, whose official names are “AT rich interactive domain 1B (SWI1-like)” and “zona pellucida-like domain containing 1”, respectively. The rs2677879 polymorphism is located in the  METTL4  gene on chromosome 18, whose official name is “methyltransferase like 4”. The GWAS of sensory disturbances after BSSRO revealed associations between genetic polymorphisms located in the flanking region of the  ARID1B  and  ZPLD1  genes and hypoesthesia and between a nonsynonymous genetic polymorphism in the  METTL4  gene and dysesthesia.\n",
      "\n",
      "22792071\n",
      "cortical thickness http://www.ebi.ac.uk/efo/EFO_0004840\n",
      "bone density http://www.ebi.ac.uk/efo/EFO_0003923\n",
      "bone density http://www.ebi.ac.uk/efo/EFO_0003923\n",
      "density\n",
      "WNT16 influences bone mineral density, cortical bone thickness, bone strength, and osteoporotic fracture risk.\n",
      "WNT16  Influences Bone Mineral Density, Cortical Bone Thickness, Bone Strength, and Osteoporotic Fracture Risk. We aimed to identify genetic variants associated with cortical bone thickness (CBT) and bone mineral density (BMD) by performing two separate genome-wide association study (GWAS) meta-analyses for CBT in 3 cohorts comprising 5,878 European subjects and for BMD in 5 cohorts comprising 5,672 individuals. We then assessed selected single-nucleotide polymorphisms (SNPs) for osteoporotic fracture in 2,023 cases and 3,740 controls. Association with CBT and forearm BMD was tested for ∼2.5 million SNPs in each cohort separately, and results were meta-analyzed using fixed effect meta-analysis. We identified a missense SNP (Thr>Ile; rs2707466) located in the  WNT16  gene (7q31), associated with CBT (effect size of −0.11 standard deviations [SD] per C allele, P = 6.2×10 −9 ). This SNP, as well as another nonsynonymous SNP rs2908004 (Gly>Arg), also had genome-wide significant association with forearm BMD (−0.14 SD per C allele, P = 2.3×10 −12 , and −0.16 SD per G allele, P = 1.2×10 −15 , respectively). Four genome-wide significant SNPs arising from BMD meta-analysis were tested for association with forearm fracture. SNP rs7776725 in  FAM3C , a gene adjacent to  WNT16 , was associated with a genome-wide significant increased risk of forearm fracture (OR = 1.33, P = 7.3×10 −9 ), with genome-wide suggestive signals from the two missense variants in  WNT16  (rs2908004: OR = 1.22, P = 4.9×10 −6  and rs2707466: OR = 1.22, P = 7.2×10 −6 ). We next generated a homozygous mouse with targeted disruption of  Wnt16 . Female  Wnt16 −/−  mice had 27% (P<0.001) thinner cortical bones at the femur midshaft, and bone strength measures were reduced between 43%–61% (6.5×10 −13 <P<5.9×10 −4 ) at both femur and tibia, compared with their wild-type littermates. Natural variation in humans and targeted disruption in mice demonstrate that  WNT16  is an important determinant of CBT, BMD, bone strength, and risk of fracture. Bone traits are highly dependent on genetic factors. To date, numerous genetic loci for bone mineral density (BMD) and only one locus for osteoporotic fracture have been previously identified to be genome-wide significant. Cortical bone has been reported to be an important determinant of bone strength; so far, no genome-wide association studies (GWAS) have been performed for cortical bone thickness (CBT) of the tibial and radial diaphysis or BMD at forearm, a skeletal site rich in cortical bone. Therefore, we performed two separated meta-analyses of GWAS for cortical thickness of the tibia in 3 independent cohorts of 5,878 men and women, and for forearm BMD in 5 cohorts of 5,672 individuals. We identified the 7q31 locus, which contains  WNT16 , to be associated with CBT and BMD. Four SNPs from this locus were then tested in 2,023 osteoporotic fracture cases and 3,740 controls. One of these SNPs was genome-wide significant, and two were genome-wide suggestive, for forearm fracture. Generating a mouse with targeted disruption of  Wnt16 , we also demonstrated that mice lacking this protein had substantially thinner bone cortices and reduced bone strength than their wild-type littermates. These findings highlight  WNT16  as a clinically relevant member of the Wnt signaling pathway and increase our understanding of the etiology of osteoporosis-related phenotypes and fracture.\n",
      "\n",
      "24096698\n",
      "endometrial neoplasm http://www.ebi.ac.uk/efo/EFO_0004230\n",
      "endometrial neoplasm http://www.ebi.ac.uk/efo/EFO_0004230\n",
      "ovarian endometrioid neoplasm|endometrial neoplasms\n",
      "Genome-wide association study of endometrial cancer in E2C2.\n",
      "Genome-wide association study of endometrial cancer in E2C2. Endometrial cancer (EC), a neoplasm of the uterine epithelial lining, is the most common gynecological malignancy in developed countries and the fourth most common cancer among US women. Women with a family history of EC have an increased risk for the disease, suggesting that inherited genetic factors play a role. We conducted a two-stage genome-wide association study of Type I EC. Stage 1 included 5,472 women (2,695 cases and 2,777 controls) of European ancestry from seven studies. We selected independent single-nucleotide polymorphisms (SNPs) that displayed the most significant associations with EC in Stage 1 for replication among 17,948 women (4,382 cases and 13,566 controls) in a multiethnic population (African America, Asian, Latina, Hawaiian and European ancestry), from nine studies. Although no novel variants reached genome-wide significance, we replicated previously identified associations with genetic markers near the  HNF1B  locus. Our findings suggest that larger studies with specific tumor classification are necessary to identify novel genetic polymorphisms associated with EC susceptibility. The online version of this article (doi:10.1007/s00439-013-1369-1) contains supplementary material, which is available to authorized users.\n",
      "\n",
      "19570815\n",
      "body height http://www.ebi.ac.uk/efo/EFO_0004339\n",
      "anthropometry http://www.ebi.ac.uk/efo/EFO_0004302\n",
      "anthropometry http://www.ebi.ac.uk/efo/EFO_0004302\n",
      "anthropometric traits\n",
      "A genome-wide association study of northwestern Europeans involves the C-type natriuretic peptide signaling pathway in the etiology of human height variation.\n",
      "A genome-wide association study of northwestern Europeans involves the C-type natriuretic peptide signaling pathway in the etiology of human height variation. Northwestern Europeans are among the tallest of human populations. The increase in body height in these people appears to have reached a plateau, suggesting the ubiquitous presence of an optimal environment in which genetic factors may have exerted a particularly strong influence on human growth. Therefore, we performed a genome-wide association study (GWAS) of body height using 2.2 million markers in 10 074 individuals from three Dutch and one German population-based cohorts. Upon genotyping, the 12 most significantly height-associated single nucleotide polymorphisms (SNPs) from this GWAS in 6912 additional individuals of Dutch and Swedish origin, a genetic variant (rs6717918) on chromosome 2q37.1 was found to be associated with height at a genome-wide significance level ( P combined  = 3.4 × 10 −9 ). Notably, a second SNP (rs6718438) located ∼450 bp away and in strong LD ( r 2  = 0.77) with rs6717918 was previously found to be suggestive of a height association in 29 820 individuals of mainly northwestern European ancestry, and the over-expression of a nearby natriuretic peptide precursor type C ( NPPC ) gene, has been associated with overgrowth and skeletal anomalies. We also found a SNP (rs10472828) located on 5p14 near the natriuretic peptide receptor 3 ( NPR3 ) gene, encoding a receptor of the NPPC ligand, to be associated with body height ( P combined  = 2.1 × 10 −7 ). Taken together, these results suggest that variation in the C-type natriuretic peptide signaling pathway, involving the  NPPC  and  NPR3  genes, plays an important role in determining human body height.\n",
      "\n",
      "23386860\n",
      "type ii diabetes mellitus http://www.ebi.ac.uk/efo/EFO_0001360\n",
      "type ii diabetes mellitus http://www.ebi.ac.uk/efo/EFO_0001360\n",
      "diabetes mellitus type 02|non-insulin dependent diabetes mellitus|stable diabetes mellitus|diabetes mellitus\n",
      "Exploring genome-wide - dietary heme iron intake interactions and the risk of type 2 diabetes.\n",
      "Exploring genome-wide – dietary heme iron intake interactions and the risk of type 2 diabetes. Aims/hypothesis:  Genome-wide association studies have identified over 50 new genetic loci for type 2 diabetes (T2D). Several studies conclude that higher dietary heme iron intake increases the risk of T2D. Therefore we assessed whether the relation between genetic loci and T2D is modified by dietary heme iron intake. Methods:  We used Affymetrix Genome-Wide Human 6.0 array data [681,770 single nucleotide polymorphisms (SNPs)] and dietary information collected in the Health Professionals Follow-up Study ( n  = 725 cases;  n  = 1,273 controls) and the Nurses’ Health Study ( n  = 1,081 cases;  n  = 1,692 controls). We assessed whether genome-wide SNPs or iron metabolism SNPs interacted with dietary heme iron intake in relation to T2D, testing for associations in each cohort separately and then meta-analyzing to pool the results. Finally, we created 1,000 synthetic pathways matched to an iron metabolism pathway on number of genes, and number of SNPs in each gene. We compared the iron metabolic pathway SNPs with these synthetic SNP assemblies in their relation to T2D to assess if the pathway as a whole interacts with dietary heme iron intake. Results:  Using a genomic approach, we found no significant gene–environment interactions with dietary heme iron intake in relation to T2D at a Bonferroni corrected genome-wide significance level of 7.33 ×10 - 8  (top SNP in pooled analysis: intergenic rs10980508;  p  = 1.03 × 10 - 6 ). Furthermore, no SNP in the iron metabolic pathway significantly interacted with dietary heme iron intake at a Bonferroni corrected significance level of 2.10 × 10 - 4  (top SNP in pooled analysis: rs1805313;  p  = 1.14 × 10 - 3 ). Finally, neither the main genetic effects (pooled empirical  p  by SNP = 0.41), nor gene – dietary heme–iron interactions (pooled empirical  p -value for the interactions = 0.72) were significant for the iron metabolic pathway as a whole. Conclusions:  We found no significant interactions between dietary heme iron intake and common SNPs in relation to T2D.\n",
      "\n",
      "21909115\n",
      "hypertension http://www.ebi.ac.uk/efo/EFO_0000537\n",
      "systolic blood pressure http://www.ebi.ac.uk/efo/EFO_0006335\n",
      "diastolic blood pressure http://www.ebi.ac.uk/efo/EFO_0006336\n",
      "diastolic blood pressure http://www.ebi.ac.uk/efo/EFO_0006336\n",
      "diastolic pressure|diabp\n",
      "Genetic variants in novel pathways influence blood pressure and cardiovascular disease risk.\n",
      "Genetic Variants in Novel Pathways Influence Blood Pressure and Cardiovascular Disease Risk. Blood pressure (BP) is a heritable trait 1  influenced by multiple biological pathways and is responsive to environmental stimuli. Over one billion people worldwide have hypertension (BP ≥140 mm Hg systolic [SBP] or ≥90 mm Hg diastolic [DBP]) 2 . Even small increments in BP are associated with increased risk of cardiovascular events 3 . This genome-wide association study of SBP and DBP, which used a multi-stage design in 200,000 individuals of European descent, identified 16 novel loci: six of these loci contain genes previously known or suspected to regulate BP ( GUCY1A3-GUCY1B3; NPR3-C5orf23; ADM; FURIN-FES; GOSR2; GNAS-EDN3 ); the other 10 provide new clues to BP physiology. A genetic risk score based on 29 genome-wide significant variants was associated with hypertension, left ventricular wall thickness, stroke, and coronary artery disease, but not kidney disease or kidney function. We also observed associations with BP in East Asian, South Asian, and African ancestry individuals. Our findings provide new insights into the genetics and biology of BP, and suggest novel potential therapeutic pathways for cardiovascular disease prevention.\n",
      "\n",
      "22685421\n",
      "pathological myopia http://www.ebi.ac.uk/efo/EFO_0004207\n",
      "pathological myopia http://www.ebi.ac.uk/efo/EFO_0004207\n",
      "myopia (pathological)|degenerative myopia|myopia\n",
      "Genetic variants on chromosome 1q41 influence ocular axial length and high myopia.\n",
      "Genetic Variants on Chromosome 1q41 Influence Ocular Axial Length and High Myopia. As one of the leading causes of visual impairment and blindness, myopia poses a significant public health burden in Asia. The primary determinant of myopia is an elongated ocular axial length (AL). Here we report a meta-analysis of three genome-wide association studies on AL conducted in 1,860 Chinese adults, 929 Chinese children, and 2,155 Malay adults. We identified a genetic locus on chromosome 1q41 harboring the zinc-finger 11B pseudogene  ZC3H11B  showing genome-wide significant association with AL variation (rs4373767, β = −0.16 mm per minor allele,  P meta  = 2.69×10 −10 ). The minor C allele of rs4373767 was also observed to significantly associate with decreased susceptibility to high myopia (per-allele odds ratio (OR) = 0.75, 95% CI: 0.68–0.84,  P meta  = 4.38×10 −7 ) in 1,118 highly myopic cases and 5,433 controls. ZC3H11B  and two neighboring genes  SLC30A10  and  LYPLAL1  were expressed in the human neural retina, retinal pigment epithelium, and sclera. In an experimental myopia mouse model, we observed significant alterations to gene and protein expression in the retina and sclera of the unilateral induced myopic eyes for the murine genes  ZC3H11A ,  SLC30A10 , and  LYPLAL1 . This supports the likely role of genetic variants at chromosome 1q41 in influencing AL variation and high myopia. Myopic individuals exhibit an increase in ocular axial length (AL). As a highly heritable ocular biometry of refractive error, identification of quantitative trait loci influencing AL variation would be valuable in informing the biological etiology of myopia. We have determined that a genetic locus on chromosome 1q41 containing zinc-finger pseudogene  ZC3H11B  is associated with AL and high myopia through a meta-analysis of three genome-wide association scans on AL in Chinese and Malays, with validation for high myopia association in two additional Japanese cohorts. In addition, variations in the expression of murine gene  ZC3H11A  and two neighboring genes  SLC30A10  and  LYPLAL1  in the retina and sclera in a myopic mouse model implicate the role of these genes in myopia onset. To our knowledge, this is the first genome-wide survey of single nucleotide polymorphism (SNP) variation of AL in Asians. Our results suggest that genetic variants at chromosome 1q41 have potential roles in both common and high myopia.\n",
      "\n",
      "25317111\n",
      "medication adherence behavior http://www.ebi.ac.uk/efo/EFO_0006344\n",
      "hypertension http://www.ebi.ac.uk/efo/EFO_0000537\n",
      "diabetes mellitus http://www.ebi.ac.uk/efo/EFO_0000400\n",
      "diabetes mellitus http://www.ebi.ac.uk/efo/EFO_0000400\n",
      "dm - diabetes mellitus|diabetes mellitus\n",
      "Genome-wide association study of medication adherence in chronic diseases in the korean population.\n",
      "Genome-Wide Association Study of Medication Adherence in Chronic Diseases in the Korean Population. Medication adherence is generally defined as the extent of voluntary cooperation of a patient in taking medicine as prescribed. Adherence to long-term treatment with chronic disease is essential for reducing disease comorbidity and mortality. However, medication non-adherence in chronic disease averages 50%. This study was conducted a genome-wide association study to identify the genetic basis of medication adherence. A total of 235 medication non-adherents and 1,067 medication adherents with hypertension or diabetes were used from the Korean Association Resource project data according to the self-reported treatment status of each chronic disease, respectively. We identified four single nucleotide polymorphisms with suggestive genome-wide association. The most significant single nucleotide polymorphism was rs6978712 (chromosome 7, p = 4.87 × 10 -7 ), which is located proximal to the  GCC1  gene, which was previously implicated in decision-making capability in drug abusers. Two suggestive single nucleotide polymorphisms were in strong linkage disequilibrium (r 2  > 0.8) with rs6978712. Thus, in the aspect of decision-making in adherence behavior, the association between medication adherence and three loci proximal to the  GCC1  gene seems worthy of further research. However, to overcome a few limitations in this study, defining the standardized phenotype criteria for self-reported adherence should be performed before replicating association studies.\n",
      "\n",
      "17903297\n",
      "neuropsychological test http://www.ebi.ac.uk/efo/EFO_0003926\n",
      "volumetric brain mri http://www.ebi.ac.uk/efo/EFO_0004868\n",
      "neuropsychological test http://www.ebi.ac.uk/efo/EFO_0003926\n",
      "test\n",
      "Genetic correlates of brain aging on MRI and cognitive test measures: a genome-wide association and linkage analysis in the Framingham Study.\n",
      "Genetic correlates of brain aging on MRI and cognitive test measures: a genome-wide association and linkage analysis in the Framingham study. Brain magnetic resonance imaging (MRI) and cognitive tests can identify heritable endophenotypes associated with an increased risk of developing stroke, dementia and Alzheimer's disease (AD). We conducted a genome-wide association (GWA) and linkage analysis exploring the genetic basis of these endophenotypes in a community-based sample. A total of 705 stroke- and dementia-free Framingham participants (age 62 +9 yrs, 50% male) who underwent volumetric brain MRI and cognitive testing (1999–2002) were genotyped. We used linear models adjusting for first degree relationships via generalized estimating equations (GEE) and family based association tests (FBAT) in additive models to relate qualifying single nucleotide polymorphisms (SNPs, 70,987 autosomal on Affymetrix 100K Human Gene Chip with minor allele frequency ≥ 0.10, genotypic call rate ≥ 0.80, and Hardy-Weinberg equilibrium p-value ≥ 0.001) to multivariable-adjusted residuals of 9 MRI measures including total cerebral brain (TCBV), lobar, ventricular and white matter hyperintensity (WMH) volumes, and 6 cognitive factors/tests assessing verbal and visuospatial memory, visual scanning and motor speed, reading, abstract reasoning and naming. We determined multipoint identity-by-descent utilizing 10,592 informative SNPs and 613 short tandem repeats and used variance component analyses to compute LOD scores. The strongest gene-phenotype association in FBAT analyses was between  SORL1  (rs1131497; p = 3.2 × 10 -6 ) and abstract reasoning, and in GEE analyses between  CDH4  (rs1970546; p = 3.7 × 10 -8 ) and TCBV. SORL1  plays a role in amyloid precursor protein processing and has been associated with the risk of AD. Among the 50 strongest associations (25 each by GEE and FBAT) were other biologically interesting genes. Polymorphisms within 28 of 163 candidate genes for stroke, AD and memory impairment were associated with the endophenotypes studied at p < 0.001. We confirmed our previously reported linkage of WMH on chromosome 4 and describe linkage of reading performance to a marker on chromosome 18 (GATA11A06), previously linked to dyslexia (LOD scores = 2.2 and 5.1). Our results suggest that genes associated with clinical neurological disease also have detectable effects on subclinical phenotypes. These hypothesis generating data illustrate the use of an unbiased approach to discover novel pathways that may be involved in brain aging, and could be used to replicate observations made in other studies.\n",
      "\n",
      "20675712\n",
      "sensory perception of bitter taste http://purl.obolibrary.org/obo/GO_0050913\n",
      "sensory perception of bitter taste http://purl.obolibrary.org/obo/GO_0050913\n",
      "bitter taste sensitivity|bitter taste perception\n",
      "The perception of quinine taste intensity is associated with common genetic variants in a bitter receptor cluster on chromosome 12.\n",
      "The perception of quinine taste intensity is associated with common genetic variants in a bitter receptor cluster on chromosome 12. The perceived taste intensities of quinine HCl, caffeine, sucrose octaacetate (SOA) and propylthiouracil (PROP) solutions were examined in 1457 twins and their siblings. Previous heritability modeling of these bitter stimuli indicated a common genetic factor for quinine, caffeine and SOA (22–28%), as well as separate specific genetic factors for PROP (72%) and quinine (15%). To identify the genes involved, we performed a genome-wide association study with the same sample as the modeling analysis, genotyped for approximately 610 000 single-nucleotide polymorphisms (SNPs). For caffeine and SOA, no SNP association reached a genome-wide statistical criterion. For PROP, the peak association was within  TAS2R38  ( rs713598 , A49P,  P  = 1.6 × 10 −104 ), which accounted for 45.9% of the trait variance. For quinine, the peak association was centered in a region that contains bitter receptor as well as salivary protein genes and explained 5.8% of the trait variance ( TAS2R19 ,  rs10772420 , R299C,  P  = 1.8 × 10 −15 ). We confirmed this association in a replication sample of twins of similar ancestry ( P  = 0.00001). The specific genetic factor for the perceived intensity of PROP was identified as the gene previously implicated in this trait ( TAS2R38 ). For quinine, one or more bitter receptor or salivary proline-rich protein genes on chromosome 12 have alleles which affect its perception but tight linkage among very similar genes precludes the identification of a single causal genetic variant.\n",
      "\n",
      "22479191\n",
      "cystatin c measurement http://www.ebi.ac.uk/efo/EFO_0004617\n",
      "chronic kidney disease http://www.ebi.ac.uk/efo/EFO_0003884\n",
      "serum creatinine measurement http://www.ebi.ac.uk/efo/EFO_0004518\n",
      "cystatin c measurement http://www.ebi.ac.uk/efo/EFO_0004617\n",
      "\n",
      "Genome-wide association and functional follow-up reveals new loci for kidney function.\n",
      "Genome-Wide Association and Functional Follow-Up Reveals New Loci for Kidney Function. Chronic kidney disease (CKD) is an important public health problem with a genetic component. We performed genome-wide association studies in up to 130,600 European ancestry participants overall, and stratified for key CKD risk factors. We uncovered 6 new loci in association with estimated glomerular filtration rate (eGFR), the primary clinical measure of CKD, in or near  MPPED2 ,  DDX1 ,  SLC47A1 ,  CDK12 ,  CASP9 , and  INO80 . Morpholino knockdown of  mpped2  and  casp9  in zebrafish embryos revealed podocyte and tubular abnormalities with altered dextran clearance, suggesting a role for these genes in renal function. By providing new insights into genes that regulate renal function, these results could further our understanding of the pathogenesis of CKD. Chronic kidney disease (CKD) is an important public health problem with a hereditary component. We performed a new genome-wide association study in up to 130,600 European ancestry individuals to identify genes that may influence kidney function, specifically genes that may influence kidney function differently depending on sex, age, hypertension, and diabetes status of individuals. We uncovered 6 new loci associated with estimated glomerular filtration rate (eGFR), the primary measure of renal function, in or near  MPPED2 ,  DDX1 ,  SLC47A1 ,  CDK12 ,  CASP9 , and  INO80 . CDK12  effect was stronger in younger and absent in older individuals. MPPED2 ,  DDX1 ,  SLC47A1 , and  CDK12  loci were associated with eGFR in African ancestry samples as well, highlighting the cross-ethnicity validity of our findings. Using the zebrafish model, we performed morpholino knockdown of  mpped2  and  casp9  in zebrafish embryos and revealed podocyte and tubular abnormalities with altered dextran clearance, suggesting a role for these genes in renal function. These results further our understanding of the pathogenesis of CKD and provide insights into potential novel mechanisms of disease.\n",
      "\n",
      "20038947\n",
      "unipolar depression http://www.ebi.ac.uk/efo/EFO_0003761\n",
      "unipolar depression http://www.ebi.ac.uk/efo/EFO_0003761\n",
      "melancholia\n",
      "Novel loci for major depression identified by genome-wide association study of Sequenced Treatment Alternatives to Relieve Depression and meta-analysis of three studies.\n",
      "Novel loci for major depression identified by genome-wide association study of STAR*D and meta-analysis of three studies. We report a genome-wide association study (GWAS) of major depressive disorder (MDD) in 1,221 cases from the Sequenced Treatment Alternatives to Relieve Depression (STAR*D) study and 1,636 screened controls. No genome-wide evidence for association was detected. We also carried out a meta-analysis of three European-ancestry MDD GWAS datasets: STAR*D, Genetics of Recurrent Early-Onset Depression (GenRED) and the publicly-available Genetic Association Information Network MDD dataset (GAIN-MDD). These datasets, totaling 3,957 cases and 3,428 controls, were genotyped using four different platforms (Affymetrix 6.0, 5.0 and 500K, and Perlegen). For each of 2.4 million HapMap II SNPs, using genotyped data where available and imputed data otherwise, single-SNP association tests were carried out in each sample with correction for ancestry-informative principal components. The strongest evidence for association in the meta-analysis was observed for intronic SNPs in  ATP6V1B2  (P = 6.78 × 10 −7 ),  SP4  ( P  = 7.68 × 10 −7 ) and  GRM7  ( P  = 1.11 × 10 −6 ). Additional exploratory analyses were carried out for a narrower phenotype (recurrent MDD with onset before age 31, N = 2,191 cases), and separately for males and females. Several of the best findings were supported primarily by evidence from narrow cases or from either males or females. Based on previous biological evidence, we consider  GRM7  a strong MDD candidate gene. Larger samples will be required to determine whether any common SNPs are significantly associated with MDD.\n",
      "\n",
      "24564958\n",
      "social communication impairment http://www.ebi.ac.uk/efo/EFO_0005427\n",
      "social communication impairment http://www.ebi.ac.uk/efo/EFO_0005427\n",
      "\n",
      "Variability in the common genetic architecture of social-communication spectrum phenotypes during childhood and adolescence.\n",
      "Variability in the common genetic architecture of social-communication spectrum phenotypes during childhood and adolescence. Social-communication abilities are heritable traits, and their impairments overlap with the autism continuum. To characterise the genetic architecture of social-communication difficulties developmentally and identify genetic links with the autistic dimension, we conducted a genome-wide screen of social-communication problems at multiple time-points during childhood and adolescence. Social-communication difficulties were ascertained at ages 8, 11, 14 and 17 years in a UK population-based birth cohort (Avon Longitudinal Study of Parents and Children;  N  ≤ 5,628) using mother-reported Social Communication Disorder Checklist scores. Genome-wide Complex Trait Analysis (GCTA) was conducted for all phenotypes. The time-points with the highest GCTA heritability were subsequently analysed for single SNP association genome-wide. Type I error in the presence of measurement relatedness and the likelihood of observing SNP signals near known autism susceptibility loci (co-location) were assessed via large-scale, genome-wide permutations. Association signals ( P  ≤ 10 −5 ) were also followed up in Autism Genetic Resource Exchange pedigrees ( N  = 793) and the Autism Case Control cohort ( N cases / N controls  = 1,204/6,491). GCTA heritability was strongest in childhood ( h 2 (8 years)  = 0.24) and especially in later adolescence ( h 2 (17 years)  = 0.45), with a marked drop during early to middle adolescence ( h 2 (11 years)  = 0.16 and h 2 (14 years)  = 0.08). Genome-wide screens at ages 8 and 17 years identified for the latter time-point evidence for association at 3p22.2 near  SCN11A  (rs4453791,  P  = 9.3 × 10 −9 ; genome-wide empirical  P  = 0.011) and suggestive evidence at 20p12.3 at  PLCB1  (rs3761168,  P  = 7.9 × 10 −8 ; genome-wide empirical  P  = 0.085). None of these signals contributed to risk for autism. However, the co-location of population-based signals and autism susceptibility loci harbouring rare mutations, such as  PLCB1 , is unlikely to be due to chance (genome-wide empirical  P co-location  = 0.007). Our findings suggest that measurable common genetic effects for social-communication difficulties vary developmentally and that these changes may affect detectable overlaps with the autism spectrum.\n",
      "\n",
      "25085501\n",
      "seasonal allergic rhinitis http://www.ebi.ac.uk/efo/EFO_0003956\n",
      "asthma http://www.ebi.ac.uk/efo/EFO_0000270\n",
      "allergic rhinitis http://www.ebi.ac.uk/efo/EFO_0005854\n",
      "allergic rhinitis http://www.ebi.ac.uk/efo/EFO_0005854\n",
      "\n",
      "Integrated Genome-wide Association, Coexpression Network, and Expression Single Nucleotide Polymorphism Analysis Identifies Novel Pathway in Allergic Rhinitis.\n",
      "Integrated genome-wide association, coexpression network, and expression single nucleotide polymorphism analysis identifies novel pathway in allergic rhinitis. Allergic rhinitis is a common disease whose genetic basis is incompletely explained. We report an integrated genomic analysis of allergic rhinitis. We performed genome wide association studies (GWAS) of allergic rhinitis in 5633 ethnically diverse North American subjects. Next, we profiled gene expression in disease-relevant tissue (peripheral blood CD4+ lymphocytes) collected from subjects who had been genotyped. We then integrated the GWAS and gene expression data using expression single nucleotide (eSNP), coexpression network, and pathway approaches to identify the biologic relevance of our GWAS. GWAS revealed ethnicity-specific findings, with 4 genome-wide significant loci among Latinos and 1 genome-wide significant locus in the GWAS meta-analysis across ethnic groups. To identify biologic context for these results, we constructed a coexpression network to define modules of genes with similar patterns of CD4+ gene expression (coexpression modules) that could serve as constructs of broader gene expression. 6 of the 22 GWAS loci with P-value ≤ 1x10 −6  tagged one particular coexpression module (4.0-fold enrichment, P-value 0.0029), and this module also had the greatest enrichment (3.4-fold enrichment, P-value 2.6 × 10 −24 ) for allergic rhinitis-associated eSNPs (genetic variants associated with both gene expression and allergic rhinitis). The integrated GWAS, coexpression network, and eSNP results therefore supported this coexpression module as an allergic rhinitis module. Pathway analysis revealed that the module was enriched for mitochondrial pathways (8.6-fold enrichment, P-value 4.5 × 10 −72 ). Our results highlight mitochondrial pathways as a target for further investigation of allergic rhinitis mechanism and treatment. Our integrated approach can be applied to provide biologic context for GWAS of other diseases.\n",
      "\n",
      "21483430\n",
      "neuroimaging measurement http://www.ebi.ac.uk/efo/EFO_0004346\n",
      "south texas assessment of neurocognition http://www.ebi.ac.uk/efo/EFO_0004785\n",
      "south texas assessment of neurocognition http://www.ebi.ac.uk/efo/EFO_0004785\n",
      "\n",
      "Impact of DISC1 variation on neuroanatomical and neurocognitive phenotypes.\n",
      "Impact of  DISC1  variation on neuroanatomical and neurocognitive phenotypes. Although  DISC1  has been implicated in many psychiatric disorders, including schizophrenia, bipolar disorder, schizoaffective disorder and major depression, its biological role in these disorders is unclear. To better understand this gene and its role in psychiatric disease, we conducted transcriptional profiling and genome-wide association analysis in 1 232 pedigreed Mexican American individuals for whom we have neuroanatomic images, neurocognitive assessments and neuropsychiatric diagnoses. SOLAR was used to determine heritability, identify gene expression patterns and perform association analyses on 188 quantitative brain-related phenotypes. We found that the  DISC1  transcript is highly heritable (h 2 =0.50; p=1.97 × 10 −22 ), and that gene expression is strongly  cis -regulated ( cis -LOD=3.89) but is also influenced by  trans -effects. We identified several  DISC1  polymorphisms that were associated with cortical gray-matter thickness within the parietal, temporal and frontal lobes. Associated regions affiliated with memory included the entorhinal cortex (rs821639, p=4.11 × 10 −5 ; rs2356606, p=4.71 × 10 −4 ), cingulate cortex (rs16856322, p=2.88 × 10 −4 ) and parahippocampal gyrus (rs821639, p=4.95 × 10 −4 ); those affiliated with executive and other cognitive processing included the transverse temporal gyrus (rs9661837, p=5.21 × 10 −4 ; rs17773946, p=6.23 × 10 −4 ), anterior cingulate cortex (rs2487453, p=; 4.79 × 10 −4 ; rs3738401, p= 5.43 × 10 −4 ) and medial orbitofrontal cortex (rs9661837; p=7.40 × 10 −4 ). Cognitive measures of working memory (rs2793094, p=3.38 × 10 −4 ), as well as lifetime history of depression (rs4658966, p=4.33 × 10 −4 ; rs12137417, p=4.93 × 10 −4 ) and panic (rs12137417, p=7.41 × 10 −4 ) were associated with  DISC1  sequence variation. DISC1  has well-defined genetic regulation and clearly influences important phenotypes related to psychiatric disease.\n",
      "\n",
      "21076409\n",
      "qrs duration http://www.ebi.ac.uk/efo/EFO_0005055\n",
      "heart function measurement http://www.ebi.ac.uk/efo/EFO_0004311\n",
      "heart function measurement http://www.ebi.ac.uk/efo/EFO_0004311\n",
      "heart function tests|cardiac function tests\n",
      "Common variants in 22 loci are associated with QRS duration and cardiac ventricular conduction.\n",
      "Common variants in 22 loci are associated with QRS duration and cardiac ventricular conduction. QRS interval on the electrocardiogram reflects ventricular depolarization and conduction time, and is a risk factor for mortality, sudden death, and heart failure. We performed a genome-wide association meta-analysis in 40,407 European-descent individuals from 14 studies, with further genotyping in 7170 additional Europeans, and identified 22 loci associated with QRS duration ( P  < 5 × 10 −8 ). These loci map in or near genes in pathways with established roles in ventricular conduction such as sodium channels, transcription factors, and calcium-handling proteins, but also point to novel biologic processes, such as kinase inhibitors and genes related to tumorigenesis. We demonstrate that  SCN10A , a gene at our most significant locus, is expressed in the mouse ventricular conduction system, and treatment with a selective  SCN10A  blocker prolongs QRS duration. These findings extend our current knowledge of ventricular depolarization and conduction.\n",
      "\n",
      "21829393\n",
      "autoantibody measurement http://www.ebi.ac.uk/efo/EFO_0004866\n",
      "type i diabetes mellitus http://www.ebi.ac.uk/efo/EFO_0001359\n",
      "type i diabetes mellitus http://www.ebi.ac.uk/efo/EFO_0001359\n",
      "iddm|diabetes mellitus\n",
      "Genome-wide association analysis of autoantibody positivity in type 1 diabetes cases.\n",
      "Genome-Wide Association Analysis of Autoantibody Positivity in Type 1 Diabetes Cases. The genetic basis of autoantibody production is largely unknown outside of associations located in the major histocompatibility complex (MHC) human leukocyte antigen (HLA) region. The aim of this study is the discovery of new genetic associations with autoantibody positivity using genome-wide association scan single nucleotide polymorphism (SNP) data in type 1 diabetes (T1D) patients with autoantibody measurements. We measured two anti-islet autoantibodies, glutamate decarboxylase (GADA, n = 2,506), insulinoma-associated antigen 2 (IA-2A, n = 2,498), antibodies to the autoimmune thyroid (Graves') disease (AITD) autoantigen thyroid peroxidase (TPOA, n = 8,300), and antibodies against gastric parietal cells (PCA, n = 4,328) that are associated with autoimmune gastritis. Two loci passed a stringent genome-wide significance level ( p <10 −10 ): 1q23/ FCRL3  with IA-2A and 9q34/ ABO  with PCA. Eleven of 52 non-MHC T1D loci showed evidence of association with at least one autoantibody at a false discovery rate of 16%: 16p11/ IL27 -IA-2A, 2q24/ IFIH1 -IA-2A and PCA, 2q32/ STAT4 -TPOA, 10p15/ IL2RA -GADA, 6q15/ BACH2 -TPOA, 21q22/ UBASH3A -TPOA, 1p13/ PTPN22 -TPOA, 2q33/ CTLA4 -TPOA, 4q27/ IL2 /TPOA, 15q14/ RASGRP1 /TPOA, and 12q24/ SH2B3 -GADA and TPOA. Analysis of the TPOA-associated loci in 2,477 cases with Graves' disease identified two new AITD loci ( BACH2  and  UBASH3A ). Autoantibodies are important markers for autoimmune diseases such as type 1 diabetes and Graves' disease. However, little is known about the genetic factors that control their production. To improve our understanding of this genetic basis, we measured four autoantibodies in a collection of up to 8,300 type 1 diabetes cases plasma samples. We combined these measurements with genome-wide genotype data to conduct four independent genome-wide association studies. Two loci showed unequivocal evidence of autoantibody association: the  FCRL3  locus and the  ABO  blood group locus. Variants in the  FCRL3  gene have been previously associated with autoimmune diseases, but such associations have not been reported for  ABO  blood group genotypes. In addition, we found extensive overlap between type 1 diabetes and autoantibody loci, and these findings provide new information about the role of these risk variants. Lastly, we hypothesized that loci associated with thyroid autoantibodies are strong candidates for association with thyroid autoimmune disorders. We confirmed this hypothesis by genotyping these variants in an independent cohort of Graves' disease cases, and we found evidence for two new Graves' disease loci.\n",
      "\n",
      "24204295\n",
      "narcolepsy with cataplexy http://www.ebi.ac.uk/efo/EFO_0000614\n",
      "age at onset http://www.ebi.ac.uk/efo/EFO_0004847\n",
      "narcolepsy with cataplexy http://www.ebi.ac.uk/efo/EFO_0000614\n",
      "gelineau's syndrome|narcolepsy nos|paroxysmal sleep|gelineaus syndrome|syndrome\n",
      "Genome wide analysis of narcolepsy in China implicates novel immune loci and reveals changes in association prior to versus after the 2009 H1N1 influenza pandemic.\n",
      "Genome Wide Analysis of Narcolepsy in China Implicates Novel Immune Loci and Reveals Changes in Association Prior to Versus After the 2009 H1N1 Influenza Pandemic. Previous studies in narcolepsy, an autoimmune disorder affecting hypocretin (orexin) neurons and recently associated with H1N1 influenza, have demonstrated significant associations with five loci. Using a well-characterized Chinese cohort, we refined known associations in TRA@ and P2RY11-DNMT1 and identified new associations in the TCR beta (TRB@; rs9648789 max P = 3.7×10 −9  OR 0.77), ZNF365 (rs10995245 max P = 1.2×10 −11  OR 1.23), and IL10RB-IFNAR1 loci (rs2252931 max P = 2.2×10 −9  OR 0.75). Variants in the Human Leukocyte Antigen (HLA)- DQ region were associated with age of onset (rs7744020 P = 7.9×10 −9  beta −1.9 years) and varied significantly among cases with onset after the 2009 H1N1 influenza pandemic compared to previous years (rs9271117 P = 7.8×10 −10  OR 0.57). These reflected an association of DQB1*03:01 with earlier onset and decreased DQB1*06:02 homozygosity following 2009. Our results illustrate how genetic association can change in the presence of new environmental challenges and suggest that the monitoring of genetic architecture over time may help reveal the appearance of novel triggers for autoimmune diseases. Narcolepsy-hypocretin deficiency results from a highly specific autoimmune attack on hypocretin cells. Recent studies have established antigen presentation by specific class II proteins encoded by (HLA DQB1*06:02 and DQA1*01:02) to the cognate T cell receptor as the main disease pathway, with a role for H1N1 influenza in the triggering process. Here, we have used a large and well-characterized cohort of Chinese narcolepsy cases to examine genetic architecture not observed in European samples. We confirmed previously implicated susceptibility genes (T cell receptor alpha, P2RY11), and identify new loci (ZNF365, IL10RB-IFNAR1), most notably, variants at the beta chain of the T cell receptor. We found that one HLA variant, (DQB1*03:01), is associated with dramatically earlier disease onset (nearly 2 years). We also identified differences in HLA haplotype frequencies among cases with onset following the 2009 H1N1 influenza pandemic as compared to before the outbreak, with fewer HLA DQB1*06:02 homozygotes. This may be the first demonstration of such an effect, and suggests that the study of changes in GWAS signals over time could help identify environmental factors in other autoimmune diseases.\n",
      "\n",
      "21383967\n",
      "immune system disease http://www.ebi.ac.uk/efo/EFO_0000540\n",
      "immune system disease http://www.ebi.ac.uk/efo/EFO_0000540\n",
      "[x]disorder involving the immune mechanism\n",
      "Meta-analysis of genome-wide association studies in celiac disease and rheumatoid arthritis identifies fourteen non-HLA shared loci.\n",
      "Meta-Analysis of Genome-Wide Association Studies in Celiac Disease and Rheumatoid Arthritis Identifies Fourteen Non-HLA Shared Loci. Epidemiology and candidate gene studies indicate a shared genetic basis for celiac disease (CD) and rheumatoid arthritis (RA), but the extent of this sharing has not been systematically explored. Previous studies demonstrate that 6 of the established non-HLA CD and RA risk loci (out of 26 loci for each disease) are shared between both diseases. We hypothesized that there are additional shared risk alleles and that combining genome-wide association study (GWAS) data from each disease would increase power to identify these shared risk alleles. We performed a meta-analysis of two published GWAS on CD (4,533 cases and 10,750 controls) and RA (5,539 cases and 17,231 controls). After genotyping the top associated SNPs in 2,169 CD cases and 2,255 controls, and 2,845 RA cases and 4,944 controls, 8 additional SNPs demonstrated  P <5×10 −8  in a combined analysis of all 50,266 samples, including four SNPs that have not been previously confirmed in either disease: rs10892279 near the  DDX6  gene ( P combined  = 1.2×10 −12 ), rs864537 near  CD247  ( P combined  = 2.2×10 −11 ), rs2298428 near  UBE2L3  ( P combined  = 2.5×10 −10 ), and rs11203203 near  UBASH3A  ( P combined  = 1.1×10 −8 ). We also confirmed that 4 gene loci previously established in either CD or RA are associated with the other autoimmune disease at combined  P <5×10 −8  ( SH2B3 ,  8q24 ,  STAT4 , and  TRAF1-C5 ). From the 14 shared gene loci, 7 SNPs showed a genome-wide significant effect on expression of one or more transcripts in the linkage disequilibrium (LD) block around the SNP. These associations implicate antigen presentation and T-cell activation as a shared mechanism of disease pathogenesis and underscore the utility of cross-disease meta-analysis for identification of genetic risk factors with pleiotropic effects between two clinically distinct diseases. Celiac disease (CD) and rheumatoid arthritis (RA) are two autoimmune diseases characterized by distinct clinical features but increased co-occurrence in families and individuals. Genome-wide association studies (GWAS) performed in CD and RA have identified the HLA region and 26 non-HLA genetic risk loci in each disease. Of the 26 CD and 26 RA risk loci, previous studies have shown that six are shared between the two diseases. In this study we aimed to identify additional shared risk alleles and, in doing so, gain more insight into shared disease pathogenesis. We first empirically investigated the distribution of putative risk alleles from GWAS across both diseases (after removing known risk loci for both diseases). We found that CD risk alleles are non-randomly distributed in the RA GWAS (and vice versa), indicating that CD risk alleles have an increased prior probability of being associated with RA (and vice versa). Next, we performed a GWAS meta-analysis to search for shared risk alleles by combing the RA and CD GWAS, performing both directional and opposite allelic effect analyses, followed by replication testing in independent case-control datasets in both diseases. In addition to the already established six non-HLA shared risk loci, we observed statistically robust associations at eight SNPs, thereby increasing the number of shared non-HLA risk loci to fourteen. Finally, we used gene expression studies and pathway analysis tools to identify the plausible candidate genes in the fourteen associated loci. We observed remarkable overrepresentation of T-cell signaling molecules among the shared genes.\n",
      "\n",
      "21998595\n",
      "body height http://www.ebi.ac.uk/efo/EFO_0004339\n",
      "anthropometry http://www.ebi.ac.uk/efo/EFO_0004302\n",
      "anthropometry http://www.ebi.ac.uk/efo/EFO_0004302\n",
      "anthropometric traits\n",
      "Identification, replication, and fine-mapping of Loci associated with adult height in individuals of african ancestry.\n",
      "Identification, Replication, and Fine-Mapping of Loci Associated with Adult Height in Individuals of African Ancestry. Adult height is a classic polygenic trait of high heritability ( h 2  ∼0.8). More than 180 single nucleotide polymorphisms (SNPs), identified mostly in populations of European descent, are associated with height. These variants convey modest effects and explain ∼10% of the variance in height. Discovery efforts in other populations, while limited, have revealed loci for height not previously implicated in individuals of European ancestry. Here, we performed a meta-analysis of genome-wide association (GWA) results for adult height in 20,427 individuals of African ancestry with replication in up to 16,436 African Americans. We found two novel height loci (Xp22-rs12393627,  P  = 3.4×10 −12  and 2p14-rs4315565,  P  = 1.2×10 −8 ). As a group, height associations discovered in European-ancestry samples replicate in individuals of African ancestry ( P  = 1.7×10 −4  for overall replication). Fine-mapping of the European height loci in African-ancestry individuals showed an enrichment of SNPs that are associated with expression of nearby genes when compared to the index European height SNPs ( P <0.01). Our results highlight the utility of genetic studies in non-European populations to understand the etiology of complex human diseases and traits. Adult height is an ideal phenotype to improve our understanding of the genetic architecture of complex diseases and traits: it is easily measured and usually available in large cohorts, relatively stable, and mostly influenced by genetics (narrow-sense heritability of height  h 2 ∼0.8). Genome-wide association (GWA) studies in individuals of European ancestry have identified >180 single nucleotide polymorphisms (SNPs) associated with height. In the current study, we continued to use height as a model polygenic trait and explored the genetic influence in populations of African ancestry through a meta-analysis of GWA height results from 20,809 individuals of African descent. We identified two novel height loci not previously found in Europeans. We also replicated the European height signals, suggesting that many of the genetic variants that are associated with height are shared between individuals of European and African descent. Finally, in fine-mapping the European height loci in African-ancestry individuals, we found SNPs more likely to be associated with the expression of nearby genes than the SNPs originally found in Europeans. Thus, our results support the utility of performing genetic studies in non-European populations to gain insights into complex human diseases and traits.\n",
      "\n",
      "25742292\n",
      "female fertility http://www.ebi.ac.uk/efo/EFO_0006918\n",
      "fertility measurement http://www.ebi.ac.uk/efo/EFO_0006923\n",
      "female fertility http://www.ebi.ac.uk/efo/EFO_0006918\n",
      "\n",
      "Genome-wide association study of parity in bangladeshi women.\n",
      "Genome-Wide Association Study of Parity in Bangladeshi Women. Human fertility is a complex trait determined by gene-environment interactions in which genetic factors represent a significant component. To better understand inter-individual variability in fertility, we performed one of the first genome-wide association studies (GWAS) of common fertility phenotypes, lifetime number of pregnancies and number of children in a developing country population. The fertility phenotype data and DNA samples were obtained at baseline recruitment from individuals participating in a large prospective cohort study in Bangladesh. GWAS analyses of fertility phenotypes were conducted among 1,686 married women. One SNP on chromosome 4 was non-significantly associated with number of children at P <10 -7  and number of pregnancies at P <10 -6 . This SNP is located in a region without a gene within 1 Mb. One SNP on chromosome 6 was non-significantly associated with extreme number of children at P <10 -6 . The closest gene to this SNP is  HDGFL1 , a hepatoma-derived growth factor. When we excluded hormonal contraceptive users, a SNP on chromosome 5 was non-significantly associated at P <10 -5  for number of children and number of pregnancies. This SNP is located near  C5orf64 , an open reading frame, and  ZSWIM6 , a zinc ion binding gene. We also estimated the heritability of these phenotypes from our genotype data using GCTA (Genome-wide Complex Trait Analysis) for number of children (h g 2  = 0.149, SE = 0.24, p-value = 0.265) and number of pregnancies (h g 2  = 0.007, SE = 0.22, p-value = 0.487). Our genome-wide association study and heritability estimates of number of pregnancies and number of children in Bangladesh did not confer strong evidence of common variants for parity variation. However, our results suggest that future studies may want to consider the role of 3 notable SNPs in their analysis.\n",
      "\n",
      "20174558\n",
      "type ii diabetes mellitus http://www.ebi.ac.uk/efo/EFO_0001360\n",
      "type ii diabetes mellitus http://www.ebi.ac.uk/efo/EFO_0001360\n",
      "diabetes mellitus type 02|non-insulin dependent diabetes mellitus|stable diabetes mellitus|diabetes mellitus\n",
      "A genome-wide association study identifies susceptibility variants for type 2 diabetes in Han Chinese.\n",
      "A Genome-Wide Association Study Identifies Susceptibility Variants for Type 2 Diabetes in Han Chinese. To investigate the underlying mechanisms of T2D pathogenesis, we looked for diabetes susceptibility genes that increase the risk of type 2 diabetes (T2D) in a Han Chinese population. A two-stage genome-wide association (GWA) study was conducted, in which 995 patients and 894 controls were genotyped using the Illumina HumanHap550-Duo BeadChip for the first genome scan stage. This was further replicated in 1,803 patients and 1,473 controls in stage 2. We found two loci not previously associated with diabetes susceptibility in and around the genes protein tyrosine phosphatase receptor type D ( PTPRD ) ( P  = 8.54×10 −10 ; odds ratio [OR] = 1.57; 95% confidence interval [CI] = 1.36–1.82), and serine racemase ( SRR ) ( P  = 3.06×10 −9 ; OR = 1.28; 95% CI = 1.18–1.39). We also confirmed that variants in  KCNQ1  were associated with T2D risk, with the strongest signal at rs2237895 ( P  = 9.65×10 −10 ; OR = 1.29, 95% CI = 1.19–1.40). By identifying two novel genetic susceptibility loci in a Han Chinese population and confirming the involvement of  KCNQ1 , which was previously reported to be associated with T2D in Japanese and European descent populations, our results may lead to a better understanding of differences in the molecular pathogenesis of T2D among various populations. Type 2 diabetes (T2D) is a complex disease that involves many genes and environmental factors. Genome-wide and candidate-gene association studies have thus far identified at least 19 regions containing genes that may confer a risk for T2D. However, most of these studies were conducted with patients of European descent. We studied Chinese patients with T2D and identified two genes, PTPRD and SRR, that were not previously known to be involved in diabetes and are involved in biological pathways different from those implicated in T2D by previous association reports. PTPRD is a protein tyrosine phosphatase and may affect insulin signaling on its target cells. SRR encodes a serine racemase that synthesizes D-serine from L-serine. Both D-serine (coagonist) and the neurotransmitter glutamate bind to NMDA receptors and trigger excitatory neurotransmission in the brain. Glutamate signaling also regulates insulin and glucagon secretion in pancreatic islets. Thus, SRR and D-serine, in addition to regulating insulin and glucagon secretion, may play a role in the etiology of T2D. Our study suggests that, in different patient populations, different genes may confer risks for diabetes. Our findings may lead to a better understanding of the molecular pathogenesis of T2D.\n",
      "\n",
      "20169177\n",
      "rheumatoid arthritis http://www.ebi.ac.uk/efo/EFO_0000685\n",
      "systemic lupus erythematosus http://www.ebi.ac.uk/efo/EFO_0002690\n",
      "rheumatoid arthritis http://www.ebi.ac.uk/efo/EFO_0000685\n",
      "proliferative arthritis|rheumatoid arthritis (disorder)|arthritis\n",
      "Genome-wide association study in Asian populations identifies variants in ETS1 and WDFY4 associated with systemic lupus erythematosus.\n",
      "Genome-Wide Association Study in Asian Populations Identifies Variants in  ETS1  and  WDFY4  Associated with Systemic Lupus Erythematosus. Systemic lupus erythematosus is a complex and potentially fatal autoimmune disease, characterized by autoantibody production and multi-organ damage. By a genome-wide association study (320 patients and 1,500 controls) and subsequent replication altogether involving a total of 3,300 Asian SLE patients from Hong Kong, Mainland China, and Thailand, as well as 4,200 ethnically and geographically matched controls, genetic variants in  ETS1  and  WDFY4  were found to be associated with SLE ( ETS1 : rs1128334,  P  = 2.33×10 −11 , OR = 1.29;  WDFY4 : rs7097397,  P  = 8.15×10 −12 , OR = 1.30). ETS1  encodes for a transcription factor known to be involved in a wide range of immune functions, including Th17 cell development and terminal differentiation of B lymphocytes. SNP rs1128334 is located in the 3′-UTR of  ETS1 , and allelic expression analysis from peripheral blood mononuclear cells showed significantly lower expression level from the risk allele. WDFY4 is a conserved protein with unknown function, but is predominantly expressed in primary and secondary immune tissues, and rs7097397 in  WDFY4  changes an arginine residue to glutamine (R1816Q) in this protein. Our study also confirmed association of the  HLA  locus,  STAT4 ,  TNFSF4 ,  BLK ,  BANK1 ,  IRF5 , and  TNFAIP3  with SLE in Asians. These new genetic findings may help us to gain a better understanding of the disease and the functions of the genes involved. In this study, we first conducted a genome-wide association study in a Hong Kong Chinese population, followed by replication in three other cohorts from Mainland China and a cohort from Thailand, which totaled 3,300 Asian patients and 4,200 ethnically and geographically matched controls. We identified novel variants in  ETS1  and  WDFY4  associated with SLE with genome-wide significance and confirmed the association of  HLA  locus,  STAT4 ,  BLK ,  IRF5 ,  BANK1 ,  TNFSF , and  IRF5  with the disease. ETS1  encodes a critical transcription factor involved in Th17 and B cell development. Allelic expression study showed a significantly lower expression of  ETS1  from the risk allele, which provided functional support to the genetic findings. WDFY4 is a huge protein with unknown function but is predominantly expressed in primary and secondary immune tissues, and a nonsynonymous SNP in this gene was found to be highly associated with SLE susceptibility. Our findings shed new light on the function of these genes as well as the mechanism of this devastating disease.\n",
      "\n",
      "22911880\n",
      "mental or behavioural disorder http://www.ebi.ac.uk/efo/EFO_0000677\n",
      "eating disorder http://www.ebi.ac.uk/efo/EFO_0005203\n",
      "mental or behavioural disorder http://www.ebi.ac.uk/efo/EFO_0000677\n",
      "disorders\n",
      "Genome-wide association analysis of eating disorder-related symptoms, behaviors, and personality traits.\n",
      "Genome-Wide Association Analysis of Eating Disorder-Related Symptoms, Behaviors, and Personality Traits. Eating disorders (EDs) are common, complex psychiatric disorders thought to be caused by both genetic and environmental factors. They share many symptoms, behaviors, and personality traits, which may have overlapping heritability. The aim of the present study is to perform a genome-wide association scan (GWAS) of six ED phenotypes comprising three symptom traits from the Eating Disorders Inventory 2 [Drive for Thinness (DT), Body Dissatisfaction (BD), and Bulimia], Weight Fluctuation symptom, Breakfast Skipping behavior and Childhood Obsessive-Compulsive Personality Disorder trait (CHIRP). Investigated traits were derived from standardized self-report questionnaires completed by the TwinsUK population-based cohort. We tested 283,744 directly typed SNPs across six phenotypes of interest in the TwinsUK discovery dataset and followed-up signals from various strata using a two-stage replication strategy in two independent cohorts of European ancestry. We meta-analyzed a total of 2,698 individuals for DT, 2,680 for BD, 2,789 (821 cases/1,968 controls) for Bulimia, 1,360 (633 cases/727 controls) for Childhood Obsessive-Compulsive Personality Disorder trait, 2,773 (761 cases/2,012 controls) for Breakfast Skipping, and 2,967 (798 cases/2,169 controls) for Weight Fluctuation symptom. In this GWAS analysis of six ED-related phenotypes, we detected association of eight genetic variants with  P  < 10 −5 . Genetic variants that showed suggestive evidence of association were previously associated with several psychiatric disorders and ED-related phenotypes. Our study indicates that larger-scale collaborative studies will be needed to achieve the necessary power to detect loci underlying ED-related traits. © 2012 Wiley Periodicals, Inc.\n",
      "\n",
      "25226531\n",
      "infant expressive language ability http://www.ebi.ac.uk/efo/EFO_0006316\n",
      "infant expressive language ability http://www.ebi.ac.uk/efo/EFO_0006316\n",
      "\n",
      "Common variation near ROBO2 is associated with expressive vocabulary in infancy.\n",
      "Common variation near  ROBO2  is associated with expressive vocabulary in infancy. Twin studies suggest that expressive vocabulary at ~24 months is modestly heritable. However, the genes influencing this early linguistic phenotype are unknown. Here we conduct a genome-wide screen and follow-up study of expressive vocabulary in toddlers of European descent from up to four studies of the EArly Genetics and Lifecourse Epidemiology consortium, analysing an early (15–18 months, ‘one-word stage’,  N Total =8,889) and a later (24–30 months, ‘two-word stage’,  N Total =10,819) phase of language acquisition. For the early phase, one single-nucleotide polymorphism (rs7642482) at 3p12.3 near  ROBO2 , encoding a conserved axon-binding receptor, reaches the genome-wide significance level ( P =1.3 × 10 −8 ) in the combined sample. This association links language-related common genetic variation in the general population to a potential autism susceptibility locus and a linkage region for dyslexia, speech-sound disorder and reading. The contribution of common genetic influences is, although modest, supported by genome-wide complex trait analysis (meta-GCTA  h 2 15–18-months =0.13, meta-GCTA  h 2 24–30-months =0.14) and in concordance with additional twin analysis (5,733 pairs of European descent,  h 2 24-months =0.20). The genetic basis of expressive vocabulary in children around 2 years old is poorly understood. Here, the authors show that a genetic variant near the  ROBO2  gene is associated with early language acquisition in the general population and highlight a potential genetic link between language-related common genetic variation and a linkage region for dyslexia, speech-sound disorder and reading.\n",
      "\n",
      "20861866\n",
      "glaucoma http://www.ebi.ac.uk/efo/EFO_0000516\n",
      "age-related macular degeneration http://www.ebi.ac.uk/efo/EFO_0001365\n",
      "glaucoma http://www.ebi.ac.uk/efo/EFO_0000516\n",
      "glaucoma associated with other ocular disorders|other specified glaucoma nos|glaucoma nec|glaucoma nos|other specified glaucoma nos (disorder)|glaucoma (disorder)|glaucoma associated with other ocular disorders (disorder)|glaucoma\n",
      "Genome-wide association identifies SKIV2L and MYRIP as protective factors for age-related macular degeneration.\n",
      "Genome-wide association identifies SKIV2L and MYRIP as protective factors for age-related macular degeneration. Age-related macular degeneration (AMD) is the leading cause of blindness in the elderly in the developed world. We conducted a genome-wide association study in a series of families enriched for AMD and completed a meta-analysis of this new data with results from reanalysis of an existing study of a late-stage case/control cohort. We tested the top findings for replication in 1 896 cases and 1 866 controls and identified two novel genetic protective factors for AMD. In addition to the  CFH  (p=2.3×10 −64 ) and  ARMS2  (p=1.2×10 −60 ) loci, we observed a protective effect at rs429608, an intronic SNP in  SKIV2L  (p=5.3×10 −15 ), a gene near the  C2/BF  locus, that indicates the protective effect may be mediated by variants other than the C2/BF variants previously studied. Haplotype analysis at this locus identified three protective haplotypes defined by the rs429608 protective allele. We also identified a new potentially protective effect at rs2679798 in  MYRIP  (p=2.9×10 −4 ), a gene involved in retinal pigment epithelium melanosome trafficking. Interestingly,  MYRIP  was initially identified in the family-based scan and was confirmed in the case-control set. From these efforts, we report the identification of two novel protective factors for AMD and confirm the previously known associations at  CFH ,  ARMS2  and  C3 .\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from extractor.util import gold_phen_recall\n",
    "kb = KnowledgeBase()\n",
    "id2doc = {doc.name : doc for doc in corpus.documents}\n",
    "gold_dict_phen = { doc_id : set() for doc_id, phen_id in gold_set_phens }\n",
    "for doc_id, phen_id in gold_set_phens:\n",
    "    gold_dict_phen[doc_id].add(phen_id)\n",
    "\n",
    "phen_not_found = list(gold_phen_recall(phen_c, gold_set_phens, phen2id))\n",
    "print len(phen_not_found)\n",
    "for doc_id, phen_id in phen_not_found[:100]:\n",
    "    if len(gold_dict_phen[doc_id]) > 3: continue\n",
    "    print doc_id\n",
    "    for phen_id2 in gold_dict_phen[doc_id]:\n",
    "        print id2phen[phen_id2].name, phen_id2\n",
    "#     print gold_dict_phen[doc_id]\n",
    "    print id2phen[phen_id].name, id2phen[phen_id].ontology_ref\n",
    "    print id2phen[phen_id].synonyms    \n",
    "    print kb.title_by_pmid(doc_id)\n",
    "    print ' '.join(s.text for s in id2doc[doc_id].sentences)\n",
    "    print"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Over aggregate phenotypes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We say that a mention for an aggregate phenotype is correct, if it corresponds to the name of the GWC phenotype or to the phenotype of any equivalent EFO phenotype."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 1282 gold mentions, e.g.:\n",
      "[('23251661', u'http://www.ebi.ac.uk/efo/EFO_0004749'), ('19587794', u'http://www.ebi.ac.uk/efo/EFO_0004682'), ('24376456', u'http://www.ebi.ac.uk/efo/EFO_0006340'), ('20081858', u'http://www.ebi.ac.uk/efo/EFO_0004501'), ('23776548', u'http://www.ebi.ac.uk/efo/EFO_0004731')]\n",
      "1063 535\n",
      "Statistics over EFO phenotypes:\n",
      "# of gold annotations\t= 9747\n",
      "# of candidates\t\t= 27861\n",
      "# of correct candidates\t= 9017\n",
      "Candidate recall\t= 0.925\n",
      "Candidate precision\t= 0.324\n"
     ]
    }
   ],
   "source": [
    "from db.kb import KnowledgeBase\n",
    "from nltk.stem import PorterStemmer\n",
    "from extractor.util import change_name\n",
    "\n",
    "kb = KnowledgeBase() # reload\n",
    "gold_set_agg_phens = frozenset \\\n",
    "([ \n",
    "    (doc.name, phen.id) for doc in corpus.documents \n",
    "                        for phen in kb.phen_by_pmid(doc.name, source='gwas_catalog')\n",
    "])\n",
    "\n",
    "# map phenotype names to their id (EFO syn -> GWC id)\n",
    "agg_phen2id = dict()\n",
    "for doc in corpus.documents:\n",
    "    for phen in kb.phen_by_pmid(doc.name, source='gwas_catalog'):\n",
    "        for eq_phen in phen.equivalents:\n",
    "            for syn in [phen.name] + [eq_phen.name] + eq_phen.synonyms.split('|'):\n",
    "                syn_name = change_name(syn)\n",
    "                if syn_name not in agg_phen2id: agg_phen2id[syn_name] = set()\n",
    "                agg_phen2id[syn_name].add(phen.id)\n",
    "\n",
    "# map ids to phenotypes (GWC id -> GWC phen obj)                \n",
    "agg_id2phen = \\\n",
    "{\n",
    "    phen.id : phen for doc in corpus.documents\n",
    "                   for phen in kb.phen_by_pmid(doc.name, source='gwas_catalog')\n",
    "}\n",
    "\n",
    "print 'Found %d gold mentions, e.g.:' % len(gold_set_phens)\n",
    "print list(gold_set_phens)[:5]\n",
    "print len(phen2id), len(id2phen)\n",
    "\n",
    "from extractor.util import gold_phen_stats, gold_agg_phen_stats\n",
    "gold_agg_phen_stats(phen_c, gold_set_agg_phens, agg_phen2id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Why is the recall low?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1003\n",
      "22383897\n",
      "breast cancer 25571\n",
      "\tbreast carcinoma http://www.ebi.ac.uk/efo/EFO_0000305 breast neoplasms|neoplasm\n",
      "breast cancer 25572\n",
      "\tbreast carcinoma http://www.ebi.ac.uk/efo/EFO_0000305 breast neoplasms|neoplasm\n",
      "breast cancer 25573\n",
      "\tbreast carcinoma http://www.ebi.ac.uk/efo/EFO_0000305 breast neoplasms|neoplasm\n",
      "breast cancer\n",
      "Genome-Wide Association Study in East Asians Identifies Novel Susceptibility Loci for Breast Cancer. Genetic factors play an important role in the etiology of both sporadic and familial breast cancer. We aimed to discover novel genetic susceptibility loci for breast cancer. We conducted a four-stage genome-wide association study (GWAS) in 19,091 cases and 20,606 controls of East-Asian descent including Chinese, Korean, and Japanese women. After analyzing 690,947 SNPs in 2,918 cases and 2,324 controls, we evaluated 5,365 SNPs for replication in 3,972 cases and 3,852 controls. Ninety-four SNPs were further evaluated in 5,203 cases and 5,138 controls, and finally the top 22 SNPs were investigated in up to 17,423 additional subjects (7,489 cases and 9,934 controls). SNP rs9485372, near the TGF-β activated kinase ( TAB2 ) gene in chromosome 6q25.1, showed a consistent association with breast cancer risk across all four stages, with a  P -value of 3.8×10 −12  in the combined analysis of all samples. Adjusted odds ratios (95% confidence intervals) were 0.89 (0.85–0.94) and 0.80 (0.75–0.86) for the A/G and A/A genotypes, respectively, compared with the genotype G/G. SNP rs9383951 ( P  = 1.9×10 −6  from the combined analysis of all samples), located in intron 5 of the  ESR1  gene, and SNP rs7107217 ( P  = 4.6×10 −7 ), located at 11q24.3, also showed a consistent association in each of the four stages. This study provides strong evidence for a novel breast cancer susceptibility locus represented by rs9485372, near the  TAB2  gene (6q25.1), and identifies two possible susceptibility loci located in the  ESR1  gene and 11q24.3, respectively. Breast cancer is one of the most common malignancies among women worldwide. Genetic factors play an important role in the etiology of breast cancer. To identify common genetic susceptibility alleles for breast cancer, we performed a four-stage genome-wide association study in 19,091 cases and 20,606 controls among East-Asian women. Single nucleotide polymorphism (SNP) rs9485372, near the TGF-beta activated kinase 1 ( TAB2 ) gene at chromosome 6q25.1, was associated with breast cancer risk ( P  = 3.8×10 −12 ). SNPs rs9383951, located in intron 5 of the estrogen receptor 1 ( ESR1 ) gene, and rs7107217, located at 11q24.3, were also consistently associated with breast cancer risk in all four stages with a combined  P  of 1.9×10 −6  and 4.6×10 −7 , respectively. This study provides strong evidence for a novel breast cancer susceptibility locus represented by rs9485372, near the  TAB2  gene (6q25.1), and identifies two possible susceptibility loci located in the  ESR1  gene and 11q24.3, respectively.\n",
      "\n",
      "19578366\n",
      "glioma (high-grade) 20856\n",
      "\tcentral nervous system cancer http://www.ebi.ac.uk/efo/EFO_0000326 cns cancer|malignant central nervous system neoplasm|malignant tumor of cns|cancer of cns|malignant tumor of central nervous system|malignant cns neoplasms|malignant neoplasm of cns|malignant tumor of the central nervous system|malignant cns tumor|malignant central nervous system tumor|malignant neoplasm of the cns|malignant neoplasm of the central nervous system|malignant cns neoplasm|central nervous system neoplasms\n",
      "glioma (high-grade) 20857\n",
      "\tcentral nervous system cancer http://www.ebi.ac.uk/efo/EFO_0000326 cns cancer|malignant central nervous system neoplasm|malignant tumor of cns|cancer of cns|malignant tumor of central nervous system|malignant cns neoplasms|malignant neoplasm of cns|malignant tumor of the central nervous system|malignant cns tumor|malignant central nervous system tumor|malignant neoplasm of the cns|malignant neoplasm of the central nervous system|malignant cns neoplasm|central nervous system neoplasms\n",
      "glioma (high-grade) 20855\n",
      "\tcentral nervous system cancer http://www.ebi.ac.uk/efo/EFO_0000326 cns cancer|malignant central nervous system neoplasm|malignant tumor of cns|cancer of cns|malignant tumor of central nervous system|malignant cns neoplasms|malignant neoplasm of cns|malignant tumor of the central nervous system|malignant cns tumor|malignant central nervous system tumor|malignant neoplasm of the cns|malignant neoplasm of the central nervous system|malignant cns neoplasm|central nervous system neoplasms\n",
      "glioma (high-grade)\n",
      "Variants in the  CDKN2B  and  RTEL1  regions are associated with high grade glioma susceptibility. The causes of glioblastoma and other devastating gliomas remain obscure  1 , 2 . To discover new candidate genes that might influence glioma susceptibility, we conducted a principal component  3  adjusted genome wide association study (GWAS) of 275,895 autosomal variants among 692 adult high grade glioma patients (622 from the San Francisco Adult Glioma Study (AGS) and 70 from the Cancer Genome Atlas (TCGA)  4  versus 3992 controls (602 from AGS and 3390 from Illumina iControlDB (iControls)). For replication, we then analyzed the 13 SNPs with p<10 −6  using completely independent data from 176 high grade glioma patients versus 174 controls from the Mayo Clinic. Rs1412829 in chromosome 9p21 (near  CDKN2B ) had discovery p-value 3.4*10 −8 , replication p-value 0.0038, and combined p-value 1.85*10 −10 . Rs6010620 intronic to  RTEL1  had discovery p-value 1.5*10 −7 , replication p-value 0.00035, and combined p-value 3.40*10 −9 . For both SNPs, the direction of association was the same in discovery and replication phases.\n",
      "\n",
      "19412176\n",
      "narcolepsy 20551\n",
      "\tnarcolepsy with cataplexy http://www.ebi.ac.uk/efo/EFO_0000614 gelineau's syndrome|narcolepsy nos|paroxysmal sleep|gelineaus syndrome|syndrome\n",
      "narcolepsy\n",
      "Narcolepsy is strongly associated with the TCR alpha locus. Narcolepsy-cataplexy, characterized by sleepiness and rapid onset into REM sleep, affects 1 in 2,000 individuals 1 , 2 . Narcolepsy was first shown to be tightly associated with HLA-DR2 3 , and later sublocalized to DQB1*0602 4 . Following studies in dogs 5  and mice 6 , a 95% loss of hypocretin-producing cells in human postmortem hypothalami was reported 7 , 8 , Using Genome Wide Association (GWA) in Caucasians with replication in three ethnic groups, we found association with polymorphisms in the T-Cell receptor alpha ( TCRA ) locus, with highest significance at rs1154155 (average allelic odds ratio 1.69, genotype odds ratios 1.94 and 2.55, p<10 −21 , 1830 cases, 2164 controls). This is the first documented genetic involvement of the  TCRA  locus, the major receptor for HLA-peptide presentation, in any disease. It is still unclear how specific HLA alleles confer susceptibility to over 100 HLA-associated disorders 9 , thus narcolepsy will provide new insights on how HLA-TCR interactions contribute to organ specific autoimmune targeting.\n",
      "\n",
      "23001122\n",
      "breast cancer (male) 21123\n",
      "\tbreast carcinoma http://www.ebi.ac.uk/efo/EFO_0000305 breast neoplasms|neoplasm\n",
      "breast cancer (male) 21124\n",
      "\tbreast carcinoma http://www.ebi.ac.uk/efo/EFO_0000305 breast neoplasms|neoplasm\n",
      "breast cancer (male) 21125\n",
      "\tbreast carcinoma http://www.ebi.ac.uk/efo/EFO_0000305 breast neoplasms|neoplasm\n",
      "breast cancer (male)\n",
      "Genome-wide association study identifies a novel variant in  RAD51B  associated with male breast cancer risk. We conducted a genome-wide association study of male breast cancer using 823 cases and 2,795 controls of European ancestry with validation in independent sample sets totalling 438 cases and 474 controls. A novel variant in  RAD51B  (14q24.1) was significantly associated with male breast cancer risk ( P  = 3.02 ×10 −13 , odds ratio (OR) = 1.57). TOX3  (16q12.1) was also a susceptibility locus ( P  = 3.87 ×10 −15 , OR = 1.50).\n",
      "\n",
      "23183491\n",
      "opioid sensitivity 21902\n",
      "\tresponse to opiate http://purl.obolibrary.org/obo/GO_0014072 response to isoquinoline alkaloid\n",
      "opioid sensitivity\n",
      "Genome-wide association study identifies a potent locus associated with human opioid sensitivity. Opioids, such as morphine and fentanyl, are widely used as effective analgesics for the treatment of acute and chronic pain. In addition, the opioid system has a key role in the rewarding effects of morphine, ethanol, cocaine and various other drugs. Although opioid sensitivity is well known to vary widely among individual subjects, several candidate genetic polymorphisms reported so far are not sufficient for fully understanding the wide range of interindividual differences in human opioid sensitivity. By conducting a multistage genome-wide association study (GWAS) in healthy subjects, we found that genetic polymorphisms within a linkage disequilibrium block that spans 2q33.3–2q34 were strongly associated with the requirements for postoperative opioid analgesics after painful cosmetic surgery. The C allele of the best candidate single-nucleotide polymorphism (SNP), rs2952768, was associated with more analgesic requirements, and consistent results were obtained in patients who underwent abdominal surgery. In addition, carriers of the C allele in this SNP exhibited less vulnerability to severe drug dependence in patients with methamphetamine dependence, alcohol dependence, and eating disorders and a lower ‘Reward Dependence' score on a personality questionnaire in healthy subjects. Furthermore, the C/C genotype of this SNP was significantly associated with the elevated expression of a neighboring gene,  CREB1 . These results show that SNPs in this locus are the most potent genetic factors associated with human opioid sensitivity known to date, affecting both the efficacy of opioid analgesics and liability to severe substance dependence. Our findings provide valuable information for the personalized treatment of pain and drug dependence.\n",
      "\n",
      "22383897\n",
      "breast cancer 25571\n",
      "\tbreast carcinoma http://www.ebi.ac.uk/efo/EFO_0000305 breast neoplasms|neoplasm\n",
      "breast cancer 25572\n",
      "\tbreast carcinoma http://www.ebi.ac.uk/efo/EFO_0000305 breast neoplasms|neoplasm\n",
      "breast cancer 25573\n",
      "\tbreast carcinoma http://www.ebi.ac.uk/efo/EFO_0000305 breast neoplasms|neoplasm\n",
      "breast cancer\n",
      "Genome-Wide Association Study in East Asians Identifies Novel Susceptibility Loci for Breast Cancer. Genetic factors play an important role in the etiology of both sporadic and familial breast cancer. We aimed to discover novel genetic susceptibility loci for breast cancer. We conducted a four-stage genome-wide association study (GWAS) in 19,091 cases and 20,606 controls of East-Asian descent including Chinese, Korean, and Japanese women. After analyzing 690,947 SNPs in 2,918 cases and 2,324 controls, we evaluated 5,365 SNPs for replication in 3,972 cases and 3,852 controls. Ninety-four SNPs were further evaluated in 5,203 cases and 5,138 controls, and finally the top 22 SNPs were investigated in up to 17,423 additional subjects (7,489 cases and 9,934 controls). SNP rs9485372, near the TGF-β activated kinase ( TAB2 ) gene in chromosome 6q25.1, showed a consistent association with breast cancer risk across all four stages, with a  P -value of 3.8×10 −12  in the combined analysis of all samples. Adjusted odds ratios (95% confidence intervals) were 0.89 (0.85–0.94) and 0.80 (0.75–0.86) for the A/G and A/A genotypes, respectively, compared with the genotype G/G. SNP rs9383951 ( P  = 1.9×10 −6  from the combined analysis of all samples), located in intron 5 of the  ESR1  gene, and SNP rs7107217 ( P  = 4.6×10 −7 ), located at 11q24.3, also showed a consistent association in each of the four stages. This study provides strong evidence for a novel breast cancer susceptibility locus represented by rs9485372, near the  TAB2  gene (6q25.1), and identifies two possible susceptibility loci located in the  ESR1  gene and 11q24.3, respectively. Breast cancer is one of the most common malignancies among women worldwide. Genetic factors play an important role in the etiology of breast cancer. To identify common genetic susceptibility alleles for breast cancer, we performed a four-stage genome-wide association study in 19,091 cases and 20,606 controls among East-Asian women. Single nucleotide polymorphism (SNP) rs9485372, near the TGF-beta activated kinase 1 ( TAB2 ) gene at chromosome 6q25.1, was associated with breast cancer risk ( P  = 3.8×10 −12 ). SNPs rs9383951, located in intron 5 of the estrogen receptor 1 ( ESR1 ) gene, and rs7107217, located at 11q24.3, were also consistently associated with breast cancer risk in all four stages with a combined  P  of 1.9×10 −6  and 4.6×10 −7 , respectively. This study provides strong evidence for a novel breast cancer susceptibility locus represented by rs9485372, near the  TAB2  gene (6q25.1), and identifies two possible susceptibility loci located in the  ESR1  gene and 11q24.3, respectively.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from extractor.util import gold_agg_phen_recall\n",
    "kb = KnowledgeBase()\n",
    "id2doc = {doc.name : doc for doc in corpus.documents}\n",
    "gold_agg_dict_phen = { doc_id : set() for doc_id, phen_id in gold_set_agg_phens }\n",
    "for doc_id, phen_id in gold_set_agg_phens:\n",
    "    gold_agg_dict_phen[doc_id].add(phen_id)\n",
    "\n",
    "# this contains (doc_id, gwc_id) pairs from gold_set_agg_phens that haven't been found\n",
    "phen_not_found = list(gold_agg_phen_recall(phen_c, gold_set_agg_phens, agg_phen2id))\n",
    "print len(phen_not_found)\n",
    "for doc_id, phen_id in phen_not_found[:100]:\n",
    "    if len(gold_agg_dict_phen[doc_id]) > 3: continue # skip if >3 gwc_phen_ids in doc\n",
    "    print doc_id\n",
    "    for phen_id2 in gold_agg_dict_phen[doc_id]: # iterate over gwc_id's in doc\n",
    "        print agg_id2phen[phen_id2].name, phen_id2 # print its name and id\n",
    "        for eq_phen in agg_id2phen[phen_id2].equivalents: # loop over equivalent phens\n",
    "            print '\\t', eq_phen.name, eq_phen.ontology_ref, eq_phen.synonyms # print name, ref, synonyms\n",
    "    print agg_id2phen[phen_id].name # print the one that we haven't found\n",
    "    print ' '.join(s.text for s in id2doc[doc_id].sentences)\n",
    "    print"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[u'Genome-wide', u'association', u'study', u'identifies', u'multiple', u'susceptibility', u'loci', u'for', u'pulmonary', u'fibrosis', u'.']\n"
     ]
    }
   ],
   "source": [
    "print id2doc[doc_id].sentences[0].words"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sentence(Document('23583980', Corpus (GWAS Corpus)), 0, u'Genome-wide association study identifies multiple susceptibility loci for pulmonary fibrosis.')\n",
      "Genome-wide association study identifies multiple susceptibility loci\n",
      "Genome-wide association study identifies multiple susceptibility loci genom wide associ studi identifi multipl suscept loci\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "association study identifies multiple susceptibility loci for\n",
      "association study identifies multiple susceptibility loci for associ studi identifi multipl suscept loci for\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "study identifies multiple susceptibility loci for pulmonary\n",
      "study identifies multiple susceptibility loci for pulmonary studi identifi multipl suscept loci for pulmonari\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "identifies multiple susceptibility loci for pulmonary fibrosis\n",
      "identifies multiple susceptibility loci for pulmonary fibrosis identifi multipl suscept loci for pulmonari fibrosi\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "multiple susceptibility loci for pulmonary fibrosis.\n",
      "multiple susceptibility loci for pulmonary fibrosis. multipl suscept loci for pulmonari fibrosi\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "Genome-wide association study identifies multiple susceptibility\n",
      "Genome-wide association study identifies multiple susceptibility genom wide associ studi identifi multipl suscept\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "association study identifies multiple susceptibility loci\n",
      "association study identifies multiple susceptibility loci associ studi identifi multipl suscept loci\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "study identifies multiple susceptibility loci for\n",
      "study identifies multiple susceptibility loci for studi identifi multipl suscept loci for\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "identifies multiple susceptibility loci for pulmonary\n",
      "identifies multiple susceptibility loci for pulmonary identifi multipl suscept loci for pulmonari\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "multiple susceptibility loci for pulmonary fibrosis\n",
      "multiple susceptibility loci for pulmonary fibrosis multipl suscept loci for pulmonari fibrosi\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "susceptibility loci for pulmonary fibrosis.\n",
      "susceptibility loci for pulmonary fibrosis. suscept loci for pulmonari fibrosi\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "Genome-wide association study identifies multiple\n",
      "Genome-wide association study identifies multiple genom wide associ studi identifi multipl\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "association study identifies multiple susceptibility\n",
      "association study identifies multiple susceptibility associ studi identifi multipl suscept\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "study identifies multiple susceptibility loci\n",
      "study identifies multiple susceptibility loci studi identifi multipl suscept loci\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "identifies multiple susceptibility loci for\n",
      "identifies multiple susceptibility loci for identifi multipl suscept loci for\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "multiple susceptibility loci for pulmonary\n",
      "multiple susceptibility loci for pulmonary multipl suscept loci for pulmonari\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "susceptibility loci for pulmonary fibrosis\n",
      "susceptibility loci for pulmonary fibrosis suscept loci for pulmonari fibrosi\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "loci for pulmonary fibrosis.\n",
      "loci for pulmonary fibrosis. loci for pulmonari fibrosi\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "Genome-wide association study identifies\n",
      "Genome-wide association study identifies genom wide associ studi identifi\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "association study identifies multiple\n",
      "association study identifies multiple associ studi identifi multipl\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "study identifies multiple susceptibility\n",
      "study identifies multiple susceptibility studi identifi multipl suscept\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "identifies multiple susceptibility loci\n",
      "identifies multiple susceptibility loci identifi multipl suscept loci\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "multiple susceptibility loci for\n",
      "multiple susceptibility loci for multipl suscept loci for\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "susceptibility loci for pulmonary\n",
      "susceptibility loci for pulmonary suscept loci for pulmonari\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "loci for pulmonary fibrosis\n",
      "loci for pulmonary fibrosis loci for pulmonari fibrosi\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "for pulmonary fibrosis.\n",
      "for pulmonary fibrosis. for pulmonari fibrosi\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "Genome-wide association study\n",
      "Genome-wide association study genom wide associ studi\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "association study identifies\n",
      "association study identifies associ studi identifi\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "study identifies multiple\n",
      "study identifies multiple studi identifi multipl\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "identifies multiple susceptibility\n",
      "identifies multiple susceptibility identifi multipl suscept\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "multiple susceptibility loci\n",
      "multiple susceptibility loci multipl suscept loci\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "susceptibility loci for\n",
      "susceptibility loci for suscept loci for\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "loci for pulmonary\n",
      "loci for pulmonary loci for pulmonari\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "for pulmonary fibrosis\n",
      "for pulmonary fibrosis for pulmonari fibrosi\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "pulmonary fibrosis.\n",
      "pulmonary fibrosis. pulmonari fibrosi\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "Genome-wide association\n",
      "Genome-wide association genom wide associ\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "association study\n",
      "association study associ studi\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "study identifies\n",
      "study identifies studi identifi\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "identifies multiple\n",
      "identifies multiple identifi multipl\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "multiple susceptibility\n",
      "multiple susceptibility multipl suscept\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "susceptibility loci\n",
      "susceptibility loci suscept loci\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "loci for\n",
      "loci for loci for\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "for pulmonary\n",
      "for pulmonary for pulmonari\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "pulmonary fibrosis\n",
      "pulmonary fibrosis pulmonari fibrosi\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "fibrosis.\n",
      "fibrosis. fibrosi\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "Genome-wide\n",
      "Genome-wide genom wide\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "association\n",
      "association associ\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "study\n",
      "study studi\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "identifies\n",
      "identifies identifi\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "multiple\n",
      "multiple multipl\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "susceptibility\n",
      "susceptibility suscept\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "loci\n",
      "loci loci\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "for\n",
      "for for\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "pulmonary\n",
      "pulmonary pulmonari\n",
      "... False False False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      "fibrosis\n",
      "fibrosis fibrosi\n",
      "... True True False False False\n",
      "None\n",
      "23583980 None\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n",
      ".\n",
      ". \n",
      "... False False True False True\n",
      "http://www.ebi.ac.uk/efo/EFO_0006918\n",
      "23583980 http://www.ebi.ac.uk/efo/EFO_0006918\n",
      "set([u'http://www.ebi.ac.uk/efo/EFO_0004244'])\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# FOR DEBUGGING WHY SPANS ARENT MATCHED\n",
    "from extractor.util import change_name\n",
    "\n",
    "doc_id = '23583980'\n",
    "ngrams = Ngrams(n_max=7)\n",
    "print id2doc[doc_id].sentences[0]\n",
    "for span in ngrams.apply(id2doc[doc_id].sentences[0]):\n",
    "    print span.get_span()\n",
    "    if phen_matcher._f(span):    \n",
    "        phen_name = span.get_span()\n",
    "        print phen_name, change_name(phen_name)\n",
    "        print '...', phen_name == 'fibrosis', phen_name in phenotype_list, change_name(phen_name) in phenotype_list, phen_name in efo_phen_matcher.d, change_name(phen_name) in efo_phen_matcher.d\n",
    "        phen_id = phen2id.get(change_name(phen_name), None)\n",
    "        print phen_id\n",
    "        if not phen_id or phen_id not in gold_dict_phen[span.context.document.name]:\n",
    "            print span.context.document.name, phen_id\n",
    "            print gold_dict_phen[span.context.document.name]\n",
    "        \n",
    "        print"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[u'cd8_alpha-negative plasmactyoid dendritic cell', u'alpha-hydro-omega-hydroxypoly(oxyethylene)', u'ajmalan-17alpha', u'(-)-(s)-alpha-ethyl-2-oxo-1-pyrrolidineacetamide', u'alpha-l-fucosidase deficiency', u'branched chain alpha-ketoacid dehydrogenase complex deficiency', u'tumor necrosis factor-alpha', u'primary alpha-dystroglycanopathy', u'alpha-thalassemia - x-linked intellectual disability syndrome', u' alpha-beta intraepithelial t cell', u'alpha-n-acetylgalactosaminidase deficiency', u'(3alpha', u'cd8-alpha-beta-positive', u'l-alpha-acetamido-beta-mercaptopropionic acid', u'fc-epsilon rialpha-high basophil progenitor cell', u'cd8_alpha-positive cd11b-negative dendritic cell', u'alpha globulin measurement', u'hyperphenylalaninemia due to pterin-4-alpha-carbinolamine dehydratase deficiency', u'alpha-tocopherol metabolic process', u'(22s)-5alpha-campestane-3beta', u'(s)-(+)-alpha-amino-4-carboxy-2-methylbenzeneacetic', u't-b+ scid due to il-7ralpha deficiency', u'cd8alpha-negative thymic conventional dendritic cell', u'alpha-b crystallin-related late-onset distal myopathy', u'mature alpha-beta t cell', u'hemorrhagic disease due to alpha-1-antitrypsin pittsburgh mutation', u'alpha mating type (yeast)', u'alpha-phthalimidoglutarimide', u'dl-alpha-tocopherol', u'(s)-alpha-methyl-4-carboxyphenylglycine', u'interleukin-2 receptor alpha chain deficiency', u'3-(alpha-phenyl-beta-acetylethyl)-4-hydroxycoumarin', u'immature cd8_alpha-positive cd11b-negative dendritic cell', u'3-carboxamido-4-hydroxy-alpha-((1-methyl-3-phenylpropylamino)methyl)benzyl alcohol', u'pyruvate dehydrogenase e1-alpha deficiency', u'familial hypoalphalipoproteinemia', u'qualitative or quantitative defects of alpha-actin', u'alpha-methyl-p-tyrosine', u'17alpha-ethynylestradiol', u'mating_type_alpha', u'congenital adrenal hyperplasia due to 17-alpha-hydroxylase deficiency', u'hereditary persistence of alpha-fetoprotein', u'hereditary renal amyloidosis due to fibrinogen a alpha-chain variant', u'3-(alpha-acetonyl-p-nitrobenzyl)-4-hydroxycoumarin', u'analphalipoproteinemia', u'(s)-alpha-methyl-4-(2-methylpropyl)benzeneacetic acid', u'alpha-naphthaleneacetic acid', u'(s)-6-methoxy-alpha-methyl-2-naphthaleneacetic acid', u'7alpha-{9-[(4', u'alpha-adrenergic receptor blockaders', u'(alphae)-2-[[6-(2-cyanophenoxy)-4-pyrimidinyl]oxy]-alpha-(methoxymethylene) benzeneacetic acid methyl ester', u'(5alpha', u'3-(alpha-acetonyl-4-nitrobenzyl)-4-hydroxycoumarin', u'alpha granule disease', u'alpha cell', u'7 alpha-hydroxy-4-cholesten-3-one 12-alpha-hydroxylase', u' cd8_alpha-negative', u'tumor necrosis factor-alpha measurement', u'alpha-adrenergic agonist', u'streptomyces griseus subsp. alpha', u'qualitative or quantitative defects of integrin alpha-7', u'immature alpha-beta t cell', u'hereditary amyloid nephropathy due to fibrinogen a alpha-chain variant', u'n-(alpha', u'alpha-thalassemia major', u'mating type alpha', u'17alpha-methyl-delta(4)-androsten-17beta-ol-3-one', u'n-[(1s)-1-(dihydroxyboranyl)-3-methylbutyl]-n(alpha)-(pyrazin-2-ylcarbonyl)-d-phenylalaninamide', u'1-(o-chloro-alpha', u'alpha-1', u'17alpha-estradiol', u'cd8_alpha-positive plasmactyoid dendritic cell', u'n-acetyl-alpha-glucosaminidase deficiency', u' alpha-beta cytokine secreting effector t cell', u'alpha-butylene', u'3-(alpha-p-nitrophenyl-beta-acetylethyl)-4-hydroxycoumarin', u'[2r-(2alpha', u'gamma-diisopropylamino-alpha-phenyl-alpha-(2-pyridyl)butyramide', u'tcr-alpha-beta+ t-cell deficiency', u'immature cd8_alpha-negative cd11b-positive dendritic cell', u'l-alpha-methyltyrosine', u'17alpha-ethynylestra-1', u'qualitative or quantitative defects of protein involved in o-glycosylation of alpha-dystroglycan', u'secondary alpha-dystroglycanopathy', u'interferon alpha', u'alphavirus infections', u'(rs)-4-hydroxy-alpha(1)-({[6-(4-phenylbutoxy)hexyl]amino}methyl)-1', u'4-alpha-hydroxyphenylpyruvate hydroxylase deficiency', u'alpha dense granule deficiency', u'alpha peak frequency measurement', u'l-histidylglycyl-l-alpha-glutamylglycyl-l-threonyl-l-phenylalanyl-l-threonyl-l-seryl-l-alpha-aspartyl-l-leucyl-l-seryl-l-lysyl-l-glutaminyl-l-methionyl-l-alpha-glutamyl-l-alpha-glutamyl-l-alpha-glutamyl-l-alanyl-l-valyl-l-arginyl-l-leucyl-l-phenylalanyl-l-isoleucyl-l-alpha-glutamyl-l-tryptophyl-l-leucyl-l-lysyl-l-asparaginylglycylglycyl-l-prolyl-l-seryl-l-serylglycyl-l-alanyl-l-prolyl-l-prolyl-l-prolyl-l-serinamide', u'qualitative or quantitative defects of alpha-sarcoglycan', u'lysosomal alpha-d-mannosidase deficiency', u'heparan-alpha-glucosaminide n-acetyltransferase deficiency', u'alpha-1-antitrypsin deficiency (disorder)', u'alpha-thalassemia - myelodysplastic syndrome', u'alpha-methyl-acetoacetyl-coa thiolase deficiency', u'mature cd8_alpha-negative cd11b-negative dendritic cell', u'cd8alpha-positive thymic conventional dendritic cell', u'platelet alpha-granule deficiency', u'alpha-ketopropionaldehyde', u'alpha-methyldopa', u'familial amyloid nephropathy due to fibrinogen a alpha-chain variant', u'(s)-(-)-alpha-ethyl-2-oxo-1-pyrrolidineacetamide', u'hypoalphalipoproteinemia', u'17(alpha)-methyl-delta(4)-androsten-17(beta)-ol-3-one', u'3-(alpha-(p-nitrophenol)-beta-acetylethyl)-4-hydroxycoumarin', u'alpha-adrenergic antagonists', u'mature cd8_alpha-negative cd11b-positive dendritic cell', u'alpha-methyltyrosine', u'(24r)-5alpha-ergostan-3beta-ol', u'4-androstene-17alpha-methyl-17beta-ol-3-one', u'alpha-methyl-beta-(3', u' alpha-beta cytotoxic t cell', u'disorder of peroxisomal alpha-', u'congenital disorder of glycosylation due to steroid 5alpha-reductase type 3 deficiency', u'(2s-(2alpha', u'oxysterol 7-alpha-hydroxylase deficiency', u'immature cd8_alpha-negative cd11b-negative dendritic cell', u'serum alpha-1-antitrypsin measurement', u\"3-(alpha-(4'-nitrophenyl)-beta-acetylethyl)-4-hydroxycoumarin\", u'alpha lipoprotein cholesterol', u'alpha-thalassemia and related diseases', u'adult-onset alpha-n-acetylgalactosaminidase deficiency', u'qualitative or quantitative defects of alphab-cristallin', u'alpha-thalassemia - intellectual disability syndrome linked to chromosome 16', u'(2alpha', u'pregnenolone 16alpha-carbonitrile', u'mature cd8_alpha-positive cd11b-negative dendritic cell', u'congenital muscular alpha-dystroglycanopathy with brain and eye anomalies', u'5alpha-dihydrotestosterone', u'1-dehydro-16alpha-methyl-9alpha-fluorohydrocortisone', u'alpha-thalassemia hydrops fetalis', u'bleeding diathesis due to integrin alpha2-beta1 deficiency', u'bilirubin ixalpha', u'alpha-1-antichymotrypsin deficiency', u'obsolete_alpha-1-antitrypsin deficiency', u'alpha thalassemia - retardation syndrome', u'(-)-alpha-methyl-l-tyrosine', u'congenital muscular dystrophy with integrin alpha-7 deficiency', u'interferon alpha measurement', u'familial hyperalphalipoproteinemia', u'alpha-l-iduronidase deficiency', u'alpha-1-antitrypsin def', u'cd8_alpha-low langerhans cell', u'alpha-amino-gamma-methylmercaptobutyric acid', u'fc-epsilon rialpha-low mast cell progenitor', u'soluble interleukin-2 receptor subunit alpha measurement', u'l-seryl-l-tyrosyl-l-seryl-l-methionyl-l-alpha-glutamyl-l-histidyl-l-phenylalanyl-l-arginyl-l-tryptophylglycyl-l-lysyl-l-prolyl-l-valylglycyl-l-lysyl-l-lysyl-l-arginyl-l-arginyl-l-prolyl-l-valyl-l-lysyl-l-valyl-l-tyrosyl-l-prolyl-l-alpha-aspartylglycyl-l-alanyl-l-alpha-glutamyl-l-alpha-aspartyl-l-alpha-glutamyl-l-seryl-l-alanyl-l-alpha-glutamyl-l-alanyl-l-phenylalanyl-l-prolyl-l-leucyl-l-alpha-glutamyl-l-phenylalanine', u'cd8-alpha-alpha-positive', u'cah due to 17-alpha-hydroxylase deficiency', u' alpha-beta regulatory t cell', u'immature cd8_alpha-low langerhans cell', u'susceptibility to respiratory infections associated with cd8alpha chain mutation', u'sterol 12 alpha hydroxylase', u'tcr-alpha-beta-positive t-cell deficiency', u'alpha-n-acetylgalactosaminidase deficiency type 2', u'alpha-crystallinopathy', u'alpha wave measurement', u'primary qualitative or quantitative defects of alpha-dystroglycan', u'steroid 12-alpha-hydroxylase', u'(s)-alpha-ethyl-2-oxo-1-pyrrolidineacetamide', u'hypercholesterolemia due to cholesterol 7alpha-hydroxylase deficiency', u'alpha macroglobulin measurement', u' alpha-beta thymocyte', u'(alphas)-n', u'alpha-thalassemia', u'17alpha-ethinyl estradiol', u'congenital deficiency in alpha-fetoprotein', u'alpha-beta t cell', u'alpha-tocopherol metabolism', u'alpha-(p-chlorophenoxy)isobutyric acid', u'alpha-methyl-l-p-tyrosine', u'alpha-naphthyl n-methylcarbamate', u'alpha-methyl-acyl-coa racemase deficiency', u'qualitative or quantitative defects of alpha-dystroglycan', u'lysine alpha-ketoglutarate reductase deficiency', u'alpha-amanitin', u'alpha-methyl-l-dopa', u'7-amino-9alpha-methoxymitosane', u'alpha-thalassemia-related diseases', u'alpha delta granule deficiency', u'(+-)-n-methyl-3-phenyl-3-((alpha', u'hyperalphalipoproteinemia', u'alpha-d-glucose', u'17alpha-methyltestosterone', u'1-dehydro-6alpha-methylhydrocortisone', u't-b+ severe combined immunodeficiency due to il-7ralpha deficiency', u'xy disorder of sex development due to 5-alpha-reductase 2 deficiency', u'alpha-n-acetylgalactosaminidase deficiency type 1', u'alpha-n-acetylgalactosaminidase deficiency type 3', u'alpha-naphthyl isothiocyanate', u'6-(p-hydroxy-alpha-aminophenylacetamido)penicillanic acid', u'alpha-dystroglycanopathy', u'l-alpha-pyrrolidinecarboxylic acid', u\"(5'alpha)-2-bromo-12'-hydroxy-5'-isobutyl-2'-isopropyl-3'\", u'delta(1)-6alpha-methylhydrocortisone', u'17alpha-methyl-3-oxo-4-androsten-17beta-ol', u'alpha-adrenergic blockers', u'vgamma5-positive cd8alpha alpha positive gamma-delta intraepithelial t cell', u'17beta-hydroxy-5alpha-androstan-3-one', u'(s)-(+)-alpha-methyl-4-carboxyphenylglycine', u'fibrinogen a alpha-chain amyloidosis', u'tnf-alpha; tumor necrosis factor', u'alpha-sarcoglycanopathy', u'alpha-p-chlorophenoxyisobutyryl ethyl ester', u'alpha-(4-chlorophenoxy)-alpha-methylpropionic acid', u'alpha-aminoadipic aciduria', u' alpha-beta t cell', u'alpha-galactosidase a deficiency', u'alphavirus infectious disease', u'streptomyces griseus subsp. alpha (ciferri 1927) pridham 1970 (approved lists 1980)', u'6alpha-methylprednisolone', u'(+)-alpha-tocopherol', u'alpha-(2-(diisopropylamino)ethyl)-alpha-phenyl-2-pyridineacetamide', u'alpha', u'alpha-methyl-l-tyrosine', u'homozygous alpha0-thalassemia', u'mature cd8_alpha-low langerhans cell', u'alpha-thalassemia intermedia', u'alpha-beta intraepithelial t cell', u'cd8-alpha alpha positive', u'alpha-adrenoceptor antagonists', u'alpha-mannosidosis', u'(e)-(+-)-beta-((4-chlorophenyl)methylene)-alpha-(1', u'alpha-ketoglutarate dehydrogenase deficiency', u'(s)-(alpha)-methyl-4-carboxyphenylglycine', u'alpha-kainic acid', u'5alpha-campestan-3beta-ol', u'alpha-tocopherol', u'familial renal amyloidosis due to fibrinogen a alpha-chain variant', u'interferon alpha 2a', u'(s)-alpha-methyltyrosine', u'(+)-(s)-6-methoxy-alpha-methyl-2-naphthaleneacetic acid', u'(s)-alpha-amino-5-fluoro-3', u'alpha-methylguanidino acetic acid', u'congenital alpha2 antiplasmin deficiency', u'cd8_alpha-negative cd11b-negative dendritic cell', u'alpha 1-antitrypsin deficiency', u'1-alpha-hydroxylase deficiency', u'alpha-methyl-para-tyrosine', u'pancreatic alpha cell', u'alpha-adrenergic blocker', u'alpha-butene', u' alpha-beta memory t cell', u'(+)-alpha-methyl-4-carboxyphenylglycine', u'alpha-amino-beta-phenylpropionic acid', u'alpha-adrenergic antagonist', u'alpha-lipoprotein cholesterol', u'cytochrome p450 subfamily viiib (sterol 12-alpha-hydroxylase) polypeptide 1', u'alpha-1-antitrypsin deficiency', u'congenital muscular dystrophy due to laminin alpha2 deficiency', u'alpha storage pool deficiency', u'vgamma5-negative cd8alpha alpha positive gamma-delta intraepithelial t cell', u'limb-girdle muscular dystrophy due to alpha-sarcoglycan deficiency', u'pyruvate dehydrogenase complex e1 component subunit alpha deficiency']\n",
      "[u'alpha chain defici', u'due to cholesterol 7alpha hydroxylas defici', u'dehydrogenas e1 alpha defici', u'4 alpha hydroxyphenylpyruv hydroxylas defici', u'alphaviru infecti diseas', u'hereditari amyloid nephropathi due to fibrinogen a alpha chain', u'alpha 2 6 2 cyanophenoxi 4 pyrimidinyloxi alpha methoxymethylen benzeneacet acid', u'disord of sex develop due to 5 alpha reductas 2 defici', u'protein involv in o glycosyl of alpha dystroglycan', u'congenit muscular alpha dystroglycanopathi with brain and', u'adren hyperplasia due to 17 alpha hydroxylas', u'amyloid nephropathi due to fibrinogen a alpha chain variant', u'matur cd8alpha neg cd11b neg dendrit cell', u'alpha thalassemia hydrop fetali', u'to alpha sarcoglycan defici', u'to respiratori infect associ with cd8alpha chain mutat', u'or quantit defect of protein involv in o glycosyl of alpha dystroglycan', u'alpha beta intraepitheli t', u'alpha thalassemia myelodysplast', u'sterol 12 alpha hydroxylas', u'p450 subfamili viiib sterol 12 alpha hydroxylas polypeptid 1', u'alpha b crystallin relat late onset distal', u'or quantit defect of integrin alpha 7', u'congenit alpha2 antiplasmin', u'diathesi due to integrin alpha2 beta1 defici', u'to laminin alpha2', u'interleukin 2 receptor alpha chain defici', u'to cholesterol 7alpha hydroxylas defici', u'congenit muscular alpha dystroglycanopathi with brain and eye', u'alpha amino gamma methylmercaptobutyr acid', u'famili hypoalphalipoproteinemia', u'serum alpha 1 antitrypsin', u'glycosyl due to steroid 5alpha reductas', u'fibrinogen a alpha chain', u'fc epsilon rialpha low mast', u'due to pterin 4 alpha carbinolamin dehydratas defici', u'amyloidosi due to fibrinogen a alpha chain variant', u'5alpha reductas type 3 defici', u'alpha adrenerg antagonist', u'of integrin alpha 7', u'cd8alpha neg thymic convent dendrit cell', u'tcr alpha beta t cell defici', u'streptomyc griseu subsp alpha ciferri 1927 pridham 1970', u'12 alpha hydroxylas polypeptid 1', u'due to steroid 5alpha reductas type', u'with cd8alpha chain mutat', u'quantit defect of protein involv in o glycosyl of alpha dystroglycan', u'associ with cd8alpha', u'due to steroid 5alpha reductas type 3 defici', u'immatur cd8alpha neg cd11b posit', u'matur alpha beta t cell', u'or quantit defect of alpha dystroglycan', u'vgamma5 neg cd8alpha alpha posit gamma delta intraepitheli t', u's alpha ethyl 2 oxo 1 pyrrolidineacetamid', u'alpha storag pool defici', u'cd8alpha neg plasmactyoid dendrit cell', u'combin immunodefici due to il 7ralpha defici', u'alpha thalassemia intellectu disabl syndrom link', u'cd8alpha neg cd11b neg dendrit', u'to steroid 5alpha reductas', u'bleed diathesi due to integrin alpha2 beta1', u'homozyg alpha0 thalassemia', u'matur cd8alpha posit cd11b neg dendrit cell', u'alpha methyl 4 carboxyphenylglycin', u'rs 4 hydroxi alpha1 6 4 phenylbutoxyhexylaminomethyl 1', u'qualit or quantit defect of alpha sarcoglycan', u'alpha beta t cell', u'xy disord of sex develop due to 5 alpha reductas', u'alpha dystroglycanopathi with brain', u'alpha thalassemia and relat', u'famili renal amyloidosi due to fibrinogen a alpha chain variant', u'alpha thalassemia', u'alpha 1 antitrypsin pittsburgh mutat', u'quantit defect of alpha dystroglycan', u'due to alpha 1 antitrypsin', u'cd8alpha posit thymic convent dendrit cell', u'respiratori infect associ with cd8alpha', u'diseas due to alpha 1 antitrypsin pittsburgh', u'quantit defect of alphab cristallin', u'immunodefici due to il 7ralpha defici', u'or quantit defect of alpha actin', u'suscept to respiratori infect associ with cd8alpha', u'cd8alpha alpha posit gamma delta', u'immatur cd8alpha neg cd11b neg dendrit cell', u'dehydrogenas complex e1 compon subunit alpha defici', u'immatur cd8alpha neg cd11b neg dendrit', u'renal amyloidosi due to fibrinogen a alpha chain', u'3 carboxamido 4 hydroxi alpha 1 methyl 3 phenylpropylaminomethylbenzyl alcohol', u'pterin 4 alpha carbinolamin dehydratas defici', u'of protein involv in o glycosyl of alpha dystroglycan', u'l histidylglycyl l alpha glutamylglycyl l threonyl l phenylalanyl l threonyl l seryl l alpha aspartyl l leucyl l seryl l lysyl l glutaminyl l methionyl l alpha glutamyl l alpha glutamyl l alpha glutamyl l alanyl l valyl l arginyl l leucyl l phenylalanyl l isoleucyl l alpha glutamyl l tryptophyl l leucyl l lysyl l asparaginylglycylglycyl l prolyl l seryl l serylglycyl l alanyl l prolyl l prolyl l prolyl l serinamid', u'matur cd8alpha low langerhan cell', u'3 alpha acetonyl 4 nitrobenzyl 4 hydroxycoumarin', u'diseas due to alpha 1 antitrypsin pittsburgh mutat', u'alpha methyl l p tyrosin', u't b scid due to il 7ralpha defici', u'diathesi due to integrin alpha2 beta1', u'tnf alpha tumor necrosi', u'alpha posit gamma delta intraepitheli t', u'sex develop due to 5 alpha reductas 2', u'12 alpha hydroxylas', u'muscular dystrophi with integrin alpha 7', u'disord of glycosyl due to steroid 5alpha reductas', u'due to alpha 1 antitrypsin pittsburgh', u'oxysterol 7 alpha hydroxylas defici', u'of peroxisom alpha', u'alpha ketoacid dehydrogenas complex defici', u'due to integrin alpha2 beta1 defici', u't b sever combin immunodefici due to il 7ralpha', u'alpha galactosidas a defici', u'alpha ketoglutar dehydrogenas defici', u'alpha naphthaleneacet acid', u'hypoalphalipoproteinemia', u'famili amyloid nephropathi due to fibrinogen a alpha chain', u'ciferri 1927 griseu subsp alpha', u'or quantit defect of alpha sarcoglycan', u'alpha delta granul defici', u'lysin alpha ketoglutar reductas', u'alpha 2 diisopropylaminoethyl alpha phenyl 2 pyridineacetamid', u's alpha methyltyrosin', u'to 5 alpha reductas 2', u'griseu subsp alpha ciferri 1927 pridham 1970 approv', u'quantit defect of integrin alpha 7', u'congenit muscular alpha dystroglycanopathi with brain and eye anomali', u'alpha mate type', u'chain alpha ketoacid dehydrogenas complex defici', u'alpha dystroglycanopathi with brain and eye', u'hypercholesterolemia due to cholesterol 7alpha hydroxylas', u'alpha methyl l dopa', u'alpha crystallinopathi', u'a alpha chain variant', u'glycosyl due to steroid 5alpha reductas type', u'disord of glycosyl due to steroid 5alpha reductas type 3', u'griseu subsp alpha ciferri', u'chain alpha ketoacid dehydrogenas complex', u'22 5alpha campestan 3beta', u'vgamma5 neg cd8alpha alpha posit gamma delta intraepitheli t cell', u's alpha methyl 4 2 methylpropylbenzeneacet acid', u'e beta 4 chlorophenylmethylen alpha 1', u'complex e1 compon subunit alpha defici', u'due to 5 alpha reductas 2', u'receptor alpha chain', u'alpha thalassemia x link', u'respiratori infect associ with cd8alpha chain mutat', u's 6 methoxi alpha methyl 2 naphthaleneacet acid', u'cd8alpha neg cd11b posit dendrit', u'to cholesterol 7alpha hydroxylas', u'alpha beta cytotox t', u'l alpha pyrrolidinecarboxyl acid', u'to 17 alpha hydroxylas defici', u'subfamili viiib sterol 12 alpha hydroxylas polypeptid 1', u'alpha beta memori t cell', u'hyperplasia due to 17 alpha hydroxylas', u'limb girdl muscular dystrophi due to alpha sarcoglycan defici', u'matur cd8alpha low langerhan', u'1 dehydro 16alpha methyl 9alpha fluorohydrocortison', u'congenit muscular alpha dystroglycanopathi with', u'ciferri 1927 alpha', u'or quantit defect of alphab cristallin', u'subfamili viiib sterol 12 alpha hydroxylas polypeptid', u'griseu subsp alpha ciferri 1927 pridham 1970 approv list', u'sex develop due to 5 alpha reductas', u'branch chain alpha ketoacid dehydrogenas complex', u'cd8alpha chain mutat', u'cd8alpha neg plasmactyoid dendrit', u'solubl interleukin 2 receptor subunit alpha', u'integrin alpha 7 defici', u't b scid due to il 7ralpha', u'matingtypealpha', u'alpha p chlorophenoxyisobutyryl ethyl ester', u'dystrophi with integrin alpha 7 defici', u'6alpha methylprednisolon', u'3 alpha p nitrophenol beta acetylethyl 4 hydroxycoumarin', u'defici in alpha fetoprotein', u'muscular alpha dystroglycanopathi with brain and', u'alpha storag pool', u'cd8 alpha beta posit', u'alpha naphthyl n methylcarbam', u'secondari alpha dystroglycanopathi', u'amyloidosi due to fibrinogen a alpha chain', u'alpha mannosidosi', u'approv list 1980 subsp alpha ciferri 1927 pridham 1970', u'alpha aminoadip aciduria', u'alpha beta cytokin secret effector t', u'approv list 1980 alpha ciferri 1927 pridham 1970', u'2s 2alpha', u'congenit adren hyperplasia due to 17 alpha hydroxylas', u'17beta hydroxi 5alpha androstan 3 one', u'tnf alpha tumor necrosi factor', u'alpha thalassemia intellectu disabl syndrom link to chromosom 16', u'streptomyc griseu subsp alpha ciferri', u'xy disord of sex develop due to 5 alpha reductas 2 defici', u'congenit adren hyperplasia due to 17 alpha hydroxylas defici', u'alpha adrenoceptor antagonist', u'alpha ciferri 1927 pridham 1970', u'sterol 12 alpha', u'alpha thalassemia and relat diseas', u'2r 2alpha', u'alpha adrenerg blocker', u'alpha wave', u'alpha 1 antitrypsin defici', u'alpha methyl l tyrosin', u'to laminin alpha2 defici', u'due to 17 alpha hydroxylas', u'lysin alpha ketoglutar reductas defici', u'griseu subsp alpha', u'gamma diisopropylamino alpha phenyl alpha 2 pyridylbutyramid', u'dystrophi due to laminin alpha2 defici', u'of sex develop due to 5 alpha reductas', u'alpha sarcoglycanopathi', u'nephropathi due to fibrinogen a alpha chain variant', u'alpha ketoacid dehydrogenas complex', u'5alpha 2 bromo 12 hydroxi 5 isobutyl 2 isopropyl 3', u'alpha n acetylgalactosaminidas defici type 2', u'alpha n acetylgalactosaminidas defici type 3', u'alpha n acetylgalactosaminidas defici type 1', u'hemorrhag diseas due to alpha 1 antitrypsin', u'immunodefici due to il 7ralpha', u'hyperalphalipoproteinemia', u'defect of protein involv in o glycosyl of alpha dystroglycan', u'primari qualit or quantit defect of alpha dystroglycan', u'scid due to il 7ralpha', u'alpha thalassemia intellectu disabl', u'muscular dystrophi due to laminin alpha2', u'due to 17 alpha hydroxylas defici', u'renal amyloidosi due to fibrinogen a alpha chain variant', u'subsp alpha ciferri 1927 pridham', u'sterol 12 alpha hydroxylas viiib', u'vgamma5 posit cd8alpha alpha posit gamma delta intraepitheli', u'dl alpha tocopherol', u'branch chain alpha ketoacid dehydrogenas complex defici', u'cd8alpha posit thymic convent dendrit', u'with integrin alpha 7', u'sterol 12 alpha hydroxylas subfamili viiib', u'5alpha', u'o glycosyl of alpha dystroglycan', u'3 alpha p nitrophenyl beta acetylethyl 4 hydroxycoumarin', u'necrosi factor alpha', u'alpha dens granul defici', u'interleukin 2 receptor subunit alpha', u'alpha posit gamma delta intraepitheli t cell', u'alpha thalassemia x link intellectu disabl syndrom', u'alpha p chlorophenoxyisobutyr acid', u'associ with cd8alpha chain mutat', u'alpha beta intraepitheli t cell', u'alpha b crystallin relat late onset distal myopathi', u'24r 5alpha ergostan 3beta ol', u'muscular dystrophi due to laminin alpha2 defici', u'qualit or quantit defect of integrin alpha 7', u'qualit or quantit defect of alphab cristallin', u'alpha', u'infect associ with cd8alpha', u'defect of alphab cristallin', u'alpha thalassemia intellectu disabl syndrom link to', u'of sex develop due to 5 alpha reductas 2 defici', u'muscular alpha dystroglycanopathi with brain and eye anomali', u'fc epsilon rialpha high basophil progenitor cell', u'sever combin immunodefici due to il 7ralpha defici', u'scid due to il 7ralpha defici', u'pyruv dehydrogenas complex e1 compon subunit alpha defici', u'alpha delta granul', u'cah due to 17 alpha hydroxylas', u'alpha methyl acetoacetyl coa thiolas defici', u'alpha butylen', u'alpha tocopherol metabol process', u'to alpha 1 antitrypsin pittsburgh', u'alpha ketoglutar reductas defici', u'with integrin alpha 7 defici', u'5alpha campestan 3beta ol', u'vgamma5 neg cd8alpha alpha posit gamma delta intraepitheli', u'cd8alpha neg cd11b posit dendrit cell', u'interferon alpha 2a', u'in o glycosyl of alpha dystroglycan', u'hemorrhag diseas due to alpha 1 antitrypsin pittsburgh', u'vgamma5 neg cd8alpha alpha', u'immatur cd8alpha low langerhan', u'immatur alpha beta t', u'alpha beta thymocyt', u'to integrin alpha2 beta1', u'pancreat alpha cell', u'7alpha 9 4', u'alpha methyltyrosin', u'due to steroid 5alpha reductas type 3', u'alpha 2 6 2 cyanophenoxi 4 pyrimidinyloxi alpha methoxymethylen benzeneacet acid methyl', u'17alpha ethynylestradiol', u'laminin alpha2 defici', u'vgamma5 posit cd8alpha alpha posit', u'6 p hydroxi alpha aminophenylacetamidopenicillan acid', u'alpha posit gamma delta', u'immatur cd8alpha neg cd11b neg', u'vgamma5 posit cd8alpha alpha posit gamma delta intraepitheli t cell', u'muscular dystrophi due to alpha sarcoglycan', u'alpha dystroglycanopathi', u'sterol 12 alpha hydroxylas polypeptid 1', u'integrin alpha2 beta1 defici', u'dehydrogenas complex e1 compon subunit alpha', u'17alpha methyl delta4 androsten 17beta ol 3 one', u'alpha hydro omega hydroxypolyoxyethylen', u'immatur cd8alpha posit cd11b neg', u'compon subunit alpha', u'alpha macroglobulin', u'vgamma5 posit cd8alpha alpha', u'immatur alpha beta t cell', u'muscular dystrophi due to alpha sarcoglycan defici', u'a alpha chain amyloidosi', u'rialpha high basophil progenitor', u'alpha beta cytokin secret effector', u'interferon alpha', u'alpha n acetylgalactosaminidas defici', u'sterol 12 alpha hydroxylas p450 subfamili viiib', u'hyperphenylalaninemia due to pterin 4 alpha carbinolamin dehydratas defici', u'disord of peroxisom alpha', u'alpha thalassemia retard', u'hyperplasia due to 17 alpha hydroxylas defici', u'hereditari persist of alpha fetoprotein', u'alpha l iduronidas defici', u'to steroid 5alpha reductas type 3', u'compon subunit alpha defici', u'immatur cd8alpha neg cd11b posit dendrit cell', u'alpha globulin', u'matur cd8alpha neg cd11b neg', u'alpha adrenerg agonist', u'sterol 12 alpha hydroxylas cytochrom p450 subfamili viiib', u'griseu subsp alpha ciferri 1927 pridham', u'due to laminin alpha2', u'immatur cd8alpha posit cd11b neg dendrit cell', u'alpha thalassemia relat diseas', u'17alpha estradiol', u'suscept to respiratori infect associ with cd8alpha chain', u'cytochrom p450 subfamili viiib sterol 12 alpha hydroxylas polypeptid 1', u'famili amyloid nephropathi due to fibrinogen a alpha chain variant', u'alpha ciferri 1927 pridham', u'alpha dystroglycanopathi with brain and', u'alpha b crystallin relat late onset', u'alpha thalassemia intellectu', u'quantit defect of alpha actin', u'analphalipoproteinemia', u'streptomyc griseu subsp alpha', u'develop due to 5 alpha reductas 2 defici', u'congenit disord of glycosyl due to steroid 5alpha reductas type', u'limb girdl muscular dystrophi due to alpha sarcoglycan', u'rialpha low mast cell', u'to pterin 4 alpha carbinolamin dehydratas', u'vgamma5 posit cd8alpha alpha posit gamma delta', u'alpha adrenerg receptor blockad', u'l alpha acetamido beta mercaptopropion acid', u'streptomyc griseu subsp alpha ciferri 1927 pridham 1970 approv', u'cd8alpha posit cd11b neg dendrit', u'alpha cell', u'alpha amino beta phenylpropion acid', u'cd8alpha posit thymic convent', u'cd8alpha alpha posit gamma delta intraepitheli t cell', u'qualit or quantit defect of protein involv in o glycosyl of alpha dystroglycan', u'immatur cd8alpha neg cd11b posit dendrit', u'infect associ with cd8alpha chain', u'1 o chloro alpha', u'alpha methyl beta 3', u'alpha beta cytokin secret', u'alpha beta regulatori t cell', u'subsp alpha ciferri 1927 pridham 1970', u'defect of alpha dystroglycan', u'interleukin 2 receptor alpha', u'subunit alpha defici', u'tcr alpha beta posit t cell defici', u'fc epsilon rialpha high basophil progenitor', u'cah due to 17 alpha hydroxylas defici', u'disord alpha 1 antitrypsin defici', u'to respiratori infect associ with cd8alpha', u'disord of glycosyl due to steroid 5alpha reductas type 3 defici', u'cd8alpha low langerhan cell', u'alpha ciferri 1927 pridham 1970 approv list', u'steroid 5alpha reductas type 3', u'alpha thalassemia myelodysplast syndrom', u'fibrinogen a alpha chain variant', u'muscular alpha dystroglycanopathi with brain', u'4 androsten 17alpha methyl 17beta ol 3 one', u'rialpha high basophil progenitor cell', u'amyloid nephropathi due to fibrinogen a alpha chain', u'dystrophi with integrin alpha 7', u'to pterin 4 alpha carbinolamin dehydratas defici', u'nephropathi due to fibrinogen a alpha chain', u'congenit alpha2 antiplasmin defici', u'due to alpha 1 antitrypsin pittsburgh mutat', u'due to 5 alpha reductas', u'cd8alpha neg thymic convent dendrit', u'2alpha', u'alpha beta regulatori t', u'involv in o glycosyl of alpha dystroglycan', u'3alpha', u'to integrin alpha2 beta1 defici', u'to steroid 5alpha reductas type 3 defici', u'quantit defect of alpha sarcoglycan', u's alpha methyl 4 carboxyphenylglycin', u'to il 7ralpha defici', u'alpha thalassemia intellectu disabl syndrom', u'n methyl 3 phenyl 3 alpha', u'congenit muscular alpha dystroglycanopathi with brain', u'hereditari renal amyloidosi due to fibrinogen a alpha chain', u'to 5 alpha reductas 2 defici', u'7 amino 9alpha methoxymitosan', u'congenit defici in alpha fetoprotein', u'alpha d glucos', u'3 alpha 4 nitrophenyl beta acetylethyl 4 hydroxycoumarin', u'alpha ciferri 1927 pridham 1970 approv', u'immatur cd8alpha low langerhan cell', u'ajmalan 17alpha', u's alpha amino 5 fluoro 3', u'pyruv dehydrogenas e1 alpha defici', u'famili renal amyloidosi due to fibrinogen a alpha chain', u'muscular dystrophi with integrin alpha 7 defici', u'alpha methyl acyl coa racemas defici', u'heparan alpha glucosaminid n acetyltransferas defici', u'l alpha methyltyrosin', u'alpha ketopropionaldehyd', u'due to steroid 5alpha reductas', u'adult onset alpha n acetylgalactosaminidas defici', u'bleed diathesi due to integrin alpha2 beta1 defici', u'receptor alpha chain defici', u'3 alpha phenyl beta acetylethyl 4 hydroxycoumarin', u'of glycosyl due to steroid 5alpha reductas type 3 defici', u'qualit or quantit defect of alpha actin', u'cd8alpha neg cd11b neg dendrit cell', u'qualit or quantit defect of alpha dystroglycan', u'alpha 1 antitrypsin def', u'e1 compon subunit alpha defici', u'alpha n acetylgalactosaminidas defici type', u'cd8alpha alpha posit', u'alpha lipoprotein cholesterol', u'associ with cd8alpha chain', u'with cd8alpha chain', u'receptor subunit alpha', u'cd8alpha posit plasmactyoid dendrit', u'defect of alpha sarcoglycan', u'of sex develop due to 5 alpha reductas 2', u'5 alpha reductas 2 defici', u'to steroid 5alpha reductas type', u'matur cd8alpha neg cd11b posit dendrit', u'cd8alpha posit cd11b neg dendrit cell', u'dystrophi due to laminin alpha2', u'subsp alpha ciferri 1927 pridham 1970 approv list', u'alpha methyl para tyrosin', u'alphaviru infect', u'alpha beta cytokin secret effector t cell', u'defect of alpha actin', u'alpha beta cytotox t cell', u'matur alpha beta t', u'17alpha methyl 3 oxo 4 androsten 17beta ol', u'matur cd8alpha neg cd11b neg dendrit', u'bilirubin ixalpha', u'obsoletealpha 1 antitrypsin defici', u'due to laminin alpha2 defici', u'due to integrin alpha2 beta1', u'alpha 1 antichymotrypsin defici', u'chain alpha ketoacid dehydrogenas', u'immatur cd8alpha posit cd11b neg dendrit', u'persist of alpha fetoprotein', u'vgamma5 posit cd8alpha alpha posit gamma delta intraepitheli t', u'alpha methylguanidino acet acid', u'adren hyperplasia due to 17 alpha hydroxylas defici', u'muscular alpha dystroglycanopathi with brain and eye', u'griseu subsp alpha ciferri 1927 pridham 1970', u't b sever combin immunodefici due to il 7ralpha defici', u'17alpha ethinyl estradiol', u'glycosyl due to steroid 5alpha reductas type 3 defici', u's alpha amino 4 carboxi 2 methylbenzeneacet', u'alpha 4 chlorophenoxi alpha methylpropion acid', u'alpha buten', u'fc epsilon rialpha low mast cell progenitor', u'complex e1 compon subunit alpha', u'due to 5 alpha reductas 2 defici', u'steroid 12 alpha hydroxylas', u'alpha thalassemia intellectu disabl syndrom link to chromosom', u'combin immunodefici due to il 7ralpha', u'sterol 12 alpha hydroxylas polypeptid', u'17alpha methyltestosteron', u'disord of sex develop due to 5 alpha reductas 2', u'cd8alpha alpha posit gamma delta intraepitheli', u'ciferri 1927 streptomyc griseu subsp alpha', u'lysosom alpha d mannosidas defici', u'dystrophi due to alpha sarcoglycan defici', u'due to alpha sarcoglycan defici', u'subsp alpha ciferri', u'alpha methyl p tyrosin', u'famili hyperalphalipoproteinemia', u'alpha 2 6 2 cyanophenoxi 4 pyrimidinyloxi alpha methoxymethylen benzeneacet acid methyl ester', u'infect associ with cd8alpha chain mutat', u'alpha amanitin', u'hereditari renal amyloidosi due to fibrinogen a alpha chain variant', u'branch chain alpha ketoacid dehydrogenas', u'cholesterol 7alpha hydroxylas defici', u'to alpha 1 antitrypsin pittsburgh mutat', u'due to alpha sarcoglycan', u'disord of glycosyl due to steroid 5alpha reductas type', u'alpha2 antiplasmin defici', u'of glycosyl due to steroid 5alpha reductas type 3', u'due to fibrinogen a alpha chain variant', u'alpha posit gamma delta intraepitheli', u'to fibrinogen a alpha chain', u'ciferri 1927 subsp alpha', u'steroid 5alpha reductas type 3 defici', u'1 alpha hydroxylas defici', u'fc epsilon rialpha high basophil', u'alpha l fucosidas defici', u'xy disord of sex develop due to 5 alpha reductas 2', u'dystrophi due to alpha sarcoglycan', u'glycosyl due to steroid 5alpha reductas type 3', u'mate type alpha', u'alpha peak frequenc', u'of glycosyl due to steroid 5alpha reductas', u'alpha thalassemia retard syndrom', u'pregnenolon 16alpha carbonitril', u'hyperphenylalaninemia due to pterin 4 alpha carbinolamin dehydratas', u'subsp alpha ciferri 1927 pridham 1970 approv', u'congenit muscular dystrophi with integrin alpha 7 defici', u'of glycosyl due to steroid 5alpha reductas type', u'n alpha', u'alpha thalassemia x link intellectu', u'platelet alpha granul defici', u'alpha 1', u'alpha n', u'n acetyl alpha glucosaminidas defici', u'tumor necrosi factor alpha', u'cd8 alpha alpha posit', u'alpha naphthyl isothiocyan', u'cytochrom p450 subfamili viiib sterol 12 alpha hydroxylas polypeptid', u'congenit disord of glycosyl due to steroid 5alpha reductas', u'matur cd8alpha posit cd11b neg', u'n 1s 1 dihydroxyboranyl 3 methylbutyl nalpha pyrazin 2 ylcarbonyl d phenylalaninamid', u'vgamma5 neg cd8alpha alpha posit gamma delta', u'l seryl l tyrosyl l seryl l methionyl l alpha glutamyl l histidyl l phenylalanyl l arginyl l tryptophylglycyl l lysyl l prolyl l valylglycyl l lysyl l lysyl l arginyl l arginyl l prolyl l valyl l lysyl l valyl l tyrosyl l prolyl l alpha aspartylglycyl l alanyl l alpha glutamyl l alpha aspartyl l alpha glutamyl l seryl l alanyl l alpha glutamyl l alanyl l phenylalanyl l prolyl l leucyl l alpha glutamyl l phenylalanin', u'alpha tocopherol', u'e1 compon subunit alpha', u'due to fibrinogen a alpha chain', u'due to il 7ralpha', u'pyruv dehydrogenas e1 alpha', u'steroid 5alpha reductas type', u'cd8alpha neg', u'matur cd8alpha posit cd11b neg dendrit', u'congenit muscular dystrophi with integrin alpha 7', u'alpha phthalimidoglutarimid', u'fc epsilon rialpha low mast cell', u'primari alpha dystroglycanopathi', u'vgamma5 neg cd8alpha alpha posit', u'alpha kainic acid', u'7 alpha hydroxi 4 cholesten 3 one 12 alpha hydroxylas', u'delta1 6alpha methylhydrocortison', u'congenit muscular dystrophi due to laminin alpha2 defici', u'defect of integrin alpha 7', u'suscept to respiratori infect associ with cd8alpha chain mutat', u'3 alpha acetonyl p nitrobenzyl 4 hydroxycoumarin', u'17alpha ethynylestra 1', u'yeast alpha mate type', u'alpha thalassemia intermedia', u'1 dehydro 6alpha methylhydrocortison', u'subunit alpha', u'alpha dens granul', u'alpha tocopherol metabol', u'due to pterin 4 alpha carbinolamin dehydratas', u'viiib sterol 12 alpha hydroxylas polypeptid 1', u'develop due to 5 alpha reductas', u'cd8alpha neg thymic convent', u'to respiratori infect associ with cd8alpha chain', u'develop due to 5 alpha reductas 2', u'diseas due to alpha 1 antitrypsin', u'sex develop due to 5 alpha reductas 2 defici', u'alpha thalassemia x link intellectu disabl', u'5alpha reductas type 3', u'to fibrinogen a alpha chain variant', u'matur cd8alpha neg cd11b posit', u'hereditari amyloid nephropathi due to fibrinogen a alpha chain variant', u'sever combin immunodefici due to il 7ralpha', u'due to cholesterol 7alpha hydroxylas', u'alpha methyldopa', u'disord of sex develop due to 5 alpha reductas', u'rialpha low mast cell progenitor', u'hemorrhag diseas due to alpha 1 antitrypsin pittsburgh mutat', u'alpha beta memori t', u'congenit muscular dystrophi due to laminin alpha2', u'branch chain alpha ketoacid', u'approv list 1980 streptomyc griseu subsp alpha ciferri 1927 pridham 1970', u'alpha granul diseas', u'congenit disord of glycosyl due to steroid 5alpha reductas type 3', u'respiratori infect associ with cd8alpha chain', u'streptomyc griseu subsp alpha ciferri 1927 pridham 1970 approv list', u'due to pterin 4 alpha carbinolamin', u'cd8alpha posit plasmactyoid dendrit cell', u'cd8alpha alpha posit gamma delta intraepitheli t', u'congenit disord of glycosyl due to steroid 5alpha reductas type 3 defici', u'due to il 7ralpha defici', u'pyruv dehydrogenas complex e1 compon subunit alpha', u'hypercholesterolemia due to cholesterol 7alpha hydroxylas defici', u'matur cd8alpha neg cd11b posit dendrit cell', u'streptomyc griseu subsp alpha ciferri 1927 pridham', u'hyperphenylalaninemia due to pterin 4 alpha carbinolamin', u'interleukin 2 receptor alpha chain', u'muscular alpha dystroglycanopathi with', u'congenit muscular alpha dystroglycanopathi', u'p450 subfamili viiib sterol 12 alpha hydroxylas polypeptid', u'approv list 1980 griseu subsp alpha ciferri 1927 pridham 1970', u'viiib sterol 12 alpha hydroxylas polypeptid', u'5alpha dihydrotestosteron', u'alpha dystroglycanopathi with brain and eye anomali', u'fibrinogen a alpha chain amyloidosi']\n"
     ]
    }
   ],
   "source": [
    "print [ph for ph in efo_phenotype_list0 if 'alpha' in ph]\n",
    "print [ph for ph in efo_phen_matcher.d if 'alpha' in ph]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'NoneType' object has no attribute 'split'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-83-9f9ccf33050a>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      9\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mphenotype\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     10\u001b[0m         \u001b[0mphenotype_names\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0madd\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mphenotype\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 11\u001b[0;31m         \u001b[0msynonyms\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msyn\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0msyn\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mphenotype\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msynonyms\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'|'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     12\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mquery_word\u001b[0m \u001b[0;32min\u001b[0m \u001b[0msynonyms\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0mquery_word\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0mphenotype\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     13\u001b[0m             \u001b[0;32mprint\u001b[0m \u001b[0mphenotype\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mphenotype\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0montology_ref\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'NoneType' object has no attribute 'split'"
     ]
    }
   ],
   "source": [
    "query_word = 'trait in'\n",
    "from db import db_session\n",
    "from db.schema import *\n",
    "\n",
    "phenotypes = db_session.query(Phenotype).filter(Phenotype.source=='snomed').all()\n",
    "# phenotypes == kb.get_phenotype_candidates_cheating()\n",
    "phenotype_names = set()\n",
    "for phenotype in phenotypes:\n",
    "    if phenotype.name:\n",
    "        phenotype_names.add((phenotype.name))\n",
    "        synonyms = [(syn) for syn in phenotype.synonyms.split('|')]\n",
    "        if query_word in synonyms or query_word == phenotype.name:\n",
    "            print phenotype.name, phenotype.ontology_ref\n",
    "        phenotype_names.update(synonyms)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(u'borderline personality disorder', u'borderlin person')]"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# print len(phenotype_names)\n",
    "[(word, change_name(word)) for word in snomed_phenotype_list if change_name(word) == change_name('borderline personality')]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extraction performance"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, extract some features for each candidate mention."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Building feature index...\n",
      "Extracting features...\n",
      "0/637836\n",
      "5000/637836\n",
      "10000/637836\n",
      "15000/637836\n",
      "20000/637836\n",
      "25000/637836\n",
      "30000/637836\n",
      "35000/637836\n",
      "40000/637836\n",
      "45000/637836\n",
      "50000/637836\n",
      "55000/637836\n",
      "60000/637836\n",
      "65000/637836\n",
      "70000/637836\n",
      "75000/637836\n",
      "80000/637836\n",
      "85000/637836\n",
      "90000/637836\n",
      "95000/637836\n",
      "100000/637836\n",
      "105000/637836\n",
      "110000/637836\n",
      "115000/637836\n",
      "120000/637836\n",
      "125000/637836\n",
      "130000/637836\n",
      "135000/637836\n",
      "140000/637836\n",
      "145000/637836\n",
      "150000/637836\n",
      "155000/637836\n",
      "160000/637836\n",
      "165000/637836\n",
      "170000/637836\n",
      "175000/637836\n",
      "180000/637836\n",
      "185000/637836\n",
      "190000/637836\n",
      "195000/637836\n",
      "200000/637836\n",
      "205000/637836\n",
      "210000/637836\n",
      "215000/637836\n",
      "220000/637836\n",
      "225000/637836\n",
      "230000/637836\n",
      "235000/637836\n",
      "240000/637836\n",
      "245000/637836\n",
      "250000/637836\n",
      "255000/637836\n",
      "260000/637836\n",
      "265000/637836\n",
      "270000/637836\n",
      "275000/637836\n",
      "280000/637836\n",
      "285000/637836\n",
      "290000/637836\n",
      "295000/637836\n",
      "300000/637836\n",
      "305000/637836\n",
      "310000/637836\n",
      "315000/637836\n",
      "320000/637836\n",
      "325000/637836\n",
      "330000/637836\n",
      "335000/637836\n",
      "340000/637836\n",
      "345000/637836\n",
      "350000/637836\n",
      "355000/637836\n",
      "360000/637836\n",
      "365000/637836\n",
      "370000/637836\n",
      "375000/637836\n",
      "380000/637836\n",
      "385000/637836\n",
      "390000/637836\n",
      "395000/637836\n",
      "400000/637836\n",
      "405000/637836\n",
      "410000/637836\n",
      "415000/637836\n",
      "420000/637836\n",
      "425000/637836\n",
      "430000/637836\n",
      "435000/637836\n",
      "440000/637836\n",
      "445000/637836\n",
      "450000/637836\n",
      "455000/637836\n",
      "460000/637836\n",
      "465000/637836\n",
      "470000/637836\n",
      "475000/637836\n",
      "480000/637836\n",
      "485000/637836\n",
      "490000/637836\n",
      "495000/637836\n",
      "500000/637836\n",
      "505000/637836\n",
      "510000/637836\n",
      "515000/637836\n",
      "520000/637836\n",
      "525000/637836\n",
      "530000/637836\n",
      "535000/637836\n",
      "540000/637836\n",
      "545000/637836\n",
      "550000/637836\n",
      "555000/637836\n",
      "560000/637836\n",
      "565000/637836\n",
      "570000/637836\n",
      "575000/637836\n",
      "580000/637836\n",
      "585000/637836\n",
      "590000/637836\n",
      "595000/637836\n",
      "600000/637836\n",
      "605000/637836\n",
      "610000/637836\n",
      "615000/637836\n",
      "620000/637836\n",
      "625000/637836\n",
      "630000/637836\n",
      "635000/637836\n",
      "Example features: [u'DDLIB_WORD_SEQ_[Crohn Disease]', u'DDLIB_LEMMA_SEQ_[crohn disease]', u'DDLIB_POS_SEQ_[NN NN]', u'DDLIB_DEP_SEQ_[compound compound]', u'DDLIB_W_LEFT_1_[novel]', u'DDLIB_W_LEFT_POS_1_[JJ]', u'DDLIB_W_LEFT_2_[. novel]', u'DDLIB_W_LEFT_POS_2_[. JJ]', u'DDLIB_W_LEFT_3_[ptger4 . novel]', u'DDLIB_W_LEFT_POS_3_[NN . JJ]']\n"
     ]
    }
   ],
   "source": [
    "import cPickle        \n",
    "from snorkel.features import NgramFeaturizer\n",
    "\n",
    "# pkl_f = 'phenotype_feats.pkl'\n",
    "# try:\n",
    "#     with open(pkl_f, 'rb') as f:\n",
    "#         featurizer = cPickle.load(f)\n",
    "# except:\n",
    "featurizer = NgramFeaturizer()\n",
    "featurizer.fit_transform(phen_c)\n",
    "\n",
    "print 'Example features:', [f for f in featurizer.get_features_by_candidate(phen_c[0])[:10]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Defined 2747 positive and 48286 negative gold mentions\n"
     ]
    }
   ],
   "source": [
    "# classify candidates as correct or not\n",
    "\n",
    "# doc_id -> set of correct gwc_ids\n",
    "gold_agg_dict_phen = { doc_id : set() for doc_id, phen_id in gold_set_agg_phens }\n",
    "for doc_id, phen_id in gold_set_agg_phens:\n",
    "    gold_agg_dict_phen[doc_id].add(phen_id)\n",
    "\n",
    "def c2uid(candidate):\n",
    "    return candidate.context.document.name, candidate.context.position, candidate.char_start, candidate.char_end\n",
    "    \n",
    "gt_dict_pos = dict()\n",
    "gt_dict_neg = dict()\n",
    "for candidate in phen_c:\n",
    "    doc_id = candidate.context.document.name\n",
    "    agg_ids = agg_phen2id.get(change_name(candidate.get_span()), set())\n",
    "    uid = c2uid(candidate)\n",
    "    if agg_ids & gold_agg_dict_phen[doc_id]:\n",
    "        gt_dict_pos[uid] = +1\n",
    "    else:\n",
    "        gt_dict_neg[uid] = -1\n",
    "\n",
    "gt_dict = dict(gt_dict_pos.items() + gt_dict_neg.items())\n",
    "print 'Defined %d positive and %d negative gold mentions' % (len(gt_dict_pos), len(gt_dict_neg))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "17447842\n",
      "crohn's disease 19827\n",
      "\tcrohn's disease http://www.ebi.ac.uk/efo/EFO_0000384 gastritis associated with crohn's disease|ileitis\n",
      "crohn's disease 19333\n",
      "\tcrohn's disease http://www.ebi.ac.uk/efo/EFO_0000384 gastritis associated with crohn's disease|ileitis\n",
      "crohn's disease 19486\n",
      "\tcrohn's disease http://www.ebi.ac.uk/efo/EFO_0000384 gastritis associated with crohn's disease|ileitis\n",
      "Novel Crohn Disease Locus Identified by Genome-Wide Association Maps to a Gene Desert on 5p13.1 and Modulates Expression of  PTGER4. To identify novel susceptibility loci for Crohn disease (CD), we undertook a genome-wide association study with more than 300,000 SNPs characterized in 547 patients and 928 controls. We found three chromosome regions that provided evidence of disease association with  p -values between 10 −6  and 10 −9 . Two of these ( IL23R  on Chromosome 1 and  CARD15  on Chromosome 16) correspond to genes previously reported to be associated with CD. In addition, a 250-kb region of Chromosome 5p13.1 was found to contain multiple markers with strongly suggestive evidence of disease association (including four markers with  p  < 10 −7 ). We replicated the results for 5p13.1 by studying 1,266 additional CD patients, 559 additional controls, and 428 trios. Significant evidence of association ( p  < 4 × 10 −4 ) was found in case/control comparisons with the replication data, while associated alleles were over-transmitted to affected offspring ( p  < 0.05), thus confirming that the 5p13.1 locus contributes to CD susceptibility. The CD-associated 250-kb region was saturated with 111 SNP markers. Haplotype analysis supports a complex locus architecture with multiple variants contributing to disease susceptibility. The novel 5p13.1 CD locus is contained within a 1.25-Mb gene desert. We present evidence that disease-associated alleles correlate with quantitative expression levels of the prostaglandin receptor EP4,  PTGER4,  the gene that resides closest to the associated region. Our results identify a major new susceptibility locus for CD, and suggest that genetic variants associated with disease risk at this locus could modulate  cis -acting regulatory elements of  PTGER4 . Individual susceptibility to many common diseases is determined by a combination of environmental and genetic factors. Identifying these genetic risk factors is one of the most important objectives of modern medical genetics, as it paves the way towards personalized medicine and drug target identification. Recent advances in SNP genotyping technology allows systematic association scanning of the entire genome for the detection of novel susceptibility loci. We herein apply this approach to Crohn disease, which afflicts an estimated 0.15% of the people in the developed world and identify a novel susceptibility locus on Chromosome 5. A unique feature of the novel 5p13.1 locus is that it coincides with a 1.25-Mb gene desert. We present evidence that genetic variants at this locus influence the expression levels of the closest gene,  PTGER4,  located 270 kb away, in the direction of the centromere. PTGER4  encodes the prostaglandin receptor EP4 and is a strong candidate susceptibility gene for Crohn disease as PTGER4 knock-out mice have increased susceptibility to colitis.\n",
      "1 Crohn Disease | crohn's disease http://www.ebi.ac.uk/efo/EFO_0000384\n",
      "-1 Maps to | mean arterial pressure http://www.ebi.ac.uk/efo/EFO_0006340\n",
      "-1 to a\n",
      "-1 Association\n",
      "-1 and\n",
      "-1 of\n",
      "-1 Crohn disease (CD),\n",
      "-1 susceptibility\n",
      "-1 a\n",
      "-1 association\n",
      "-1 in\n",
      "-1 and\n",
      "-1 disease association with\n",
      "-1 of disease\n",
      "-1 p -\n",
      "-1 -values\n",
      "-1 chromosome\n",
      "-1 and\n",
      "-1 Chromosome 16\n",
      "-1 associated with\n",
      "-1 of\n",
      "-1 Chromosome\n",
      "-1 and\n",
      "-1 , a\n",
      "-1 of Chromosome\n",
      "-1 of disease\n",
      "-1 disease association\n",
      "-1 In\n",
      "-1 was\n",
      "-1 suggestive\n",
      "-1 p\n",
      "-1 , and\n",
      "-1 to affected\n",
      "-1 of\n",
      "-1 association\n",
      "-1 p\n",
      "-1 ×\n",
      "-1 was\n",
      "-1 in\n",
      "-1 associated\n",
      "-1 p\n",
      "-1 confirming\n",
      "-1 susceptibility\n",
      "-1 was\n",
      "-1 to disease susceptibility\n",
      "-1 analysis\n",
      "-1 a\n",
      "-1 a\n",
      "-1 levels of the\n",
      "-1 prostaglandin receptor\n",
      "-1 disease-associated\n",
      "-1 associated\n",
      "-1 associated with disease\n",
      "-1 a major\n",
      "-1 , and\n",
      "-1 susceptibility\n",
      "-1 suggest\n",
      "-1 genetic\n",
      "-1 of\n",
      "-1 susceptibility to\n",
      "-1 common diseases\n",
      "-1 a\n",
      "-1 of\n",
      "-1 and\n",
      "-1 genetic\n",
      "-1 risk factors\n",
      "-1 of the\n",
      "-1 genetics,\n",
      "-1 , as\n",
      "-1 genetic\n",
      "-1 of\n",
      "-1 medical\n",
      "-1 personalized | personality http://www.ebi.ac.uk/efo/EFO_0003947\n",
      "-1 and\n",
      "-1 drug | response to drug http://purl.obolibrary.org/obo/GO_0042493\n",
      "-1 of the\n",
      "-1 in\n",
      "-1 genotyping\n",
      "-1 association\n",
      "-1 genome\n",
      "-1 detection\n",
      "-1 of\n",
      "-1 susceptibility\n",
      "1 to Crohn disease, | crohn's disease http://www.ebi.ac.uk/efo/EFO_0000384\n",
      "-1 of the\n",
      "-1 Chromosome 5\n",
      "-1 an\n",
      "-1 in\n",
      "-1 developed\n",
      "-1 and\n",
      "-1 a\n",
      "-1 susceptibility\n",
      "-1 of the\n",
      "-1 A\n",
      "-1 a\n",
      "-1 levels of the\n",
      "-1 , in\n",
      "-1 of the\n",
      "-1 genetic\n",
      "-1 prostaglandin receptor\n",
      "1 Crohn disease | crohn's disease http://www.ebi.ac.uk/efo/EFO_0000384\n",
      "-1 disease as\n",
      "-1 susceptibility to\n",
      "-1 to colitis\n",
      "-1 and\n",
      "-1 a\n",
      "-1 susceptibility\n",
      "-1 increased\n",
      "\n",
      "17658951\n",
      "obesity-related traits 19256\n",
      "\tadiponectin measurement http://www.ebi.ac.uk/efo/EFO_0004502 adiponectin levels\n",
      "\thoma-ir http://www.ebi.ac.uk/efo/EFO_0004501 homeostasis model assessment insulin resistance\n",
      "\tccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749 monocyte chemotactic protein-1 measurement|chemokine ccl2 level|monocyte chemoattractant protein-1 measurement|chemokine ccl2 levels|monocyte chemotactic protein-1 levels|monocyte chemotactic protein-1 level\n",
      "\tthyroid stimulating hormone measurement http://www.ebi.ac.uk/efo/EFO_0004748 thyrotropin measurement|thyroid stimulating hormone level|thyroid stimulating hormone levels\n",
      "\tprotein measurement http://www.ebi.ac.uk/efo/EFO_0004747 protein levels\n",
      "\tblood urea nitrogen measurement http://www.ebi.ac.uk/efo/EFO_0004741 \n",
      "\tlean body mass http://www.ebi.ac.uk/efo/EFO_0004995 \n",
      "\tserum alanine aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004735 alanine transaminase measurement|alanine aminotransferase measurement\n",
      "\taspartate aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004736 aspartate transaminase measurement\n",
      "\thormone measurement http://www.ebi.ac.uk/efo/EFO_0004730 \n",
      "\tfat body mass http://www.ebi.ac.uk/efo/EFO_0005409 total body fat measurement\n",
      "\tadipose tissue measurement http://www.ebi.ac.uk/efo/EFO_0004764 \n",
      "\ticam-1 measurement http://www.ebi.ac.uk/efo/EFO_0004520 soluble icam1 measurement|soluble intercellular adhesion molecule-1|sicam-1\n",
      "\ttriglyceride measurement http://www.ebi.ac.uk/efo/EFO_0004530 triglyceride levels\n",
      "\twaist circumference http://www.ebi.ac.uk/efo/EFO_0004342 \n",
      "\tbody mass index http://www.ebi.ac.uk/efo/EFO_0004340 quetelet's index|bmi\n",
      "\tobesity http://www.ebi.ac.uk/efo/EFO_0001073 overweight and obesity|obesity (disorder)|obese|obesity nos|obese (finding)|adiposity|adiposis|obesity [ambiguous]|obesity\n",
      "\tbirth weight http://www.ebi.ac.uk/efo/EFO_0004344 birthweight\n",
      "\twaist-hip ratio http://www.ebi.ac.uk/efo/EFO_0004343 \n",
      "\theart rate http://www.ebi.ac.uk/efo/EFO_0004326 pulse rate|cardiac chronotropy\n",
      "\tbody height http://www.ebi.ac.uk/efo/EFO_0004339 height\n",
      "\tbody weight http://www.ebi.ac.uk/efo/EFO_0004338 weight\n",
      "\thomocysteine measurement http://www.ebi.ac.uk/efo/EFO_0004578 homocysteine levels\n",
      "\ttotal cholesterol measurement http://www.ebi.ac.uk/efo/EFO_0004574 cholesterol\n",
      "\tccl5 measurement http://www.ebi.ac.uk/efo/EFO_0005117 regulated on activation\n",
      "\tigfbp-1 measurement http://www.ebi.ac.uk/efo/EFO_0005118 \n",
      "\tantioxidant measurement http://www.ebi.ac.uk/efo/EFO_0005119 \n",
      "\thead circumference http://www.ebi.ac.uk/efo/EFO_0005114 \n",
      "\tmetabolic rate measurement http://www.ebi.ac.uk/efo/EFO_0005115 \n",
      "\turinary metabolite measurement http://www.ebi.ac.uk/efo/EFO_0005116 \n",
      "\tfatty acid measurement http://www.ebi.ac.uk/efo/EFO_0005110 \n",
      "\tfolic acid measurement http://www.ebi.ac.uk/efo/EFO_0005111 folate measurement|vitamin b9 measurement\n",
      "\tgestational age http://www.ebi.ac.uk/efo/EFO_0005112 \n",
      "\tbody composition measurement http://www.ebi.ac.uk/efo/EFO_0005106 \n",
      "\tarm span http://www.ebi.ac.uk/efo/EFO_0005108 \n",
      "\tenergy expenditure http://www.ebi.ac.uk/efo/EFO_0005109 \n",
      "\tamino acid measurement http://www.ebi.ac.uk/efo/EFO_0005134 \n",
      "\tc-peptide measurement http://www.ebi.ac.uk/efo/EFO_0005187 \n",
      "\tccl11 measurement http://www.ebi.ac.uk/efo/EFO_0005188 \n",
      "\trespiratory quotient http://www.ebi.ac.uk/efo/EFO_0005189 \n",
      "\turinary nitrogen measurement http://www.ebi.ac.uk/efo/EFO_0005190 \n",
      "\twaist height ratio http://www.ebi.ac.uk/efo/EFO_0005191 \n",
      "\ttestosterone measurement http://www.ebi.ac.uk/efo/EFO_0004908 testosterone levels|testosterone level\n",
      "\tinterleukin-6 measurement http://www.ebi.ac.uk/efo/EFO_0004810 il-6 levels|il-6 level|il-6 measurement\n",
      "\ttransforming growth factor beta measurement http://www.ebi.ac.uk/efo/EFO_0004818 \n",
      "\tigf-1 measurement http://www.ebi.ac.uk/efo/EFO_0004627 insulin-like growth factor 1 levels\n",
      "\tigfbp-3 measurement http://www.ebi.ac.uk/efo/EFO_0004626 insulin-like growth factor binding protein 3 levels\n",
      "\tvitamin b12 measurement http://www.ebi.ac.uk/efo/EFO_0004620 vitamin b12 levels\n",
      "\tsleep measurement http://www.ebi.ac.uk/efo/EFO_0004870 \n",
      "\thigh density lipoprotein cholesterol measurement http://www.ebi.ac.uk/efo/EFO_0004612 hdl measurement\n",
      "\tlow density lipoprotein cholesterol measurement http://www.ebi.ac.uk/efo/EFO_0004611 ldl measurement\n",
      "\tmaximal oxygen uptake measurement http://www.ebi.ac.uk/efo/EFO_0004887 maximal oxygen consumption|peak oxygen uptake|maximal oxygen uptake|maximal aerobic capacity|vo2 max\n",
      "\tcortisol secretion measurement http://www.ebi.ac.uk/efo/EFO_0004642 \n",
      "\tfasting blood insulin measurement http://www.ebi.ac.uk/efo/EFO_0004466 fasting blood insulin|fasting blood insulin level|fasting insulin-related traits\n",
      "\tfasting blood glucose measurement http://www.ebi.ac.uk/efo/EFO_0004465 fasting glucose-related traits|fasting blood glucose level|fasting plasma glucose\n",
      "\tinsulin sensitivity measurement http://www.ebi.ac.uk/efo/EFO_0004471 insulin sensitivity test\n",
      "\ttumor necrosis factor-alpha measurement http://www.ebi.ac.uk/efo/EFO_0004684 \n",
      "\tc-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458 c-reactive protein level\n",
      "\testradiol measurement http://www.ebi.ac.uk/efo/EFO_0004697 oestrogen level|oestrogen levels|estradiol levels\n",
      "\tsystolic blood pressure http://www.ebi.ac.uk/efo/EFO_0006335 sysbp|systolic pressure\n",
      "\tdiastolic blood pressure http://www.ebi.ac.uk/efo/EFO_0006336 diastolic pressure|diabp\n",
      "\tleptin measurement http://www.ebi.ac.uk/efo/EFO_0005000 \n",
      "\thip circumference http://www.ebi.ac.uk/efo/EFO_0005093 \n",
      "\tenergy intake http://www.ebi.ac.uk/efo/EFO_0003939 caloric intake\n",
      "\tphysical activity http://www.ebi.ac.uk/efo/EFO_0003940 motor activity|activities\n",
      "\tbone density http://www.ebi.ac.uk/efo/EFO_0003923 density\n",
      "obesity-related traits 19254\n",
      "\tadiponectin measurement http://www.ebi.ac.uk/efo/EFO_0004502 adiponectin levels\n",
      "\thoma-ir http://www.ebi.ac.uk/efo/EFO_0004501 homeostasis model assessment insulin resistance\n",
      "\tccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749 monocyte chemotactic protein-1 measurement|chemokine ccl2 level|monocyte chemoattractant protein-1 measurement|chemokine ccl2 levels|monocyte chemotactic protein-1 levels|monocyte chemotactic protein-1 level\n",
      "\tthyroid stimulating hormone measurement http://www.ebi.ac.uk/efo/EFO_0004748 thyrotropin measurement|thyroid stimulating hormone level|thyroid stimulating hormone levels\n",
      "\tprotein measurement http://www.ebi.ac.uk/efo/EFO_0004747 protein levels\n",
      "\tblood urea nitrogen measurement http://www.ebi.ac.uk/efo/EFO_0004741 \n",
      "\tlean body mass http://www.ebi.ac.uk/efo/EFO_0004995 \n",
      "\tserum alanine aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004735 alanine transaminase measurement|alanine aminotransferase measurement\n",
      "\taspartate aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004736 aspartate transaminase measurement\n",
      "\thormone measurement http://www.ebi.ac.uk/efo/EFO_0004730 \n",
      "\tfat body mass http://www.ebi.ac.uk/efo/EFO_0005409 total body fat measurement\n",
      "\tadipose tissue measurement http://www.ebi.ac.uk/efo/EFO_0004764 \n",
      "\ticam-1 measurement http://www.ebi.ac.uk/efo/EFO_0004520 soluble icam1 measurement|soluble intercellular adhesion molecule-1|sicam-1\n",
      "\ttriglyceride measurement http://www.ebi.ac.uk/efo/EFO_0004530 triglyceride levels\n",
      "\twaist circumference http://www.ebi.ac.uk/efo/EFO_0004342 \n",
      "\tbody mass index http://www.ebi.ac.uk/efo/EFO_0004340 quetelet's index|bmi\n",
      "\tobesity http://www.ebi.ac.uk/efo/EFO_0001073 overweight and obesity|obesity (disorder)|obese|obesity nos|obese (finding)|adiposity|adiposis|obesity [ambiguous]|obesity\n",
      "\tbirth weight http://www.ebi.ac.uk/efo/EFO_0004344 birthweight\n",
      "\twaist-hip ratio http://www.ebi.ac.uk/efo/EFO_0004343 \n",
      "\theart rate http://www.ebi.ac.uk/efo/EFO_0004326 pulse rate|cardiac chronotropy\n",
      "\tbody height http://www.ebi.ac.uk/efo/EFO_0004339 height\n",
      "\tbody weight http://www.ebi.ac.uk/efo/EFO_0004338 weight\n",
      "\thomocysteine measurement http://www.ebi.ac.uk/efo/EFO_0004578 homocysteine levels\n",
      "\ttotal cholesterol measurement http://www.ebi.ac.uk/efo/EFO_0004574 cholesterol\n",
      "\tccl5 measurement http://www.ebi.ac.uk/efo/EFO_0005117 regulated on activation\n",
      "\tigfbp-1 measurement http://www.ebi.ac.uk/efo/EFO_0005118 \n",
      "\tantioxidant measurement http://www.ebi.ac.uk/efo/EFO_0005119 \n",
      "\thead circumference http://www.ebi.ac.uk/efo/EFO_0005114 \n",
      "\tmetabolic rate measurement http://www.ebi.ac.uk/efo/EFO_0005115 \n",
      "\turinary metabolite measurement http://www.ebi.ac.uk/efo/EFO_0005116 \n",
      "\tfatty acid measurement http://www.ebi.ac.uk/efo/EFO_0005110 \n",
      "\tfolic acid measurement http://www.ebi.ac.uk/efo/EFO_0005111 folate measurement|vitamin b9 measurement\n",
      "\tgestational age http://www.ebi.ac.uk/efo/EFO_0005112 \n",
      "\tbody composition measurement http://www.ebi.ac.uk/efo/EFO_0005106 \n",
      "\tarm span http://www.ebi.ac.uk/efo/EFO_0005108 \n",
      "\tenergy expenditure http://www.ebi.ac.uk/efo/EFO_0005109 \n",
      "\tamino acid measurement http://www.ebi.ac.uk/efo/EFO_0005134 \n",
      "\tc-peptide measurement http://www.ebi.ac.uk/efo/EFO_0005187 \n",
      "\tccl11 measurement http://www.ebi.ac.uk/efo/EFO_0005188 \n",
      "\trespiratory quotient http://www.ebi.ac.uk/efo/EFO_0005189 \n",
      "\turinary nitrogen measurement http://www.ebi.ac.uk/efo/EFO_0005190 \n",
      "\twaist height ratio http://www.ebi.ac.uk/efo/EFO_0005191 \n",
      "\ttestosterone measurement http://www.ebi.ac.uk/efo/EFO_0004908 testosterone levels|testosterone level\n",
      "\tinterleukin-6 measurement http://www.ebi.ac.uk/efo/EFO_0004810 il-6 levels|il-6 level|il-6 measurement\n",
      "\ttransforming growth factor beta measurement http://www.ebi.ac.uk/efo/EFO_0004818 \n",
      "\tigf-1 measurement http://www.ebi.ac.uk/efo/EFO_0004627 insulin-like growth factor 1 levels\n",
      "\tigfbp-3 measurement http://www.ebi.ac.uk/efo/EFO_0004626 insulin-like growth factor binding protein 3 levels\n",
      "\tvitamin b12 measurement http://www.ebi.ac.uk/efo/EFO_0004620 vitamin b12 levels\n",
      "\tsleep measurement http://www.ebi.ac.uk/efo/EFO_0004870 \n",
      "\thigh density lipoprotein cholesterol measurement http://www.ebi.ac.uk/efo/EFO_0004612 hdl measurement\n",
      "\tlow density lipoprotein cholesterol measurement http://www.ebi.ac.uk/efo/EFO_0004611 ldl measurement\n",
      "\tmaximal oxygen uptake measurement http://www.ebi.ac.uk/efo/EFO_0004887 maximal oxygen consumption|peak oxygen uptake|maximal oxygen uptake|maximal aerobic capacity|vo2 max\n",
      "\tcortisol secretion measurement http://www.ebi.ac.uk/efo/EFO_0004642 \n",
      "\tfasting blood insulin measurement http://www.ebi.ac.uk/efo/EFO_0004466 fasting blood insulin|fasting blood insulin level|fasting insulin-related traits\n",
      "\tfasting blood glucose measurement http://www.ebi.ac.uk/efo/EFO_0004465 fasting glucose-related traits|fasting blood glucose level|fasting plasma glucose\n",
      "\tinsulin sensitivity measurement http://www.ebi.ac.uk/efo/EFO_0004471 insulin sensitivity test\n",
      "\ttumor necrosis factor-alpha measurement http://www.ebi.ac.uk/efo/EFO_0004684 \n",
      "\tc-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458 c-reactive protein level\n",
      "\testradiol measurement http://www.ebi.ac.uk/efo/EFO_0004697 oestrogen level|oestrogen levels|estradiol levels\n",
      "\tsystolic blood pressure http://www.ebi.ac.uk/efo/EFO_0006335 sysbp|systolic pressure\n",
      "\tdiastolic blood pressure http://www.ebi.ac.uk/efo/EFO_0006336 diastolic pressure|diabp\n",
      "\tleptin measurement http://www.ebi.ac.uk/efo/EFO_0005000 \n",
      "\thip circumference http://www.ebi.ac.uk/efo/EFO_0005093 \n",
      "\tenergy intake http://www.ebi.ac.uk/efo/EFO_0003939 caloric intake\n",
      "\tphysical activity http://www.ebi.ac.uk/efo/EFO_0003940 motor activity|activities\n",
      "\tbone density http://www.ebi.ac.uk/efo/EFO_0003923 density\n",
      "obesity-related traits 19255\n",
      "\tadiponectin measurement http://www.ebi.ac.uk/efo/EFO_0004502 adiponectin levels\n",
      "\thoma-ir http://www.ebi.ac.uk/efo/EFO_0004501 homeostasis model assessment insulin resistance\n",
      "\tccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749 monocyte chemotactic protein-1 measurement|chemokine ccl2 level|monocyte chemoattractant protein-1 measurement|chemokine ccl2 levels|monocyte chemotactic protein-1 levels|monocyte chemotactic protein-1 level\n",
      "\tthyroid stimulating hormone measurement http://www.ebi.ac.uk/efo/EFO_0004748 thyrotropin measurement|thyroid stimulating hormone level|thyroid stimulating hormone levels\n",
      "\tprotein measurement http://www.ebi.ac.uk/efo/EFO_0004747 protein levels\n",
      "\tblood urea nitrogen measurement http://www.ebi.ac.uk/efo/EFO_0004741 \n",
      "\tlean body mass http://www.ebi.ac.uk/efo/EFO_0004995 \n",
      "\tserum alanine aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004735 alanine transaminase measurement|alanine aminotransferase measurement\n",
      "\taspartate aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004736 aspartate transaminase measurement\n",
      "\thormone measurement http://www.ebi.ac.uk/efo/EFO_0004730 \n",
      "\tfat body mass http://www.ebi.ac.uk/efo/EFO_0005409 total body fat measurement\n",
      "\tadipose tissue measurement http://www.ebi.ac.uk/efo/EFO_0004764 \n",
      "\ticam-1 measurement http://www.ebi.ac.uk/efo/EFO_0004520 soluble icam1 measurement|soluble intercellular adhesion molecule-1|sicam-1\n",
      "\ttriglyceride measurement http://www.ebi.ac.uk/efo/EFO_0004530 triglyceride levels\n",
      "\twaist circumference http://www.ebi.ac.uk/efo/EFO_0004342 \n",
      "\tbody mass index http://www.ebi.ac.uk/efo/EFO_0004340 quetelet's index|bmi\n",
      "\tobesity http://www.ebi.ac.uk/efo/EFO_0001073 overweight and obesity|obesity (disorder)|obese|obesity nos|obese (finding)|adiposity|adiposis|obesity [ambiguous]|obesity\n",
      "\tbirth weight http://www.ebi.ac.uk/efo/EFO_0004344 birthweight\n",
      "\twaist-hip ratio http://www.ebi.ac.uk/efo/EFO_0004343 \n",
      "\theart rate http://www.ebi.ac.uk/efo/EFO_0004326 pulse rate|cardiac chronotropy\n",
      "\tbody height http://www.ebi.ac.uk/efo/EFO_0004339 height\n",
      "\tbody weight http://www.ebi.ac.uk/efo/EFO_0004338 weight\n",
      "\thomocysteine measurement http://www.ebi.ac.uk/efo/EFO_0004578 homocysteine levels\n",
      "\ttotal cholesterol measurement http://www.ebi.ac.uk/efo/EFO_0004574 cholesterol\n",
      "\tccl5 measurement http://www.ebi.ac.uk/efo/EFO_0005117 regulated on activation\n",
      "\tigfbp-1 measurement http://www.ebi.ac.uk/efo/EFO_0005118 \n",
      "\tantioxidant measurement http://www.ebi.ac.uk/efo/EFO_0005119 \n",
      "\thead circumference http://www.ebi.ac.uk/efo/EFO_0005114 \n",
      "\tmetabolic rate measurement http://www.ebi.ac.uk/efo/EFO_0005115 \n",
      "\turinary metabolite measurement http://www.ebi.ac.uk/efo/EFO_0005116 \n",
      "\tfatty acid measurement http://www.ebi.ac.uk/efo/EFO_0005110 \n",
      "\tfolic acid measurement http://www.ebi.ac.uk/efo/EFO_0005111 folate measurement|vitamin b9 measurement\n",
      "\tgestational age http://www.ebi.ac.uk/efo/EFO_0005112 \n",
      "\tbody composition measurement http://www.ebi.ac.uk/efo/EFO_0005106 \n",
      "\tarm span http://www.ebi.ac.uk/efo/EFO_0005108 \n",
      "\tenergy expenditure http://www.ebi.ac.uk/efo/EFO_0005109 \n",
      "\tamino acid measurement http://www.ebi.ac.uk/efo/EFO_0005134 \n",
      "\tc-peptide measurement http://www.ebi.ac.uk/efo/EFO_0005187 \n",
      "\tccl11 measurement http://www.ebi.ac.uk/efo/EFO_0005188 \n",
      "\trespiratory quotient http://www.ebi.ac.uk/efo/EFO_0005189 \n",
      "\turinary nitrogen measurement http://www.ebi.ac.uk/efo/EFO_0005190 \n",
      "\twaist height ratio http://www.ebi.ac.uk/efo/EFO_0005191 \n",
      "\ttestosterone measurement http://www.ebi.ac.uk/efo/EFO_0004908 testosterone levels|testosterone level\n",
      "\tinterleukin-6 measurement http://www.ebi.ac.uk/efo/EFO_0004810 il-6 levels|il-6 level|il-6 measurement\n",
      "\ttransforming growth factor beta measurement http://www.ebi.ac.uk/efo/EFO_0004818 \n",
      "\tigf-1 measurement http://www.ebi.ac.uk/efo/EFO_0004627 insulin-like growth factor 1 levels\n",
      "\tigfbp-3 measurement http://www.ebi.ac.uk/efo/EFO_0004626 insulin-like growth factor binding protein 3 levels\n",
      "\tvitamin b12 measurement http://www.ebi.ac.uk/efo/EFO_0004620 vitamin b12 levels\n",
      "\tsleep measurement http://www.ebi.ac.uk/efo/EFO_0004870 \n",
      "\thigh density lipoprotein cholesterol measurement http://www.ebi.ac.uk/efo/EFO_0004612 hdl measurement\n",
      "\tlow density lipoprotein cholesterol measurement http://www.ebi.ac.uk/efo/EFO_0004611 ldl measurement\n",
      "\tmaximal oxygen uptake measurement http://www.ebi.ac.uk/efo/EFO_0004887 maximal oxygen consumption|peak oxygen uptake|maximal oxygen uptake|maximal aerobic capacity|vo2 max\n",
      "\tcortisol secretion measurement http://www.ebi.ac.uk/efo/EFO_0004642 \n",
      "\tfasting blood insulin measurement http://www.ebi.ac.uk/efo/EFO_0004466 fasting blood insulin|fasting blood insulin level|fasting insulin-related traits\n",
      "\tfasting blood glucose measurement http://www.ebi.ac.uk/efo/EFO_0004465 fasting glucose-related traits|fasting blood glucose level|fasting plasma glucose\n",
      "\tinsulin sensitivity measurement http://www.ebi.ac.uk/efo/EFO_0004471 insulin sensitivity test\n",
      "\ttumor necrosis factor-alpha measurement http://www.ebi.ac.uk/efo/EFO_0004684 \n",
      "\tc-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458 c-reactive protein level\n",
      "\testradiol measurement http://www.ebi.ac.uk/efo/EFO_0004697 oestrogen level|oestrogen levels|estradiol levels\n",
      "\tsystolic blood pressure http://www.ebi.ac.uk/efo/EFO_0006335 sysbp|systolic pressure\n",
      "\tdiastolic blood pressure http://www.ebi.ac.uk/efo/EFO_0006336 diastolic pressure|diabp\n",
      "\tleptin measurement http://www.ebi.ac.uk/efo/EFO_0005000 \n",
      "\thip circumference http://www.ebi.ac.uk/efo/EFO_0005093 \n",
      "\tenergy intake http://www.ebi.ac.uk/efo/EFO_0003939 caloric intake\n",
      "\tphysical activity http://www.ebi.ac.uk/efo/EFO_0003940 motor activity|activities\n",
      "\tbone density http://www.ebi.ac.uk/efo/EFO_0003923 density\n",
      "Genome-Wide Association Scan Shows Genetic Variants in the  FTO  Gene Are Associated with Obesity-Related Traits. The obesity epidemic is responsible for a substantial economic burden in developed countries and is a major risk factor for type 2 diabetes and cardiovascular disease. The disease is the result not only of several environmental risk factors, but also of genetic predisposition. To take advantage of recent advances in gene-mapping technology, we executed a genome-wide association scan to identify genetic variants associated with obesity-related quantitative traits in the genetically isolated population of Sardinia. Initial analysis suggested that several SNPs in the  FTO  and  PFKP  genes were associated with increased BMI, hip circumference, and weight. Within the  FTO  gene, rs9930506 showed the strongest association with BMI ( p =  8.6 ×10 − 7 ), hip circumference ( p =  3.4 × 10 − 8 ), and weight ( p =  9.1 × 10 − 7 ). In Sardinia, homozygotes for the rare “G” allele of this SNP (minor allele frequency = 0.46) were 1.3 BMI units heavier than homozygotes for the common “A” allele. Within the PFKP gene, rs6602024 showed very strong association with BMI ( p =  4.9 × 10 − 6 ). Homozygotes for the rare “A” allele of this SNP (minor allele frequency = 0.12) were 1.8 BMI units heavier than homozygotes for the common “G” allele. To replicate our findings, we genotyped these two SNPs in the GenNet study. In European Americans ( N  = 1,496) and in Hispanic Americans ( N  = 839), we replicated significant association between rs9930506 in the FTO gene and BMI ( p -value for meta-analysis of European American and Hispanic American follow-up samples,  p =  0.001), weight ( p =  0.001), and hip circumference ( p =  0.0005). We did not replicate association between rs6602024 and obesity-related traits in the GenNet sample, although we found that in European Americans, Hispanic Americans, and African Americans, homozygotes for the rare “A” allele were, on average, 1.0–3.0 BMI units heavier than homozygotes for the more common “G” allele. In summary, we have completed a whole genome–association scan for three obesity-related quantitative traits and report that common genetic variants in the  FTO  gene are associated with substantial changes in BMI, hip circumference, and body weight. These changes could have a significant impact on the risk of obesity-related morbidity in the general population. Although twin and family studies have clearly shown that genes play a role in obesity, it has proven quite difficult to identify the specific genetic variants involved. Here, we take advantage of recent technical and methodological advances to examine the role of common genetic variants on several obesity-related traits. By examining >4,000 Sardinians, we show that a specific genetic variant, rs9930506, and other nearby variants on human Chromosome 16 are associated with body mass index, hip circumference, and total body weight. The variants overlap  FTO,  a gene with poorly understood function. Further studies of the region may implicate new biological pathways affecting susceptibility to obesity. We also show that the association is not restricted to Sardinia but is also seen in independent samples of European Americans and Hispanic Americans. This finding is particularly important because obesity is associated with increased risk of cardiovascular disease and diabetes.\n",
      "-1 Associated with\n",
      "1 Obesity-Related Traits\n",
      "-1 Association\n",
      "-1 Genetic\n",
      "-1 in\n",
      "-1 major risk factor\n",
      "-1 type 2 diabetes\n",
      "-1 a major\n",
      "-1 cardiovascular disease | cardiovascular disease http://www.ebi.ac.uk/efo/EFO_0000319\n",
      "1 obesity | obesity http://www.ebi.ac.uk/efo/EFO_0001073\n",
      "-1 epidemic\n",
      "-1 responsible\n",
      "-1 a\n",
      "-1 in\n",
      "-1 developed\n",
      "-1 and\n",
      "-1 and\n",
      "-1 risk factors,\n",
      "-1 The disease\n",
      "-1 , but\n",
      "-1 of genetic\n",
      "-1 not\n",
      "-1 of\n",
      "-1 several\n",
      "-1 predisposition\n",
      "-1 obesity-related quantitative traits\n",
      "-1 associated with\n",
      "-1 of\n",
      "-1 in\n",
      "-1 a\n",
      "-1 association\n",
      "-1 genetic\n",
      "-1 in\n",
      "-1 genetically\n",
      "-1 population\n",
      "-1 of\n",
      "1 , hip circumference | hip circumference http://www.ebi.ac.uk/efo/EFO_0005093\n",
      "1 hip circumference, | hip circumference http://www.ebi.ac.uk/efo/EFO_0005093\n",
      "-1 associated with\n",
      "1 BMI, | body mass index http://www.ebi.ac.uk/efo/EFO_0004340\n",
      "-1 , and\n",
      "-1 analysis\n",
      "-1 suggested\n",
      "-1 several\n",
      "-1 in\n",
      "-1 and\n",
      "-1 increased\n",
      "1 weight | body weight http://www.ebi.ac.uk/efo/EFO_0004338\n",
      "1 ), hip circumference ( | hip circumference http://www.ebi.ac.uk/efo/EFO_0005093\n",
      "-1 association with\n",
      "-1 , and\n",
      "1 BMI | body mass index http://www.ebi.ac.uk/efo/EFO_0004340\n",
      "-1 p\n",
      "-1 ×\n",
      "-1 p\n",
      "-1 ×\n",
      "1 weight | body weight http://www.ebi.ac.uk/efo/EFO_0004338\n",
      "-1 p\n",
      "-1 ×\n",
      "-1 In\n",
      "-1 of\n",
      "1 BMI | body mass index http://www.ebi.ac.uk/efo/EFO_0004340\n",
      "-1 A\n",
      "-1 association with\n",
      "1 BMI | body mass index http://www.ebi.ac.uk/efo/EFO_0004340\n",
      "-1 p\n",
      "-1 ×\n",
      "-1 A\n",
      "-1 of\n",
      "1 BMI | body mass index http://www.ebi.ac.uk/efo/EFO_0004340\n",
      "-1 findings,\n",
      "-1 genotyped\n",
      "-1 in\n",
      "1 ), weight ( | body weight http://www.ebi.ac.uk/efo/EFO_0004338\n",
      "-1 p -\n",
      "-1 -value\n",
      "-1 ,  p\n",
      "-1 , and\n",
      "1 hip circumference | hip circumference http://www.ebi.ac.uk/efo/EFO_0005093\n",
      "-1 In\n",
      "-1 N\n",
      "-1 and\n",
      "-1 in\n",
      "-1 N\n",
      "-1 association\n",
      "-1 in\n",
      "-1 and\n",
      "1 BMI | body mass index http://www.ebi.ac.uk/efo/EFO_0004340\n",
      "-1 of\n",
      "-1 and\n",
      "-1 follow-up\n",
      "-1 p\n",
      "-1 p\n",
      "1 obesity-related traits\n",
      "-1 , and\n",
      "-1 not\n",
      "-1 association\n",
      "-1 and\n",
      "-1 in\n",
      "-1 in\n",
      "-1 A\n",
      "1 BMI | body mass index http://www.ebi.ac.uk/efo/EFO_0004340\n",
      "-1 obesity-related quantitative traits\n",
      "1 , hip circumference | hip circumference http://www.ebi.ac.uk/efo/EFO_0005093\n",
      "1 hip circumference, | hip circumference http://www.ebi.ac.uk/efo/EFO_0005093\n",
      "-1 associated with\n",
      "1 BMI, | body mass index http://www.ebi.ac.uk/efo/EFO_0004340\n",
      "-1 , and\n",
      "1 body weight | body weight http://www.ebi.ac.uk/efo/EFO_0004338\n",
      "-1 In\n",
      "-1 a\n",
      "-1 genome\n",
      "-1 association\n",
      "-1 and\n",
      "-1 genetic\n",
      "-1 in\n",
      "-1 changes\n",
      "-1 in\n",
      "-1 risk of\n",
      "-1 changes\n",
      "-1 a\n",
      "-1 in\n",
      "-1 general\n",
      "-1 population\n",
      "1 obesity, | obesity http://www.ebi.ac.uk/efo/EFO_0001073\n",
      "-1 and\n",
      "-1 family\n",
      "-1 play\n",
      "-1 a\n",
      "-1 in\n",
      "-1 specific\n",
      "-1 genetic\n",
      "1 obesity-related traits\n",
      "-1 of\n",
      "-1 and\n",
      "-1 of\n",
      "-1 genetic\n",
      "-1 several\n",
      "1 body mass index, | body mass index http://www.ebi.ac.uk/efo/EFO_0004340\n",
      "1 , hip circumference | hip circumference http://www.ebi.ac.uk/efo/EFO_0005093\n",
      "1 hip circumference, | hip circumference http://www.ebi.ac.uk/efo/EFO_0005093\n",
      "-1 , and\n",
      "-1 Chromosome 16\n",
      "-1 associated with\n",
      "-1 , and\n",
      "1 body weight | body weight http://www.ebi.ac.uk/efo/EFO_0004338\n",
      "-1 a\n",
      "-1 specific\n",
      "-1 genetic\n",
      "-1 human\n",
      "-1 ,  a\n",
      "-1 of the\n",
      "-1 susceptibility to\n",
      "1 to obesity | obesity http://www.ebi.ac.uk/efo/EFO_0001073\n",
      "-1 biological\n",
      "-1 affecting\n",
      "-1 association\n",
      "-1 not\n",
      "-1 but\n",
      "-1 in\n",
      "-1 independent\n",
      "-1 of\n",
      "-1 and\n",
      "-1 with increased risk of\n",
      "-1 of cardiovascular disease\n",
      "-1 associated with\n",
      "-1 disease and\n",
      "-1 finding\n",
      "1 obesity | obesity http://www.ebi.ac.uk/efo/EFO_0001073\n",
      "-1 diabetes\n",
      "\n",
      "17684544\n",
      "crohn's disease 19532\n",
      "\tcrohn's disease http://www.ebi.ac.uk/efo/EFO_0000384 gastritis associated with crohn's disease|ileitis\n",
      "crohn's disease 19845\n",
      "\tcrohn's disease http://www.ebi.ac.uk/efo/EFO_0000384 gastritis associated with crohn's disease|ileitis\n",
      "crohn's disease 19846\n",
      "\tcrohn's disease http://www.ebi.ac.uk/efo/EFO_0000384 gastritis associated with crohn's disease|ileitis\n",
      "Systematic Association Mapping Identifies  NELL1  as a Novel IBD Disease Gene. Crohn disease (CD), a sub-entity of inflammatory bowel disease (IBD), is a complex polygenic disorder. Although recent studies have successfully identified CD-associated genetic variants, these susceptibility loci explain only a fraction of the heritability of the disease. Here, we report on a multi-stage genome-wide scan of 393 German CD cases and 399 controls. Among the 116,161 single-nucleotide polymorphisms tested, an association with the known CD susceptibility gene  NOD2 , the  5q31  haplotype, and the recently reported CD locus at  5p13.1  was confirmed. In addition, SNP rs1793004 in the gene encoding nel-like 1 precursor ( NELL1,  chromosome  11p15.1 ) showed a consistent disease-association in independent German population- and family-based samples (942 cases, 1082 controls, 375 trios). Subsequent fine mapping and replication in an independent sample of 454 French/Canadian CD trios supported the authenticity of the  NELL1  association. Further confirmation in a large German ulcerative colitis (UC) sample indicated that  NELL1  is a ubiquitous IBD susceptibility locus (combined p<10 −6 ; OR = 1.66, 95% CI: 1.30–2.11). The novel  5p13.1  locus was also replicated in the French/Canadian sample and in an independent UK CD patient panel (453 cases, 521 controls, combined p<10 −6  for SNP rs1992660). Several associations were replicated in at least one independent sample, point to an involvement of  ITGB6  (upstream),  GRM8  (downstream),  OR5V1  (downstream),  PPP3R2  (downstream),  NM_152575  (upstream) and  HNF4G  (intron).\n",
      "-1 as a\n",
      "-1 IBD Disease\n",
      "-1 Association\n",
      "-1 Mapping | mean arterial pressure http://www.ebi.ac.uk/efo/EFO_0006340\n",
      "-1 inflammatory bowel disease (IBD),\n",
      "-1 Crohn disease (CD),\n",
      "-1 , a\n",
      "-1 of\n",
      "-1 a\n",
      "-1 of the disease\n",
      "-1 of the\n",
      "-1 genetic\n",
      "-1 susceptibility\n",
      "-1 a\n",
      "-1 a\n",
      "-1 of\n",
      "-1 and\n",
      "-1 , an\n",
      "-1 association with\n",
      "-1 , and\n",
      "-1 susceptibility\n",
      "-1 was\n",
      "-1 confirmed\n",
      "-1 ,  chromosome\n",
      "-1 population-\n",
      "-1 - and\n",
      "-1 In\n",
      "-1 in\n",
      "-1 a\n",
      "-1 disease-association\n",
      "-1 in\n",
      "-1 independent\n",
      "-1 of the\n",
      "-1 mapping | mean arterial pressure http://www.ebi.ac.uk/efo/EFO_0006340\n",
      "-1 and\n",
      "-1 in\n",
      "-1 an\n",
      "-1 independent\n",
      "-1 of\n",
      "-1 association\n",
      "-1 ulcerative colitis (UC)\n",
      "-1 confirmation\n",
      "-1 in\n",
      "-1 a\n",
      "-1 indicated | ankle brachial index http://www.ebi.ac.uk/efo/EFO_0003912\n",
      "-1 a\n",
      "-1 IBD\n",
      "-1 susceptibility\n",
      "-1 p\n",
      "-1 was\n",
      "-1 in\n",
      "-1 and\n",
      "-1 in\n",
      "-1 an\n",
      "-1 independent\n",
      "-1 p\n",
      "-1 in at\n",
      "-1 to an\n",
      "-1 Several\n",
      "-1 associations\n",
      "-1 independent\n",
      "-1 of\n",
      "-1 and\n",
      "\n",
      "17903292\n",
      "cystatin c 19169\n",
      "\tcystatin c measurement http://www.ebi.ac.uk/efo/EFO_0004617 \n",
      "thyroid stimulating hormone 19170\n",
      "\tthyroid stimulating hormone measurement http://www.ebi.ac.uk/efo/EFO_0004748 thyrotropin measurement|thyroid stimulating hormone level|thyroid stimulating hormone levels\n",
      "\tthyroid function http://www.ebi.ac.uk/efo/EFO_0004296 \n",
      "thyroid stimulating hormone 19171\n",
      "\tthyroid stimulating hormone measurement http://www.ebi.ac.uk/efo/EFO_0004748 thyrotropin measurement|thyroid stimulating hormone level|thyroid stimulating hormone levels\n",
      "\tthyroid function http://www.ebi.ac.uk/efo/EFO_0004296 \n",
      "thyroid stimulating hormone 19172\n",
      "\tthyroid stimulating hormone measurement http://www.ebi.ac.uk/efo/EFO_0004748 thyrotropin measurement|thyroid stimulating hormone level|thyroid stimulating hormone levels\n",
      "\tthyroid function http://www.ebi.ac.uk/efo/EFO_0004296 \n",
      "urinary albumin excretion 19173\n",
      "\talbuminuria http://www.ebi.ac.uk/efo/EFO_0004285 urinary albumin excretion\n",
      "urinary albumin excretion 19626\n",
      "\talbuminuria http://www.ebi.ac.uk/efo/EFO_0004285 urinary albumin excretion\n",
      "A genome-wide association for kidney function and endocrine-related traits in the NHLBI's Framingham Heart Study. Glomerular filtration rate (GFR) and urinary albumin excretion (UAE) are markers of kidney function that are known to be heritable. Many endocrine conditions have strong familial components. We tested for association between the Affymetrix GeneChip Human Mapping 100K single nucleotide polymorphism (SNP) set and measures of kidney function and endocrine traits. Genotype information on the Affymetrix GeneChip Human Mapping 100K SNP set was available on 1345 participants. Serum creatinine and cystatin-C (cysC; n = 981) were measured at the seventh examination cycle (1998–2001); GFR (n = 1010) was estimated via the Modification of Diet in Renal Disease (MDRD) equation; UAE was measured on spot urine samples during the sixth examination cycle (1995–1998) and was indexed to urinary creatinine (n = 822). Thyroid stimulating hormone (TSH) was measured at the third and fourth examination cycles (1981–1984; 1984–1987) and mean value of the measurements were used (n = 810). Age-sex-adjusted and multivariable-adjusted residuals for these measurements were used in association with genotype data using generalized estimating equations (GEE) and family-based association tests (FBAT) models. We presented the results for association tests using additive allele model. We evaluated associations with 70,987 SNPs on autosomes with minor allele frequencies of at least 0.10, Hardy-Weinberg Equilibrium p-value ≥ 0.001, and call rates of at least 80%. The top SNPs associated with these traits using the GEE method were rs2839235 with GFR (p-value 1.6*10 -05 ), rs1158167 with cysC (p-value 8.5*10 -09 ), rs1712790 with UAE (p-value 1.9*10 -06 ), and rs6977660 with TSH (p-value 3.7*10 -06 ), respectively. The top SNPs associated with these traits using the FBAT method were rs6434804 with GFR(p-value 2.4*10 -5 ), rs563754 with cysC (p-value 4.7*10 -5 ), rs1243400 with UAE (p-value 4.8*10 -6 ), and rs4128956 with TSH (p-value 3.6*10 -5 ), respectively. Detailed association test results can be found at  . Four SNPs in or near the  CST 3 gene were highly associated with cysC levels (p-value 8.5*10 -09  to 0.007). Kidney function traits and TSH are associated with SNPs on the Affymetrix GeneChip Human Mapping 100K SNP set. These data will serve as a valuable resource for replication as more SNPs associated with kidney function and endocrine traits are identified.\n",
      "-1 endocrine-related traits\n",
      "-1 A\n",
      "-1 association\n",
      "-1 kidney\n",
      "-1 and\n",
      "-1 in\n",
      "-1 Heart\n",
      "-1 Glomerular filtration rate | glomerular filtration rate http://www.ebi.ac.uk/efo/EFO_0005208\n",
      "1 urinary albumin excretion | albuminuria http://www.ebi.ac.uk/efo/EFO_0004285\n",
      "-1 of kidney\n",
      "-1 GFR | glomerular filtration rate http://www.ebi.ac.uk/efo/EFO_0005208\n",
      "-1 and\n",
      "-1 endocrine\n",
      "-1 familial\n",
      "-1 of kidney\n",
      "-1 endocrine traits\n",
      "-1 association\n",
      "-1 Human\n",
      "-1 Mapping | mean arterial pressure http://www.ebi.ac.uk/efo/EFO_0006340\n",
      "-1 single | herpes zoster http://www.ebi.ac.uk/efo/EFO_0006510\n",
      "-1 and\n",
      "-1 and\n",
      "-1 Genotype\n",
      "-1 Human\n",
      "-1 Mapping | mean arterial pressure http://www.ebi.ac.uk/efo/EFO_0006340\n",
      "-1 was\n",
      "-1 Serum creatinine | serum creatinine measurement http://www.ebi.ac.uk/efo/EFO_0004518\n",
      "-1 Renal Disease\n",
      "-1 and\n",
      "1 cystatin-C | cystatin c measurement http://www.ebi.ac.uk/efo/EFO_0004617\n",
      "-1 n\n",
      "-1 GFR | glomerular filtration rate http://www.ebi.ac.uk/efo/EFO_0005208\n",
      "-1 n\n",
      "-1 was\n",
      "-1 of\n",
      "-1 in\n",
      "-1 was\n",
      "-1 spot\n",
      "-1 urine\n",
      "-1 sixth\n",
      "-1 and\n",
      "-1 was\n",
      "-1 creatinine | serum creatinine measurement http://www.ebi.ac.uk/efo/EFO_0004518\n",
      "-1 n\n",
      "1 Thyroid stimulating hormone | thyroid stimulating hormone measurement http://www.ebi.ac.uk/efo/EFO_0004748\n",
      "-1 of the\n",
      "-1 the measurements\n",
      "-1 TSH\n",
      "-1 was\n",
      "-1 and\n",
      "-1 fourth\n",
      "-1 and\n",
      "-1 value\n",
      "-1 n\n",
      "-1 these measurements\n",
      "-1 association with\n",
      "-1 and\n",
      "-1 in\n",
      "-1 genotype\n",
      "-1 generalized\n",
      "-1 and\n",
      "-1 association\n",
      "-1 association\n",
      "-1 associations with\n",
      "-1 of at\n",
      "-1 , and\n",
      "-1 of at\n",
      "-1 rates\n",
      "-1 associated with\n",
      "-1 these traits\n",
      "-1 , and\n",
      "-1 GFR | glomerular filtration rate http://www.ebi.ac.uk/efo/EFO_0005208\n",
      "-1 TSH\n",
      "-1 associated with\n",
      "-1 these traits\n",
      "-1 , and\n",
      "-1 GFR | glomerular filtration rate http://www.ebi.ac.uk/efo/EFO_0005208\n",
      "-1 -5\n",
      "-1 -5\n",
      "-1 TSH\n",
      "-1 -5\n",
      "-1 association test\n",
      "-1 associated with\n",
      "-1 in\n",
      "-1 near\n",
      "-1 function traits\n",
      "-1 associated with\n",
      "-1 Kidney\n",
      "-1 and\n",
      "-1 TSH\n",
      "-1 Human\n",
      "-1 Mapping | mean arterial pressure http://www.ebi.ac.uk/efo/EFO_0006340\n",
      "-1 as a\n",
      "-1 associated with\n",
      "-1 endocrine traits\n",
      "-1 will\n",
      "-1 as\n",
      "-1 kidney\n",
      "-1 and\n",
      "\n",
      "17903293\n",
      "select biomarker traits 19638\n",
      "\tccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749 monocyte chemotactic protein-1 measurement|chemokine ccl2 level|monocyte chemoattractant protein-1 measurement|chemokine ccl2 levels|monocyte chemotactic protein-1 levels|monocyte chemotactic protein-1 level\n",
      "\tserum alanine aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004735 alanine transaminase measurement|alanine aminotransferase measurement\n",
      "\tatrial natriuretic factor measurement http://www.ebi.ac.uk/efo/EFO_0004789 \n",
      "\tcd40 ligand measurement http://www.ebi.ac.uk/efo/EFO_0004790 cd154 ligand measurement\n",
      "\talkaline phosphatase measurement http://www.ebi.ac.uk/efo/EFO_0004533 ap\n",
      "\tbilirubin measurement http://www.ebi.ac.uk/efo/EFO_0004570 bilirubin levels\n",
      "\tvitamin d measurement http://www.ebi.ac.uk/efo/EFO_0004631 \n",
      "\tvitamin k measurement http://www.ebi.ac.uk/efo/EFO_0004618 phylloquinone measurement\n",
      "\ttumor necrosis factor-alpha measurement http://www.ebi.ac.uk/efo/EFO_0004684 \n",
      "\tc-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458 c-reactive protein level\n",
      "select biomarker traits 19633\n",
      "\tccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749 monocyte chemotactic protein-1 measurement|chemokine ccl2 level|monocyte chemoattractant protein-1 measurement|chemokine ccl2 levels|monocyte chemotactic protein-1 levels|monocyte chemotactic protein-1 level\n",
      "\tserum alanine aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004735 alanine transaminase measurement|alanine aminotransferase measurement\n",
      "\tatrial natriuretic factor measurement http://www.ebi.ac.uk/efo/EFO_0004789 \n",
      "\tcd40 ligand measurement http://www.ebi.ac.uk/efo/EFO_0004790 cd154 ligand measurement\n",
      "\talkaline phosphatase measurement http://www.ebi.ac.uk/efo/EFO_0004533 ap\n",
      "\tbilirubin measurement http://www.ebi.ac.uk/efo/EFO_0004570 bilirubin levels\n",
      "\tvitamin d measurement http://www.ebi.ac.uk/efo/EFO_0004631 \n",
      "\tvitamin k measurement http://www.ebi.ac.uk/efo/EFO_0004618 phylloquinone measurement\n",
      "\ttumor necrosis factor-alpha measurement http://www.ebi.ac.uk/efo/EFO_0004684 \n",
      "\tc-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458 c-reactive protein level\n",
      "select biomarker traits 19161\n",
      "\tccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749 monocyte chemotactic protein-1 measurement|chemokine ccl2 level|monocyte chemoattractant protein-1 measurement|chemokine ccl2 levels|monocyte chemotactic protein-1 levels|monocyte chemotactic protein-1 level\n",
      "\tserum alanine aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004735 alanine transaminase measurement|alanine aminotransferase measurement\n",
      "\tatrial natriuretic factor measurement http://www.ebi.ac.uk/efo/EFO_0004789 \n",
      "\tcd40 ligand measurement http://www.ebi.ac.uk/efo/EFO_0004790 cd154 ligand measurement\n",
      "\talkaline phosphatase measurement http://www.ebi.ac.uk/efo/EFO_0004533 ap\n",
      "\tbilirubin measurement http://www.ebi.ac.uk/efo/EFO_0004570 bilirubin levels\n",
      "\tvitamin d measurement http://www.ebi.ac.uk/efo/EFO_0004631 \n",
      "\tvitamin k measurement http://www.ebi.ac.uk/efo/EFO_0004618 phylloquinone measurement\n",
      "\ttumor necrosis factor-alpha measurement http://www.ebi.ac.uk/efo/EFO_0004684 \n",
      "\tc-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458 c-reactive protein level\n",
      "select biomarker traits 19627\n",
      "\tccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749 monocyte chemotactic protein-1 measurement|chemokine ccl2 level|monocyte chemoattractant protein-1 measurement|chemokine ccl2 levels|monocyte chemotactic protein-1 levels|monocyte chemotactic protein-1 level\n",
      "\tserum alanine aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004735 alanine transaminase measurement|alanine aminotransferase measurement\n",
      "\tatrial natriuretic factor measurement http://www.ebi.ac.uk/efo/EFO_0004789 \n",
      "\tcd40 ligand measurement http://www.ebi.ac.uk/efo/EFO_0004790 cd154 ligand measurement\n",
      "\talkaline phosphatase measurement http://www.ebi.ac.uk/efo/EFO_0004533 ap\n",
      "\tbilirubin measurement http://www.ebi.ac.uk/efo/EFO_0004570 bilirubin levels\n",
      "\tvitamin d measurement http://www.ebi.ac.uk/efo/EFO_0004631 \n",
      "\tvitamin k measurement http://www.ebi.ac.uk/efo/EFO_0004618 phylloquinone measurement\n",
      "\ttumor necrosis factor-alpha measurement http://www.ebi.ac.uk/efo/EFO_0004684 \n",
      "\tc-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458 c-reactive protein level\n",
      "select biomarker traits 19628\n",
      "\tccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749 monocyte chemotactic protein-1 measurement|chemokine ccl2 level|monocyte chemoattractant protein-1 measurement|chemokine ccl2 levels|monocyte chemotactic protein-1 levels|monocyte chemotactic protein-1 level\n",
      "\tserum alanine aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004735 alanine transaminase measurement|alanine aminotransferase measurement\n",
      "\tatrial natriuretic factor measurement http://www.ebi.ac.uk/efo/EFO_0004789 \n",
      "\tcd40 ligand measurement http://www.ebi.ac.uk/efo/EFO_0004790 cd154 ligand measurement\n",
      "\talkaline phosphatase measurement http://www.ebi.ac.uk/efo/EFO_0004533 ap\n",
      "\tbilirubin measurement http://www.ebi.ac.uk/efo/EFO_0004570 bilirubin levels\n",
      "\tvitamin d measurement http://www.ebi.ac.uk/efo/EFO_0004631 \n",
      "\tvitamin k measurement http://www.ebi.ac.uk/efo/EFO_0004618 phylloquinone measurement\n",
      "\ttumor necrosis factor-alpha measurement http://www.ebi.ac.uk/efo/EFO_0004684 \n",
      "\tc-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458 c-reactive protein level\n",
      "select biomarker traits 19629\n",
      "\tccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749 monocyte chemotactic protein-1 measurement|chemokine ccl2 level|monocyte chemoattractant protein-1 measurement|chemokine ccl2 levels|monocyte chemotactic protein-1 levels|monocyte chemotactic protein-1 level\n",
      "\tserum alanine aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004735 alanine transaminase measurement|alanine aminotransferase measurement\n",
      "\tatrial natriuretic factor measurement http://www.ebi.ac.uk/efo/EFO_0004789 \n",
      "\tcd40 ligand measurement http://www.ebi.ac.uk/efo/EFO_0004790 cd154 ligand measurement\n",
      "\talkaline phosphatase measurement http://www.ebi.ac.uk/efo/EFO_0004533 ap\n",
      "\tbilirubin measurement http://www.ebi.ac.uk/efo/EFO_0004570 bilirubin levels\n",
      "\tvitamin d measurement http://www.ebi.ac.uk/efo/EFO_0004631 \n",
      "\tvitamin k measurement http://www.ebi.ac.uk/efo/EFO_0004618 phylloquinone measurement\n",
      "\ttumor necrosis factor-alpha measurement http://www.ebi.ac.uk/efo/EFO_0004684 \n",
      "\tc-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458 c-reactive protein level\n",
      "select biomarker traits 19630\n",
      "\tccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749 monocyte chemotactic protein-1 measurement|chemokine ccl2 level|monocyte chemoattractant protein-1 measurement|chemokine ccl2 levels|monocyte chemotactic protein-1 levels|monocyte chemotactic protein-1 level\n",
      "\tserum alanine aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004735 alanine transaminase measurement|alanine aminotransferase measurement\n",
      "\tatrial natriuretic factor measurement http://www.ebi.ac.uk/efo/EFO_0004789 \n",
      "\tcd40 ligand measurement http://www.ebi.ac.uk/efo/EFO_0004790 cd154 ligand measurement\n",
      "\talkaline phosphatase measurement http://www.ebi.ac.uk/efo/EFO_0004533 ap\n",
      "\tbilirubin measurement http://www.ebi.ac.uk/efo/EFO_0004570 bilirubin levels\n",
      "\tvitamin d measurement http://www.ebi.ac.uk/efo/EFO_0004631 \n",
      "\tvitamin k measurement http://www.ebi.ac.uk/efo/EFO_0004618 phylloquinone measurement\n",
      "\ttumor necrosis factor-alpha measurement http://www.ebi.ac.uk/efo/EFO_0004684 \n",
      "\tc-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458 c-reactive protein level\n",
      "select biomarker traits 19631\n",
      "\tccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749 monocyte chemotactic protein-1 measurement|chemokine ccl2 level|monocyte chemoattractant protein-1 measurement|chemokine ccl2 levels|monocyte chemotactic protein-1 levels|monocyte chemotactic protein-1 level\n",
      "\tserum alanine aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004735 alanine transaminase measurement|alanine aminotransferase measurement\n",
      "\tatrial natriuretic factor measurement http://www.ebi.ac.uk/efo/EFO_0004789 \n",
      "\tcd40 ligand measurement http://www.ebi.ac.uk/efo/EFO_0004790 cd154 ligand measurement\n",
      "\talkaline phosphatase measurement http://www.ebi.ac.uk/efo/EFO_0004533 ap\n",
      "\tbilirubin measurement http://www.ebi.ac.uk/efo/EFO_0004570 bilirubin levels\n",
      "\tvitamin d measurement http://www.ebi.ac.uk/efo/EFO_0004631 \n",
      "\tvitamin k measurement http://www.ebi.ac.uk/efo/EFO_0004618 phylloquinone measurement\n",
      "\ttumor necrosis factor-alpha measurement http://www.ebi.ac.uk/efo/EFO_0004684 \n",
      "\tc-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458 c-reactive protein level\n",
      "select biomarker traits 19632\n",
      "\tccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749 monocyte chemotactic protein-1 measurement|chemokine ccl2 level|monocyte chemoattractant protein-1 measurement|chemokine ccl2 levels|monocyte chemotactic protein-1 levels|monocyte chemotactic protein-1 level\n",
      "\tserum alanine aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004735 alanine transaminase measurement|alanine aminotransferase measurement\n",
      "\tatrial natriuretic factor measurement http://www.ebi.ac.uk/efo/EFO_0004789 \n",
      "\tcd40 ligand measurement http://www.ebi.ac.uk/efo/EFO_0004790 cd154 ligand measurement\n",
      "\talkaline phosphatase measurement http://www.ebi.ac.uk/efo/EFO_0004533 ap\n",
      "\tbilirubin measurement http://www.ebi.ac.uk/efo/EFO_0004570 bilirubin levels\n",
      "\tvitamin d measurement http://www.ebi.ac.uk/efo/EFO_0004631 \n",
      "\tvitamin k measurement http://www.ebi.ac.uk/efo/EFO_0004618 phylloquinone measurement\n",
      "\ttumor necrosis factor-alpha measurement http://www.ebi.ac.uk/efo/EFO_0004684 \n",
      "\tc-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458 c-reactive protein level\n",
      "select biomarker traits 19160\n",
      "\tccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749 monocyte chemotactic protein-1 measurement|chemokine ccl2 level|monocyte chemoattractant protein-1 measurement|chemokine ccl2 levels|monocyte chemotactic protein-1 levels|monocyte chemotactic protein-1 level\n",
      "\tserum alanine aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004735 alanine transaminase measurement|alanine aminotransferase measurement\n",
      "\tatrial natriuretic factor measurement http://www.ebi.ac.uk/efo/EFO_0004789 \n",
      "\tcd40 ligand measurement http://www.ebi.ac.uk/efo/EFO_0004790 cd154 ligand measurement\n",
      "\talkaline phosphatase measurement http://www.ebi.ac.uk/efo/EFO_0004533 ap\n",
      "\tbilirubin measurement http://www.ebi.ac.uk/efo/EFO_0004570 bilirubin levels\n",
      "\tvitamin d measurement http://www.ebi.ac.uk/efo/EFO_0004631 \n",
      "\tvitamin k measurement http://www.ebi.ac.uk/efo/EFO_0004618 phylloquinone measurement\n",
      "\ttumor necrosis factor-alpha measurement http://www.ebi.ac.uk/efo/EFO_0004684 \n",
      "\tc-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458 c-reactive protein level\n",
      "select biomarker traits 19634\n",
      "\tccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749 monocyte chemotactic protein-1 measurement|chemokine ccl2 level|monocyte chemoattractant protein-1 measurement|chemokine ccl2 levels|monocyte chemotactic protein-1 levels|monocyte chemotactic protein-1 level\n",
      "\tserum alanine aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004735 alanine transaminase measurement|alanine aminotransferase measurement\n",
      "\tatrial natriuretic factor measurement http://www.ebi.ac.uk/efo/EFO_0004789 \n",
      "\tcd40 ligand measurement http://www.ebi.ac.uk/efo/EFO_0004790 cd154 ligand measurement\n",
      "\talkaline phosphatase measurement http://www.ebi.ac.uk/efo/EFO_0004533 ap\n",
      "\tbilirubin measurement http://www.ebi.ac.uk/efo/EFO_0004570 bilirubin levels\n",
      "\tvitamin d measurement http://www.ebi.ac.uk/efo/EFO_0004631 \n",
      "\tvitamin k measurement http://www.ebi.ac.uk/efo/EFO_0004618 phylloquinone measurement\n",
      "\ttumor necrosis factor-alpha measurement http://www.ebi.ac.uk/efo/EFO_0004684 \n",
      "\tc-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458 c-reactive protein level\n",
      "select biomarker traits 19635\n",
      "\tccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749 monocyte chemotactic protein-1 measurement|chemokine ccl2 level|monocyte chemoattractant protein-1 measurement|chemokine ccl2 levels|monocyte chemotactic protein-1 levels|monocyte chemotactic protein-1 level\n",
      "\tserum alanine aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004735 alanine transaminase measurement|alanine aminotransferase measurement\n",
      "\tatrial natriuretic factor measurement http://www.ebi.ac.uk/efo/EFO_0004789 \n",
      "\tcd40 ligand measurement http://www.ebi.ac.uk/efo/EFO_0004790 cd154 ligand measurement\n",
      "\talkaline phosphatase measurement http://www.ebi.ac.uk/efo/EFO_0004533 ap\n",
      "\tbilirubin measurement http://www.ebi.ac.uk/efo/EFO_0004570 bilirubin levels\n",
      "\tvitamin d measurement http://www.ebi.ac.uk/efo/EFO_0004631 \n",
      "\tvitamin k measurement http://www.ebi.ac.uk/efo/EFO_0004618 phylloquinone measurement\n",
      "\ttumor necrosis factor-alpha measurement http://www.ebi.ac.uk/efo/EFO_0004684 \n",
      "\tc-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458 c-reactive protein level\n",
      "select biomarker traits 19636\n",
      "\tccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749 monocyte chemotactic protein-1 measurement|chemokine ccl2 level|monocyte chemoattractant protein-1 measurement|chemokine ccl2 levels|monocyte chemotactic protein-1 levels|monocyte chemotactic protein-1 level\n",
      "\tserum alanine aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004735 alanine transaminase measurement|alanine aminotransferase measurement\n",
      "\tatrial natriuretic factor measurement http://www.ebi.ac.uk/efo/EFO_0004789 \n",
      "\tcd40 ligand measurement http://www.ebi.ac.uk/efo/EFO_0004790 cd154 ligand measurement\n",
      "\talkaline phosphatase measurement http://www.ebi.ac.uk/efo/EFO_0004533 ap\n",
      "\tbilirubin measurement http://www.ebi.ac.uk/efo/EFO_0004570 bilirubin levels\n",
      "\tvitamin d measurement http://www.ebi.ac.uk/efo/EFO_0004631 \n",
      "\tvitamin k measurement http://www.ebi.ac.uk/efo/EFO_0004618 phylloquinone measurement\n",
      "\ttumor necrosis factor-alpha measurement http://www.ebi.ac.uk/efo/EFO_0004684 \n",
      "\tc-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458 c-reactive protein level\n",
      "select biomarker traits 19157\n",
      "\tccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749 monocyte chemotactic protein-1 measurement|chemokine ccl2 level|monocyte chemoattractant protein-1 measurement|chemokine ccl2 levels|monocyte chemotactic protein-1 levels|monocyte chemotactic protein-1 level\n",
      "\tserum alanine aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004735 alanine transaminase measurement|alanine aminotransferase measurement\n",
      "\tatrial natriuretic factor measurement http://www.ebi.ac.uk/efo/EFO_0004789 \n",
      "\tcd40 ligand measurement http://www.ebi.ac.uk/efo/EFO_0004790 cd154 ligand measurement\n",
      "\talkaline phosphatase measurement http://www.ebi.ac.uk/efo/EFO_0004533 ap\n",
      "\tbilirubin measurement http://www.ebi.ac.uk/efo/EFO_0004570 bilirubin levels\n",
      "\tvitamin d measurement http://www.ebi.ac.uk/efo/EFO_0004631 \n",
      "\tvitamin k measurement http://www.ebi.ac.uk/efo/EFO_0004618 phylloquinone measurement\n",
      "\ttumor necrosis factor-alpha measurement http://www.ebi.ac.uk/efo/EFO_0004684 \n",
      "\tc-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458 c-reactive protein level\n",
      "select biomarker traits 19158\n",
      "\tccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749 monocyte chemotactic protein-1 measurement|chemokine ccl2 level|monocyte chemoattractant protein-1 measurement|chemokine ccl2 levels|monocyte chemotactic protein-1 levels|monocyte chemotactic protein-1 level\n",
      "\tserum alanine aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004735 alanine transaminase measurement|alanine aminotransferase measurement\n",
      "\tatrial natriuretic factor measurement http://www.ebi.ac.uk/efo/EFO_0004789 \n",
      "\tcd40 ligand measurement http://www.ebi.ac.uk/efo/EFO_0004790 cd154 ligand measurement\n",
      "\talkaline phosphatase measurement http://www.ebi.ac.uk/efo/EFO_0004533 ap\n",
      "\tbilirubin measurement http://www.ebi.ac.uk/efo/EFO_0004570 bilirubin levels\n",
      "\tvitamin d measurement http://www.ebi.ac.uk/efo/EFO_0004631 \n",
      "\tvitamin k measurement http://www.ebi.ac.uk/efo/EFO_0004618 phylloquinone measurement\n",
      "\ttumor necrosis factor-alpha measurement http://www.ebi.ac.uk/efo/EFO_0004684 \n",
      "\tc-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458 c-reactive protein level\n",
      "select biomarker traits 19159\n",
      "\tccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749 monocyte chemotactic protein-1 measurement|chemokine ccl2 level|monocyte chemoattractant protein-1 measurement|chemokine ccl2 levels|monocyte chemotactic protein-1 levels|monocyte chemotactic protein-1 level\n",
      "\tserum alanine aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004735 alanine transaminase measurement|alanine aminotransferase measurement\n",
      "\tatrial natriuretic factor measurement http://www.ebi.ac.uk/efo/EFO_0004789 \n",
      "\tcd40 ligand measurement http://www.ebi.ac.uk/efo/EFO_0004790 cd154 ligand measurement\n",
      "\talkaline phosphatase measurement http://www.ebi.ac.uk/efo/EFO_0004533 ap\n",
      "\tbilirubin measurement http://www.ebi.ac.uk/efo/EFO_0004570 bilirubin levels\n",
      "\tvitamin d measurement http://www.ebi.ac.uk/efo/EFO_0004631 \n",
      "\tvitamin k measurement http://www.ebi.ac.uk/efo/EFO_0004618 phylloquinone measurement\n",
      "\ttumor necrosis factor-alpha measurement http://www.ebi.ac.uk/efo/EFO_0004684 \n",
      "\tc-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458 c-reactive protein level\n",
      "select biomarker traits 19640\n",
      "\tccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749 monocyte chemotactic protein-1 measurement|chemokine ccl2 level|monocyte chemoattractant protein-1 measurement|chemokine ccl2 levels|monocyte chemotactic protein-1 levels|monocyte chemotactic protein-1 level\n",
      "\tserum alanine aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004735 alanine transaminase measurement|alanine aminotransferase measurement\n",
      "\tatrial natriuretic factor measurement http://www.ebi.ac.uk/efo/EFO_0004789 \n",
      "\tcd40 ligand measurement http://www.ebi.ac.uk/efo/EFO_0004790 cd154 ligand measurement\n",
      "\talkaline phosphatase measurement http://www.ebi.ac.uk/efo/EFO_0004533 ap\n",
      "\tbilirubin measurement http://www.ebi.ac.uk/efo/EFO_0004570 bilirubin levels\n",
      "\tvitamin d measurement http://www.ebi.ac.uk/efo/EFO_0004631 \n",
      "\tvitamin k measurement http://www.ebi.ac.uk/efo/EFO_0004618 phylloquinone measurement\n",
      "\ttumor necrosis factor-alpha measurement http://www.ebi.ac.uk/efo/EFO_0004684 \n",
      "\tc-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458 c-reactive protein level\n",
      "select biomarker traits 19641\n",
      "\tccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749 monocyte chemotactic protein-1 measurement|chemokine ccl2 level|monocyte chemoattractant protein-1 measurement|chemokine ccl2 levels|monocyte chemotactic protein-1 levels|monocyte chemotactic protein-1 level\n",
      "\tserum alanine aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004735 alanine transaminase measurement|alanine aminotransferase measurement\n",
      "\tatrial natriuretic factor measurement http://www.ebi.ac.uk/efo/EFO_0004789 \n",
      "\tcd40 ligand measurement http://www.ebi.ac.uk/efo/EFO_0004790 cd154 ligand measurement\n",
      "\talkaline phosphatase measurement http://www.ebi.ac.uk/efo/EFO_0004533 ap\n",
      "\tbilirubin measurement http://www.ebi.ac.uk/efo/EFO_0004570 bilirubin levels\n",
      "\tvitamin d measurement http://www.ebi.ac.uk/efo/EFO_0004631 \n",
      "\tvitamin k measurement http://www.ebi.ac.uk/efo/EFO_0004618 phylloquinone measurement\n",
      "\ttumor necrosis factor-alpha measurement http://www.ebi.ac.uk/efo/EFO_0004684 \n",
      "\tc-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458 c-reactive protein level\n",
      "select biomarker traits 19642\n",
      "\tccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749 monocyte chemotactic protein-1 measurement|chemokine ccl2 level|monocyte chemoattractant protein-1 measurement|chemokine ccl2 levels|monocyte chemotactic protein-1 levels|monocyte chemotactic protein-1 level\n",
      "\tserum alanine aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004735 alanine transaminase measurement|alanine aminotransferase measurement\n",
      "\tatrial natriuretic factor measurement http://www.ebi.ac.uk/efo/EFO_0004789 \n",
      "\tcd40 ligand measurement http://www.ebi.ac.uk/efo/EFO_0004790 cd154 ligand measurement\n",
      "\talkaline phosphatase measurement http://www.ebi.ac.uk/efo/EFO_0004533 ap\n",
      "\tbilirubin measurement http://www.ebi.ac.uk/efo/EFO_0004570 bilirubin levels\n",
      "\tvitamin d measurement http://www.ebi.ac.uk/efo/EFO_0004631 \n",
      "\tvitamin k measurement http://www.ebi.ac.uk/efo/EFO_0004618 phylloquinone measurement\n",
      "\ttumor necrosis factor-alpha measurement http://www.ebi.ac.uk/efo/EFO_0004684 \n",
      "\tc-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458 c-reactive protein level\n",
      "select biomarker traits 19639\n",
      "\tccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749 monocyte chemotactic protein-1 measurement|chemokine ccl2 level|monocyte chemoattractant protein-1 measurement|chemokine ccl2 levels|monocyte chemotactic protein-1 levels|monocyte chemotactic protein-1 level\n",
      "\tserum alanine aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004735 alanine transaminase measurement|alanine aminotransferase measurement\n",
      "\tatrial natriuretic factor measurement http://www.ebi.ac.uk/efo/EFO_0004789 \n",
      "\tcd40 ligand measurement http://www.ebi.ac.uk/efo/EFO_0004790 cd154 ligand measurement\n",
      "\talkaline phosphatase measurement http://www.ebi.ac.uk/efo/EFO_0004533 ap\n",
      "\tbilirubin measurement http://www.ebi.ac.uk/efo/EFO_0004570 bilirubin levels\n",
      "\tvitamin d measurement http://www.ebi.ac.uk/efo/EFO_0004631 \n",
      "\tvitamin k measurement http://www.ebi.ac.uk/efo/EFO_0004618 phylloquinone measurement\n",
      "\ttumor necrosis factor-alpha measurement http://www.ebi.ac.uk/efo/EFO_0004684 \n",
      "\tc-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458 c-reactive protein level\n",
      "select biomarker traits 19637\n",
      "\tccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749 monocyte chemotactic protein-1 measurement|chemokine ccl2 level|monocyte chemoattractant protein-1 measurement|chemokine ccl2 levels|monocyte chemotactic protein-1 levels|monocyte chemotactic protein-1 level\n",
      "\tserum alanine aminotransferase measurement http://www.ebi.ac.uk/efo/EFO_0004735 alanine transaminase measurement|alanine aminotransferase measurement\n",
      "\tatrial natriuretic factor measurement http://www.ebi.ac.uk/efo/EFO_0004789 \n",
      "\tcd40 ligand measurement http://www.ebi.ac.uk/efo/EFO_0004790 cd154 ligand measurement\n",
      "\talkaline phosphatase measurement http://www.ebi.ac.uk/efo/EFO_0004533 ap\n",
      "\tbilirubin measurement http://www.ebi.ac.uk/efo/EFO_0004570 bilirubin levels\n",
      "\tvitamin d measurement http://www.ebi.ac.uk/efo/EFO_0004631 \n",
      "\tvitamin k measurement http://www.ebi.ac.uk/efo/EFO_0004618 phylloquinone measurement\n",
      "\ttumor necrosis factor-alpha measurement http://www.ebi.ac.uk/efo/EFO_0004684 \n",
      "\tc-reactive protein measurement http://www.ebi.ac.uk/efo/EFO_0004458 c-reactive protein level\n",
      "Genome-wide association with select biomarker traits in the Framingham Heart Study. Systemic biomarkers provide insights into disease pathogenesis, diagnosis, and risk stratification. Many systemic biomarker concentrations are heritable phenotypes. Genome-wide association studies (GWAS) provide mechanisms to investigate the genetic contributions to biomarker variability unconstrained by current knowledge of physiological relations. We examined the association of Affymetrix 100K GeneChip single nucleotide polymorphisms (SNPs) to 22 systemic biomarker concentrations in 4 biological domains: inflammation/oxidative stress; natriuretic peptides; liver function; and vitamins. Related members of the Framingham Offspring cohort (n = 1012; mean age 59 ± 10 years, 51% women) had both phenotype and genotype data (minimum-maximum per phenotype n = 507–1008). We used Generalized Estimating Equations (GEE), Family Based Association Tests (FBAT) and variance components linkage to relate SNPs to multivariable-adjusted biomarker residuals. Autosomal SNPs (n = 70,987) meeting the following criteria were studied: minor allele frequency ≥ 10%, call rate ≥ 80% and Hardy-Weinberg equilibrium p ≥ 0.001. With  GEE , 58 SNPs had p < 10 -6 : the top SNPs were rs2494250 (p = 1.00*10 -14 ) and rs4128725 (p = 3.68*10 -12 ) for monocyte chemoattractant protein-1 (MCP1), and rs2794520 (p = 2.83*10 -8 ) and rs2808629 (p = 3.19*10 -8 ) for C-reactive protein (CRP) averaged from 3 examinations (over about 20 years). With  FBAT , 11 SNPs had p < 10 -6 : the top SNPs were the same for MCP1 (rs4128725, p = 3.28*10 -8 , and rs2494250, p = 3.55*10 -8 ), and also included B-type natriuretic peptide (rs437021, p = 1.01*10 -6 ) and Vitamin K percent undercarboxylated osteocalcin (rs2052028, p = 1.07*10 -6 ). The peak  LOD  (logarithm of the odds) scores were for MCP1 (4.38, chromosome 1) and CRP (3.28, chromosome 1; previously described) concentrations; of note the 1.5 support interval included the MCP1 and CRP SNPs reported above (GEE model). Previous candidate SNP associations with circulating CRP concentrations were replicated at p < 0.05; the SNPs rs2794520 and rs2808629 are in linkage disequilibrium with previously reported SNPs. GEE, FBAT and linkage results are posted at  . The Framingham GWAS represents a resource to describe potentially novel genetic influences on systemic biomarker variability. The newly described associations will need to be replicated in other studies.\n",
      "-1 association with\n",
      "-1 biomarker traits\n",
      "-1 select\n",
      "-1 in\n",
      "-1 Heart\n",
      "-1 into disease\n",
      "-1 , diagnosis\n",
      "-1 diagnosis,\n",
      "-1 , and\n",
      "-1 Systemic\n",
      "-1 biomarkers\n",
      "-1 heritable phenotypes\n",
      "-1 systemic\n",
      "-1 biomarker\n",
      "-1 to biomarker\n",
      "-1 association\n",
      "-1 genetic\n",
      "-1 current\n",
      "-1 knowledge\n",
      "-1 of\n",
      "-1 physiological\n",
      "-1 relations\n",
      "-1 natriuretic peptides\n",
      "-1 liver function\n",
      "-1 association\n",
      "-1 of\n",
      "-1 single | herpes zoster http://www.ebi.ac.uk/efo/EFO_0006510\n",
      "-1 systemic\n",
      "-1 biomarker\n",
      "-1 in\n",
      "-1 biological\n",
      "-1 stress | exercise test http://www.ebi.ac.uk/efo/EFO_0004328\n",
      "-1 and\n",
      "-1 vitamins\n",
      "-1 of the\n",
      "-1 both phenotype\n",
      "-1 per phenotype\n",
      "-1 Related\n",
      "-1 n\n",
      "-1 and\n",
      "-1 genotype\n",
      "-1 n\n",
      "-1 , Family\n",
      "-1 to relate\n",
      "-1 Generalized\n",
      "-1 Association\n",
      "-1 and\n",
      "-1 biomarker\n",
      "-1 n\n",
      "-1 rate\n",
      "-1 and\n",
      "-1 p\n",
      "-1 C-reactive protein (CRP)\n",
      "1 monocyte chemoattractant protein-1 | ccl2 measurement http://www.ebi.ac.uk/efo/EFO_0004749\n",
      "-1 , and\n",
      "-1 p\n",
      "-1 p\n",
      "-1 and\n",
      "-1 p\n",
      "-1 p\n",
      "-1 and\n",
      "-1 p\n",
      "-1 over\n",
      "-1 B-type natriuretic peptide\n",
      "-1 , p\n",
      "-1 , and\n",
      "-1 , p\n",
      "-1 , and\n",
      "-1 , p\n",
      "1 Vitamin K | vitamin k measurement http://www.ebi.ac.uk/efo/EFO_0004618\n",
      "-1 , p\n",
      "-1 p\n",
      "-1 and\n",
      "-1 osteocalcin\n",
      "-1 of the\n",
      "-1 , chromosome\n",
      "-1 , chromosome\n",
      "-1 and\n",
      "-1 of\n",
      "-1 and\n",
      "-1 associations with\n",
      "-1 at p\n",
      "-1 and\n",
      "-1 in\n",
      "-1 disequilibrium\n",
      "-1 and\n",
      "-1 a\n",
      "-1 genetic\n",
      "-1 systemic\n",
      "-1 biomarker\n",
      "-1 associations\n",
      "-1 will\n",
      "-1 in\n",
      "\n",
      "17903294\n",
      "hemostatic factors and hematological phenotypes 19584\n",
      "\thematological measurement http://www.ebi.ac.uk/efo/EFO_0004503 \n",
      "\themoglobin measurement http://www.ebi.ac.uk/efo/EFO_0004509 hb|haemoglobin measurement|hemoglobin count\n",
      "\tplatelet aggregation http://purl.obolibrary.org/obo/GO_0070527 \n",
      "\terythrocyte count http://www.ebi.ac.uk/efo/EFO_0004305 erythrocyte number|red blood cell count|rbc\n",
      "\tblood viscosity http://www.ebi.ac.uk/efo/EFO_0004301 \n",
      "\tplasminogen activator inhibitor 1 measurement http://www.ebi.ac.uk/efo/EFO_0004792 \n",
      "\ttissue plasminogen activator measurement http://www.ebi.ac.uk/efo/EFO_0004791 \n",
      "\tmean corpuscular hemoglobin http://www.ebi.ac.uk/efo/EFO_0004527 mch|mean corpuscular haemoglobin\n",
      "\tprotein c measurement http://www.ebi.ac.uk/efo/EFO_0004633 coagulation factor xiv\n",
      "\tprotein s measurement http://www.ebi.ac.uk/efo/EFO_0004637 \n",
      "hemostatic factors and hematological phenotypes 19585\n",
      "\thematological measurement http://www.ebi.ac.uk/efo/EFO_0004503 \n",
      "\themoglobin measurement http://www.ebi.ac.uk/efo/EFO_0004509 hb|haemoglobin measurement|hemoglobin count\n",
      "\tplatelet aggregation http://purl.obolibrary.org/obo/GO_0070527 \n",
      "\terythrocyte count http://www.ebi.ac.uk/efo/EFO_0004305 erythrocyte number|red blood cell count|rbc\n",
      "\tblood viscosity http://www.ebi.ac.uk/efo/EFO_0004301 \n",
      "\tplasminogen activator inhibitor 1 measurement http://www.ebi.ac.uk/efo/EFO_0004792 \n",
      "\ttissue plasminogen activator measurement http://www.ebi.ac.uk/efo/EFO_0004791 \n",
      "\tmean corpuscular hemoglobin http://www.ebi.ac.uk/efo/EFO_0004527 mch|mean corpuscular haemoglobin\n",
      "\tprotein c measurement http://www.ebi.ac.uk/efo/EFO_0004633 coagulation factor xiv\n",
      "\tprotein s measurement http://www.ebi.ac.uk/efo/EFO_0004637 \n",
      "hemostatic factors and hematological phenotypes 19586\n",
      "\thematological measurement http://www.ebi.ac.uk/efo/EFO_0004503 \n",
      "\themoglobin measurement http://www.ebi.ac.uk/efo/EFO_0004509 hb|haemoglobin measurement|hemoglobin count\n",
      "\tplatelet aggregation http://purl.obolibrary.org/obo/GO_0070527 \n",
      "\terythrocyte count http://www.ebi.ac.uk/efo/EFO_0004305 erythrocyte number|red blood cell count|rbc\n",
      "\tblood viscosity http://www.ebi.ac.uk/efo/EFO_0004301 \n",
      "\tplasminogen activator inhibitor 1 measurement http://www.ebi.ac.uk/efo/EFO_0004792 \n",
      "\ttissue plasminogen activator measurement http://www.ebi.ac.uk/efo/EFO_0004791 \n",
      "\tmean corpuscular hemoglobin http://www.ebi.ac.uk/efo/EFO_0004527 mch|mean corpuscular haemoglobin\n",
      "\tprotein c measurement http://www.ebi.ac.uk/efo/EFO_0004633 coagulation factor xiv\n",
      "\tprotein s measurement http://www.ebi.ac.uk/efo/EFO_0004637 \n",
      "hemostatic factors and hematological phenotypes 19587\n",
      "\thematological measurement http://www.ebi.ac.uk/efo/EFO_0004503 \n",
      "\themoglobin measurement http://www.ebi.ac.uk/efo/EFO_0004509 hb|haemoglobin measurement|hemoglobin count\n",
      "\tplatelet aggregation http://purl.obolibrary.org/obo/GO_0070527 \n",
      "\terythrocyte count http://www.ebi.ac.uk/efo/EFO_0004305 erythrocyte number|red blood cell count|rbc\n",
      "\tblood viscosity http://www.ebi.ac.uk/efo/EFO_0004301 \n",
      "\tplasminogen activator inhibitor 1 measurement http://www.ebi.ac.uk/efo/EFO_0004792 \n",
      "\ttissue plasminogen activator measurement http://www.ebi.ac.uk/efo/EFO_0004791 \n",
      "\tmean corpuscular hemoglobin http://www.ebi.ac.uk/efo/EFO_0004527 mch|mean corpuscular haemoglobin\n",
      "\tprotein c measurement http://www.ebi.ac.uk/efo/EFO_0004633 coagulation factor xiv\n",
      "\tprotein s measurement http://www.ebi.ac.uk/efo/EFO_0004637 \n",
      "hemostatic factors and hematological phenotypes 19588\n",
      "\thematological measurement http://www.ebi.ac.uk/efo/EFO_0004503 \n",
      "\themoglobin measurement http://www.ebi.ac.uk/efo/EFO_0004509 hb|haemoglobin measurement|hemoglobin count\n",
      "\tplatelet aggregation http://purl.obolibrary.org/obo/GO_0070527 \n",
      "\terythrocyte count http://www.ebi.ac.uk/efo/EFO_0004305 erythrocyte number|red blood cell count|rbc\n",
      "\tblood viscosity http://www.ebi.ac.uk/efo/EFO_0004301 \n",
      "\tplasminogen activator inhibitor 1 measurement http://www.ebi.ac.uk/efo/EFO_0004792 \n",
      "\ttissue plasminogen activator measurement http://www.ebi.ac.uk/efo/EFO_0004791 \n",
      "\tmean corpuscular hemoglobin http://www.ebi.ac.uk/efo/EFO_0004527 mch|mean corpuscular haemoglobin\n",
      "\tprotein c measurement http://www.ebi.ac.uk/efo/EFO_0004633 coagulation factor xiv\n",
      "\tprotein s measurement http://www.ebi.ac.uk/efo/EFO_0004637 \n",
      "hemostatic factors and hematological phenotypes 19589\n",
      "\thematological measurement http://www.ebi.ac.uk/efo/EFO_0004503 \n",
      "\themoglobin measurement http://www.ebi.ac.uk/efo/EFO_0004509 hb|haemoglobin measurement|hemoglobin count\n",
      "\tplatelet aggregation http://purl.obolibrary.org/obo/GO_0070527 \n",
      "\terythrocyte count http://www.ebi.ac.uk/efo/EFO_0004305 erythrocyte number|red blood cell count|rbc\n",
      "\tblood viscosity http://www.ebi.ac.uk/efo/EFO_0004301 \n",
      "\tplasminogen activator inhibitor 1 measurement http://www.ebi.ac.uk/efo/EFO_0004792 \n",
      "\ttissue plasminogen activator measurement http://www.ebi.ac.uk/efo/EFO_0004791 \n",
      "\tmean corpuscular hemoglobin http://www.ebi.ac.uk/efo/EFO_0004527 mch|mean corpuscular haemoglobin\n",
      "\tprotein c measurement http://www.ebi.ac.uk/efo/EFO_0004633 coagulation factor xiv\n",
      "\tprotein s measurement http://www.ebi.ac.uk/efo/EFO_0004637 \n",
      "hemostatic factors and hematological phenotypes 19590\n",
      "\thematological measurement http://www.ebi.ac.uk/efo/EFO_0004503 \n",
      "\themoglobin measurement http://www.ebi.ac.uk/efo/EFO_0004509 hb|haemoglobin measurement|hemoglobin count\n",
      "\tplatelet aggregation http://purl.obolibrary.org/obo/GO_0070527 \n",
      "\terythrocyte count http://www.ebi.ac.uk/efo/EFO_0004305 erythrocyte number|red blood cell count|rbc\n",
      "\tblood viscosity http://www.ebi.ac.uk/efo/EFO_0004301 \n",
      "\tplasminogen activator inhibitor 1 measurement http://www.ebi.ac.uk/efo/EFO_0004792 \n",
      "\ttissue plasminogen activator measurement http://www.ebi.ac.uk/efo/EFO_0004791 \n",
      "\tmean corpuscular hemoglobin http://www.ebi.ac.uk/efo/EFO_0004527 mch|mean corpuscular haemoglobin\n",
      "\tprotein c measurement http://www.ebi.ac.uk/efo/EFO_0004633 coagulation factor xiv\n",
      "\tprotein s measurement http://www.ebi.ac.uk/efo/EFO_0004637 \n",
      "hemostatic factors and hematological phenotypes 19591\n",
      "\thematological measurement http://www.ebi.ac.uk/efo/EFO_0004503 \n",
      "\themoglobin measurement http://www.ebi.ac.uk/efo/EFO_0004509 hb|haemoglobin measurement|hemoglobin count\n",
      "\tplatelet aggregation http://purl.obolibrary.org/obo/GO_0070527 \n",
      "\terythrocyte count http://www.ebi.ac.uk/efo/EFO_0004305 erythrocyte number|red blood cell count|rbc\n",
      "\tblood viscosity http://www.ebi.ac.uk/efo/EFO_0004301 \n",
      "\tplasminogen activator inhibitor 1 measurement http://www.ebi.ac.uk/efo/EFO_0004792 \n",
      "\ttissue plasminogen activator measurement http://www.ebi.ac.uk/efo/EFO_0004791 \n",
      "\tmean corpuscular hemoglobin http://www.ebi.ac.uk/efo/EFO_0004527 mch|mean corpuscular haemoglobin\n",
      "\tprotein c measurement http://www.ebi.ac.uk/efo/EFO_0004633 coagulation factor xiv\n",
      "\tprotein s measurement http://www.ebi.ac.uk/efo/EFO_0004637 \n",
      "factor vii 19233\n",
      "\tfactor vii measurement http://www.ebi.ac.uk/efo/EFO_0004619 \n",
      "factor vii 19234\n",
      "\tfactor vii measurement http://www.ebi.ac.uk/efo/EFO_0004619 \n",
      "hemostatic factors and hematological phenotypes 19235\n",
      "\thematological measurement http://www.ebi.ac.uk/efo/EFO_0004503 \n",
      "\themoglobin measurement http://www.ebi.ac.uk/efo/EFO_0004509 hb|haemoglobin measurement|hemoglobin count\n",
      "\tplatelet aggregation http://purl.obolibrary.org/obo/GO_0070527 \n",
      "\terythrocyte count http://www.ebi.ac.uk/efo/EFO_0004305 erythrocyte number|red blood cell count|rbc\n",
      "\tblood viscosity http://www.ebi.ac.uk/efo/EFO_0004301 \n",
      "\tplasminogen activator inhibitor 1 measurement http://www.ebi.ac.uk/efo/EFO_0004792 \n",
      "\ttissue plasminogen activator measurement http://www.ebi.ac.uk/efo/EFO_0004791 \n",
      "\tmean corpuscular hemoglobin http://www.ebi.ac.uk/efo/EFO_0004527 mch|mean corpuscular haemoglobin\n",
      "\tprotein c measurement http://www.ebi.ac.uk/efo/EFO_0004633 coagulation factor xiv\n",
      "\tprotein s measurement http://www.ebi.ac.uk/efo/EFO_0004637 \n",
      "hemostatic factors and hematological phenotypes 19236\n",
      "\thematological measurement http://www.ebi.ac.uk/efo/EFO_0004503 \n",
      "\themoglobin measurement http://www.ebi.ac.uk/efo/EFO_0004509 hb|haemoglobin measurement|hemoglobin count\n",
      "\tplatelet aggregation http://purl.obolibrary.org/obo/GO_0070527 \n",
      "\terythrocyte count http://www.ebi.ac.uk/efo/EFO_0004305 erythrocyte number|red blood cell count|rbc\n",
      "\tblood viscosity http://www.ebi.ac.uk/efo/EFO_0004301 \n",
      "\tplasminogen activator inhibitor 1 measurement http://www.ebi.ac.uk/efo/EFO_0004792 \n",
      "\ttissue plasminogen activator measurement http://www.ebi.ac.uk/efo/EFO_0004791 \n",
      "\tmean corpuscular hemoglobin http://www.ebi.ac.uk/efo/EFO_0004527 mch|mean corpuscular haemoglobin\n",
      "\tprotein c measurement http://www.ebi.ac.uk/efo/EFO_0004633 coagulation factor xiv\n",
      "\tprotein s measurement http://www.ebi.ac.uk/efo/EFO_0004637 \n",
      "factor vii 19431\n",
      "\tfactor vii measurement http://www.ebi.ac.uk/efo/EFO_0004619 \n",
      "factor vii 19575\n",
      "\tfactor vii measurement http://www.ebi.ac.uk/efo/EFO_0004619 \n",
      "hemostatic factors and hematological phenotypes 19576\n",
      "\thematological measurement http://www.ebi.ac.uk/efo/EFO_0004503 \n",
      "\themoglobin measurement http://www.ebi.ac.uk/efo/EFO_0004509 hb|haemoglobin measurement|hemoglobin count\n",
      "\tplatelet aggregation http://purl.obolibrary.org/obo/GO_0070527 \n",
      "\terythrocyte count http://www.ebi.ac.uk/efo/EFO_0004305 erythrocyte number|red blood cell count|rbc\n",
      "\tblood viscosity http://www.ebi.ac.uk/efo/EFO_0004301 \n",
      "\tplasminogen activator inhibitor 1 measurement http://www.ebi.ac.uk/efo/EFO_0004792 \n",
      "\ttissue plasminogen activator measurement http://www.ebi.ac.uk/efo/EFO_0004791 \n",
      "\tmean corpuscular hemoglobin http://www.ebi.ac.uk/efo/EFO_0004527 mch|mean corpuscular haemoglobin\n",
      "\tprotein c measurement http://www.ebi.ac.uk/efo/EFO_0004633 coagulation factor xiv\n",
      "\tprotein s measurement http://www.ebi.ac.uk/efo/EFO_0004637 \n",
      "hemostatic factors and hematological phenotypes 19577\n",
      "\thematological measurement http://www.ebi.ac.uk/efo/EFO_0004503 \n",
      "\themoglobin measurement http://www.ebi.ac.uk/efo/EFO_0004509 hb|haemoglobin measurement|hemoglobin count\n",
      "\tplatelet aggregation http://purl.obolibrary.org/obo/GO_0070527 \n",
      "\terythrocyte count http://www.ebi.ac.uk/efo/EFO_0004305 erythrocyte number|red blood cell count|rbc\n",
      "\tblood viscosity http://www.ebi.ac.uk/efo/EFO_0004301 \n",
      "\tplasminogen activator inhibitor 1 measurement http://www.ebi.ac.uk/efo/EFO_0004792 \n",
      "\ttissue plasminogen activator measurement http://www.ebi.ac.uk/efo/EFO_0004791 \n",
      "\tmean corpuscular hemoglobin http://www.ebi.ac.uk/efo/EFO_0004527 mch|mean corpuscular haemoglobin\n",
      "\tprotein c measurement http://www.ebi.ac.uk/efo/EFO_0004633 coagulation factor xiv\n",
      "\tprotein s measurement http://www.ebi.ac.uk/efo/EFO_0004637 \n",
      "hemostatic factors and hematological phenotypes 19578\n",
      "\thematological measurement http://www.ebi.ac.uk/efo/EFO_0004503 \n",
      "\themoglobin measurement http://www.ebi.ac.uk/efo/EFO_0004509 hb|haemoglobin measurement|hemoglobin count\n",
      "\tplatelet aggregation http://purl.obolibrary.org/obo/GO_0070527 \n",
      "\terythrocyte count http://www.ebi.ac.uk/efo/EFO_0004305 erythrocyte number|red blood cell count|rbc\n",
      "\tblood viscosity http://www.ebi.ac.uk/efo/EFO_0004301 \n",
      "\tplasminogen activator inhibitor 1 measurement http://www.ebi.ac.uk/efo/EFO_0004792 \n",
      "\ttissue plasminogen activator measurement http://www.ebi.ac.uk/efo/EFO_0004791 \n",
      "\tmean corpuscular hemoglobin http://www.ebi.ac.uk/efo/EFO_0004527 mch|mean corpuscular haemoglobin\n",
      "\tprotein c measurement http://www.ebi.ac.uk/efo/EFO_0004633 coagulation factor xiv\n",
      "\tprotein s measurement http://www.ebi.ac.uk/efo/EFO_0004637 \n",
      "hemostatic factors and hematological phenotypes 19579\n",
      "\thematological measurement http://www.ebi.ac.uk/efo/EFO_0004503 \n",
      "\themoglobin measurement http://www.ebi.ac.uk/efo/EFO_0004509 hb|haemoglobin measurement|hemoglobin count\n",
      "\tplatelet aggregation http://purl.obolibrary.org/obo/GO_0070527 \n",
      "\terythrocyte count http://www.ebi.ac.uk/efo/EFO_0004305 erythrocyte number|red blood cell count|rbc\n",
      "\tblood viscosity http://www.ebi.ac.uk/efo/EFO_0004301 \n",
      "\tplasminogen activator inhibitor 1 measurement http://www.ebi.ac.uk/efo/EFO_0004792 \n",
      "\ttissue plasminogen activator measurement http://www.ebi.ac.uk/efo/EFO_0004791 \n",
      "\tmean corpuscular hemoglobin http://www.ebi.ac.uk/efo/EFO_0004527 mch|mean corpuscular haemoglobin\n",
      "\tprotein c measurement http://www.ebi.ac.uk/efo/EFO_0004633 coagulation factor xiv\n",
      "\tprotein s measurement http://www.ebi.ac.uk/efo/EFO_0004637 \n",
      "hemostatic factors and hematological phenotypes 19580\n",
      "\thematological measurement http://www.ebi.ac.uk/efo/EFO_0004503 \n",
      "\themoglobin measurement http://www.ebi.ac.uk/efo/EFO_0004509 hb|haemoglobin measurement|hemoglobin count\n",
      "\tplatelet aggregation http://purl.obolibrary.org/obo/GO_0070527 \n",
      "\terythrocyte count http://www.ebi.ac.uk/efo/EFO_0004305 erythrocyte number|red blood cell count|rbc\n",
      "\tblood viscosity http://www.ebi.ac.uk/efo/EFO_0004301 \n",
      "\tplasminogen activator inhibitor 1 measurement http://www.ebi.ac.uk/efo/EFO_0004792 \n",
      "\ttissue plasminogen activator measurement http://www.ebi.ac.uk/efo/EFO_0004791 \n",
      "\tmean corpuscular hemoglobin http://www.ebi.ac.uk/efo/EFO_0004527 mch|mean corpuscular haemoglobin\n",
      "\tprotein c measurement http://www.ebi.ac.uk/efo/EFO_0004633 coagulation factor xiv\n",
      "\tprotein s measurement http://www.ebi.ac.uk/efo/EFO_0004637 \n",
      "hemostatic factors and hematological phenotypes 19581\n",
      "\thematological measurement http://www.ebi.ac.uk/efo/EFO_0004503 \n",
      "\themoglobin measurement http://www.ebi.ac.uk/efo/EFO_0004509 hb|haemoglobin measurement|hemoglobin count\n",
      "\tplatelet aggregation http://purl.obolibrary.org/obo/GO_0070527 \n",
      "\terythrocyte count http://www.ebi.ac.uk/efo/EFO_0004305 erythrocyte number|red blood cell count|rbc\n",
      "\tblood viscosity http://www.ebi.ac.uk/efo/EFO_0004301 \n",
      "\tplasminogen activator inhibitor 1 measurement http://www.ebi.ac.uk/efo/EFO_0004792 \n",
      "\ttissue plasminogen activator measurement http://www.ebi.ac.uk/efo/EFO_0004791 \n",
      "\tmean corpuscular hemoglobin http://www.ebi.ac.uk/efo/EFO_0004527 mch|mean corpuscular haemoglobin\n",
      "\tprotein c measurement http://www.ebi.ac.uk/efo/EFO_0004633 coagulation factor xiv\n",
      "\tprotein s measurement http://www.ebi.ac.uk/efo/EFO_0004637 \n",
      "hemostatic factors and hematological phenotypes 19582\n",
      "\thematological measurement http://www.ebi.ac.uk/efo/EFO_0004503 \n",
      "\themoglobin measurement http://www.ebi.ac.uk/efo/EFO_0004509 hb|haemoglobin measurement|hemoglobin count\n",
      "\tplatelet aggregation http://purl.obolibrary.org/obo/GO_0070527 \n",
      "\terythrocyte count http://www.ebi.ac.uk/efo/EFO_0004305 erythrocyte number|red blood cell count|rbc\n",
      "\tblood viscosity http://www.ebi.ac.uk/efo/EFO_0004301 \n",
      "\tplasminogen activator inhibitor 1 measurement http://www.ebi.ac.uk/efo/EFO_0004792 \n",
      "\ttissue plasminogen activator measurement http://www.ebi.ac.uk/efo/EFO_0004791 \n",
      "\tmean corpuscular hemoglobin http://www.ebi.ac.uk/efo/EFO_0004527 mch|mean corpuscular haemoglobin\n",
      "\tprotein c measurement http://www.ebi.ac.uk/efo/EFO_0004633 coagulation factor xiv\n",
      "\tprotein s measurement http://www.ebi.ac.uk/efo/EFO_0004637 \n",
      "hemostatic factors and hematological phenotypes 19583\n",
      "\thematological measurement http://www.ebi.ac.uk/efo/EFO_0004503 \n",
      "\themoglobin measurement http://www.ebi.ac.uk/efo/EFO_0004509 hb|haemoglobin measurement|hemoglobin count\n",
      "\tplatelet aggregation http://purl.obolibrary.org/obo/GO_0070527 \n",
      "\terythrocyte count http://www.ebi.ac.uk/efo/EFO_0004305 erythrocyte number|red blood cell count|rbc\n",
      "\tblood viscosity http://www.ebi.ac.uk/efo/EFO_0004301 \n",
      "\tplasminogen activator inhibitor 1 measurement http://www.ebi.ac.uk/efo/EFO_0004792 \n",
      "\ttissue plasminogen activator measurement http://www.ebi.ac.uk/efo/EFO_0004791 \n",
      "\tmean corpuscular hemoglobin http://www.ebi.ac.uk/efo/EFO_0004527 mch|mean corpuscular haemoglobin\n",
      "\tprotein c measurement http://www.ebi.ac.uk/efo/EFO_0004633 coagulation factor xiv\n",
      "\tprotein s measurement http://www.ebi.ac.uk/efo/EFO_0004637 \n",
      "Genome-wide association and linkage analyses of hemostatic factors and hematological phenotypes in the Framingham Heart Study. Increased circulating levels of hemostatic factors as well as anemia have been associated with increased risk of cardiovascular disease (CVD). Known associations between hemostatic factors and sequence variants at genes encoding these factors explain only a small proportion of total phenotypic variation. We sought to confirm known putative loci and identify novel loci that may influence either trait in genome-wide association and linkage analyses using the Affymetrix GeneChip 100K single nucleotide polymorphism (SNP) set. Plasma levels of circulating hemostatic factors (fibrinogen, factor VII, plasminogen activator inhibitor-1, von Willebrand factor, tissue plasminogen activator, D-dimer) and hematological phenotypes (platelet aggregation, viscosity, hemoglobin, red blood cell count, mean corpuscular volume, mean corpuscular hemoglobin concentration) were obtained in approximately 1000 Framingham Heart Study (FHS) participants from 310 families. Population-based association analyses using the generalized estimating equations (GEE), family-based association test (FBAT), and multipoint variance components linkage analyses were performed on the multivariable adjusted residuals of hemostatic and hematological phenotypes. In association analysis, the lowest GEE p-value for hemostatic factors was p = 4.5*10 -16  for factor VII at SNP rs561241, a variant located near the  F7  gene and in complete linkage disequilibrium (LD) (r 2  = 1) with the Arg353Gln  F7  SNP previously shown to account for 9% of total phenotypic variance. The lowest GEE p-value for hematological phenotypes was 7*10 -8  at SNP rs2412522 on chromosome 4 for mean corpuscular hemoglobin concentration. We presented top 25 most significant GEE results with p-values in the range of 10 -6  to 10 -5  for hemostatic or hematological phenotypes. In relating 100K SNPs to known candidate genes, we identified two SNPs (rs1582055, rs4897475) in erythrocyte membrane protein band 4.1-like 2 ( EPB41L2 ) associated with hematological phenotypes (GEE p < 10 -3 ). In linkage analyses, the highest linkage LOD score for hemostatic factors was 3.3 for factor VII on chromosome 10 around 15 Mb, and for hematological phenotypes, LOD 3.4 for hemoglobin on chromosome 4 around 55 Mb. All GEE and FBAT association and variance components linkage results can be found at  Using genome-wide association methodology, we have successfully identified a SNP in complete LD with a sequence variant previously shown to be strongly associated with factor VII, providing proof of principle for this approach. Further study of additional strongly associated SNPs and linked regions may identify novel variants that influence the inter-individual variability in hemostatic factors and hematological phenotypes.\n",
      "-1 hematological phenotypes\n",
      "-1 association\n",
      "-1 and\n",
      "-1 of\n",
      "-1 and\n",
      "-1 in\n",
      "-1 Heart\n",
      "-1 cardiovascular disease (CVD)\n",
      "-1 with increased risk of\n",
      "-1 of cardiovascular disease\n",
      "-1 levels of\n",
      "-1 associated with\n",
      "-1 Increased\n",
      "-1 as\n",
      "-1 as\n",
      "-1 anemia | anemia http://www.ebi.ac.uk/efo/EFO_0004272\n",
      "-1 associations\n",
      "-1 and\n",
      "-1 a\n",
      "-1 small\n",
      "-1 of\n",
      "-1 phenotypic\n",
      "-1 to confirm\n",
      "-1 either trait\n",
      "-1 trait in\n",
      "-1 and\n",
      "-1 association\n",
      "-1 and\n",
      "-1 single | herpes zoster http://www.ebi.ac.uk/efo/EFO_0006510\n",
      "1 , red blood cell count | erythrocyte count http://www.ebi.ac.uk/efo/EFO_0004305\n",
      "1 red blood cell count, | erythrocyte count http://www.ebi.ac.uk/efo/EFO_0004305\n",
      "-1 , mean corpuscular hemoglobin concentration | mean corpuscular hemoglobin concentration http://www.ebi.ac.uk/efo/EFO_0004528\n",
      "1 , plasminogen activator inhibitor-1 | plasminogen activator inhibitor 1 measurement http://www.ebi.ac.uk/efo/EFO_0004792\n",
      "1 plasminogen activator inhibitor-1, | plasminogen activator inhibitor 1 measurement http://www.ebi.ac.uk/efo/EFO_0004792\n",
      "-1 , von Willebrand factor | von willebrand factor measurement http://www.ebi.ac.uk/efo/EFO_0004629\n",
      "-1 von Willebrand factor, | von willebrand factor measurement http://www.ebi.ac.uk/efo/EFO_0004629\n",
      "1 , tissue plasminogen activator | tissue plasminogen activator measurement http://www.ebi.ac.uk/efo/EFO_0004791\n",
      "1 tissue plasminogen activator, | tissue plasminogen activator measurement http://www.ebi.ac.uk/efo/EFO_0004791\n",
      "-1 , mean corpuscular volume | mean corpuscular volume http://www.ebi.ac.uk/efo/EFO_0004526\n",
      "-1 mean corpuscular volume, | mean corpuscular volume http://www.ebi.ac.uk/efo/EFO_0004526\n",
      "-1 corpuscular volume, mean | mean corpuscular volume http://www.ebi.ac.uk/efo/EFO_0004526\n",
      "-1 volume, mean corpuscular | mean corpuscular volume http://www.ebi.ac.uk/efo/EFO_0004526\n",
      "1 , factor VII | factor vii measurement http://www.ebi.ac.uk/efo/EFO_0004619\n",
      "1 factor VII, | factor vii measurement http://www.ebi.ac.uk/efo/EFO_0004619\n",
      "-1 factor, tissue\n",
      "1 platelet aggregation, | platelet aggregation http://purl.obolibrary.org/obo/GO_0070527\n",
      "-1 levels of\n",
      "-1 fibrinogen, | fibrinogen measurement http://www.ebi.ac.uk/efo/EFO_0004623\n",
      "-1 , D-dimer\n",
      "-1 hematological phenotypes\n",
      "-1 , viscosity\n",
      "-1 viscosity,\n",
      "1 , hemoglobin | hemoglobin measurement http://www.ebi.ac.uk/efo/EFO_0004509\n",
      "1 hemoglobin, | hemoglobin measurement http://www.ebi.ac.uk/efo/EFO_0004509\n",
      "-1 and\n",
      "-1 in\n",
      "-1 Heart\n",
      "-1 FHS\n",
      "-1 families\n",
      "-1 association test\n",
      "-1 , and\n",
      "-1 hematological phenotypes\n",
      "-1 association\n",
      "-1 generalized\n",
      "-1 adjusted\n",
      "-1 of\n",
      "-1 and\n",
      "1 factor VII at | factor vii measurement http://www.ebi.ac.uk/efo/EFO_0004619\n",
      "-1 analysis,\n",
      "-1 , a\n",
      "-1 In\n",
      "-1 association\n",
      "-1 was\n",
      "-1 p\n",
      "-1 near\n",
      "-1 and\n",
      "-1 in\n",
      "-1 disequilibrium\n",
      "-1 LD\n",
      "-1 of\n",
      "-1 phenotypic\n",
      "-1 mean corpuscular hemoglobin concentration | mean corpuscular hemoglobin concentration http://www.ebi.ac.uk/efo/EFO_0004528\n",
      "-1 hematological phenotypes\n",
      "-1 chromosome 4\n",
      "-1 was\n",
      "-1 hematological phenotypes\n",
      "-1 in\n",
      "-1 of\n",
      "-1 -5\n",
      "-1 associated with\n",
      "-1 hematological phenotypes\n",
      "-1 In\n",
      "-1 relating\n",
      "-1 in\n",
      "-1 erythrocyte | erythrocyte measurement http://www.ebi.ac.uk/efo/EFO_0005047\n",
      "-1 protein | protein measurement http://www.ebi.ac.uk/efo/EFO_0004747\n",
      "-1 p\n",
      "1 factor VII | factor vii measurement http://www.ebi.ac.uk/efo/EFO_0004619\n",
      "-1 , and\n",
      "-1 hematological phenotypes\n",
      "-1 phenotypes,\n",
      "-1 chromosome 4\n",
      "-1 In\n",
      "-1 was\n",
      "-1 chromosome\n",
      "1 hemoglobin | hemoglobin measurement http://www.ebi.ac.uk/efo/EFO_0004509\n",
      "1 factor VII, | factor vii measurement http://www.ebi.ac.uk/efo/EFO_0004619\n",
      "-1 associated with\n",
      "-1 All\n",
      "-1 and\n",
      "-1 association\n",
      "-1 and\n",
      "-1 association\n",
      "-1 a\n",
      "-1 in\n",
      "-1 LD\n",
      "-1 a\n",
      "-1 of\n",
      "-1 hematological phenotypes\n",
      "-1 of\n",
      "-1 associated\n",
      "-1 and\n",
      "-1 in\n",
      "-1 and\n",
      "\n",
      "17903295\n",
      "aging traits 19811\n",
      "\tage at menopause http://www.ebi.ac.uk/efo/EFO_0004704 \n",
      "\texercise test http://www.ebi.ac.uk/efo/EFO_0004328 stress test\n",
      "\taging http://purl.obolibrary.org/obo/GO_0007568 senescence|aging\n",
      "\tage at death http://www.ebi.ac.uk/efo/EFO_0005056 \n",
      "aging traits 19812\n",
      "\tage at menopause http://www.ebi.ac.uk/efo/EFO_0004704 \n",
      "\texercise test http://www.ebi.ac.uk/efo/EFO_0004328 stress test\n",
      "\taging http://purl.obolibrary.org/obo/GO_0007568 senescence|aging\n",
      "\tage at death http://www.ebi.ac.uk/efo/EFO_0005056 \n",
      "aging traits 19813\n",
      "\tage at menopause http://www.ebi.ac.uk/efo/EFO_0004704 \n",
      "\texercise test http://www.ebi.ac.uk/efo/EFO_0004328 stress test\n",
      "\taging http://purl.obolibrary.org/obo/GO_0007568 senescence|aging\n",
      "\tage at death http://www.ebi.ac.uk/efo/EFO_0005056 \n",
      "aging traits 19814\n",
      "\tage at menopause http://www.ebi.ac.uk/efo/EFO_0004704 \n",
      "\texercise test http://www.ebi.ac.uk/efo/EFO_0004328 stress test\n",
      "\taging http://purl.obolibrary.org/obo/GO_0007568 senescence|aging\n",
      "\tage at death http://www.ebi.ac.uk/efo/EFO_0005056 \n",
      "aging traits 19815\n",
      "\tage at menopause http://www.ebi.ac.uk/efo/EFO_0004704 \n",
      "\texercise test http://www.ebi.ac.uk/efo/EFO_0004328 stress test\n",
      "\taging http://purl.obolibrary.org/obo/GO_0007568 senescence|aging\n",
      "\tage at death http://www.ebi.ac.uk/efo/EFO_0005056 \n",
      "aging traits 19816\n",
      "\tage at menopause http://www.ebi.ac.uk/efo/EFO_0004704 \n",
      "\texercise test http://www.ebi.ac.uk/efo/EFO_0004328 stress test\n",
      "\taging http://purl.obolibrary.org/obo/GO_0007568 senescence|aging\n",
      "\tage at death http://www.ebi.ac.uk/efo/EFO_0005056 \n",
      "aging traits 19817\n",
      "\tage at menopause http://www.ebi.ac.uk/efo/EFO_0004704 \n",
      "\texercise test http://www.ebi.ac.uk/efo/EFO_0004328 stress test\n",
      "\taging http://purl.obolibrary.org/obo/GO_0007568 senescence|aging\n",
      "\tage at death http://www.ebi.ac.uk/efo/EFO_0005056 \n",
      "morbidity-free survival 19818\n",
      "\tdisease free survival http://www.ebi.ac.uk/efo/EFO_0000409  disease-free survival|morbidity-free survival\n",
      "morbidity-free survival 19819\n",
      "\tdisease free survival http://www.ebi.ac.uk/efo/EFO_0000409  disease-free survival|morbidity-free survival\n",
      "morbidity-free survival 19820\n",
      "\tdisease free survival http://www.ebi.ac.uk/efo/EFO_0000409  disease-free survival|morbidity-free survival\n",
      "morbidity-free survival 19189\n",
      "\tdisease free survival http://www.ebi.ac.uk/efo/EFO_0000409  disease-free survival|morbidity-free survival\n",
      "morbidity-free survival 19190\n",
      "\tdisease free survival http://www.ebi.ac.uk/efo/EFO_0000409  disease-free survival|morbidity-free survival\n",
      "aging traits 19191\n",
      "\tage at menopause http://www.ebi.ac.uk/efo/EFO_0004704 \n",
      "\texercise test http://www.ebi.ac.uk/efo/EFO_0004328 stress test\n",
      "\taging http://purl.obolibrary.org/obo/GO_0007568 senescence|aging\n",
      "\tage at death http://www.ebi.ac.uk/efo/EFO_0005056 \n",
      "aging traits 19192\n",
      "\tage at menopause http://www.ebi.ac.uk/efo/EFO_0004704 \n",
      "\texercise test http://www.ebi.ac.uk/efo/EFO_0004328 stress test\n",
      "\taging http://purl.obolibrary.org/obo/GO_0007568 senescence|aging\n",
      "\tage at death http://www.ebi.ac.uk/efo/EFO_0005056 \n",
      "aging traits 19193\n",
      "\tage at menopause http://www.ebi.ac.uk/efo/EFO_0004704 \n",
      "\texercise test http://www.ebi.ac.uk/efo/EFO_0004328 stress test\n",
      "\taging http://purl.obolibrary.org/obo/GO_0007568 senescence|aging\n",
      "\tage at death http://www.ebi.ac.uk/efo/EFO_0005056 \n",
      "Genetic correlates of longevity and selected age-related phenotypes: a genome-wide association study in the Framingham Study. Family studies and heritability estimates provide evidence for a genetic contribution to variation in the human life span. We conducted a genome wide association study (Affymetrix 100K SNP GeneChip) for longevity-related traits in a community-based sample. We report on 5 longevity and aging traits in up to 1345 Framingham Study participants from 330 families. Multivariable-adjusted residuals were computed using appropriate models (Cox proportional hazards, logistic, or linear regression) and the residuals from these models were used to test for association with qualifying SNPs (70, 987 autosomal SNPs with genotypic call rate ≥80%, minor allele frequency ≥10%, Hardy-Weinberg test p ≥ 0.001). In family-based association test (FBAT) models, 8 SNPs in two regions approximately 500 kb apart on chromosome 1 (physical positions 73,091,610 and 73, 527,652) were associated with age at death (p-value < 10 -5 ). The two sets of SNPs were in high linkage disequilibrium (minimum r 2  = 0.58). The top 30 SNPs for generalized estimating equation (GEE) tests of association with age at death included rs10507486 (p = 0.0001) and rs4943794 (p = 0.0002), SNPs intronic to  FOXO1A , a gene implicated in lifespan extension in animal models. FBAT models identified 7 SNPs and GEE models identified 9 SNPs associated with both age at death and morbidity-free survival at age 65 including rs2374983 near  PON1 . In the analysis of selected candidate genes, SNP associations (FBAT or GEE p-value < 0.01) were identified for age at death in or near the following genes:  FOXO1A, GAPDH, KL, LEPR, PON1, PSEN1, SOD2, and WRN . Top ranked SNP associations in the GEE model for age at natural menopause included rs6910534 (p = 0.00003) near  FOXO3a  and rs3751591 (p = 0.00006) in  CYP19A1 . Results of all longevity phenotype-genotype associations for all autosomal SNPs are web posted at  . Longevity and aging traits are associated with SNPs on the Affymetrix 100K GeneChip. None of the associations achieved genome-wide significance. These data generate hypotheses and serve as a resource for replication as more genes and biologic pathways are proposed as contributing to longevity and healthy aging.\n",
      "-1 age-related phenotypes\n",
      "-1 Genetic\n",
      "-1 of\n",
      "-1 longevity | longevity http://www.ebi.ac.uk/efo/EFO_0004300\n",
      "-1 and\n",
      "-1 selected\n",
      "-1 a\n",
      "-1 association\n",
      "-1 in\n",
      "-1 Family\n",
      "-1 and\n",
      "-1 a\n",
      "-1 genetic\n",
      "-1 in\n",
      "-1 human\n",
      "-1 longevity-related traits\n",
      "-1 conducted\n",
      "-1 a\n",
      "-1 genome\n",
      "-1 association\n",
      "-1 in\n",
      "-1 a\n",
      "1 aging traits\n",
      "-1 5\n",
      "-1 longevity | longevity http://www.ebi.ac.uk/efo/EFO_0004300\n",
      "-1 and\n",
      "-1 in\n",
      "-1 families\n",
      "-1 association with\n",
      "-1 test p\n",
      "-1 and\n",
      "-1 genotypic\n",
      "-1 rate\n",
      "-1 associated with age at\n",
      "1 age at death | sudden cardiac arrest http://www.ebi.ac.uk/efo/EFO_0004278\n",
      "-1 association test\n",
      "-1 In\n",
      "-1 in\n",
      "-1 chromosome\n",
      "-1 and\n",
      "-1 -5\n",
      "-1 of\n",
      "-1 in\n",
      "-1 high\n",
      "-1 disequilibrium\n",
      "-1 association with age at\n",
      "1 age at death | sudden cardiac arrest http://www.ebi.ac.uk/efo/EFO_0004278\n",
      "-1 , a\n",
      "-1 generalized\n",
      "-1 of\n",
      "-1 p\n",
      "-1 and\n",
      "-1 p\n",
      "-1 in\n",
      "-1 in\n",
      "-1 animal\n",
      "1 morbidity-free survival at age | disease free survival http://www.ebi.ac.uk/efo/EFO_0000409\n",
      "1 age at death | sudden cardiac arrest http://www.ebi.ac.uk/efo/EFO_0004278\n",
      "-1 associated with\n",
      "-1 and\n",
      "-1 and\n",
      "-1 near\n",
      "1 age at death | sudden cardiac arrest http://www.ebi.ac.uk/efo/EFO_0004278\n",
      "-1 , and\n",
      "-1 In\n",
      "-1 analysis\n",
      "-1 of\n",
      "-1 selected\n",
      "-1 associations\n",
      "-1 in\n",
      "-1 near\n",
      "-1 associations\n",
      "-1 in\n",
      "1 menopause | age at menopause http://www.ebi.ac.uk/efo/EFO_0004704\n",
      "-1 p\n",
      "-1 near\n",
      "-1 and\n",
      "-1 p\n",
      "-1 in\n",
      "-1 CYP19A1\n",
      "-1 of\n",
      "-1 all\n",
      "-1 longevity | longevity http://www.ebi.ac.uk/efo/EFO_0004300\n",
      "-1 associations\n",
      "-1 all\n",
      "1 aging traits\n",
      "-1 associated with\n",
      "-1 Longevity | longevity http://www.ebi.ac.uk/efo/EFO_0004300\n",
      "-1 and\n",
      "-1 of the\n",
      "-1 associations\n",
      "-1 as a\n",
      "-1 to longevity | longevity http://www.ebi.ac.uk/efo/EFO_0004300\n",
      "-1 generate\n",
      "-1 and\n",
      "-1 as\n",
      "-1 and\n",
      "-1 biologic\n",
      "-1 as\n",
      "-1 and\n",
      "1 aging | aging http://purl.obolibrary.org/obo/GO_0007568\n",
      "\n",
      "17903296\n",
      "bone mineral density 19843\n",
      "\tbone density http://www.ebi.ac.uk/efo/EFO_0003923 density\n",
      "bone mineral density 19844\n",
      "\tbone density http://www.ebi.ac.uk/efo/EFO_0003923 density\n",
      "hip geometry 19598\n",
      "\thip geometry http://www.ebi.ac.uk/efo/EFO_0004685 \n",
      "hip geometry 19599\n",
      "\thip geometry http://www.ebi.ac.uk/efo/EFO_0004685 \n",
      "hip geometry 19600\n",
      "\thip geometry http://www.ebi.ac.uk/efo/EFO_0004685 \n",
      "hip geometry 19601\n",
      "\thip geometry http://www.ebi.ac.uk/efo/EFO_0004685 \n",
      "hip geometry 19602\n",
      "\thip geometry http://www.ebi.ac.uk/efo/EFO_0004685 \n",
      "hip geometry 19603\n",
      "\thip geometry http://www.ebi.ac.uk/efo/EFO_0004685 \n",
      "hip geometry 19604\n",
      "\thip geometry http://www.ebi.ac.uk/efo/EFO_0004685 \n",
      "hip geometry 19605\n",
      "\thip geometry http://www.ebi.ac.uk/efo/EFO_0004685 \n",
      "hip geometry 19606\n",
      "\thip geometry http://www.ebi.ac.uk/efo/EFO_0004685 \n",
      "hip geometry 19607\n",
      "\thip geometry http://www.ebi.ac.uk/efo/EFO_0004685 \n",
      "hip geometry 19608\n",
      "\thip geometry http://www.ebi.ac.uk/efo/EFO_0004685 \n",
      "hip geometry 19609\n",
      "\thip geometry http://www.ebi.ac.uk/efo/EFO_0004685 \n",
      "bone mineral density 19610\n",
      "\tbone density http://www.ebi.ac.uk/efo/EFO_0003923 density\n",
      "bone mineral density 19611\n",
      "\tbone density http://www.ebi.ac.uk/efo/EFO_0003923 density\n",
      "bone mineral density 19612\n",
      "\tbone density http://www.ebi.ac.uk/efo/EFO_0003923 density\n",
      "bone mineral density 19613\n",
      "\tbone density http://www.ebi.ac.uk/efo/EFO_0003923 density\n",
      "bone mineral density 19614\n",
      "\tbone density http://www.ebi.ac.uk/efo/EFO_0003923 density\n",
      "bone mineral density 19174\n",
      "\tbone density http://www.ebi.ac.uk/efo/EFO_0003923 density\n",
      "hip geometry 19175\n",
      "\thip geometry http://www.ebi.ac.uk/efo/EFO_0004685 \n",
      "hip geometry 19176\n",
      "\thip geometry http://www.ebi.ac.uk/efo/EFO_0004685 \n",
      "hip geometry 19177\n",
      "\thip geometry http://www.ebi.ac.uk/efo/EFO_0004685 \n",
      "hip geometry 19178\n",
      "\thip geometry http://www.ebi.ac.uk/efo/EFO_0004685 \n",
      "Genome-wide association with bone mass and geometry in the Framingham Heart Study. Osteoporosis is characterized by low bone mass and compromised bone structure, heritable traits that contribute to fracture risk. There have been no genome-wide association and linkage studies for these traits using high-density genotyping platforms. We used the Affymetrix 100K SNP GeneChip marker set in the Framingham Heart Study (FHS) to examine genetic associations with ten primary quantitative traits: bone mineral density (BMD), calcaneal ultrasound, and geometric indices of the hip. To test associations with multivariable-adjusted residual trait values, we used additive generalized estimating equation (GEE) and family-based association tests (FBAT) models within each sex as well as sexes combined. We evaluated 70,987 autosomal SNPs with genotypic call rates ≥80%, HWE p ≥ 0.001, and MAF ≥10% in up to 1141 phenotyped individuals (495 men and 646 women, mean age 62.5 yrs). Variance component linkage analysis was performed using 11,200 markers. Heritability estimates for all bone phenotypes were 30–66%. LOD scores ≥3.0 were found on chromosomes 15 (1.5 LOD confidence interval: 51,336,679–58,934,236 bp) and 22 (35,890,398–48,603,847 bp) for femoral shaft section modulus. The ten primary phenotypes had 12 associations with 100K SNPs in GEE models at p < 0.000001 and 2 associations in FBAT models at p < 0.000001. The 25 most significant p-values for GEE and FBAT were all less than 3.5 × 10 -6  and 2.5 × 10 -5 , respectively. Of the 40 top SNPs with the greatest numbers of significantly associated BMD traits (including femoral neck, trochanter, and lumbar spine), one half to two-thirds were in or near genes that have  not  previously been studied for osteoporosis. Notably, pleiotropic associations between BMD and bone geometric traits were uncommon. Evidence for association (FBAT or GEE p < 0.05) was observed for several SNPs in candidate genes for osteoporosis, such as rs1801133 in  MTHFR ; rs1884052 and rs3778099 in  ESR1 ; rs4988300 in  LRP5 ; rs2189480 in  VDR ; rs2075555 in  COLIA1 ; rs10519297 and rs2008691 in  CYP19 , as well as SNPs in  PPARG  (rs10510418 and rs2938392) and  ANKH  (rs2454873 and rs379016). All GEE, FBAT and linkage results are provided as an open-access results resource at  . The FHS 100K SNP project offers an unbiased genome-wide strategy to identify new candidate loci and to replicate previously suggested candidate genes for osteoporosis.\n",
      "-1 association with bone\n",
      "-1 mass\n",
      "-1 and\n",
      "-1 in\n",
      "-1 Heart\n",
      "-1 is characterized by\n",
      "-1 heritable traits\n",
      "-1 Osteoporosis | osteoporosis http://www.ebi.ac.uk/efo/EFO_0003882\n",
      "-1 low\n",
      "-1 bone\n",
      "-1 mass\n",
      "-1 and\n",
      "-1 bone\n",
      "-1 these traits\n",
      "-1 no\n",
      "-1 association\n",
      "-1 and\n",
      "-1 genotyping\n",
      "-1 primary quantitative traits\n",
      "1 bone mineral density\n",
      "-1 of the hip\n",
      "-1 associations with\n",
      "-1 , and\n",
      "-1 in\n",
      "-1 Heart\n",
      "-1 FHS\n",
      "-1 genetic\n",
      "-1 BMD\n",
      "-1 indices | ankle brachial index http://www.ebi.ac.uk/efo/EFO_0003912\n",
      "-1 To test associations with\n",
      "-1 trait values,\n",
      "-1 residual trait\n",
      "-1 generalized\n",
      "-1 and\n",
      "-1 association\n",
      "-1 sex\n",
      "-1 as\n",
      "-1 as\n",
      "-1 sexes\n",
      "-1 , and\n",
      "-1 genotypic\n",
      "-1 rates\n",
      "-1 p\n",
      "-1 in\n",
      "-1 phenotyped\n",
      "-1 and\n",
      "-1 analysis\n",
      "-1 was\n",
      "-1 bone phenotypes\n",
      "-1 all\n",
      "-1 chromosomes\n",
      "-1 and\n",
      "-1 primary phenotypes\n",
      "-1 associations with\n",
      "-1 at p\n",
      "-1 at p\n",
      "-1 in\n",
      "-1 and\n",
      "-1 associations\n",
      "-1 in\n",
      "-1 -5 ,\n",
      "-1 and\n",
      "-1 all\n",
      "-1 ×\n",
      "-1 and\n",
      "-1 ×\n",
      "-1 Of the\n",
      "-1 BMD traits\n",
      "-1 , and\n",
      "-1 of\n",
      "-1 associated\n",
      "-1 in\n",
      "-1 near\n",
      "-1 not\n",
      "-1 osteoporosis | osteoporosis http://www.ebi.ac.uk/efo/EFO_0003882\n",
      "-1 geometric traits\n",
      "-1 associations\n",
      "-1 BMD\n",
      "-1 and\n",
      "-1 bone\n",
      "-1 osteoporosis, | osteoporosis http://www.ebi.ac.uk/efo/EFO_0003882\n",
      "-1 , as\n",
      "-1 association\n",
      "-1 p\n",
      "-1 was\n",
      "-1 several\n",
      "-1 in\n",
      "-1 as\n",
      "-1 in\n",
      "-1 and\n",
      "-1 in\n",
      "-1 in\n",
      "-1 in\n",
      "-1 in\n",
      "-1 and\n",
      "-1 in\n",
      "-1 as\n",
      "-1 in\n",
      "-1 and\n",
      "-1 and\n",
      "-1 and\n",
      "-1 All\n",
      "-1 and\n",
      "-1 as\n",
      "-1 an\n",
      "-1 and to\n",
      "-1 FHS\n",
      "-1 offers\n",
      "-1 an\n",
      "-1 suggested\n",
      "-1 osteoporosis | osteoporosis http://www.ebi.ac.uk/efo/EFO_0003882\n",
      "\n",
      "17903297\n",
      "volumetric brain mri 19552\n",
      "\tvolumetric brain mri http://www.ebi.ac.uk/efo/EFO_0004868 \n",
      "volumetric brain mri 19553\n",
      "\tvolumetric brain mri http://www.ebi.ac.uk/efo/EFO_0004868 \n",
      "volumetric brain mri 19554\n",
      "\tvolumetric brain mri http://www.ebi.ac.uk/efo/EFO_0004868 \n",
      "cognitive test performance 19555\n",
      "\tneuropsychological test http://www.ebi.ac.uk/efo/EFO_0003926 test\n",
      "cognitive test performance 19556\n",
      "\tneuropsychological test http://www.ebi.ac.uk/efo/EFO_0003926 test\n",
      "cognitive test performance 19557\n",
      "\tneuropsychological test http://www.ebi.ac.uk/efo/EFO_0003926 test\n",
      "cognitive test performance 19558\n",
      "\tneuropsychological test http://www.ebi.ac.uk/efo/EFO_0003926 test\n",
      "cognitive test performance 19559\n",
      "\tneuropsychological test http://www.ebi.ac.uk/efo/EFO_0003926 test\n",
      "cognitive test performance 19560\n",
      "\tneuropsychological test http://www.ebi.ac.uk/efo/EFO_0003926 test\n",
      "cognitive test performance 19561\n",
      "\tneuropsychological test http://www.ebi.ac.uk/efo/EFO_0003926 test\n",
      "cognitive test performance 19562\n",
      "\tneuropsychological test http://www.ebi.ac.uk/efo/EFO_0003926 test\n",
      "cognitive test performance 19563\n",
      "\tneuropsychological test http://www.ebi.ac.uk/efo/EFO_0003926 test\n",
      "cognitive test performance 19564\n",
      "\tneuropsychological test http://www.ebi.ac.uk/efo/EFO_0003926 test\n",
      "cognitive test performance 19218\n",
      "\tneuropsychological test http://www.ebi.ac.uk/efo/EFO_0003926 test\n",
      "cognitive test performance 19219\n",
      "\tneuropsychological test http://www.ebi.ac.uk/efo/EFO_0003926 test\n",
      "cognitive test performance 19220\n",
      "\tneuropsychological test http://www.ebi.ac.uk/efo/EFO_0003926 test\n",
      "cognitive test performance 19221\n",
      "\tneuropsychological test http://www.ebi.ac.uk/efo/EFO_0003926 test\n",
      "volumetric brain mri 19222\n",
      "\tvolumetric brain mri http://www.ebi.ac.uk/efo/EFO_0004868 \n",
      "volumetric brain mri 19551\n",
      "\tvolumetric brain mri http://www.ebi.ac.uk/efo/EFO_0004868 \n",
      "Genetic correlates of brain aging on MRI and cognitive test measures: a genome-wide association and linkage analysis in the Framingham study. Brain magnetic resonance imaging (MRI) and cognitive tests can identify heritable endophenotypes associated with an increased risk of developing stroke, dementia and Alzheimer's disease (AD). We conducted a genome-wide association (GWA) and linkage analysis exploring the genetic basis of these endophenotypes in a community-based sample. A total of 705 stroke- and dementia-free Framingham participants (age 62 +9 yrs, 50% male) who underwent volumetric brain MRI and cognitive testing (1999–2002) were genotyped. We used linear models adjusting for first degree relationships via generalized estimating equations (GEE) and family based association tests (FBAT) in additive models to relate qualifying single nucleotide polymorphisms (SNPs, 70,987 autosomal on Affymetrix 100K Human Gene Chip with minor allele frequency ≥ 0.10, genotypic call rate ≥ 0.80, and Hardy-Weinberg equilibrium p-value ≥ 0.001) to multivariable-adjusted residuals of 9 MRI measures including total cerebral brain (TCBV), lobar, ventricular and white matter hyperintensity (WMH) volumes, and 6 cognitive factors/tests assessing verbal and visuospatial memory, visual scanning and motor speed, reading, abstract reasoning and naming. We determined multipoint identity-by-descent utilizing 10,592 informative SNPs and 613 short tandem repeats and used variance component analyses to compute LOD scores. The strongest gene-phenotype association in FBAT analyses was between  SORL1  (rs1131497; p = 3.2 × 10 -6 ) and abstract reasoning, and in GEE analyses between  CDH4  (rs1970546; p = 3.7 × 10 -8 ) and TCBV. SORL1  plays a role in amyloid precursor protein processing and has been associated with the risk of AD. Among the 50 strongest associations (25 each by GEE and FBAT) were other biologically interesting genes. Polymorphisms within 28 of 163 candidate genes for stroke, AD and memory impairment were associated with the endophenotypes studied at p < 0.001. We confirmed our previously reported linkage of WMH on chromosome 4 and describe linkage of reading performance to a marker on chromosome 18 (GATA11A06), previously linked to dyslexia (LOD scores = 2.2 and 5.1). Our results suggest that genes associated with clinical neurological disease also have detectable effects on subclinical phenotypes. These hypothesis generating data illustrate the use of an unbiased approach to discover novel pathways that may be involved in brain aging, and could be used to replicate observations made in other studies.\n",
      "-1 of brain\n",
      "-1 cognitive test\n",
      "-1 Genetic\n",
      "-1 aging | aging http://purl.obolibrary.org/obo/GO_0007568\n",
      "-1 and\n",
      "-1 a\n",
      "-1 association\n",
      "-1 and\n",
      "-1 analysis\n",
      "-1 in\n",
      "-1 Alzheimer's disease (AD)\n",
      "-1 increased risk of developing\n",
      "-1 associated with an\n",
      "-1 stroke, | stroke http://www.ebi.ac.uk/efo/EFO_0000712\n",
      "-1 , dementia | alzheimers disease http://www.ebi.ac.uk/efo/EFO_0000249\n",
      "-1 Brain | brain measurement http://www.ebi.ac.uk/efo/EFO_0004464\n",
      "-1 resonance\n",
      "-1 and\n",
      "-1 cognitive\n",
      "-1 endophenotypes\n",
      "-1 and\n",
      "-1 conducted\n",
      "-1 a\n",
      "-1 association\n",
      "-1 and\n",
      "-1 analysis\n",
      "-1 genetic\n",
      "-1 of\n",
      "-1 endophenotypes\n",
      "-1 in\n",
      "-1 a\n",
      "1 volumetric brain MRI | volumetric brain mri http://www.ebi.ac.uk/efo/EFO_0004868\n",
      "-1 stroke- | stroke http://www.ebi.ac.uk/efo/EFO_0000712\n",
      "-1 - and\n",
      "-1 A\n",
      "-1 of\n",
      "-1 and\n",
      "-1 cognitive\n",
      "-1 genotyped\n",
      "-1 and white matter\n",
      "-1 white matter hyperintensity\n",
      "-1 to relate\n",
      "-1 , genotypic\n",
      "-1 , and\n",
      "-1 , and\n",
      "-1 memory,\n",
      "-1 , reading | reading http://www.ebi.ac.uk/efo/EFO_0005229\n",
      "-1 reading, | reading http://www.ebi.ac.uk/efo/EFO_0005229\n",
      "-1 adjusting\n",
      "-1 generalized\n",
      "-1 and\n",
      "-1 family\n",
      "-1 association\n",
      "-1 in\n",
      "-1 single | herpes zoster http://www.ebi.ac.uk/efo/EFO_0006510\n",
      "-1 Human\n",
      "-1 rate\n",
      "-1 of\n",
      "-1 cerebral\n",
      "-1 brain | brain measurement http://www.ebi.ac.uk/efo/EFO_0004464\n",
      "-1 cognitive\n",
      "-1 and\n",
      "-1 and\n",
      "-1 reasoning\n",
      "-1 and\n",
      "-1 and\n",
      "-1 short\n",
      "-1 and\n",
      "-1 reasoning,\n",
      "-1 , and\n",
      "-1 association\n",
      "-1 in\n",
      "-1 was\n",
      "-1 p\n",
      "-1 ×\n",
      "-1 and\n",
      "-1 in\n",
      "-1 p\n",
      "-1 ×\n",
      "-1 and\n",
      "-1 associated with\n",
      "-1 risk of\n",
      "-1 plays\n",
      "-1 a\n",
      "-1 in\n",
      "-1 protein | protein measurement http://www.ebi.ac.uk/efo/EFO_0004747\n",
      "-1 and\n",
      "-1 AD\n",
      "-1 associations\n",
      "-1 and\n",
      "-1 biologically\n",
      "-1 stroke, | stroke http://www.ebi.ac.uk/efo/EFO_0000712\n",
      "-1 , AD\n",
      "-1 memory impairment\n",
      "-1 associated with\n",
      "-1 at p\n",
      "-1 of\n",
      "-1 and\n",
      "-1 endophenotypes\n",
      "-1 chromosome 4\n",
      "-1 to a\n",
      "-1 chromosome 18\n",
      "-1 to dyslexia | dyslexia http://www.ebi.ac.uk/efo/EFO_0005424\n",
      "-1 confirmed\n",
      "-1 of\n",
      "-1 and\n",
      "-1 of\n",
      "-1 reading | reading http://www.ebi.ac.uk/efo/EFO_0005229\n",
      "-1 and\n",
      "-1 associated with\n",
      "-1 neurological disease\n",
      "-1 subclinical phenotypes\n",
      "-1 suggest\n",
      "-1 clinical\n",
      "-1 detectable\n",
      "-1 use of\n",
      "-1 aging, | aging http://purl.obolibrary.org/obo/GO_0007568\n",
      "-1 , and\n",
      "-1 generating\n",
      "-1 an\n",
      "-1 in\n",
      "-1 brain | brain measurement http://www.ebi.ac.uk/efo/EFO_0004464\n",
      "-1 in\n",
      "\n",
      "17903298\n",
      "diabetes related insulin traits 19592\n",
      "\ta1c measurement http://www.ebi.ac.uk/efo/EFO_0004541 hba1c levels|glycosylated haemoglobin a level|glycosylated hemoglobin a level|glycated hemoglobin levels\n",
      "\tfasting blood insulin measurement http://www.ebi.ac.uk/efo/EFO_0004466 fasting blood insulin|fasting blood insulin level|fasting insulin-related traits\n",
      "\tinsulin sensitivity measurement http://www.ebi.ac.uk/efo/EFO_0004471 insulin sensitivity test\n",
      "\tinsulin resistance http://www.ebi.ac.uk/efo/EFO_0002614 drug resistance to insulin (disorder)\n",
      "diabetes related insulin traits 19593\n",
      "\ta1c measurement http://www.ebi.ac.uk/efo/EFO_0004541 hba1c levels|glycosylated haemoglobin a level|glycosylated hemoglobin a level|glycated hemoglobin levels\n",
      "\tfasting blood insulin measurement http://www.ebi.ac.uk/efo/EFO_0004466 fasting blood insulin|fasting blood insulin level|fasting insulin-related traits\n",
      "\tinsulin sensitivity measurement http://www.ebi.ac.uk/efo/EFO_0004471 insulin sensitivity test\n",
      "\tinsulin resistance http://www.ebi.ac.uk/efo/EFO_0002614 drug resistance to insulin (disorder)\n",
      "fasting plasma glucose 19594\n",
      "\tfasting blood glucose measurement http://www.ebi.ac.uk/efo/EFO_0004465 fasting glucose-related traits|fasting blood glucose level|fasting plasma glucose\n",
      "fasting plasma glucose 19595\n",
      "\tfasting blood glucose measurement http://www.ebi.ac.uk/efo/EFO_0004465 fasting glucose-related traits|fasting blood glucose level|fasting plasma glucose\n",
      "fasting plasma glucose 19596\n",
      "\tfasting blood glucose measurement http://www.ebi.ac.uk/efo/EFO_0004465 fasting glucose-related traits|fasting blood glucose level|fasting plasma glucose\n",
      "fasting plasma glucose 19597\n",
      "\tfasting blood glucose measurement http://www.ebi.ac.uk/efo/EFO_0004465 fasting glucose-related traits|fasting blood glucose level|fasting plasma glucose\n",
      "diabetes related insulin traits 19194\n",
      "\ta1c measurement http://www.ebi.ac.uk/efo/EFO_0004541 hba1c levels|glycosylated haemoglobin a level|glycosylated hemoglobin a level|glycated hemoglobin levels\n",
      "\tfasting blood insulin measurement http://www.ebi.ac.uk/efo/EFO_0004466 fasting blood insulin|fasting blood insulin level|fasting insulin-related traits\n",
      "\tinsulin sensitivity measurement http://www.ebi.ac.uk/efo/EFO_0004471 insulin sensitivity test\n",
      "\tinsulin resistance http://www.ebi.ac.uk/efo/EFO_0002614 drug resistance to insulin (disorder)\n",
      "diabetes related insulin traits 19195\n",
      "\ta1c measurement http://www.ebi.ac.uk/efo/EFO_0004541 hba1c levels|glycosylated haemoglobin a level|glycosylated hemoglobin a level|glycated hemoglobin levels\n",
      "\tfasting blood insulin measurement http://www.ebi.ac.uk/efo/EFO_0004466 fasting blood insulin|fasting blood insulin level|fasting insulin-related traits\n",
      "\tinsulin sensitivity measurement http://www.ebi.ac.uk/efo/EFO_0004471 insulin sensitivity test\n",
      "\tinsulin resistance http://www.ebi.ac.uk/efo/EFO_0002614 drug resistance to insulin (disorder)\n",
      "fasting plasma glucose 19196\n",
      "\tfasting blood glucose measurement http://www.ebi.ac.uk/efo/EFO_0004465 fasting glucose-related traits|fasting blood glucose level|fasting plasma glucose\n",
      "fasting plasma glucose 19197\n",
      "\tfasting blood glucose measurement http://www.ebi.ac.uk/efo/EFO_0004465 fasting glucose-related traits|fasting blood glucose level|fasting plasma glucose\n",
      "diabetes (incident) 19198\n",
      "\tdiabetes mellitus http://www.ebi.ac.uk/efo/EFO_0000400 dm - diabetes mellitus|diabetes mellitus\n",
      "Genome-wide association with diabetes-related traits in the Framingham Heart Study. Susceptibility to type 2 diabetes may be conferred by genetic variants having modest effects on risk. Genome-wide fixed marker arrays offer a novel approach to detect these variants. We used the Affymetrix 100K SNP array in 1,087 Framingham Offspring Study family members to examine genetic associations with three diabetes-related quantitative glucose traits (fasting plasma glucose (FPG), hemoglobin A1c, 28-yr time-averaged FPG (tFPG)), three insulin traits (fasting insulin, HOMA-insulin resistance, and 0–120 min insulin sensitivity index); and with risk for diabetes. We used additive generalized estimating equations (GEE) and family-based association test (FBAT) models to test associations of SNP genotypes with sex-age-age 2 -adjusted residual trait values, and Cox survival models to test incident diabetes. We found 415 SNPs associated (at p < 0.001) with at least one of the six quantitative traits in GEE, 242 in FBAT (18 overlapped with GEE for 639 non-overlapping SNPs), and 128 associated with incident diabetes (31 overlapped with the 639) giving 736 non-overlapping SNPs. Of these 736 SNPs, 439 were within 60 kb of a known gene. Additionally, 53 SNPs (of which 42 had r 2  < 0.80 with each other) had p < 0.01 for incident diabetes AND (all 3 glucose traits OR all 3 insulin traits, OR 2 glucose traits and 2 insulin traits); of these, 36 overlapped with the 736 other SNPs. Of 100K SNPs, one (rs7100927) was in moderate LD (r 2  = 0.50) with  TCF7L2  (rs7903146), and was associated with risk of diabetes (Cox p-value 0.007, additive hazard ratio for diabetes = 1.56) and with tFPG (GEE p-value 0.03). There were no common (MAF > 1%) 100K SNPs in LD (r 2  > 0.05) with  ABCC8  A1369S (rs757110),  KCNJ11  E23K (rs5219), or SNPs in  CAPN10  or  HNFa . PPARG  P12A (rs1801282) was not significantly associated with diabetes or related traits. Framingham 100K SNP data is a resource for association tests of known and novel genes with diabetes and related traits posted at  . Framingham 100K data replicate the  TCF7L2  association with diabetes.\n",
      "-1 association with\n",
      "-1 diabetes-related traits\n",
      "-1 in\n",
      "-1 Heart\n",
      "-1 to type 2 diabetes\n",
      "-1 Susceptibility to\n",
      "-1 genetic\n",
      "-1 to detect\n",
      "-1 offer\n",
      "-1 a\n",
      "1 fasting plasma glucose | fasting blood glucose measurement http://www.ebi.ac.uk/efo/EFO_0004465\n",
      "-1 , hemoglobin A1c\n",
      "-1 hemoglobin A1c,\n",
      "-1 associations with\n",
      "-1 quantitative glucose\n",
      "-1 glucose traits\n",
      "-1 insulin traits\n",
      "-1 insulin, | insulin measurement http://www.ebi.ac.uk/efo/EFO_0004467\n",
      "-1 resistance,\n",
      "-1 , and\n",
      "1 insulin sensitivity | insulin sensitivity measurement http://www.ebi.ac.uk/efo/EFO_0004471\n",
      "-1 risk for\n",
      "-1 in\n",
      "-1 family\n",
      "-1 genetic\n",
      "-1 fasting\n",
      "-1 and\n",
      "-1 diabetes\n",
      "-1 to test associations\n",
      "-1 trait values,\n",
      "-1 association test\n",
      "-1 -adjusted\n",
      "-1 residual trait\n",
      "-1 , and\n",
      "-1 generalized\n",
      "-1 and\n",
      "-1 of\n",
      "-1 genotypes\n",
      "-1 sex-age-age\n",
      "-1 survival | survival time http://www.ebi.ac.uk/efo/EFO_0000714\n",
      "-1 diabetes\n",
      "-1 six quantitative traits\n",
      "-1 at p\n",
      "-1 of the\n",
      "-1 , and\n",
      "-1 associated with\n",
      "-1 associated\n",
      "-1 in\n",
      "-1 in\n",
      "-1 diabetes\n",
      "-1 Of\n",
      "-1 of\n",
      "-1 a\n",
      "-1 glucose traits\n",
      "-1 insulin traits\n",
      "-1 glucose traits\n",
      "-1 insulin traits\n",
      "-1 of\n",
      "-1 p\n",
      "-1 diabetes\n",
      "-1 AND\n",
      "-1 all\n",
      "-1 all\n",
      "-1 and\n",
      "-1 of\n",
      "-1 , and\n",
      "-1 associated with\n",
      "-1 risk of\n",
      "-1 Of\n",
      "-1 was\n",
      "-1 in\n",
      "-1 moderate\n",
      "-1 LD\n",
      "-1 was\n",
      "-1 diabetes\n",
      "-1 ratio\n",
      "-1 diabetes\n",
      "-1 and\n",
      "-1 no\n",
      "-1 in\n",
      "-1 LD\n",
      "-1 in\n",
      "-1 associated with diabetes\n",
      "-1 related traits\n",
      "-1 was\n",
      "-1 not\n",
      "-1 and related\n",
      "-1 related traits\n",
      "-1 a\n",
      "-1 association\n",
      "-1 of\n",
      "-1 and\n",
      "-1 diabetes\n",
      "-1 association with diabetes\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# let's look at a few examples\n",
    "for doc in corpus.documents[:10]:\n",
    "    doc_id = doc.name\n",
    "    print doc.name\n",
    "    for phen_id2 in gold_agg_dict_phen[doc_id]: # iterate over gwc_id's in doc\n",
    "        print agg_id2phen[phen_id2].name, phen_id2 # print its name and id\n",
    "        for eq_phen in agg_id2phen[phen_id2].equivalents: # loop over equivalent phens\n",
    "            print '\\t', eq_phen.name, eq_phen.ontology_ref, eq_phen.synonyms # print name, ref, synonyms\n",
    "    print ' '.join(s.text for s in id2doc[doc_id].sentences)\n",
    "    for candidate in phen_c:\n",
    "        if candidate.context.document != doc: continue\n",
    "#         print gt_dict[candidate.uid], candidate.get_span(), [agg_id2phen[phen_id].name for phen_id in phen2id.get(change_name(candidate.get_span()),set())]\n",
    "        print gt_dict[c2uid(candidate)], candidate.get_span(), \n",
    "        phen_id = phen2id.get(change_name(candidate.get_span()), None)\n",
    "        if phen_id: print '|', id2phen[phen_id].name, id2phen[phen_id].ontology_ref,\n",
    "        print\n",
    "    print"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create training set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Training set size: 25517\n",
      "Gold set size: 25516\n",
      "Positive labels in training set: 1324\n",
      "Negative labels in training set: 24193\n",
      "Positive labels in gold set: 1423\n",
      "Negative labels in gold set: 24093\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "\n",
    "# Split into train and test set\n",
    "candidates = phen_c\n",
    "training_candidates = []\n",
    "gold_candidates     = []\n",
    "gold_labels         = []\n",
    "n_half = len(candidates)/2\n",
    "for c in candidates[:n_half]:\n",
    "    uid = c2uid(c)\n",
    "    if uid in gt_dict:\n",
    "        gold_candidates.append(c)\n",
    "        gold_labels.append(gt_dict[uid])\n",
    "    else:\n",
    "        training_candidates.append(c)\n",
    "training_candidates.extend(candidates[n_half:])\n",
    "gold_labels = np.array(gold_labels)\n",
    "print \"Training set size: %s\" % len(training_candidates)\n",
    "print \"Gold set size: %s\" % len(gold_candidates)\n",
    "print \"Positive labels in training set: %s\" % len([c for c in training_candidates if gt_dict.get(c2uid(c),0)==1])\n",
    "print \"Negative labels in training set: %s\" % len([c for c in training_candidates if gt_dict.get(c2uid(c),0)==-1])\n",
    "print \"Positive labels in gold set: %s\" % len([c for c in gold_candidates if gt_dict[c2uid(c)]==1])\n",
    "print \"Negative labels in gold set: %s\" % len([c for c in gold_candidates if gt_dict[c2uid(c)]==-1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from nltk.stem import PorterStemmer\n",
    "import re\n",
    "import string\n",
    "punctuation = set(string.punctuation)\n",
    "stemmer = PorterStemmer()\n",
    "\n",
    "# load set of dictionary phenotypes\n",
    "kb = KnowledgeBase()\n",
    "phenotype_list = kb.get_phenotype_candidates() # TODO: load disease names from NCBI\n",
    "phenotype_list = [phenotype for phenotype in phenotype_list]\n",
    "phenotype_set = set(phenotype_list)\n",
    "\n",
    "# load stopwords\n",
    "with open('../data/phenotypes/snorkel/dicts/manual_stopwords.txt') as f:\n",
    "    stopwords = {line.strip() for line in f}\n",
    "stopwords.update(['analysis', 'age', 'drug', 'community', 'detect', 'activity', 'genome',\n",
    "                  'genetic', 'phenotype', 'response', 'population', 'parameter', 'diagnosis',\n",
    "                  'level', 'survival', 'maternal', 'paternal', 'clinical', 'joint', 'related',\n",
    "                  'status', 'risk', 'protein', 'association', 'signal', 'pathway', 'genotype', 'scale',\n",
    "                  'human', 'family', 'heart', 'general', 'chromosome', 'susceptibility', 'select', \n",
    "                  'medical', 'system', 'trait', 'suggest', 'confirm', 'subclinical', 'receptor', \n",
    "                  'class', 'adult', 'affecting'])\n",
    "from nltk.corpus import stopwords as nltk_stopwords\n",
    "stopwords.update(nltk_stopwords.words('english'))\n",
    "stopwords = {stemmer.stem(word) for word in stopwords}\n",
    "\n",
    "# small helpers\n",
    "def get_phenotype(entity, stem=False):\n",
    "    phenotype = entity.get_span()\n",
    "    if stem: phenotype = stemmer.stem(phenotype)\n",
    "    return phenotype.lower()\n",
    "\n",
    "def stem_list(L):\n",
    "    return [stemmer.stem(l.lower()) for l in L]\n",
    "\n",
    "def LF_gt(m):\n",
    "    return gt_dict.get(m.uid, 0) if m in training_candidates else 0\n",
    "\n",
    "# positive LFs\n",
    "def LF_first_sentence(m):\n",
    "    return +10 if m.context.position == 0 else 0\n",
    "def LF_from_regex(m):\n",
    "    if m.context.position == 0 and not regex_phen_matcher._f(m) and not LF_bad_words(m): return +5\n",
    "    else: return 0\n",
    "def LF_with_acronym(m):\n",
    "    post_txt = ''.join(m.post_window('words',d=5))\n",
    "    return +1 if re.search(r'\\([A-Z]{2,4}\\)', post_txt) else 0\n",
    "def LF_many_words(m):\n",
    "    return +1 if len(m.get_span().split()) >= 3 else 0\n",
    "def LF_start_of_sentence(m):\n",
    "    return +1 if m.get_word_start() <= 3 else 0\n",
    "\n",
    "LFs_pos = [LF_first_sentence, LF_with_acronym, LF_from_regex, LF_many_words]\n",
    "\n",
    "# negative LFs\n",
    "def LF_bad_words(m):\n",
    "    bad_words = ['disease', 'single', 'map', 'genetic variation', '( p <']\n",
    "    return -100 if any(m.get_span().lower().startswith(b) for b in bad_words) else 0\n",
    "def LF_short(m):\n",
    "    txt = m.get_attrib_span('words', 3)\n",
    "    return -50 if len(txt) < 5 else 0\n",
    "def LF_no_nouns(m):\n",
    "    return -10 if not any(t.startswith('NN') for t in m.get_attrib_tokens('poses')) else 0\n",
    "def LF_pvalue(m):\n",
    "    txt = m.get_span().lower()\n",
    "    return -100 if 'p <' in txt or 'p =' in txt else 0\n",
    "def LF_stopwords(m):\n",
    "    txt = m.get_span()\n",
    "    txt = ''.join(ch for ch in txt if ch not in punctuation)\n",
    "    words = txt.lower().split()\n",
    "    return -10 if all(word in stopwords for word in words) or \\\n",
    "                  all(stemmer.stem(word) in stopwords for word in words) or \\\n",
    "                  all(change_name(word) in stopwords for word in words) else 0\n",
    "\n",
    "\n",
    "LFs_neg = [LF_bad_words, LF_short, LF_no_nouns, LF_pvalue, LF_stopwords]\n",
    "\n",
    "LFs = LFs_pos + LFs_neg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Applying LFs...\n",
      "Featurizing...\n",
      "Building feature index...\n",
      "Extracting features...\n",
      "0/351112\n",
      "5000/351112\n",
      "10000/351112\n",
      "15000/351112\n",
      "20000/351112\n",
      "25000/351112\n",
      "30000/351112\n",
      "35000/351112\n",
      "40000/351112\n",
      "45000/351112\n",
      "50000/351112\n",
      "55000/351112\n",
      "60000/351112\n",
      "65000/351112\n",
      "70000/351112\n",
      "75000/351112\n",
      "80000/351112\n",
      "85000/351112\n",
      "90000/351112\n",
      "95000/351112\n",
      "100000/351112\n",
      "105000/351112\n",
      "110000/351112\n",
      "115000/351112\n",
      "120000/351112\n",
      "125000/351112\n",
      "130000/351112\n",
      "135000/351112\n",
      "140000/351112\n",
      "145000/351112\n",
      "150000/351112\n",
      "155000/351112\n",
      "160000/351112\n",
      "165000/351112\n",
      "170000/351112\n",
      "175000/351112\n",
      "180000/351112\n",
      "185000/351112\n",
      "190000/351112\n",
      "195000/351112\n",
      "200000/351112\n",
      "205000/351112\n",
      "210000/351112\n",
      "215000/351112\n",
      "220000/351112\n",
      "225000/351112\n"
     ]
    }
   ],
   "source": [
    "from snorkel.snorkel import TrainingSet\n",
    "from snorkel.features import NgramFeaturizer\n",
    "\n",
    "training_set = TrainingSet(training_candidates, LFs, featurizer=NgramFeaturizer())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "lf_stats = training_set.lf_stats()\n",
    "lf_stats[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from snorkel.snorkel import Learner, PipelinedLearner\n",
    "import snorkel.learning\n",
    "from snorkel.learning import LogReg\n",
    "\n",
    "learner = Learner(training_set, model=snorkel.learning.LogReg())\n",
    "\n",
    "# Splitting into CV and test set\n",
    "n_half = len(gold_candidates)/2\n",
    "test_candidates = gold_candidates[:n_half]\n",
    "test_labels     = gold_labels[:n_half]\n",
    "cv_candidates   = gold_candidates[n_half:]\n",
    "cv_labels       = gold_labels[n_half:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from snorkel.learning_utils import GridSearch\n",
    "\n",
    "gs       = GridSearch(learner, ['mu', 'lf_w0'], [[1e-5, 1e-7],[1.0,2.0]])\n",
    "gs_stats = gs.fit(cv_candidates, cv_labels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "gs_stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "learner.test_wmv(test_candidates, test_labels)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save and analyze the results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Analyze / Visualize"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "If a mention occurs in the title, its probably correct, we can take it.\n",
    "\n",
    "Question: what papers did not have any disease mentions in the title?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "preds = learner.predict_wmv(phen_c)\n",
    "results = [c for p, c in zip(preds, phen_c) if p > 0 and c.context.position == 0]\n",
    "doc_set = {c.context.document.name for c in results}\n",
    "missing_docs = {doc.name for doc in corpus.documents} - doc_set\n",
    "docs = sorted(list(missing_docs))\n",
    "print len(docs)\n",
    "for d in missing_docs:\n",
    "    print d, kb.paper_by_pmid(d).title"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's not visualize what we found."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "scores = learner.score_wmv(phen_c)\n",
    "score_dict = { doc.name : list() for doc in corpus.documents }\n",
    "for s, c in zip(scores, phen_c):\n",
    "    score_dict[c.context.document.name].append((s,c))\n",
    "\n",
    "results = dict()\n",
    "for pmid, preds in score_dict.items():\n",
    "    if preds: \n",
    "        best_c = sorted(preds, reverse=True)[0][1]\n",
    "        results[best_c.context.document.name] = best_c\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# doc_set = {c.context.document.name for c in results}\n",
    "# missing_docs = {doc.name for doc in corpus.documents} - doc_set\n",
    "# docs = sorted(list(missing_docs))\n",
    "# print len(docs)\n",
    "for d in corpus.documents:\n",
    "    print d, kb.paper_by_pmid(d.name).title\n",
    "    print unicode(results.get(d.name, None)), LF_stopwords(results.get(d.name))\n",
    "    try:\n",
    "        print sorted(score_dict[d.name], reverse=True)[:5]\n",
    "    except UnicodeEncodeError:\n",
    "        print 'Unicode error'\n",
    "    print"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# print 'waist circumference' in phenotype_list\n",
    "# [ph for ph in phenotype_list if 'sex' in ph]\n",
    "[ph for ph in efo_phenotype_list if 'waist' in ph]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "\n",
    "### Save results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from nltk.corpus import stopwords as nltk_stopwords\n",
    "nltk_stopword_set = set(nltk_stopwords.words('english'))\n",
    "\n",
    "def clean_stopwords(txt):\n",
    "    words = txt.split()\n",
    "    i = 0\n",
    "    new_words = []\n",
    "    while i < len(words):\n",
    "        i += 1\n",
    "        if words[i-1] in nltk_stopword_set: continue\n",
    "        new_words.append(words[i-1])\n",
    "    return ' '.join(new_words)\n",
    "\n",
    "with open('phenotypes.extracted.tsv', 'w') as f:\n",
    "    for d in corpus.documents:\n",
    "        # pick the top two results:\n",
    "        best = sorted(score_dict[d.name], reverse=True)[:3]\n",
    "        # if both are in title, report both, otherwise report only the best one\n",
    "        if len(best) == 3 and best[2][1].context.position == 0 and best[1][0] - best[2][0] < 1:\n",
    "            (_, r1), (_, r2), (_, r3) = best\n",
    "            phen = clean_stopwords(r1.get_span()) + '|' \\\n",
    "                 + clean_stopwords(r2.get_span()) + '|' \\\n",
    "                 + clean_stopwords(r3.get_span())\n",
    "        elif len(best) == 2 and best[1][1].context.position == 0 and best[1][0] > 5:\n",
    "            (_, r1), (_, r2) = best\n",
    "            phen = clean_stopwords(r1.get_span()) + '|' \\\n",
    "                 + clean_stopwords(r2.get_span())\n",
    "        else:\n",
    "            phen = clean_stopwords(best[0][1].get_span())\n",
    "        out_str = u'%s\\t%s\\t\\n' % (d.name, phen)        \n",
    "        f.write(out_str.encode(\"UTF-8\"))\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
